// igniteui-react-core/esm5/lib/AnchoredRadialSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { RadialBaseDescription } from "./RadialBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var AnchoredRadialSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AnchoredRadialSeriesDescription, _super);
    function AnchoredRadialSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.eo = null;
        _this.em = null;
        _this.en = null;
        _this.el = null;
        _this.ek = null;
        _this.ej = null;
        _this.ed = 0;
        _this.eb = null;
        _this.ef = 0;
        _this.eg = 0;
        return _this;
    }
    AnchoredRadialSeriesDescription.prototype.get_type = function () {
        return "AnchoredRadialSeries";
    };
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.eo;
        },
        set: function (a) {
            this.eo = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberAsLegendLabel", {
        get: function () {
            return this.em;
        },
        set: function (a) {
            this.em = a;
            this.g("ValueMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberAsLegendUnit", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            this.en = a;
            this.g("ValueMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineType", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineZIndex", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.g("TrendLineZIndex");
        },
        enumerable: false,
        configurable: true
    });
    AnchoredRadialSeriesDescription.$t = markType(AnchoredRadialSeriesDescription, 'AnchoredRadialSeriesDescription', RadialBaseDescription.$);
    return AnchoredRadialSeriesDescription;
}(RadialBaseDescription));
export { AnchoredRadialSeriesDescription };
