// igniteui-react-core/esm5/lib/AssigningSeriesShapeStyleEventArgsBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AssigningSeriesStyleEventArgsBaseDescription } from "./AssigningSeriesStyleEventArgsBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var AssigningSeriesShapeStyleEventArgsBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AssigningSeriesShapeStyleEventArgsBaseDescription, _super);
    function AssigningSeriesShapeStyleEventArgsBaseDescription() {
        var _this = _super.call(this) || this;
        _this.au = 0;
        _this.aq = null;
        _this.as = 0;
        _this.at = 0;
        return _this;
    }
    AssigningSeriesShapeStyleEventArgsBaseDescription.prototype.get_type = function () {
        return "AssigningSeriesShapeStyleEventArgsBase";
    };
    Object.defineProperty(AssigningSeriesShapeStyleEventArgsBaseDescription.prototype, "strokeThickness", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("StrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesShapeStyleEventArgsBaseDescription.prototype, "strokeDashArray", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("StrokeDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesShapeStyleEventArgsBaseDescription.prototype, "radiusX", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("RadiusX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesShapeStyleEventArgsBaseDescription.prototype, "radiusY", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("RadiusY");
        },
        enumerable: false,
        configurable: true
    });
    AssigningSeriesShapeStyleEventArgsBaseDescription.$t = markType(AssigningSeriesShapeStyleEventArgsBaseDescription, 'AssigningSeriesShapeStyleEventArgsBaseDescription', AssigningSeriesStyleEventArgsBaseDescription.$);
    return AssigningSeriesShapeStyleEventArgsBaseDescription;
}(AssigningSeriesStyleEventArgsBaseDescription));
export { AssigningSeriesShapeStyleEventArgsBaseDescription };
