// igniteui-react-core/esm5/lib/AssigningSeriesStyleEventArgsBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var AssigningSeriesStyleEventArgsBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AssigningSeriesStyleEventArgsBaseDescription, _super);
    function AssigningSeriesStyleEventArgsBaseDescription() {
        var _this = _super.call(this) || this;
        _this.ag = 0;
        _this.af = 0;
        _this.s = new Date();
        _this.r = new Date();
        _this.ak = null;
        _this.aj = null;
        _this.al = null;
        _this.x = 0;
        _this.v = 0;
        _this.h = null;
        _this.w = 0;
        _this.y = 0;
        _this.z = 0;
        _this.k = false;
        _this.j = false;
        _this.l = false;
        _this.m = false;
        return _this;
    }
    AssigningSeriesStyleEventArgsBaseDescription.prototype.get_type = function () {
        return "AssigningSeriesStyleEventArgsBase";
    };
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "startIndex", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("StartIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "endIndex", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("EndIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "startDate", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("StartDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "endDate", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("EndDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "getItemsRef", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("GetItemsRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "fill", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("Fill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "stroke", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("Stroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "opacity", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Opacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "fadeOpacity", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("FadeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "highlightingInfo", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("HighlightingInfo");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "maxAllSeriesHighlightingProgress", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("MaxAllSeriesHighlightingProgress");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "sumAllSeriesHighlightingProgress", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("SumAllSeriesHighlightingProgress");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "totalAllSeriesHighlightingProgress", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("TotalAllSeriesHighlightingProgress");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "highlightingHandled", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("HighlightingHandled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "hasDateRange", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("HasDateRange");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "isNegativeShape", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsNegativeShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AssigningSeriesStyleEventArgsBaseDescription.prototype, "isThumbnail", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("IsThumbnail");
        },
        enumerable: false,
        configurable: true
    });
    AssigningSeriesStyleEventArgsBaseDescription.$t = markType(AssigningSeriesStyleEventArgsBaseDescription, 'AssigningSeriesStyleEventArgsBaseDescription', Description.$);
    return AssigningSeriesStyleEventArgsBaseDescription;
}(Description));
export { AssigningSeriesStyleEventArgsBaseDescription };
