// igniteui-react-core/esm5/lib/CategoryItemHighlightLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var CategoryItemHighlightLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryItemHighlightLayerDescription, _super);
    function CategoryItemHighlightLayerDescription() {
        var _this = _super.call(this) || this;
        _this.ev = null;
        _this.eu = null;
        _this.ek = false;
        _this.ep = null;
        _this.et = null;
        _this.eq = null;
        _this.er = null;
        _this.es = null;
        _this.en = 0;
        _this.ej = false;
        return _this;
    }
    CategoryItemHighlightLayerDescription.prototype.get_type = function () {
        return "CategoryItemHighlightLayer";
    };
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "targetSeriesRef", {
        get: function () {
            return this.ev;
        },
        set: function (a) {
            this.ev = a;
            this.g("TargetSeriesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "targetSeriesName", {
        get: function () {
            return this.eu;
        },
        set: function (a) {
            this.eu = a;
            this.g("TargetSeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "useInterpolation", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("UseInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "highlightType", {
        get: function () {
            return this.ep;
        },
        set: function (a) {
            this.ep = a;
            this.g("HighlightType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "markerType", {
        get: function () {
            return this.et;
        },
        set: function (a) {
            this.et = a;
            this.g("MarkerType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "markerBrush", {
        get: function () {
            return this.eq;
        },
        set: function (a) {
            this.eq = a;
            this.g("MarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "markerOutline", {
        get: function () {
            return this.er;
        },
        set: function (a) {
            this.er = a;
            this.g("MarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "markerTemplateRef", {
        get: function () {
            return this.es;
        },
        set: function (a) {
            this.es = a;
            this.g("MarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "bandHighlightWidth", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            this.en = a;
            this.g("BandHighlightWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "skipUnknownValues", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("SkipUnknownValues");
        },
        enumerable: false,
        configurable: true
    });
    CategoryItemHighlightLayerDescription.$t = markType(CategoryItemHighlightLayerDescription, 'CategoryItemHighlightLayerDescription', AnnotationLayerDescription.$);
    return CategoryItemHighlightLayerDescription;
}(AnnotationLayerDescription));
export { CategoryItemHighlightLayerDescription };
