// igniteui-react-core/esm5/lib/CategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var CategorySeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategorySeriesDescription, _super);
    function CategorySeriesDescription() {
        var _this = _super.call(this) || this;
        _this.dy = false;
        _this.dx = false;
        _this.d0 = false;
        _this.d7 = null;
        _this.dz = false;
        _this.d6 = null;
        _this.d5 = null;
        return _this;
    }
    CategorySeriesDescription.prototype.get_type = function () {
        return "CategorySeries";
    };
    Object.defineProperty(CategorySeriesDescription.prototype, "isCustomCategoryStyleAllowed", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("IsCustomCategoryStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "isCustomCategoryMarkerStyleAllowed", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("IsCustomCategoryMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.g("UseHighMarkerFidelity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "transitionInMode", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("TransitionInMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "isTransitionInEnabled", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("IsTransitionInEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "assigningCategoryStyleRef", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.g("AssigningCategoryStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "assigningCategoryMarkerStyleRef", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("AssigningCategoryMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    CategorySeriesDescription.$t = markType(CategorySeriesDescription, 'CategorySeriesDescription', MarkerSeriesDescription.$);
    return CategorySeriesDescription;
}(MarkerSeriesDescription));
export { CategorySeriesDescription };
