// igniteui-react-core/esm5/lib/DataToolTipLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var DataToolTipLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataToolTipLayerDescription, _super);
    function DataToolTipLayerDescription() {
        var _this = _super.call(this) || this;
        _this.jg = null;
        _this.ex = false;
        _this.ig = null;
        _this.ib = null;
        _this.fr = 0;
        _this.fs = 0;
        _this.fb = 0;
        _this.fc = 0;
        _this.id = null;
        _this.h9 = null;
        _this.ie = null;
        _this.ia = null;
        _this.ew = false;
        _this.en = null;
        _this.ek = null;
        _this.em = null;
        _this.ej = null;
        _this.jr = null;
        _this.h5 = 0;
        _this.h6 = 0;
        _this.jt = null;
        _this.js = null;
        _this.ey = false;
        _this.ju = null;
        _this.eo = null;
        _this.gf = 0;
        _this.gg = 0;
        _this.gh = 0;
        _this.gi = 0;
        _this.ez = false;
        _this.jz = null;
        _this.e0 = false;
        _this.gk = 0;
        _this.gl = 0;
        _this.gm = 0;
        _this.gn = 0;
        _this.jv = null;
        _this.jw = null;
        _this.gj = 0;
        _this.jx = null;
        _this.jy = null;
        _this.ij = null;
        _this.el = null;
        _this.ih = null;
        _this.ii = null;
        _this.ik = null;
        _this.il = null;
        _this.im = null;
        _this.fi = 0;
        _this.fj = 0;
        _this.fk = 0;
        _this.fl = 0;
        _this.fd = 0;
        _this.fe = 0;
        _this.ff = 0;
        _this.fg = 0;
        _this.ev = false;
        _this.io = null;
        _this.fh = 0;
        _this.ip = null;
        _this.iq = null;
        _this.i2 = null;
        _this.i3 = null;
        _this.fy = 0;
        _this.i4 = null;
        _this.i5 = null;
        _this.i6 = null;
        _this.i1 = null;
        _this.fz = 0;
        _this.f0 = 0;
        _this.f1 = 0;
        _this.f2 = 0;
        _this.fu = 0;
        _this.fv = 0;
        _this.fw = 0;
        _this.fx = 0;
        _this.jc = null;
        _this.jd = null;
        _this.f4 = 0;
        _this.je = null;
        _this.jf = null;
        _this.iw = null;
        _this.ix = null;
        _this.iy = null;
        _this.ft = 0;
        _this.iz = null;
        _this.i0 = null;
        _this.i7 = null;
        _this.i8 = null;
        _this.i9 = null;
        _this.f3 = 0;
        _this.ja = null;
        _this.jb = null;
        _this.e7 = 0;
        _this.e8 = 0;
        _this.e9 = 0;
        _this.fa = 0;
        _this.ic = null;
        _this.jl = null;
        _this.jm = null;
        _this.jn = null;
        _this.gb = 0;
        _this.gc = 0;
        _this.gd = 0;
        _this.ge = 0;
        _this.jo = null;
        _this.ga = 0;
        _this.jp = null;
        _this.jq = null;
        _this.f6 = 0;
        _this.f7 = 0;
        _this.f8 = 0;
        _this.f9 = 0;
        _this.jh = null;
        _this.ji = null;
        _this.f5 = 0;
        _this.jj = null;
        _this.jk = null;
        _this.ir = null;
        _this.is = null;
        _this.fn = 0;
        _this.fo = 0;
        _this.fp = 0;
        _this.fq = 0;
        _this.it = null;
        _this.fm = 0;
        _this.iu = null;
        _this.iv = null;
        return _this;
    }
    DataToolTipLayerDescription.prototype.get_type = function () {
        return "DataToolTipLayer";
    };
    Object.defineProperty(DataToolTipLayerDescription.prototype, "targetAxisRef", {
        get: function () {
            return this.jg;
        },
        set: function (a) {
            this.jg = a;
            this.g("TargetAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "useInterpolation", {
        get: function () {
            return this.ex;
        },
        set: function (a) {
            this.ex = a;
            this.g("UseInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupingMode", {
        get: function () {
            return this.ig;
        },
        set: function (a) {
            this.ig = a;
            this.g("GroupingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "actualGroupingMode", {
        get: function () {
            return this.ib;
        },
        set: function (a) {
            this.ib = a;
            this.g("ActualGroupingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "positionOffsetX", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("PositionOffsetX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "positionOffsetY", {
        get: function () {
            return this.fs;
        },
        set: function (a) {
            this.fs = a;
            this.g("PositionOffsetY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "defaultPositionOffsetX", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("DefaultPositionOffsetX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "defaultPositionOffsetY", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("DefaultPositionOffsetY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupedPositionModeX", {
        get: function () {
            return this.id;
        },
        set: function (a) {
            this.id = a;
            this.g("GroupedPositionModeX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "actualGroupedPositionModeX", {
        get: function () {
            return this.h9;
        },
        set: function (a) {
            this.h9 = a;
            this.g("ActualGroupedPositionModeX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupedPositionModeY", {
        get: function () {
            return this.ie;
        },
        set: function (a) {
            this.ie = a;
            this.g("GroupedPositionModeY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "actualGroupedPositionModeY", {
        get: function () {
            return this.ia;
        },
        set: function (a) {
            this.ia = a;
            this.g("ActualGroupedPositionModeY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "shouldUpdateWhenSeriesDataChanges", {
        get: function () {
            return this.ew;
        },
        set: function (a) {
            this.ew = a;
            this.g("ShouldUpdateWhenSeriesDataChanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "includedSeries", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            this.en = a;
            this.g("IncludedSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "excludedSeries", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("ExcludedSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "includedColumns", {
        get: function () {
            return this.em;
        },
        set: function (a) {
            this.em = a;
            this.g("IncludedColumns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "excludedColumns", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("ExcludedColumns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatAbbreviation", {
        get: function () {
            return this.jr;
        },
        set: function (a) {
            this.jr = a;
            this.g("ValueFormatAbbreviation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatMaxFractions", {
        get: function () {
            return this.h5;
        },
        set: function (a) {
            this.h5 = a;
            this.g("ValueFormatMaxFractions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatMinFractions", {
        get: function () {
            return this.h6;
        },
        set: function (a) {
            this.h6 = a;
            this.g("ValueFormatMinFractions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatMode", {
        get: function () {
            return this.jt;
        },
        set: function (a) {
            this.jt = a;
            this.g("ValueFormatMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatCulture", {
        get: function () {
            return this.js;
        },
        set: function (a) {
            this.js = a;
            this.g("ValueFormatCulture");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatUseGrouping", {
        get: function () {
            return this.ey;
        },
        set: function (a) {
            this.ey = a;
            this.g("ValueFormatUseGrouping");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatString", {
        get: function () {
            return this.ju;
        },
        set: function (a) {
            this.ju = a;
            this.g("ValueFormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatSpecifiers", {
        get: function () {
            return this.eo;
        },
        set: function (a) {
            this.eo = a;
            this.g("ValueFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowMarginBottom", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("ValueRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowMarginLeft", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("ValueRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowMarginRight", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("ValueRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowMarginTop", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("ValueRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowVisible", {
        get: function () {
            return this.ez;
        },
        set: function (a) {
            this.ez = a;
            this.g("ValueRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextWhenMissingData", {
        get: function () {
            return this.jz;
        },
        set: function (a) {
            this.jz = a;
            this.g("ValueTextWhenMissingData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextUseSeriesColors", {
        get: function () {
            return this.e0;
        },
        set: function (a) {
            this.e0 = a;
            this.g("ValueTextUseSeriesColors");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextMarginBottom", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("ValueTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextMarginLeft", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("ValueTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextMarginRight", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("ValueTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextMarginTop", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("ValueTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextColor", {
        get: function () {
            return this.jv;
        },
        set: function (a) {
            this.jv = a;
            this.g("ValueTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextFontFamily", {
        get: function () {
            return this.jw;
        },
        set: function (a) {
            this.jw = a;
            this.g("ValueTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextFontSize", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("ValueTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextFontStyle", {
        get: function () {
            return this.jx;
        },
        set: function (a) {
            this.jx = a;
            this.g("ValueTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextFontWeight", {
        get: function () {
            return this.jy;
        },
        set: function (a) {
            this.jy = a;
            this.g("ValueTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatString", {
        get: function () {
            return this.ij;
        },
        set: function (a) {
            this.ij = a;
            this.g("HeaderFormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatSpecifiers", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("HeaderFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatCulture", {
        get: function () {
            return this.ih;
        },
        set: function (a) {
            this.ih = a;
            this.g("HeaderFormatCulture");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatDate", {
        get: function () {
            return this.ii;
        },
        set: function (a) {
            this.ii = a;
            this.g("HeaderFormatDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatTime", {
        get: function () {
            return this.ik;
        },
        set: function (a) {
            this.ik = a;
            this.g("HeaderFormatTime");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerText", {
        get: function () {
            return this.il;
        },
        set: function (a) {
            this.il = a;
            this.g("HeaderText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextColor", {
        get: function () {
            return this.im;
        },
        set: function (a) {
            this.im = a;
            this.g("HeaderTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextMarginBottom", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("HeaderTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextMarginLeft", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("HeaderTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextMarginRight", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("HeaderTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextMarginTop", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("HeaderTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowMarginBottom", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("HeaderRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowMarginLeft", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("HeaderRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowMarginRight", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("HeaderRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowMarginTop", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("HeaderRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowVisible", {
        get: function () {
            return this.ev;
        },
        set: function (a) {
            this.ev = a;
            this.g("HeaderRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextFontFamily", {
        get: function () {
            return this.io;
        },
        set: function (a) {
            this.io = a;
            this.g("HeaderTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextFontSize", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("HeaderTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextFontStyle", {
        get: function () {
            return this.ip;
        },
        set: function (a) {
            this.ip = a;
            this.g("HeaderTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextFontWeight", {
        get: function () {
            return this.iq;
        },
        set: function (a) {
            this.iq = a;
            this.g("HeaderTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextColor", {
        get: function () {
            return this.i2;
        },
        set: function (a) {
            this.i2 = a;
            this.g("SummaryTitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextFontFamily", {
        get: function () {
            return this.i3;
        },
        set: function (a) {
            this.i3 = a;
            this.g("SummaryTitleTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextFontSize", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("SummaryTitleTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextFontStyle", {
        get: function () {
            return this.i4;
        },
        set: function (a) {
            this.i4 = a;
            this.g("SummaryTitleTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextFontWeight", {
        get: function () {
            return this.i5;
        },
        set: function (a) {
            this.i5 = a;
            this.g("SummaryTitleTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryType", {
        get: function () {
            return this.i6;
        },
        set: function (a) {
            this.i6 = a;
            this.g("SummaryType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleText", {
        get: function () {
            return this.i1;
        },
        set: function (a) {
            this.i1 = a;
            this.g("SummaryTitleText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextMarginBottom", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("SummaryTitleTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextMarginLeft", {
        get: function () {
            return this.f0;
        },
        set: function (a) {
            this.f0 = a;
            this.g("SummaryTitleTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextMarginRight", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("SummaryTitleTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextMarginTop", {
        get: function () {
            return this.f2;
        },
        set: function (a) {
            this.f2 = a;
            this.g("SummaryTitleTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryRowMarginBottom", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("SummaryRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryRowMarginLeft", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("SummaryRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryRowMarginRight", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            this.fw = a;
            this.g("SummaryRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryRowMarginTop", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("SummaryRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextColor", {
        get: function () {
            return this.jc;
        },
        set: function (a) {
            this.jc = a;
            this.g("SummaryValueTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextFontFamily", {
        get: function () {
            return this.jd;
        },
        set: function (a) {
            this.jd = a;
            this.g("SummaryValueTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextFontSize", {
        get: function () {
            return this.f4;
        },
        set: function (a) {
            this.f4 = a;
            this.g("SummaryValueTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextFontStyle", {
        get: function () {
            return this.je;
        },
        set: function (a) {
            this.je = a;
            this.g("SummaryValueTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextFontWeight", {
        get: function () {
            return this.jf;
        },
        set: function (a) {
            this.jf = a;
            this.g("SummaryValueTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelText", {
        get: function () {
            return this.iw;
        },
        set: function (a) {
            this.iw = a;
            this.g("SummaryLabelText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextColor", {
        get: function () {
            return this.ix;
        },
        set: function (a) {
            this.ix = a;
            this.g("SummaryLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextFontFamily", {
        get: function () {
            return this.iy;
        },
        set: function (a) {
            this.iy = a;
            this.g("SummaryLabelTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextFontSize", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("SummaryLabelTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextFontStyle", {
        get: function () {
            return this.iz;
        },
        set: function (a) {
            this.iz = a;
            this.g("SummaryLabelTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextFontWeight", {
        get: function () {
            return this.i0;
        },
        set: function (a) {
            this.i0 = a;
            this.g("SummaryLabelTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsText", {
        get: function () {
            return this.i7;
        },
        set: function (a) {
            this.i7 = a;
            this.g("SummaryUnitsText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextColor", {
        get: function () {
            return this.i8;
        },
        set: function (a) {
            this.i8 = a;
            this.g("SummaryUnitsTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextFontFamily", {
        get: function () {
            return this.i9;
        },
        set: function (a) {
            this.i9 = a;
            this.g("SummaryUnitsTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextFontSize", {
        get: function () {
            return this.f3;
        },
        set: function (a) {
            this.f3 = a;
            this.g("SummaryUnitsTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextFontStyle", {
        get: function () {
            return this.ja;
        },
        set: function (a) {
            this.ja = a;
            this.g("SummaryUnitsTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextFontWeight", {
        get: function () {
            return this.jb;
        },
        set: function (a) {
            this.jb = a;
            this.g("SummaryUnitsTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeMarginBottom", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("BadgeMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeMarginLeft", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("BadgeMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeMarginRight", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("BadgeMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeMarginTop", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("BadgeMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeShape", {
        get: function () {
            return this.ic;
        },
        set: function (a) {
            this.ic = a;
            this.g("BadgeShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsDisplayMode", {
        get: function () {
            return this.jl;
        },
        set: function (a) {
            this.jl = a;
            this.g("UnitsDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsText", {
        get: function () {
            return this.jm;
        },
        set: function (a) {
            this.jm = a;
            this.g("UnitsText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextColor", {
        get: function () {
            return this.jn;
        },
        set: function (a) {
            this.jn = a;
            this.g("UnitsTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextMarginBottom", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("UnitsTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextMarginLeft", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("UnitsTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextMarginRight", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("UnitsTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextMarginTop", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("UnitsTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextFontFamily", {
        get: function () {
            return this.jo;
        },
        set: function (a) {
            this.jo = a;
            this.g("UnitsTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextFontSize", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("UnitsTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextFontStyle", {
        get: function () {
            return this.jp;
        },
        set: function (a) {
            this.jp = a;
            this.g("UnitsTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextFontWeight", {
        get: function () {
            return this.jq;
        },
        set: function (a) {
            this.jq = a;
            this.g("UnitsTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextMarginBottom", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("TitleTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextMarginLeft", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("TitleTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextMarginRight", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("TitleTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextMarginTop", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("TitleTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextColor", {
        get: function () {
            return this.jh;
        },
        set: function (a) {
            this.jh = a;
            this.g("TitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextFontFamily", {
        get: function () {
            return this.ji;
        },
        set: function (a) {
            this.ji = a;
            this.g("TitleTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextFontSize", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("TitleTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextFontStyle", {
        get: function () {
            return this.jj;
        },
        set: function (a) {
            this.jj = a;
            this.g("TitleTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextFontWeight", {
        get: function () {
            return this.jk;
        },
        set: function (a) {
            this.jk = a;
            this.g("TitleTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelDisplayMode", {
        get: function () {
            return this.ir;
        },
        set: function (a) {
            this.ir = a;
            this.g("LabelDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextColor", {
        get: function () {
            return this.is;
        },
        set: function (a) {
            this.is = a;
            this.g("LabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextMarginBottom", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("LabelTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextMarginLeft", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("LabelTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextMarginRight", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("LabelTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextMarginTop", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("LabelTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextFontFamily", {
        get: function () {
            return this.it;
        },
        set: function (a) {
            this.it = a;
            this.g("LabelTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextFontSize", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("LabelTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextFontStyle", {
        get: function () {
            return this.iu;
        },
        set: function (a) {
            this.iu = a;
            this.g("LabelTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextFontWeight", {
        get: function () {
            return this.iv;
        },
        set: function (a) {
            this.iv = a;
            this.g("LabelTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    DataToolTipLayerDescription.$t = markType(DataToolTipLayerDescription, 'DataToolTipLayerDescription', AnnotationLayerDescription.$);
    return DataToolTipLayerDescription;
}(AnnotationLayerDescription));
export { DataToolTipLayerDescription };
