// igniteui-react-core/esm5/lib/DatePickerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DatePickerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DatePickerDescription, _super);
    function DatePickerDescription() {
        var _this = _super.call(this) || this;
        _this.w = new Date();
        _this.v = new Date();
        _this.ap = null;
        _this.at = null;
        _this.aq = null;
        _this.ab = 0;
        _this.ar = null;
        _this.as = null;
        _this.av = null;
        _this.u = new Date();
        _this.t = new Date();
        _this.ai = null;
        _this.af = null;
        _this.ay = null;
        _this.ac = 0;
        _this.az = null;
        _this.a0 = null;
        _this.ax = null;
        _this.an = null;
        _this.k = false;
        _this.l = false;
        _this.h = false;
        _this.j = false;
        _this.aj = null;
        _this.ak = null;
        _this.m = false;
        _this.ah = null;
        _this.al = null;
        _this.i = false;
        _this.aw = null;
        _this.am = null;
        _this.au = null;
        _this.ao = null;
        _this.ag = null;
        _this.a1 = null;
        return _this;
    }
    DatePickerDescription.prototype.get_type = function () {
        return "DatePicker";
    };
    Object.defineProperty(DatePickerDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "value", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "today", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Today");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "label", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelTextColor", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("LabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelFontFamily", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("LabelFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelFontSize", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("LabelFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelFontStyle", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("LabelFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelFontWeight", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("LabelFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "placeholder", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("Placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "minDate", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("MinDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "maxDate", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("MaxDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "density", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "baseTheme", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textFontFamily", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("TextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textFontSize", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("TextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textFontStyle", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("TextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textFontWeight", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("TextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textColor", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "iconColor", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("IconColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "showClearButton", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("ShowClearButton");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "showTodayButton", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ShowTodayButton");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "allowTextInput", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("AllowTextInput");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "openOnFocus", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("OpenOnFocus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "firstDayOfWeek", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("FirstDayOfWeek");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "firstWeekOfYear", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("FirstWeekOfYear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "showWeekNumbers", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("ShowWeekNumbers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "dateFormat", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("DateFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "formatString", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("FormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "isDisabled", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("IsDisabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "selectedValueChangedRef", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("SelectedValueChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "gotFocusRef", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("GotFocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "lostFocusRef", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("LostFocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "keyDownRef", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("KeyDownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "changingRef", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("ChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "valueChangeRef", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("ValueChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    DatePickerDescription.$t = markType(DatePickerDescription, 'DatePickerDescription', Description.$);
    return DatePickerDescription;
}(Description));
export { DatePickerDescription };
