// igniteui-react-core/esm5/lib/FinancialPriceSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { FinancialSeriesDescription } from "./FinancialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var FinancialPriceSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialPriceSeriesDescription, _super);
    function FinancialPriceSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.en = null;
        _this.em = null;
        _this.eb = null;
        _this.d7 = 0;
        _this.d5 = null;
        _this.d9 = 0;
        _this.ee = null;
        _this.ej = null;
        _this.eh = null;
        _this.ef = null;
        _this.ec = null;
        _this.ek = null;
        _this.ei = null;
        _this.eg = null;
        _this.ed = null;
        _this.el = null;
        return _this;
    }
    FinancialPriceSeriesDescription.prototype.get_type = function () {
        return "FinancialPriceSeries";
    };
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLineType", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            this.en = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.em;
        },
        set: function (a) {
            this.em = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "displayType", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.g("DisplayType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "negativeOutline", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("NegativeOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "lowMemberAsLegendLabel", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.g("LowMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "highMemberAsLegendLabel", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("HighMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "closeMemberAsLegendLabel", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.g("CloseMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "openMemberAsLegendLabel", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("OpenMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "lowMemberAsLegendUnit", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            this.ei = a;
            this.g("LowMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "highMemberAsLegendUnit", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.g("HighMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "closeMemberAsLegendUnit", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("CloseMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "openMemberAsLegendUnit", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("OpenMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    FinancialPriceSeriesDescription.$t = markType(FinancialPriceSeriesDescription, 'FinancialPriceSeriesDescription', FinancialSeriesDescription.$);
    return FinancialPriceSeriesDescription;
}(FinancialSeriesDescription));
export { FinancialPriceSeriesDescription };
