// igniteui-react-core/esm5/lib/FinancialIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { FinancialSeriesDescription } from "./FinancialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var FinancialIndicatorDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialIndicatorDescription, _super);
    function FinancialIndicatorDescription() {
        var _this = _super.call(this) || this;
        _this.ee = null;
        _this.d9 = 0;
        _this.eg = null;
        _this.ef = null;
        _this.ed = null;
        _this.d7 = 0;
        _this.d5 = null;
        _this.ea = 0;
        return _this;
    }
    FinancialIndicatorDescription.prototype.get_type = function () {
        return "FinancialIndicator";
    };
    Object.defineProperty(FinancialIndicatorDescription.prototype, "displayType", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.g("DisplayType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "ignoreFirst", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("IgnoreFirst");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLineType", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    FinancialIndicatorDescription.$t = markType(FinancialIndicatorDescription, 'FinancialIndicatorDescription', FinancialSeriesDescription.$);
    return FinancialIndicatorDescription;
}(FinancialSeriesDescription));
export { FinancialIndicatorDescription };
