// igniteui-react-core/esm5/lib/GeographicHighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicHighDensityScatterSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicHighDensityScatterSeriesDescription, _super);
    function GeographicHighDensityScatterSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.dv = null;
        _this.dw = null;
        _this.dh = false;
        _this.dg = false;
        _this.dl = 0;
        _this.dk = 0;
        _this.du = null;
        _this.dt = null;
        _this.dp = 0;
        _this.dq = 0;
        _this.dx = null;
        return _this;
    }
    GeographicHighDensityScatterSeriesDescription.prototype.get_type = function () {
        return "GeographicHighDensityScatterSeries";
    };
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "latitudeMemberPath", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("LatitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "longitudeMemberPath", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("LongitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "useBruteForce", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            this.dh = a;
            this.g("UseBruteForce");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveLoad", {
        get: function () {
            return this.dg;
        },
        set: function (a) {
            this.dg = a;
            this.g("ProgressiveLoad");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMinimum", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("HeatMinimum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMaximum", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("HeatMaximum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMinimumColor", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("HeatMinimumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMaximumColor", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("HeatMaximumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "pointExtent", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("PointExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveStatus", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("ProgressiveStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveLoadStatusChangedRef", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("ProgressiveLoadStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicHighDensityScatterSeriesDescription.$t = markType(GeographicHighDensityScatterSeriesDescription, 'GeographicHighDensityScatterSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicHighDensityScatterSeriesDescription;
}(GeographicMapSeriesHostDescription));
export { GeographicHighDensityScatterSeriesDescription };
