// igniteui-react-core/esm5/lib/GeographicProportionalSymbolSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMarkerSeriesDescription } from "./GeographicMarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicProportionalSymbolSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicProportionalSymbolSeriesDescription, _super);
    function GeographicProportionalSymbolSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.es = null;
        _this.et = null;
        _this.eu = null;
        _this.d5 = null;
        _this.er = null;
        _this.ep = null;
        _this.d3 = null;
        _this.d7 = false;
        _this.ea = false;
        _this.ei = 0;
        _this.eh = 0;
        _this.eb = false;
        _this.d9 = false;
        _this.d8 = false;
        _this.eq = null;
        _this.el = 0;
        _this.eo = null;
        _this.en = null;
        return _this;
    }
    GeographicProportionalSymbolSeriesDescription.prototype.get_type = function () {
        return "GeographicProportionalSymbolSeries";
    };
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "latitudeMemberPath", {
        get: function () {
            return this.es;
        },
        set: function (a) {
            this.es = a;
            this.g("LatitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "longitudeMemberPath", {
        get: function () {
            return this.et;
        },
        set: function (a) {
            this.et = a;
            this.g("LongitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "radiusMemberPath", {
        get: function () {
            return this.eu;
        },
        set: function (a) {
            this.eu = a;
            this.g("RadiusMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "radiusScale", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("RadiusScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "labelMemberPath", {
        get: function () {
            return this.er;
        },
        set: function (a) {
            this.er = a;
            this.g("LabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "fillMemberPath", {
        get: function () {
            return this.ep;
        },
        set: function (a) {
            this.ep = a;
            this.g("FillMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "fillScale", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("FillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "fillScaleUseGlobalValues", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("FillScaleUseGlobalValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "markerOutlineUsesFillScale", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("MarkerOutlineUsesFillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "markerOutlineBrightness", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            this.ei = a;
            this.g("MarkerOutlineBrightness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "markerBrushBrightness", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.g("MarkerBrushBrightness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "radiusScaleUseGlobalValues", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("RadiusScaleUseGlobalValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "isCustomScatterStyleAllowed", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("IsCustomScatterStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "isCustomScatterMarkerStyleAllowed", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.g("IsCustomScatterMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.eq;
        },
        set: function (a) {
            this.eq = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "assigningScatterStyleRef", {
        get: function () {
            return this.eo;
        },
        set: function (a) {
            this.eo = a;
            this.g("AssigningScatterStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "assigningScatterMarkerStyleRef", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            this.en = a;
            this.g("AssigningScatterMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicProportionalSymbolSeriesDescription.$t = markType(GeographicProportionalSymbolSeriesDescription, 'GeographicProportionalSymbolSeriesDescription', GeographicMarkerSeriesDescription.$);
    return GeographicProportionalSymbolSeriesDescription;
}(GeographicMarkerSeriesDescription));
export { GeographicProportionalSymbolSeriesDescription };
