// igniteui-react-core/esm5/lib/GeographicShapeSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicShapeSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicShapeSeriesDescription, _super);
    function GeographicShapeSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.ej = null;
        _this.eh = null;
        _this.ei = null;
        _this.d3 = 0;
        _this.d2 = 0;
        _this.eg = null;
        _this.ee = null;
        _this.ec = null;
        _this.ef = null;
        _this.ea = null;
        _this.ed = null;
        _this.d1 = 0;
        _this.eb = null;
        _this.d9 = null;
        _this.d7 = null;
        _this.d8 = null;
        _this.ek = null;
        return _this;
    }
    GeographicShapeSeriesDescription.prototype.get_type = function () {
        return "GeographicShapeSeries";
    };
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "shapeStyleSelectorRef", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("ShapeStyleSelectorRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "shapeFill", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.g("ShapeFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "shapeStroke", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            this.ei = a;
            this.g("ShapeStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "shapeStrokeThickness", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("ShapeStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "shapeOpacity", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.g("ShapeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "markerType", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.g("MarkerType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "markerOutlineMode", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.g("MarkerOutlineMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "markerFillMode", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.g("MarkerFillMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "markerTemplateRef", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("MarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "markerBrush", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("MarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "markerOutline", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("MarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "markerThickness", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.g("MarkerThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "markerCollisionAvoidance", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("MarkerCollisionAvoidance");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "actualMarkerTemplateRef", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("ActualMarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "actualMarkerBrush", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("ActualMarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "actualMarkerOutline", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.g("ActualMarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesDescription.prototype, "styleShapeRef", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("StyleShapeRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicShapeSeriesDescription.$t = markType(GeographicShapeSeriesDescription, 'GeographicShapeSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicShapeSeriesDescription;
}(GeographicShapeSeriesBaseDescription));
export { GeographicShapeSeriesDescription };
