// igniteui-react-core/esm5/lib/HighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var HighDensityScatterSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HighDensityScatterSeriesDescription, _super);
    function HighDensityScatterSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.ds = null;
        _this.dw = null;
        _this.dv = null;
        _this.dt = null;
        _this.dx = null;
        _this.du = null;
        _this.dy = null;
        _this.dz = null;
        _this.dd = false;
        _this.dc = false;
        _this.dh = 0;
        _this.dg = 0;
        _this.dq = null;
        _this.dp = null;
        _this.dk = 0;
        _this.dl = 0;
        _this.dr = null;
        return _this;
    }
    HighDensityScatterSeriesDescription.prototype.get_type = function () {
        return "HighDensityScatterSeries";
    };
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xAxisRef", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yAxisRef", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberPath", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("XMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberAsLegendLabel", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("XMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberAsLegendLabel", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("YMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberAsLegendUnit", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("XMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberAsLegendUnit", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("YMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberPath", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("YMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "useBruteForce", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            this.dd = a;
            this.g("UseBruteForce");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveLoad", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("ProgressiveLoad");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMinimum", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            this.dh = a;
            this.g("HeatMinimum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMaximum", {
        get: function () {
            return this.dg;
        },
        set: function (a) {
            this.dg = a;
            this.g("HeatMaximum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMinimumColor", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("HeatMinimumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMaximumColor", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("HeatMaximumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "pointExtent", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("PointExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveStatus", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("ProgressiveStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveLoadStatusChangedRef", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("ProgressiveLoadStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    HighDensityScatterSeriesDescription.$t = markType(HighDensityScatterSeriesDescription, 'HighDensityScatterSeriesDescription', SeriesDescription.$);
    return HighDensityScatterSeriesDescription;
}(SeriesDescription));
export { HighDensityScatterSeriesDescription };
