// igniteui-react-core/esm5/lib/HighlightingInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var HighlightingInfoDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HighlightingInfoDescription, _super);
    function HighlightingInfoDescription() {
        var _this = _super.call(this) || this;
        _this.s = 0;
        _this.r = 0;
        _this.v = null;
        _this.p = 0;
        _this.l = false;
        _this.k = false;
        _this.j = false;
        _this.h = null;
        return _this;
    }
    HighlightingInfoDescription.prototype.get_type = function () {
        return "HighlightingInfo";
    };
    Object.defineProperty(HighlightingInfoDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighlightingInfoDescription.prototype, "startIndex", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("StartIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighlightingInfoDescription.prototype, "endIndex", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("EndIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighlightingInfoDescription.prototype, "state", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("State");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighlightingInfoDescription.prototype, "progress", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Progress");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighlightingInfoDescription.prototype, "isMarker", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsMarker");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighlightingInfoDescription.prototype, "isFullRange", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("IsFullRange");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighlightingInfoDescription.prototype, "isExclusive", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("IsExclusive");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighlightingInfoDescription.prototype, "series", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    HighlightingInfoDescription.$t = markType(HighlightingInfoDescription, 'HighlightingInfoDescription', Description.$);
    return HighlightingInfoDescription;
}(Description));
export { HighlightingInfoDescription };
