// igniteui-react-core/esm5/lib/PolarBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var PolarBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarBaseDescription, _super);
    function PolarBaseDescription() {
        var _this = _super.call(this) || this;
        _this.em = null;
        _this.et = null;
        _this.ej = null;
        _this.eq = null;
        _this.d2 = false;
        _this.ea = 0;
        _this.ev = null;
        _this.eu = null;
        _this.ei = null;
        _this.d7 = 0;
        _this.dx = null;
        _this.eb = 0;
        _this.ec = 0;
        _this.dz = false;
        _this.eh = null;
        _this.ep = null;
        _this.d9 = 0;
        _this.d1 = false;
        _this.d0 = false;
        _this.er = null;
        _this.ek = null;
        _this.es = null;
        _this.el = null;
        _this.eo = null;
        _this.en = null;
        return _this;
    }
    PolarBaseDescription.prototype.get_type = function () {
        return "PolarBase";
    };
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberPath", {
        get: function () {
            return this.em;
        },
        set: function (a) {
            this.em = a;
            this.g("AngleMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberPath", {
        get: function () {
            return this.et;
        },
        set: function (a) {
            this.et = a;
            this.g("RadiusMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleAxisRef", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("AngleAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusAxisRef", {
        get: function () {
            return this.eq;
        },
        set: function (a) {
            this.eq = a;
            this.g("RadiusAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "useCartesianInterpolation", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.g("UseCartesianInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "maximumMarkers", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("MaximumMarkers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineType", {
        get: function () {
            return this.ev;
        },
        set: function (a) {
            this.ev = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.eu;
        },
        set: function (a) {
            this.eu = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            this.ei = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineZIndex", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.g("TrendLineZIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "clipSeriesToBounds", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("ClipSeriesToBounds");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "actualItemSearchMode", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.g("ActualItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.ep;
        },
        set: function (a) {
            this.ep = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "isCustomPolarStyleAllowed", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.g("IsCustomPolarStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "isCustomPolarMarkerStyleAllowed", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.g("IsCustomPolarMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberAsLegendLabel", {
        get: function () {
            return this.er;
        },
        set: function (a) {
            this.er = a;
            this.g("RadiusMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberAsLegendLabel", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("AngleMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberAsLegendUnit", {
        get: function () {
            return this.es;
        },
        set: function (a) {
            this.es = a;
            this.g("RadiusMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberAsLegendUnit", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("AngleMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "assigningPolarStyleRef", {
        get: function () {
            return this.eo;
        },
        set: function (a) {
            this.eo = a;
            this.g("AssigningPolarStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "assigningPolarMarkerStyleRef", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            this.en = a;
            this.g("AssigningPolarMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    PolarBaseDescription.$t = markType(PolarBaseDescription, 'PolarBaseDescription', MarkerSeriesDescription.$);
    return PolarBaseDescription;
}(MarkerSeriesDescription));
export { PolarBaseDescription };
