// igniteui-react-core/esm5/lib/RadialGaugeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var RadialGaugeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialGaugeDescription, _super);
    function RadialGaugeDescription() {
        var _this = _super.call(this) || this;
        _this.cv = null;
        _this.ad = 0;
        _this.cx = null;
        _this.cw = null;
        _this.de = null;
        _this.cz = null;
        _this.co = null;
        _this.i = null;
        _this.j = null;
        _this.ai = 0;
        _this.u = 0;
        _this.ah = 0;
        _this.t = 0;
        _this.ae = 0;
        _this.ab = 0;
        _this.ac = 0;
        _this.ba = 0;
        _this.a4 = 0;
        _this.a1 = 0;
        _this.da = null;
        _this.cj = 0;
        _this.dc = null;
        _this.c2 = null;
        _this.c3 = null;
        _this.aw = 0;
        _this.ap = 0;
        _this.c7 = null;
        _this.ax = 0;
        _this.aq = 0;
        _this.ao = 0;
        _this.an = 0;
        _this.av = 0;
        _this.au = 0;
        _this.at = 0;
        _this.ar = 0;
        _this.c6 = null;
        _this.a5 = 0;
        _this.c4 = null;
        _this.c5 = null;
        _this.ay = 0;
        _this.as = 0;
        _this.a2 = 0;
        _this.af = 0;
        _this.ag = 0;
        _this.a7 = 0;
        _this.a6 = 0;
        _this.a8 = 0;
        _this.db = null;
        _this.cu = null;
        _this.al = 0;
        _this.ak = 0;
        _this.am = 0;
        _this.c1 = null;
        _this.aj = 0;
        _this.c8 = null;
        _this.cp = null;
        _this.cq = null;
        _this.aa = 0;
        _this.y = 0;
        _this.z = 0;
        _this.a3 = 0;
        _this.c9 = null;
        _this.w = 0;
        _this.x = 0;
        _this.cr = null;
        _this.k = null;
        _this.a0 = 0;
        _this.cs = null;
        _this.q = false;
        _this.p = false;
        _this.ct = null;
        _this.c0 = null;
        _this.h = null;
        _this.a9 = 0;
        _this.az = 0;
        _this.v = 0;
        _this.cy = null;
        _this.cn = null;
        _this.dd = null;
        _this.cm = null;
        _this.cl = null;
        return _this;
    }
    Object.defineProperty(RadialGaugeDescription.prototype, "fontFamily", {
        get: function () {
            return this.cv;
        },
        set: function (a) {
            this.cv = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "fontSize", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "fontWeight", {
        get: function () {
            return this.cx;
        },
        set: function (a) {
            this.cx = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "fontStyle", {
        get: function () {
            return this.cw;
        },
        set: function (a) {
            this.cw = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "width", {
        get: function () {
            return this.de;
        },
        set: function (a) {
            this.de = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "height", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "background", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    RadialGaugeDescription.prototype.get_type = function () {
        return "RadialGauge";
    };
    Object.defineProperty(RadialGaugeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "rangeBrushes", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("RangeBrushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "rangeOutlines", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("RangeOutlines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "minimumValue", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("MinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "actualMinimumValue", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("ActualMinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "maximumValue", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("MaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "actualMaximumValue", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ActualMaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "interval", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Interval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "centerX", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("CenterX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "centerY", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("CenterY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "value", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "scaleStartAngle", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("ScaleStartAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "scaleEndAngle", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("ScaleEndAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "scaleSweepDirection", {
        get: function () {
            return this.da;
        },
        set: function (a) {
            this.da = a;
            this.g("ScaleSweepDirection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "transitionDuration", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.g("TransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "transitionEasingFunctionRef", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("TransitionEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleBrush", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("NeedleBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleOutline", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.g("NeedleOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleStartExtent", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("NeedleStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleEndExtent", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("NeedleEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleShape", {
        get: function () {
            return this.c7;
        },
        set: function (a) {
            this.c7 = a;
            this.g("NeedleShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleStartWidthRatio", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("NeedleStartWidthRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleEndWidthRatio", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("NeedleEndWidthRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleBaseFeatureWidthRatio", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("NeedleBaseFeatureWidthRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleBaseFeatureExtent", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("NeedleBaseFeatureExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needlePointFeatureWidthRatio", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("NeedlePointFeatureWidthRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needlePointFeatureExtent", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("NeedlePointFeatureExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needlePivotWidthRatio", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("NeedlePivotWidthRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needlePivotInnerWidthRatio", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("NeedlePivotInnerWidthRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needlePivotShape", {
        get: function () {
            return this.c6;
        },
        set: function (a) {
            this.c6 = a;
            this.g("NeedlePivotShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "scaleStartExtent", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("ScaleStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needlePivotBrush", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.g("NeedlePivotBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needlePivotOutline", {
        get: function () {
            return this.c5;
        },
        set: function (a) {
            this.c5 = a;
            this.g("NeedlePivotOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needleStrokeThickness", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("NeedleStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "needlePivotStrokeThickness", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("NeedlePivotStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "scaleEndExtent", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("ScaleEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "labelExtent", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("LabelExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "labelInterval", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("LabelInterval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "tickStartExtent", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("TickStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "tickEndExtent", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("TickEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "tickStrokeThickness", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("TickStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "tickBrush", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.g("TickBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "fontBrush", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("FontBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "minorTickStartExtent", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("MinorTickStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "minorTickEndExtent", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("MinorTickEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "minorTickStrokeThickness", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("MinorTickStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "minorTickBrush", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("MinorTickBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "minorTickCount", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("MinorTickCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "scaleBrush", {
        get: function () {
            return this.c8;
        },
        set: function (a) {
            this.c8 = a;
            this.g("ScaleBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "backingBrush", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("BackingBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "backingOutline", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("BackingOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "backingStrokeThickness", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("BackingStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "backingOuterExtent", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("BackingOuterExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "backingOversweep", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("BackingOversweep");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "scaleOversweep", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("ScaleOversweep");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "scaleOversweepShape", {
        get: function () {
            return this.c9;
        },
        set: function (a) {
            this.c9 = a;
            this.g("ScaleOversweepShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "backingCornerRadius", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("BackingCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "backingInnerExtent", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("BackingInnerExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "backingShape", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("BackingShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "ranges", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Ranges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "radiusMultiplier", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("RadiusMultiplier");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "duplicateLabelOmissionStrategy", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("DuplicateLabelOmissionStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "isNeedleDraggingEnabled", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("IsNeedleDraggingEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "isNeedleDraggingConstrained", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IsNeedleDraggingConstrained");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "font", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("Font");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "labelFormat", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("LabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "labelFormatSpecifiers", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("LabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "transitionProgress", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("TransitionProgress");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "formatLabelRef", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("FormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "alignLabelRef", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.g("AlignLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "valueChangedRef", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            this.dd = a;
            this.g("ValueChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "actualMinimumValueChangeRef", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("ActualMinimumValueChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialGaugeDescription.prototype, "actualMaximumValueChangeRef", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("ActualMaximumValueChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    RadialGaugeDescription.$t = markType(RadialGaugeDescription, 'RadialGaugeDescription', Description.$);
    return RadialGaugeDescription;
}(Description));
export { RadialGaugeDescription };
