// igniteui-react-core/esm5/lib/ScatterPolygonSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ShapeSeriesBaseDescription } from "./ShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ScatterPolygonSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolygonSeriesDescription, _super);
    function ScatterPolygonSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.d3 = null;
        _this.eg = null;
        _this.ef = null;
        _this.d5 = 0;
        _this.d9 = null;
        _this.ea = null;
        _this.d7 = null;
        _this.ed = null;
        _this.d8 = null;
        _this.eb = null;
        _this.ee = null;
        _this.ec = null;
        return _this;
    }
    ScatterPolygonSeriesDescription.prototype.get_type = function () {
        return "ScatterPolygonSeries";
    };
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "shapeStyle", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("ShapeStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerType", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.g("MarkerType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerTemplateRef", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("MarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerThickness", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("MarkerThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerTemplateRef", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("ActualMarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerBrush", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("MarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerBrush", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("ActualMarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerOutline", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("MarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerOutline", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.g("ActualMarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerCollisionAvoidance", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("MarkerCollisionAvoidance");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerOutlineMode", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.g("MarkerOutlineMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerFillMode", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.g("MarkerFillMode");
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeriesDescription.$t = markType(ScatterPolygonSeriesDescription, 'ScatterPolygonSeriesDescription', ShapeSeriesBaseDescription.$);
    return ScatterPolygonSeriesDescription;
}(ShapeSeriesBaseDescription));
export { ScatterPolygonSeriesDescription };
