// igniteui-react-core/esm5/lib/ScatterContourSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ScatterTriangulationSeriesDescription } from "./ScatterTriangulationSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ScatterContourSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterContourSeriesDescription, _super);
    function ScatterContourSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.eb = null;
        _this.d9 = null;
        _this.ea = null;
        _this.d6 = null;
        _this.d5 = null;
        _this.d3 = null;
        return _this;
    }
    ScatterContourSeriesDescription.prototype.get_type = function () {
        return "ScatterContourSeries";
    };
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "valueMemberAsLegendLabel", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("ValueMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "valueMemberAsLegendUnit", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("ValueMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "fillScale", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.g("FillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "actualFillScale", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("ActualFillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "valueResolver", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("ValueResolver");
        },
        enumerable: false,
        configurable: true
    });
    ScatterContourSeriesDescription.$t = markType(ScatterContourSeriesDescription, 'ScatterContourSeriesDescription', ScatterTriangulationSeriesDescription.$);
    return ScatterContourSeriesDescription;
}(ScatterTriangulationSeriesDescription));
export { ScatterContourSeriesDescription };
