// igniteui-react-core/esm5/lib/ShapeSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ShapeSeriesBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ShapeSeriesBaseDescription, _super);
    function ShapeSeriesBaseDescription() {
        var _this = _super.call(this) || this;
        _this.dd = false;
        _this.dc = false;
        _this.dm = null;
        _this.dq = null;
        _this.dj = 0;
        _this.di = 0;
        _this.dr = null;
        _this.dt = null;
        _this.du = null;
        _this.dg = 0;
        _this.dp = null;
        _this.dn = null;
        _this.ds = null;
        return _this;
    }
    ShapeSeriesBaseDescription.prototype.get_type = function () {
        return "ShapeSeriesBase";
    };
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "isCustomShapeStyleAllowed", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            this.dd = a;
            this.g("IsCustomShapeStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "isCustomShapeMarkerStyleAllowed", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("IsCustomShapeMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "actualItemSearchMode", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("ActualItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "itemSearchPointsThreshold", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.g("ItemSearchPointsThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "shapeMemberPath", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("ShapeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "xAxisRef", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "yAxisRef", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "shapeFilterResolution", {
        get: function () {
            return this.dg;
        },
        set: function (a) {
            this.dg = a;
            this.g("ShapeFilterResolution");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "assigningShapeStyleRef", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("AssigningShapeStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "assigningShapeMarkerStyleRef", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("AssigningShapeMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "styleShapeRef", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("StyleShapeRef");
        },
        enumerable: false,
        configurable: true
    });
    ShapeSeriesBaseDescription.$t = markType(ShapeSeriesBaseDescription, 'ShapeSeriesBaseDescription', SeriesDescription.$);
    return ShapeSeriesBaseDescription;
}(SeriesDescription));
export { ShapeSeriesBaseDescription };
