// igniteui-react-core/esm5/lib/TreemapNodeStyleMappingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { TreemapNodeStyleDescription } from "./TreemapNodeStyleDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var TreemapNodeStyleMappingDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TreemapNodeStyleMappingDescription, _super);
    function TreemapNodeStyleMappingDescription() {
        var _this = _super.call(this) || this;
        _this.a3 = null;
        _this.ax = 0;
        _this.aw = 0;
        _this.a0 = null;
        _this.a2 = null;
        return _this;
    }
    TreemapNodeStyleMappingDescription.prototype.get_type = function () {
        return "TreemapNodeStyleMapping";
    };
    Object.defineProperty(TreemapNodeStyleMappingDescription.prototype, "targetType", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("TargetType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodeStyleMappingDescription.prototype, "minimumValue", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("MinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodeStyleMappingDescription.prototype, "maximumValue", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("MaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodeStyleMappingDescription.prototype, "value", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodeStyleMappingDescription.prototype, "mappingMode", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("MappingMode");
        },
        enumerable: false,
        configurable: true
    });
    TreemapNodeStyleMappingDescription.$t = markType(TreemapNodeStyleMappingDescription, 'TreemapNodeStyleMappingDescription', TreemapNodeStyleDescription.$);
    return TreemapNodeStyleMappingDescription;
}(TreemapNodeStyleDescription));
export { TreemapNodeStyleMappingDescription };
