// igniteui-react-core/esm5/lib/ValueLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ValueLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ValueLayerDescription, _super);
    function ValueLayerDescription() {
        var _this = _super.call(this) || this;
        _this.fm = null;
        _this.fh = null;
        _this.fn = null;
        _this.fk = null;
        _this.fl = null;
        _this.fi = null;
        _this.fj = null;
        _this.el = false;
        _this.ej = false;
        _this.fq = null;
        _this.fo = null;
        _this.ep = 0;
        _this.fd = 0;
        _this.fp = null;
        _this.er = 0;
        _this.et = 0;
        _this.es = 0;
        _this.eq = 0;
        _this.ex = 0;
        _this.ez = 0;
        _this.ey = 0;
        _this.ew = 0;
        _this.eu = 0;
        _this.ft = null;
        _this.fr = null;
        _this.ev = 0;
        _this.fe = 0;
        _this.fs = null;
        _this.e0 = 0;
        _this.ek = false;
        return _this;
    }
    ValueLayerDescription.prototype.get_type = function () {
        return "ValueLayer";
    };
    Object.defineProperty(ValueLayerDescription.prototype, "valueMode", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("ValueMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "horizontalLineStroke", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("HorizontalLineStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "verticalLineStroke", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("VerticalLineStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "targetSeriesName", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("TargetSeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "targetSeriesRef", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("TargetSeriesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "targetAxisName", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("TargetAxisName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "targetAxisRef", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("TargetAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "useInterpolation", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("UseInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "isAxisAnnotationEnabled", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("IsAxisAnnotationEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationTextColor", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("XAxisAnnotationTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationBackground", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("XAxisAnnotationBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationBackgroundCornerRadius", {
        get: function () {
            return this.ep;
        },
        set: function (a) {
            this.ep = a;
            this.g("XAxisAnnotationBackgroundCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationInterpolatedValuePrecision", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("XAxisAnnotationInterpolatedValuePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationOutline", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("XAxisAnnotationOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationPaddingLeft", {
        get: function () {
            return this.er;
        },
        set: function (a) {
            this.er = a;
            this.g("XAxisAnnotationPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationPaddingTop", {
        get: function () {
            return this.et;
        },
        set: function (a) {
            this.et = a;
            this.g("XAxisAnnotationPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationPaddingRight", {
        get: function () {
            return this.es;
        },
        set: function (a) {
            this.es = a;
            this.g("XAxisAnnotationPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationPaddingBottom", {
        get: function () {
            return this.eq;
        },
        set: function (a) {
            this.eq = a;
            this.g("XAxisAnnotationPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationPaddingLeft", {
        get: function () {
            return this.ex;
        },
        set: function (a) {
            this.ex = a;
            this.g("YAxisAnnotationPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationPaddingTop", {
        get: function () {
            return this.ez;
        },
        set: function (a) {
            this.ez = a;
            this.g("YAxisAnnotationPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationPaddingRight", {
        get: function () {
            return this.ey;
        },
        set: function (a) {
            this.ey = a;
            this.g("YAxisAnnotationPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationPaddingBottom", {
        get: function () {
            return this.ew;
        },
        set: function (a) {
            this.ew = a;
            this.g("YAxisAnnotationPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationStrokeThickness", {
        get: function () {
            return this.eu;
        },
        set: function (a) {
            this.eu = a;
            this.g("XAxisAnnotationStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationTextColor", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("YAxisAnnotationTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationBackground", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("YAxisAnnotationBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationBackgroundCornerRadius", {
        get: function () {
            return this.ev;
        },
        set: function (a) {
            this.ev = a;
            this.g("YAxisAnnotationBackgroundCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationInterpolatedValuePrecision", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("YAxisAnnotationInterpolatedValuePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationOutline", {
        get: function () {
            return this.fs;
        },
        set: function (a) {
            this.fs = a;
            this.g("YAxisAnnotationOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationStrokeThickness", {
        get: function () {
            return this.e0;
        },
        set: function (a) {
            this.e0 = a;
            this.g("YAxisAnnotationStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "skipUnknownValues", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("SkipUnknownValues");
        },
        enumerable: false,
        configurable: true
    });
    ValueLayerDescription.$t = markType(ValueLayerDescription, 'ValueLayerDescription', AnnotationLayerDescription.$);
    return ValueLayerDescription;
}(AnnotationLayerDescription));
export { ValueLayerDescription };
