// igniteui-react-core/esm5/lib/WebActionStripDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebDisplayDensityBaseElementDescription } from "./WebDisplayDensityBaseElementDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebActionStripDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebActionStripDescription, _super);
    function WebActionStripDescription() {
        var _this = _super.call(this) || this;
        _this.s = null;
        _this.m = null;
        _this.q = false;
        _this.o = null;
        _this.u = null;
        return _this;
    }
    WebActionStripDescription.prototype.get_type = function () {
        return "WebActionStrip";
    };
    Object.defineProperty(WebActionStripDescription.prototype, "context", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Context");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebActionStripDescription.prototype, "actionButtons", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("ActionButtons");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebActionStripDescription.prototype, "hidden", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Hidden");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebActionStripDescription.prototype, "resourceStrings", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("ResourceStrings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebActionStripDescription.prototype, "name", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    WebActionStripDescription.$t = markType(WebActionStripDescription, 'WebActionStripDescription', WebDisplayDensityBaseElementDescription.$);
    return WebActionStripDescription;
}(WebDisplayDensityBaseElementDescription));
export { WebActionStripDescription };
