// igniteui-react-core/esm5/lib/WebButtonBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebButtonBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebButtonBaseDescription, _super);
    function WebButtonBaseDescription() {
        var _this = _super.call(this) || this;
        _this.m = null;
        _this.p = null;
        _this.n = null;
        _this.s = null;
        _this.q = null;
        _this.h = false;
        _this.j = null;
        _this.r = null;
        _this.l = null;
        _this.o = null;
        _this.k = null;
        return _this;
    }
    WebButtonBaseDescription.prototype.get_type = function () {
        return "WebButtonBase";
    };
    Object.defineProperty(WebButtonBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "displayType", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("DisplayType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "href", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Href");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "download", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Download");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "target", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Target");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "rel", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Rel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "disabled", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "ariaLabel", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("AriaLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "size", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Size");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "clickedRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ClickedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "focusRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("FocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonBaseDescription.prototype, "blurRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("BlurRef");
        },
        enumerable: false,
        configurable: true
    });
    WebButtonBaseDescription.$t = markType(WebButtonBaseDescription, 'WebButtonBaseDescription', Description.$);
    return WebButtonBaseDescription;
}(Description));
export { WebButtonBaseDescription };
