// igniteui-react-core/esm5/lib/WebCellTypeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebCellTypeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebCellTypeDescription, _super);
    function WebCellTypeDescription() {
        var _this = _super.call(this) || this;
        _this.af = null;
        _this.ad = null;
        _this.u = false;
        _this.q = false;
        _this.r = false;
        _this.s = false;
        _this.h = null;
        _this.j = null;
        _this.l = null;
        _this.ac = null;
        _this.n = null;
        _this.t = false;
        _this.ae = null;
        _this.al = null;
        _this.aa = 0;
        return _this;
    }
    WebCellTypeDescription.prototype.get_type = function () {
        return "WebCellType";
    };
    Object.defineProperty(WebCellTypeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "value", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "editValue", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("EditValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "selected", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Selected");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "active", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Active");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "editable", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Editable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "editMode", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("EditMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "column", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "grid", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Grid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "id", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "cellID", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("CellID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "validation", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Validation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "readonly", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("Readonly");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "title", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "width", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCellTypeDescription.prototype, "visibleColumnIndex", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("VisibleColumnIndex");
        },
        enumerable: false,
        configurable: true
    });
    WebCellTypeDescription.$t = markType(WebCellTypeDescription, 'WebCellTypeDescription', Description.$);
    WebCellTypeDescription.__marshalByValue = true;
    WebCellTypeDescription.__marshalByValueAlias = "CellType";
    return WebCellTypeDescription;
}(Description));
export { WebCellTypeDescription };
