// igniteui-react-core/esm5/lib/WebColumnSelectionEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebColumnSelectionEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebColumnSelectionEventArgsDetailDescription, _super);
    function WebColumnSelectionEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.o = null;
        _this.n = null;
        _this.m = null;
        _this.p = null;
        return _this;
    }
    WebColumnSelectionEventArgsDetailDescription.prototype.get_type = function () {
        return "WebColumnSelectionEventArgsDetail";
    };
    Object.defineProperty(WebColumnSelectionEventArgsDetailDescription.prototype, "oldSelection", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("OldSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnSelectionEventArgsDetailDescription.prototype, "newSelection", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("NewSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnSelectionEventArgsDetailDescription.prototype, "added", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Added");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnSelectionEventArgsDetailDescription.prototype, "removed", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Removed");
        },
        enumerable: false,
        configurable: true
    });
    WebColumnSelectionEventArgsDetailDescription.$t = markType(WebColumnSelectionEventArgsDetailDescription, 'WebColumnSelectionEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebColumnSelectionEventArgsDetailDescription.__marshalByValue1 = true;
    WebColumnSelectionEventArgsDetailDescription.__marshalByValueAlias1 = "ColumnSelectionEventArgsDetail";
    return WebColumnSelectionEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebColumnSelectionEventArgsDetailDescription };
