// igniteui-react-core/esm5/lib/WebComboItemDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebComboItemDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebComboItemDescription, _super);
    function WebComboItemDescription() {
        var _this = _super.call(this) || this;
        _this.n = 0;
        _this.j = false;
        _this.h = false;
        _this.i = false;
        return _this;
    }
    WebComboItemDescription.prototype.get_type = function () {
        return "WebComboItem";
    };
    Object.defineProperty(WebComboItemDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboItemDescription.prototype, "index", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Index");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboItemDescription.prototype, "selected", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Selected");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboItemDescription.prototype, "active", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Active");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboItemDescription.prototype, "hideCheckbox", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("HideCheckbox");
        },
        enumerable: false,
        configurable: true
    });
    WebComboItemDescription.$t = markType(WebComboItemDescription, 'WebComboItemDescription', Description.$);
    return WebComboItemDescription;
}(Description));
export { WebComboItemDescription };
