// igniteui-react-core/esm5/lib/WebDropdownDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebDropdownDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebDropdownDescription, _super);
    function WebDropdownDescription() {
        var _this = _super.call(this) || this;
        _this.j = false;
        _this.k = false;
        _this.u = null;
        _this.v = null;
        _this.w = null;
        _this.h = false;
        _this.r = 0;
        _this.i = false;
        _this.l = false;
        _this.x = null;
        _this.t = null;
        return _this;
    }
    WebDropdownDescription.prototype.get_type = function () {
        return "WebDropdown";
    };
    Object.defineProperty(WebDropdownDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "keepOpenOnSelect", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("KeepOpenOnSelect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "open", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Open");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "placement", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Placement");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "positionStrategy", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("PositionStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "scrollStrategy", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("ScrollStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "flip", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Flip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "distance", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Distance");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "keepOpenOnOutsideClick", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("KeepOpenOnOutsideClick");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "sameWidth", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("SameWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "size", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Size");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDropdownDescription.prototype, "changeRef", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebDropdownDescription.$t = markType(WebDropdownDescription, 'WebDropdownDescription', Description.$);
    return WebDropdownDescription;
}(Description));
export { WebDropdownDescription };
