// igniteui-react-core/esm5/lib/WebExporterEventDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebExporterEventDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebExporterEventDescription, _super);
    function WebExporterEventDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.j = null;
        _this.l = null;
        _this.n = false;
        return _this;
    }
    WebExporterEventDescription.prototype.get_type = function () {
        return "WebExporterEvent";
    };
    Object.defineProperty(WebExporterEventDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebExporterEventDescription.prototype, "exporter", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Exporter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebExporterEventDescription.prototype, "options", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Options");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebExporterEventDescription.prototype, "grid", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Grid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebExporterEventDescription.prototype, "cancel", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Cancel");
        },
        enumerable: false,
        configurable: true
    });
    WebExporterEventDescription.$t = markType(WebExporterEventDescription, 'WebExporterEventDescription', Description.$);
    return WebExporterEventDescription;
}(Description));
export { WebExporterEventDescription };
