// igniteui-react-core/esm5/lib/WebGridDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebGridBaseDirectiveDescription } from "./WebGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGridDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridDescription, _super);
    function WebGridDescription() {
        var _this = _super.call(this) || this;
        _this.gp = false;
        _this.g1 = null;
        _this.gz = null;
        _this.g5 = null;
        _this.gy = null;
        _this.gv = 0;
        _this.gk = null;
        _this.gj = null;
        _this.gq = false;
        _this.gn = null;
        _this.g0 = null;
        _this.g2 = null;
        _this.g4 = null;
        _this.gr = false;
        _this.gx = null;
        _this.g3 = null;
        return _this;
    }
    WebGridDescription.prototype.get_type = function () {
        return "WebGrid";
    };
    Object.defineProperty(WebGridDescription.prototype, "groupsExpanded", {
        get: function () {
            return this.gp;
        },
        set: function (a) {
            this.gp = a;
            this.g("GroupsExpanded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "dropAreaTemplateRef", {
        get: function () {
            return this.g1;
        },
        set: function (a) {
            this.g1 = a;
            this.g("DropAreaTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "detailTemplateRef", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("DetailTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "id", {
        get: function () {
            return this.g5;
        },
        set: function (a) {
            this.g5 = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "dataRef", {
        get: function () {
            return this.gy;
        },
        set: function (a) {
            this.gy = a;
            this.g("DataRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "totalItemCount", {
        get: function () {
            return this.gv;
        },
        set: function (a) {
            this.gv = a;
            this.g("TotalItemCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "groupingExpressions", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("GroupingExpressions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "groupingExpansionState", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("GroupingExpansionState");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "hideGroupedColumns", {
        get: function () {
            return this.gq;
        },
        set: function (a) {
            this.gq = a;
            this.g("HideGroupedColumns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "groupStrategy", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("GroupStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "dropAreaMessage", {
        get: function () {
            return this.g0;
        },
        set: function (a) {
            this.g0 = a;
            this.g("DropAreaMessage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "groupByRowSelectorTemplateRef", {
        get: function () {
            return this.g2;
        },
        set: function (a) {
            this.g2 = a;
            this.g("GroupByRowSelectorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "groupRowTemplateRef", {
        get: function () {
            return this.g4;
        },
        set: function (a) {
            this.g4 = a;
            this.g("GroupRowTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "showGroupArea", {
        get: function () {
            return this.gr;
        },
        set: function (a) {
            this.gr = a;
            this.g("ShowGroupArea");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "dataPreLoadRef", {
        get: function () {
            return this.gx;
        },
        set: function (a) {
            this.gx = a;
            this.g("DataPreLoadRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridDescription.prototype, "groupingDoneRef", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("GroupingDoneRef");
        },
        enumerable: false,
        configurable: true
    });
    WebGridDescription.$t = markType(WebGridDescription, 'WebGridDescription', WebGridBaseDirectiveDescription.$);
    return WebGridDescription;
}(WebGridBaseDirectiveDescription));
export { WebGridDescription };
