// igniteui-react-core/esm5/lib/WebPivotGridDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebGridBaseDirectiveDescription } from "./WebGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPivotGridDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPivotGridDescription, _super);
    function WebPivotGridDescription() {
        var _this = _super.call(this) || this;
        _this.gx = null;
        _this.gj = null;
        _this.gm = false;
        _this.gn = false;
        _this.gl = false;
        _this.gr = null;
        _this.gv = null;
        _this.gt = null;
        _this.gw = null;
        _this.gs = null;
        _this.gy = null;
        _this.gu = null;
        _this.gz = null;
        return _this;
    }
    WebPivotGridDescription.prototype.get_type = function () {
        return "WebPivotGrid";
    };
    Object.defineProperty(WebPivotGridDescription.prototype, "valueChipTemplateRef", {
        get: function () {
            return this.gx;
        },
        set: function (a) {
            this.gx = a;
            this.g("ValueChipTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "pivotConfiguration", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("PivotConfiguration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "showPivotConfigurationUI", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("ShowPivotConfigurationUI");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "superCompactMode", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("SuperCompactMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "defaultExpandState", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("DefaultExpandState");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "dataRef", {
        get: function () {
            return this.gr;
        },
        set: function (a) {
            this.gr = a;
            this.g("DataRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "emptyPivotGridTemplateRef", {
        get: function () {
            return this.gv;
        },
        set: function (a) {
            this.gv = a;
            this.g("EmptyPivotGridTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "dimensionsChangeRef", {
        get: function () {
            return this.gt;
        },
        set: function (a) {
            this.gt = a;
            this.g("DimensionsChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "pivotConfigurationChangeRef", {
        get: function () {
            return this.gw;
        },
        set: function (a) {
            this.gw = a;
            this.g("PivotConfigurationChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "dimensionInitRef", {
        get: function () {
            return this.gs;
        },
        set: function (a) {
            this.gs = a;
            this.g("DimensionInitRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "valueInitRef", {
        get: function () {
            return this.gy;
        },
        set: function (a) {
            this.gy = a;
            this.g("ValueInitRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "dimensionsSortingExpressionsChangeRef", {
        get: function () {
            return this.gu;
        },
        set: function (a) {
            this.gu = a;
            this.g("DimensionsSortingExpressionsChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "valuesChangeRef", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("ValuesChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebPivotGridDescription.$t = markType(WebPivotGridDescription, 'WebPivotGridDescription', WebGridBaseDirectiveDescription.$);
    return WebPivotGridDescription;
}(WebGridBaseDirectiveDescription));
export { WebPivotGridDescription };
