// igniteui-react-core/esm5/lib/WebPaginatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebDisplayDensityBaseElementDescription } from "./WebDisplayDensityBaseElementDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPaginatorDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPaginatorDescription, _super);
    function WebPaginatorDescription() {
        var _this = _super.call(this) || this;
        _this.u = 0;
        _this.s = 0;
        _this.t = 0;
        _this.v = 0;
        _this.m = null;
        _this.o = null;
        _this.q = null;
        _this.aa = null;
        _this.ae = null;
        _this.ab = null;
        _this.ac = null;
        _this.ad = null;
        return _this;
    }
    WebPaginatorDescription.prototype.get_type = function () {
        return "WebPaginator";
    };
    Object.defineProperty(WebPaginatorDescription.prototype, "totalPages", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("TotalPages");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "page", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Page");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "perPage", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("PerPage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "totalRecords", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("TotalRecords");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "selectOptions", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("SelectOptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "overlaySettings", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("OverlaySettings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "resourceStrings", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("ResourceStrings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "name", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "perPageChangeRef", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("PerPageChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "pageChangeRef", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("PageChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "pagingRef", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("PagingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaginatorDescription.prototype, "pagingDoneRef", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("PagingDoneRef");
        },
        enumerable: false,
        configurable: true
    });
    WebPaginatorDescription.$t = markType(WebPaginatorDescription, 'WebPaginatorDescription', WebDisplayDensityBaseElementDescription.$);
    return WebPaginatorDescription;
}(WebDisplayDensityBaseElementDescription));
export { WebPaginatorDescription };
