// igniteui-react-core/esm5/lib/WebPivotDateDimensionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebPivotDimensionDescription } from "./WebPivotDimensionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPivotDateDimensionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPivotDateDimensionDescription, _super);
    function WebPivotDateDimensionDescription() {
        var _this = _super.call(this) || this;
        _this.aj = false;
        _this.am = null;
        _this.ac = null;
        _this.ag = null;
        _this.ae = null;
        return _this;
    }
    WebPivotDateDimensionDescription.prototype.get_type = function () {
        return "WebPivotDateDimension";
    };
    Object.defineProperty(WebPivotDateDimensionDescription.prototype, "enabled", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("Enabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDateDimensionDescription.prototype, "dataType", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("DataType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDateDimensionDescription.prototype, "resourceStrings", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("ResourceStrings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDateDimensionDescription.prototype, "baseDimension", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("BaseDimension");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDateDimensionDescription.prototype, "options", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Options");
        },
        enumerable: false,
        configurable: true
    });
    WebPivotDateDimensionDescription.$t = markType(WebPivotDateDimensionDescription, 'WebPivotDateDimensionDescription', WebPivotDimensionDescription.$);
    WebPivotDateDimensionDescription.__marshalByValue1 = true;
    WebPivotDateDimensionDescription.__marshalByValueAlias1 = "PivotDateDimension";
    return WebPivotDateDimensionDescription;
}(WebPivotDimensionDescription));
export { WebPivotDateDimensionDescription };
