// igniteui-react-core/esm5/lib/WebPinRowEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebCancelableEventArgsDetailDescription } from "./WebCancelableEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPinRowEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPinRowEventArgsDetailDescription, _super);
    function WebPinRowEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.t = null;
        _this.m = null;
        _this.r = 0;
        _this.p = false;
        return _this;
    }
    WebPinRowEventArgsDetailDescription.prototype.get_type = function () {
        return "WebPinRowEventArgsDetail";
    };
    Object.defineProperty(WebPinRowEventArgsDetailDescription.prototype, "rowID", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("RowID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPinRowEventArgsDetailDescription.prototype, "row", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Row");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPinRowEventArgsDetailDescription.prototype, "insertAtIndex", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("InsertAtIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPinRowEventArgsDetailDescription.prototype, "isPinned", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IsPinned");
        },
        enumerable: false,
        configurable: true
    });
    WebPinRowEventArgsDetailDescription.$t = markType(WebPinRowEventArgsDetailDescription, 'WebPinRowEventArgsDetailDescription', WebCancelableEventArgsDetailDescription.$);
    WebPinRowEventArgsDetailDescription.__marshalByValue1 = true;
    WebPinRowEventArgsDetailDescription.__marshalByValueAlias1 = "PinRowEventArgsDetail";
    return WebPinRowEventArgsDetailDescription;
}(WebCancelableEventArgsDetailDescription));
export { WebPinRowEventArgsDetailDescription };
