// igniteui-react-core/esm5/lib/WebStepDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebStepDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebStepDescription, _super);
    function WebStepDescription() {
        var _this = _super.call(this) || this;
        _this.l = false;
        _this.h = false;
        _this.n = false;
        _this.k = false;
        _this.i = false;
        _this.o = false;
        _this.ac = null;
        _this.ad = null;
        _this.ab = null;
        _this.z = 0;
        _this.j = false;
        _this.m = false;
        _this.p = false;
        return _this;
    }
    WebStepDescription.prototype.get_type = function () {
        return "WebStep";
    };
    Object.defineProperty(WebStepDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "invalid", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Invalid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "active", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Active");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "optional", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Optional");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "disabled", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "complete", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Complete");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "previousComplete", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("PreviousComplete");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "stepType", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("StepType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "titlePosition", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("TitlePosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "orientation", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "index", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("Index");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "contentTop", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ContentTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "linearDisabled", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("LinearDisabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepDescription.prototype, "visited", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Visited");
        },
        enumerable: false,
        configurable: true
    });
    WebStepDescription.$t = markType(WebStepDescription, 'WebStepDescription', Description.$);
    return WebStepDescription;
}(Description));
export { WebStepDescription };
