// igniteui-react-core/esm5/lib/WebStepperDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebStepperDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebStepperDescription, _super);
    function WebStepperDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.q = null;
        _this.r = null;
        _this.k = false;
        _this.j = false;
        _this.p = null;
        _this.s = null;
        _this.o = null;
        _this.n = null;
        return _this;
    }
    WebStepperDescription.prototype.get_type = function () {
        return "WebStepper";
    };
    Object.defineProperty(WebStepperDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "steps", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Steps");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "orientation", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "stepType", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("StepType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "linear", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Linear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "contentTop", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ContentTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "dir", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Dir");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "titlePosition", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("TitlePosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "activeStepChangingRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("ActiveStepChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebStepperDescription.prototype, "activeStepChangedRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ActiveStepChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    WebStepperDescription.$t = markType(WebStepperDescription, 'WebStepperDescription', Description.$);
    return WebStepperDescription;
}(Description));
export { WebStepperDescription };
