// igniteui-react-core/esm5/lib/WebSnackbarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSnackbarDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebSnackbarDescription, _super);
    function WebSnackbarDescription() {
        var _this = _super.call(this) || this;
        _this.i = false;
        _this.l = 0;
        _this.h = false;
        _this.o = null;
        _this.n = null;
        return _this;
    }
    WebSnackbarDescription.prototype.get_type = function () {
        return "WebSnackbar";
    };
    Object.defineProperty(WebSnackbarDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSnackbarDescription.prototype, "open", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Open");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSnackbarDescription.prototype, "displayTime", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("DisplayTime");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSnackbarDescription.prototype, "keepOpen", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("KeepOpen");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSnackbarDescription.prototype, "actionText", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("ActionText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSnackbarDescription.prototype, "actionRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ActionRef");
        },
        enumerable: false,
        configurable: true
    });
    WebSnackbarDescription.$t = markType(WebSnackbarDescription, 'WebSnackbarDescription', Description.$);
    return WebSnackbarDescription;
}(Description));
export { WebSnackbarDescription };
