// igniteui-react-core/esm5/lib/WebTreeGridDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebGridBaseDirectiveDescription } from "./WebGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebTreeGridDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebTreeGridDescription, _super);
    function WebTreeGridDescription() {
        var _this = _super.call(this) || this;
        _this.gr = null;
        _this.gt = null;
        _this.gu = null;
        _this.gn = false;
        _this.gv = null;
        _this.gk = null;
        _this.gj = null;
        _this.gs = null;
        _this.gp = 0;
        _this.gw = null;
        return _this;
    }
    WebTreeGridDescription.prototype.get_type = function () {
        return "WebTreeGrid";
    };
    Object.defineProperty(WebTreeGridDescription.prototype, "childDataKey", {
        get: function () {
            return this.gr;
        },
        set: function (a) {
            this.gr = a;
            this.g("ChildDataKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "foreignKey", {
        get: function () {
            return this.gt;
        },
        set: function (a) {
            this.gt = a;
            this.g("ForeignKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "hasChildrenKey", {
        get: function () {
            return this.gu;
        },
        set: function (a) {
            this.gu = a;
            this.g("HasChildrenKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "cascadeOnDelete", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("CascadeOnDelete");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "id", {
        get: function () {
            return this.gv;
        },
        set: function (a) {
            this.gv = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "rootRecords", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("RootRecords");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "processedRootRecords", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("ProcessedRootRecords");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "dataRef", {
        get: function () {
            return this.gs;
        },
        set: function (a) {
            this.gs = a;
            this.g("DataRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "expansionDepth", {
        get: function () {
            return this.gp;
        },
        set: function (a) {
            this.gp = a;
            this.g("ExpansionDepth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTreeGridDescription.prototype, "rowLoadingIndicatorTemplateRef", {
        get: function () {
            return this.gw;
        },
        set: function (a) {
            this.gw = a;
            this.g("RowLoadingIndicatorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    WebTreeGridDescription.$t = markType(WebTreeGridDescription, 'WebTreeGridDescription', WebGridBaseDirectiveDescription.$);
    return WebTreeGridDescription;
}(WebGridBaseDirectiveDescription));
export { WebTreeGridDescription };
