// igniteui-react-core/esm5/lib/WebToggleDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebToggleDirectiveDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebToggleDirectiveDescription, _super);
    function WebToggleDirectiveDescription() {
        var _this = _super.call(this) || this;
        _this.k = null;
        _this.l = null;
        _this.m = null;
        _this.i = null;
        _this.j = null;
        _this.h = null;
        return _this;
    }
    WebToggleDirectiveDescription.prototype.get_type = function () {
        return "WebToggleDirective";
    };
    Object.defineProperty(WebToggleDirectiveDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebToggleDirectiveDescription.prototype, "id", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebToggleDirectiveDescription.prototype, "openedRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("OpenedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebToggleDirectiveDescription.prototype, "openingRef", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("OpeningRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebToggleDirectiveDescription.prototype, "closedRef", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ClosedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebToggleDirectiveDescription.prototype, "closingRef", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ClosingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebToggleDirectiveDescription.prototype, "appendedRef", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("AppendedRef");
        },
        enumerable: false,
        configurable: true
    });
    WebToggleDirectiveDescription.$t = markType(WebToggleDirectiveDescription, 'WebToggleDirectiveDescription', Description.$);
    return WebToggleDirectiveDescription;
}(Description));
export { WebToggleDirectiveDescription };
