// igniteui-react-charts/esm5/lib/CategoryToolTipLayer_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { AnnotationLayer } from "./AnnotationLayer";
import { Axis } from "./Axis";
import { DependencyProperty } from "igniteui-react-core";
import { CategoryTooltipLayerPosition_$type } from "./CategoryTooltipLayerPosition";
import { CategoryToolTipLayerFrame } from "./CategoryToolTipLayerFrame";
import { Base, enumGetBox, typeCast, fromEnum, Point_$type, markType, markDep, String_$type } from "igniteui-react-core";
import { Series } from "./Series";
import { CategoryTooltipSeriesInfo } from "./CategoryTooltipSeriesInfo";
import { XamDataChart } from "./XamDataChart";
import { List$1 } from "igniteui-react-core";
import { DataContext } from "igniteui-react-core";
import { Size } from "igniteui-react-core";
import { DeviceUtils } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
import { AnnotationLayerView } from "./AnnotationLayerView";
import { PointerTooltip } from "./PointerTooltip";
/**
 * @hidden
 */
var CategoryToolTipLayer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryToolTipLayer, _super);
    function CategoryToolTipLayer() {
        var _this = _super.call(this) || this;
        var a = new CategoryToolTipLayerFrame();
        var b = new CategoryToolTipLayerFrame();
        var c = new CategoryToolTipLayerFrame();
        _this.qq = a;
        _this.qp = b;
        _this.qr = c;
        _this.ab = CategoryToolTipLayer.$;
        return _this;
    }
    CategoryToolTipLayer.prototype.get_dc = function () {
        return true;
    };
    CategoryToolTipLayer.prototype.by = function () {
        return new CategoryToolTipLayerView(this);
    };
    CategoryToolTipLayer.prototype.l4 = function (a) {
        _super.prototype.l4.call(this, a);
        this.td = a;
    };
    CategoryToolTipLayer.prototype.get_dn = function () {
        return true;
    };
    CategoryToolTipLayer.prototype.l9 = function (a, b, c, d) {
        _super.prototype.l9.call(this, a, b, c, d);
        switch (b) {
            case CategoryToolTipLayer.$$p[0]:
                this.mn(true);
                break;
            case CategoryToolTipLayer.$$p[2]:
                this.mn(true);
                break;
            case CategoryToolTipLayer.$$p[1]:
                this.mn(true);
                break;
            case "SeriesViewer":
                this.td.cf();
                break;
        }
    };
    CategoryToolTipLayer.prototype.r2 = function (a, b) {
        if (this.tb == a) {
            if (!this.at.c) {
                this.mn(b);
            }
        }
    };
    CategoryToolTipLayer.prototype.r3 = function (a, b) {
        if (!this.at.c) {
            this.mn(b);
        }
    };
    CategoryToolTipLayer.prototype.r5 = function (a, b) {
        var e_1, _a, e_2, _b;
        _super.prototype.r5.call(this, a, b);
        var c = a;
        var d = this.tf;
        c.l = NaN;
        c.k = NaN;
        c.g = NaN;
        c.h = NaN;
        c.k = NaN;
        c.l = NaN;
        if (isNaN_(this.s8.x) && isNaN_(this.s8.y)) {
            this.td.cd(this.td.ca());
            return;
        }
        if (this.tb != null) {
            var e = this.tb;
            if (e == null || !e.ca || e.b8) {
                return;
            }
            if (!e.cn()) {
                return;
            }
            var f = e;
            var g = this.te(f.dd, d);
            this.tk(g, c, d);
        }
        else {
            if (typeCast(XamDataChart.$, this.ce) !== null) {
                var h = this.ce;
                var i = true;
                var j = false;
                var k = new List$1(Series.$, 0);
                try {
                    for (var _c = __values(fromEnum(h.axes)), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var l = _d.value;
                        if (l.ca && !l.b8) {
                            if (!l.cn()) {
                                continue;
                            }
                            if (i) {
                                i = false;
                                j = l.co;
                            }
                            else {
                                if (l.co != j) {
                                    continue;
                                }
                            }
                            try {
                                for (var _e = (e_2 = void 0, __values(fromEnum(l.dd))), _f = _e.next(); !_f.done; _f = _e.next()) {
                                    var m = _f.value;
                                    k.add(m);
                                }
                            }
                            catch (e_2_1) {
                                e_2 = { error: e_2_1 };
                            }
                            finally {
                                try {
                                    if (_f && !_f.done && (_b = _e.return))
                                        _b.call(_e);
                                }
                                finally {
                                    if (e_2)
                                        throw e_2.error;
                                }
                            }
                        }
                    }
                }
                catch (e_1_1) {
                    e_1 = { error: e_1_1 };
                }
                finally {
                    try {
                        if (_d && !_d.done && (_a = _c.return))
                            _a.call(_c);
                    }
                    finally {
                        if (e_1)
                            throw e_1.error;
                    }
                }
                var n = this.te(k, d);
                this.tk(n, c, d);
            }
        }
    };
    CategoryToolTipLayer.prototype.te = function (a, b) {
        var c = 1.7976931348623157E+308;
        var d = -1.7976931348623157E+308;
        var e = new List$1(DataContext.$, 0);
        var f = new List$1(Base.$, 0);
        var g = false;
        var h = true;
        for (var i = 0; i < a.count; i++) {
            var j = a.item(i);
            if (!j.ds && !j.dj) {
                continue;
            }
            if (j.isStacked) {
                continue;
            }
            var k = j;
            if (!k.categoryAxis.cn()) {
                continue;
            }
            if (h) {
                g = j.isVertical;
            }
            else {
                if (g != j.isVertical) {
                    continue;
                }
            }
            var l = this.ry(j);
            if (!l.c) {
                continue;
            }
            var m = j.ps(l.d, b, false);
            if (g) {
                if (!isNaN_(m.y)) {
                    c = Math.min(c, m.y);
                    d = Math.max(d, m.y);
                }
            }
            else {
                if (!isNaN_(m.x)) {
                    c = Math.min(c, m.x);
                    d = Math.max(d, m.x);
                }
            }
            var n = this.tg(j);
            if (n == null) {
                continue;
            }
            var o = null;
            var p = j.hy(this.s8);
            if (j.dp && p != null) {
                o = j.av(p);
            }
            else {
                o = new DataContext();
                o.item = p;
                o.series = j;
            }
            e.add(o);
            f.add1(n);
        }
        var q = 0;
        var r = 0;
        if (g) {
            r = (c + d) / 2;
            q = this.p7.right;
            if (this.tc == 2 || this.tc == 1) {
                q = this.p7.left;
            }
            if (r < 0 || r > this.p7.bottom) {
                r = NaN;
            }
        }
        else {
            q = (c + d) / 2;
            r = this.p7.top;
            if (this.tc == 2 || this.tc == 1) {
                r = this.p7.bottom;
            }
            if (q < 0 || q > this.p7.right) {
                q = NaN;
            }
        }
        var s = new CategoryTooltipSeriesInfo();
        s.c = { $type: Point_$type, x: q, y: r };
        s.a = e;
        s.b = f;
        return s;
    };
    CategoryToolTipLayer.prototype.li = function () {
        this.s8 = { $type: Point_$type, x: NaN, y: NaN };
        this.b1.ap();
    };
    CategoryToolTipLayer.prototype.tk = function (a, b, c) {
        var d = a.c;
        var e = a.a;
        var f = a.b;
        d = this.td.cg(d);
        var g = 0;
        var h = this.b1.bu;
        var i = this.td.ca();
        this.td.cc();
        var j = false;
        if (e.count == 0) {
            this.b1.ap();
            return;
        }
        var k = false;
        for (var l = 0; l < e.count; l++) {
            var m = e._inner[l].series;
            j = m.isVertical;
            var n = e._inner[l];
            var o = f._inner[l];
            if (this.td.b7(m, o, n)) {
                k = true;
            }
            g++;
        }
        if (!k) {
            this.td.cd(i);
            return;
        }
        var p = this.td.ch(i);
        var q = 0;
        var r = 0;
        var s = 0;
        var t = 0;
        var u = false;
        var v = false;
        var w = this.tc;
        if (w == 2 || w == 3) {
            v = true;
        }
        if (w == 2 || w == 1) {
            u = true;
        }
        var x = this.td.ci();
        if (j) {
            if (!u && !v && d.x + (p.width + DeviceUtils.g(10)) > x.width) {
                v = true;
            }
            if (u && !v && d.x - (p.width + DeviceUtils.g(10)) < 0) {
                v = true;
            }
        }
        else {
            if (!u && !v && d.y - (p.height + 10) < 0) {
                v = true;
            }
            if (u && !v && d.y + (p.height + 10) > x.height) {
                v = true;
            }
        }
        var y = DeviceUtils.g(10);
        var z = DeviceUtils.g(5);
        if (!u) {
            if (!j) {
                z = p.height + y;
            }
            if (v) {
                if (j) {
                    z = z + p.width + y;
                }
                else {
                    z = 0;
                }
            }
        }
        else {
            if (j) {
                z = p.width + y;
                if (v) {
                    z = 0;
                }
            }
            else {
                z = DeviceUtils.g(5);
                if (v) {
                    z = z + p.height + y;
                }
            }
        }
        if (j) {
            r = 0 - (y + DeviceUtils.g(5));
            if (u) {
                r = (r * -1) + p.width;
            }
            q = p.height / 2;
            t = d.y - p.height / 2;
            s = d.x - z;
        }
        else {
            q = p.height + DeviceUtils.g(10) + DeviceUtils.g(5);
            if (u) {
                q = 0 - (y + DeviceUtils.g(5));
            }
            r = p.width / 2;
            s = d.x - p.width / 2;
            t = d.y - z;
        }
        if (j) {
            if (t < 0) {
                q -= 0 - t;
                t = 0;
            }
            if (t + p.height > x.height) {
                q += (t + p.height) - x.height;
                t = x.height - p.height;
            }
        }
        else {
            if (s < 0) {
                r -= 0 - s;
                s = 0;
            }
            if (s + p.width > x.width) {
                r += (s + p.width) - x.width;
                s = x.width - p.width;
            }
        }
        b.k = s;
        b.l = t;
        b.g = r;
        b.h = q;
        b.j = p.width;
        b.i = p.height;
    };
    CategoryToolTipLayer.prototype.tg = function (a) {
        return a.toolTip;
    };
    CategoryToolTipLayer.prototype.r8 = function (a, b) {
        _super.prototype.r8.call(this, a, b);
        var c = a;
        var d = b.bu.left;
        var e = b.bu.right;
        var f = b.bu.top;
        var g = b.bu.bottom;
        var h = this.td.ca();
        if (isNaN_(c.k) || isNaN_(c.l) || isNaN_(c.g) || isNaN_(c.h)) {
            this.td.cd(h);
            return;
        }
        var i = c.k;
        var j = c.l;
        var k = c.g;
        var l = c.h;
        if (h != null) {
            this.td.ce(h, i, j, k, l);
        }
    };
    CategoryToolTipLayer.prototype.ep = function () {
        return true;
    };
    CategoryToolTipLayer.$t = markType(CategoryToolTipLayer, 'CategoryToolTipLayer', AnnotationLayer.$);
    CategoryToolTipLayer.$$p = markDep(DependencyProperty, PropertyMetadata, CategoryToolTipLayer, 'raisePropertyChanged', ['TargetAxis:tb:tl', [Axis.$, null], 'ToolTipPosition:tc:tm', [CategoryTooltipLayerPosition_$type, enumGetBox(CategoryTooltipLayerPosition_$type, 0)], 'UseInterpolation:tf:tn', [0, false]]);
    return CategoryToolTipLayer;
}(AnnotationLayer));
export { CategoryToolTipLayer };
/**
 * @hidden
 */
var CategoryToolTipLayerView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryToolTipLayerView, _super);
    function CategoryToolTipLayerView(a) {
        var _this = _super.call(this, a) || this;
        _this.b4 = null;
        _this.b5 = null;
        _this._tooltips = null;
        _this.b8 = null;
        _this.b6 = false;
        _this._htmlTest = null;
        _this.b4 = a;
        _this.htmlTest = /^[^<]*(<[\w\W]+>)[^>]*$/;
        return _this;
    }
    CategoryToolTipLayerView.prototype.ca = function () {
        if (this.b5 == null) {
            this.b5 = new PointerTooltip();
            this.b5.bw(this.e.he);
            this.b5.ap = this.e.ce.cu.k;
            this._tooltips = this.e.he.createElement("div");
            this.b8 = this.e.he.createElement("div");
            this.b5._visibility = 0;
        }
        return this.b5;
    };
    CategoryToolTipLayerView.prototype.as = function () {
        _super.prototype.as.call(this);
        if (this.b5 != null) {
            this.b5._visibility = 1;
        }
    };
    CategoryToolTipLayerView.prototype.aj = function () {
        _super.prototype.aj.call(this);
        if (this.b5 != null) {
            this.b5.br();
            this.b5 = null;
        }
    };
    Object.defineProperty(CategoryToolTipLayerView.prototype, "htmlTest", {
        get: function () {
            return this._htmlTest;
        },
        set: function (a) {
            this._htmlTest = a;
        },
        enumerable: false,
        configurable: true
    });
    CategoryToolTipLayerView.prototype.b7 = function (a, b, c) {
        if (a.h7 == "default") {
            a.h7 = a.b1.getDefaultTooltipTemplate();
        }
        var template_ = a.h7;
        var context_ = c;
        if (a.toolTip == null) {
            return false;
        }
        var tooltip_ = a.toolTip;
        if ((tooltip_.updateToolTip)) {
            tooltip_._inPointer = true;
            tooltip_.updateToolTip(context_, true);
            tooltip_.setStyleProperty('position', 'relative');
            tooltip_.setStyleProperty('left', '');
            tooltip_.setStyleProperty('top', '');
            tooltip_.remove();
        }
        this._tooltips.append(a.toolTip);
        return true;
    };
    CategoryToolTipLayerView.prototype.ch = function (a) {
        var b = a;
        b.content = this._tooltips;
        var c = new Size(1, Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY);
        b.cy = 1;
        var d = b.cx(new Size(1, Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY));
        b.cy = 0;
        return d;
    };
    CategoryToolTipLayerView.prototype.cg = function (a) {
        var b = 0;
        var c = 0;
        if (this.e.ce != null) {
            var d = this.e.ce.s3();
            b += d.x;
            c += d.y;
        }
        return { $type: Point_$type, x: b + a.x, y: c + a.y };
    };
    CategoryToolTipLayerView.prototype.ce = function (a, b, c, d, e) {
        var f = a;
        var g = f.content;
        f.cs = { $type: Point_$type, x: d, y: e };
        f._visibility = 0;
        f.n = b;
        f.o = c;
    };
    CategoryToolTipLayerView.prototype.ci = function () {
        var a = (window.innerWidth);
        var b = (window.innerHeight);
        return new Size(1, a, b);
    };
    CategoryToolTipLayerView.prototype.cd = function (a) {
        this._tooltips.removeChildren();
        this.b8.removeChildren();
        this.b5._visibility = 1;
    };
    CategoryToolTipLayerView.prototype.cc = function () {
        if (this._tooltips != null) {
            this._tooltips.removeChildren();
        }
    };
    CategoryToolTipLayerView.prototype.cf = function () {
    };
    CategoryToolTipLayerView.prototype.a6 = function (a, b) {
        _super.prototype.a6.call(this, a, b);
        if (b) {
            return;
        }
        var c = null;
        if (this.e.ce != null) {
            c = this.e.ce.cu.j.i();
        }
        var d = 0;
        var e = 0;
        var f = this.e.ce;
        if (f != null) {
            var g = f.cu.j.b7();
            d += g.x;
            e += g.y;
        }
        if (this.b5 != null) {
            this.b5.aq = c;
            this.b5.by(this.b5.n, this.b5.o);
        }
    };
    CategoryToolTipLayerView.prototype.al = function (a) {
        _super.prototype.al.call(this, a);
        var b = this.b5;
        var c = 0;
        var d = 0;
        if (this.e.ce != null) {
            var e = this.e.ce.s3();
            c += e.x;
            d += e.y;
        }
        var f = b.a5();
        var g = b.n - c;
        var h = b.o - d;
        var i = new List$1(String_$type, 0);
        var j = this._tooltips.getChildCount();
        for (var k = 0; k < j; k++) {
            var indexliteral_ = k.toString();
            var l = this._tooltips.getChildAt(k).getText();
            var m = l.toString();
            i.add(m);
        }
        this.e.a9.addPointerTooltipData(a, f, g, h, i);
    };
    CategoryToolTipLayerView.$t = markType(CategoryToolTipLayerView, 'CategoryToolTipLayerView', AnnotationLayerView.$);
    return CategoryToolTipLayerView;
}(AnnotationLayerView));
export { CategoryToolTipLayerView };
