// igniteui-react-charts/esm5/lib/CommodityChannelIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { CommodityChannelIndexIndicatorStrategy } from "./CommodityChannelIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var CommodityChannelIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CommodityChannelIndexIndicator, _super);
    function CommodityChannelIndexIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CommodityChannelIndexIndicator.prototype.get_up = function () {
        return new CommodityChannelIndexIndicatorStrategy();
    };
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "up", {
        get: function () {
            return this.get_up();
        },
        enumerable: false,
        configurable: true
    });
    CommodityChannelIndexIndicator.prototype.get_u2 = function () {
        return CommodityChannelIndexIndicator.$;
    };
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "u2", {
        get: function () {
            return this.get_u2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "va", {
        get: function () {
            return typeGetValue(this.c(CommodityChannelIndexIndicator.vb));
        },
        set: function (a) {
            this.h(CommodityChannelIndexIndicator.vb, a);
        },
        enumerable: false,
        configurable: true
    });
    CommodityChannelIndexIndicator.prototype.uv = function () {
        return this.va;
    };
    CommodityChannelIndexIndicator.prototype.u4 = function (a) {
        this.va = a;
    };
    CommodityChannelIndexIndicator.$t = markType(CommodityChannelIndexIndicator, 'CommodityChannelIndexIndicator', StrategyBasedIndicator.$);
    CommodityChannelIndexIndicator.vb = StrategyBasedIndicator.u7(20, CommodityChannelIndexIndicator.$);
    return CommodityChannelIndexIndicator;
}(StrategyBasedIndicator));
export { CommodityChannelIndexIndicator };
