// igniteui-react-charts/esm5/lib/DetrendedPriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { DetrendedPriceOscillatorIndicatorStrategy } from "./DetrendedPriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var DetrendedPriceOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DetrendedPriceOscillatorIndicator, _super);
    function DetrendedPriceOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DetrendedPriceOscillatorIndicator.prototype.get_up = function () {
        return new DetrendedPriceOscillatorIndicatorStrategy();
    };
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "up", {
        get: function () {
            return this.get_up();
        },
        enumerable: false,
        configurable: true
    });
    DetrendedPriceOscillatorIndicator.prototype.get_u2 = function () {
        return DetrendedPriceOscillatorIndicator.$;
    };
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "u2", {
        get: function () {
            return this.get_u2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "va", {
        get: function () {
            return typeGetValue(this.c(DetrendedPriceOscillatorIndicator.vb));
        },
        set: function (a) {
            this.h(DetrendedPriceOscillatorIndicator.vb, a);
        },
        enumerable: false,
        configurable: true
    });
    DetrendedPriceOscillatorIndicator.prototype.uv = function () {
        return this.va;
    };
    DetrendedPriceOscillatorIndicator.prototype.u4 = function (a) {
        this.va = a;
    };
    DetrendedPriceOscillatorIndicator.$t = markType(DetrendedPriceOscillatorIndicator, 'DetrendedPriceOscillatorIndicator', StrategyBasedIndicator.$);
    DetrendedPriceOscillatorIndicator.vb = StrategyBasedIndicator.u7(20, DetrendedPriceOscillatorIndicator.$);
    return DetrendedPriceOscillatorIndicator;
}(StrategyBasedIndicator));
export { DetrendedPriceOscillatorIndicator };
