// igniteui-react-charts/esm5/lib/FastStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FastStochasticOscillatorIndicatorStrategy } from "./FastStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
var FastStochasticOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FastStochasticOscillatorIndicator, _super);
    function FastStochasticOscillatorIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 3;
        return _this;
    }
    FastStochasticOscillatorIndicator.prototype.get_up = function () {
        return new FastStochasticOscillatorIndicatorStrategy();
    };
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "up", {
        get: function () {
            return this.get_up();
        },
        enumerable: false,
        configurable: true
    });
    FastStochasticOscillatorIndicator.prototype.get_u2 = function () {
        return FastStochasticOscillatorIndicator.$;
    };
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "u2", {
        get: function () {
            return this.get_u2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "va", {
        get: function () {
            return typeGetValue(this.c(FastStochasticOscillatorIndicator.vb));
        },
        set: function (a) {
            this.h(FastStochasticOscillatorIndicator.vb, a);
        },
        enumerable: false,
        configurable: true
    });
    FastStochasticOscillatorIndicator.prototype.uv = function () {
        return this.va;
    };
    FastStochasticOscillatorIndicator.prototype.u4 = function (a) {
        this.va = a;
    };
    FastStochasticOscillatorIndicator.$t = markType(FastStochasticOscillatorIndicator, 'FastStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FastStochasticOscillatorIndicator.vb = StrategyBasedIndicator.u7(14, FastStochasticOscillatorIndicator.$);
    return FastStochasticOscillatorIndicator;
}(StrategyBasedIndicator));
export { FastStochasticOscillatorIndicator };
