// igniteui-react-charts/esm5/lib/ForceIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ForceIndexIndicatorStrategy } from "./ForceIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var ForceIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ForceIndexIndicator, _super);
    function ForceIndexIndicator() {
        return _super.call(this) || this;
    }
    ForceIndexIndicator.prototype.get_tw = function () {
        return 1;
    };
    ForceIndexIndicator.prototype.get_up = function () {
        return new ForceIndexIndicatorStrategy();
    };
    Object.defineProperty(ForceIndexIndicator.prototype, "up", {
        get: function () {
            return this.get_up();
        },
        enumerable: false,
        configurable: true
    });
    ForceIndexIndicator.prototype.get_u2 = function () {
        return ForceIndexIndicator.$;
    };
    Object.defineProperty(ForceIndexIndicator.prototype, "u2", {
        get: function () {
            return this.get_u2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ForceIndexIndicator.prototype, "va", {
        get: function () {
            return typeGetValue(this.c(ForceIndexIndicator.vb));
        },
        set: function (a) {
            this.h(ForceIndexIndicator.vb, a);
        },
        enumerable: false,
        configurable: true
    });
    ForceIndexIndicator.prototype.uv = function () {
        return this.va;
    };
    ForceIndexIndicator.prototype.u4 = function (a) {
        this.va = a;
    };
    ForceIndexIndicator.$t = markType(ForceIndexIndicator, 'ForceIndexIndicator', StrategyBasedIndicator.$);
    ForceIndexIndicator.vb = StrategyBasedIndicator.u7(0, ForceIndexIndicator.$);
    return ForceIndexIndicator;
}(StrategyBasedIndicator));
export { ForceIndexIndicator };
