// igniteui-react-charts/esm5/lib/FullStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FullStochasticOscillatorIndicatorStrategy } from "./FullStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
var FullStochasticOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FullStochasticOscillatorIndicator, _super);
    function FullStochasticOscillatorIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 3;
        return _this;
    }
    FullStochasticOscillatorIndicator.prototype.get_up = function () {
        return new FullStochasticOscillatorIndicatorStrategy();
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "up", {
        get: function () {
            return this.get_up();
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.get_u2 = function () {
        return FullStochasticOscillatorIndicator.$;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "u2", {
        get: function () {
            return this.get_u2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "va", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.vf));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.vf, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.uv = function () {
        return this.va;
    };
    FullStochasticOscillatorIndicator.prototype.u4 = function (a) {
        this.va = a;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "smoothingPeriod", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.vg));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.vg, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.ux = function () {
        return this.smoothingPeriod;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "vc", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.vh));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.vh, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.uu = function () {
        return this.vc;
    };
    FullStochasticOscillatorIndicator.prototype.t5 = function () {
        return this.vc;
    };
    FullStochasticOscillatorIndicator.$t = markType(FullStochasticOscillatorIndicator, 'FullStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FullStochasticOscillatorIndicator.vf = StrategyBasedIndicator.u7(14, FullStochasticOscillatorIndicator.$);
    FullStochasticOscillatorIndicator.vg = StrategyBasedIndicator.u8(3, FullStochasticOscillatorIndicator.$, "SmoothingPeriod");
    FullStochasticOscillatorIndicator.vh = StrategyBasedIndicator.u8(3, FullStochasticOscillatorIndicator.$, "TriggerPeriod");
    return FullStochasticOscillatorIndicator;
}(StrategyBasedIndicator));
export { FullStochasticOscillatorIndicator };
