// igniteui-react-charts/esm5/lib/igr-value-overlay.js
import { __extends } from "tslib";
import { IgrAxis } from "./igr-axis";
import { IgrSeries } from "./igr-series";
import { ValueOverlay } from "./ValueOverlay";
import { ensureBool, brushToString, stringToBrush, toPoint, fromPoint } from "igniteui-react-core";
/**
 * Represents the class of the value overlay. The value overlay is a line or circle representing a value on an axis.
*/
var IgrValueOverlay = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgrValueOverlay, _super);
    function IgrValueOverlay(props) {
        var _this = _super.call(this, props) || this;
        _this._axisName = null;
        return _this;
    }
    IgrValueOverlay.prototype.createImplementation = function () {
        return new ValueOverlay();
    };
    Object.defineProperty(IgrValueOverlay.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axis", {
        /**
         * Gets or sets the axis used by the value overlay.
        */
        get: function () {
            var r = this.i.qc;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgrAxis._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.qc = null : this.i.qc = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisName", {
        /**
        * Gets or sets the name to use to resolve axis from markup.
        */
        get: function () {
            return this._axisName;
        },
        set: function (v) {
            this._axisName = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "value", {
        /**
         * Gets or sets the value of the overlay.
        */
        get: function () {
            return this.i.qv;
        },
        set: function (v) {
            this.i.qv = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "isAxisAnnotationEnabled", {
        /**
         * Gets or sets whether to draw annotations over the axes where the crosshair meets with them.
        */
        get: function () {
            return this.i.qg;
        },
        set: function (v) {
            this.i.qg = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationFormatLabel", {
        /**
         * Sets or gets a function which takes an object that produces a formatted label for the axis annotation.
        */
        get: function () {
            return this.i.qe;
        },
        set: function (v) {
            this.i.qe = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationTextColor", {
        /**
         * Gets or sets the color to use for the axis annotation text. Leave unset for an automatic value.
        */
        get: function () {
            return brushToString(this.i.rv);
        },
        set: function (v) {
            this.i.rv = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationBackground", {
        /**
         * Gets or sets the color to use for the axis annotation backing. Leave unset for an automatic value.
        */
        get: function () {
            return brushToString(this.i.rt);
        },
        set: function (v) {
            this.i.rt = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationBackgroundCornerRadius", {
        /**
         * Gets or sets the corner radius to use for the axis annotation backing. Leave unset for an automatic value.
        */
        get: function () {
            return this.i.qk;
        },
        set: function (v) {
            this.i.qk = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationInterpolatedValuePrecision", {
        /**
         * Gets or sets the precision to use displaying values for interpolated positions.
        */
        get: function () {
            return this.i.qw;
        },
        set: function (v) {
            this.i.qw = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationOutline", {
        /**
         * Gets or sets the color to use for the axis annotation outline. Leave unset for an automatic value.
        */
        get: function () {
            return brushToString(this.i.ru);
        },
        set: function (v) {
            this.i.ru = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationPaddingLeft", {
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.qm;
        },
        set: function (v) {
            this.i.qm = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationPaddingTop", {
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.qo;
        },
        set: function (v) {
            this.i.qo = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationPaddingRight", {
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.qn;
        },
        set: function (v) {
            this.i.qn = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationPaddingBottom", {
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.ql;
        },
        set: function (v) {
            this.i.ql = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "axisAnnotationStrokeThickness", {
        /**
         * Gets or sets the stroke thickness for the axis annotation backing. Leave unset for an automatic value.
        */
        get: function () {
            return this.i.qp;
        },
        set: function (v) {
            this.i.qp = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "isValueOverlay", {
        /**
         * Gets whether the series is a value overlay.
        */
        get: function () {
            return this.i.isValueOverlay;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrValueOverlay.prototype, "labelResolved", {
        get: function () {
            return this.i.rb;
        },
        enumerable: false,
        configurable: true
    });
    IgrValueOverlay.prototype.bindAxes = function (axes) {
        _super.prototype.bindAxes.call(this, axes);
        for (var i = 0; i < axes.length; i++) {
            if (this.axisName && this.axisName.length > 0 &&
                axes[i].name == this.axisName) {
                this.axis = axes[i];
            }
        }
    };
    IgrValueOverlay.prototype.findByName = function (name) {
        var baseResult = _super.prototype.findByName.call(this, name);
        if (baseResult) {
            return baseResult;
        }
        if (this.axis && this.axis.name && this.axis.name == name) {
            return this.axis;
        }
        return null;
    };
    IgrValueOverlay.prototype._styling = function (container, component, parent) {
        _super.prototype._styling.call(this, container, component, parent);
        this._inStyling = true;
        if (this.axis && this.axis._styling) {
            this.axis._styling(container, component, this);
        }
        this._inStyling = false;
    };
    IgrValueOverlay.prototype.getSeriesValue = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.gn(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    };
    IgrValueOverlay.prototype.getSeriesValuePosition = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.ps(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    };
    /**
     * Gets the item index associated with the specified world position
    
    * @param world
    */
    IgrValueOverlay.prototype.getItemIndex = function (world) {
        var iv = this.i.hi(toPoint(world));
        return (iv);
    };
    /**
     * Gets the item that is the best match for the specified world coordinates.
    
    * @param world  * The world coordinates to use.
    */
    IgrValueOverlay.prototype.getItem = function (world) {
        var iv = this.i.hy(toPoint(world));
        return (iv);
    };
    /**
     * Requests that the provided item should be brought into view if possible.
    
    * @param item  * The item to attempt to bring into view.
    */
    IgrValueOverlay.prototype.scrollIntoView = function (item) {
        var iv = this.i.eq(item);
        return (iv);
    };
    /**
     * Gets the label for a data item.
    
    * @param value  * The unscaled value to get a label for.
    */
    IgrValueOverlay.prototype.getLabel = function (value) {
        var iv = this.i.q9(value);
        return (iv);
    };
    return IgrValueOverlay;
}(IgrSeries));
export { IgrValueOverlay };
