// igniteui-react-charts/esm5/lib/MoneyFlowIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { MoneyFlowIndexIndicatorStrategy } from "./MoneyFlowIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var MoneyFlowIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(MoneyFlowIndexIndicator, _super);
    function MoneyFlowIndexIndicator() {
        return _super.call(this) || this;
    }
    MoneyFlowIndexIndicator.prototype.get_tw = function () {
        return 1;
    };
    MoneyFlowIndexIndicator.prototype.get_up = function () {
        return new MoneyFlowIndexIndicatorStrategy();
    };
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "up", {
        get: function () {
            return this.get_up();
        },
        enumerable: false,
        configurable: true
    });
    MoneyFlowIndexIndicator.prototype.get_u2 = function () {
        return MoneyFlowIndexIndicator.$;
    };
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "u2", {
        get: function () {
            return this.get_u2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "va", {
        get: function () {
            return typeGetValue(this.c(MoneyFlowIndexIndicator.vb));
        },
        set: function (a) {
            this.h(MoneyFlowIndexIndicator.vb, a);
        },
        enumerable: false,
        configurable: true
    });
    MoneyFlowIndexIndicator.prototype.uv = function () {
        return this.va;
    };
    MoneyFlowIndexIndicator.prototype.u4 = function (a) {
        this.va = a;
    };
    MoneyFlowIndexIndicator.$t = markType(MoneyFlowIndexIndicator, 'MoneyFlowIndexIndicator', StrategyBasedIndicator.$);
    MoneyFlowIndexIndicator.vb = StrategyBasedIndicator.u7(14, MoneyFlowIndexIndicator.$);
    return MoneyFlowIndexIndicator;
}(StrategyBasedIndicator));
export { MoneyFlowIndexIndicator };
