// igniteui-react-charts/esm5/lib/PointSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { HorizontalAnchoredCategorySeries } from "./HorizontalAnchoredCategorySeries";
import { markType } from "igniteui-react-core";
import { AnchoredCategorySeriesView } from "./AnchoredCategorySeriesView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
var PointSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PointSeries, _super);
    function PointSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PointSeries.$;
        return _this;
    }
    PointSeries.prototype.get_c5 = function () {
        return true;
    };
    PointSeries.prototype.by = function () {
        return new PointSeriesView(this);
    };
    PointSeries.prototype.l4 = function (a) {
        _super.prototype.l4.call(this, a);
        this.vj = a;
    };
    PointSeries.prototype.g0 = function () {
        return 0.7;
    };
    PointSeries.prototype.qf = function () {
        if (this.qe == 0) {
            return 2;
        }
        return _super.prototype.qf.call(this);
    };
    PointSeries.prototype.ez = function (a, b) {
        var c = this.ps(this.px(a), true, false);
        if (this.qo(a, b)) {
            return true;
        }
        return false;
    };
    PointSeries.prototype.l9 = function (a, b, c, d) {
        _super.prototype.l9.call(this, a, b, c, d);
    };
    Object.defineProperty(PointSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.ss;
        },
        enumerable: false,
        configurable: true
    });
    PointSeries.$t = markType(PointSeries, 'PointSeries', HorizontalAnchoredCategorySeries.$);
    return PointSeries;
}(HorizontalAnchoredCategorySeries));
export { PointSeries };
/**
 * @hidden
 */
var PointSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PointSeriesView, _super);
    function PointSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.da = null;
        _this.da = a;
        return _this;
    }
    PointSeriesView.prototype.as = function () {
        _super.prototype.as.call(this);
        if (!this.q) {
            this.b1.ql = Defaults.e;
        }
    };
    PointSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.ph = a;
        this.e.g1 = 3;
        this.e.g2 = 2;
        this.e.g3 = 2;
        this.e.fb = false;
    };
    PointSeriesView.$t = markType(PointSeriesView, 'PointSeriesView', AnchoredCategorySeriesView.$);
    return PointSeriesView;
}(AnchoredCategorySeriesView));
export { PointSeriesView };
