// igniteui-react-charts/esm5/lib/PolarLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { UnknownValuePlotting_$type } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { runOn, enumGetBox, markType, markDep } from "igniteui-react-core";
import { Path } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var PolarLineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarLineSeries, _super);
    function PolarLineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarLineSeries.$;
        return _this;
    }
    PolarLineSeries.prototype.get_d4 = function () {
        return true;
    };
    PolarLineSeries.prototype.by = function () {
        return new PolarLineSeriesView(this);
    };
    PolarLineSeries.prototype.l4 = function (a) {
        _super.prototype.l4.call(this, a);
        this.vo = a;
    };
    PolarLineSeries.prototype.vm = function (a) {
        var b = a;
        b.c5();
    };
    PolarLineSeries.prototype.vn = function (a, b) {
        var c = b;
        c.c6(a.m, this.fx);
        this.se.az(this, this.s1(), runOn(this, this.rr));
        var d = false;
        var e = this.se.b;
        if (e != null) {
            d = true;
        }
        if (d) {
            this.t9(-1, this.sn.count, b.q);
        }
        this.se.ad(b.c8, true, false, true, false);
    };
    PolarLineSeries.prototype.ve = function () {
        return this.vp;
    };
    PolarLineSeries.prototype.l9 = function (a, b, c, d) {
        _super.prototype.l9.call(this, a, b, c, d);
        switch (b) {
            case PolarLineSeries.$$p[0]:
                this.mn(false);
                this.lu();
                break;
        }
    };
    PolarLineSeries.$t = markType(PolarLineSeries, 'PolarLineSeries', PolarLineSeriesBase.$);
    PolarLineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, PolarLineSeries, 'raisePropertyChanged', ['UnknownValuePlotting:vp:vr', [UnknownValuePlotting_$type, enumGetBox(UnknownValuePlotting_$type, 1)]]);
    return PolarLineSeries;
}(PolarLineSeriesBase));
export { PolarLineSeries };
/**
 * @hidden
 */
var PolarLineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarLineSeriesView, _super);
    function PolarLineSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c4 = null;
        _this.c8 = new Path();
        _this.c7 = new Path();
        _this.c4 = a;
        return _this;
    }
    PolarLineSeriesView.prototype.as = function () {
        _super.prototype.as.call(this);
        if (!this.q) {
            this.b1.ql = Defaults.e;
        }
    };
    PolarLineSeriesView.prototype.c5 = function () {
        this.c8.an = null;
        this.c7.an = null;
        this.aq();
    };
    PolarLineSeriesView.prototype.c6 = function (a, b) {
        this.c4.vb.d(this.c8, a, b);
        this.aq();
    };
    PolarLineSeriesView.prototype.bb = function () {
        _super.prototype.bb.call(this);
    };
    PolarLineSeriesView.prototype.bc = function () {
        _super.prototype.bc.call(this);
        var a = this.bo();
        this.c7.an = this.c8.an;
        this.c7._stroke = a;
        this.c7.ad = this.e.fy + 3;
    };
    PolarLineSeriesView.prototype.a6 = function (a, b) {
        _super.prototype.a6.call(this, a, b);
        if (a.d) {
            if (b) {
                a.t(this.c7);
            }
            else {
                a.t(this.c8);
            }
        }
    };
    PolarLineSeriesView.prototype.al = function (a) {
        _super.prototype.al.call(this, a);
        this.e.a9.exportPathData(a, this.c8, "polarShape", ["Main"]);
    };
    PolarLineSeriesView.$t = markType(PolarLineSeriesView, 'PolarLineSeriesView', PolarLineSeriesBaseView.$);
    return PolarLineSeriesView;
}(PolarLineSeriesBaseView));
export { PolarLineSeriesView };
