// igniteui-react-charts/esm5/lib/PolarSplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { DependencyProperty } from "igniteui-react-core";
import { runOn, markType, markDep } from "igniteui-react-core";
import { Path } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { PathGeometry } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { Extensions } from "igniteui-react-core";
import { Numeric } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
var PolarSplineAreaSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarSplineAreaSeries, _super);
    function PolarSplineAreaSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarSplineAreaSeries.$;
        return _this;
    }
    PolarSplineAreaSeries.prototype.get_df = function () {
        return true;
    };
    PolarSplineAreaSeries.prototype.get_ej = function () {
        return true;
    };
    PolarSplineAreaSeries.prototype.by = function () {
        return new PolarSplineAreaSeriesView(this);
    };
    PolarSplineAreaSeries.prototype.l4 = function (a) {
        _super.prototype.l4.call(this, a);
        this.vo = a;
    };
    PolarSplineAreaSeries.prototype.vm = function (a) {
        var b = a;
        b.c5();
    };
    PolarSplineAreaSeries.prototype.vn = function (a, b) {
        this.vm(b);
        if (a.m.count < 1) {
            return;
        }
        var c = b;
        c.c6(a.m, this.vp);
        this.se.az(this, this.s1(), runOn(this, this.rr));
        var d = false;
        var e = this.se.b;
        if (e != null) {
            d = true;
        }
        if (d) {
            this.t9(-1, this.sn.count, b.q);
        }
        var f = c.dc;
        var g = c.db;
        this.se.ad(f, true, false, true, true);
        this.se.ad(g, false, true, false, false);
        g._opacity = this.se.j * this.fu;
    };
    PolarSplineAreaSeries.prototype.get_vj = function () {
        return true;
    };
    PolarSplineAreaSeries.prototype.l9 = function (a, b, c, d) {
        _super.prototype.l9.call(this, a, b, c, d);
        switch (b) {
            case PolarSplineAreaSeries.$$p[0]:
                this.mn(false);
                break;
        }
    };
    PolarSplineAreaSeries.$t = markType(PolarSplineAreaSeries, 'PolarSplineAreaSeries', PolarLineSeriesBase.$);
    PolarSplineAreaSeries.$$p = markDep(DependencyProperty, PropertyMetadata, PolarSplineAreaSeries, 'raisePropertyChanged', ['Stiffness:vp:vr', [1, 0.5]]);
    return PolarSplineAreaSeries;
}(PolarLineSeriesBase));
export { PolarSplineAreaSeries };
/**
 * @hidden
 */
var PolarSplineAreaSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarSplineAreaSeriesView, _super);
    function PolarSplineAreaSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c4 = null;
        _this.da = ((function () {
            var $ret = new Path();
            $ret.an = new PathGeometry();
            return $ret;
        })());
        _this.c9 = ((function () {
            var $ret = new Path();
            $ret.an = new PathGeometry();
            return $ret;
        })());
        _this.c8 = ((function () {
            var $ret = new Path();
            $ret.an = new PathGeometry();
            return $ret;
        })());
        _this.c7 = ((function () {
            var $ret = new Path();
            $ret.an = new PathGeometry();
            return $ret;
        })());
        _this.c4 = a;
        return _this;
    }
    PolarSplineAreaSeriesView.prototype.as = function () {
        _super.prototype.as.call(this);
        if (!this.q) {
            this.c4.ql = Defaults.e;
        }
    };
    Object.defineProperty(PolarSplineAreaSeriesView.prototype, "db", {
        get: function () {
            return this.c9;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarSplineAreaSeriesView.prototype, "dc", {
        get: function () {
            return this.da;
        },
        enumerable: false,
        configurable: true
    });
    PolarSplineAreaSeriesView.prototype.c5 = function () {
        Extensions.e(this.c9.an);
        Extensions.e(this.da.an);
        Extensions.e(this.c7.an);
        Extensions.e(this.c8.an);
        this.aq();
    };
    PolarSplineAreaSeriesView.prototype.c6 = function (a, b) {
        this.da.an.c = Numeric.h(a.count + 1, function (c) { return c < a.count ? a._inner[c].x : a._inner[c - a.count].x; }, function (c) { return c < a.count ? a._inner[c].y : a._inner[c - a.count].y; }, b);
        this.c9.an.c = Numeric.h(a.count + 1, function (c) { return c < a.count ? a._inner[c].x : a._inner[c - a.count].x; }, function (c) { return c < a.count ? a._inner[c].y : a._inner[c - a.count].y; }, b);
        this.aq();
    };
    PolarSplineAreaSeriesView.prototype.bb = function () {
        _super.prototype.bb.call(this);
    };
    PolarSplineAreaSeriesView.prototype.bc = function () {
        _super.prototype.bc.call(this);
        var a = this.bo();
        this.c7.an = this.c9.an;
        this.c8.an = this.da.an;
        this.c7._fill = a;
        this.c7._opacity = 1;
        this.c8._stroke = a;
        this.c8.ad = this.e.fy + 3;
    };
    PolarSplineAreaSeriesView.prototype.a6 = function (a, b) {
        _super.prototype.a6.call(this, a, b);
        if (a.d) {
            if (b) {
                a.t(this.c7);
                a.t(this.c8);
            }
            else {
                a.t(this.c9);
                a.t(this.da);
            }
        }
    };
    PolarSplineAreaSeriesView.prototype.al = function (a) {
        _super.prototype.al.call(this, a);
        this.e.a9.exportPathData(a, this.da, "polarShape", ["Main"]);
        this.e.a9.exportPathData(a, this.c9, "fillShape", ["Fill"]);
    };
    PolarSplineAreaSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.ph = a;
        this.e.g1 = 5;
        this.e.g2 = 5;
        this.e.g3 = 5;
    };
    PolarSplineAreaSeriesView.$t = markType(PolarSplineAreaSeriesView, 'PolarSplineAreaSeriesView', PolarLineSeriesBaseView.$);
    return PolarSplineAreaSeriesView;
}(PolarLineSeriesBaseView));
export { PolarSplineAreaSeriesView };
