// igniteui-react-charts/esm5/lib/RadialBucketCalculator_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { Base, markType, runOn, EnumUtil, typeCast, fromEnum, markDep } from "igniteui-react-core";
import { IBucketizer_$type } from "./IBucketizer";
import { NotImplementedException } from "igniteui-react-core";
import { CategoryAngleAxis } from "./CategoryAngleAxis";
import { NumericRadiusAxis } from "./NumericRadiusAxis";
import { truncate, intDivide, isNaN_ } from "igniteui-react-core";
import { MarkerSeries } from "./MarkerSeries";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { SeriesRenderer$2 } from "./SeriesRenderer$2";
import { RadialFrame } from "./RadialFrame";
import { DependencyProperty } from "igniteui-react-core";
import { HighlightingInfo } from "./HighlightingInfo";
import { RadialSeriesRenderManager } from "./RadialSeriesRenderManager";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { ScalerParams } from "./ScalerParams";
import { Marker } from "./Marker";
import { MarkerSeriesView } from "./MarkerSeriesView";
import { RadialAxes } from "./RadialAxes";
import { Visibility_$type } from "igniteui-react-core";
import { SeriesRenderingArguments } from "./SeriesRenderingArguments";
import { PropertyMetadata } from "igniteui-react-core";
import { round10N } from "igniteui-react-core";
import { Pool$1 } from "igniteui-react-core";
/**
 * @hidden
 */
var RadialBucketCalculator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialBucketCalculator, _super);
    function RadialBucketCalculator(a) {
        var _this = _super.call(this) || this;
        _this.a = null;
        _this.c = 0;
        _this.d = 0;
        _this.b = 0;
        _this.a = a;
        return _this;
    }
    RadialBucketCalculator.prototype.getBucket = function (a) {
        throw new NotImplementedException(0);
    };
    RadialBucketCalculator.prototype.getBucketWithoutUnknowns = function (a) {
        return this.getBucket(a);
    };
    RadialBucketCalculator.prototype.getErrorBucket = function (a, b) {
        return NaN;
    };
    RadialBucketCalculator.prototype.getBucketInfo = function (a, b, c, d) {
        a = this.c;
        b = this.d;
        c = this.b;
        d = this.a.cr.fx;
        return {
            p0: a,
            p1: b,
            p2: c,
            p3: d
        };
    };
    RadialBucketCalculator.prototype.e = function (a) {
        var b = this.a.bv;
        var c = this.a.bu;
        var d = this.a.cr.rt;
        if (b.isEmpty || c.isEmpty || d == null || this.a.cr.cl == null || this.a.cr.cl.count == 0) {
            this.b = 0;
            return;
        }
        var e = Math.floor(d.mw(c, b));
        var f = Math.ceil(d.mv(c, b));
        if (d.cg) {
            f = Math.ceil(d.mw(c, b));
            e = Math.floor(d.mv(c, b));
        }
        if (f < e) {
            f = d.le + f;
        }
        var g = this.a.cr.r0;
        var h = g != null ? g.nn : 0.75;
        var i = Math.min(c.width, c.height) * 0.5 * (h) * 2 * Math.PI;
        var j = Math.floor((f - e + 1) * a / i);
        this.b = truncate(Math.max(1, j));
        this.c = truncate(Math.max(0, Math.floor(e / this.b) - 1));
        this.d = truncate(Math.ceil(f / this.b));
    };
    RadialBucketCalculator.prototype.cacheValues = function () {
    };
    RadialBucketCalculator.prototype.unCacheValues = function () {
    };
    RadialBucketCalculator.$t = markType(RadialBucketCalculator, 'RadialBucketCalculator', Base.$, [IBucketizer_$type]);
    return RadialBucketCalculator;
}(Base));
export { RadialBucketCalculator };
/**
 * @hidden
 */
var RadialBase = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialBase, _super);
    function RadialBase() {
        var _this = _super.call(this) || this;
        _this.r7 = null;
        _this.r8 = null;
        _this.sa = false;
        _this.r9 = false;
        _this.ry = null;
        _this.rz = null;
        _this.assigningRadialStyle = null;
        _this.assigningRadialMarkerStyle = null;
        _this.r5 = new RadialFrame(3);
        _this.r6 = new RadialFrame(3);
        _this.r3 = new RadialFrame(3);
        _this.r1 = null;
        _this.r7 = new RadialSeriesRenderManager();
        _this.r8 = new SeriesRenderer$2(RadialFrame.$, RadialBaseView.$, 1, runOn(_this, _this.sv), runOn(_this, _this.sy), runOn(_this, _this.cw), runOn(_this, _this.mw), runOn(_this, _this.su), function (a) { return _this.r2.cs.e(_this.fx); });
        return _this;
    }
    RadialBase.prototype.l4 = function (a) {
        _super.prototype.l4.call(this, a);
        this.r2 = a;
    };
    RadialBase.prototype.su = function () {
        if (this.cf.g) {
            this.cf.t();
        }
    };
    Object.defineProperty(RadialBase.prototype, "categoryAxis", {
        get: function () {
            return this.rt;
        },
        enumerable: false,
        configurable: true
    });
    RadialBase.prototype.as = function () {
        return this.rt;
    };
    RadialBase.prototype.rq = function (a, b) {
        return null;
    };
    RadialBase.prototype.c2 = function (a, b, c, d) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var e = this.rt;
        var f = new ScalerParams(1, this.ce.ta, this.p7, e.cg);
        f.c = this.getEffectiveViewport1(this.b1);
        var g = this.r0;
        var h = new ScalerParams(1, this.ce.ta, this.p7, g.cg);
        h.c = this.getEffectiveViewport1(this.b1);
        a.clear();
        this.qk.cf(function (i) {
            if (i._visibility == 0) {
                var j = new SeriesPointOfInterest();
                var k = i.n;
                var l = i.o;
                if (k < _this.p7.left || k > _this.p7.right || l < _this.p7.top || l > _this.p7.bottom) {
                    return;
                }
                var m = void 0;
                var n = void 0;
                var o = _this.r1.g(k, l, _this.b1.bv, _this.b1.bu, m, n);
                m = o.p4;
                n = o.p5;
                j.b = m;
                j.c = n;
                if (b) {
                    if (c) {
                        j.d = _this.sp(j.c, d);
                    }
                    else {
                        var p = _this.gt(i);
                        if (!isNaN_(p)) {
                            j.d = _this.sp(p, d);
                        }
                        else {
                            j.d = _this.sp(j.c, d);
                        }
                    }
                }
                else {
                    j.d = e.gc(i.ak);
                }
                j.a = i.content;
                a.add(j);
            }
        });
        return true;
    };
    RadialBase.prototype.sp = function (a, b) {
        if (b >= 0) {
            return round10N(a, b).toString();
        }
        else {
            return a.toString();
        }
    };
    RadialBase.prototype.get_ed = function () {
        return true;
    };
    RadialBase.prototype.get_d0 = function () {
        return true;
    };
    RadialBase.prototype.rv = function (a, b, c) {
        var d = new HighlightingInfo();
        d.b = this;
        d.h = 0;
        d.g = this.cl.count - 1;
        d.d = true;
        if (c != null && c.h == d.h && c.g == d.g) {
            return c;
        }
        return d;
    };
    RadialBase.prototype.rw = function (a, b, c) {
        var d = this.r2.cs.c;
        var e = this.r2.cs.d;
        var f = this.r2.cs.b;
        var g = this.r4();
        var h = this.cl.count;
        var i = this.rt;
        var j = -1;
        var k = -1;
        var l = this.hi(b);
        j = truncate(Math.floor((intDivide(l, f)))) * f;
        k = j + (f - 1);
        if (c != null && c.h == j && c.g == k) {
            return c;
        }
        var m = new HighlightingInfo();
        m.b = this;
        m.h = j;
        m.g = k;
        return m;
    };
    RadialBase.prototype.rx = function (a, b, c) {
        var d = this.r2.cs.c;
        var e = this.r2.cs.d;
        var f = this.r2.cs.b;
        var g = this.r4();
        var h = this.cl.count;
        var i = this.rt;
        var j = -1;
        var k = -1;
        var l = this.hi(b);
        j = truncate(Math.floor((intDivide(l, f)))) * f;
        k = j + (f - 1);
        if (c != null && c.h == j && c.g == k) {
            return c;
        }
        var m = new HighlightingInfo();
        m.e = true;
        m.b = this;
        m.h = j;
        m.g = k;
        return m;
    };
    RadialBase.prototype.ax = function (a, b, c) {
        if (this.c3 && this.ei(this.bl) && !c) {
            var d = this.rw(a, b, this.ry);
            this.ry = d;
            return d;
        }
        else {
            var e = this.rv(a, b, this.ry);
            this.ry = e;
            return e;
        }
    };
    RadialBase.prototype.ay = function (a, b, c) {
        if (this.ei(this.bl) && !c) {
            var d = this.rx(a, b, this.rz);
            this.rz = d;
            return d;
        }
        else {
            var e = this.rv(a, b, this.rz);
            e.e = true;
            this.rz = e;
            return e;
        }
    };
    RadialBase.prototype.sj = function () {
        return (this.assigningRadialStyle != null && this.sa) || (this.bl != 1);
    };
    RadialBase.prototype.si = function () {
        return (this.assigningRadialMarkerStyle != null && this.r9) || (this.bl != 1);
    };
    RadialBase.prototype.sx = function (a) {
        if (this.assigningRadialStyle != null && this.sa) {
            this.assigningRadialStyle(this, a);
        }
    };
    RadialBase.prototype.sw = function (a) {
        if (this.assigningRadialMarkerStyle != null && this.r9) {
            this.assigningRadialMarkerStyle(this, a);
        }
    };
    RadialBase.prototype.sl = function (a) {
        return a.cs.b;
    };
    RadialBase.prototype.sm = function (a) {
        return a.cs.c;
    };
    RadialBase.prototype.ag = function () {
        return [this.rt, this.r0];
    };
    RadialBase.prototype.nb = function (a, b) {
        this.r2.cs.e(this.fx);
        this.mn(false);
    };
    RadialBase.prototype.na = function (a, b) {
        this.r2.cs.e(this.fx);
        this.mn(false);
    };
    RadialBase.prototype.l9 = function (a, b, c, d) {
        _super.prototype.l9.call(this, a, b, c, d);
        switch (b) {
            case "SeriesViewer":
                if (c != null && d == null) {
                    if (this.rt != null) {
                        this.rt.bz(this);
                    }
                    if (this.r0 != null) {
                        this.r0.bz(this);
                    }
                }
                if (c == null && d != null) {
                    if (this.rt != null) {
                        this.rt.cs(this);
                    }
                    if (this.r0 != null) {
                        this.r0.cs(this);
                    }
                }
                this.r2.cs.e(this.fx);
                this.mn(false);
                break;
            case RadialBase.$$p[0]:
                if (this.rt != null && this.r0 != null) {
                    this.r1 = new RadialAxes(this.r0, this.rt);
                }
                if (c != null) {
                    c.bz(this);
                }
                if (d != null) {
                    d.cs(this);
                }
                this.r2.cs.e(this.fx);
                this.mn(false);
                break;
            case RadialBase.$$p[4]:
                if (this.rt != null && this.r0 != null) {
                    this.r1 = new RadialAxes(this.r0, this.rt);
                }
                if (c != null) {
                    c.bz(this);
                }
                if (d != null) {
                    d.cs(this);
                }
                this.r2.cs.e(this.fx);
                if (this.r0 == null || !this.r0.c0()) {
                    this.mn(false);
                }
                break;
            case "SyncLink":
                if (this.b9 != null && this.ce != null) {
                    this.r2.cs.e(this.fx);
                    this.mn(false);
                }
                break;
            case "TransitionProgress":
                this.r6.c(this.g5, this.r5, this.r3);
                if (this.cy(this.b1)) {
                    return;
                }
                if (this.g5 == 1) {
                    this.sy(this.r3, this.r2);
                }
                else {
                    this.sy(this.r6, this.r2);
                }
                if (this.ce != null) {
                    this.ce.ob(this);
                }
                break;
            case RadialBase.$$p[1]:
                this.mn(false);
                this.lu();
                break;
            case "VisibilityProxy":
                if (EnumUtil.getEnumValue(Visibility_$type, c) != 0 && EnumUtil.getEnumValue(Visibility_$type, d) == 0) {
                    this.r2.cs.e(this.fx);
                }
                break;
            case RadialBase.$$p[3]:
                this.sa = this.sh();
                this.mn(false);
                break;
            case RadialBase.$$p[2]:
                this.r9 = this.sg();
                this.mn(false);
                break;
        }
    };
    RadialBase.prototype.sh = function () {
        return this.sf;
    };
    RadialBase.prototype.sg = function () {
        return this.se;
    };
    RadialBase.prototype.sc = function (a) {
        if (typeCast(NumericRadiusAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    RadialBase.prototype.sb = function (a) {
        if (typeCast(CategoryAngleAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    RadialBase.prototype.hy = function (a) {
        var b = this.hi(a);
        return b >= 0 && this.cl != null && b < this.cl.count ? this.cl.item(b) : null;
    };
    RadialBase.prototype.sk = function (a) {
        var b = this.b1.bv;
        var c = this.b1.bu;
        if (this.rt != null && !b.isEmpty && !c.isEmpty && this.r1 != null) {
            var d = this.r1.c(a);
            if (d < 0) {
                d += Math.PI * 2;
            }
            if (d > Math.PI * 2) {
                d -= Math.PI * 2;
            }
            return d;
        }
        return NaN;
    };
    RadialBase.prototype.f9 = function (a) {
        var b = this.b1.bv;
        var c = this.b1.bu;
        var d = -1;
        if (this.rt != null && !b.isEmpty && !c.isEmpty && this.r1 != null) {
            var e = this.r1.c(a);
            if (e < 0) {
                e += Math.PI * 2;
            }
            if (e > Math.PI * 2) {
                e -= Math.PI * 2;
            }
            var f = this.rt.getUnscaledAngle(e);
            if (this.rt.categoryMode != 0) {
                f -= 0.5;
            }
            d = f;
        }
        return d;
    };
    RadialBase.prototype.hi = function (a) {
        var b = truncate(Math.round(this.f9(a)));
        if (this.cl != null && b == this.cl.count) {
            b = 0;
        }
        return b;
    };
    RadialBase.prototype.eq = function (a) {
        return false;
    };
    RadialBase.prototype.r4 = function () {
        var a = this.r3;
        if (this.cw()) {
            if (this.cf.g) {
                this.cf.t();
            }
            a = this.r6;
        }
        return a;
    };
    RadialBase.prototype.lk = function () {
        _super.prototype.lk.call(this);
        if (this.rt != null) {
            this.rt.h7(false);
        }
        if (this.r0 != null) {
            this.r0.h7(false);
        }
    };
    RadialBase.prototype.fd = function (a, b, c) {
        var d = true;
        var e = c;
        if (!_super.prototype.fd.call(this, a, b, c) || !c.o() || b.isEmpty || a.isEmpty || this.rt == null || this.rt.itemsSource == null || this.r0 == null || this.r1 == null || this.cl == null || this.rt.bh == null || this.r0.bh == null || this.r0.lw == this.r0.lv) {
            e.cs.b = 0;
            d = false;
        }
        return d;
    };
    RadialBase.prototype.mo = function (a) {
        var b;
        var c;
        var d = this.getViewInfo(c, b);
        c = d.p0;
        b = d.p1;
        if (!this.fd(c, b, this.b1)) {
            this.k2(true, this.b1);
            return;
        }
        var e = new SeriesRenderingArguments(this, c, b, a, this.ew);
        var f = this.r8.g(e, this.r5, this.r3, this.r6, this.r2);
        this.r5 = f.p1;
        this.r3 = f.p2;
        this.r6 = f.p3;
    };
    RadialBase.prototype.sn = function () {
        var e_1, _a;
        var a = 0;
        try {
            for (var _b = __values(fromEnum(this.ce.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                if (b == this) {
                    return a;
                }
                var c = typeCast(RadialBase.$, b);
                if (c != null && c.rt == this.rt && c.preferredCategoryMode(c.rt) == 2) {
                    a++;
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        return -1;
    };
    RadialBase.$t = markType(RadialBase, 'RadialBase', MarkerSeries.$, [IHasCategoryModePreference_$type]);
    RadialBase.$$p = markDep(DependencyProperty, PropertyMetadata, RadialBase, 'raisePropertyChanged', ['AngleAxis:rt:sz', [CategoryAngleAxis.$, null], 'ClipSeriesToBounds:sd:s0', [0, false], 'IsCustomRadialMarkerStyleAllowed:se:s1', [0, false], 'IsCustomRadialStyleAllowed:sf:s2', [0, false], 'ValueAxis:r0:s3', [NumericRadiusAxis.$, null]]);
    return RadialBase;
}(MarkerSeries));
export { RadialBase };
/**
 * @hidden
 */
var RadialBaseView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialBaseView, _super);
    function RadialBaseView(a) {
        var _this = _super.call(this, a) || this;
        _this.cs = null;
        _this.cr = null;
        _this.ct = null;
        _this.cu = 0;
        _this.cv = null;
        _this.cr = a;
        _this.ct = new Pool$1(Marker.$);
        _this.ci(_this.ct);
        return _this;
    }
    RadialBaseView.prototype.cf = function (a) {
        this.ct.c(a);
    };
    RadialBaseView.$t = markType(RadialBaseView, 'RadialBaseView', MarkerSeriesView.$);
    return RadialBaseView;
}(MarkerSeriesView));
export { RadialBaseView };
