// igniteui-react-charts/esm5/lib/ScatterPolylineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ShapeSeriesBase } from "./ShapeSeriesBase";
import { StyleSelector } from "./StyleSelector";
import { DependencyProperty } from "igniteui-react-core";
import { Style } from "igniteui-react-core";
import { markType, markDep, typeCast } from "igniteui-react-core";
import { ShapeSeriesViewBase } from "./ShapeSeriesViewBase";
import { PropertyMetadata } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { Path } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
/**
 * @hidden
 */
var ScatterPolylineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolylineSeries, _super);
    function ScatterPolylineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = ScatterPolylineSeries.$;
        return _this;
    }
    ScatterPolylineSeries.prototype.get_eb = function () {
        return true;
    };
    ScatterPolylineSeries.prototype.by = function () {
        return new ScatterPolylineSeriesView(this);
    };
    ScatterPolylineSeries.prototype.get_q6 = function () {
        return false;
    };
    ScatterPolylineSeries.prototype.l4 = function (a) {
        _super.prototype.l4.call(this, a);
        this.sz = a;
    };
    ScatterPolylineSeries.prototype.get_rb = function () {
        return false;
    };
    ScatterPolylineSeries.prototype.l9 = function (a, b, c, d) {
        _super.prototype.l9.call(this, a, b, c, d);
        switch (b) {
            case ScatterPolylineSeries.$$p[0]:
                this.sz.cv(this.s5);
                this.mn(false);
                this.lu();
                break;
            case ScatterPolylineSeries.$$p[1]:
                this.sz.cw(this.s0);
                this.mn(false);
                this.lu();
                break;
        }
    };
    ScatterPolylineSeries.prototype.sj = function (a) {
        var b = a;
        b.cv(this.s5);
        b.cw(this.s0);
        _super.prototype.sj.call(this, a);
    };
    ScatterPolylineSeries.prototype.au = function (a, b) {
        var c = this.sz.c0(a, b, true);
        if (c != null) {
            return c.dataContext;
        }
        return _super.prototype.au.call(this, a, b);
    };
    ScatterPolylineSeries.$t = markType(ScatterPolylineSeries, 'ScatterPolylineSeries', ShapeSeriesBase.$);
    ScatterPolylineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterPolylineSeries, 'raisePropertyChanged', ['ShapeStyle:s5:s3', [Style.$, null], 'ShapeStyleSelector:s0:s4', [StyleSelector.$, null]]);
    return ScatterPolylineSeries;
}(ShapeSeriesBase));
export { ScatterPolylineSeries };
/**
 * @hidden
 */
var ScatterPolylineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolylineSeriesView, _super);
    function ScatterPolylineSeriesView(a) {
        return _super.call(this, a) || this;
    }
    ScatterPolylineSeriesView.prototype.as = function () {
        _super.prototype.as.call(this);
        if (!this.q) {
            this.b3.g4 = Defaults.aa;
        }
    };
    ScatterPolylineSeriesView.prototype.c2 = function (a, b) {
        return this.c3(a, b, false);
    };
    ScatterPolylineSeriesView.prototype.cj = function (a, b, c) {
        _super.prototype.cj.call(this, a, b, c);
        ShapeSeriesViewBase.ci(this, a, b);
    };
    ScatterPolylineSeriesView.prototype.ch = function (a, b) {
        var c = typeCast(Path.$, a);
        if (c == null) {
            return;
        }
        c.an = b;
    };
    ScatterPolylineSeriesView.prototype.ac = function (a) {
        var b = _super.prototype.ac.call(this, a);
        b._fill = null;
        return b;
    };
    ScatterPolylineSeriesView.prototype.ab = function (a) {
        var b = _super.prototype.ab.call(this, a);
        b._fill = null;
        return b;
    };
    ScatterPolylineSeriesView.prototype.b8 = function (a) {
        return a.width >= this.b3.rs || a.height >= this.b3.rs;
    };
    ScatterPolylineSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.a7.a$r(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.b3.title + "</span><br/>";
        a += "</div>";
        return a;
    };
    ScatterPolylineSeriesView.$t = markType(ScatterPolylineSeriesView, 'ScatterPolylineSeriesView', ShapeSeriesViewBase.$);
    return ScatterPolylineSeriesView;
}(ShapeSeriesViewBase));
export { ScatterPolylineSeriesView };
