// igniteui-react-charts/esm5/lib/ScatterSplineSeriesView.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ScatterBaseView } from "./ScatterBaseView";
import { Path } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
var ScatterSplineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterSplineSeriesView, _super);
    function ScatterSplineSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.dg = new Path();
        _this.df = new Path();
        return _this;
    }
    ScatterSplineSeriesView.prototype.as = function () {
        _super.prototype.as.call(this);
        if (!this.q) {
            this.b1.ql = Defaults.e;
        }
    };
    ScatterSplineSeriesView.prototype.bb = function () {
        _super.prototype.bb.call(this);
    };
    ScatterSplineSeriesView.prototype.bc = function () {
        _super.prototype.bc.call(this);
        this.df.an = this.dg.an;
        var a = this.bo();
        this.df._stroke = a;
        this.df.ad = this.e.fy + 3;
    };
    ScatterSplineSeriesView.prototype.a6 = function (a, b) {
        _super.prototype.a6.call(this, a, b);
        if (a.d) {
            if (b) {
                a.t(this.df);
            }
            else {
                a.t(this.dg);
            }
        }
    };
    ScatterSplineSeriesView.prototype.c4 = function (a) {
        _super.prototype.c4.call(this, a);
        this.dg.an = null;
        this.df.an = null;
    };
    ScatterSplineSeriesView.prototype.al = function (a) {
        _super.prototype.al.call(this, a);
        this.e.a9.exportPathData(a, this.dg, "shape", ["Main"]);
    };
    ScatterSplineSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.ph = a;
        this.e.g1 = 3;
        this.e.g2 = 1;
        this.e.g3 = 4;
        this.e.fb = false;
    };
    ScatterSplineSeriesView.$t = markType(ScatterSplineSeriesView, 'ScatterSplineSeriesView', ScatterBaseView.$);
    return ScatterSplineSeriesView;
}(ScatterBaseView));
export { ScatterSplineSeriesView };
