// igniteui-react-charts/esm5/lib/ScatterAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ScatterTriangulationSeries } from "./ScatterTriangulationSeries";
import { ColorScale } from "./ColorScale";
import { runOn, delegateRemove, delegateCombine, typeCast, Point_$type, markType, markDep } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { FastItemIntColumn } from "igniteui-react-core";
import { IFastItemsSource_$type } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { SeriesView } from "./SeriesView";
import { TriangleRasterizer } from "./TriangleRasterizer";
import { ScalerParams } from "./ScalerParams";
import { ObservableColorCollection } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
import { ColorUtil } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { CustomPaletteColorScale } from "./CustomPaletteColorScale";
import { DataLegendSeriesValueInfo } from "igniteui-react-core";
import { DataLegendSeriesContext } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { truncate, intDivide } from "igniteui-react-core";
import { boxArray$1 } from "igniteui-react-core";
import { RenderingContext } from "igniteui-react-core";
import { CanvasViewRenderer } from "igniteui-react-core";
import { Rectangle } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
/**
 * @hidden
 */
var ScatterAreaSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterAreaSeries, _super);
    function ScatterAreaSeries() {
        var _this = _super.call(this) || this;
        _this.r7 = null;
        _this.sc = null;
        _this.r5 = null;
        _this.r6 = null;
        FastItemIntColumn.m();
        _this.ab = ScatterAreaSeries.$;
        return _this;
    }
    ScatterAreaSeries.prototype.get_df = function () {
        return true;
    };
    Object.defineProperty(ScatterAreaSeries.prototype, "r9", {
        get: function () {
            return this.r7;
        },
        set: function (a) {
            var b = this.r7 != a;
            if (b) {
                var c = this.r7;
                if (this.r7 != null) {
                    var d = this.r7;
                    d.propertyChanged = delegateRemove(d.propertyChanged, runOn(this, this.sq));
                }
                this.r7 = a;
                if (this.r7 != null) {
                    var e = this.r7;
                    e.propertyChanged = delegateCombine(e.propertyChanged, runOn(this, this.sq));
                }
                this.raisePropertyChanged("ColorScale", c, this.r7);
            }
        },
        enumerable: false,
        configurable: true
    });
    ScatterAreaSeries.prototype.sq = function (a, b) {
        this.ss();
        this.mn(false);
    };
    ScatterAreaSeries.prototype.h1 = function (a, b) {
        var c = this.jv(b);
        if (c == ScatterAreaSeries.$$p[3]) {
            return this.h2(a, b, this.sl);
        }
        return _super.prototype.h1.call(this, a, b);
    };
    Object.defineProperty(ScatterAreaSeries.prototype, "sd", {
        get: function () {
            return this.sc;
        },
        set: function (a) {
            var b = this.sd != a;
            if (b) {
                var c = this.sd;
                this.sc = a;
                this.raisePropertyChanged("ColorColumn", c, this.sd);
            }
        },
        enumerable: false,
        configurable: true
    });
    ScatterAreaSeries.prototype.get_e7 = function () {
        return true;
    };
    ScatterAreaSeries.prototype.set_e7 = function (a) {
        _super.prototype.set_e7.call(this, a);
    };
    ScatterAreaSeries.prototype.l9 = function (a, b, c, d) {
        _super.prototype.l9.call(this, a, b, c, d);
        switch (b) {
            case "FastItemsSource":
                var e = typeCast(IFastItemsSource_$type, c);
                if (e != null) {
                    e.deregisterColumn(this.sd);
                    this.sd = null;
                }
                var f = typeCast(IFastItemsSource_$type, d);
                if (f != null) {
                    this.sd = this.ci(this.sl);
                }
                this.mn(false);
                break;
            case ScatterAreaSeries.$$p[3]:
                if (this.cl != null) {
                    this.cl.deregisterColumn(this.sd);
                    this.sd = this.ci(this.sl);
                }
                this.mn(false);
                this.lu();
                break;
            case "ColorScale":
                this.ss();
                this.mn(false);
                this.lu();
                break;
        }
    };
    ScatterAreaSeries.prototype.mo = function (a) {
        _super.prototype.mo.call(this, a);
        this.sr(this.b1);
    };
    ScatterAreaSeries.prototype.mm = function (a, b, c, d, e) {
        _super.prototype.mm.call(this, a, b, c, d, e);
        var f = this.a7.a$h.item(d);
        f.a1(c);
        if (this.cy(f)) {
            return;
        }
        this.sr(f);
    };
    ScatterAreaSeries.prototype.sr = function (a) {
        if (this.cy(a)) {
            return;
        }
        var b;
        var c;
        var d = a.getViewInfo(c, b);
        c = d.p0;
        b = d.p1;
        var e = a;
        var f = null;
        var g = null;
        var h = this.qc;
        var i = this.qd;
        var j = this.getEffectiveViewport1(a);
        var k = new ScalerParams(1, b, c, h.cg);
        k.c = j;
        var l = new ScalerParams(1, b, c, i.cg);
        l.c = j;
        g = this.sb(a, k, l);
        if (g != null) {
            e.b8(g);
        }
        else {
            e.cb(f);
        }
    };
    ScatterAreaSeries.prototype.so = function (a) {
        var b;
        var c;
        var d = a.getViewInfo(b, c);
        b = d.p0;
        c = d.p1;
        var e = truncate(Math.round(b.width)) * truncate(Math.round(b.height));
        var f = a.b0;
        if (f == null || f.length != e) {
            f = new Array(e);
            for (var g = 0; g < e; g++) {
                f[g] = 0;
            }
        }
        else {
            for (var h = 0; h < e; h++) {
                f[h] = 0;
            }
        }
        a.b0 = f;
    };
    ScatterAreaSeries.prototype.hy = function (a) {
        var b = truncate(Math.round(this.p7.width));
        var c = truncate(Math.round(this.p7.height));
        var d = this.b1.b0;
        if (d == null || d.length != (b * c) || this.qm == null || this.qm.count < 1 || this.qn == null || this.qn.count < 1 || this.qo == null || this.qo.count < 1) {
            return null;
        }
        var e = this.ce.tq;
        var f = (a.x - e.left) / e.width;
        var g = (a.y - e.top) / e.height;
        var h = truncate(Math.round(b * f));
        var i = truncate(Math.round(c * g));
        var j = (b * i) + h;
        if (j < 0 || j > d.length - 1) {
            return null;
        }
        var k = d[j] - 1;
        if (k < 0 || k > this.qm.count) {
            return null;
        }
        var l = this.r5.length;
        var m = this.qm.item(k);
        var n = this.qn.item(k);
        var o = this.qo.item(k);
        if ((m >= l) || (n >= l) || (o >= l)) {
            return null;
        }
        var p = (this.r5[m] - this.p7.left) / this.p7.width;
        var q = (this.r6[m] - this.p7.top) / this.p7.height;
        var r = (this.r5[n] - this.p7.left) / this.p7.width;
        var s = (this.r6[n] - this.p7.top) / this.p7.height;
        var t = (this.r5[o] - this.p7.left) / this.p7.width;
        var u = (this.r6[o] - this.p7.top) / this.p7.height;
        var v = Math.pow(p - f, 2) + Math.pow(q - g, 2);
        var w = Math.pow(r - f, 2) + Math.pow(s - g, 2);
        var x = Math.pow(t - f, 2) + Math.pow(u - g, 2);
        if (v < w && v < x) {
            return this.cl.item(m);
        }
        if (w < v && w < x) {
            return this.cl.item(n);
        }
        if (x < v && x < w) {
            return this.cl.item(o);
        }
        return this.cl.item(m);
    };
    ScatterAreaSeries.prototype.sb = function (a, b, c) {
        var d;
        var e;
        var f = a.getViewInfo(e, d);
        e = f.p0;
        d = f.p1;
        e = new Rect(0, 0, 0, e.width, e.height);
        b = new ScalerParams(0, b.e, e, b.a, b.c);
        c = new ScalerParams(0, c.e, e, c.a, c.c);
        c.b = this.gg();
        var g = a;
        var h = this.qc;
        var i = this.qd;
        var j = this.qm.count;
        var k;
        var xArr_ = this.qp.asArray();
        k = (xArr_.slice(0));
        var l;
        var yArr_ = this.qq.asArray();
        l = (yArr_.slice(0));
        var m = this.sd.asArray();
        var n = this.qm.asArray();
        var o = this.qn.asArray();
        var p = this.qo.asArray();
        this.r5 = k;
        this.r6 = l;
        h.hm(boxArray$1(k), 0, k.length, b);
        i.hm(boxArray$1(l), 0, l.length, c);
        var q = this.r8;
        var r = g.b2(b.d, q, this.sd);
        var s = e.top;
        var t = e.left;
        var u = e.right;
        var v = e.bottom;
        var w;
        var x;
        var y;
        var z;
        this.so(g);
        var aa = g.b0;
        var ab = k.length;
        for (var ac = 0; ac < j; ac++) {
            var ad = n[ac];
            var ae = o[ac];
            var af = p[ac];
            if ((ad >= ab) || (ae >= ab) || (af >= ab) || (ad < 0) || (ae < 0) || (af < 0)) {
                continue;
            }
            var ag = k[ad];
            var ah = l[ad];
            var ai = { $type: Point_$type, x: ag, y: ah };
            var aj = k[ae];
            var ak = l[ae];
            var al = { $type: Point_$type, x: aj, y: ak };
            var am = k[af];
            var an = l[af];
            var ao = { $type: Point_$type, x: am, y: an };
            w = al.y < ao.y ? al.y : ao.y;
            w = ai.y < w ? ai.y : w;
            x = al.y > ao.y ? al.y : ao.y;
            x = ai.y > x ? ai.y : x;
            y = al.x < ao.x ? al.x : ao.x;
            y = ai.x < y ? ai.x : y;
            z = al.x > ao.x ? al.x : ao.x;
            z = ai.x > z ? ai.x : z;
            if (w < v && x > s && y < u && z > t) {
                var ap = m[ad];
                var aq = m[ae];
                var ar = m[af];
                r.p(ac, aa, ai, al, ao, ap, aq, ar);
            }
        }
        return r;
    };
    ScatterAreaSeries.prototype.se = function () {
        var a = new Color();
        var b = this.o5;
        if (b == null) {
            a = ColorUtil.r("#000000");
        }
        else {
            a = ColorUtil.q(b);
        }
        var c = new ObservableColorCollection();
        c.add(ColorUtil.v(a, -0.5));
        c.add(a);
        return c;
    };
    ScatterAreaSeries.prototype.ss = function () {
        var _this = this;
        var a = this.r9;
        if (a == null) {
            a = ((function () {
                var $ret = new CustomPaletteColorScale();
                $ret.palette = _this.se();
                $ret.interpolationMode = 2;
                return $ret;
            })());
        }
        else if (typeCast(CustomPaletteColorScale.$, a) !== null) {
            var b = a;
            if (b.palette == null || b.palette.count == 0) {
                b.palette = this.se();
            }
            a = b;
        }
        var c = new ObservableColorCollection();
        if (typeCast(CustomPaletteColorScale.$, a) !== null) {
            c = a.palette;
        }
        else {
            c = this.se();
        }
        this.r8 = a;
    };
    ScatterAreaSeries.prototype.m1 = function () {
        _super.prototype.m1.call(this);
        this.ss();
    };
    ScatterAreaSeries.prototype.k2 = function (a, b) {
        _super.prototype.k2.call(this, a, b);
        var c = b;
        c.b7();
    };
    ScatterAreaSeries.prototype.by = function () {
        return new ScatterAreaSeriesView(this);
    };
    ScatterAreaSeries.prototype.l4 = function (a) {
        _super.prototype.l4.call(this, a);
        this.sa = typeCast(ScatterAreaSeriesView.$, a);
    };
    ScatterAreaSeries.prototype.sp = function (a) {
        if (this.sa != null) {
            this.sa.b6(a);
        }
    };
    ScatterAreaSeries.prototype.getDataLegendSeriesContextAt = function (a, b) {
        var c = { $type: Point_$type, x: a, y: b };
        var d = this.hy(c);
        var e = d == null ? NaN : this.gf(d, "XMemberPath");
        var f = d == null ? NaN : this.gf(d, "YMemberPath");
        var g = d == null ? NaN : this.gf(d, "ColorMemberPath");
        var h = new DataLegendSeriesValueInfo();
        h.l = e;
        h.a = 13;
        h.p = "X";
        h.q = this.re != null ? this.re : "X";
        h.r = this.ra != null ? this.ra : "X:";
        h.s = this.rc != null ? this.rc : "";
        var i = new DataLegendSeriesValueInfo();
        i.l = f;
        i.a = 14;
        i.p = "Y";
        i.q = this.rm != null ? this.rm : "Y";
        i.r = this.ri != null ? this.ri : "Y:";
        i.s = this.rk != null ? this.rk : "";
        var j = new DataLegendSeriesValueInfo();
        j.l = g;
        j.a = 0;
        j.p = "Value";
        j.q = this.sl != null ? this.sl : "Value";
        j.r = this.sh != null ? this.sh : "V:";
        j.s = this.sj != null ? this.sj : "";
        var k = new DataLegendSeriesContext();
        k.f = ((function () {
            var $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
            $ret.add(h);
            $ret.add(i);
            $ret.add(j);
            return $ret;
        })());
        k.b = this.aw();
        return k;
    };
    ScatterAreaSeries.$t = markType(ScatterAreaSeries, 'ScatterAreaSeries', ScatterTriangulationSeries.$);
    ScatterAreaSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterAreaSeries, 'raisePropertyChanged', ['ActualColorScale:r8:st', [ColorScale.$, null], 'ColorMemberAsLegendLabel:sh:su', [2, null], 'ColorMemberAsLegendUnit:sj:sv', [2, null], 'ColorMemberPath:sl:sw', [2, null]]);
    return ScatterAreaSeries;
}(ScatterTriangulationSeries));
export { ScatterAreaSeries };
/**
 * @hidden
 */
var ScatterAreaSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterAreaSeriesView, _super);
    function ScatterAreaSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.b1 = null;
        _this.bx = null;
        _this.b5 = null;
        _this.b4 = null;
        _this.b3 = null;
        _this.b1 = a;
        return _this;
    }
    Object.defineProperty(ScatterAreaSeriesView.prototype, "b0", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
        },
        enumerable: false,
        configurable: true
    });
    ScatterAreaSeriesView.prototype.ca = function (a, b) {
    };
    ScatterAreaSeriesView.prototype.b6 = function (a) {
    };
    ScatterAreaSeriesView.prototype.by = function () {
        return this.bz(true);
    };
    ScatterAreaSeriesView.prototype.bz = function (a) {
        this.b9();
        var b = truncate(Math.round(this.bu.left));
        var c = truncate(Math.round(this.bu.left));
        var d = truncate(Math.round(this.bu.width));
        var e = truncate(Math.round(this.bu.height));
        var f = d.toString();
        var g = e.toString();
        if (this.b4.getAttribute("width") != f) {
            this.b4.setAttribute("width", d.toString());
        }
        if (this.b4.getAttribute("height") != g) {
            this.b4.setAttribute("height", e.toString());
        }
        var h = this.b3.h();
        if (a) {
            h.clearRect(0, 0, d, e);
        }
        this.b5 = h.getImageData(0, 0, d, e);
        var imageDataArray_ = this.b5.data;
        return imageDataArray_;
    };
    ScatterAreaSeriesView.prototype.b9 = function () {
        if (this.b4 == null) {
            this.b4 = this.e.he.createElement("canvas");
            var a = this.e.he.get2DCanvasContext(this.b4);
            this.b3 = new RenderingContext(new CanvasViewRenderer(), a);
        }
    };
    ScatterAreaSeriesView.prototype.b2 = function (a, b, c) {
        var d = truncate(Math.round(this.bu.left));
        var e = truncate(Math.round(this.bu.left));
        var f = truncate(Math.round(this.bu.width));
        var g = truncate(Math.round(this.bu.height));
        return new TriangleRasterizer(this.by(), b, c.minimum, c.maximum, c, f, g);
    };
    ScatterAreaSeriesView.prototype.b8 = function (a) {
        this.cb(null);
    };
    ScatterAreaSeriesView.prototype.b7 = function () {
        if (this.b3 != null) {
            this.b3.l(this.bu.left, this.bu.top, this.bu.width, this.bu.height);
        }
        this.aq();
    };
    ScatterAreaSeriesView.prototype.a6 = function (a, b) {
        _super.prototype.a6.call(this, a, b);
        if (this.b4 == null) {
            return;
        }
        if (b) {
            var c = new Rectangle();
            c.n = this.bu.left;
            c.o = this.bu.top;
            c.width = this.bu.width;
            c.height = this.bu.height;
            c._fill = this.bo();
            a.w(c);
        }
        else {
            var d = truncate(Math.round(this.bu.left));
            var e = truncate(Math.round(this.bu.top));
            var f = truncate(Math.round(this.bu.width));
            var g = truncate(Math.round(this.bu.height));
            a.n(this.b4.getNativeElement(), 1, d, e, f, g);
        }
    };
    ScatterAreaSeriesView.prototype.cb = function (a) {
        var b = truncate(Math.round(this.bu.left));
        var c = truncate(Math.round(this.bu.left));
        var d = truncate(Math.round(this.bu.width));
        var e = truncate(Math.round(this.bu.height));
        this.b9();
        this.b3.h().putImageData(this.b5, 0, 0);
        this.aq();
    };
    ScatterAreaSeriesView.prototype.al = function (a) {
        _super.prototype.al.call(this, a);
        var b = truncate(Math.round(this.bu.width));
        var c = truncate(Math.round(this.bu.height));
        var d = this.bz(false);
        if (d == null || d.length == 0) {
            this.e.a9.exportSeriesPixelsData(a, null, b, c);
            return;
        }
        var e = new Array(intDivide(d.length, 4));
        var f;
        for (var g = 0; g < intDivide(d.length, 4); g++) {
            f = g * 4;
            e[g] = d[f + 3] << 24 | d[f] << 16 | d[f + 1] << 8 | d[f + 2];
        }
        this.e.a9.exportSeriesPixelsData(a, e, b, c);
    };
    ScatterAreaSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.a7.a$r(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.b1.title + "</span><br/><span>" + "(${item." + this.b1.re + "}, ${item." + this.b1.rm + "})</span>";
        if (!stringIsNullOrEmpty(this.b1.sl)) {
            a += "<span>, Value: ${item." + this.b1.sl + "}</span>";
        }
        a += "</div>";
        return a;
    };
    ScatterAreaSeriesView.$t = markType(ScatterAreaSeriesView, 'ScatterAreaSeriesView', SeriesView.$);
    return ScatterAreaSeriesView;
}(SeriesView));
export { ScatterAreaSeriesView };
