// igniteui-react-charts/esm5/lib/Stacked100BarSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedBarSeries } from "./StackedBarSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { isNaN_ } from "igniteui-react-core";
import { StackedBarSeriesView } from "./StackedBarSeriesView";
import { Stacked100BarBucketCalculator } from "./Stacked100BarBucketCalculator";
/**
 * @hidden
 */
var Stacked100BarSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100BarSeries, _super);
    function Stacked100BarSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100BarSeries.$;
        return _this;
    }
    Stacked100BarSeries.prototype.get_tx = function () {
        return true;
    };
    Stacked100BarSeries.prototype.gz = function () {
        if (!isNaN_(this.gy)) {
            return _super.prototype.gz.call(this);
        }
        return Defaults.h;
    };
    Stacked100BarSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.mn(false);
    };
    Stacked100BarSeries.prototype.by = function () {
        return new Stacked100BarSeriesView(this);
    };
    Stacked100BarSeries.prototype.l4 = function (a) {
        _super.prototype.l4.call(this, a);
        this.u1 = a;
    };
    Stacked100BarSeries.prototype.tp = function () {
        return this.u1;
    };
    Stacked100BarSeries.prototype.ud = function () {
        _super.prototype.ud.call(this);
        Stacked100DataPreparer.a(this);
    };
    Object.defineProperty(Stacked100BarSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.ss;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100BarSeries.$t = markType(Stacked100BarSeries, 'Stacked100BarSeries', StackedBarSeries.$, [IStacked100Series_$type]);
    return Stacked100BarSeries;
}(StackedBarSeries));
export { Stacked100BarSeries };
/**
 * @hidden
 */
var Stacked100BarSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100BarSeriesView, _super);
    function Stacked100BarSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c6 = null;
        _this.c6 = a;
        return _this;
    }
    Stacked100BarSeriesView.prototype.cs = function () {
        return new Stacked100BarBucketCalculator(this);
    };
    Stacked100BarSeriesView.prototype.as = function () {
        _super.prototype.as.call(this);
        if (!this.q) {
            this.c0.ql = Defaults.c;
            this.c0.g4 = Defaults.i;
        }
    };
    Stacked100BarSeriesView.$t = markType(Stacked100BarSeriesView, 'Stacked100BarSeriesView', StackedBarSeriesView.$);
    return Stacked100BarSeriesView;
}(StackedBarSeriesView));
export { Stacked100BarSeriesView };
