// igniteui-react-charts/esm5/lib/StochRSIIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StochRSIIndicatorStrategy } from "./StochRSIIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var StochRSIIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StochRSIIndicator, _super);
    function StochRSIIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StochRSIIndicator.prototype.get_up = function () {
        return new StochRSIIndicatorStrategy();
    };
    Object.defineProperty(StochRSIIndicator.prototype, "up", {
        get: function () {
            return this.get_up();
        },
        enumerable: false,
        configurable: true
    });
    StochRSIIndicator.prototype.get_u2 = function () {
        return StochRSIIndicator.$;
    };
    Object.defineProperty(StochRSIIndicator.prototype, "u2", {
        get: function () {
            return this.get_u2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StochRSIIndicator.prototype, "va", {
        get: function () {
            return typeGetValue(this.c(StochRSIIndicator.vb));
        },
        set: function (a) {
            this.h(StochRSIIndicator.vb, a);
        },
        enumerable: false,
        configurable: true
    });
    StochRSIIndicator.prototype.uv = function () {
        return this.va;
    };
    StochRSIIndicator.prototype.u4 = function (a) {
        this.va = a;
    };
    StochRSIIndicator.$t = markType(StochRSIIndicator, 'StochRSIIndicator', StrategyBasedIndicator.$);
    StochRSIIndicator.vb = StrategyBasedIndicator.u7(14, StochRSIIndicator.$);
    return StochRSIIndicator;
}(StrategyBasedIndicator));
export { StochRSIIndicator };
