// igniteui-react-charts/esm5/lib/WilliamsPercentRIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { WilliamsPercentRIndicatorStrategy } from "./WilliamsPercentRIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var WilliamsPercentRIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WilliamsPercentRIndicator, _super);
    function WilliamsPercentRIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    WilliamsPercentRIndicator.prototype.get_up = function () {
        return new WilliamsPercentRIndicatorStrategy();
    };
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "up", {
        get: function () {
            return this.get_up();
        },
        enumerable: false,
        configurable: true
    });
    WilliamsPercentRIndicator.prototype.get_u2 = function () {
        return WilliamsPercentRIndicator.$;
    };
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "u2", {
        get: function () {
            return this.get_u2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "va", {
        get: function () {
            return typeGetValue(this.c(WilliamsPercentRIndicator.vb));
        },
        set: function (a) {
            this.h(WilliamsPercentRIndicator.vb, a);
        },
        enumerable: false,
        configurable: true
    });
    WilliamsPercentRIndicator.prototype.uv = function () {
        return this.va;
    };
    WilliamsPercentRIndicator.prototype.u4 = function (a) {
        this.va = a;
    };
    WilliamsPercentRIndicator.$t = markType(WilliamsPercentRIndicator, 'WilliamsPercentRIndicator', StrategyBasedIndicator.$);
    WilliamsPercentRIndicator.vb = StrategyBasedIndicator.u7(14, WilliamsPercentRIndicator.$);
    return WilliamsPercentRIndicator;
}(StrategyBasedIndicator));
export { WilliamsPercentRIndicator };
