// igniteui-react-charts/esm5/lib/XYChart.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { DomainChart } from "./DomainChart";
import { fromEnum, enumGetBox, markType } from "igniteui-react-core";
import { HorizontalAlignment_$type } from "igniteui-react-core";
import { VerticalAlignment_$type } from "igniteui-react-core";
import { Visibility_$type } from "igniteui-react-core";
import { YAxisLabelLocation_$type } from "./YAxisLabelLocation";
import { XAxisLabelLocation_$type } from "./XAxisLabelLocation";
import { AxisLabelSettings } from "./AxisLabelSettings";
import { AxisDefaults } from "./AxisDefaults";
import { TitleSettings } from "./TitleSettings";
import { ScalerParams } from "./ScalerParams";
import { DeviceUtils } from "igniteui-react-core";
/**
 * @hidden
 */
var XYChart = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(XYChart, _super);
    function XYChart() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.v5 = null;
        _this.v6 = null;
        _this.wm = 0;
        _this.wo = 5;
        _this.wn = 0;
        _this.wl = 5;
        _this.w3 = 5;
        _this.w5 = 0;
        _this.w4 = 5;
        _this.w2 = 0;
        _this.abw = null;
        _this.ab3 = null;
        _this.abu = null;
        _this.abv = null;
        _this.wx = 0;
        _this.xe = 0;
        _this.ww = NaN;
        _this.xd = NaN;
        _this.wz = NaN;
        _this.xg = NaN;
        _this.wy = NaN;
        _this.xf = NaN;
        _this.wv = NaN;
        _this.xc = NaN;
        _this.ab2 = null;
        _this.ab9 = null;
        _this.yo = null;
        _this.ys = null;
        _this.yq = null;
        _this.yu = null;
        _this.yj = null;
        _this.yk = null;
        _this.abx = null;
        _this.ab4 = null;
        _this.wp = DeviceUtils.g(1);
        _this.w6 = DeviceUtils.g(1);
        _this.wq = DeviceUtils.g(1);
        _this.w7 = DeviceUtils.g(1);
        _this.abz = null;
        _this.ab6 = null;
        _this.ab0 = null;
        _this.ab7 = null;
        _this.wr = DeviceUtils.g(1);
        _this.w8 = DeviceUtils.g(1);
        _this.ws = DeviceUtils.g(8);
        _this.w9 = 0;
        _this.ab1 = null;
        _this.ab8 = null;
        _this.wt = DeviceUtils.g(1);
        _this.xa = DeviceUtils.g(1);
        _this.yp = null;
        _this.yt = null;
        _this.aby = null;
        _this.ab5 = null;
        _this.wk = 0;
        _this.w1 = 0;
        _this.wj = NaN;
        _this.w0 = NaN;
        _this.wu = 0;
        _this.xb = -90;
        _this.wf = false;
        _this.wg = false;
        _this.abp = 1;
        _this.acs = 1;
        _this.abo = 1;
        _this.abq = 2;
        _this.acq = 0;
        _this.acr = 1;
        _this.acw = 0;
        _this.acx = 0;
        _this.wd = 4;
        _this.wb = 4;
        _this.yn = null;
        _this.vz = null;
        _this.yr = null;
        _this.v0 = null;
        return _this;
    }
    XYChart.prototype.provideContainer = function (a) {
        if (a != null) {
            var b = this.v3();
            var c = this.v4();
            if (b != null) {
                b.provideRenderer(a);
            }
            if (c != null) {
                c.provideRenderer(a);
            }
        }
        _super.prototype.provideContainer.call(this, a);
    };
    XYChart.prototype.r9 = function () {
        _super.prototype.r9.call(this);
        var a = this.v3();
        var b = this.v4();
        if (a != null && a.ay != null) {
            this.acc = a.ay.textColor;
        }
        if (b != null && b.ay != null) {
            this.acj = b.ay.textColor;
        }
        this.ach = AxisDefaults.j;
        this.aco = AxisDefaults.j;
        this.acd = AxisDefaults.h;
        this.ack = AxisDefaults.h;
        this.acg = AxisDefaults.g;
        this.acn = AxisDefaults.g;
    };
    XYChart.prototype.sy = function (a, b, c) {
        var _this = this;
        _super.prototype.sy.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelLeftMargin":
                this.abg(function (d) { return d.leftMargin = _this.xo; });
                break;
            case "XAxisLabelTopMargin":
                this.abg(function (d) { return d.topMargin = _this.xq; });
                break;
            case "XAxisLabelRightMargin":
                this.abg(function (d) { return d.rightMargin = _this.xp; });
                break;
            case "XAxisLabelBottomMargin":
                this.abg(function (d) { return d.bottomMargin = _this.xn; });
                break;
            case "YAxisLabelLeftMargin":
                this.abl(function (d) { return d.leftMargin = _this.x5; });
                break;
            case "YAxisLabelTopMargin":
                this.abl(function (d) { return d.topMargin = _this.x7; });
                break;
            case "YAxisLabelRightMargin":
                this.abl(function (d) { return d.rightMargin = _this.x6; });
                break;
            case "YAxisLabelBottomMargin":
                this.abl(function (d) { return d.bottomMargin = _this.x4; });
                break;
            case "ActualItemsSource":
            case "ShouldDisplayMockData":
            case "ShouldUseSkeletonStyleForMockData":
                this.aca = this.u6(this.acc);
                this.acb = this.u6(this.acj);
                break;
            case "XAxisLabelTextColor":
                this.aca = this.u6(this.acc);
                break;
            case "YAxisLabelTextColor":
                this.acb = this.u6(this.acj);
                break;
            case "ActualXAxisLabelTextColor":
                this.abg(function (d) { return d.textColor = _this.aca; });
                break;
            case "ActualYAxisLabelTextColor":
                this.abl(function (d) { return d.textColor = _this.acb; });
                break;
            case "XAxisTitleMargin":
                this.abi(function (d) { return d.margin = _this.xz; });
                break;
            case "YAxisTitleMargin":
                this.abn(function (d) { return d.margin = _this.yg; });
                break;
            case "XAxisTitleLeftMargin":
                this.abi(function (d) { return d.leftMargin = _this.xy; });
                break;
            case "YAxisTitleLeftMargin":
                this.abn(function (d) { return d.leftMargin = _this.yf; });
                break;
            case "XAxisTitleTopMargin":
                this.abi(function (d) { return d.topMargin = _this.x1; });
                break;
            case "YAxisTitleTopMargin":
                this.abn(function (d) { return d.topMargin = _this.yi; });
                break;
            case "XAxisTitleRightMargin":
                this.abi(function (d) { return d.rightMargin = _this.x0; });
                break;
            case "YAxisTitleRightMargin":
                this.abn(function (d) { return d.rightMargin = _this.yh; });
                break;
            case "XAxisTitleBottomMargin":
                this.abi(function (d) { return d.bottomMargin = _this.xx; });
                break;
            case "YAxisTitleBottomMargin":
                this.abn(function (d) { return d.bottomMargin = _this.ye; });
                break;
            case "XAxisTitleTextColor":
                this.abi(function (d) { return d.textColor = _this.aci; });
                break;
            case "YAxisTitleTextColor":
                this.abn(function (d) { return d.textColor = _this.acp; });
                break;
            case "XAxisFormatLabel":
                this.abe(function (d) { return d.at = _this.v7; });
                break;
            case "YAxisFormatLabel":
                this.abj(function (d) { return d.at = _this.v8; });
                break;
        }
    };
    Object.defineProperty(XYChart.prototype, "v7", {
        get: function () {
            return this.v5;
        },
        set: function (a) {
            var b = this.v7;
            if (a != b) {
                this.v5 = a;
                this.sv("XAxisFormatLabel", b, this.v7);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "v8", {
        get: function () {
            return this.v6;
        },
        set: function (a) {
            var b = this.v8;
            if (a != b) {
                this.v6 = a;
                this.sv("YAxisFormatLabel", b, this.v8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xo", {
        get: function () {
            return this.wm;
        },
        set: function (a) {
            var b = this.xo;
            if (a != b) {
                this.wm = a;
                this.sv("XAxisLabelLeftMargin", b, this.xo);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xq", {
        get: function () {
            return this.wo;
        },
        set: function (a) {
            var b = this.xq;
            if (a != b) {
                this.wo = a;
                this.sv("XAxisLabelTopMargin", b, this.xq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xp", {
        get: function () {
            return this.wn;
        },
        set: function (a) {
            var b = this.xp;
            if (a != b) {
                this.wn = a;
                this.sv("XAxisLabelRightMargin", b, this.xp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xn", {
        get: function () {
            return this.wl;
        },
        set: function (a) {
            var b = this.xn;
            if (a != b) {
                this.wl = a;
                this.sv("XAxisLabelBottomMargin", b, this.xn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x5", {
        get: function () {
            return this.w3;
        },
        set: function (a) {
            var b = this.x5;
            if (a != b) {
                this.w3 = a;
                this.sv("YAxisLabelLeftMargin", b, this.x5);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x7", {
        get: function () {
            return this.w5;
        },
        set: function (a) {
            var b = this.x7;
            if (a != b) {
                this.w5 = a;
                this.sv("YAxisLabelTopMargin", b, this.x7);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x6", {
        get: function () {
            return this.w4;
        },
        set: function (a) {
            var b = this.x6;
            if (a != b) {
                this.w4 = a;
                this.sv("YAxisLabelRightMargin", b, this.x6);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x4", {
        get: function () {
            return this.w2;
        },
        set: function (a) {
            var b = this.x4;
            if (a != b) {
                this.w2 = a;
                this.sv("YAxisLabelBottomMargin", b, this.x4);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acc", {
        get: function () {
            return this.abw;
        },
        set: function (a) {
            var b = this.acc;
            if (a != b) {
                this.abw = a;
                this.sv("XAxisLabelTextColor", b, this.acc);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acj", {
        get: function () {
            return this.ab3;
        },
        set: function (a) {
            var b = this.acj;
            if (a != b) {
                this.ab3 = a;
                this.sv("YAxisLabelTextColor", b, this.acj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aca", {
        get: function () {
            return this.abu;
        },
        set: function (a) {
            var b = this.abu;
            if (a != b) {
                this.abu = a;
                this.sv("ActualXAxisLabelTextColor", b, this.aca);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acb", {
        get: function () {
            return this.abv;
        },
        set: function (a) {
            var b = this.abv;
            if (a != b) {
                this.abv = a;
                this.sv("ActualYAxisLabelTextColor", b, this.acb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xz", {
        get: function () {
            return this.wx;
        },
        set: function (a) {
            var b = this.xz;
            if (a != b) {
                this.wx = a;
                this.sv("XAxisTitleMargin", b, this.xz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "yg", {
        get: function () {
            return this.xe;
        },
        set: function (a) {
            var b = this.yg;
            if (a != b) {
                this.xe = a;
                this.sv("YAxisTitleMargin", b, this.yg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xy", {
        get: function () {
            return this.ww;
        },
        set: function (a) {
            var b = this.xy;
            if (a != b) {
                this.ww = a;
                this.sv("XAxisTitleLeftMargin", b, this.xy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "yf", {
        get: function () {
            return this.xd;
        },
        set: function (a) {
            var b = this.yf;
            if (a != b) {
                this.xd = a;
                this.sv("YAxisTitleLeftMargin", b, this.yf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x1", {
        get: function () {
            return this.wz;
        },
        set: function (a) {
            var b = this.x1;
            if (a != b) {
                this.wz = a;
                this.sv("XAxisTitleTopMargin", b, this.x1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "yi", {
        get: function () {
            return this.xg;
        },
        set: function (a) {
            var b = this.yi;
            if (a != b) {
                this.xg = a;
                this.sv("YAxisTitleTopMargin", b, this.yi);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x0", {
        get: function () {
            return this.wy;
        },
        set: function (a) {
            var b = this.x0;
            if (a != b) {
                this.wy = a;
                this.sv("XAxisTitleRightMargin", b, this.x0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "yh", {
        get: function () {
            return this.xf;
        },
        set: function (a) {
            var b = this.yh;
            if (a != b) {
                this.xf = a;
                this.sv("YAxisTitleRightMargin", b, this.yh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xx", {
        get: function () {
            return this.wv;
        },
        set: function (a) {
            var b = this.xx;
            if (a != b) {
                this.wv = a;
                this.sv("XAxisTitleBottomMargin", b, this.xx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ye", {
        get: function () {
            return this.xc;
        },
        set: function (a) {
            var b = this.ye;
            if (a != b) {
                this.xc = a;
                this.sv("YAxisTitleBottomMargin", b, this.ye);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aci", {
        get: function () {
            return this.ab2;
        },
        set: function (a) {
            var b = this.aci;
            if (a != b) {
                this.ab2 = a;
                this.sv("XAxisTitleTextColor", b, this.aci);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acp", {
        get: function () {
            return this.ab9;
        },
        set: function (a) {
            var b = this.acp;
            if (a != b) {
                this.ab9 = a;
                this.sv("YAxisTitleTextColor", b, this.acp);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.sz = function (a, b, c) {
        var _this = this;
        _super.prototype.sz.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelTextStyle":
                this.abg(function (d) { return d.textStyle = _this.zc; });
                break;
            case "YAxisLabelTextStyle":
                this.abl(function (d) { return d.textStyle = _this.aah; });
                break;
            case "XAxisTitleTextStyle":
                this.abi(function (d) { return d.textStyle = _this.z0; });
                break;
            case "YAxisTitleTextStyle":
                this.abn(function (d) { return d.textStyle = _this.aa5; });
                break;
        }
    };
    Object.defineProperty(XYChart.prototype, "zc", {
        get: function () {
            return this.yo;
        },
        set: function (a) {
            var b = this.zc;
            if (a != b) {
                this.yo = a;
                this.sv("XAxisLabelTextStyle", b, this.zc);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aah", {
        get: function () {
            return this.ys;
        },
        set: function (a) {
            var b = this.aah;
            if (a != b) {
                this.ys = a;
                this.sv("YAxisLabelTextStyle", b, this.aah);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "z0", {
        get: function () {
            return this.yq;
        },
        set: function (a) {
            var b = this.z0;
            if (a != b) {
                this.yq = a;
                this.sv("XAxisTitleTextStyle", b, this.z0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa5", {
        get: function () {
            return this.yu;
        },
        set: function (a) {
            var b = this.aa5;
            if (a != b) {
                this.yu = a;
                this.sv("YAxisTitleTextStyle", b, this.aa5);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.abb = function (a) {
        var _this = this;
        this.aa8(a, function (b) { return b.leftMargin = _this.xo; });
        this.aa8(a, function (b) { return b.topMargin = _this.xq; });
        this.aa8(a, function (b) { return b.rightMargin = _this.xp; });
        this.aa8(a, function (b) { return b.bottomMargin = _this.xn; });
        this.aa8(a, function (b) { return b.textColor = _this.aca; });
        this.aa9(a, function (b) { return b.textColor = _this.aci; });
        this.aa9(a, function (b) { return b.leftMargin = _this.xy; });
        this.aa9(a, function (b) { return b.topMargin = _this.x1; });
        this.aa9(a, function (b) { return b.rightMargin = _this.x0; });
        this.aa9(a, function (b) { return b.bottomMargin = _this.xx; });
        this.aa8(a, function (b) { return b.textStyle = _this.zc; });
        this.aa9(a, function (b) { return b.textStyle = _this.z0; });
    };
    XYChart.prototype.abd = function (a) {
        var _this = this;
        this.aa8(a, function (b) { return b.leftMargin = _this.x5; });
        this.aa8(a, function (b) { return b.topMargin = _this.x7; });
        this.aa8(a, function (b) { return b.rightMargin = _this.x6; });
        this.aa8(a, function (b) { return b.bottomMargin = _this.x4; });
        this.aa8(a, function (b) { return b.textColor = _this.acb; });
        this.aa9(a, function (b) { return b.textColor = _this.acp; });
        this.aa9(a, function (b) { return b.leftMargin = _this.yf; });
        this.aa9(a, function (b) { return b.topMargin = _this.yi; });
        this.aa9(a, function (b) { return b.rightMargin = _this.yh; });
        this.aa9(a, function (b) { return b.bottomMargin = _this.ye; });
        this.aa8(a, function (b) { return b.textStyle = _this.aah; });
        this.aa9(a, function (b) { return b.textStyle = _this.aa5; });
    };
    XYChart.prototype.tg = function (a) {
        var b = this.bc(a);
        var c = this.c1(a, b);
        if (c == null) {
            return;
        }
        c.itemsSource = a.data;
        c.name = this.oe(a);
        c.title = this.of(a, b);
        this.dataChart.series.add(c);
    };
    XYChart.prototype.abe = function (a) {
        this.abf(a, null);
    };
    XYChart.prototype.abf = function (a, b) {
        a(this.v3());
    };
    XYChart.prototype.abk = function (a, b) {
        a(this.v4());
    };
    XYChart.prototype.abj = function (a) {
        this.abk(a, true);
    };
    XYChart.prototype.rt = function (a, b) {
        this.abe(a);
        this.abk(a, b);
    };
    XYChart.prototype.abi = function (a) {
        this.aa9(this.v3(), a);
    };
    XYChart.prototype.abn = function (a) {
        this.aa9(this.v4(), a);
    };
    XYChart.prototype.abl = function (a) {
        this.aa8(this.v4(), a);
    };
    XYChart.prototype.abg = function (a) {
        this.aa8(this.v3(), a);
    };
    XYChart.prototype.xh = function (a) {
        var b = this.v3();
        return b.ei(a, new ScalerParams(1, this.ve, this.vd, b.cg));
    };
    XYChart.prototype.xj = function (a) {
        var b = this.v3();
        return b.el(a, new ScalerParams(1, this.ve, this.vd, b.cg));
    };
    XYChart.prototype.xi = function (a) {
        var b = this.v4();
        return b.ei(a, new ScalerParams(1, this.ve, this.vd, b.cg));
    };
    XYChart.prototype.xk = function (a) {
        var b = this.v4();
        return b.el(a, new ScalerParams(1, this.ve, this.vd, b.cg));
    };
    XYChart.prototype.td = function (a) {
        _super.prototype.td.call(this, a);
        if (a == null) {
            return;
        }
        if (a.axes != null) {
            var b = this.v3();
            var c = this.v4();
            if (a.axes.contains(b)) {
                a.axes.remove(b);
            }
            if (a.axes.contains(c)) {
                a.axes.remove(c);
            }
        }
    };
    XYChart.yx = function (a, b) {
        var e_1, _a;
        try {
            for (var _b = __values(fromEnum(b)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                var d = c.findMatchingHint(16);
                if (d != null) {
                    var e = d.path;
                    return e;
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        return null;
    };
    XYChart.prototype.r4 = function (a) {
        _super.prototype.r4.call(this, a);
        if (a == this.dataChart) {
            this.dataChart.axes.add(this.v3());
            this.dataChart.axes.add(this.v4());
        }
    };
    XYChart.prototype.aba = function (a) {
        var _this = this;
        if (a != null && this.j8 != null && a.fk == null) {
            a.provideRenderer(this.j8);
        }
        a.jv = this.acd;
        a.eq = this.xr;
        a.jw = this.ace;
        a.et = this.xs;
        a.jy = this.acg;
        a.ff = this.xt;
        a.jx = this.acf;
        a.fg = this.xu;
        a.jz = this.ach;
        a.fh = this.xv;
        a.c3 = true;
        a.ct = this.gb;
        a.cv = this.gb;
        var b = a == this.v3();
        if (b) {
            a.g5 = this.zr;
        }
        a.fz = this.yl;
        a.gf = this.y3;
        a.ag = this.v1;
        this.aa8(a, function (c) { return c.angle = _this.xm; });
        this.aa8(a, function (c) {
            if (c.extent != _this.xl) {
                c.extent = _this.xl;
            }
        });
        this.aa8(a, function (c) { return c.horizontalAlignment = _this.abr; });
        this.aa8(a, function (c) { return c.verticalAlignment = _this.act; });
        this.aa8(a, function (c) { return c.visibility = _this.acy; });
        this.aa8(a, function (c) { return c.location = _this.v9(_this.wc); });
        this.aa9(a, function (c) { return c.horizontalAlignment = _this.abs; });
        this.aa9(a, function (c) { return c.angle = _this.xw; });
        this.abb(a);
    };
    XYChart.prototype.abc = function (a) {
        var _this = this;
        if (a == null) {
            return;
        }
        if (this.j8 != null && a.fk == null) {
            a.provideRenderer(this.j8);
        }
        a.ct = this.gb;
        a.cv = this.gb;
        this.aa8(a, function (b) { return b.angle = _this.x3; });
        this.aa8(a, function (b) {
            if (b.extent != _this.x2) {
                b.extent = _this.x2;
            }
        });
        this.aa8(a, function (b) { return b.horizontalAlignment = _this.abt; });
        this.aa8(a, function (b) { return b.verticalAlignment = _this.acu; });
        this.aa8(a, function (b) { return b.visibility = _this.acz; });
        this.aa8(a, function (b) { return b.location = _this.wa(_this.we); });
        this.aa9(a, function (b) { return b.verticalAlignment = _this.acv; });
        this.aa9(a, function (b) { return b.angle = _this.yd; });
        a.jv = this.ack;
        a.eq = this.x8;
        a.jw = this.acl;
        a.et = this.x9;
        a.jy = this.acn;
        a.ff = this.ya;
        a.jx = this.acm;
        a.fg = this.yb;
        a.jz = this.aco;
        a.fh = this.yc;
        var b = a == this.v4();
        if (b) {
            a.g5 = this.aaw;
        }
        a.gf = this.z8;
        a.ag = this.v2;
        this.abd(a);
    };
    XYChart.prototype.r1 = function () {
        _super.prototype.r1.call(this);
        this.aba(this.v3());
        this.abc(this.v4());
    };
    XYChart.prototype.aa8 = function (a, b) {
        if (a == null) {
            return;
        }
        if (a.ay == null) {
            a.ay = new AxisLabelSettings();
            a.ay.angle = a.d4();
            a.ay.bottomMargin = a.d5();
            a.ay.leftMargin = a.d6();
            a.ay.rightMargin = a.d7();
            a.ay.topMargin = a.d8();
            a.ay.horizontalAlignment = a.jq();
            a.ay.location = a.az();
            a.ay.verticalAlignment = a.kj();
        }
        b(a.ay);
    };
    XYChart.prototype.aa9 = function (a, b) {
        if (a == null) {
            return;
        }
        if (a.be == null) {
            a.be = new TitleSettings();
            a.be.bottomMargin = a.ea();
            a.be.leftMargin = a.eb();
            a.be.rightMargin = a.ec();
            a.be.topMargin = a.ed();
            a.be.location = a.a0();
            a.be.angle = a.d9();
            a.be.horizontalAlignment = a.jr();
            a.be.verticalAlignment = a.kk();
        }
        b(a.be);
    };
    XYChart.prototype.sv = function (a, b, c) {
        var _this = this;
        _super.prototype.sv.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelAngle":
                this.abg(function (d) { return d.angle = _this.xm; });
                break;
            case "YAxisLabelAngle":
                this.abl(function (d) { return d.angle = _this.x3; });
                break;
            case "XAxisLabelHorizontalAlignment":
                this.abg(function (d) { return d.horizontalAlignment = _this.abr; });
                break;
            case "YAxisLabelHorizontalAlignment":
                this.abl(function (d) { return d.horizontalAlignment = _this.abt; });
                break;
            case "XAxisLabelVerticalAlignment":
                this.abg(function (d) { return d.verticalAlignment = _this.act; });
                break;
            case "YAxisLabelVerticalAlignment":
                this.abl(function (d) { return d.verticalAlignment = _this.acu; });
                break;
            case "XAxisLabelVisibility":
                this.abg(function (d) { return d.visibility = _this.acy; });
                break;
            case "YAxisLabelVisibility":
                this.abl(function (d) { return d.visibility = _this.acz; });
                break;
            case "XAxisLabel":
                this.abe(function (d) { return d.fz = _this.yl; });
                break;
            case "YAxisLabel":
                this.abj(function (d) { return d.fz = _this.ym; });
                break;
            case "XAxisLabelFormat":
                this.abe(function (d) { return d.gf = _this.y3; });
                break;
            case "XAxisLabelFormatSpecifiers":
                this.abe(function (d) { return d.ag = _this.v1; });
                break;
            case "YAxisLabelFormat":
                this.abj(function (d) { return d.gf = _this.z8; });
                break;
            case "YAxisLabelFormatSpecifiers":
                this.abj(function (d) { return d.ag = _this.v2; });
                break;
            case "XAxisTitleAngle":
                this.abi(function (d) { return d.angle = _this.xw; });
                break;
            case "YAxisTitleAngle":
                this.abn(function (d) { return d.angle = _this.yd; });
                break;
            case "XAxisTitleAlignment":
                this.abi(function (d) { return d.horizontalAlignment = _this.abs; });
                break;
            case "YAxisTitleAlignment":
                this.abn(function (d) { return d.verticalAlignment = _this.acv; });
                break;
            case "XAxisTitle":
                this.abe(function (d) { return d.g5 = _this.zr; });
                break;
            case "YAxisTitle":
                this.abk(function (d) { return d.g5 = _this.aaw; }, false);
                break;
            case "XAxisMajorStroke":
                this.abf(function (d) { return d.jv = _this.acd; }, a);
                break;
            case "YAxisMajorStroke":
                this.abj(function (d) { return d.jv = _this.ack; });
                break;
            case "XAxisMajorStrokeThickness":
                this.abf(function (d) { return d.eq = _this.xr; }, a);
                break;
            case "YAxisMajorStrokeThickness":
                this.abj(function (d) { return d.eq = _this.x8; });
                break;
            case "XAxisMinorStroke":
                this.abe(function (d) { return d.jw = _this.ace; });
                break;
            case "YAxisMinorStroke":
                this.abj(function (d) { return d.jw = _this.acl; });
                break;
            case "XAxisMinorStrokeThickness":
                this.abe(function (d) { return d.et = _this.xs; });
                break;
            case "YAxisMinorStrokeThickness":
                this.abj(function (d) { return d.et = _this.x9; });
                break;
            case "XAxisTickLength":
                this.abe(function (d) { return d.fg = _this.xu; });
                break;
            case "YAxisTickLength":
                this.abj(function (d) { return d.fg = _this.yb; });
                break;
            case "XAxisTickStroke":
                this.abe(function (d) { return d.jz = _this.ach; });
                break;
            case "YAxisTickStroke":
                this.abj(function (d) { return d.jz = _this.aco; });
                break;
            case "XAxisTickStrokeThickness":
                this.abe(function (d) { return d.fh = _this.xv; });
                break;
            case "YAxisTickStrokeThickness":
                this.abj(function (d) { return d.fh = _this.yc; });
                break;
            case "XAxisStroke":
                this.abe(function (d) { return d.jy = _this.acg; });
                break;
            case "YAxisStroke":
                this.abj(function (d) { return d.jy = _this.acn; });
                break;
            case "XAxisStrokeThickness":
                this.abe(function (d) { return d.ff = _this.xt; });
                break;
            case "YAxisStrokeThickness":
                this.abj(function (d) { return d.ff = _this.ya; });
                break;
            case "XAxisExtent":
                this.abg(function (d) { return d.extent = _this.xl; });
                break;
            case "YAxisExtent":
                this.abl(function (d) { return d.extent = _this.x2; });
                break;
            case "XAxisStrip":
                this.abe(function (d) { return d.jx = _this.acf; });
                break;
            case "YAxisStrip":
                this.abj(function (d) { return d.jx = _this.acm; });
                break;
            case "XAxisInverted":
                this.abe(function (d) { return d.cg = _this.wh; });
                break;
            case "YAxisInverted":
                this.abj(function (d) { return d.cg = _this.wi; });
                break;
            case "YAxisLabelLocation":
                this.abm();
                break;
            case "XAxisLabelLocation":
                this.abh();
                break;
        }
    };
    XYChart.prototype.abm = function () {
        var a = this.wa(this.we);
        if (a == 2 || a == 7) {
            this.abt = 2;
        }
        else {
            this.abt = 0;
        }
        this.abl(function (b) { return b.location = a; });
    };
    XYChart.prototype.abh = function () {
        var a = this.v9(this.wc);
        if (a == 1 || a == 5) {
            this.act = 0;
        }
        else {
            this.act = 2;
        }
        this.abg(function (b) { return b.location = a; });
    };
    XYChart.prototype.wa = function (a) {
        if (a == 0) {
            return 2;
        }
        else if (a == 1) {
            return 3;
        }
        else if (a == 3) {
            return 7;
        }
        else if (a == 2) {
            return 6;
        }
        else {
            if (this.b1 == 1) {
                return 3;
            }
            else {
                return 2;
            }
        }
    };
    XYChart.prototype.v9 = function (a) {
        if (a == 1) {
            return 1;
        }
        else if (a == 0) {
            return 0;
        }
        else if (a == 3) {
            return 5;
        }
        else if (a == 2) {
            return 4;
        }
        else {
            return 1;
        }
    };
    Object.defineProperty(XYChart.prototype, "yl", {
        get: function () {
            return this.yj;
        },
        set: function (a) {
            var b = this.yl;
            if (a != b) {
                this.yj = a;
                this.sv("XAxisLabel", b, this.yl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ym", {
        get: function () {
            return this.yk;
        },
        set: function (a) {
            var b = this.ym;
            if (a != b) {
                this.yk = a;
                this.sv("YAxisLabel", b, this.ym);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acd", {
        get: function () {
            return this.abx;
        },
        set: function (a) {
            var b = this.acd;
            if (a != b) {
                this.abx = a;
                this.sv("XAxisMajorStroke", b, this.acd);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ack", {
        get: function () {
            return this.ab4;
        },
        set: function (a) {
            var b = this.ack;
            if (a != b) {
                this.ab4 = a;
                this.sv("YAxisMajorStroke", b, this.ack);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xr", {
        get: function () {
            return this.wp;
        },
        set: function (a) {
            var b = this.xr;
            if (a != b) {
                this.wp = a;
                this.sv("XAxisMajorStrokeThickness", b, this.xr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x8", {
        get: function () {
            return this.w6;
        },
        set: function (a) {
            var b = this.x8;
            if (a != b) {
                this.w6 = a;
                this.sv("YAxisMajorStrokeThickness", b, this.x8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xs", {
        get: function () {
            return this.wq;
        },
        set: function (a) {
            var b = this.xs;
            if (a != b) {
                this.wq = a;
                this.sv("XAxisMinorStrokeThickness", b, this.xs);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x9", {
        get: function () {
            return this.w7;
        },
        set: function (a) {
            var b = this.x9;
            if (a != b) {
                this.w7 = a;
                this.sv("YAxisMinorStrokeThickness", b, this.x9);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acf", {
        get: function () {
            return this.abz;
        },
        set: function (a) {
            var b = this.acf;
            if (a != b) {
                this.abz = a;
                this.sv("XAxisStrip", b, this.acf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acm", {
        get: function () {
            return this.ab6;
        },
        set: function (a) {
            var b = this.acm;
            if (a != b) {
                this.ab6 = a;
                this.sv("YAxisStrip", b, this.acm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acg", {
        get: function () {
            return this.ab0;
        },
        set: function (a) {
            var b = this.acg;
            if (a != b) {
                this.ab0 = a;
                this.sv("XAxisStroke", b, this.acg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acn", {
        get: function () {
            return this.ab7;
        },
        set: function (a) {
            var b = this.acn;
            if (a != b) {
                this.ab7 = a;
                this.sv("YAxisStroke", b, this.acn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xt", {
        get: function () {
            return this.wr;
        },
        set: function (a) {
            var b = this.xt;
            if (a != b) {
                this.wr = a;
                this.sv("XAxisStrokeThickness", b, this.xt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ya", {
        get: function () {
            return this.w8;
        },
        set: function (a) {
            var b = this.ya;
            if (a != b) {
                this.w8 = a;
                this.sv("YAxisStrokeThickness", b, this.ya);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xu", {
        get: function () {
            return this.ws;
        },
        set: function (a) {
            var b = this.xu;
            if (a != b) {
                this.ws = a;
                this.sv("XAxisTickLength", b, this.xu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "yb", {
        get: function () {
            return this.w9;
        },
        set: function (a) {
            var b = this.yb;
            if (a != b) {
                this.w9 = a;
                this.sv("YAxisTickLength", b, this.yb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ach", {
        get: function () {
            return this.ab1;
        },
        set: function (a) {
            var b = this.ach;
            if (a != b) {
                this.ab1 = a;
                this.sv("XAxisTickStroke", b, this.ach);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aco", {
        get: function () {
            return this.ab8;
        },
        set: function (a) {
            var b = this.aco;
            if (a != b) {
                this.ab8 = a;
                this.sv("YAxisTickStroke", b, this.aco);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xv", {
        get: function () {
            return this.wt;
        },
        set: function (a) {
            var b = this.xv;
            if (a != b) {
                this.wt = a;
                this.sv("XAxisTickStrokeThickness", b, this.xv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "yc", {
        get: function () {
            return this.xa;
        },
        set: function (a) {
            var b = this.yc;
            if (a != b) {
                this.xa = a;
                this.sv("YAxisTickStrokeThickness", b, this.yc);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "zr", {
        get: function () {
            return this.yp;
        },
        set: function (a) {
            var b = this.zr;
            if (a != b) {
                this.yp = a;
                this.sv("XAxisTitle", b, this.zr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aaw", {
        get: function () {
            return this.yt;
        },
        set: function (a) {
            var b = this.aaw;
            if (a != b) {
                this.yt = a;
                this.sv("YAxisTitle", b, this.aaw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ace", {
        get: function () {
            return this.aby;
        },
        set: function (a) {
            var b = this.ace;
            if (a != b) {
                this.aby = a;
                this.sv("XAxisMinorStroke", b, this.ace);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acl", {
        get: function () {
            return this.ab5;
        },
        set: function (a) {
            var b = this.acl;
            if (a != b) {
                this.ab5 = a;
                this.sv("YAxisMinorStroke", b, this.acl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xm", {
        get: function () {
            return this.wk;
        },
        set: function (a) {
            var b = this.xm;
            if (a != b) {
                this.wk = a;
                this.sv("XAxisLabelAngle", b, this.xm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x3", {
        get: function () {
            return this.w1;
        },
        set: function (a) {
            var b = this.x3;
            if (a != b) {
                this.w1 = a;
                this.sv("YAxisLabelAngle", b, this.x3);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xl", {
        get: function () {
            return this.wj;
        },
        set: function (a) {
            var b = this.xl;
            if (a != b) {
                this.wj = a;
                this.sv("XAxisExtent", b, this.xl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "x2", {
        get: function () {
            return this.w0;
        },
        set: function (a) {
            var b = this.x2;
            if (a != b) {
                this.w0 = a;
                this.sv("YAxisExtent", b, this.x2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "xw", {
        get: function () {
            return this.wu;
        },
        set: function (a) {
            var b = this.xw;
            if (a != b) {
                this.wu = a;
                this.sv("XAxisTitleAngle", b, this.xw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "yd", {
        get: function () {
            return this.xb;
        },
        set: function (a) {
            var b = this.yd;
            if (a != b) {
                this.xb = a;
                this.sv("YAxisTitleAngle", b, this.yd);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "wh", {
        get: function () {
            return this.wf;
        },
        set: function (a) {
            var b = this.wh;
            if (a != b) {
                this.wf = a;
                this.sv("XAxisInverted", b, this.wh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "wi", {
        get: function () {
            return this.wg;
        },
        set: function (a) {
            var b = this.wi;
            if (a != b) {
                this.wg = a;
                this.sv("YAxisInverted", b, this.wi);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abs", {
        get: function () {
            return this.abp;
        },
        set: function (a) {
            var b = this.abs;
            if (a != b) {
                this.abp = a;
                this.sv("XAxisTitleAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.abs));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acv", {
        get: function () {
            return this.acs;
        },
        set: function (a) {
            var b = this.acv;
            if (a != b) {
                this.acs = a;
                this.sv("YAxisTitleAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.acv));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abr", {
        get: function () {
            return this.abo;
        },
        set: function (a) {
            var b = this.abr;
            if (a != b) {
                this.abo = a;
                this.sv("XAxisLabelHorizontalAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.abr));
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_abt = function () {
        return this.abq;
    };
    XYChart.prototype.set_abt = function (a) {
        var b = this.abt;
        if (a != b) {
            this.abq = a;
            this.sv("YAxisLabelHorizontalAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.abt));
        }
    };
    Object.defineProperty(XYChart.prototype, "abt", {
        get: function () {
            return this.get_abt();
        },
        set: function (a) {
            this.set_abt(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "act", {
        get: function () {
            return this.acq;
        },
        set: function (a) {
            var b = this.act;
            if (a != b) {
                this.acq = a;
                this.sv("XAxisLabelVerticalAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.act));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acu", {
        get: function () {
            return this.acr;
        },
        set: function (a) {
            var b = this.acu;
            if (a != b) {
                this.acr = a;
                this.sv("YAxisLabelVerticalAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.acu));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acy", {
        get: function () {
            return this.acw;
        },
        set: function (a) {
            var b = this.acy;
            if (a != b) {
                this.acw = a;
                this.sv("XAxisLabelVisibility", enumGetBox(Visibility_$type, b), enumGetBox(Visibility_$type, this.acy));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acz", {
        get: function () {
            return this.acx;
        },
        set: function (a) {
            var b = this.acz;
            if (a != b) {
                this.acx = a;
                this.sv("YAxisLabelVisibility", enumGetBox(Visibility_$type, b), enumGetBox(Visibility_$type, this.acz));
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_we = function () {
        return this.wd;
    };
    XYChart.prototype.set_we = function (a) {
        var b = this.we;
        if (a != b) {
            this.wd = a;
            this.sv("YAxisLabelLocation", enumGetBox(YAxisLabelLocation_$type, b), enumGetBox(YAxisLabelLocation_$type, this.we));
        }
    };
    Object.defineProperty(XYChart.prototype, "we", {
        get: function () {
            return this.get_we();
        },
        set: function (a) {
            this.set_we(a);
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_wc = function () {
        return this.wb;
    };
    XYChart.prototype.set_wc = function (a) {
        var b = this.wc;
        if (a != b) {
            this.wb = a;
            this.sv("XAxisLabelLocation", enumGetBox(XAxisLabelLocation_$type, b), enumGetBox(XAxisLabelLocation_$type, this.wc));
        }
    };
    Object.defineProperty(XYChart.prototype, "wc", {
        get: function () {
            return this.get_wc();
        },
        set: function (a) {
            this.set_wc(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "y3", {
        get: function () {
            return this.yn;
        },
        set: function (a) {
            var b = this.yn;
            if (a != b) {
                this.yn = a;
                this.sv("XAxisLabelFormat", b, this.yn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "v1", {
        get: function () {
            return this.vz;
        },
        set: function (a) {
            var b = this.vz;
            if (a != b) {
                this.vz = a;
                this.sv("XAxisLabelFormatSpecifiers", b, this.vz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "z8", {
        get: function () {
            return this.yr;
        },
        set: function (a) {
            var b = this.yr;
            if (a != b) {
                this.yr = a;
                this.sv("YAxisLabelFormat", b, this.yr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "v2", {
        get: function () {
            return this.v0;
        },
        set: function (a) {
            var b = this.v0;
            if (a != b) {
                this.v0 = a;
                this.sv("YAxisLabelFormatSpecifiers", b, this.v0);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.$t = markType(XYChart, 'XYChart', DomainChart.$);
    return XYChart;
}(DomainChart));
export { XYChart };
