// igniteui-react-charts/esm2015/lib/AbsoluteVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { AbsoluteVolumeOscillatorIndicatorStrategy } from "./AbsoluteVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AbsoluteVolumeOscillatorIndicator = /*@__PURE__*/ (() => {
    class AbsoluteVolumeOscillatorIndicator extends StrategyBasedIndicator {
        get_up() {
            return new AbsoluteVolumeOscillatorIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return AbsoluteVolumeOscillatorIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get shortPeriod() {
            return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.vb));
        }
        set shortPeriod(a) {
            this.h(AbsoluteVolumeOscillatorIndicator.vb, a);
        }
        ux() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.va));
        }
        set longPeriod(a) {
            this.h(AbsoluteVolumeOscillatorIndicator.va, a);
        }
        uu() {
            return this.longPeriod;
        }
    }
    AbsoluteVolumeOscillatorIndicator.$t = /*@__PURE__*/ markType(AbsoluteVolumeOscillatorIndicator, 'AbsoluteVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    AbsoluteVolumeOscillatorIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u9(10, AbsoluteVolumeOscillatorIndicator.$);
    AbsoluteVolumeOscillatorIndicator.va = /*@__PURE__*/ StrategyBasedIndicator.u5(30, AbsoluteVolumeOscillatorIndicator.$);
    return AbsoluteVolumeOscillatorIndicator;
})();
