// igniteui-react-charts/esm2015/lib/AreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeries } from "./HorizontalAnchoredCategorySeries";
import { IUnknownValuePlotting_$type } from "igniteui-react-core";
import { CategoryLineRasterizer } from "./CategoryLineRasterizer";
import { UnknownValuePlotting_$type } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { typeCast, runOn, enumGetBox, markType, markDep } from "igniteui-react-core";
import { Path } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { PropertyMetadata } from "igniteui-react-core";
import { AnchoredCategorySeriesView } from "./AnchoredCategorySeriesView";
import { Defaults } from "./Defaults";
import { ISortingAxis_$type } from "./ISortingAxis";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let AreaSeries = /*@__PURE__*/ (() => {
    class AreaSeries extends HorizontalAnchoredCategorySeries {
        by() {
            return new AreaSeriesView(this);
        }
        r7() {
            return 1;
        }
        b4() {
            if (this.b5 == 0) {
                return 2;
            }
            return super.b4();
        }
        l4(a) {
            super.l4(a);
            this.vj = a;
        }
        get_dg() {
            return true;
        }
        get_df() {
            return true;
        }
        constructor() {
            super();
            this.vk = null;
            this.vk = new CategoryLineRasterizer();
            this.ab = AreaSeries.$;
            this.u3 = 0;
        }
        preferredCategoryMode(a) {
            return this.u3;
        }
        ez(a, b) {
            let c = this.ps(this.px(a), true, false);
            if (this.va(a, c, b)) {
                return true;
            }
            if (this.ey(a, b)) {
                return true;
            }
            if (this.qo(a, b)) {
                return true;
            }
            return false;
        }
        k2(a, b) {
            super.k2(a, b);
            let c = b;
            c.db();
        }
        tf(a, b) {
            super.tf(a, b);
            let c = b.cr.d;
            let d = typeCast(AreaSeriesView.$, b);
            let e = a.f;
            let f = this.getEffectiveViewport1(d);
            this.r5.a2(this, this.sp(), runOn(this, this.rq), this.sx(b), this.sy(b));
            let g = false;
            let h = this.r5.b;
            if (h != null) {
                g = true;
            }
            if (g) {
                let i = new ScalerParams(0, b.bv, b.bu, this.u0.cg, f);
                this.tb(e, -1, this.valueColumn.count, this.u0, i, b.q);
            }
            let j = d.dm;
            let k = d.dn;
            let l = d.dp;
            let m = d.dq;
            this.r5.ad(j, true, false, true, true);
            this.r5.ad(k, true, false, true, true);
            this.r5.ad(l, false, true, false, false);
            this.r5.ad(m, false, true, false, false);
            if (b.cv(a)) {
                d.dd(e.count, e, true, c, this.fx, (n, o, p, q, r) => this.uh(n, a.f.count, b), this.unknownValuePlotting, e.count == 1 ? null : this.to(e, e.count - 1, b.bu, b.bv, true));
                b.cy(a);
            }
            let n = this.fetchYAxis();
            l._opacity = this.r5.j * this.fu;
            m._opacity = 0.5 * this.r5.j * this.fu;
        }
        l9(a, b, c, d) {
            super.l9(a, b, c, d);
            switch (b) {
                case AreaSeries.$$p[0]:
                    this.r2.m = this.unknownValuePlotting == 0;
                    this.mn(false);
                    this.lu();
                    break;
            }
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    AreaSeries.$t = /*@__PURE__*/ markType(AreaSeries, 'AreaSeries', HorizontalAnchoredCategorySeries.$, [IUnknownValuePlotting_$type]);
    AreaSeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, AreaSeries, 'raisePropertyChanged', ['UnknownValuePlotting::vm', [UnknownValuePlotting_$type, /*@__PURE__*/ enumGetBox(UnknownValuePlotting_$type, 1)]]);
    return AreaSeries;
})();
/**
 * @hidden
 */
export let AreaSeriesView = /*@__PURE__*/ (() => {
    class AreaSeriesView extends AnchoredCategorySeriesView {
        constructor(a) {
            super(a);
            this.da = null;
            this.di = new Path();
            this.dk = new Path();
            this.dj = new Path();
            this.dl = new Path();
            this.de = new Path();
            this.dg = new Path();
            this.df = new Path();
            this.dh = new Path();
            this.da = a;
        }
        as() {
            super.as();
            if (!this.q) {
                this.da.ql = Defaults.e;
            }
        }
        get dm() {
            return this.dk;
        }
        get dn() {
            return this.dl;
        }
        get dp() {
            return this.di;
        }
        get dq() {
            return this.dj;
        }
        db() {
            this.di.an = null;
            this.dj.an = null;
            this.dk.an = null;
            this.dl.an = null;
            this.aq();
        }
        dc(a, b, c, d, e, f, g) {
            this.dd(a, b, c, d, e, f, g, null);
        }
        dd(a, b, c, d, e, f, g, h) {
            this.da.vk.a = typeCast(ISortingAxis_$type, this.ct.fetchXAxis()) !== null ? true : false;
            this.da.vk.j(this.di, this.dk, this.dj, this.dl, a, b, c, d, e, f, g, h);
            this.aq();
        }
        bc() {
            super.bc();
            this.de.an = this.di.an;
            this.dg.an = this.dk.an;
            this.df.an = this.dj.an;
            this.dh.an = this.dl.an;
            let a = this.bo();
            this.de._fill = a;
            this.de._opacity = 1;
            this.df._fill = a;
            this.df._opacity = 1;
            this.dg._stroke = a;
            this.dg.ad = this.e.fy + 3;
            this.dh._stroke = a;
            this.dh.ad = this.e.fy + 3;
        }
        a6(a, b) {
            super.a6(a, b);
            if (b) {
                a.t(this.de);
                a.t(this.df);
                a.t(this.dg);
                a.t(this.dh);
            }
            else {
                a.t(this.di);
                a.t(this.dj);
                a.t(this.dk);
                a.t(this.dl);
            }
        }
        al(a) {
            super.al(a);
            this.e.a9.exportPathData(a, this.dk, "lowerShape", ["Lower"]);
            this.e.a9.exportPathData(a, this.dl, "upperShape", ["Upper", "Main"]);
            this.e.a9.exportPathData(a, this.dj, "translucentShape", ["Translucent"]);
            this.e.a9.exportPathData(a, this.di, "fillShape", ["Fill"]);
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.ph = a;
            this.e.g1 = 5;
            this.e.g2 = 1;
            this.e.g3 = -3;
        }
    }
    AreaSeriesView.$t = /*@__PURE__*/ markType(AreaSeriesView, 'AreaSeriesView', AnchoredCategorySeriesView.$);
    return AreaSeriesView;
})();
