// igniteui-react-charts/esm2015/lib/AverageTrueRangeIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageTrueRangeIndicatorStrategy } from "./AverageTrueRangeIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AverageTrueRangeIndicator = /*@__PURE__*/ (() => {
    class AverageTrueRangeIndicator extends StrategyBasedIndicator {
        get_up() {
            return new AverageTrueRangeIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return AverageTrueRangeIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get va() {
            return typeGetValue(this.c(AverageTrueRangeIndicator.vb));
        }
        set va(a) {
            this.h(AverageTrueRangeIndicator.vb, a);
        }
        uv() {
            return this.va;
        }
        u4(a) {
            this.va = a;
        }
    }
    AverageTrueRangeIndicator.$t = /*@__PURE__*/ markType(AverageTrueRangeIndicator, 'AverageTrueRangeIndicator', StrategyBasedIndicator.$);
    AverageTrueRangeIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u7(14, AverageTrueRangeIndicator.$);
    return AverageTrueRangeIndicator;
})();
