// igniteui-react-charts/esm2015/lib/BollingerBandWidthIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { BollingerBandWidthIndicatorStrategy } from "./BollingerBandWidthIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let BollingerBandWidthIndicator = /*@__PURE__*/ (() => {
    class BollingerBandWidthIndicator extends StrategyBasedIndicator {
        get_up() {
            return new BollingerBandWidthIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return BollingerBandWidthIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get vb() {
            return typeGetValue(this.c(BollingerBandWidthIndicator.ve));
        }
        set vb(a) {
            this.h(BollingerBandWidthIndicator.ve, a);
        }
        uv() {
            return this.vb;
        }
        get va() {
            return this.c(BollingerBandWidthIndicator.vd);
        }
        set va(a) {
            this.h(BollingerBandWidthIndicator.vd, a);
        }
        ut() {
            return this.va;
        }
    }
    BollingerBandWidthIndicator.$t = /*@__PURE__*/ markType(BollingerBandWidthIndicator, 'BollingerBandWidthIndicator', StrategyBasedIndicator.$);
    BollingerBandWidthIndicator.ve = /*@__PURE__*/ StrategyBasedIndicator.u7(20, BollingerBandWidthIndicator.$);
    BollingerBandWidthIndicator.vd = /*@__PURE__*/ StrategyBasedIndicator.u6(2, BollingerBandWidthIndicator.$);
    return BollingerBandWidthIndicator;
})();
