// igniteui-react-charts/esm2015/lib/ChaikinOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { ChaikinOscillatorIndicatorStrategy } from "./ChaikinOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ChaikinOscillatorIndicator = /*@__PURE__*/ (() => {
    class ChaikinOscillatorIndicator extends StrategyBasedIndicator {
        get_up() {
            return new ChaikinOscillatorIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return ChaikinOscillatorIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get shortPeriod() {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.vb));
        }
        set shortPeriod(a) {
            this.h(ChaikinOscillatorIndicator.vb, a);
        }
        ux() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.va));
        }
        set longPeriod(a) {
            this.h(ChaikinOscillatorIndicator.va, a);
        }
        uu() {
            return this.longPeriod;
        }
    }
    ChaikinOscillatorIndicator.$t = /*@__PURE__*/ markType(ChaikinOscillatorIndicator, 'ChaikinOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    ChaikinOscillatorIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u9(3, ChaikinOscillatorIndicator.$);
    ChaikinOscillatorIndicator.va = /*@__PURE__*/ StrategyBasedIndicator.u5(10, ChaikinOscillatorIndicator.$);
    return ChaikinOscillatorIndicator;
})();
