// igniteui-react-charts/esm2015/lib/ChaikinVolatilityIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ChaikinVolatilityIndicatorStrategy } from "./ChaikinVolatilityIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ChaikinVolatilityIndicator = /*@__PURE__*/ (() => {
    class ChaikinVolatilityIndicator extends StrategyBasedIndicator {
        get_up() {
            return new ChaikinVolatilityIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return ChaikinVolatilityIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get va() {
            return typeGetValue(this.c(ChaikinVolatilityIndicator.vb));
        }
        set va(a) {
            this.h(ChaikinVolatilityIndicator.vb, a);
        }
        uv() {
            return this.va;
        }
        u4(a) {
            this.va = a;
        }
    }
    ChaikinVolatilityIndicator.$t = /*@__PURE__*/ markType(ChaikinVolatilityIndicator, 'ChaikinVolatilityIndicator', StrategyBasedIndicator.$);
    ChaikinVolatilityIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u7(10, ChaikinVolatilityIndicator.$);
    return ChaikinVolatilityIndicator;
})();
