// igniteui-react-charts/esm2015/lib/FinancialPriceSeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { IFinancialPriceSeriesProxy_$type } from "./IFinancialPriceSeriesProxy";
/**
 * @hidden
 */
export let FinancialPriceSeriesProxy = /*@__PURE__*/ (() => {
    class FinancialPriceSeriesProxy extends Base {
        getCloseMemberPath(a) {
            return a.so;
        }
        getHighMemberPath(a) {
            return a.sr;
        }
        getLowMemberPath(a) {
            return a.sw;
        }
        getOpenMemberPath(a) {
            return a.s0;
        }
        getVolumeMemberPath(a) {
            return a.s4;
        }
        matchesType(a) {
            if (!a.ds) {
                return false;
            }
            if (a.dt || a.du) {
                return false;
            }
            return true;
        }
        setCategoryMode(a, b) {
            a.qs = b;
        }
        setCloseMemberPath(a, b) {
            a.so = b;
        }
        setDisplayType(a, b) {
            a.tw = b;
        }
        setHighMemberPath(a, b) {
            a.sr = b;
        }
        setLowMemberPath(a, b) {
            a.sw = b;
        }
        setOpenMemberPath(a, b) {
            a.s0 = b;
        }
        setVolumeMemberPath(a, b) {
            a.s4 = b;
        }
        setXAxis(a, b) {
            a.qk = b;
        }
        setYAxis(a, b) {
            a.q6 = b;
        }
        setTrendLineType(a, b) {
            if (a.dv) {
                a.trendLineType = b;
            }
        }
    }
    FinancialPriceSeriesProxy.$t = /*@__PURE__*/ markType(FinancialPriceSeriesProxy, 'FinancialPriceSeriesProxy', Base.$, [IFinancialPriceSeriesProxy_$type]);
    return FinancialPriceSeriesProxy;
})();
