// igniteui-react-charts/esm2015/lib/HorizontalRangeCategorySeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RangeCategorySeries } from "./RangeCategorySeries";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { DependencyProperty } from "igniteui-react-core";
import { NumericYAxis } from "./NumericYAxis";
import { Axis } from "./Axis";
import { Point_$type, runOn, typeCast, markType, markDep } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { RangeValueList } from "./RangeValueList";
import { ScalerParams } from "./ScalerParams";
import { CategoryXAxis } from "./CategoryXAxis";
import { DataLegendSeriesValueInfo } from "igniteui-react-core";
import { DataLegendSeriesContext } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
/**
 * @hidden
 */
export let HorizontalRangeCategorySeries = /*@__PURE__*/ (() => {
    class HorizontalRangeCategorySeries extends RangeCategorySeries {
        fetchXAxis() {
            return this.xAxis;
        }
        fetchYAxis() {
            return this.yAxis;
        }
        kz() {
            super.kz();
            this.xAxis = null;
            this.yAxis = null;
        }
        ah(a) {
            let b = this.aj(this.rw(), this.rx.f, this.sx(this.b1), this.sy(this.b1), this.px(a), true);
            if (b == null) {
                return null;
            }
            let c = b[0];
            let d = b[1];
            let e = new Array(1);
            let f = new Array(2);
            f[0] = { $type: Point_$type, x: c[0], y: c[2] };
            f[1] = { $type: Point_$type, x: d[0], y: d[2] };
            e[0] = f;
            return e;
        }
        ai(a) {
            let b = this.aj(this.rw(), this.rx.f, this.sx(this.b1), this.sy(this.b1), this.px(a), true);
            if (b == null) {
                return null;
            }
            let c = b[0];
            let d = b[1];
            let e = new Array(1);
            let f = new Array(2);
            f[0] = { $type: Point_$type, x: c[0], y: c[1] };
            f[1] = { $type: Point_$type, x: d[0], y: d[1] };
            e[0] = f;
            return e;
        }
        get_ee() {
            return true;
        }
        bq() {
            if (this.br == 0) {
                return 1;
            }
            else {
                return super.bq();
            }
        }
        getOffsetValue() {
            return this.RangeFramePreparer.r(this.xAxis, this.b1.bv, this.b1.bu, this.getEffectiveViewport1(this.b1));
        }
        getCategoryWidth() {
            return this.xAxis.getCategorySize(this.b1.bv, this.b1.bu, this.getEffectiveViewport1(this.b1));
        }
        hm(a, b) {
            return this.hn(a, b, this.xAxis, runOn(this, this.sv), new RangeValueList(this.highColumn, this.lowColumn));
        }
        ho(a, b) {
            return this.hp(a, b, this.xAxis, runOn(this, this.sv), new RangeValueList(this.highColumn, this.lowColumn));
        }
        f7(a, b, c, d, e) {
            if (c == null) {
                return Number.POSITIVE_INFINITY;
            }
            let f = this.xAxis.le;
            return this.f8(a, b, this.xAxis, d, e, f, runOn(this, this.sv));
        }
        gn(a, b, c) {
            if (this.ce == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.b1);
            let e = new ScalerParams(0, this.ce.ta, this.b1.bu, this.xAxis.cg, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.ce.ta, this.b1.bu, d);
            return this.gp(new RangeValueList(this.highColumn, this.lowColumn), a, this.xAxis, e, f, runOn(this, this.sv), b, c);
        }
        gl(a, b, c) {
            if (this.ce == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.b1);
            let e = new ScalerParams(0, this.ce.ta, this.b1.bu, this.xAxis.cg, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.ce.ta, this.b1.bu, d);
            return this.gp(this.lowColumn, a, this.xAxis, e, f, runOn(this, this.sv), b, c);
        }
        gj(a, b, c) {
            if (this.ce == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.b1);
            let e = new ScalerParams(0, this.ce.ta, this.b1.bu, this.xAxis.cg, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.ce.ta, this.b1.bu, d);
            return this.gp(this.highColumn, a, this.xAxis, e, f, runOn(this, this.sv), b, c);
        }
        po(a, b, c) {
            return this.pu(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.b1.bv, this.b1.bu, this.getEffectiveViewport1(this.b1)), this.yAxis, this.xAxis, runOn(this, this.gj), (d, e) => this.hp(d, e, this.xAxis, runOn(this, this.sv), this.highColumn), (d, e) => this.hn(d, e, this.xAxis, runOn(this, this.sv), this.highColumn));
        }
        pq(a, b, c) {
            return this.pu(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.b1.bv, this.b1.bu, this.getEffectiveViewport1(this.b1)), this.yAxis, this.xAxis, runOn(this, this.gl), (d, e) => this.hp(d, e, this.xAxis, runOn(this, this.sv), this.lowColumn), (d, e) => this.hn(d, e, this.xAxis, runOn(this, this.sv), this.lowColumn));
        }
        ps(a, b, c) {
            return this.pu(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.b1.bv, this.b1.bu, this.getEffectiveViewport1(this.b1)), this.yAxis, this.xAxis, null, null, null);
        }
        sr() {
            return this.yAxis != null && this.yAxis.c0();
        }
        l9(a, b, c, d) {
            super.l9(a, b, c, d);
            switch (b) {
                case HorizontalRangeCategorySeries.$$p[4]:
                    this.k6(typeCast(Axis.$, c));
                    this.me(typeCast(Axis.$, d));
                    this.r6.cr.g(this.fx);
                    this.mn(false);
                    this.lu();
                    break;
                case HorizontalRangeCategorySeries.$$p[5]:
                    this.k6(typeCast(Axis.$, c));
                    this.me(typeCast(Axis.$, d));
                    this.r6.cr.g(this.fx);
                    this.sr();
                    this.mn(false);
                    this.lu();
                    break;
            }
        }
        t5(a) {
            if (typeCast(NumericYAxis.$, a) !== null) {
                return true;
            }
            return false;
        }
        t4(a) {
            if (typeCast(CategoryXAxis.$, a) !== null || a.cd) {
                return true;
            }
            return false;
        }
        getDataLegendSeriesContextAt(a, b) {
            let c = { $type: Point_$type, x: a, y: b };
            let d = this.gl(c, false, false);
            let e = this.gj(c, false, false);
            let f = new DataLegendSeriesValueInfo();
            f.a = 1;
            f.l = e > d ? d : e;
            f.p = "Low";
            f.q = this.tx != null ? this.tx : "Low";
            f.r = this.ua != null ? this.ua : "L:";
            f.s = this.uc != null ? this.uc : "";
            let g = new DataLegendSeriesValueInfo();
            g.a = 2;
            g.l = e > d ? e : d;
            g.p = "High";
            g.q = this.tu != null ? this.tu : "High";
            g.r = this.t6 != null ? this.t6 : "H:";
            g.s = this.t8 != null ? this.t8 : "";
            let h = new DataLegendSeriesValueInfo();
            h.a = 7;
            h.l = Math.abs(g.l - f.l);
            h.p = "Range";
            h.q = "Range";
            h.r = "R:";
            h.k = true;
            let i = new DataLegendSeriesValueInfo();
            i.a = 6;
            i.l = f.l + (h.l / 2);
            i.p = "Average";
            i.q = "Average";
            i.r = "A:";
            i.k = true;
            let j = new DataLegendSeriesContext();
            j.f = ((() => {
                let $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
                $ret.add(g);
                $ret.add(f);
                $ret.add(h);
                $ret.add(i);
                return $ret;
            })());
            j.b = this.aw();
            return j;
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    HorizontalRangeCategorySeries.$t = /*@__PURE__*/ markType(HorizontalRangeCategorySeries, 'HorizontalRangeCategorySeries', RangeCategorySeries.$);
    HorizontalRangeCategorySeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, HorizontalRangeCategorySeries, 'raisePropertyChanged', ['HighMemberAsLegendLabel:t6:ug', [2, null], 'HighMemberAsLegendUnit:t8:uh', [2, null], 'LowMemberAsLegendLabel:ua:ui', [2, null], 'LowMemberAsLegendUnit:uc:uj', [2, null], 'XAxis::uk', [CategoryAxisBase.$, null], 'YAxis::ul', [NumericYAxis.$, null]]);
    return HorizontalRangeCategorySeries;
})();
