// igniteui-react-charts/esm2015/lib/igr-value-layer.js
import { ValueLayerValueMode_$type } from "./ValueLayerValueMode";
import { IgrSeries } from "./igr-series";
import { IgrAxis } from "./igr-axis";
import { IgrAnnotationLayer } from "./igr-annotation-layer";
import { ValueLayer } from "./ValueLayer";
import { ensureBool, ensureEnum, brushToString, stringToBrush } from "igniteui-react-core";
/**
 * Represents an annotation layer that displays crosshair lines that cross through the closest value of the target series under the cursor.
*/
export class IgrValueLayer extends IgrAnnotationLayer {
    createImplementation() {
        return new ValueLayer();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the series is an crosshair annotation layer.
    */
    get isAnnotationValueLayer() {
        return this.i.de;
    }
    /**
     * Gets or sets the value mode for the overlay.
    */
    get valueMode() {
        return this.i.tg;
    }
    set valueMode(v) {
        this.i.tg = ensureEnum(ValueLayerValueMode_$type, v);
    }
    /**
     * Gets or sets the color to use for the horizontal line. Leave null for an automatic value.
    */
    get horizontalLineStroke() {
        return brushToString(this.i.v2);
    }
    set horizontalLineStroke(v) {
        this.i.v2 = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for vertical line. Leave null for an automatic value.
    */
    get verticalLineStroke() {
        return brushToString(this.i.v3);
    }
    set verticalLineStroke(v) {
        this.i.v3 = stringToBrush(v);
    }
    /**
     * Gets or sets the name of the series to target this annotation to. If null, this annotation targets all series simultaneously.
    */
    get targetSeriesName() {
        return this.i.ub;
    }
    set targetSeriesName(v) {
        this.i.ub = v;
    }
    /**
     * Gets or sets the series to target this annotation to. If null, this annotation targets all series simultaneously.
    */
    get targetSeries() {
        const r = this.i.te;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrSeries._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set targetSeries(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.te = null : this.i.te = v.i;
    }
    /**
     * Gets or sets the name of the axis to target this annotation to. If null, this annotation targets all axis simultaneously.
    */
    get targetAxisName() {
        return this.i.t8;
    }
    set targetAxisName(v) {
        this.i.t8 = v;
    }
    /**
     * Gets or sets the axis to target this annotation to. If null, this annotation targets all value axis simultaneously.
    */
    get targetAxis() {
        const r = this.i.tb;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrAxis._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set targetAxis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.tb = null : this.i.tb = v.i;
    }
    /**
     * Gets or sets whether to use value interpolation when drawing a line through the best value for the cursor position.
    */
    get useInterpolation() {
        return this.i.tn;
    }
    set useInterpolation(v) {
        this.i.tn = ensureBool(v);
    }
    /**
     * Gets or sets whether to draw annotations over the axes where the crosshair meets with them.
    */
    get isAxisAnnotationEnabled() {
        return this.i.tk;
    }
    set isAxisAnnotationEnabled(v) {
        this.i.tk = ensureBool(v);
    }
    /**
     * Gets or sets the color to use for the x axis annotation text. Leave unset for an automatic value.
    */
    get xAxisAnnotationTextColor() {
        return brushToString(this.i.v6);
    }
    set xAxisAnnotationTextColor(v) {
        this.i.v6 = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationBackground() {
        return brushToString(this.i.v4);
    }
    set xAxisAnnotationBackground(v) {
        this.i.v4 = stringToBrush(v);
    }
    /**
     * Gets or sets the corner radius to use for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationBackgroundCornerRadius() {
        return this.i.tq;
    }
    set xAxisAnnotationBackgroundCornerRadius(v) {
        this.i.tq = +v;
    }
    /**
     * Gets or sets the precision to use displaying values for interpolated crosshair positions.
    */
    get xAxisAnnotationInterpolatedValuePrecision() {
        return this.i.t2;
    }
    set xAxisAnnotationInterpolatedValuePrecision(v) {
        this.i.t2 = +v;
    }
    /**
     * Gets or sets the color to use for the x axis annotation outline. Leave unset for an automatic value.
    */
    get xAxisAnnotationOutline() {
        return brushToString(this.i.v5);
    }
    set xAxisAnnotationOutline(v) {
        this.i.v5 = stringToBrush(v);
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingLeft() {
        return this.i.ts;
    }
    set xAxisAnnotationPaddingLeft(v) {
        this.i.ts = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingTop() {
        return this.i.tu;
    }
    set xAxisAnnotationPaddingTop(v) {
        this.i.tu = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingRight() {
        return this.i.tt;
    }
    set xAxisAnnotationPaddingRight(v) {
        this.i.tt = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingBottom() {
        return this.i.tr;
    }
    set xAxisAnnotationPaddingBottom(v) {
        this.i.tr = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingLeft() {
        return this.i.ty;
    }
    set yAxisAnnotationPaddingLeft(v) {
        this.i.ty = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingTop() {
        return this.i.t0;
    }
    set yAxisAnnotationPaddingTop(v) {
        this.i.t0 = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingRight() {
        return this.i.tz;
    }
    set yAxisAnnotationPaddingRight(v) {
        this.i.tz = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingBottom() {
        return this.i.tx;
    }
    set yAxisAnnotationPaddingBottom(v) {
        this.i.tx = +v;
    }
    /**
     * Gets or sets the stroke thickness for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationStrokeThickness() {
        return this.i.tv;
    }
    set xAxisAnnotationStrokeThickness(v) {
        this.i.tv = +v;
    }
    /**
     * Gets or sets the color to use for the y axis annotation text. Leave unset for an automatic value.
    */
    get yAxisAnnotationTextColor() {
        return brushToString(this.i.v9);
    }
    set yAxisAnnotationTextColor(v) {
        this.i.v9 = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationBackground() {
        return brushToString(this.i.v7);
    }
    set yAxisAnnotationBackground(v) {
        this.i.v7 = stringToBrush(v);
    }
    /**
     * Gets or sets the corner radius to use for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationBackgroundCornerRadius() {
        return this.i.tw;
    }
    set yAxisAnnotationBackgroundCornerRadius(v) {
        this.i.tw = +v;
    }
    /**
     * Gets or sets the precision to use displaying values for interpolated crosshair positions.
    */
    get yAxisAnnotationInterpolatedValuePrecision() {
        return this.i.t3;
    }
    set yAxisAnnotationInterpolatedValuePrecision(v) {
        this.i.t3 = +v;
    }
    /**
     * Gets or sets the color to use for the y axis annotation outline. Leave unset for an automatic value.
    */
    get yAxisAnnotationOutline() {
        return brushToString(this.i.v8);
    }
    set yAxisAnnotationOutline(v) {
        this.i.v8 = stringToBrush(v);
    }
    /**
     * Gets or sets the stroke thickness for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationStrokeThickness() {
        return this.i.t1;
    }
    set yAxisAnnotationStrokeThickness(v) {
        this.i.t1 = +v;
    }
    /**
     * Gets or sets whether to skip unknown values when searching for series values.
    */
    get skipUnknownValues() {
        return this.i.tm;
    }
    set skipUnknownValues(v) {
        this.i.tm = ensureBool(v);
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.targetSeries && this.targetSeries.name && this.targetSeries.name == name) {
            return this.targetSeries;
        }
        if (this.targetAxis && this.targetAxis.name && this.targetAxis.name == name) {
            return this.targetAxis;
        }
        return null;
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.targetSeries && this.targetSeries._styling) {
            this.targetSeries._styling(container, component, this);
        }
        if (this.targetAxis && this.targetAxis._styling) {
            this.targetAxis._styling(container, component, this);
        }
        this._inStyling = false;
    }
}
