// igniteui-react-charts/esm2015/lib/igr-marker-series.js
import { MarkerType_$type } from "./MarkerType";
import { MarkerOutlineMode_$type } from "./MarkerOutlineMode";
import { MarkerFillMode_$type } from "./MarkerFillMode";
import { IgrSeries } from "./igr-series";
import { ensureEnum, brushToString, stringToBrush } from "igniteui-react-core";
/**
 * Represents the base class for series containing markers.
*
* MarkerSeries represents the base class for series containing markers.
*
* Example:
*
* ```ts
*  <IgrDataChart
*     dataSource={this.state.dataSource} >
*
*     <IgrCategoryXAxis name="xAxis" />
*     <IgrNumericYAxis name="yAxis" />
*
*     <IgrColumnSeries
*      name="series1"
*     xAxisName="xAxis"
* 	yAxisName="yAxis"
* 	valueMemberPath="Value" />
*  </IgrDataChart>
* ```
*
* ```ts
* this.series = new IgrMarkerSeries({name: "series"});
* this.series.xAxis = this.categoryXAxis;
* this.series.yAxis = this.numericYAxis;
* this.series.xAxisName = "timeXAxis";
* this.series.yAxisName = "numericYAxis";
* this.series.volumeMemberPath="Volume";
* ```
*/
export class IgrMarkerSeries extends IgrSeries {
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Returns whether the current series supports visual markers.
    *
    * `HasMarkers` property is overriden, it returns whether the current series supports visual markers. You can use it like this:
    *
    * ```ts
    *  let gotMarkers: boolean = this.series.hasMarkers;
    * ```
    */
    get hasMarkers() {
        return this.i.hasMarkers;
    }
    /**
     * Gets or sets the marker type for the current series object.
     * This property is ignored when the MarkerTemplate property is set
    *
    * `MarkerType` property gets or sets the marker type for the current series object. For example, you can set it like this:
    *
    * ```ts
    * this.series.markerType = "circle";
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrCategoryXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrColumnSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    * 	markerType = "circle"/>
    *  </IgrDataChart>
    * ```
    */
    get markerType() {
        return this.i.ql;
    }
    set markerType(v) {
        this.i.ql = ensureEnum(MarkerType_$type, v);
    }
    /**
     * Gets or sets whether the marker outline is based on the marker brush of the series rather than the marker outlines collection.
    */
    get markerOutlineMode() {
        return this.i.qi;
    }
    set markerOutlineMode(v) {
        this.i.qi = ensureEnum(MarkerOutlineMode_$type, v);
    }
    /**
     * Gets or sets whether the marker fill is based on the marker outline of the series rather than the marker brushes collection.
    */
    get markerFillMode() {
        return this.i.qe;
    }
    set markerFillMode(v) {
        this.i.qe = ensureEnum(MarkerFillMode_$type, v);
    }
    /**
     * Gets or sets the MarkerTemplate for the current series object.
    *
    * `MarkerTemplate` property gets or sets the MarkerTemplate for the current series object. For example, you can use it like this:
    */
    get markerTemplate() {
        return this.i.q7;
    }
    set markerTemplate(v) {
        this.i.q7 = v;
    }
    /**
     * Gets or sets thickness of the marker outline
    */
    get markerThickness() {
        return this.i.qq;
    }
    set markerThickness(v) {
        this.i.qq = +v;
    }
    /**
     * Gets the effective marker template for the current series object.
    *
    * `ActualMarkerTemplate` property gets the effective marker template for the current series object. You can use it like this:
    */
    get actualMarkerTemplate() {
        return this.i.q5;
    }
    set actualMarkerTemplate(v) {
        this.i.q5 = v;
    }
    /**
     * Gets or sets the brush that specifies how the current series object's marker interiors are painted.
    *
    * `MarkerBrush` property gets or sets the brush that specifies how the current series object's marker interiors are painted. You can use it like this:
    *
    * ```ts
    * this.series.markerBrush = "red";
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrCategoryXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrColumnSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    * 	markerBrush = "blue" />
    *  </IgrDataChart>
    * ```
    */
    get markerBrush() {
        return brushToString(this.i.rn);
    }
    set markerBrush(v) {
        this.i.rn = stringToBrush(v);
    }
    /**
     * Gets the effective marker brush for the current series object.
    *
    * `ActualMarkerBrush` property gets the effective marker brush for the current series object. For example, you can use it like this:
    *
    * ```ts
    *  let effectiveMarkerBrush: string = this.series.actualMarkerBrush;
    *
    * ```
    */
    get actualMarkerBrush() {
        return brushToString(this.i.rl);
    }
    set actualMarkerBrush(v) {
        this.i.rl = stringToBrush(v);
    }
    /**
     * Gets or sets the brush that specifies how the current series object's marker outlines are painted.
    *
    * `MarkerOutline` property gets or sets the brush that specifies how the current series object's marker outlines are painted. You can use it like this:
    *
    * ```ts
    * this.series.markerOutline = "red";
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrCategoryXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrColumnSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    * 	markerOutline = "blue" />
    *  </IgrDataChart>
    * ```
    */
    get markerOutline() {
        return brushToString(this.i.ro);
    }
    set markerOutline(v) {
        this.i.ro = stringToBrush(v);
    }
    /**
     * Gets the effective marker outline for the current series object.
    *
    *     `ActualMarkerOutline` property gets the effective marker outline for the current series object. You can use it like this:
    *
    * ```ts
    *  let effectiveMarkerOutline: string = this.series.actualMarkerOutline;
    *
    * ```
    */
    get actualMarkerOutline() {
        return brushToString(this.i.rm);
    }
    set actualMarkerOutline(v) {
        this.i.rm = stringToBrush(v);
    }
}
