// igniteui-react-charts/esm2015/lib/igr-radial-pie-series.js
import { IgrAnchoredRadialSeries } from "./igr-anchored-radial-series";
import { RadialPieSeries } from "./RadialPieSeries";
/**
 * Represents a IgxDataChartComponent radial pie series.
 * Compare values across categories by using radial pie slices.
*/
export class IgrRadialPieSeries extends IgrAnchoredRadialSeries {
    createImplementation() {
        return new RadialPieSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the current series shows a pie shape.
    */
    get isPie() {
        return this.i.d7;
    }
    /**
     * Overridden by derived series classes to indicate when marker-less display is preferred or not.
    */
    get isMarkerlessDisplayPreferred() {
        return this.i.d5;
    }
    /**
     * Gets or sets the x-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusX() {
        return this.i.ug;
    }
    set radiusX(v) {
        this.i.ug = +v;
    }
    /**
     * Gets or sets the y-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusY() {
        return this.i.uh;
    }
    set radiusY(v) {
        this.i.uh = +v;
    }
    /**
     * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
    
    */
    getItemSpan() {
        let iv = this.i.ge();
        return (iv);
    }
}
