// igniteui-react-charts/esm2015/lib/igr-data-tool-tip-layer.js
import { IgrAxis } from "./igr-axis";
import { DataToolTipLayerGroupingMode_$type } from "igniteui-react-core";
import { DataTooltipGroupedPositionX_$type } from "igniteui-react-core";
import { DataTooltipGroupedPositionY_$type } from "igniteui-react-core";
import { DataAbbreviationMode_$type } from "igniteui-react-core";
import { DataLegendValueMode_$type } from "igniteui-react-core";
import { DataLegendHeaderDateMode_$type } from "igniteui-react-core";
import { DataLegendHeaderTimeMode_$type } from "igniteui-react-core";
import { DataLegendSummaryType_$type } from "igniteui-react-core";
import { LegendItemBadgeShape_$type } from "igniteui-react-core";
import { DataLegendUnitsMode_$type } from "igniteui-react-core";
import { DataLegendLabelMode_$type } from "igniteui-react-core";
import { IgrAnnotationLayer } from "./igr-annotation-layer";
import { DataToolTipLayer } from "./DataToolTipLayer";
import { ensureBool, ensureEnum, brushToString, stringToBrush, arrayFindByName } from "igniteui-react-core";
import { Thickness } from "igniteui-react-core";
import { FontInfo } from "igniteui-react-core";
/**
 * Represents an annotation layer that displays grouped tooltips with summaries for series.
*/
export class IgrDataToolTipLayer extends IgrAnnotationLayer {
    createImplementation() {
        return new DataToolTipLayer();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the series is an annotation layer displayed only when hovering over the chart.
    */
    get isAnnotationHoverLayer() {
        return this.i.dc;
    }
    /**
     * Gets whether the default tooltip behaviors for the chart are disabled if this layer is present.
    */
    get isDefaultTooltipBehaviorDisabled() {
        return this.i.dn;
    }
    /**
     * Gets or sets the Axis to target this annotation to. If null, this annotation layer will not render content.
    */
    get targetAxis() {
        const r = this.i.tk;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrAxis._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set targetAxis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.tk = null : this.i.tk = v.i;
    }
    /**
     * Gets or sets whether to use value interpolation when drawing the tooltips.
    */
    get useInterpolation() {
        return this.i.uy;
    }
    set useInterpolation(v) {
        this.i.uy = ensureBool(v);
    }
    /**
     * Gets or sets the grouping mode to apply to use for the tooltip layer.
    */
    get groupingMode() {
        return this.i.t2;
    }
    set groupingMode(v) {
        this.i.t2 = ensureEnum(DataToolTipLayerGroupingMode_$type, v);
    }
    /**
     * Gets the actual resolved tooltip layer grouping mode.
    */
    get actualGroupingMode() {
        return this.i.t1;
    }
    set actualGroupingMode(v) {
        this.i.t1 = ensureEnum(DataToolTipLayerGroupingMode_$type, v);
    }
    /**
     * Gets or sets the offset of the tooltip layer on the X axis.
    */
    get positionOffsetX() {
        return this.i.u7;
    }
    set positionOffsetX(v) {
        this.i.u7 = +v;
    }
    /**
     * Gets or sets the offset of the tooltip layer on the Y axis.
    */
    get positionOffsetY() {
        return this.i.u8;
    }
    set positionOffsetY(v) {
        this.i.u8 = +v;
    }
    /**
     * Gets or sets the default offset of the tooltip layer on the X axis.
    */
    get defaultPositionOffsetX() {
        return this.i.u5;
    }
    set defaultPositionOffsetX(v) {
        this.i.u5 = +v;
    }
    /**
     * Gets or sets the default offset of the tooltip layer on the Y axis.
    */
    get defaultPositionOffsetY() {
        return this.i.u6;
    }
    set defaultPositionOffsetY(v) {
        this.i.u6 = +v;
    }
    /**
     * Gets or sets the tooltip position mode on the X axis for grouped series.
    */
    get groupedPositionModeX() {
        return this.i.tu;
    }
    set groupedPositionModeX(v) {
        this.i.tu = ensureEnum(DataTooltipGroupedPositionX_$type, v);
    }
    /**
     * Gets the actual resolved tooltip position mode on the X axis for grouped series.
    */
    get actualGroupedPositionModeX() {
        return this.i.tt;
    }
    set actualGroupedPositionModeX(v) {
        this.i.tt = ensureEnum(DataTooltipGroupedPositionX_$type, v);
    }
    /**
     * Gets or sets the tooltip position mode on the Y axis for grouped series.
    */
    get groupedPositionModeY() {
        return this.i.ty;
    }
    set groupedPositionModeY(v) {
        this.i.ty = ensureEnum(DataTooltipGroupedPositionY_$type, v);
    }
    /**
     * Gets the actual resolved tooltip position mode on the Y axis for grouped series.
    */
    get actualGroupedPositionModeY() {
        return this.i.tx;
    }
    set actualGroupedPositionModeY(v) {
        this.i.tx = ensureEnum(DataTooltipGroupedPositionY_$type, v);
    }
    /**
     * Gets or sets whether the data legend should update when the series data is mutated.
    */
    get shouldUpdateWhenSeriesDataChanges() {
        return this.i.ux;
    }
    set shouldUpdateWhenSeriesDataChanges(v) {
        this.i.ux = ensureBool(v);
    }
    /**
     * Gets or sets indexes or names of series to include in displaying in the data legend, e.g. "0, 1"
    */
    get includedSeries() {
        return this.i.th;
    }
    set includedSeries(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.th = v;
    }
    /**
     * Gets or sets indexes or names of series to exclude from displaying in the data legend, e.g. "0, 1"
     * The ExcludedSeries property takes precedences over values of IncludedSeries property
    */
    get excludedSeries() {
        return this.i.tc;
    }
    set excludedSeries(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.tc = v;
    }
    /**
     * Gets or sets names of data columns or their labels to include in displaying in the data legend, e.g. "High, Low" or "H, L"
     * The ExcludedColumns property takes precedences over values of IncludedColumns property
    */
    get includedColumns() {
        return this.i.tg;
    }
    set includedColumns(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.tg = v;
    }
    /**
     * Gets or sets names of data columns or their labels to exclude from displaying in the data legend, e.g. "High, Low" or "H, L"
     * The ExcludedColumns property takes precedences over values of IncludedColumns property
    */
    get excludedColumns() {
        return this.i.tb;
    }
    set excludedColumns(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.tb = v;
    }
    /**
     * Gets or sets mode for abbreviating large numbers displayed in the legend
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get valueFormatAbbreviation() {
        return this.i.tl;
    }
    set valueFormatAbbreviation(v) {
        this.i.tl = ensureEnum(DataAbbreviationMode_$type, v);
    }
    /**
     * Gets or sets maximum digits for formating numbers displayed in the legend
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get valueFormatMaxFractions() {
        return this.i.v7;
    }
    set valueFormatMaxFractions(v) {
        this.i.v7 = +v;
    }
    /**
     * Gets or sets minimum digits for formating numbers displayed in the legend
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get valueFormatMinFractions() {
        return this.i.v8;
    }
    set valueFormatMinFractions(v) {
        this.i.v8 = +v;
    }
    /**
     * Gets or sets the mode for displaying values in the data legend, e.g. Currency ($500.25), Decimal (500.25), Integer (500)
    */
    get valueFormatMode() {
        return this.i.tr;
    }
    set valueFormatMode(v) {
        this.i.tr = ensureEnum(DataLegendValueMode_$type, v);
    }
    /**
     * Gets or sets globalization culture when displaying values as currencies, e.g. use "en-GB" to display British pound symbol when the ValueFormatMode property is set to 'Currency' mode
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get valueFormatCulture() {
        return this.i.x2;
    }
    set valueFormatCulture(v) {
        this.i.x2 = v;
    }
    /**
     * Gets or sets whether or not use grouping separator, e.g, 15,000 for 15000
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get valueFormatUseGrouping() {
        return this.i.uz;
    }
    set valueFormatUseGrouping(v) {
        this.i.uz = ensureBool(v);
    }
    /**
     * Gets or sets the format string for values displayed in the data legend.
    */
    get valueFormatString() {
        return this.i.x8;
    }
    set valueFormatString(v) {
        this.i.x8 = v;
    }
    /**
     * Gets or sets the format specifiers to use with the ValueFormatString string.
    */
    get valueFormatSpecifiers() {
        return this.i.ti;
    }
    set valueFormatSpecifiers(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.ti = v;
    }
    get valueRowMarginBottom() {
        return this.i.aaz ? this.i.aaz.bottom : NaN;
    }
    set valueRowMarginBottom(v) {
        this.ensureValueRowMargin();
        this.i.aaz.bottom = +v;
        this.i.aaz = this.i.aaz;
    }
    get valueRowMarginLeft() {
        return this.i.aaz ? this.i.aaz.left : NaN;
    }
    set valueRowMarginLeft(v) {
        this.ensureValueRowMargin();
        this.i.aaz.left = +v;
        this.i.aaz = this.i.aaz;
    }
    get valueRowMarginRight() {
        return this.i.aaz ? this.i.aaz.right : NaN;
    }
    set valueRowMarginRight(v) {
        this.ensureValueRowMargin();
        this.i.aaz.right = +v;
        this.i.aaz = this.i.aaz;
    }
    get valueRowMarginTop() {
        return this.i.aaz ? this.i.aaz.top : NaN;
    }
    set valueRowMarginTop(v) {
        this.ensureValueRowMargin();
        this.i.aaz.top = +v;
        this.i.aaz = this.i.aaz;
    }
    ensureValueRowMargin() {
        if (this.i.aaz) {
            return;
        }
        this.i.aaz = new Thickness(2);
    }
    /**
     * Gets or sets whether to show series rows.
    */
    get valueRowVisible() {
        return this.i.u0;
    }
    set valueRowVisible(v) {
        this.i.u0 = ensureBool(v);
    }
    /**
     * Gets or sets text displayed when data column is missing a value, e.g. "no data"
    */
    get valueTextWhenMissingData() {
        return this.i.yh;
    }
    set valueTextWhenMissingData(v) {
        this.i.yh = v;
    }
    /**
     * Gets or sets whether to use series colors when displaying values in the legend
    */
    get valueTextUseSeriesColors() {
        return this.i.u1;
    }
    set valueTextUseSeriesColors(v) {
        this.i.u1 = ensureBool(v);
    }
    get valueTextMarginBottom() {
        return this.i.aa0 ? this.i.aa0.bottom : NaN;
    }
    set valueTextMarginBottom(v) {
        this.ensureValueTextMargin();
        this.i.aa0.bottom = +v;
        this.i.aa0 = this.i.aa0;
    }
    get valueTextMarginLeft() {
        return this.i.aa0 ? this.i.aa0.left : NaN;
    }
    set valueTextMarginLeft(v) {
        this.ensureValueTextMargin();
        this.i.aa0.left = +v;
        this.i.aa0 = this.i.aa0;
    }
    get valueTextMarginRight() {
        return this.i.aa0 ? this.i.aa0.right : NaN;
    }
    set valueTextMarginRight(v) {
        this.ensureValueTextMargin();
        this.i.aa0.right = +v;
        this.i.aa0 = this.i.aa0;
    }
    get valueTextMarginTop() {
        return this.i.aa0 ? this.i.aa0.top : NaN;
    }
    set valueTextMarginTop(v) {
        this.ensureValueTextMargin();
        this.i.aa0.top = +v;
        this.i.aa0 = this.i.aa0;
    }
    ensureValueTextMargin() {
        if (this.i.aa0) {
            return;
        }
        this.i.aa0 = new Thickness(2);
    }
    /**
     * Gets or sets the units text color.
    */
    get valueTextColor() {
        return brushToString(this.i.aak);
    }
    set valueTextColor(v) {
        this.i.aak = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get valueTextStyle() {
        if (this.i.uo == null) {
            return null;
        }
        return this.i.uo.fontString;
    }
    set valueTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.uo = fi;
    }
    /**
     * Gets or sets the format string for header text displayed in the data legend.
    */
    get headerFormatString() {
        return this.i.wq;
    }
    set headerFormatString(v) {
        this.i.wq = v;
    }
    /**
     * Gets or sets the format specifiers to use with the HeaderFormatString string.
    */
    get headerFormatSpecifiers() {
        return this.i.tf;
    }
    set headerFormatSpecifiers(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.tf = v;
    }
    /**
     * Gets or sets globalization culture when displaying header as date time
     * This property is ignored when setting HeaderFormatSpecifiers or HeaderFormatString properties
    */
    get headerFormatCulture() {
        return this.i.wm;
    }
    set headerFormatCulture(v) {
        this.i.wm = v;
    }
    /**
     * Gets or sets date format for the header
     * This property is ignored when setting HeaderFormatSpecifiers or HeaderFormatString properties
    */
    get headerFormatDate() {
        return this.i.tm;
    }
    set headerFormatDate(v) {
        this.i.tm = ensureEnum(DataLegendHeaderDateMode_$type, v);
    }
    /**
     * Gets or sets time format for the header
     * This property is ignored when setting HeaderFormatSpecifiers or HeaderFormatString properties
    */
    get headerFormatTime() {
        return this.i.tn;
    }
    set headerFormatTime(v) {
        this.i.tn = ensureEnum(DataLegendHeaderTimeMode_$type, v);
    }
    /**
     * Gets or sets the HeaderText for the data legend.
    */
    get headerText() {
        return this.i.wv;
    }
    set headerText(v) {
        this.i.wv = v;
    }
    /**
     * Gets or sets the header text color.
    */
    get headerTextColor() {
        return brushToString(this.i.aac);
    }
    set headerTextColor(v) {
        this.i.aac = stringToBrush(v);
    }
    get headerTextMarginBottom() {
        return this.i.aat ? this.i.aat.bottom : NaN;
    }
    set headerTextMarginBottom(v) {
        this.ensureHeaderTextMargin();
        this.i.aat.bottom = +v;
        this.i.aat = this.i.aat;
    }
    get headerTextMarginLeft() {
        return this.i.aat ? this.i.aat.left : NaN;
    }
    set headerTextMarginLeft(v) {
        this.ensureHeaderTextMargin();
        this.i.aat.left = +v;
        this.i.aat = this.i.aat;
    }
    get headerTextMarginRight() {
        return this.i.aat ? this.i.aat.right : NaN;
    }
    set headerTextMarginRight(v) {
        this.ensureHeaderTextMargin();
        this.i.aat.right = +v;
        this.i.aat = this.i.aat;
    }
    get headerTextMarginTop() {
        return this.i.aat ? this.i.aat.top : NaN;
    }
    set headerTextMarginTop(v) {
        this.ensureHeaderTextMargin();
        this.i.aat.top = +v;
        this.i.aat = this.i.aat;
    }
    ensureHeaderTextMargin() {
        if (this.i.aat) {
            return;
        }
        this.i.aat = new Thickness(2);
    }
    get headerRowMarginBottom() {
        return this.i.aas ? this.i.aas.bottom : NaN;
    }
    set headerRowMarginBottom(v) {
        this.ensureHeaderRowMargin();
        this.i.aas.bottom = +v;
        this.i.aas = this.i.aas;
    }
    get headerRowMarginLeft() {
        return this.i.aas ? this.i.aas.left : NaN;
    }
    set headerRowMarginLeft(v) {
        this.ensureHeaderRowMargin();
        this.i.aas.left = +v;
        this.i.aas = this.i.aas;
    }
    get headerRowMarginRight() {
        return this.i.aas ? this.i.aas.right : NaN;
    }
    set headerRowMarginRight(v) {
        this.ensureHeaderRowMargin();
        this.i.aas.right = +v;
        this.i.aas = this.i.aas;
    }
    get headerRowMarginTop() {
        return this.i.aas ? this.i.aas.top : NaN;
    }
    set headerRowMarginTop(v) {
        this.ensureHeaderRowMargin();
        this.i.aas.top = +v;
        this.i.aas = this.i.aas;
    }
    ensureHeaderRowMargin() {
        if (this.i.aas) {
            return;
        }
        this.i.aas = new Thickness(2);
    }
    /**
     * Gets or sets whether to show Header row.
    */
    get headerRowVisible() {
        return this.i.uq;
    }
    set headerRowVisible(v) {
        this.i.uq = ensureBool(v);
    }
    /**
     * Gets or Sets the style to use for the header text.
    */
    get headerTextStyle() {
        if (this.i.ug == null) {
            return null;
        }
        return this.i.ug.fontString;
    }
    set headerTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.ug = fi;
    }
    /**
     * Gets or sets the summary text color.
    */
    get summaryTitleTextColor() {
        return brushToString(this.i.aaf);
    }
    set summaryTitleTextColor(v) {
        this.i.aaf = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the summary text.
    */
    get summaryTitleTextStyle() {
        if (this.i.uj == null) {
            return null;
        }
        return this.i.uj.fontString;
    }
    set summaryTitleTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.uj = fi;
    }
    /**
     * Gets or sets the SummaryType for the data legend.
    */
    get summaryType() {
        return this.i.tp;
    }
    set summaryType(v) {
        this.i.tp = ensureEnum(DataLegendSummaryType_$type, v);
    }
    /**
     * Gets or sets the SummaryTitleText for the data legend.
    */
    get summaryTitleText() {
        return this.i.xe;
    }
    set summaryTitleText(v) {
        this.i.xe = v;
    }
    get summaryTitleTextMarginBottom() {
        return this.i.aaw ? this.i.aaw.bottom : NaN;
    }
    set summaryTitleTextMarginBottom(v) {
        this.ensureSummaryTitleTextMargin();
        this.i.aaw.bottom = +v;
        this.i.aaw = this.i.aaw;
    }
    get summaryTitleTextMarginLeft() {
        return this.i.aaw ? this.i.aaw.left : NaN;
    }
    set summaryTitleTextMarginLeft(v) {
        this.ensureSummaryTitleTextMargin();
        this.i.aaw.left = +v;
        this.i.aaw = this.i.aaw;
    }
    get summaryTitleTextMarginRight() {
        return this.i.aaw ? this.i.aaw.right : NaN;
    }
    set summaryTitleTextMarginRight(v) {
        this.ensureSummaryTitleTextMargin();
        this.i.aaw.right = +v;
        this.i.aaw = this.i.aaw;
    }
    get summaryTitleTextMarginTop() {
        return this.i.aaw ? this.i.aaw.top : NaN;
    }
    set summaryTitleTextMarginTop(v) {
        this.ensureSummaryTitleTextMargin();
        this.i.aaw.top = +v;
        this.i.aaw = this.i.aaw;
    }
    ensureSummaryTitleTextMargin() {
        if (this.i.aaw) {
            return;
        }
        this.i.aaw = new Thickness(2);
    }
    get summaryRowMarginBottom() {
        return this.i.aav ? this.i.aav.bottom : NaN;
    }
    set summaryRowMarginBottom(v) {
        this.ensureSummaryRowMargin();
        this.i.aav.bottom = +v;
        this.i.aav = this.i.aav;
    }
    get summaryRowMarginLeft() {
        return this.i.aav ? this.i.aav.left : NaN;
    }
    set summaryRowMarginLeft(v) {
        this.ensureSummaryRowMargin();
        this.i.aav.left = +v;
        this.i.aav = this.i.aav;
    }
    get summaryRowMarginRight() {
        return this.i.aav ? this.i.aav.right : NaN;
    }
    set summaryRowMarginRight(v) {
        this.ensureSummaryRowMargin();
        this.i.aav.right = +v;
        this.i.aav = this.i.aav;
    }
    get summaryRowMarginTop() {
        return this.i.aav ? this.i.aav.top : NaN;
    }
    set summaryRowMarginTop(v) {
        this.ensureSummaryRowMargin();
        this.i.aav.top = +v;
        this.i.aav = this.i.aav;
    }
    ensureSummaryRowMargin() {
        if (this.i.aav) {
            return;
        }
        this.i.aav = new Thickness(2);
    }
    /**
     * Gets or sets the units text color.
    */
    get summaryValueTextColor() {
        return brushToString(this.i.aah);
    }
    set summaryValueTextColor(v) {
        this.i.aah = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get summaryValueTextStyle() {
        if (this.i.ul == null) {
            return null;
        }
        return this.i.ul.fontString;
    }
    set summaryValueTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.ul = fi;
    }
    /**
     * Gets or sets the units text for the data legend.
    */
    get summaryLabelText() {
        return this.i.w9;
    }
    set summaryLabelText(v) {
        this.i.w9 = v;
    }
    /**
     * Gets or sets the units text color.
    */
    get summaryLabelTextColor() {
        return brushToString(this.i.aae);
    }
    set summaryLabelTextColor(v) {
        this.i.aae = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get summaryLabelTextStyle() {
        if (this.i.ui == null) {
            return null;
        }
        return this.i.ui.fontString;
    }
    set summaryLabelTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.ui = fi;
    }
    /**
     * Gets or sets the units text for the data legend.
    */
    get summaryUnitsText() {
        return this.i.xk;
    }
    set summaryUnitsText(v) {
        this.i.xk = v;
    }
    /**
     * Gets or sets the units text color.
    */
    get summaryUnitsTextColor() {
        return brushToString(this.i.aag);
    }
    set summaryUnitsTextColor(v) {
        this.i.aag = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get summaryUnitsTextStyle() {
        if (this.i.uk == null) {
            return null;
        }
        return this.i.uk.fontString;
    }
    set summaryUnitsTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.uk = fi;
    }
    get badgeMarginBottom() {
        return this.i.aar ? this.i.aar.bottom : NaN;
    }
    set badgeMarginBottom(v) {
        this.ensureBadgeMargin();
        this.i.aar.bottom = +v;
        this.i.aar = this.i.aar;
    }
    get badgeMarginLeft() {
        return this.i.aar ? this.i.aar.left : NaN;
    }
    set badgeMarginLeft(v) {
        this.ensureBadgeMargin();
        this.i.aar.left = +v;
        this.i.aar = this.i.aar;
    }
    get badgeMarginRight() {
        return this.i.aar ? this.i.aar.right : NaN;
    }
    set badgeMarginRight(v) {
        this.ensureBadgeMargin();
        this.i.aar.right = +v;
        this.i.aar = this.i.aar;
    }
    get badgeMarginTop() {
        return this.i.aar ? this.i.aar.top : NaN;
    }
    set badgeMarginTop(v) {
        this.ensureBadgeMargin();
        this.i.aar.top = +v;
        this.i.aar = this.i.aar;
    }
    ensureBadgeMargin() {
        if (this.i.aar) {
            return;
        }
        this.i.aar = new Thickness(2);
    }
    /**
     * Gets or sets the BadgeShape for the data legend.
    */
    get badgeShape() {
        return this.i.t6;
    }
    set badgeShape(v) {
        this.i.t6 = ensureEnum(LegendItemBadgeShape_$type, v);
    }
    /**
     * Gets or sets the UnitsMode for the data legend.
    */
    get unitsDisplayMode() {
        return this.i.tq;
    }
    set unitsDisplayMode(v) {
        this.i.tq = ensureEnum(DataLegendUnitsMode_$type, v);
    }
    /**
     * Gets or sets the units text for the data legend.
    */
    get unitsText() {
        return this.i.xv;
    }
    set unitsText(v) {
        this.i.xv = v;
    }
    /**
     * Gets or sets the units text color.
    */
    get unitsTextColor() {
        return brushToString(this.i.aaj);
    }
    set unitsTextColor(v) {
        this.i.aaj = stringToBrush(v);
    }
    get unitsTextMarginBottom() {
        return this.i.aay ? this.i.aay.bottom : NaN;
    }
    set unitsTextMarginBottom(v) {
        this.ensureUnitsTextMargin();
        this.i.aay.bottom = +v;
        this.i.aay = this.i.aay;
    }
    get unitsTextMarginLeft() {
        return this.i.aay ? this.i.aay.left : NaN;
    }
    set unitsTextMarginLeft(v) {
        this.ensureUnitsTextMargin();
        this.i.aay.left = +v;
        this.i.aay = this.i.aay;
    }
    get unitsTextMarginRight() {
        return this.i.aay ? this.i.aay.right : NaN;
    }
    set unitsTextMarginRight(v) {
        this.ensureUnitsTextMargin();
        this.i.aay.right = +v;
        this.i.aay = this.i.aay;
    }
    get unitsTextMarginTop() {
        return this.i.aay ? this.i.aay.top : NaN;
    }
    set unitsTextMarginTop(v) {
        this.ensureUnitsTextMargin();
        this.i.aay.top = +v;
        this.i.aay = this.i.aay;
    }
    ensureUnitsTextMargin() {
        if (this.i.aay) {
            return;
        }
        this.i.aay = new Thickness(2);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get unitsTextStyle() {
        if (this.i.un == null) {
            return null;
        }
        return this.i.un.fontString;
    }
    set unitsTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.un = fi;
    }
    get titleTextMarginBottom() {
        return this.i.aax ? this.i.aax.bottom : NaN;
    }
    set titleTextMarginBottom(v) {
        this.ensureTitleTextMargin();
        this.i.aax.bottom = +v;
        this.i.aax = this.i.aax;
    }
    get titleTextMarginLeft() {
        return this.i.aax ? this.i.aax.left : NaN;
    }
    set titleTextMarginLeft(v) {
        this.ensureTitleTextMargin();
        this.i.aax.left = +v;
        this.i.aax = this.i.aax;
    }
    get titleTextMarginRight() {
        return this.i.aax ? this.i.aax.right : NaN;
    }
    set titleTextMarginRight(v) {
        this.ensureTitleTextMargin();
        this.i.aax.right = +v;
        this.i.aax = this.i.aax;
    }
    get titleTextMarginTop() {
        return this.i.aax ? this.i.aax.top : NaN;
    }
    set titleTextMarginTop(v) {
        this.ensureTitleTextMargin();
        this.i.aax.top = +v;
        this.i.aax = this.i.aax;
    }
    ensureTitleTextMargin() {
        if (this.i.aax) {
            return;
        }
        this.i.aax = new Thickness(2);
    }
    /**
     * Gets or sets the display text color.
    */
    get titleTextColor() {
        return brushToString(this.i.aai);
    }
    set titleTextColor(v) {
        this.i.aai = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the display text.
    */
    get titleTextStyle() {
        if (this.i.um == null) {
            return null;
        }
        return this.i.um.fontString;
    }
    set titleTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.um = fi;
    }
    /**
     * Gets or sets the mode for displaying labels before series values in the data legend, e.g. O: H: L: C: for financial series
    */
    get labelDisplayMode() {
        return this.i.to;
    }
    set labelDisplayMode(v) {
        this.i.to = ensureEnum(DataLegendLabelMode_$type, v);
    }
    /**
     * Gets or sets the units text color.
    */
    get labelTextColor() {
        return brushToString(this.i.aad);
    }
    set labelTextColor(v) {
        this.i.aad = stringToBrush(v);
    }
    get labelTextMarginBottom() {
        return this.i.aau ? this.i.aau.bottom : NaN;
    }
    set labelTextMarginBottom(v) {
        this.ensureLabelTextMargin();
        this.i.aau.bottom = +v;
        this.i.aau = this.i.aau;
    }
    get labelTextMarginLeft() {
        return this.i.aau ? this.i.aau.left : NaN;
    }
    set labelTextMarginLeft(v) {
        this.ensureLabelTextMargin();
        this.i.aau.left = +v;
        this.i.aau = this.i.aau;
    }
    get labelTextMarginRight() {
        return this.i.aau ? this.i.aau.right : NaN;
    }
    set labelTextMarginRight(v) {
        this.ensureLabelTextMargin();
        this.i.aau.right = +v;
        this.i.aau = this.i.aau;
    }
    get labelTextMarginTop() {
        return this.i.aau ? this.i.aau.top : NaN;
    }
    set labelTextMarginTop(v) {
        this.ensureLabelTextMargin();
        this.i.aau.top = +v;
        this.i.aau = this.i.aau;
    }
    ensureLabelTextMargin() {
        if (this.i.aau) {
            return;
        }
        this.i.aau = new Thickness(2);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get labelTextStyle() {
        if (this.i.uh == null) {
            return null;
        }
        return this.i.uh.fontString;
    }
    set labelTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.uh = fi;
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.targetAxis && this.targetAxis.name && this.targetAxis.name == name) {
            return this.targetAxis;
        }
        if (this.valueFormatSpecifiers != null && arrayFindByName(this.valueFormatSpecifiers, name)) {
            return arrayFindByName(this.valueFormatSpecifiers, name);
        }
        if (this.headerFormatSpecifiers != null && arrayFindByName(this.headerFormatSpecifiers, name)) {
            return arrayFindByName(this.headerFormatSpecifiers, name);
        }
        return null;
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.targetAxis && this.targetAxis._styling) {
            this.targetAxis._styling(container, component, this);
        }
        this._inStyling = false;
    }
    /**
     * Hides any tooltips presented by the layer, if any.
    
    */
    hideToolTips() {
        this.i.li();
    }
    /**
     * Hides any tooltips presented by the layer, if any.
    
    */
    hideToolTipsImmediate() {
        this.i.lj();
    }
}
