// igniteui-react-charts/esm2015/lib/igr-radial-base.js
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { IgrCategoryAngleAxis } from "./igr-category-angle-axis";
import { IgrNumericRadiusAxis } from "./igr-numeric-radius-axis";
import { IgrAssigningRadialStyleEventArgs } from "./igr-assigning-radial-style-event-args";
import { IgrAssigningRadialMarkerStyleEventArgs } from "./igr-assigning-radial-marker-style-event-args";
import { IgrMarkerSeries } from "./igr-marker-series";
import { ensureBool, toPoint } from "igniteui-react-core";
/**
 * Represents the base class for all IgxDataChartComponent radial series
*/
export class IgrRadialBase extends IgrMarkerSeries {
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
        this._angleAxisName = null;
        this._valueAxisName = null;
        this._assigningRadialStyle = null;
        this._assigningRadialStyle_wrapped = null;
        this._assigningRadialMarkerStyle = null;
        this._assigningRadialMarkerStyle_wrapped = null;
    }
    /**
     * Checks if this series is a radial series
    */
    get isRadial() {
        return this.i.ed;
    }
    /**
     * Gets or sets whether this Radial series should allow custom style overrides of its individual visuals.
    */
    get isCustomRadialStyleAllowed() {
        return this.i.sf;
    }
    set isCustomRadialStyleAllowed(v) {
        this.i.sf = ensureBool(v);
    }
    /**
     * Gets or sets whether this Radial series should allow custom style overrides of its individual marker visuals.
    */
    get isCustomRadialMarkerStyleAllowed() {
        return this.i.se;
    }
    set isCustomRadialMarkerStyleAllowed(v) {
        this.i.se = ensureBool(v);
    }
    /**
     * Gets the effective angle axis for the current series object.
    */
    get angleAxis() {
        const r = this.i.rt;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrCategoryAngleAxis._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set angleAxis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.rt = null : this.i.rt = v.i;
    }
    /**
    * Gets or sets the name to use to resolve angleAxis from markup.
    */
    get angleAxisName() {
        return this._angleAxisName;
    }
    set angleAxisName(v) {
        this._angleAxisName = v;
    }
    /**
     * Gets the effective value axis for the current series object.
    */
    get valueAxis() {
        const r = this.i.r0;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrNumericRadiusAxis._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set valueAxis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.r0 = null : this.i.r0 = v.i;
    }
    /**
    * Gets or sets the name to use to resolve valueAxis from markup.
    */
    get valueAxisName() {
        return this._valueAxisName;
    }
    set valueAxisName(v) {
        this._valueAxisName = v;
    }
    /**
     * Gets or sets whether to clip the series to the bounds.
     * Setting this property to true will decrease chart performance.
    */
    get clipSeriesToBounds() {
        return this.i.sd;
    }
    set clipSeriesToBounds(v) {
        this.i.sd = ensureBool(v);
    }
    bindAxes(axes) {
        super.bindAxes(axes);
        for (let i = 0; i < axes.length; i++) {
            if (this.angleAxisName && this.angleAxisName.length > 0 &&
                axes[i].name == this.angleAxisName) {
                this.angleAxis = axes[i];
            }
        }
        for (let i = 0; i < axes.length; i++) {
            if (this.valueAxisName && this.valueAxisName.length > 0 &&
                axes[i].name == this.valueAxisName) {
                this.valueAxis = axes[i];
            }
        }
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.angleAxis && this.angleAxis.name && this.angleAxis.name == name) {
            return this.angleAxis;
        }
        if (this.valueAxis && this.valueAxis.name && this.valueAxis.name == name) {
            return this.valueAxis;
        }
        return null;
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.angleAxis && this.angleAxis._styling) {
            this.angleAxis._styling(container, component, this);
        }
        if (this.valueAxis && this.valueAxis._styling) {
            this.valueAxis._styling(container, component, this);
        }
        this._inStyling = false;
    }
    /**
     * Determine if object can be used as ValueAxis
    
    * @param axis  * The object to check
    */
    canUseAsValueAxis(axis) {
        let iv = this.i.sc(axis);
        return (iv);
    }
    /**
     * Determine if object can be used as AngleAxis
    
    * @param axis  * The object to check
    */
    canUseAsAngleAxis(axis) {
        let iv = this.i.sb(axis);
        return (iv);
    }
    /**
     * Gets the item that is the best match for the specified world coordinates.
    
    * @param world  * The world coordinates to use.
    */
    getItem(world) {
        let iv = this.i.hy(toPoint(world));
        return (iv);
    }
    /**
     * Gets the angle to the provided world position from the center of the radial series.
    
    * @param world  * The world position to get the angle for
    */
    getAngleFromWorld(world) {
        let iv = this.i.sk(toPoint(world));
        return (iv);
    }
    /**
     * Gets the precise item index, if possible, based on the closeness to the previous or next whole integer. If the series cannot provide this information, GetExactItemIndex will return the same integer value as GetItemIndex.
    
    * @param world  * The world position for which to return the index.
    */
    getExactItemIndex(world) {
        let iv = this.i.f9(toPoint(world));
        return (iv);
    }
    /**
     * Get the index of the item near the provided world coordinates.
    
    * @param world  * The world coordinates for which to getch the item index.
    */
    getItemIndex(world) {
        let iv = this.i.hi(toPoint(world));
        return (iv);
    }
    /**
     * Scrolls the specified item into the view.
    
    * @param item  * The item to scroll into view.
    */
    scrollIntoView(item) {
        let iv = this.i.eq(item);
        return (iv);
    }
    /**
     * Returns the offset value for this series if grouped on a category axis.
    
    */
    getOffsetValue() {
        let iv = this.i.getOffsetValue();
        return (iv);
    }
    /**
     * Returns the width of the category grouping this series is in.
    
    */
    getCategoryWidth() {
        let iv = this.i.getCategoryWidth();
        return (iv);
    }
    /**
     * Event raised when Assigning Category Style
    */
    get assigningRadialStyle() {
        return this._assigningRadialStyle;
    }
    set assigningRadialStyle(ev) {
        if (this._assigningRadialStyle_wrapped !== null) {
            this.i.assigningRadialStyle = delegateRemove(this.i.assigningRadialStyle, this._assigningRadialStyle_wrapped);
            this._assigningRadialStyle_wrapped = null;
            this._assigningRadialStyle = null;
        }
        this._assigningRadialStyle = ev;
        this._assigningRadialStyle_wrapped = (o, e) => {
            let outerArgs = new IgrAssigningRadialStyleEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeAssigningRadialStyle) {
                this.beforeAssigningRadialStyle(this, outerArgs);
            }
            if (this._assigningRadialStyle) {
                this._assigningRadialStyle(this, outerArgs);
            }
        };
        this.i.assigningRadialStyle = delegateCombine(this.i.assigningRadialStyle, this._assigningRadialStyle_wrapped);
        ;
    }
    /**
     * Event raised when Assigning Radial Marker Style
    */
    get assigningRadialMarkerStyle() {
        return this._assigningRadialMarkerStyle;
    }
    set assigningRadialMarkerStyle(ev) {
        if (this._assigningRadialMarkerStyle_wrapped !== null) {
            this.i.assigningRadialMarkerStyle = delegateRemove(this.i.assigningRadialMarkerStyle, this._assigningRadialMarkerStyle_wrapped);
            this._assigningRadialMarkerStyle_wrapped = null;
            this._assigningRadialMarkerStyle = null;
        }
        this._assigningRadialMarkerStyle = ev;
        this._assigningRadialMarkerStyle_wrapped = (o, e) => {
            let outerArgs = new IgrAssigningRadialMarkerStyleEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeAssigningRadialMarkerStyle) {
                this.beforeAssigningRadialMarkerStyle(this, outerArgs);
            }
            if (this._assigningRadialMarkerStyle) {
                this._assigningRadialMarkerStyle(this, outerArgs);
            }
        };
        this.i.assigningRadialMarkerStyle = delegateCombine(this.i.assigningRadialMarkerStyle, this._assigningRadialMarkerStyle_wrapped);
        ;
    }
}
