// igniteui-react-charts/esm2015/lib/MovingAverageConvergenceDivergenceIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { MovingAverageConvergenceDivergenceIndicatorStrategy } from "./MovingAverageConvergenceDivergenceIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
export let MovingAverageConvergenceDivergenceIndicator = /*@__PURE__*/ (() => {
    class MovingAverageConvergenceDivergenceIndicator extends StrategyBasedIndicator {
        get_tw() {
            return 1;
        }
        constructor() {
            super();
            this.trendLineBrush = ((() => {
                let $ret = new Brush();
                $ret.fill = "#565656";
                return $ret;
            })());
            this.trendLineType = 10;
            this.trendLinePeriod = 9;
        }
        get_up() {
            return new MovingAverageConvergenceDivergenceIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return MovingAverageConvergenceDivergenceIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get shortPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.vd));
        }
        set shortPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.vd, a);
        }
        ux() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.vc));
        }
        set longPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.vc, a);
        }
        uu() {
            return this.longPeriod;
        }
        get signalPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.ve));
        }
        set signalPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.ve, a);
        }
        t5() {
            return this.signalPeriod;
        }
    }
    MovingAverageConvergenceDivergenceIndicator.$t = /*@__PURE__*/ markType(MovingAverageConvergenceDivergenceIndicator, 'MovingAverageConvergenceDivergenceIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    MovingAverageConvergenceDivergenceIndicator.vd = /*@__PURE__*/ StrategyBasedIndicator.u9(10, MovingAverageConvergenceDivergenceIndicator.$);
    MovingAverageConvergenceDivergenceIndicator.vc = /*@__PURE__*/ StrategyBasedIndicator.u5(30, MovingAverageConvergenceDivergenceIndicator.$);
    MovingAverageConvergenceDivergenceIndicator.ve = /*@__PURE__*/ StrategyBasedIndicator.u8(9, MovingAverageConvergenceDivergenceIndicator.$, "SignalPeriod");
    return MovingAverageConvergenceDivergenceIndicator;
})();
