// igniteui-react-charts/esm2015/lib/OverlayProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType, getInstanceType } from "igniteui-react-core";
import { IOverlayProxy_$type } from "./IOverlayProxy";
/**
 * @hidden
 */
export let OverlayProxy = /*@__PURE__*/ (() => {
    class OverlayProxy extends Base {
        matchesType(a) {
            if (a.du) {
                return true;
            }
            return false;
        }
        setCloseMemberPath(a, b) {
            a.so = b;
        }
        setHighMemberPath(a, b) {
            a.sr = b;
        }
        setLowMemberPath(a, b) {
            a.sw = b;
        }
        setMultiplier(a, b) {
            if (getInstanceType(a).typeName == "BollingerBandsOverlay") {
                let s_ = a;
                let p_ = b;
                s_['multiplier'] = p_;
            }
        }
        setOpenMemberPath(a, b) {
            a.s0 = b;
        }
        setVolumeMemberPath(a, b) {
            a.s4 = b;
        }
        setXAxis(a, b) {
            a.qk = b;
        }
        setYAxis(a, b) {
            a.q6 = b;
        }
    }
    OverlayProxy.$t = /*@__PURE__*/ markType(OverlayProxy, 'OverlayProxy', Base.$, [IOverlayProxy_$type]);
    return OverlayProxy;
})();
