// igniteui-react-charts/esm2015/lib/PercentageVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentageVolumeOscillatorIndicatorStrategy } from "./PercentageVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PercentageVolumeOscillatorIndicator = /*@__PURE__*/ (() => {
    class PercentageVolumeOscillatorIndicator extends StrategyBasedIndicator {
        get_up() {
            return new PercentageVolumeOscillatorIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return PercentageVolumeOscillatorIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get shortPeriod() {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.vb));
        }
        set shortPeriod(a) {
            this.h(PercentageVolumeOscillatorIndicator.vb, a);
        }
        ux() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.va));
        }
        set longPeriod(a) {
            this.h(PercentageVolumeOscillatorIndicator.va, a);
        }
        uu() {
            return this.longPeriod;
        }
    }
    PercentageVolumeOscillatorIndicator.$t = /*@__PURE__*/ markType(PercentageVolumeOscillatorIndicator, 'PercentageVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentageVolumeOscillatorIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u9(10, PercentageVolumeOscillatorIndicator.$);
    PercentageVolumeOscillatorIndicator.va = /*@__PURE__*/ StrategyBasedIndicator.u5(30, PercentageVolumeOscillatorIndicator.$);
    return PercentageVolumeOscillatorIndicator;
})();
