// igniteui-react-charts/esm2015/lib/PercentagePriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentagePriceOscillatorIndicatorStrategy } from "./PercentagePriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PercentagePriceOscillatorIndicator = /*@__PURE__*/ (() => {
    class PercentagePriceOscillatorIndicator extends StrategyBasedIndicator {
        get_up() {
            return new PercentagePriceOscillatorIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return PercentagePriceOscillatorIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get shortPeriod() {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.vb));
        }
        set shortPeriod(a) {
            this.h(PercentagePriceOscillatorIndicator.vb, a);
        }
        ux() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.va));
        }
        set longPeriod(a) {
            this.h(PercentagePriceOscillatorIndicator.va, a);
        }
        uu() {
            return this.longPeriod;
        }
    }
    PercentagePriceOscillatorIndicator.$t = /*@__PURE__*/ markType(PercentagePriceOscillatorIndicator, 'PercentagePriceOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentagePriceOscillatorIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u9(10, PercentagePriceOscillatorIndicator.$);
    PercentagePriceOscillatorIndicator.va = /*@__PURE__*/ StrategyBasedIndicator.u5(30, PercentagePriceOscillatorIndicator.$);
    return PercentagePriceOscillatorIndicator;
})();
