// igniteui-react-charts/esm2015/lib/PolarScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PolarBase } from "./PolarBase";
import { markType } from "igniteui-react-core";
import { PolarBaseView } from "./PolarBaseView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let PolarScatterSeries = /*@__PURE__*/ (() => {
    class PolarScatterSeries extends PolarBase {
        get_c5() {
            return true;
        }
        by() {
            return new PolarScatterSeriesView(this);
        }
        l4(a) {
            super.l4(a);
            this.va = a;
        }
        g0() {
            return 0.7;
        }
        qf() {
            if (this.qe == 0) {
                return 2;
            }
            return super.qf();
        }
        constructor() {
            super();
            this.ab = PolarScatterSeries.$;
        }
    }
    PolarScatterSeries.$t = /*@__PURE__*/ markType(PolarScatterSeries, 'PolarScatterSeries', PolarBase.$);
    return PolarScatterSeries;
})();
/**
 * @hidden
 */
export let PolarScatterSeriesView = /*@__PURE__*/ (() => {
    class PolarScatterSeriesView extends PolarBaseView {
        constructor(a) {
            super(a);
            this.c3 = null;
            this.c3 = a;
        }
        as() {
            super.as();
            if (!this.q) {
                this.c3.ql = Defaults.e;
            }
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.ph = a;
            this.e.g1 = 3;
            this.e.g2 = 2;
            this.e.g3 = 2;
            this.e.fb = false;
        }
    }
    PolarScatterSeriesView.$t = /*@__PURE__*/ markType(PolarScatterSeriesView, 'PolarScatterSeriesView', PolarBaseView.$);
    return PolarScatterSeriesView;
})();
