// igniteui-react-charts/esm2015/lib/RangeCategoryBucketCalculator_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategoryBucketCalculator } from "./CategoryBucketCalculator";
import { markType, typeCast, markDep, Point_$type } from "igniteui-react-core";
import { isNaN_, truncate, isInfinity } from "igniteui-react-core";
import { CategorySeries } from "./CategorySeries";
import { IIsCategoryBased_$type } from "./IIsCategoryBased";
import { IHasHighLowValueCategory_$type } from "./IHasHighLowValueCategory";
import { RangeCategoryFramePreparer } from "./RangeCategoryFramePreparer";
import { DependencyProperty } from "igniteui-react-core";
import { Series } from "./Series";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { ICategoryScaler_$type } from "./ICategoryScaler";
import { IScaler_$type } from "./IScaler";
import { CategorySeriesView } from "./CategorySeriesView";
import { CategoryFrame } from "./CategoryFrame";
import { CategoryMarkerManager } from "./CategoryMarkerManager";
import { AxisRange } from "./AxisRange";
import { Rect } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { ISortingAxis_$type } from "./ISortingAxis";
import { IFastItemsSource_$type } from "igniteui-react-core";
import { CollisionAvoider } from "./CollisionAvoider";
import { ISupportsMarkers_$type } from "./ISupportsMarkers";
import { PropertyMetadata } from "igniteui-react-core";
import { FastFlattener } from "igniteui-react-core";
/**
 * @hidden
 */
export let RangeCategoryBucketCalculator = /*@__PURE__*/ (() => {
    class RangeCategoryBucketCalculator extends CategoryBucketCalculator {
        constructor(a) {
            super(a);
            this.j = null;
            this.l = 0;
            this.k = 0;
            this.i = null;
            this.h = null;
            this.j = a;
        }
        getBucketWithoutUnknowns(a) {
            let b = false;
            let c = this.i;
            let d = this.h;
            let e = this.l;
            let f = this.k;
            let g = Math.min(e, f);
            let h = a * this.d;
            let i = Math.min(h + this.d - 1, g - 1);
            let j = 1.7976931348623157E+308;
            let k = -1.7976931348623157E+308;
            let l = true;
            let m = 0;
            let n = 0;
            let o;
            let p;
            if (h <= i) {
                for (let q = h; q <= i; ++q) {
                    m = c[q];
                    n = d[q];
                    if (m < n) {
                        o = m;
                        p = n;
                    }
                    else {
                        p = m;
                        o = n;
                    }
                    if (!l) {
                        j = j < o ? j : o;
                        k = k > o ? k : o;
                        j = j < p ? j : p;
                        k = k > p ? k : p;
                    }
                    else {
                        j = Math.min(j, o);
                        k = Math.max(k, o);
                        k = Math.max(k, p);
                        j = Math.min(j, p);
                        l = false;
                    }
                }
            }
            if (b && n < m) {
                let r = j;
                j = k;
                k = r;
            }
            if (!l) {
                let s = new Array(3);
                s[0] = 0.5 * (h + i);
                s[1] = j;
                s[2] = k;
                return s;
            }
            let t = new Array(3);
            t[0] = NaN;
            t[1] = NaN;
            t[2] = NaN;
            return t;
        }
        getBucket(a) {
            let b = false;
            let c = this.i;
            let d = this.h;
            let e = this.l;
            let f = this.k;
            let g = Math.min(e, f);
            let h = a * this.d;
            let i = Math.min(h + this.d - 1, g - 1);
            let j = NaN;
            let k = NaN;
            let l = true;
            let m = 0;
            let n = 0;
            let o;
            let p;
            if (h <= i) {
                for (let q = h; q <= i; ++q) {
                    m = c[q];
                    n = d[q];
                    if (m < n) {
                        o = m;
                        p = n;
                    }
                    else {
                        p = m;
                        o = n;
                    }
                    if (!l) {
                        if (!isNaN_(o)) {
                            j = j < o ? j : o;
                            k = k > o ? k : o;
                        }
                        if (!isNaN_(p)) {
                            j = j < p ? j : p;
                            k = k > p ? k : p;
                        }
                    }
                    else {
                        if (!isNaN_(o)) {
                            if (isNaN_(j)) {
                                j = o;
                            }
                            else {
                                j = Math.min(j, o);
                            }
                            if (!isNaN_(k)) {
                                k = Math.max(k, o);
                            }
                        }
                        if (!isNaN_(p)) {
                            if (isNaN_(k)) {
                                k = p;
                            }
                            else {
                                k = Math.max(k, p);
                            }
                            if (!isNaN_(j)) {
                                j = Math.min(j, p);
                            }
                        }
                        if (!isNaN_(j) && !isNaN_(k)) {
                            l = false;
                        }
                    }
                }
            }
            if (n < m && b) {
                let r = j;
                j = k;
                k = r;
            }
            if (!l) {
                let s = new Array(3);
                s[0] = 0.5 * (h + i);
                s[1] = j;
                s[2] = k;
                return s;
            }
            let t = new Array(3);
            t[0] = NaN;
            t[1] = NaN;
            t[2] = NaN;
            return t;
        }
        cacheValues() {
            this.l = this.j.c0.lowColumn.count;
            this.k = this.j.c0.highColumn.count;
            this.i = this.j.c0.lowColumn.asArray();
            this.h = this.j.c0.highColumn.asArray();
        }
        unCacheValues() {
            this.i = null;
            this.h = null;
        }
    }
    RangeCategoryBucketCalculator.$t = /*@__PURE__*/ markType(RangeCategoryBucketCalculator, 'RangeCategoryBucketCalculator', CategoryBucketCalculator.$);
    return RangeCategoryBucketCalculator;
})();
/**
 * @hidden
 */
export let RangeCategorySeries = /*@__PURE__*/ (() => {
    class RangeCategorySeries extends CategorySeries {
        by() {
            return new RangeCategorySeriesView(this);
        }
        getHostReferenceValue() {
            return this.gg();
        }
        l4(a) {
            super.l4(a);
            this.to = a;
        }
        constructor() {
            super();
            this._RangeFramePreparer = null;
            this.tq = null;
            this.tp = null;
            this.tm = null;
            this.RangeFramePreparer = new RangeCategoryFramePreparer(1, this, this.to, this, this, this.to.cr);
        }
        get RangeFramePreparer() {
            return this._RangeFramePreparer;
        }
        set RangeFramePreparer(a) {
            this._RangeFramePreparer = a;
        }
        h1(a, b) {
            let c = this.jv(b);
            if (c == RangeCategorySeries.$$p[0]) {
                return this.h2(a, b, this.tu);
            }
            if (c == RangeCategorySeries.$$p[1]) {
                return this.h2(a, b, this.tx);
            }
            return super.h1(a, b);
        }
        get lowColumn() {
            return this.tq;
        }
        set lowColumn(a) {
            if (this.tq != a) {
                let b = this.tq;
                this.tq = a;
                this.raisePropertyChanged("LowColumn", b, this.tq);
            }
        }
        get highColumn() {
            return this.tp;
        }
        set highColumn(a) {
            if (this.tp != a) {
                let b = this.tp;
                this.tp = a;
                this.raisePropertyChanged("HighColumn", b, this.tp);
            }
        }
        r7() {
            return 11;
        }
        tf(a, b) {
            super.tf(a, b);
            b.b3 = true;
            CategoryMarkerManager.e(this, a.m, b.cu, this.qp);
            this.s9(b, a);
        }
        tz(a, b, c, d, e, f) {
            this.to.c1(a, b, c, d, e, f);
        }
        ar(a) {
            if (a.isEmpty && this.ce != null) {
                a = this.ce.ta;
            }
            let b = this.rw();
            if (this.lowColumn == null || this.highColumn == null) {
                return null;
            }
            let c = this.pz(this.b1);
            let d = new ScalerParams(0, a, this.ce.tp, b.cg, this.getEffectiveViewport());
            let e = (b.el(c.left, d));
            let f = (b.el(c.right, d));
            if (b.cg) {
                let g = e;
                e = f;
                f = g;
            }
            let h = null;
            if (b.cm) {
                h = b;
                e = h.i$f(e);
                f = h.i$f(f);
            }
            else {
                e = Math.floor(e);
                f = Math.ceil(f);
            }
            e = Math.max(0, Math.min(this.lowColumn.count - 1, e));
            f = Math.max(0, Math.min(this.lowColumn.count - 1, f));
            let i = 1.7976931348623157E+308;
            let j = -1.7976931348623157E+308;
            for (let k = truncate(e); k <= truncate(f); k++) {
                let l = k;
                if (h != null) {
                    l = h.i$b._inner[k];
                }
                let m = this.lowColumn.item(l);
                if (this.lowColumn.mayContainUnknowns) {
                    if (isInfinity(m) || isNaN_(m)) {
                        continue;
                    }
                }
                let n = this.highColumn.item(l);
                if (this.highColumn.mayContainUnknowns) {
                    if (isInfinity(n) || isNaN_(n)) {
                        continue;
                    }
                }
                i = Math.min(i, m);
                j = Math.max(j, n);
            }
            let o = new AxisRange(i, j);
            let p = this.sh();
            o = this.ap(o, p, a);
            return o;
        }
        eq(a) {
            let b = new Rect(0, 0, 0, 1, 1);
            let c = this.b1 != null ? this.b1.bv : Rect.empty;
            c = c.copy();
            let d = this.b1 != null ? this.b1.bu : Rect.empty;
            let e = !c.isEmpty && !d.isEmpty && this.cl != null ? this.cl.indexOf(a) : -1;
            let f = this.fetchXAxis();
            let g = this.fetchYAxis();
            let h = this.p1(this.b1);
            let i;
            if (f != null) {
                let j = new ScalerParams(0, b, b, f.cg, h);
                i = f.ei(e, j);
            }
            else {
                i = NaN;
            }
            let k = f != null ? this.RangeFramePreparer.r(typeCast(ICategoryScaler_$type, f), b, b, h) : 0;
            i += k;
            if (e >= 0 && Rect.l_op_Inequality(c, null) && Rect.l_op_Inequality(d, null)) {
                if (!isNaN_(i)) {
                    if (i < c.left + 0.1 * c.width) {
                        i = i + 0.4 * c.width;
                    }
                    if (i > c.right - 0.1 * c.width) {
                        i = i - 0.4 * c.width;
                    }
                    c.x = i - 0.5 * c.width;
                }
                if (g != null && this.highColumn != null && e < this.highColumn.count) {
                    let l = new ScalerParams(0, b, b, g.cg, h);
                    l.b = this.gg();
                    let m = g.ei(this.highColumn.item(e), l);
                    let n = g.ei(this.lowColumn.item(e), l);
                    if (!isNaN_(m) && !isNaN_(n)) {
                        let o = Math.abs(n - m);
                        if (c.height < o) {
                            c.height = o;
                            c.y = Math.min(n, m);
                        }
                        else {
                            if (n < c.top + 0.1 * c.height) {
                                n = n + 0.4 * c.height;
                            }
                            if (n > c.bottom - 0.1 * c.height) {
                                n = n - 0.4 * c.height;
                            }
                            c.y = n - 0.5 * c.height;
                        }
                    }
                }
                if (this.b9 != null) {
                    this.b9.au(this.ce, c, true);
                }
            }
            return e >= 0;
        }
        l9(a, b, c, d) {
            super.l9(a, b, c, d);
            switch (b) {
                case "FastItemsSource":
                    if (typeCast(IFastItemsSource_$type, c) != null) {
                        c.deregisterColumn(this.lowColumn);
                        c.deregisterColumn(this.highColumn);
                        this.lowColumn = null;
                        this.highColumn = null;
                    }
                    if (typeCast(IFastItemsSource_$type, d) != null) {
                        this.lowColumn = this.ci(this.tx);
                        this.highColumn = this.ci(this.tu);
                    }
                    if (!this.sr()) {
                        this.r6.cr.g(this.fx);
                        this.mn(false);
                    }
                    break;
                case RangeCategorySeries.$$p[1]:
                    if (this.cl != null) {
                        this.cl.deregisterColumn(this.lowColumn);
                        this.lowColumn = this.ci(this.tx);
                    }
                    break;
                case "LowColumn":
                    if (!this.sr()) {
                        this.r6.cr.g(this.fx);
                        this.mn(false);
                    }
                    break;
                case RangeCategorySeries.$$p[0]:
                    if (this.cl != null) {
                        this.cl.deregisterColumn(this.highColumn);
                        this.highColumn = this.ci(this.tu);
                    }
                    break;
                case "HighColumn":
                    if (!this.sr()) {
                        this.r6.cr.g(this.fx);
                        this.mn(false);
                    }
                    break;
            }
        }
        aq(a) {
            if (this.lowColumn == null || this.lowColumn.count == 0 || this.highColumn == null || this.highColumn.count == 0) {
                return null;
            }
            if (a == this.fetchXAxis()) {
                let b = Math.min(this.lowColumn.count, this.highColumn.count);
                return new AxisRange(0, b - 1);
            }
            if (a == this.fetchYAxis()) {
                let c = Math.min(this.lowColumn.minimum, this.highColumn.minimum);
                let d = Math.max(this.lowColumn.maximum, this.highColumn.maximum);
                return new AxisRange(Math.min(c, d), Math.max(c, d));
            }
            return null;
        }
        k5(a, b, c, d) {
            if (this.fetchXAxis() != null && typeCast(ISortingAxis_$type, this.fetchXAxis()) !== null) {
                this.fetchXAxis().i$i();
            }
            let e = this.fetchXAxis();
            switch (a) {
                case 3:
                    if (d == this.tx || d == this.tu) {
                        if (e != null) {
                            e.c0();
                        }
                        if (!this.sr()) {
                            this.mn(true);
                        }
                    }
                    break;
                case 1:
                    if (e != null) {
                        e.c0();
                    }
                    this.r6.cr.g(this.fx);
                    if (!this.sr()) {
                        this.mn(true);
                    }
                    break;
                case 0:
                    if (e != null) {
                        e.c0();
                    }
                    this.r6.cr.g(this.fx);
                    if (!this.sr()) {
                        this.mn(true);
                    }
                    break;
                case 2:
                    if (e != null) {
                        e.c0();
                    }
                    if (this.tx != null && this.tu != null && this.r6.cr.d > 0 && !this.sr()) {
                        this.mn(true);
                    }
                    break;
                case 4:
                    if (e != null) {
                        e.c0();
                    }
                    this.r6.cr.g(this.fx);
                    if (!this.sr()) {
                        this.mn(true);
                    }
                    break;
            }
        }
        fd(a, b, c) {
            let d = super.fd(a, b, c);
            if (this.lowColumn == null || this.lowColumn.count == 0 || this.highColumn == null || this.highColumn.count == 0) {
                d = false;
            }
            return d;
        }
        tc(a, b) {
            super.tc(a, b);
            this.RangeFramePreparer.d(a, b);
        }
        get currentCategoryMode() {
            return this.preferredCategoryMode(typeCast(CategoryAxisBase.$, this.fetchXAxis()));
        }
        get scaler() {
            return typeCast(ICategoryScaler_$type, this.fetchXAxis());
        }
        get yScaler() {
            return typeCast(IScaler_$type, this.fetchYAxis());
        }
        get bucketizer() {
            return this.r6.cr;
        }
        get currentMode2Index() {
            return this.s0();
        }
        provideCollisionDetector() {
            return new CollisionAvoider();
        }
        get mayContainUnknowns() {
            return this.lowColumn == null || this.lowColumn.mayContainUnknowns || this.highColumn == null || this.highColumn.mayContainUnknowns;
        }
        mp(a, b) {
            let c = this.e3;
            super.mp(a, b);
            if (!c) {
                this.b1.a2(b);
                return;
            }
            this.b1.a2(b);
            if (this.cy(this.b0)) {
                return;
            }
            let d = new RangeCategoryFramePreparer(1, this, typeCast(ISupportsMarkers_$type, this.b0), this.ce.cu.o, this, this.b0.cr);
            if (!this.ex) {
                this.r0 = new CategoryFrame(3);
                this.r0.u();
                d.d(this.r0, this.b0);
            }
            this.ex = false;
            this.tf(this.r0, this.b0);
            this.e3 = false;
        }
        mm(a, b, c, d, e) {
            super.mm(a, b, c, d, e);
            let f = this.a7.a$h.item(d);
            let g = f;
            g.cr.g(this.fx);
            f.a1(c);
            if (this.cy(g)) {
                return;
            }
            let h = new RangeCategoryFramePreparer(1, this, typeCast(ISupportsMarkers_$type, f), f, this, f.cr);
            if (this.tm == null) {
                this.tm = new CategoryFrame(3);
            }
            this.tm.u();
            h.d(this.tm, g);
            this.tf(this.tm, g);
        }
        gg() {
            return Series.gh(this.lowColumn, typeCast(ISortingAxis_$type, this.fetchXAxis()));
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    RangeCategorySeries.$t = /*@__PURE__*/ markType(RangeCategorySeries, 'RangeCategorySeries', CategorySeries.$, [IIsCategoryBased_$type, IHasHighLowValueCategory_$type]);
    RangeCategorySeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, RangeCategorySeries, 'raisePropertyChanged', ['HighMemberPath:tu:t0', [2, null], 'LowMemberPath:tx:t1', [2, null]]);
    return RangeCategorySeries;
})();
/**
 * @hidden
 */
export let RangeCategorySeriesView = /*@__PURE__*/ (() => {
    class RangeCategorySeriesView extends CategorySeriesView {
        constructor(a) {
            super(a);
            this.c0 = null;
            this.c0 = a;
        }
        c1(a, b, c, d, e, f) {
            a.an.clear();
            b.an.clear();
            c.an.clear();
            let g = FastFlattener.b(d, e, true, f, this.e.fx);
            let h = FastFlattener.b(d, e, false, f, this.e.fx);
            let i = g.count;
            let j = h.count;
            let k;
            let l;
            let m;
            let n;
            for (let o = 0; o < i; o++) {
                k = g.item(o);
                l = e._inner[k];
                m = l[0];
                n = l[1];
                a.an.add({ $type: Point_$type, x: m, y: n });
                b.an.add({ $type: Point_$type, x: m, y: n });
            }
            let p;
            let q;
            for (let r = 0; r < j; r++) {
                k = h.item(r);
                l = e._inner[k];
                if (f) {
                    p = l[0];
                    q = l[2];
                }
                else {
                    p = l[2];
                    q = l[3];
                }
                c.an.add({ $type: Point_$type, x: p, y: q });
                b.an.add({ $type: Point_$type, x: p, y: q });
            }
            a.ab = a.an.count > 0;
            b.ab = b.an.count > 0;
            c.ab = c.an.count > 0;
        }
        cs() {
            return new RangeCategoryBucketCalculator(this);
        }
    }
    RangeCategorySeriesView.$t = /*@__PURE__*/ markType(RangeCategorySeriesView, 'RangeCategorySeriesView', CategorySeriesView.$);
    return RangeCategorySeriesView;
})();
