// igniteui-react-charts/esm2015/lib/RelativeStrengthIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RelativeStrengthIndexIndicatorStrategy } from "./RelativeStrengthIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let RelativeStrengthIndexIndicator = /*@__PURE__*/ (() => {
    class RelativeStrengthIndexIndicator extends StrategyBasedIndicator {
        get_tw() {
            return 1;
        }
        constructor() {
            super();
        }
        get_up() {
            return new RelativeStrengthIndexIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return RelativeStrengthIndexIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get va() {
            return typeGetValue(this.c(RelativeStrengthIndexIndicator.vb));
        }
        set va(a) {
            this.h(RelativeStrengthIndexIndicator.vb, a);
        }
        uv() {
            return this.va;
        }
        u4(a) {
            this.va = a;
        }
    }
    RelativeStrengthIndexIndicator.$t = /*@__PURE__*/ markType(RelativeStrengthIndexIndicator, 'RelativeStrengthIndexIndicator', StrategyBasedIndicator.$);
    RelativeStrengthIndexIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u7(14, RelativeStrengthIndexIndicator.$);
    return RelativeStrengthIndexIndicator;
})();
