// igniteui-react-charts/esm2015/lib/ScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterBase } from "./ScatterBase";
import { markType } from "igniteui-react-core";
import { ScatterFrame } from "./ScatterFrame";
import { ScatterBaseView } from "./ScatterBaseView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let ScatterSeries = /*@__PURE__*/ (() => {
    class ScatterSeries extends ScatterBase {
        get_c5() {
            return true;
        }
        by() {
            return new ScatterSeriesView(this);
        }
        g0() {
            return 0.7;
        }
        qf() {
            if (this.qe == 0) {
                return 2;
            }
            return super.qf();
        }
        l4(a) {
            super.l4(a);
            this.vg = a;
        }
        constructor() {
            super();
            this.ab = ScatterSeries.$;
            this.sb = new ScatterFrame();
            this.sd = new ScatterFrame();
            this.r9 = new ScatterFrame();
        }
    }
    ScatterSeries.$t = /*@__PURE__*/ markType(ScatterSeries, 'ScatterSeries', ScatterBase.$);
    return ScatterSeries;
})();
/**
 * @hidden
 */
export let ScatterSeriesView = /*@__PURE__*/ (() => {
    class ScatterSeriesView extends ScatterBaseView {
        constructor(a) {
            super(a);
            this.df = null;
            this.df = a;
        }
        as() {
            super.as();
            if (!this.q) {
                this.b1.ql = Defaults.e;
            }
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.ph = a;
            this.e.g1 = 3;
            this.e.g2 = 2;
            this.e.g3 = 2;
            this.e.fb = false;
        }
    }
    ScatterSeriesView.$t = /*@__PURE__*/ markType(ScatterSeriesView, 'ScatterSeriesView', ScatterBaseView.$);
    return ScatterSeriesView;
})();
