// igniteui-react-charts/esm2015/lib/SplineSeriesBase_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeries } from "./HorizontalAnchoredCategorySeries";
import { DependencyProperty } from "igniteui-react-core";
import { SplineType_$type } from "./SplineType";
import { AxisRange } from "./AxisRange";
import { Array_$type, typeCast, Point_$type, enumGetBox, markType, markDep } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { ISortingAxis_$type } from "./ISortingAxis";
import { CollisionAvoider } from "./CollisionAvoider";
import { AnchoredCategorySeriesView } from "./AnchoredCategorySeriesView";
import { ScalerParams } from "./ScalerParams";
import { SafeSortedReadOnlyDoubleCollection } from "./SafeSortedReadOnlyDoubleCollection";
import { Numeric } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { truncate, isNaN_ } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let SplineSeriesBase = /*@__PURE__*/ (() => {
    class SplineSeriesBase extends HorizontalAnchoredCategorySeries {
        constructor() {
            super(...arguments);
            this.vl = null;
            this.vr = NaN;
            this.vq = NaN;
        }
        gz() {
            if (!isNaN_(this.gy)) {
                return super.gz();
            }
            return this.vm.db();
        }
        by() {
            return new SplineSeriesBaseView(this);
        }
        l4(a) {
            super.l4(a);
            this.vm = a;
        }
        mt(a) {
            super.mt(a);
            this.vo = true;
        }
        aq(a) {
            if (!this.vo || this.vl == null || this.vl.length == 0) {
                return super.aq(a);
            }
            if (this.valueColumn == null || this.valueColumn.count == 0) {
                return null;
            }
            if (a == this.fetchXAxis()) {
                return new AxisRange(0, this.valueColumn.count - 1);
            }
            this.vy(this.vm);
            if (this.vl == null || this.vl.length == 0) {
                return null;
            }
            if (a == this.fetchYAxis()) {
                if (isNaN_(this.vu) || isNaN_(this.vv)) {
                    return null;
                }
                return new AxisRange(this.vv, this.vu);
            }
            return null;
        }
        get_dg() {
            return true;
        }
        vj(a) {
            if (a.cr.d == 0) {
                return null;
            }
            let b = new List$1(Array_$type, 0);
            let c = a.bv;
            let d = a.bu;
            let e = this.getEffectiveViewport1(a);
            let f = 0;
            let g = 0;
            let h = 0;
            let i = this;
            let j = typeCast(CategoryAxisBase.$, i.fetchXAxis());
            let k = i.pz(this.b1);
            let l = i.cl;
            if (c.isEmpty || d.isEmpty || j == null || l == null || l.count == 0) {
                f = 0;
                return null;
            }
            let m = this.fx;
            let n = typeCast(ISortingAxis_$type, j);
            if (n == null || n.i$b == null) {
                let o = new ScalerParams(0, c, d, j.cg, e);
                let p = 0;
                let q = j.lf - 1;
                let r = Math.floor((q - p + 1) * m / k.width);
                f = truncate(Math.max(1, r));
                g = truncate(Math.max(0, Math.floor(p / f) - 1));
                h = truncate(Math.ceil(q / f));
            }
            else {
                g = 0;
                h = j.lf - 1;
                f = 1;
            }
            if (n != null && n.i$b.count != this.cl.count) {
                return null;
            }
            this.vy(a);
            let s = truncate(Math.ceil(d.width / (h - g)));
            let t = new CollisionAvoider();
            if (n != null) {
                return this.vk(a);
            }
            let u = (v) => v;
            let v = (w) => this.valueColumn.item(w);
            let w = a;
            w.c6();
            for (let x = g; x <= h; ++x) {
                let y = null;
                if (f == 1) {
                    if (x >= (this.valueColumn.count - 1)) {
                        break;
                    }
                    let z = u(x);
                    let aa = v(x);
                    let ab = u(x + 1);
                    let ac = v(x + 1);
                    let ad = ab - z;
                    let ae = this.vl[x];
                    let af = this.vl[x + 1];
                    b.add([aa, ac]);
                    for (let ag = 1; ag < s; ++ag) {
                        let ah = z + ad * ag / s;
                        let ai = (ab - ah) / ad;
                        let aj = (ah - z) / ad;
                        let ak = ai * aa + aj * ac + ((ai * ai * ai - ai) * ae + (aj * aj * aj - aj) * af) * (ad * ad) / 6;
                        b.add([ak, ak]);
                    }
                }
                else {
                    y = a.cr.getBucket(x);
                    if (!isNaN_(y[0])) {
                        b.add([y[1], y[2]]);
                    }
                }
            }
            w.c8();
            return b.toArray();
        }
        vk(a) {
            let b = typeCast(ISortingAxis_$type, this.u0);
            let c = this.getEffectiveViewport1(a);
            let d = a.bv;
            let e = a.bu;
            let f = new List$1(Array_$type, 0);
            if (b == null) {
                return null;
            }
            let g = this.u0;
            let h = 0;
            let i = 0;
            let j = 0;
            let k = this.fx;
            let l = this.pz(this.b1);
            if (b == null || b.i$b == null) {
                let m = new ScalerParams(0, d, e, g.cg, c);
                let n = 0;
                let o = g.lf - 1;
                let p = Math.floor((o - n + 1) * k / l.width);
                h = truncate(Math.max(1, p));
                i = truncate(Math.max(0, Math.floor(n / h) - 1));
                j = truncate(Math.ceil(o / h));
            }
            else {
                i = 0;
                j = g.lf - 1;
                h = 1;
            }
            let q = (r) => r;
            let r = (s) => this.valueColumn.item(b.i$b._inner[s]);
            let s = truncate(Math.ceil(e.width / (j - i)));
            let t = new CollisionAvoider();
            if (h <= 0 || (i <= 0 && j <= 0)) {
                return null;
            }
            for (let u = i; u < j + 1; ++u) {
                let v = null;
                let w = u * h;
                if (b != null && b.i$b != null && w >= 0 && w < b.i$b.count) {
                    w = b.i$b._inner[w];
                }
                if (u >= (this.valueColumn.count - 1)) {
                    break;
                }
                let x = q(u);
                let y = r(u);
                let z = q(u + 1);
                let aa = r(u + 1);
                let ab = z - x;
                let ac = this.vl[u];
                let ad = this.vl[u + 1];
                f.add([y, y]);
                for (let ae = 1; ae < s; ++ae) {
                    let af = ae / s;
                    let ag = x + ab * af;
                    let ah = (z - ag) / ab;
                    let ai = (ag - x) / ab;
                    let aj = ah * y + ai * aa + ((ah * ah * ah - ah) * ac + (ai * ai * ai - ai) * ad) * (ab * ab) / 6;
                    f.add([aj, aj]);
                }
            }
            return f.toArray();
        }
        vz(a, b, c, d, e, f) {
            let g = typeCast(ISortingAxis_$type, d);
            let h = this.getEffectiveViewport1(f);
            if (g == null) {
                return;
            }
            let i = new ScalerParams(0, b, c, d.cg, h);
            let j = new ScalerParams(0, b, c, e.cg, h);
            j.b = this.gg();
            let k = this.qn();
            let l = 0;
            let m = this.vt(d, b, c, h);
            let n = (o) => o;
            let o = (p) => this.valueColumn.item(g.i$b._inner[p]);
            let p = this.r6.cr.f;
            let q = this.r6.cr.e;
            let r = truncate(Math.ceil(c.width / (p - q)));
            let s = new CollisionAvoider();
            let t = this.r6.cr.d;
            if (t <= 0 || (q <= 0 && p <= 0)) {
                this.r6.cu.count = l;
                return;
            }
            let u = 0;
            let v = 0;
            for (let w = q; w < p + 1; ++w) {
                let x = null;
                let y = w * t;
                if (g != null && g.i$b != null && y >= 0 && y < g.i$b.count) {
                    y = g.i$b._inner[y];
                }
                if (w >= (this.valueColumn.count - 1)) {
                    if (k && this.vp(a, a.f._inner[a.f.count - 1], s, Math.min(y, this.cl.count - 1), l, f, u - 1)) {
                        ++l;
                    }
                    break;
                }
                let z = n(w);
                let aa = o(w);
                let ab = n(w + 1);
                let ac = o(w + 1);
                let ad = ab - z;
                let ae = this.vl[w];
                let af = this.vl[w + 1];
                let ag = g.i$d(g.i$b._inner[w]);
                let ah = d.ei(ag, i) + m;
                let ai = e.ei(aa, j);
                a.f.add([ah, ai, ai]);
                u++;
                v = u;
                for (let aj = 1; aj < r; ++aj) {
                    let ak = aj / r;
                    let al = z + ad * ak;
                    let am = (ab - al) / ad;
                    let an = (al - z) / ad;
                    let ao = am * aa + an * ac + ((am * am * am - am) * ae + (an * an * an - an) * af) * (ad * ad) / 6;
                    let ap = g.i$d(g.i$b._inner[w]);
                    let aq = g.i$d(g.i$b._inner[w + 1]);
                    let ar = ap + (aq - ap) * ak;
                    al = d.ei(ar, i) + m;
                    ao = e.ei(ao, j);
                    a.f.add([al, ao, ao]);
                    u++;
                }
                if (k) {
                    x = [ah, ai, ai];
                }
                if (k && this.vp(a, x, s, Math.min(y, this.cl.count - 1), l, f, v - 1)) {
                    ++l;
                }
            }
            f.cu.count = l;
        }
        vp(a, b, c, d, e, f, g) {
            let h = b[0];
            let i = b[1];
            let j = new Rect(0, h - 5, i - 5, 11, 11);
            if (!isNaN_(h) && !isNaN_(i) && c.tryAdd(j)) {
                a.m.add({ $type: Point_$type, x: h, y: i });
                let k = f.cu.item(e);
                k.content.item = this.cl.item(d);
                k.al = g;
                return true;
            }
            return false;
        }
        aj(a, b, c, d, e, f) {
            if (a == null || a.cm || this.sx(this.b1) > 1 || this.sx(this.b1) == 0) {
                return super.aj(this.rw(), this.rx.f, this.sx(this.b1), this.sy(this.b1), e, f);
            }
            if (this.rx.f.count == 0) {
                return null;
            }
            let g = this.vm.cr.f;
            let h = this.b1.bu;
            let i = truncate(Math.ceil(h.width));
            if (g != d) {
                i = truncate(Math.ceil(h.width / (g - d)));
            }
            let j = -1;
            let k = -1;
            let l = this.f9(e);
            let m = this.ho(e, f);
            let n = this.hm(e, f);
            let o = truncate(Math.floor(m / this.sx(this.b1)));
            let p = truncate(Math.floor(n / this.sx(this.b1)));
            j = o - this.sy(this.b1);
            k = p - this.sy(this.b1);
            let q = j * i;
            let r = k * i;
            let s = 0;
            if (m != n) {
                s = (l - m) / (n - m);
            }
            let t = truncate(Math.floor(q + s * (r - q)));
            j = t;
            k = t + 1;
            if (j < 0) {
                j = 0;
            }
            if (j > this.rx.f.count - 1) {
                j = this.rx.f.count - 1;
            }
            if (k < 0) {
                k = 0;
            }
            if (k > this.rx.f.count - 1) {
                k = this.rx.f.count - 1;
            }
            let u = this.rx.f._inner[j];
            let v = this.rx.f._inner[k];
            let w = new Array(2);
            w[0] = u;
            w[1] = v;
            return w;
        }
        tc(a, b) {
            super.tc(a, b);
            if (a.f.count <= 1) {
                return;
            }
            if (b.cr.d == 0) {
                return;
            }
            let c = b.bv;
            let d = b.bu;
            let e = this.getEffectiveViewport1(b);
            let f = this.u0;
            let g = this.u6;
            let h = new ScalerParams(0, c, d, f.cg, e);
            let i = ((() => {
                let $ret = new ScalerParams(0, c, d, g.cg, e);
                $ret.b = this.gg();
                return $ret;
            })());
            a.f.clear();
            a.m.clear();
            let j = this.qn();
            let k = 0;
            let l = typeCast(ISortingAxis_$type, this.u0);
            if (l != null && l.i$b.count != this.cl.count) {
                return;
            }
            let m = this.vt(f, c, d, e);
            let n = b.cr.d;
            this.vy(b);
            let o = b.cr.f;
            let p = b.cr.e;
            let q = truncate(Math.ceil(d.width / (o - p)));
            let r = new CollisionAvoider();
            if (l != null) {
                this.vz(a, c, d, f, g, b);
                return;
            }
            let s = (t) => t;
            let t = (u) => this.valueColumn.item(u);
            let u = b;
            u.c6();
            let v = 0;
            let w = 0;
            for (let x = p; x <= o; ++x) {
                let y = null;
                if (n == 1) {
                    if (x >= (this.valueColumn.count - 1)) {
                        if (j && this.vp(a, a.f._inner[a.f.count - 1], r, Math.min(x * n, this.cl.count - 1), k, b, v - 1)) {
                            ++k;
                        }
                        break;
                    }
                    let z = s(x);
                    let aa = t(x);
                    let ab = s(x + 1);
                    let ac = t(x + 1);
                    let ad = ab - z;
                    let ae = this.vl[x];
                    let af = this.vl[x + 1];
                    let ag = f.ei(z, h) + m;
                    let ah = g.ei(aa, i);
                    a.f.add([ag, ah, ah]);
                    v++;
                    w = v;
                    for (let ai = 1; ai < q; ++ai) {
                        let aj = z + ad * ai / q;
                        let ak = (ab - aj) / ad;
                        let al = (aj - z) / ad;
                        let am = ak * aa + al * ac + ((ak * ak * ak - ak) * ae + (al * al * al - al) * af) * (ad * ad) / 6;
                        aj = f.ei(aj, h) + m;
                        am = g.ei(am, i);
                        a.f.add([aj, am, am]);
                        v++;
                    }
                    if (j) {
                        y = b.cr.getBucket(x);
                        y[0] = (f.ei(y[0], h) + m);
                        y[1] = g.ei(y[1], i);
                        y[2] = g.ei(y[2], i);
                    }
                }
                else {
                    y = b.cr.getBucket(x);
                    if (!isNaN_(y[0])) {
                        y[0] = (f.ei(y[0], h) + m);
                        y[1] = g.ei(y[1], i);
                        y[2] = g.ei(y[2], i);
                        a.f.add(y);
                        v++;
                        w = v;
                    }
                }
                if (j && this.vp(a, y, r, Math.min(x * n, this.cl.count - 1), k, b, w - 1)) {
                    ++k;
                }
            }
            u.c8();
            b.cu.count = k;
        }
        vy(a) {
            let b = a.cr.d;
            let c = (d) => d;
            let d = (e) => this.valueColumn.item(e);
            if (b < 1 || this.valueColumn == null || this.valueColumn.count == 0) {
                return;
            }
            if ((this.vl == null || this.vl.length != this.valueColumn.count) && b == 1) {
                let e = this.vn == 0 ? NaN : 0;
                if (this.u0 != null && typeCast(ISortingAxis_$type, this.u0) !== null && this.u0.i$b != null) {
                    let f = new SafeSortedReadOnlyDoubleCollection(1, this.valueColumn, this.u0.i$b, NaN);
                    d = (g) => f.item(g);
                }
                this.vl = Numeric.c(this.valueColumn.count, c, d, e, e);
                if (this.vo) {
                    let g = -1.7976931348623157E+308;
                    let h = 1.7976931348623157E+308;
                    let i = this.vj(a);
                    if (i == null) {
                        return;
                    }
                    for (let j = 0; j < i.length; j++) {
                        if (isNaN_(i[j][0]) || isNaN_(i[j][1])) {
                            continue;
                        }
                        if (i[j][1] > g) {
                            g = i[j][1];
                        }
                        if (i[j][0] < h) {
                            h = i[j][0];
                        }
                    }
                    this.vv = h;
                    this.vu = g;
                }
            }
        }
        vs(a, b, c, d, e, f, g) {
            let h = d - b;
            let i = b + h * a;
            let j = (d - i) / h;
            let k = (i - b) / h;
            let l = j * c + k * e + ((j * j * j - j) * f + (k * k * k - k) * g) * (h * h) / 6;
            return l;
        }
        gd(a, b, c, d, e, f, g, h) {
            let i = NaN;
            if (c >= 0 && c < b.count) {
                i = b.item(c);
            }
            let j = NaN;
            if (d >= 0 && d < b.count) {
                j = b.item(d);
            }
            if (f == 0) {
                return j;
            }
            if (e == b.count - 1) {
                return i;
            }
            if (isNaN_(j) && g != 0 && a <= 0.5 && !h) {
                return i;
            }
            if (isNaN_(i) && g != 0 && a >= 0.5 && !h) {
                return j;
            }
            let k = e;
            let l = f;
            let m = i;
            let n = j;
            let o = NaN;
            let p = NaN;
            if (this.vl != null && e >= 0 && e < this.vl.length) {
                o = this.vl[e];
            }
            if (this.vl != null && f >= 0 && f < this.vl.length) {
                p = this.vl[f];
            }
            return this.vs(a, k, m, l, n, o, p);
        }
        vt(a, b, c, d) {
            let e = this.preferredCategoryMode(a);
            if (e == 0 && a.categoryMode != 0) {
                e = 1;
            }
            let f = 0;
            switch (e) {
                case 0:
                    f = 0;
                    break;
                case 1:
                    f = 0.5 * a.getCategorySize(b, c, d);
                    break;
                case 2:
                    f = a.getGroupCenter(this.index, b, c, d);
                    break;
            }
            if (a.cg) {
                f = -f;
            }
            return f;
        }
        get vv() {
            return this.vr;
        }
        set vv(a) {
            this.vr = a;
        }
        get vu() {
            return this.vq;
        }
        set vu(a) {
            this.vq = a;
        }
        v0() {
            this.vl = null;
            this.vv = NaN;
            this.vu = NaN;
            if (this.yAxis != null) {
                this.yAxis.c0();
            }
        }
        l9(a, b, c, d) {
            switch (b) {
                case "ValueColumn":
                case "FastItemsSource":
                    this.v0();
                    break;
            }
            super.l9(a, b, c, d);
            switch (b) {
                case SplineSeriesBase.$$p[1]:
                    this.v0();
                    this.mn(false);
                    this.lu();
                    break;
            }
        }
        k5(a, b, c, d) {
            this.v0();
            super.k5(a, b, c, d);
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    SplineSeriesBase.$t = /*@__PURE__*/ markType(SplineSeriesBase, 'SplineSeriesBase', HorizontalAnchoredCategorySeries.$);
    SplineSeriesBase.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, SplineSeriesBase, 'raisePropertyChanged', ['IsSplineShapePartOfRange:vo:v1', [0, false], 'SplineType:vn:v2', [SplineType_$type, /*@__PURE__*/ enumGetBox(SplineType_$type, 0)]]);
    return SplineSeriesBase;
})();
/**
 * @hidden
 */
export let SplineSeriesBaseView = /*@__PURE__*/ (() => {
    class SplineSeriesBaseView extends AnchoredCategorySeriesView {
        constructor(a) {
            super(a);
            this.da = null;
            this.da = a;
        }
        db() {
            let a = 1;
            if (this.da.ce != null) {
                a = this.da.ce.gd;
            }
            return Math.max(1 / (a * 3), 0.3);
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.ph = a;
            this.e.g1 = 3;
            this.e.g2 = 1;
            this.e.g3 = 4;
            this.e.fb = false;
        }
    }
    SplineSeriesBaseView.$t = /*@__PURE__*/ markType(SplineSeriesBaseView, 'SplineSeriesBaseView', AnchoredCategorySeriesView.$);
    return SplineSeriesBaseView;
})();
