// igniteui-react-charts/esm2015/lib/Stacked100ColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedColumnSeries } from "./StackedColumnSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { isNaN_ } from "igniteui-react-core";
import { StackedColumnSeriesView } from "./StackedColumnSeriesView";
import { Stacked100ColumnBucketCalculator } from "./Stacked100ColumnBucketCalculator";
/**
 * @hidden
 */
export let Stacked100ColumnSeries = /*@__PURE__*/ (() => {
    class Stacked100ColumnSeries extends StackedColumnSeries {
        get_tx() {
            return true;
        }
        gz() {
            if (!isNaN_(this.gy)) {
                return super.gz();
            }
            return Defaults.h;
        }
        constructor() {
            super();
            this.ab = Stacked100ColumnSeries.$;
        }
        ac() {
            super.ac();
            this.mn(false);
        }
        by() {
            return new Stacked100ColumnSeriesView(this);
        }
        l4(a) {
            super.l4(a);
            this.u1 = a;
        }
        tp() {
            return this.u1;
        }
        ud() {
            super.ud();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    Stacked100ColumnSeries.$t = /*@__PURE__*/ markType(Stacked100ColumnSeries, 'Stacked100ColumnSeries', StackedColumnSeries.$, [IStacked100Series_$type]);
    return Stacked100ColumnSeries;
})();
/**
 * @hidden
 */
export let Stacked100ColumnSeriesView = /*@__PURE__*/ (() => {
    class Stacked100ColumnSeriesView extends StackedColumnSeriesView {
        constructor(a) {
            super(a);
            this.c5 = null;
            this.c5 = a;
        }
        cs() {
            return new Stacked100ColumnBucketCalculator(this);
        }
        as() {
            super.as();
            if (!this.q) {
                this.c0.ql = Defaults.c;
                this.c0.g4 = Defaults.i;
            }
        }
    }
    Stacked100ColumnSeriesView.$t = /*@__PURE__*/ markType(Stacked100ColumnSeriesView, 'Stacked100ColumnSeriesView', StackedColumnSeriesView.$);
    return Stacked100ColumnSeriesView;
})();
