// igniteui-react-charts/esm2015/lib/StackedLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalStackedSeriesBase } from "./HorizontalStackedSeriesBase";
import { markType } from "igniteui-react-core";
import { AxisRange } from "./AxisRange";
import { StackedSeriesView } from "./StackedSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let StackedLineSeries = /*@__PURE__*/ (() => {
    class StackedLineSeries extends HorizontalStackedSeriesBase {
        get_d4() {
            return true;
        }
        constructor() {
            super();
            this.ab = StackedLineSeries.$;
        }
        ac() {
            super.ac();
            this.mn(false);
        }
        by() {
            return new StackedLineSeriesView(this);
        }
        l4(a) {
            super.l4(a);
            this.uu = a;
        }
        aq(a) {
            if (this.itemsSource == null) {
                return null;
            }
            if (a == this.xAxis) {
                return new AxisRange(0, this.cl.count - 1);
            }
            if (a == this.yAxis) {
                this.ud();
                return new AxisRange(this.t4, this.t3);
            }
            return null;
        }
        preferredCategoryMode(a) {
            return 0;
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    StackedLineSeries.$t = /*@__PURE__*/ markType(StackedLineSeries, 'StackedLineSeries', HorizontalStackedSeriesBase.$);
    return StackedLineSeries;
})();
/**
 * @hidden
 */
export let StackedLineSeriesView = /*@__PURE__*/ (() => {
    class StackedLineSeriesView extends StackedSeriesView {
        constructor(a) {
            super(a);
        }
        as() {
            super.as();
            if (!this.q) {
                this.c0.ql = Defaults.e;
            }
        }
    }
    StackedLineSeriesView.$t = /*@__PURE__*/ markType(StackedLineSeriesView, 'StackedLineSeriesView', StackedSeriesView.$);
    return StackedLineSeriesView;
})();
