// igniteui-react-charts/esm2015/lib/Stacked100SplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedSplineAreaSeries } from "./StackedSplineAreaSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedSplineAreaSeriesView } from "./StackedSplineAreaSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let Stacked100SplineAreaSeries = /*@__PURE__*/ (() => {
    class Stacked100SplineAreaSeries extends StackedSplineAreaSeries {
        by() {
            return new Stacked100SplineAreaSeriesView(this);
        }
        l4(a) {
            super.l4(a);
            this.uy = a;
        }
        get_tx() {
            return true;
        }
        constructor() {
            super();
            this.ab = Stacked100SplineAreaSeries.$;
        }
        ac() {
            super.ac();
            this.mn(false);
        }
        ud() {
            super.ud();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    Stacked100SplineAreaSeries.$t = /*@__PURE__*/ markType(Stacked100SplineAreaSeries, 'Stacked100SplineAreaSeries', StackedSplineAreaSeries.$, [IStacked100Series_$type]);
    return Stacked100SplineAreaSeries;
})();
/**
 * @hidden
 */
export let Stacked100SplineAreaSeriesView = /*@__PURE__*/ (() => {
    class Stacked100SplineAreaSeriesView extends StackedSplineAreaSeriesView {
        constructor(a) {
            super(a);
        }
        as() {
            super.as();
            if (!this.q) {
                this.c0.ql = Defaults.e;
            }
        }
    }
    Stacked100SplineAreaSeriesView.$t = /*@__PURE__*/ markType(Stacked100SplineAreaSeriesView, 'Stacked100SplineAreaSeriesView', StackedSplineAreaSeriesView.$);
    return Stacked100SplineAreaSeriesView;
})();
