// igniteui-react-charts/esm2015/lib/Stacked100AreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedAreaSeries } from "./StackedAreaSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedAreaSeriesView } from "./StackedAreaSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let Stacked100AreaSeries = /*@__PURE__*/ (() => {
    class Stacked100AreaSeries extends StackedAreaSeries {
        get_tx() {
            return true;
        }
        constructor() {
            super();
            this.ab = Stacked100AreaSeries.$;
        }
        ac() {
            super.ac();
            this.mn(false);
        }
        by() {
            return new Stacked100AreaSeriesView(this);
        }
        l4(a) {
            super.l4(a);
            this.uv = a;
        }
        ud() {
            super.ud();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    Stacked100AreaSeries.$t = /*@__PURE__*/ markType(Stacked100AreaSeries, 'Stacked100AreaSeries', StackedAreaSeries.$, [IStacked100Series_$type]);
    return Stacked100AreaSeries;
})();
/**
 * @hidden
 */
export let Stacked100AreaSeriesView = /*@__PURE__*/ (() => {
    class Stacked100AreaSeriesView extends StackedAreaSeriesView {
        constructor(a) {
            super(a);
        }
        as() {
            super.as();
            if (!this.q) {
                this.c0.ql = Defaults.e;
            }
        }
    }
    Stacked100AreaSeriesView.$t = /*@__PURE__*/ markType(Stacked100AreaSeriesView, 'Stacked100AreaSeriesView', StackedAreaSeriesView.$);
    return Stacked100AreaSeriesView;
})();
