// igniteui-react-charts/esm2015/lib/StandardDeviationIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StandardDeviationIndicatorStrategy } from "./StandardDeviationIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let StandardDeviationIndicator = /*@__PURE__*/ (() => {
    class StandardDeviationIndicator extends StrategyBasedIndicator {
        get_up() {
            return new StandardDeviationIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return StandardDeviationIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get va() {
            return typeGetValue(this.c(StandardDeviationIndicator.vb));
        }
        set va(a) {
            this.h(StandardDeviationIndicator.vb, a);
        }
        uv() {
            return this.va;
        }
        u4(a) {
            this.va = a;
        }
    }
    StandardDeviationIndicator.$t = /*@__PURE__*/ markType(StandardDeviationIndicator, 'StandardDeviationIndicator', StrategyBasedIndicator.$);
    StandardDeviationIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u7(20, StandardDeviationIndicator.$);
    return StandardDeviationIndicator;
})();
