// igniteui-react-charts/esm2015/lib/TRIXIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { TRIXIndicatorStrategy } from "./TRIXIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
export let TRIXIndicator = /*@__PURE__*/ (() => {
    class TRIXIndicator extends StrategyBasedIndicator {
        constructor() {
            super();
            this.tx = 0;
            this.trendLineBrush = ((() => {
                let $ret = new Brush();
                $ret.fill = "#565656";
                return $ret;
            })());
            this.trendLineType = 9;
            this.trendLinePeriod = 9;
        }
        get_up() {
            return new TRIXIndicatorStrategy();
        }
        get up() {
            return this.get_up();
        }
        get_u2() {
            return TRIXIndicator.$;
        }
        get u2() {
            return this.get_u2();
        }
        get va() {
            return typeGetValue(this.c(TRIXIndicator.vb));
        }
        set va(a) {
            this.h(TRIXIndicator.vb, a);
        }
        uv() {
            return this.va;
        }
        u4(a) {
            this.va = a;
        }
    }
    TRIXIndicator.$t = /*@__PURE__*/ markType(TRIXIndicator, 'TRIXIndicator', StrategyBasedIndicator.$);
    TRIXIndicator.vb = /*@__PURE__*/ StrategyBasedIndicator.u7(15, TRIXIndicator.$);
    return TRIXIndicator;
})();
