// igniteui-react-charts/esm2015/lib/WaterfallSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeries } from "./HorizontalAnchoredCategorySeries";
import { Brush } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { runOn, markType, markDep, fromEnum } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { Pool$1 } from "igniteui-react-core";
import { Rectangle } from "igniteui-react-core";
import { ArrayExtension } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { PropertyMetadata } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
import { AnchoredCategorySeriesView } from "./AnchoredCategorySeriesView";
/**
 * @hidden
 */
export let WaterfallSeries = /*@__PURE__*/ (() => {
    class WaterfallSeries extends HorizontalAnchoredCategorySeries {
        get_en() {
            return true;
        }
        get_d5() {
            return true;
        }
        gz() {
            if (!isNaN_(this.gy)) {
                return super.gz();
            }
            return Defaults.h;
        }
        get_dw() {
            return true;
        }
        by() {
            return new WaterfallSeriesView(this);
        }
        l4(a) {
            super.l4(a);
            this.vj = a;
        }
        constructor() {
            super();
            this.ab = WaterfallSeries.$;
        }
        preferredCategoryMode(a) {
            return 2;
        }
        get_d6() {
            return true;
        }
        ms(a, b) {
            super.ms(a, b);
            this.vu = a;
            this.vv = b;
        }
        pf() {
            return this.vu;
        }
        pg() {
            return this.vv;
        }
        b4() {
            if (this.b5 == 0) {
                return 2;
            }
            return super.b4();
        }
        get_c3() {
            return true;
        }
        l9(a, b, c, d) {
            super.l9(a, b, c, d);
            switch (b) {
                case WaterfallSeries.$$p[3]:
                case WaterfallSeries.$$p[2]:
                case WaterfallSeries.$$p[0]:
                case WaterfallSeries.$$p[1]:
                    this.mn(false);
                    break;
            }
        }
        ez(a, b) {
            if (this.e2(a, b)) {
                return true;
            }
            if (this.qo(a, b)) {
                return true;
            }
            return false;
        }
        p3(a) {
            if (this.vj == null || this.vj.db == null || this.vj.db.count == 0 || this.xAxis == null) {
                return Rect.empty;
            }
            let b = this.vj.db;
            let c = this.pn(a);
            let d = -1;
            let e = this.xAxis.cg;
            let f = this.p7.right * 2;
            if (e) {
                d = ArrayExtension.binarySearch$1(Rectangle.$, b, (g) => {
                    let h = this.vj.di(g);
                    return (f - c.x) < (f - (h.left + h.width / 2)) ? -1 : ((f - c.x) > (f - (h.left + h.width / 2)) ? 1 : 0);
                });
            }
            else {
                d = ArrayExtension.binarySearch$1(Rectangle.$, b, (g) => {
                    let h = this.vj.di(g);
                    return c.x < (h.left + h.width / 2) ? -1 : (c.x > (h.left + h.width / 2) ? 1 : 0);
                });
            }
            if (d < 0) {
                d = ~d;
                d--;
            }
            if (d < 0) {
                d = 0;
            }
            if (d > b.count - 1) {
                d = b.count - 1;
            }
            let g = d;
            let h = d;
            if (g < 0) {
                g--;
            }
            if (h < b.count - 1) {
                h++;
            }
            let i = this.vw(g);
            let j = this.vw(d);
            let k = this.vw(h);
            let l = i.left + i.width / 2;
            let m = j.left + j.width / 2;
            let n = k.left + k.width / 2;
            let o = Math.abs(l - c.x);
            let p = Math.abs(m - c.x);
            let q = Math.abs(n - c.x);
            let r = i.containsPoint(c);
            let s = j.containsPoint(c);
            let t = k.containsPoint(c);
            if (p <= o && p <= q) {
                if (!e && t) {
                    return k;
                }
                if (e && r) {
                    return i;
                }
                return j;
            }
            if (o <= p && o <= q) {
                if (!e && s) {
                    return j;
                }
                return i;
            }
            if (q <= p && q <= o) {
                if (e && s) {
                    return j;
                }
                return k;
            }
            return Rect.empty;
        }
        vw(a) {
            return this.vj.dh(a);
        }
        r7() {
            return 1;
        }
        k2(a, b) {
            super.k2(a, b);
            this.vj.db.count = 0;
        }
        get_t3() {
            return 1;
        }
        ge() {
            return this.u0.k3(this.b1.bv, this.b1.bu, this.getEffectiveViewport1(this.b1));
        }
        tf(a, b) {
            let c = b;
            super.tf(a, b);
            let d = b.bv;
            let e = b.bu;
            let f = this.getEffectiveViewport1(b);
            let g = this.u0.k3(d, e, f);
            let h = this.vk;
            let i = this.vl;
            let j;
            let k = this.t1(b);
            let l = NaN;
            let m = c.db;
            let n;
            let o = 0;
            this.r5.a2(this, this.sp(), runOn(this, this.rq), this.sx(b), this.sy(b));
            this.r5.r = this.vk;
            this.r5.s = this.vl;
            this.r5.k = this.vk;
            this.r5.l = this.vl;
            let p = false;
            let q = this.r5.b;
            if (q != null) {
                p = true;
            }
            let r = this.u0.cm;
            let s = this.valueColumn.count;
            let t = a.f;
            let u = this.u0;
            let v = new ScalerParams(0, d, e, this.u0.cg, f);
            let w = this.o5;
            let x = this.vu;
            let y = this.o8;
            let z = this.vv || y;
            if (a.f.count > 0) {
                j = a.f._inner[0][0] - 0.5 * g;
                let aa = a.f._inner[0][1];
                if (!isNaN_(aa)) {
                    if (aa > k) {
                        n = m.item(o);
                        o++;
                        c.dg(n, j, k);
                        n.width = g;
                        n.height = Math.abs(k - aa);
                        this.r5.ak = x;
                        this.r5.ag = x;
                        this.r5.al = z;
                        this.r5.ah = z;
                        this.r5.f = true;
                        if (p) {
                            this.tb(t, 0, s, u, v, b.q);
                        }
                        this.r5.ad(n, false, false, false, false);
                        n.ap = this.r5.k;
                        n.aq = this.r5.l;
                    }
                    else {
                        n = m.item(o);
                        o++;
                        c.dg(n, j, aa);
                        n.width = g;
                        n.height = Math.abs(aa - k);
                        this.r5.ak = w;
                        this.r5.ag = w;
                        this.r5.al = y;
                        this.r5.ah = y;
                        this.r5.f = false;
                        if (p) {
                            this.tb(t, 0, s, u, v, b.q);
                        }
                        this.r5.ad(n, false, false, false, false);
                        n.ap = this.r5.k;
                        n.aq = this.r5.l;
                    }
                    l = aa;
                }
                else {
                    l = k;
                }
            }
            for (let ab = 1; ab < a.f.count; ++ab) {
                let ac = a.f._inner[ab];
                j = a.f._inner[ab][0] - 0.5 * g;
                let ad = ac[1];
                if (!isNaN_(ad)) {
                    n = m.item(o);
                    o++;
                    c.dg(n, j, Math.min(l, ad));
                    n.width = g;
                    n.height = Math.abs(l - ad);
                    this.r5.ak = w;
                    this.r5.ag = w;
                    this.r5.al = y;
                    this.r5.ah = y;
                    this.r5.f = false;
                    if (l > ad) {
                        if (p) {
                            this.tb(t, ab, s, u, v, b.q);
                        }
                    }
                    else {
                        this.r5.ak = x;
                        this.r5.ag = x;
                        this.r5.al = z;
                        this.r5.ah = z;
                        this.r5.f = true;
                        if (p) {
                            this.tb(t, ab, s, u, v, b.q);
                        }
                    }
                    this.r5.ad(n, false, false, false, false);
                    n.ap = this.r5.k;
                    n.aq = this.r5.l;
                    l = ad;
                }
            }
            m.count = o;
            b.cy(a);
        }
        get useHighMarkerFidelity() {
            return this.ss;
        }
    }
    WaterfallSeries.$t = /*@__PURE__*/ markType(WaterfallSeries, 'WaterfallSeries', HorizontalAnchoredCategorySeries.$);
    WaterfallSeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, WaterfallSeries, 'raisePropertyChanged', ['NegativeBrush:vu:vq', [Brush.$, null], 'NegativeOutline:vv:vr', [Brush.$, null], 'RadiusX:vk:vs', [1, 0], 'RadiusY:vl:vt', [1, 0]]);
    return WaterfallSeries;
})();
/**
 * @hidden
 */
export let WaterfallSeriesView = /*@__PURE__*/ (() => {
    class WaterfallSeriesView extends AnchoredCategorySeriesView {
        as() {
            super.as();
            this.dc = new List$1(Rectangle.$, 0);
            if (!this.q) {
                this.da.ql = Defaults.c;
                this.da.g4 = Defaults.i;
                this.da.vu = Defaults.ah;
                this.da.vv = Defaults.ai;
            }
        }
        constructor(a) {
            super(a);
            this.da = null;
            this.dc = null;
            this.dj = new Rectangle();
            this.db = null;
            this.da = a;
            this.db = ((() => {
                let $ret = new Pool$1(Rectangle.$);
                $ret.create = runOn(this, this.dk);
                $ret.activate = runOn(this, this.dd);
                $ret.disactivate = runOn(this, this.df);
                $ret.destroy = runOn(this, this.de);
                return $ret;
            })());
        }
        dk() {
            let a = new Rectangle();
            this.dc.add(a);
            a._visibility = 1;
            return a;
        }
        dd(a) {
            a._visibility = 0;
        }
        df(a) {
            a._visibility = 1;
        }
        de(a) {
            this.dc.remove(a);
        }
        dg(a, b, c) {
            let d = false;
            if (a.o != c) {
                d = true;
                a.o = c;
            }
            if (a.n != b) {
                d = true;
                a.n = b;
            }
            if (d) {
                this.aq();
            }
        }
        ac(a) {
            return this.dc._inner[a];
        }
        ab(a) {
            let b = this.dc._inner[a];
            this.dj._visibility = b._visibility;
            this.dj.n = b.n;
            this.dj.o = b.o;
            this.dj.width = b.width;
            this.dj.height = b.height;
            let c = this.bp(a);
            this.dj._fill = c;
            this.dj._stroke = c;
            this.dj.ad = this.e.fy + 3;
            return this.dj;
        }
        a6(a, b) {
            super.a6(a, b);
            if (a.d) {
                for (let c = 0; c < this.dc.count; c++) {
                    let d = this.aa(c, b);
                    this.bd(d, c, b);
                    a.w(d);
                }
            }
        }
        al(a) {
            super.al(a);
            let b = 0;
            let c = new List$1(Rectangle.$, 0);
            for (let d of fromEnum(this.db.a)) {
                c.add(d);
            }
            c.aa((e, f) => {
                if (e.n < f.n) {
                    return -1;
                }
                else if (e.n > f.n) {
                    return 1;
                }
                else {
                    return 0;
                }
            });
            for (let e of fromEnum(c)) {
                let f = "Negative";
                if (e._fill == this.e.o5) {
                    f = "Positive";
                }
                this.e.a9.exportRectangleData(a, e, "column" + b, ["Main", f]);
            }
            b++;
        }
        dh(a) {
            let b = this.db.item(a);
            return new Rect(0, b.n, b.o, b.width, b.height);
        }
        di(a) {
            return new Rect(0, a.n, a.o, a.width, a.height);
        }
    }
    WaterfallSeriesView.$t = /*@__PURE__*/ markType(WaterfallSeriesView, 'WaterfallSeriesView', AnchoredCategorySeriesView.$);
    return WaterfallSeriesView;
})();
