// igniteui-react-core/esm2015/lib/AnchoredCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AnchoredCategorySeriesDescription = /*@__PURE__*/ (() => {
    class AnchoredCategorySeriesDescription extends CategorySeriesDescription {
        get_type() {
            return "AnchoredCategorySeries";
        }
        constructor() {
            super();
            this.eo = null;
            this.em = null;
            this.en = null;
            this.el = null;
            this.ek = null;
            this.eh = null;
            this.ed = 0;
            this.eb = null;
            this.ef = 0;
            this.ej = null;
            this.ei = null;
        }
        get valueMemberPath() {
            return this.eo;
        }
        set valueMemberPath(a) {
            this.eo = a;
            this.g("ValueMemberPath");
        }
        get valueMemberAsLegendLabel() {
            return this.em;
        }
        set valueMemberAsLegendLabel(a) {
            this.em = a;
            this.g("ValueMemberAsLegendLabel");
        }
        get valueMemberAsLegendUnit() {
            return this.en;
        }
        set valueMemberAsLegendUnit(a) {
            this.en = a;
            this.g("ValueMemberAsLegendUnit");
        }
        get trendLineType() {
            return this.el;
        }
        set trendLineType(a) {
            this.el = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.ek;
        }
        set trendLineBrush(a) {
            this.ek = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.eh;
        }
        set actualTrendLineBrush(a) {
            this.eh = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.ed;
        }
        set trendLineThickness(a) {
            this.ed = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.eb;
        }
        set trendLineDashArray(a) {
            this.eb = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.ef;
        }
        set trendLinePeriod(a) {
            this.ef = a;
            this.g("TrendLinePeriod");
        }
        get markerCollisionAvoidance() {
            return this.ej;
        }
        set markerCollisionAvoidance(a) {
            this.ej = a;
            this.g("MarkerCollisionAvoidance");
        }
        get consolidatedItemHitTestBehavior() {
            return this.ei;
        }
        set consolidatedItemHitTestBehavior(a) {
            this.ei = a;
            this.g("ConsolidatedItemHitTestBehavior");
        }
    }
    AnchoredCategorySeriesDescription.$t = /*@__PURE__*/ markType(AnchoredCategorySeriesDescription, 'AnchoredCategorySeriesDescription', CategorySeriesDescription.$);
    return AnchoredCategorySeriesDescription;
})();
