// igniteui-react-core/esm2015/lib/AssigningSeriesShapeStyleEventArgsBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AssigningSeriesStyleEventArgsBaseDescription } from "./AssigningSeriesStyleEventArgsBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AssigningSeriesShapeStyleEventArgsBaseDescription = /*@__PURE__*/ (() => {
    class AssigningSeriesShapeStyleEventArgsBaseDescription extends AssigningSeriesStyleEventArgsBaseDescription {
        get_type() {
            return "AssigningSeriesShapeStyleEventArgsBase";
        }
        constructor() {
            super();
            this.au = 0;
            this.aq = null;
            this.as = 0;
            this.at = 0;
        }
        get strokeThickness() {
            return this.au;
        }
        set strokeThickness(a) {
            this.au = a;
            this.g("StrokeThickness");
        }
        get strokeDashArray() {
            return this.aq;
        }
        set strokeDashArray(a) {
            this.aq = a;
            this.g("StrokeDashArray");
        }
        get radiusX() {
            return this.as;
        }
        set radiusX(a) {
            this.as = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.at;
        }
        set radiusY(a) {
            this.at = a;
            this.g("RadiusY");
        }
    }
    AssigningSeriesShapeStyleEventArgsBaseDescription.$t = /*@__PURE__*/ markType(AssigningSeriesShapeStyleEventArgsBaseDescription, 'AssigningSeriesShapeStyleEventArgsBaseDescription', AssigningSeriesStyleEventArgsBaseDescription.$);
    return AssigningSeriesShapeStyleEventArgsBaseDescription;
})();
