// igniteui-react-core/esm2015/lib/BubbleSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterBaseDescription } from "./ScatterBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let BubbleSeriesDescription = /*@__PURE__*/ (() => {
    class BubbleSeriesDescription extends ScatterBaseDescription {
        get_type() {
            return "BubbleSeries";
        }
        constructor() {
            super();
            this.ft = null;
            this.fb = null;
            this.fq = null;
            this.fp = null;
            this.e9 = null;
            this.fd = false;
            this.fe = false;
            this.fk = 0;
            this.fj = 0;
            this.ff = false;
            this.fr = null;
            this.fn = null;
            this.fs = null;
            this.fo = null;
        }
        get radiusMemberPath() {
            return this.ft;
        }
        set radiusMemberPath(a) {
            this.ft = a;
            this.g("RadiusMemberPath");
        }
        get radiusScale() {
            return this.fb;
        }
        set radiusScale(a) {
            this.fb = a;
            this.g("RadiusScale");
        }
        get labelMemberPath() {
            return this.fq;
        }
        set labelMemberPath(a) {
            this.fq = a;
            this.g("LabelMemberPath");
        }
        get fillMemberPath() {
            return this.fp;
        }
        set fillMemberPath(a) {
            this.fp = a;
            this.g("FillMemberPath");
        }
        get fillScale() {
            return this.e9;
        }
        set fillScale(a) {
            this.e9 = a;
            this.g("FillScale");
        }
        get fillScaleUseGlobalValues() {
            return this.fd;
        }
        set fillScaleUseGlobalValues(a) {
            this.fd = a;
            this.g("FillScaleUseGlobalValues");
        }
        get markerOutlineUsesFillScale() {
            return this.fe;
        }
        set markerOutlineUsesFillScale(a) {
            this.fe = a;
            this.g("MarkerOutlineUsesFillScale");
        }
        get markerOutlineBrightness() {
            return this.fk;
        }
        set markerOutlineBrightness(a) {
            this.fk = a;
            this.g("MarkerOutlineBrightness");
        }
        get markerBrushBrightness() {
            return this.fj;
        }
        set markerBrushBrightness(a) {
            this.fj = a;
            this.g("MarkerBrushBrightness");
        }
        get radiusScaleUseGlobalValues() {
            return this.ff;
        }
        set radiusScaleUseGlobalValues(a) {
            this.ff = a;
            this.g("RadiusScaleUseGlobalValues");
        }
        get radiusMemberAsLegendLabel() {
            return this.fr;
        }
        set radiusMemberAsLegendLabel(a) {
            this.fr = a;
            this.g("RadiusMemberAsLegendLabel");
        }
        get fillMemberAsLegendLabel() {
            return this.fn;
        }
        set fillMemberAsLegendLabel(a) {
            this.fn = a;
            this.g("FillMemberAsLegendLabel");
        }
        get radiusMemberAsLegendUnit() {
            return this.fs;
        }
        set radiusMemberAsLegendUnit(a) {
            this.fs = a;
            this.g("RadiusMemberAsLegendUnit");
        }
        get fillMemberAsLegendUnit() {
            return this.fo;
        }
        set fillMemberAsLegendUnit(a) {
            this.fo = a;
            this.g("FillMemberAsLegendUnit");
        }
    }
    BubbleSeriesDescription.$t = /*@__PURE__*/ markType(BubbleSeriesDescription, 'BubbleSeriesDescription', ScatterBaseDescription.$);
    return BubbleSeriesDescription;
})();
