// igniteui-react-core/esm2015/lib/BulletGraphDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let BulletGraphDescription = /*@__PURE__*/ (() => {
    class BulletGraphDescription extends Description {
        get fontFamily() {
            return this.cl;
        }
        set fontFamily(a) {
            this.cl = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.z;
        }
        set fontSize(a) {
            this.z = a;
            this.g("FontSize");
        }
        get fontWeight() {
            return this.cn;
        }
        set fontWeight(a) {
            this.cn = a;
            this.g("FontWeight");
        }
        get fontStyle() {
            return this.cm;
        }
        set fontStyle(a) {
            this.cm = a;
            this.g("FontStyle");
        }
        get width() {
            return this.c2;
        }
        set width(a) {
            this.c2 = a;
            this.g("Width");
        }
        get height() {
            return this.cp;
        }
        set height(a) {
            this.cp = a;
            this.g("Height");
        }
        get background() {
            return this.cg;
        }
        set background(a) {
            this.cg = a;
            this.g("Background");
        }
        get_type() {
            return "BulletGraph";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.cl = null;
            this.z = 0;
            this.cn = null;
            this.cm = null;
            this.c2 = null;
            this.cp = null;
            this.cg = null;
            this.a1 = 0;
            this.cs = null;
            this.ct = null;
            this.cu = null;
            this.ao = 0;
            this.k = null;
            this.i = null;
            this.j = null;
            this.ag = 0;
            this.u = 0;
            this.af = 0;
            this.t = 0;
            this.ar = 0;
            this.cw = null;
            this.a2 = 0;
            this.c0 = null;
            this.am = 0;
            this.an = 0;
            this.a3 = 0;
            this.a4 = 0;
            this.aa = 0;
            this.ax = 0;
            this.ay = 0;
            this.ac = 0;
            this.ab = 0;
            this.ad = 0;
            this.ae = 0;
            this.ah = 0;
            this.az = 0;
            this.aw = 0;
            this.a0 = 0;
            this.cy = null;
            this.ck = null;
            this.cz = null;
            this.c1 = null;
            this.a5 = 0;
            this.aj = 0;
            this.ai = 0;
            this.ak = 0;
            this.cr = null;
            this.p = false;
            this.ch = null;
            this.ci = null;
            this.y = 0;
            this.w = 0;
            this.x = 0;
            this.aq = 0;
            this.ap = 0;
            this.cv = null;
            this.as = 0;
            this.at = 0;
            this.au = 0;
            this.cx = null;
            this.av = 0;
            this.cq = null;
            this.h = null;
            this.ca = 0;
            this.b9 = 0;
            this.q = false;
            this.cj = null;
            this.al = 0;
            this.v = 0;
            this.co = null;
            this.cf = null;
            this.ce = null;
            this.cd = null;
        }
        get transitionProgress() {
            return this.a1;
        }
        set transitionProgress(a) {
            this.a1 = a;
            this.g("TransitionProgress");
        }
        get orientation() {
            return this.cs;
        }
        set orientation(a) {
            this.cs = a;
            this.g("Orientation");
        }
        get scaleBackgroundBrush() {
            return this.ct;
        }
        set scaleBackgroundBrush(a) {
            this.ct = a;
            this.g("ScaleBackgroundBrush");
        }
        get scaleBackgroundOutline() {
            return this.cu;
        }
        set scaleBackgroundOutline(a) {
            this.cu = a;
            this.g("ScaleBackgroundOutline");
        }
        get scaleBackgroundThickness() {
            return this.ao;
        }
        set scaleBackgroundThickness(a) {
            this.ao = a;
            this.g("ScaleBackgroundThickness");
        }
        get ranges() {
            return this.k;
        }
        set ranges(a) {
            this.k = a;
            this.g("Ranges");
        }
        get rangeBrushes() {
            return this.i;
        }
        set rangeBrushes(a) {
            this.i = a;
            this.g("RangeBrushes");
        }
        get rangeOutlines() {
            return this.j;
        }
        set rangeOutlines(a) {
            this.j = a;
            this.g("RangeOutlines");
        }
        get minimumValue() {
            return this.ag;
        }
        set minimumValue(a) {
            this.ag = a;
            this.g("MinimumValue");
        }
        get actualMinimumValue() {
            return this.u;
        }
        set actualMinimumValue(a) {
            this.u = a;
            this.g("ActualMinimumValue");
        }
        get maximumValue() {
            return this.af;
        }
        set maximumValue(a) {
            this.af = a;
            this.g("MaximumValue");
        }
        get actualMaximumValue() {
            return this.t;
        }
        set actualMaximumValue(a) {
            this.t = a;
            this.g("ActualMaximumValue");
        }
        get targetValue() {
            return this.ar;
        }
        set targetValue(a) {
            this.ar = a;
            this.g("TargetValue");
        }
        get targetValueName() {
            return this.cw;
        }
        set targetValueName(a) {
            this.cw = a;
            this.g("TargetValueName");
        }
        get value() {
            return this.a2;
        }
        set value(a) {
            this.a2 = a;
            this.g("Value");
        }
        get valueName() {
            return this.c0;
        }
        set valueName(a) {
            this.c0 = a;
            this.g("ValueName");
        }
        get rangeInnerExtent() {
            return this.am;
        }
        set rangeInnerExtent(a) {
            this.am = a;
            this.g("RangeInnerExtent");
        }
        get rangeOuterExtent() {
            return this.an;
        }
        set rangeOuterExtent(a) {
            this.an = a;
            this.g("RangeOuterExtent");
        }
        get valueInnerExtent() {
            return this.a3;
        }
        set valueInnerExtent(a) {
            this.a3 = a;
            this.g("ValueInnerExtent");
        }
        get valueOuterExtent() {
            return this.a4;
        }
        set valueOuterExtent(a) {
            this.a4 = a;
            this.g("ValueOuterExtent");
        }
        get interval() {
            return this.aa;
        }
        set interval(a) {
            this.aa = a;
            this.g("Interval");
        }
        get ticksPostInitial() {
            return this.ax;
        }
        set ticksPostInitial(a) {
            this.ax = a;
            this.g("TicksPostInitial");
        }
        get ticksPreTerminal() {
            return this.ay;
        }
        set ticksPreTerminal(a) {
            this.ay = a;
            this.g("TicksPreTerminal");
        }
        get labelInterval() {
            return this.ac;
        }
        set labelInterval(a) {
            this.ac = a;
            this.g("LabelInterval");
        }
        get labelExtent() {
            return this.ab;
        }
        set labelExtent(a) {
            this.ab = a;
            this.g("LabelExtent");
        }
        get labelsPostInitial() {
            return this.ad;
        }
        set labelsPostInitial(a) {
            this.ad = a;
            this.g("LabelsPostInitial");
        }
        get labelsPreTerminal() {
            return this.ae;
        }
        set labelsPreTerminal(a) {
            this.ae = a;
            this.g("LabelsPreTerminal");
        }
        get minorTickCount() {
            return this.ah;
        }
        set minorTickCount(a) {
            this.ah = a;
            this.g("MinorTickCount");
        }
        get tickStartExtent() {
            return this.az;
        }
        set tickStartExtent(a) {
            this.az = a;
            this.g("TickStartExtent");
        }
        get tickEndExtent() {
            return this.aw;
        }
        set tickEndExtent(a) {
            this.aw = a;
            this.g("TickEndExtent");
        }
        get tickStrokeThickness() {
            return this.a0;
        }
        set tickStrokeThickness(a) {
            this.a0 = a;
            this.g("TickStrokeThickness");
        }
        get tickBrush() {
            return this.cy;
        }
        set tickBrush(a) {
            this.cy = a;
            this.g("TickBrush");
        }
        get fontBrush() {
            return this.ck;
        }
        set fontBrush(a) {
            this.ck = a;
            this.g("FontBrush");
        }
        get valueBrush() {
            return this.cz;
        }
        set valueBrush(a) {
            this.cz = a;
            this.g("ValueBrush");
        }
        get valueOutline() {
            return this.c1;
        }
        set valueOutline(a) {
            this.c1 = a;
            this.g("ValueOutline");
        }
        get valueStrokeThickness() {
            return this.a5;
        }
        set valueStrokeThickness(a) {
            this.a5 = a;
            this.g("ValueStrokeThickness");
        }
        get minorTickStartExtent() {
            return this.aj;
        }
        set minorTickStartExtent(a) {
            this.aj = a;
            this.g("MinorTickStartExtent");
        }
        get minorTickEndExtent() {
            return this.ai;
        }
        set minorTickEndExtent(a) {
            this.ai = a;
            this.g("MinorTickEndExtent");
        }
        get minorTickStrokeThickness() {
            return this.ak;
        }
        set minorTickStrokeThickness(a) {
            this.ak = a;
            this.g("MinorTickStrokeThickness");
        }
        get minorTickBrush() {
            return this.cr;
        }
        set minorTickBrush(a) {
            this.cr = a;
            this.g("MinorTickBrush");
        }
        get isScaleInverted() {
            return this.p;
        }
        set isScaleInverted(a) {
            this.p = a;
            this.g("IsScaleInverted");
        }
        get backingBrush() {
            return this.ch;
        }
        set backingBrush(a) {
            this.ch = a;
            this.g("BackingBrush");
        }
        get backingOutline() {
            return this.ci;
        }
        set backingOutline(a) {
            this.ci = a;
            this.g("BackingOutline");
        }
        get backingStrokeThickness() {
            return this.y;
        }
        set backingStrokeThickness(a) {
            this.y = a;
            this.g("BackingStrokeThickness");
        }
        get backingInnerExtent() {
            return this.w;
        }
        set backingInnerExtent(a) {
            this.w = a;
            this.g("BackingInnerExtent");
        }
        get backingOuterExtent() {
            return this.x;
        }
        set backingOuterExtent(a) {
            this.x = a;
            this.g("BackingOuterExtent");
        }
        get scaleStartExtent() {
            return this.aq;
        }
        set scaleStartExtent(a) {
            this.aq = a;
            this.g("ScaleStartExtent");
        }
        get scaleEndExtent() {
            return this.ap;
        }
        set scaleEndExtent(a) {
            this.ap = a;
            this.g("ScaleEndExtent");
        }
        get targetValueBrush() {
            return this.cv;
        }
        set targetValueBrush(a) {
            this.cv = a;
            this.g("TargetValueBrush");
        }
        get targetValueBreadth() {
            return this.as;
        }
        set targetValueBreadth(a) {
            this.as = a;
            this.g("TargetValueBreadth");
        }
        get targetValueInnerExtent() {
            return this.at;
        }
        set targetValueInnerExtent(a) {
            this.at = a;
            this.g("TargetValueInnerExtent");
        }
        get targetValueOuterExtent() {
            return this.au;
        }
        set targetValueOuterExtent(a) {
            this.au = a;
            this.g("TargetValueOuterExtent");
        }
        get targetValueOutline() {
            return this.cx;
        }
        set targetValueOutline(a) {
            this.cx = a;
            this.g("TargetValueOutline");
        }
        get targetValueStrokeThickness() {
            return this.av;
        }
        set targetValueStrokeThickness(a) {
            this.av = a;
            this.g("TargetValueStrokeThickness");
        }
        get labelFormat() {
            return this.cq;
        }
        set labelFormat(a) {
            this.cq = a;
            this.g("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.h;
        }
        set labelFormatSpecifiers(a) {
            this.h = a;
            this.g("LabelFormatSpecifiers");
        }
        get transitionDuration() {
            return this.ca;
        }
        set transitionDuration(a) {
            this.ca = a;
            this.g("TransitionDuration");
        }
        get showToolTipTimeout() {
            return this.b9;
        }
        set showToolTipTimeout(a) {
            this.b9 = a;
            this.g("ShowToolTipTimeout");
        }
        get showToolTip() {
            return this.q;
        }
        set showToolTip(a) {
            this.q = a;
            this.g("ShowToolTip");
        }
        get font() {
            return this.cj;
        }
        set font(a) {
            this.cj = a;
            this.g("Font");
        }
        get pixelScalingRatio() {
            return this.al;
        }
        set pixelScalingRatio(a) {
            this.al = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.v;
        }
        set actualPixelScalingRatio(a) {
            this.v = a;
            this.g("ActualPixelScalingRatio");
        }
        get formatLabelRef() {
            return this.co;
        }
        set formatLabelRef(a) {
            this.co = a;
            this.g("FormatLabelRef");
        }
        get alignLabelRef() {
            return this.cf;
        }
        set alignLabelRef(a) {
            this.cf = a;
            this.g("AlignLabelRef");
        }
        get actualMinimumValueChangeRef() {
            return this.ce;
        }
        set actualMinimumValueChangeRef(a) {
            this.ce = a;
            this.g("ActualMinimumValueChangeRef");
        }
        get actualMaximumValueChangeRef() {
            return this.cd;
        }
        set actualMaximumValueChangeRef(a) {
            this.cd = a;
            this.g("ActualMaximumValueChangeRef");
        }
    }
    BulletGraphDescription.$t = /*@__PURE__*/ markType(BulletGraphDescription, 'BulletGraphDescription', Description.$);
    return BulletGraphDescription;
})();
