// igniteui-react-core/esm2015/lib/CalloutLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalloutLayerDescription = /*@__PURE__*/ (() => {
    class CalloutLayerDescription extends AnnotationLayerDescription {
        get_type() {
            return "CalloutLayer";
        }
        constructor() {
            super();
            this.fv = null;
            this.fz = null;
            this.f0 = null;
            this.fu = null;
            this.em = false;
            this.ft = null;
            this.fw = null;
            this.fx = null;
            this.fs = null;
            this.ej = null;
            this.el = false;
            this.ep = false;
            this.eq = false;
            this.er = false;
            this.fh = null;
            this.ez = 0;
            this.eo = false;
            this.en = false;
            this.fd = 0;
            this.fr = null;
            this.fi = null;
            this.fe = 0;
            this.fn = null;
            this.fm = null;
            this.e1 = 0;
            this.e3 = 0;
            this.e2 = 0;
            this.e0 = 0;
            this.fj = null;
            this.e4 = 0;
            this.e5 = 0;
            this.fy = null;
            this.fq = null;
            this.fo = null;
            this.fk = null;
            this.fl = null;
            this.fp = null;
        }
        get labelMemberPath() {
            return this.fv;
        }
        set labelMemberPath(a) {
            this.fv = a;
            this.g("LabelMemberPath");
        }
        get xMemberPath() {
            return this.fz;
        }
        set xMemberPath(a) {
            this.fz = a;
            this.g("XMemberPath");
        }
        get yMemberPath() {
            return this.f0;
        }
        set yMemberPath(a) {
            this.f0 = a;
            this.g("YMemberPath");
        }
        get keyMemberPath() {
            return this.fu;
        }
        set keyMemberPath(a) {
            this.fu = a;
            this.g("KeyMemberPath");
        }
        get isCalloutOffsettingEnabled() {
            return this.em;
        }
        set isCalloutOffsettingEnabled(a) {
            this.em = a;
            this.g("IsCalloutOffsettingEnabled");
        }
        get contentMemberPath() {
            return this.ft;
        }
        set contentMemberPath(a) {
            this.ft = a;
            this.g("ContentMemberPath");
        }
        get targetSeriesName() {
            return this.fw;
        }
        set targetSeriesName(a) {
            this.fw = a;
            this.g("TargetSeriesName");
        }
        get targetSeriesRef() {
            return this.fx;
        }
        set targetSeriesRef(a) {
            this.fx = a;
            this.g("TargetSeriesRef");
        }
        get collisionChannel() {
            return this.fs;
        }
        set collisionChannel(a) {
            this.fs = a;
            this.g("CollisionChannel");
        }
        get allowedPositions() {
            return this.ej;
        }
        set allowedPositions(a) {
            this.ej = a;
            this.g("AllowedPositions");
        }
        get isAutoCalloutBehaviorEnabled() {
            return this.el;
        }
        set isAutoCalloutBehaviorEnabled(a) {
            this.el = a;
            this.g("IsAutoCalloutBehaviorEnabled");
        }
        get useInterpolatedValueForAutoCalloutLabels() {
            return this.ep;
        }
        set useInterpolatedValueForAutoCalloutLabels(a) {
            this.ep = a;
            this.g("UseInterpolatedValueForAutoCalloutLabels");
        }
        get useSeriesColorForOutline() {
            return this.eq;
        }
        set useSeriesColorForOutline(a) {
            this.eq = a;
            this.g("UseSeriesColorForOutline");
        }
        get useValueForAutoCalloutLabels() {
            return this.er;
        }
        set useValueForAutoCalloutLabels(a) {
            this.er = a;
            this.g("UseValueForAutoCalloutLabels");
        }
        get autoCalloutVisibilityMode() {
            return this.fh;
        }
        set autoCalloutVisibilityMode(a) {
            this.fh = a;
            this.g("AutoCalloutVisibilityMode");
        }
        get calloutCornerRadius() {
            return this.ez;
        }
        set calloutCornerRadius(a) {
            this.ez = a;
            this.g("CalloutCornerRadius");
        }
        get isCustomCalloutStyleEnabled() {
            return this.eo;
        }
        set isCustomCalloutStyleEnabled(a) {
            this.eo = a;
            this.g("IsCustomCalloutStyleEnabled");
        }
        get isCustomCalloutRenderStyleEnabled() {
            return this.en;
        }
        set isCustomCalloutRenderStyleEnabled(a) {
            this.en = a;
            this.g("IsCustomCalloutRenderStyleEnabled");
        }
        get autoCalloutLabelPrecision() {
            return this.fd;
        }
        set autoCalloutLabelPrecision(a) {
            this.fd = a;
            this.g("AutoCalloutLabelPrecision");
        }
        get calloutTextColor() {
            return this.fr;
        }
        set calloutTextColor(a) {
            this.fr = a;
            this.g("CalloutTextColor");
        }
        get calloutBackground() {
            return this.fi;
        }
        set calloutBackground(a) {
            this.fi = a;
            this.g("CalloutBackground");
        }
        get calloutInterpolatedValuePrecision() {
            return this.fe;
        }
        set calloutInterpolatedValuePrecision(a) {
            this.fe = a;
            this.g("CalloutInterpolatedValuePrecision");
        }
        get calloutOutline() {
            return this.fn;
        }
        set calloutOutline(a) {
            this.fn = a;
            this.g("CalloutOutline");
        }
        get calloutLeaderBrush() {
            return this.fm;
        }
        set calloutLeaderBrush(a) {
            this.fm = a;
            this.g("CalloutLeaderBrush");
        }
        get calloutPaddingLeft() {
            return this.e1;
        }
        set calloutPaddingLeft(a) {
            this.e1 = a;
            this.g("CalloutPaddingLeft");
        }
        get calloutPaddingTop() {
            return this.e3;
        }
        set calloutPaddingTop(a) {
            this.e3 = a;
            this.g("CalloutPaddingTop");
        }
        get calloutPaddingRight() {
            return this.e2;
        }
        set calloutPaddingRight(a) {
            this.e2 = a;
            this.g("CalloutPaddingRight");
        }
        get calloutPaddingBottom() {
            return this.e0;
        }
        set calloutPaddingBottom(a) {
            this.e0 = a;
            this.g("CalloutPaddingBottom");
        }
        get calloutCollisionMode() {
            return this.fj;
        }
        set calloutCollisionMode(a) {
            this.fj = a;
            this.g("CalloutCollisionMode");
        }
        get calloutPositionPadding() {
            return this.e4;
        }
        set calloutPositionPadding(a) {
            this.e4 = a;
            this.g("CalloutPositionPadding");
        }
        get calloutStrokeThickness() {
            return this.e5;
        }
        set calloutStrokeThickness(a) {
            this.e5 = a;
            this.g("CalloutStrokeThickness");
        }
        get textStyle() {
            return this.fy;
        }
        set textStyle(a) {
            this.fy = a;
            this.g("TextStyle");
        }
        get calloutStyleUpdatingRef() {
            return this.fq;
        }
        set calloutStyleUpdatingRef(a) {
            this.fq = a;
            this.g("CalloutStyleUpdatingRef");
        }
        get calloutRenderStyleUpdatingRef() {
            return this.fo;
        }
        set calloutRenderStyleUpdatingRef(a) {
            this.fo = a;
            this.g("CalloutRenderStyleUpdatingRef");
        }
        get calloutContentUpdatingRef() {
            return this.fk;
        }
        set calloutContentUpdatingRef(a) {
            this.fk = a;
            this.g("CalloutContentUpdatingRef");
        }
        get calloutLabelUpdatingRef() {
            return this.fl;
        }
        set calloutLabelUpdatingRef(a) {
            this.fl = a;
            this.g("CalloutLabelUpdatingRef");
        }
        get calloutSeriesSelectingRef() {
            return this.fp;
        }
        set calloutSeriesSelectingRef(a) {
            this.fp = a;
            this.g("CalloutSeriesSelectingRef");
        }
    }
    CalloutLayerDescription.$t = /*@__PURE__*/ markType(CalloutLayerDescription, 'CalloutLayerDescription', AnnotationLayerDescription.$);
    return CalloutLayerDescription;
})();
