// igniteui-react-core/esm2015/lib/CodeGeneratingComponentRendererCodeEmitter_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, IList_$type, NotSupportedException, String_$type, typeCast, Array_$type, fromEnum, Boolean_$type, markType, EnumUtil, fromEn, Number_$type, runOn } from "./type";
import { CodeGenerationTargetPlatforms_$type } from "./CodeGenerationTargetPlatforms";
import { CodeGeneratingRendererResult } from "./CodeGeneratingRendererResult";
import { CodeGeneratingCodeWriter } from "./CodeGeneratingCodeWriter";
import { Stack$1 } from "./Stack$1";
import { CodeGenerationItemBuilderPropertyInfo } from "./CodeGenerationItemBuilderPropertyInfo";
import { List$1 } from "./List$1";
import { TypeDescriptionContext } from "./TypeDescriptionContext";
import { TypeDescriptionMetadata } from "./TypeDescriptionMetadata";
import { CodeGenerationSpecialValue } from "./CodeGenerationSpecialValue";
import { ColorUtil } from "./ColorUtil";
import { StringBuilder } from "./StringBuilder";
import { CodeGeneratingImportManager } from "./CodeGeneratingImportManager";
import { stringIsNullOrEmpty, stringReplace, stringStartsWith, stringEndsWith, stringContains } from "./string";
import { HashSet$1 } from "./HashSet$1";
import { WebImportsHalper } from "./WebImportsHalper";
import { CodeGeneratingComponentRendererHandlerEmitter } from "./CodeGeneratingComponentRendererHandlerEmitter";
import { CodeGeneratingRegexHelper } from "./CodeGeneratingRegexHelper";
import { CodeGeneratingComponentRendererDataEmitter } from "./CodeGeneratingComponentRendererDataEmitter";
import { Dictionary$2 } from "./Dictionary$2";
import { JsonDictionaryValue } from "./JsonDictionaryValue";
import { JsonDictionaryObject } from "./JsonDictionaryObject";
import { JsonDictionaryArray } from "./JsonDictionaryArray";
import { ComponentRenderer } from "./ComponentRenderer";
import { DotNetCodeGeneratingComponentRendererHandlerEmitter } from "./DotNetCodeGeneratingComponentRendererHandlerEmitter";
import { TSCodeGeneratingComponentRendererTemplateEmitter } from "./TSCodeGeneratingComponentRendererTemplateEmitter";
import { DotNetCodeGeneratingComponentRendererTemplateEmitter } from "./DotNetCodeGeneratingComponentRendererTemplateEmitter";
import { TSCodeGeneratingComponentRendererModuleEmitter } from "./TSCodeGeneratingComponentRendererModuleEmitter";
import { DotNetCodeGeneratingComponentRendererModuleEmitter } from "./DotNetCodeGeneratingComponentRendererModuleEmitter";
import { BlazorCodeGeneratingComponentRendererModuleEmitter } from "./BlazorCodeGeneratingComponentRendererModuleEmitter";
import { CodeGenerationTransformer } from "./CodeGenerationTransformer";
import { LinearGradientBrush } from "./LinearGradientBrush";
import { IComponentRendererAdapter_$type } from "./IComponentRendererAdapter";
/**
 * @hidden
 */
export let CodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class CodeGeneratingComponentRendererCodeEmitter extends Base {
        get_ai() {
            switch (this.j.platform) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
                case 4: return "Ultra";
                case 0: return "Xam";
                case 5: return "Igb";
            }
            return "";
        }
        get ai() {
            return this.get_ai();
        }
        get_aj() {
            switch (this.j.platform) {
                case 1: return "Component";
                case 2: return "";
                case 3: return "Component";
                case 4: return "";
                case 0: return "";
            }
            return "";
        }
        get aj() {
            return this.get_aj();
        }
        get a3() {
            switch (this.j.platform) {
                case 1: return "angular";
                case 2: return "react";
                case 3: return "webcomponents";
            }
            return "";
        }
        get m() {
            return this.l;
        }
        bn(a) {
            this.j.af(a);
        }
        get k() {
            return this.j;
        }
        constructor(a, b) {
            super();
            this.l = null;
            this.j = null;
            this.y = new List$1(CodeGenerationItemBuilderPropertyInfo.$, 0);
            this.d = null;
            this.a = null;
            this.c = null;
            this.b = null;
            this.z = new List$1(String_$type, 0);
            this.ac = new Stack$1(String_$type);
            this.aa = new Stack$1(String_$type);
            this.ab = new Stack$1(Boolean_$type);
            this.a2 = null;
            this.a1 = null;
            this.l = a;
            this.j = b;
            this.d = new CodeGeneratingCodeWriter(this.l);
            this.a = new CodeGeneratingCodeWriter(this.l);
            this.c = new CodeGeneratingCodeWriter(this.l);
            this.b = new CodeGeneratingCodeWriter(this.l);
        }
        get h() {
            return this.d;
        }
        get e() {
            return this.a;
        }
        get g() {
            return this.c;
        }
        get f() {
            return this.b;
        }
        a7(a) {
            if (a == null || a == "any") {
                return;
            }
            this.z.add(a);
        }
        v(a, b) {
            if (this.r) {
                return false;
            }
            if (a.hasMustSetInCodePlatforms(b.b.owningType, b.b.propertyName)) {
                let c = a.getMustSetInCodePlatforms(b.b.owningType, b.b.propertyName);
                for (let e = 0; e < c.length; e++) {
                    let d = c[e];
                    let f;
                    if (((() => { let g = EnumUtil.tryParse$1(CodeGenerationTargetPlatforms_$type, d, true, f); f = g.p2; return g.ret; })())) {
                        if (f == this.j.platform) {
                            return true;
                        }
                    }
                }
            }
            return false;
        }
        q(a, b) {
            if (b.b != null && (b.b.knownType == 20 || b.b.knownType == 19 || b.b.knownType == 22 || b.b.knownType == 18)) {
                if (b.a != null) {
                    return true;
                }
            }
            if (b.b != null && (b.b.knownType == 8 || b.b.knownType == 9 || b.b.knownType == 10)) {
                if (this.v(a, b)) {
                    return true;
                }
            }
            return false;
        }
        t(a) {
            if (a.b != null && (a.b.knownType == 19)) {
                if (a.a != null) {
                    return true;
                }
            }
            return false;
        }
        get an() {
            return this.ac.d();
        }
        a8(a, b, c) {
            this.ac.h(a);
        }
        bm() {
        }
        bg(a) {
        }
        bh(a) {
            this.ac.e();
        }
        be(a, b, c, d) {
        }
        bf(a, b, c, d) {
            this.be(a, b, c, d);
        }
        toString() {
            return this.d.toString();
        }
        ax() {
            return "null";
        }
        a0(a, b, c) {
            if (b.b == null) {
                return b.c;
            }
            if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
                b.c = this.ad(b);
            }
            switch (b.b.knownType) {
                case 10:
                    if (this.v(a, b)) {
                        return this.al(a, b);
                    }
                    return this.aq(a, b);
                case 23:
                    if (this.v(a, b)) {
                        return this.al(a, b);
                    }
                    return this.aq(a, b);
                case 7: return b.c ? "true" : "false";
                case 4: return this.af(b);
                case 6:
                    {
                        let d = "";
                        let e = b.c;
                        for (let f = 0; f < e.count; f++) {
                            if (f >= 1) {
                                d += " ";
                            }
                            d += this.ae(e.item(f));
                        }
                        return d;
                    }
                    break;
                case 9:
                    if (this.v(a, b)) {
                        return this.al(a, b);
                    }
                    throw new NotSupportedException(1, "shouldn't get here");
                case 5: return this.ah(b);
                case 16:
                    {
                        let g = "";
                        let h = b.c;
                        for (let i = 0; i < h.count; i++) {
                            if (i >= 1) {
                                g += " ";
                            }
                            g += this.ag(h.item(i));
                        }
                        return g;
                    }
                    break;
                case 20:
                    if (b.a != null) {
                        return this.az(b);
                    }
                    if (b.c == null) {
                        return this.ax();
                    }
                    break;
                case 22:
                    if (b.a != null) {
                        return this.az(b);
                    }
                    if (b.c == null) {
                        return this.ax();
                    }
                    break;
                case 15: throw new NotSupportedException(1, "need to figure this out");
                case 3: throw new NotSupportedException(1, "need to figure this out");
                case 19:
                    if (b.a != null) {
                        return this.au(b);
                    }
                    else {
                        return this.ax();
                    }
                    break;
                case 8:
                    if (b.b.specificExternalType == "string") {
                        return this.at(b);
                    }
                    if (b.a != null) {
                        return this.av(b);
                    }
                    if (this.v(a, b)) {
                        return this.al(a, b);
                    }
                    throw new NotSupportedException(1, "shouldn't get here");
                case 13: throw new NotSupportedException(1, "shouldn't get here");
                case 18:
                    if (b.a != null) {
                        return this.aw(b);
                    }
                    else {
                        return this.ax();
                    }
                    break;
                case 1: return b.c.toString();
                case 11:
                    {
                        let j = b.c;
                        return j.x + ", " + j.y;
                    }
                case 14:
                    {
                        let k = b.c;
                        return k.left + ", " + k.top + ", " + k.width + ", " + k.height;
                    }
                case 12:
                    {
                        let l = b.c;
                        return l.width + ", " + l.height;
                    }
                case 2: return b.c != null ? b.c.toString() : this.ax();
                case 21: return b.c.toString();
                case 17:
                    if (b.c == null) {
                        return "null";
                    }
                    if (typeof b.c === 'string') {
                        return b.c.toString();
                    }
                    return b.c.toString();
                case 0: throw new NotSupportedException(1, "shouldn't get here");
            }
            return "";
        }
        al(a, b) {
            let c = this.a5(a, b);
            if (b.c != null) {
                if (typeCast(CodeGenerationItemBuilder.$, b.c) !== null) {
                    let d = this.i(this.l, this.j);
                    let e = this.ak(c, b.c, d);
                    c = this.ao(a, b.c, e, c);
                }
                else if (typeCast(Array_$type, b.c) !== null) {
                    let f = b.c;
                    let g = new List$1(CodeGenerationItemBuilder.$, 0);
                    let h = new List$1(String_$type, 0);
                    for (let i = 0; i < f.length; i++) {
                        if (typeCast(CodeGenerationItemBuilder.$, f[i]) !== null) {
                            let j = this.i(this.l, this.j);
                            let k = c;
                            let l = f[i].o();
                            if (l == null) {
                                l = f[i].m();
                                l = this.ar(l);
                            }
                            if (l != null) {
                                if (l != k) {
                                    k = l;
                                }
                                else {
                                    k = c + "_" + i.toString();
                                }
                            }
                            let m = this.ak(k, f[i], j);
                            g.add(f[i]);
                            h.add(m);
                        }
                    }
                    c = this.ap(b, a, g, h, c);
                }
            }
            return c;
        }
        ak(a, b, c) {
            b.t = a;
            b.u(c, true);
            let d = c.toString();
            for (let e of fromEnum(c.z)) {
                this.z.add(e);
            }
            return d;
        }
        ao(a, b, c, d) {
            return d;
        }
        ap(a, b, c, d, e) {
            return e;
        }
        a5(a, b) {
            if (b.c == null) {
                return this.ax();
            }
            if (typeCast(CodeGenerationItemBuilder.$, b.c) !== null) {
                let c = b.c.o();
                if (c != null) {
                    return c;
                }
                let d = b.c.m();
                let e = this.ar(d);
                return e;
            }
            else if (typeCast(IList_$type, b.c) !== null) {
                let f = b.c;
                if (f.count > 0 && typeCast(CodeGenerationItemBuilder.$, f.item(0)) !== null) {
                    let g = f.item(0);
                    let h = g.o();
                    if (h != null) {
                        return h;
                    }
                    let i = g.m();
                    let j = this.ar(i);
                    return j;
                }
                return this.ar(b.b.propertyName);
            }
            else if (typeCast(Array_$type, b.c) !== null) {
                let k = b.c;
                if (k.length > 0 && typeCast(CodeGenerationItemBuilder.$, k[0]) !== null) {
                    let l = b.c.o();
                    if (l != null) {
                        return l;
                    }
                    let m = b.c.m();
                    let n = this.ar(m);
                    return n;
                }
                return this.ar(b.b.propertyName);
            }
            return this.ax();
        }
        aq(a, b) {
            return "";
        }
        ad(a) {
            return a.c;
        }
        az(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                if (a.a.a) {
                    return this.as(a);
                }
                else {
                    return this.ay(a);
                }
            }
            return this.ax();
        }
        s(a) {
            if (a.a != null && !stringIsNullOrEmpty(a.a.e)) {
                if (a.a.a) {
                    return true;
                }
            }
            return false;
        }
        ay(a) {
            return "";
        }
        as(a) {
            return "";
        }
        av(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                return a.a.e;
            }
            return this.ax();
        }
        at(a) {
            return a.c != null ? a.c.toString() : "";
        }
        aw(a) {
            return this.au(a);
        }
        au(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.bn(a);
                return a.a.e;
            }
            return this.ax();
        }
        get_w() {
            return false;
        }
        get w() {
            return this.get_w();
        }
        ah(a) {
            let b = a.c;
            return ColorUtil.i(b, this.w);
        }
        ag(a) {
            let b = a;
            return ColorUtil.i(b, this.w);
        }
        af(a) {
            if (a.c == null) {
                return "";
            }
            let b = a.c;
            return ColorUtil.i(b.color, this.w);
        }
        ae(a) {
            if (a == null) {
                return "";
            }
            let b = a;
            return ColorUtil.i(b.color, this.w);
        }
        bl(a) {
            this.aa.h(a);
            this.ab.h(false);
        }
        bk(a) {
            let b = this.ar(a);
            this.aa.h(b);
            this.ab.h(true);
        }
        ar(a) {
            let b = stringReplace(stringReplace(a, this.a4, ""), "Component", "");
            b = CodeGeneratingComponentRenderer.d4(b);
            return this.j.v(b);
        }
        bi() {
            this.a1 = this.aa.e();
            this.ab.e();
        }
        get_a4() {
            return "";
        }
        get a4() {
            return this.get_a4();
        }
        get am() {
            if (this.aa.f > 0) {
                return this.aa.d();
            }
            let a = stringReplace(stringReplace(this.an, this.a4, ""), "Component", "");
            return CodeGeneratingComponentRenderer.d4(a);
        }
        get u() {
            return this.ac.f == 1;
        }
        get o() {
            if (this.ab.f == 0) {
                return false;
            }
            return this.ab.d();
        }
        get p() {
            if (this.aa.f > 0) {
                return true;
            }
            return false;
        }
        get_x() {
            return true;
        }
        get x() {
            return this.get_x();
        }
        get_n() {
            return this.r;
        }
        get n() {
            return this.get_n();
        }
        bj() {
            this.a1 = this.aa.e();
        }
        ba(a) {
            let b = this.f.toString();
            if (b.length > 0) {
                a.l(b);
            }
        }
        bd(a) {
            let b = this.g.toString();
            if (b.length > 0) {
                a.l(b);
            }
        }
        a9(a) {
            let b = this.e.toString();
            if (b.length > 0) {
                a.l(b);
            }
        }
        bb(a, b) {
            if (this.z.count > 0) {
                this.bc(a, this.z, b);
            }
        }
        bc(a, b, c) {
        }
    }
    CodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(CodeGeneratingComponentRendererCodeEmitter, 'CodeGeneratingComponentRendererCodeEmitter');
    return CodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let BlazorRazorCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class BlazorRazorCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        ao(a, b, c, d) {
            d = super.ao(a, b, c, d);
            this.f.l("private " + b.m() + " _" + CodeGeneratingComponentRenderer.d4(d) + " = null;");
            this.f.l("public " + b.m() + " " + CodeGeneratingComponentRenderer.d5(d));
            this.f.l("{");
            this.f.h();
            this.f.l("get");
            this.f.l("{");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.d4(d) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k(c);
            this.f.l("this._" + CodeGeneratingComponentRenderer.d4(d) + " = " + d + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.d4(d) + ";");
            this.f.f();
            this.f.l("}");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.d5(d);
        }
        ap(a, b, c, d, e) {
            e = super.ap(a, b, c, d, e);
            let f = null;
            if (c.count > 0) {
                f = c._inner[0].q() + "[]";
            }
            else {
                let g = a.b.specificExternalType;
                f = CodeGenerationItemBuilder.n(this.k.platform, g, b, this.ai, this.aj) + "[]";
            }
            this.f.l("private " + f + " _" + CodeGeneratingComponentRenderer.d4(e) + " = null;");
            this.f.l("public " + f + " " + CodeGeneratingComponentRenderer.d5(e));
            this.f.l("{");
            this.f.h();
            this.f.l("get");
            this.f.l("{");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.d4(e) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k("var " + CodeGeneratingComponentRenderer.d4(e) + " = new " + stringReplace(f, "[]", "") + "[" + c.count + "];");
            for (let h = 0; h < c.count; h++) {
                this.f.k(d._inner[h]);
                this.f.k(e + "[" + h + "] = " + c._inner[h].t + ";");
            }
            this.f.l("this._" + CodeGeneratingComponentRenderer.d4(e) + " = " + e + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.d4(e) + ";");
            this.f.f();
            this.f.l("}");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.d5(e);
        }
        i(a, b) {
            return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        get_x() {
            return false;
        }
        constructor(a, b) {
            super(a, b);
            this.bq = 0;
            this.bo = new HashSet$1(String_$type, 0);
            this.bp = new HashSet$1(String_$type, 0);
        }
        br(a, b, c) {
            if (c.c == null) {
                return "null";
            }
            let d = c.c;
            let e = b;
            let f = new CodeGenerationItemBuilderPropertyInfo();
            f.b = ((() => {
                let $ret = new TypeDescriptionMetadata();
                $ret.propertyName = c.b.propertyName;
                $ret.specificType = b;
                $ret.specificExternalType = b;
                return $ret;
            })());
            f.b.knownType = 8;
            f.b.owningType = c.b.owningType;
            if (f.b.specificType == "string") {
                f.b.knownType = 2;
            }
            if (f.b.specificType == "int") {
                f.b.knownType = 1;
            }
            if (f.b.specificType == "double") {
                f.b.knownType = 1;
            }
            if (f.b.knownType == 8) {
                e = this.a4 + e;
            }
            let g = "@(new " + e + "[] { ";
            for (let h = 0; h < d.length; h++) {
                if (h > 0) {
                    g += ", ";
                }
                f.c = d[h];
                if (f.b.knownType == 2) {
                    g += "\"" + this.a0(a, f, true) + "\"";
                }
                else {
                    g += this.a0(a, f, true);
                }
            }
            g += " })";
            return g;
        }
        aq(a, b) {
            let c = b.b.specificType;
            if (c == null) {
                c = b.b.specificExternalType;
            }
            return this.br(a, c, b);
        }
        a8(a, b, c) {
            super.a8(a, b, c);
            this.h.l("<" + a + "");
            if (!this.bo.contains(this.am)) {
                this.bo.add_1(this.am);
                if (!this.bp.contains(this.an)) {
                    this.bp.add_1(this.an);
                    this.a7(this.an);
                }
                this.f.l("private " + this.an + " " + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                if (!this.p) {
                    if (this.u) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                }
            }
        }
        bm() {
            this.h.l(">");
            this.h.h();
        }
        bh(a) {
            this.h.f();
            this.h.l("</" + a + ">");
            this.h.l("");
            super.bh(a);
        }
        be(a, b, c, d) {
            if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.d5(c.b.propertyName)) && this.q(d, c)) {
                this.bn(c);
                if (a > 0) {
                    this.h.l("");
                }
                if (c.a == null || c.a.e == null) {
                    this.h.j(b + "Name=\"@null\"");
                }
                else {
                    this.h.j(b + "Name=\"" + c.a.e + "\"");
                }
                return;
            }
            if (this.s(c)) {
                this.bn(c);
                if (!this.bo.contains(this.am)) {
                    this.bo.add_1(this.am);
                    if (!this.bp.contains(this.an)) {
                        this.bp.add_1(this.an);
                        this.a7(this.an);
                    }
                    this.f.l("private " + this.an + " " + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                    if (!this.p) {
                        if (this.u) {
                            this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                        }
                    }
                    else {
                        this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                    }
                }
                let e = CodeGeneratingComponentRenderer.d4(this.a0(d, c, true));
                let f = e;
                if (f != "null" && !stringStartsWith(f, "this.")) {
                    f = "this." + f;
                }
                this.e.l(CodeGeneratingComponentRenderer.d4(this.am) + "." + b + " = " + f + ";");
                return;
            }
            if (a > 0) {
                this.h.l("");
            }
            let g = false;
            if (b == "ComponentRenderer" && (c.b.owningType == "PropertyEditor" || c.b.owningType == "PropertyEditorPanel")) {
                g = true;
            }
            if (!g) {
                if (c.a != null && c.a.e != null) {
                    if (this.m != null) {
                        if (this.m.library != null) {
                            if (this.m.library.hasItem(c.a.e)) {
                                let h = this.m.library.getItem(c.a.e);
                                let i = h.getContentForPlatform(5);
                                if (i != null && i.isJSContent) {
                                    b = b + "Script";
                                }
                            }
                        }
                    }
                }
                this.h.j(b);
                this.h.j("=");
                this.h.j("\"");
                this.h.j(this.a0(d, c, true));
                this.h.j("\"");
            }
            if (b == "Name") {
                this.h.l("");
                this.h.j("@ref=\"" + CodeGeneratingComponentRenderer.d4(this.a0(d, c, true)) + "\"");
            }
        }
        at(a) {
            let b = a.b != null ? a.b.specificType + "." : "";
            return b + (a.c != null ? a.c.toString() : "");
        }
        a0(a, b, c) {
            if (b.b == null) {
                return b.c;
            }
            if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
                b.c = this.ad(b);
            }
            switch (b.b.knownType) {
                case 11:
                    {
                        let d = b.c;
                        return "@(new Point(" + d.x + ", " + d.y + "))";
                    }
                case 14:
                    {
                        let e = b.c;
                        return "@(new Rect(" + e.left + ", " + e.top + ", " + e.width + ", " + e.height + "))";
                    }
                case 12:
                    {
                        let f = b.c;
                        return "@(new Size(" + f.width + ", " + f.height + "))";
                    }
                case 17:
                    if (b.c == null) {
                        return "null";
                    }
                    if (typeof b.c === 'string') {
                        return "@(\"" + b.c.toString() + "\")";
                    }
                    if (typeCast(Boolean_$type, b.c) !== null) {
                        return b.c ? "true" : "false";
                    }
                    return b.c.toString();
            }
            return super.a0(a, b, c);
        }
        ad(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        ax() {
            return "@null";
        }
        as(a) {
            return CodeGeneratingComponentRenderer.d4(CodeGeneratingComponentRenderer.d4(a.a.e));
        }
        bs(a) {
            a = stringReplace(a, "-", "_");
            return CodeGeneratingComponentRenderer.d5(a);
        }
        ay(a) {
            this.bn(a);
            return CodeGeneratingComponentRenderer.d5(a.a.e);
        }
        bf(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.u(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.u(this, false);
                }
            }
        }
        get_a4() {
            return "Igb";
        }
    }
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(BlazorRazorCodeGeneratingComponentRendererCodeEmitter, 'BlazorRazorCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return BlazorRazorCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        get r() {
            return this.get_r();
        }
        i(a, b) {
            return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        get_x() {
            return false;
        }
        constructor(a, b) {
            super(a, b);
            this.bq = 0;
            this.bo = new HashSet$1(String_$type, 0);
            this.bp = new HashSet$1(String_$type, 0);
        }
        br(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "new " + a + "[] { ";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += "\"" + c[e] + "\"";
            }
            d += " }";
            return d;
        }
        aq(a, b) {
            if (b.b.specificExternalType == "string") {
                return this.br(b.b.specificExternalType, b);
            }
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        a8(a, b, c) {
            super.a8(a, b, c);
            if (this.u) {
                this.h.l("var " + this.am + " = new " + a + "();");
            }
            else {
                this.h.l("var " + this.am + " = new " + a + "();");
            }
            if (!this.bo.contains(this.am)) {
                this.bo.add_1(this.am);
                if (!this.bp.contains(this.an)) {
                    this.bp.add_1(this.an);
                    this.a7(this.an);
                }
                this.f.l("private " + this.an + " " + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                if (!this.p) {
                    if (this.u) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                }
            }
        }
        bm() {
        }
        bh(a) {
            super.bh(a);
        }
        be(a, b, c, d) {
            if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.d5(c.b.propertyName)) && this.q(d, c)) {
                this.bn(c);
                if (a > 0) {
                    this.h.l("");
                }
                if (c.a == null || c.a.e == null) {
                    this.h.j(this.am + "." + b + "Name = null;");
                }
                else {
                    this.h.j(this.am + "." + b + "Name = " + c.a.e + ";");
                }
                return;
            }
            if (this.s(c)) {
                this.bn(c);
                if (!this.bo.contains(this.am)) {
                    this.bo.add_1(this.am);
                    if (!this.bp.contains(this.an)) {
                        this.bp.add_1(this.an);
                        this.a7(this.an);
                    }
                    this.f.l("private " + this.an + " " + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                    if (!this.p) {
                        if (this.u) {
                            this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                        }
                    }
                    else {
                        this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + ";");
                    }
                }
                let e = CodeGeneratingComponentRenderer.d4(this.a0(d, c, true));
                let f = e;
                if (f != "null" && !stringStartsWith(f, "this.")) {
                    f = "this." + f;
                }
                this.e.l(CodeGeneratingComponentRenderer.d4(this.am) + "." + b + " = " + f + ";");
                return;
            }
            if (a > 0) {
                this.h.l("");
            }
            let g = false;
            if (b == "ComponentRenderer" && (c.b.owningType == "PropertyEditor" || c.b.owningType == "PropertyEditorPanel")) {
                g = true;
            }
            if (!g) {
                let h = false;
                if (c.a != null && c.a.e != null) {
                    if (this.m != null) {
                        if (this.m.library != null) {
                            if (this.m.library.hasItem(c.a.e)) {
                                let i = this.m.library.getItem(c.a.e);
                                let j = i.getContentForPlatform(5);
                                if (j != null && j.isJSContent) {
                                    b = b + "Script";
                                    h = true;
                                }
                            }
                        }
                    }
                }
                this.h.j(this.am + ".");
                this.h.j(b);
                this.h.j(" = ");
                if (h) {
                    this.h.j("\"" + this.a0(d, c, true) + "\"");
                }
                else {
                    this.h.j(this.a0(d, c, true));
                }
                this.h.j(";");
            }
        }
        at(a) {
            let b = a.b != null ? a.b.specificType + "." : "";
            return b + (a.c != null ? a.c.toString() : "");
        }
        a0(a, b, c) {
            if (b.b == null) {
                return "\"" + b.c + "\"";
            }
            if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
                b.c = this.ad(b);
            }
            switch (b.b.knownType) {
                case 11:
                    {
                        let d = b.c;
                        return "new Point(" + d.x + ", " + d.y + ")";
                    }
                case 14:
                    {
                        let e = b.c;
                        return "new Rect(" + e.left + ", " + e.top + ", " + e.width + ", " + e.height + ")";
                    }
                case 12:
                    {
                        let f = b.c;
                        return "new Size(" + f.width + ", " + f.height + ")";
                    }
                case 17:
                    if (b.c == null) {
                        return "null";
                    }
                    if (typeof b.c === 'string') {
                        return "\"" + b.c.toString() + "\"";
                    }
                    if (typeCast(Boolean_$type, b.c) !== null) {
                        return b.c ? "true" : "false";
                    }
                    return b.c.toString();
                case 5:
                case 6:
                case 16:
                case 2: return "\"" + super.a0(a, b, c) + "\"";
                case 9: throw new NotSupportedException(1, "shouldn't get here");
            }
            return super.a0(a, b, c);
        }
        ad(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        ax() {
            return "null";
        }
        as(a) {
            return CodeGeneratingComponentRenderer.d4(CodeGeneratingComponentRenderer.d4(a.a.e));
        }
        bs(a) {
            a = stringReplace(a, "-", "_");
            return CodeGeneratingComponentRenderer.d5(a);
        }
        ay(a) {
            this.bn(a);
            return CodeGeneratingComponentRenderer.d5(a.a.e);
        }
        bf(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.u(this, false);
                            this.h.l(this.am + "." + b + ".Add(" + this.a1 + ");");
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.u(this, false);
                    this.h.l(this.am + "." + b + " = " + this.a1 + ";");
                }
            }
        }
        get_a4() {
            return "Igb";
        }
    }
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let ReactMarkupCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class ReactMarkupCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        ao(a, b, c, d) {
            d = super.ao(a, b, c, d);
            let e = b.q();
            this.f.l("private " + " _" + CodeGeneratingComponentRenderer.d4(d) + ": " + e + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.d4(d) + "(): " + e + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.d4(d) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k(c);
            this.f.l("this._" + CodeGeneratingComponentRenderer.d4(d) + " = " + d + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.d4(d) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.d5(d);
        }
        ap(a, b, c, d, e) {
            e = super.ap(a, b, c, d, e);
            let f = null;
            if (c.count > 0) {
                f = c._inner[0].q() + "[]";
            }
            else {
                let g = a.b.specificExternalType;
                f = CodeGenerationItemBuilder.n(this.k.platform, g, b, this.ai, this.aj) + "[]";
            }
            this.f.l("private" + " _" + CodeGeneratingComponentRenderer.d4(e) + ": " + f + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.d4(e) + "(): " + f + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.d4(e) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k("let " + CodeGeneratingComponentRenderer.d4(e) + ": " + f + " = [];");
            for (let h = 0; h < c.count; h++) {
                this.f.k(d._inner[h]);
                this.f.k(e + ".push(" + c._inner[h].t + ")");
            }
            this.f.l("this._" + CodeGeneratingComponentRenderer.d4(e) + " = " + e + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.d4(e) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.d5(e);
        }
        i(a, b) {
            return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
            this.bo = new HashSet$1(String_$type, 0);
            this.bp = new HashSet$1(String_$type, 0);
        }
        aq(a, b) {
            if (b.c == null) {
                return "{null}";
            }
            if (b.b != null && b.b.knownType == 23) {
                let c = b.c;
                let d = "";
                for (let e = 0; e < c.length; e++) {
                    if (e > 0) {
                        d += ", ";
                    }
                    d += c[e];
                }
                return d;
            }
            let f = b.c;
            let g = "[";
            for (let h = 0; h < f.length; h++) {
                if (h > 0) {
                    g += ", ";
                }
                g += "\"" + f[h].toString() + "\"";
            }
            g += "]";
            return g;
        }
        get_w() {
            return true;
        }
        get_a4() {
            return "Igr";
        }
        a8(a, b, c) {
            super.a8(a, b, c);
            this.h.l("<" + this.br(a) + "");
            this.h.h();
            if (!this.bp.contains(this.an)) {
                this.bp.add_1(this.an);
                this.a7(this.an);
            }
            if (!this.bo.contains(this.am)) {
                this.bo.add_1(this.am);
                this.f.l("private " + CodeGeneratingComponentRenderer.d4(this.am) + ": " + this.an);
                if (this.u) {
                    this.f.l("private " + CodeGeneratingComponentRenderer.d4(this.am) + "Ref(r: " + this.an + ") {");
                    this.f.h();
                    this.f.l("this." + CodeGeneratingComponentRenderer.d4(this.am) + " = r;");
                    this.f.l("this.setState({});");
                    this.f.f();
                    this.f.l("}");
                    this.g.l("this." + CodeGeneratingComponentRenderer.d4(this.am) + "Ref = " + "this." + CodeGeneratingComponentRenderer.d4(this.am) + "Ref.bind(this);");
                }
            }
        }
        br(a) {
            if (stringEndsWith(a, "Component")) {
                a = a.substr(0, a.length - ("Component").length);
            }
            return CodeGeneratingComponentRenderer.d5(a);
        }
        bm() {
            if (!this.u && !this.p) {
                let a = this.bq();
                this.h.l("");
                this.h.l("name=\"" + a + "\"");
            }
            this.h.f();
            this.h.l(">");
            this.h.h();
        }
        bq() {
            let a = CodeGeneratingComponentRenderer.d4(stringReplace(stringReplace(this.an, this.a4, ""), "Component", ""));
            a = this.k.u(a);
            return a;
        }
        au(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.bn(a);
                if (!this.bo.contains(a.a.e)) {
                    this.g.l("this." + CodeGeneratingComponentRenderer.d4(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.d4(a.a.e) + ".bind(this);");
                }
                return "this." + CodeGeneratingComponentRenderer.d4(a.a.e);
            }
            return this.ax();
        }
        bh(a) {
            this.h.f();
            this.h.l("</" + this.br(a) + ">");
            super.bh(a);
        }
        be(a, b, c, d) {
            let e = CodeGeneratingComponentRenderer.d4(b);
            let f = this.a0(d, c, true);
            if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.d5(c.b.propertyName)) && this.q(d, c)) {
                this.bn(c);
                if (a > 0) {
                    this.h.l("");
                }
                if (c.a == null || c.a.e == null) {
                    this.h.j(e + "Name={null}");
                }
                else {
                    this.h.j(e + "Name=\"" + c.a.e + "\"");
                }
                return;
            }
            if (!(e == "name") || !this.u) {
                let g = (c.b != null && c.b.knownType == 10);
                if (this.q(d, c) || g) {
                    this.bn(c);
                    f = "{" + f + "}";
                }
                if (a > 0) {
                    this.h.l("");
                }
                this.h.j(e);
                this.h.j("=");
                if (!this.q(d, c) && !g) {
                    this.h.j("\"");
                }
                this.h.j(f);
                if (!this.q(d, c) && !g) {
                    this.h.j("\"");
                }
                a++;
            }
            if (e == "name") {
                if (this.u) {
                    if (a > 0) {
                        this.h.l("");
                    }
                    this.h.j("ref={this." + CodeGeneratingComponentRenderer.d4(this.a0(d, c, true)) + "Ref}");
                }
            }
        }
        ad(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        ax() {
            return "null";
        }
        as(a) {
            return "this." + CodeGeneratingComponentRenderer.d4(a.a.e);
        }
        ay(a) {
            return "this." + CodeGeneratingComponentRenderer.d4(a.a.e);
        }
        bf(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.u(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.u(this, false);
                }
            }
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            WebImportsHalper.a(this.k, this.a3, this.a2, this.a4, a, b, c);
        }
    }
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(ReactMarkupCodeGeneratingComponentRendererCodeEmitter, 'ReactMarkupCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return ReactMarkupCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let WebCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WebCodeBehindCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        get r() {
            return this.get_r();
        }
        constructor(a, b) {
            super(a, b);
            this.br = 0;
            this.bp = new HashSet$1(String_$type, 0);
            this.bq = new HashSet$1(String_$type, 0);
        }
        bo(a) {
            if (!a.b.owningContext.hasStringUnionPlatforms(a.b.owningType, a.b.propertyName)) {
                return false;
            }
            let b = a.b.owningContext.getStringUnionPlatforms(a.b.owningType, a.b.propertyName);
            for (let c = 0; c < b.length; c++) {
                if (b[c].toLowerCase() == this.a3) {
                    return true;
                }
            }
            return false;
        }
        at(a) {
            if (this.bo(a)) {
                return "\"" + CodeGeneratingComponentRenderer.d6(a.c.toString()) + "\"";
            }
            else {
                if (!this.bq.contains(a.b.specificType)) {
                    this.bq.add_1(a.b.specificType);
                    if (a.b.owningContext.getMetadata(a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)) != null) {
                        let b = a.b.owningContext.getMetadata(a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)).specificExternalType;
                        this.a7(a.b.specificType + ":" + b);
                    }
                    else if (a.b.owningContext.getMetadata("Web" + a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)) != null) {
                        let c = a.b.owningContext.getMetadata("Web" + a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)).specificExternalType;
                        this.a7(a.b.specificType + ":" + c);
                    }
                    else {
                        this.a7(a.b.specificType);
                    }
                }
                let d = a.b != null ? a.b.specificType + "." : "";
                return d + (a.c != null ? a.c.toString() : "");
            }
        }
        aq(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "[";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d + "]";
        }
        get_w() {
            return true;
        }
        a8(a, b, c) {
            super.a8(a, b, c);
            let d = false;
            if (c.getMetadata(b, "__isPlainObject") != null) {
                d = true;
            }
            if (d) {
                if (this.u) {
                    this.h.l("var " + this.am + ": " + a + " = {} as " + a + ";");
                }
                else {
                    this.h.l("var " + this.am + ": " + a + " = {} as " + a + ";");
                }
            }
            else {
                if (this.u) {
                    this.h.l("var " + this.am + " = new " + a + "();");
                }
                else {
                    this.h.l("var " + this.am + " = new " + a + "();");
                }
            }
            if (!this.bp.contains(this.am)) {
                this.bp.add_1(this.am);
                if (!this.bq.contains(this.an)) {
                    this.bq.add_1(this.an);
                    this.a7(this.an);
                }
                this.f.l("private " + CodeGeneratingComponentRenderer.d4(this.am) + ": " + this.an);
                if (!this.p) {
                    if (this.u) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + " = document.querySelector('" + CodeGeneratingComponentRenderer.d7(stringReplace(this.an, "Component", "")) + "') as " + this.an + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + " = document.getElementById('" + this.am + "') as " + this.an + ";");
                }
            }
        }
        bs(a) {
            if (stringEndsWith(a, "Component")) {
                a = a.substr(0, a.length - ("Component").length);
            }
            return CodeGeneratingComponentRenderer.d7(a);
        }
        bm() {
            this.h.l("");
        }
        bg(a) {
        }
        bh(a) {
            {
                this.h.l("");
            }
            super.bh(a);
        }
        be(a, b, c, d) {
            let e = CodeGeneratingComponentRenderer.d4(b);
            if (e == "name" && c.b == null) {
                return;
            }
            if (this.s(c)) {
                this.bn(c);
                if (!this.bp.contains(this.am)) {
                    this.bp.add_1(this.am);
                    if (!this.bq.contains(this.an)) {
                        this.bq.add_1(this.an);
                        this.a7(this.an);
                    }
                    this.f.l("private " + CodeGeneratingComponentRenderer.d4(this.am) + ": " + this.an);
                    if (!this.p) {
                        if (this.u) {
                            this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + " = document.querySelector('" + CodeGeneratingComponentRenderer.d7(stringReplace(this.an, "Component", "")) + "') as " + this.an + ";");
                        }
                    }
                    else {
                        this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + " = document.getElementById('" + this.am + "') as " + this.an + ";");
                    }
                }
                let f = CodeGeneratingComponentRenderer.d4(this.a0(d, c, true));
                let g = f;
                if (g != "null" && !stringStartsWith(g, "this.")) {
                    g = "this." + g;
                }
                if (this.an == "any" && stringContains(e, "-")) {
                    this.e.l(CodeGeneratingComponentRenderer.d4(this.am) + "['" + e + "'] = " + g + ";");
                }
                else {
                    if (c.b != null && c.b.isCustomEvent) {
                        this.e.l(CodeGeneratingComponentRenderer.d4(this.am) + ".addEventListener(\"" + e + "\", " + g + ");");
                    }
                    else {
                        this.e.l(CodeGeneratingComponentRenderer.d4(this.am) + "." + e + " = " + g + ";");
                    }
                }
                return;
            }
            if (a > 0) {
                this.h.l("");
            }
            if (this.an == "any" && stringContains(e, "-")) {
                this.h.j(this.am + "." + CodeGeneratingComponentRenderer.d4(this.am) + "['" + e + "']");
            }
            else {
                this.h.j(this.am + "." + CodeGeneratingComponentRenderer.d4(e));
            }
            this.h.j(" = ");
            this.h.j(this.a0(d, c, true));
            this.h.j(";");
        }
        ad(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        ax() {
            return "null";
        }
        as(a) {
            return CodeGeneratingComponentRenderer.d4(a.a.e);
        }
        ay(a) {
            return CodeGeneratingComponentRenderer.d4(a.a.e);
        }
        au(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.bn(a);
                if (!this.bp.contains(a.a.e)) {
                    this.g.l("this." + CodeGeneratingComponentRenderer.d4(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.d4(a.a.e) + ".bind(this);");
                }
                return "this." + CodeGeneratingComponentRenderer.d4(a.a.e);
            }
            return this.ax();
        }
        a0(a, b, c) {
            if (b.b == null) {
                return "\"" + b.c + "\"";
            }
            if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
                b.c = this.ad(b);
            }
            switch (b.b.knownType) {
                case 11:
                    {
                        let d = b.c;
                        return "{ x: " + d.x + ", y: " + d.y + "}";
                    }
                case 14:
                    {
                        let e = b.c;
                        return "{ left: " + e.left + ", top: " + e.top + ", width: " + e.width + ", height: " + e.height + "}";
                    }
                case 12:
                    {
                        let f = b.c;
                        return "{ width: " + f.width + ", height: " + f.height + "}";
                    }
                case 17:
                    if (b.c == null) {
                        return "null";
                    }
                    if (typeof b.c === 'string') {
                        return "\"" + b.c.toString() + "\"";
                    }
                    if (typeCast(Boolean_$type, b.c) !== null) {
                        return b.c ? "true" : "false";
                    }
                    return b.c.toString();
                case 5:
                case 23:
                case 6:
                case 16:
                case 2: return "\"" + super.a0(a, b, c) + "\"";
                case 9: throw new NotSupportedException(1, "shouldn't get here");
            }
            return super.a0(a, b, c);
        }
        bf(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    let f = true;
                    for (let g of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, g) !== null) {
                            let h = g;
                            h.u(this, false);
                            let i = c.b.owningType;
                            let j = d.getMetadata(i, "__isPlainObject");
                            let k = false;
                            if (j != null) {
                                k = true;
                            }
                            if (k) {
                                if (f) {
                                    f = false;
                                    this.h.l(this.am + "." + b + " = []");
                                }
                                this.h.l(this.am + "." + b + ".push(" + this.a1 + ");");
                            }
                            else {
                                this.h.l(this.am + "." + b + ".add(" + this.a1 + ");");
                            }
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let l = c.c;
                    l.u(this, false);
                    this.h.l(this.am + "." + b + " = " + this.a1 + ";");
                }
            }
        }
    }
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WebCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WebCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WebCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter extends WebCodeBehindCodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        i(a, b) {
            return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
        }
        get_a4() {
            return "Igc";
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            WebImportsHalper.a(this.k, this.a3, this.a2, this.a4, a, b, c);
        }
    }
    ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        bo(a) {
            if (!a.b.owningContext.hasStringUnionPlatforms(a.b.owningType, a.b.propertyName)) {
                return false;
            }
            let b = a.b.owningContext.getStringUnionPlatforms(a.b.owningType, a.b.propertyName);
            for (let c = 0; c < b.length; c++) {
                if (b[c].toLowerCase() == this.a3) {
                    return true;
                }
            }
            return false;
        }
        at(a) {
            if (this.bo(a)) {
                return CodeGeneratingComponentRenderer.d6(a.c.toString());
            }
            else {
                return super.at(a);
            }
        }
        i(a, b) {
            return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        ao(a, b, c, d) {
            d = super.ao(a, b, c, d);
            let e = b.q();
            this.f.l("private" + " _" + CodeGeneratingComponentRenderer.d4(d) + ": " + e + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.d4(d) + "(): " + e + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.d4(d) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k(c);
            this.f.l("this._" + CodeGeneratingComponentRenderer.d4(d) + " = " + d + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.d4(d) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.d5(d);
        }
        ap(a, b, c, d, e) {
            e = super.ap(a, b, c, d, e);
            let f = null;
            if (c.count > 0) {
                f = c._inner[0].q() + "[]";
            }
            else {
                let g = a.b.specificExternalType;
                f = CodeGenerationItemBuilder.n(this.k.platform, g, b, this.ai, this.aj) + "[]";
            }
            this.f.l("private" + " _" + CodeGeneratingComponentRenderer.d4(e) + ": " + f + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.d4(e) + "(): " + f + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.d4(e) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k("let " + CodeGeneratingComponentRenderer.d4(e) + ": " + f + " = [];");
            for (let h = 0; h < c.count; h++) {
                this.f.k(d._inner[h]);
                this.f.k(e + ".push(" + c._inner[h].t + ")");
            }
            this.f.l("this._" + CodeGeneratingComponentRenderer.d4(e) + " = " + e + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.d4(e) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.d5(e);
        }
        constructor(a, b) {
            super(a, b);
            this.br = 0;
            this.bp = new HashSet$1(String_$type, 0);
            this.bq = new HashSet$1(String_$type, 0);
        }
        aq(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        get_w() {
            return true;
        }
        a8(a, b, c) {
            super.a8(a, b, c);
            this.h.l("<" + this.bs(a) + "");
            if (!this.bp.contains(this.am)) {
                this.bp.add_1(this.am);
                if (!this.bq.contains(this.an)) {
                    this.bq.add_1(this.an);
                    this.a7(this.an);
                }
                this.f.l("private " + CodeGeneratingComponentRenderer.d4(this.am) + ": " + this.an);
                if (!this.p) {
                    if (this.u) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + " = document.querySelector('" + CodeGeneratingComponentRenderer.d7(stringReplace(this.an, "Component", "")) + "') as " + this.an + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + " = document.getElementById('" + this.am + "') as " + this.an + ";");
                }
            }
        }
        bs(a) {
            if (stringEndsWith(a, "Component")) {
                a = a.substr(0, a.length - ("Component").length);
            }
            return CodeGeneratingComponentRenderer.d7(a);
        }
        bm() {
            this.h.l(">");
            this.h.h();
        }
        bh(a) {
            this.h.f();
            this.h.l("</" + this.bs(a) + ">");
            super.bh(a);
        }
        get_a4() {
            return "Igc";
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            WebImportsHalper.a(this.k, this.a3, this.a2, this.a4, a, b, c);
        }
        be(a, b, c, d) {
            let e = CodeGeneratingComponentRenderer.d4(b);
            if (this.q(d, c)) {
                this.bn(c);
                if (!this.bp.contains(this.am)) {
                    this.bp.add_1(this.am);
                    if (!this.bq.contains(this.an)) {
                        this.bq.add_1(this.an);
                        this.a7(this.an);
                    }
                    this.f.l("private " + CodeGeneratingComponentRenderer.d4(this.am) + ": " + this.an);
                    if (!this.p) {
                        if (this.u) {
                            this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + " = document.querySelector('" + CodeGeneratingComponentRenderer.d7(stringReplace(this.an, "Component", "")) + "') as " + this.an + ";");
                        }
                    }
                    else {
                        this.g.l("var " + CodeGeneratingComponentRenderer.d4(this.am) + " = this." + CodeGeneratingComponentRenderer.d4(this.am) + " = document.getElementById('" + this.am + "') as " + this.an + ";");
                    }
                }
                let f = CodeGeneratingComponentRenderer.d4(this.a0(d, c, true));
                let g = f;
                if (g != "null" && !stringStartsWith(g, "this.")) {
                    g = "this." + g;
                }
                if (c.b.isCustomEvent) {
                    this.e.l(CodeGeneratingComponentRenderer.d4(this.am) + ".addEventListener(\"" + e + "\", " + g + ");");
                }
                else {
                    this.e.l(CodeGeneratingComponentRenderer.d4(this.am) + "." + e + " = " + g + ";");
                }
                return;
            }
            if (a > 0) {
                this.h.l("");
            }
            this.h.j(CodeGeneratingComponentRenderer.d7(e));
            this.h.j("=");
            this.h.j("\"");
            this.h.j(this.a0(d, c, true));
            this.h.j("\"");
            if (e == "name") {
                this.h.l("");
                this.h.j("id=\"" + this.a0(d, c, true) + "\"");
            }
        }
        ad(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        ax() {
            return "null";
        }
        as(a) {
            return CodeGeneratingComponentRenderer.d4(a.a.e);
        }
        ay(a) {
            return CodeGeneratingComponentRenderer.d4(a.a.e);
        }
        au(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.bn(a);
                if (!this.bp.contains(a.a.e)) {
                    this.g.l("this." + CodeGeneratingComponentRenderer.d4(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.d4(a.a.e) + ".bind(this);");
                }
                return "this." + CodeGeneratingComponentRenderer.d4(a.a.e);
            }
            return this.ax();
        }
        bf(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.u(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.u(this, false);
                }
            }
        }
    }
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter, 'WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter extends WebCodeBehindCodeGeneratingComponentRendererCodeEmitter {
        i(a, b) {
            return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
        }
        get_a4() {
            return "Igc";
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            WebImportsHalper.a(this.k, this.a3, this.a2, this.a4, a, b, c);
        }
    }
    WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let TSCodeGeneratingComponentRendererHandlerEmitter = /*@__PURE__*/ (() => {
    class TSCodeGeneratingComponentRendererHandlerEmitter extends CodeGeneratingComponentRendererHandlerEmitter {
        constructor(a, b, c) {
            super(a, b);
            this.w = 0;
            this.w = c;
        }
        t(a, b) {
            super.t(a, b);
            this.g.l("this." + CodeGeneratingComponentRenderer.d4(a) + "();");
        }
        u(a, b) {
            super.u(a, b);
            this.h.l("this." + CodeGeneratingComponentRenderer.d4(a) + "();");
        }
        get x() {
            switch (this.w) {
                case 1: return "angular";
                case 2: return "react";
                case 3: return "webcomponents";
            }
            return "";
        }
        get y() {
            switch (this.w) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
            }
            return "";
        }
        v(a, b) {
            super.v(a, b);
            let c = new CodeGeneratingRegexHelper("import {([^}]+)}\\s+from\\s+['\"]([^'\"]+)['\"]");
            c.execute(b, (d, e, f) => {
                let g = d(1);
                let h = d(2);
                h = stringReplace(h, "webcomponents", this.x);
                h = stringReplace(h, "react", this.x);
                h = stringReplace(h, "angular", this.x);
                let i = g.split(',');
                for (let k = 0; k < i.length; k++) {
                    let j = i[k];
                    let l = j.trim();
                    if (stringStartsWith(l, "Igc")) {
                        l = this.y + l.substr(3);
                    }
                    if (stringStartsWith(l, "Igr")) {
                        l = this.y + l.substr(3);
                    }
                    if (stringStartsWith(l, "Igx")) {
                        l = this.y + l.substr(3);
                    }
                    if (stringStartsWith(l, "Igr") && stringEndsWith(l, "Component")) {
                        l = l.substr(0, l.length - ("Component").length);
                    }
                    a.addPackageImport(h.trim(), l.trim());
                }
                return null;
            });
        }
        q(a) {
            return a;
        }
    }
    TSCodeGeneratingComponentRendererHandlerEmitter.$t = /*@__PURE__*/ markType(TSCodeGeneratingComponentRendererHandlerEmitter, 'TSCodeGeneratingComponentRendererHandlerEmitter', CodeGeneratingComponentRendererHandlerEmitter.$);
    return TSCodeGeneratingComponentRendererHandlerEmitter;
})();
/**
 * @hidden
 */
export let TSCodeGeneratingComponentRendererDataEmitter = /*@__PURE__*/ (() => {
    class TSCodeGeneratingComponentRendererDataEmitter extends CodeGeneratingComponentRendererDataEmitter {
        constructor(a, b) {
            super(a, b);
            this.al = new Dictionary$2(CodeGeneratingCodeWriter.$, Number_$type, 0);
        }
        w(a, b = false) {
            a = CodeGeneratingComponentRenderer.d5(a);
            a = stringReplace(a, "-", "_");
            if (b) {
                a += "Item";
            }
            return a;
        }
        v(a) {
            a = stringReplace(a, "-", "_");
            if (this.p.e) {
                return a;
            }
            return CodeGeneratingComponentRenderer.d4(a);
        }
        ap(a) {
            a = stringReplace(a, "-", "_");
            return "_" + CodeGeneratingComponentRenderer.d4(a);
        }
        get_s() {
            return ".ts";
        }
        ak(a, b, c, d, e) {
            super.ak(a, b, c, d, e);
            let f = this.r(this.w(a, true), this.w(a, false), b, c);
            for (let g of fromEnum(f)) {
                this.ac(g, d, e);
            }
            let h = 0;
            if (e == this.k) {
                e.addPackageImport("./" + this.w(a, false), this.w(a, false));
            }
            else {
                e.addPackageImport("./" + this.w(a, false), this.w(a, false));
            }
            d.l("export class " + this.w(a, false) + " extends Array<" + this.w(a, true) + "> {");
            d.h();
            d.l("public constructor() {");
            d.h();
            d.l("super();");
        }
        ag(a, b, c) {
            super.ag(a, b, c);
            let d = this.p.cDNBasePath + this.w(a, false) + ".json";
            let e = "";
            if (this.n.platform == 2) {
                e = "this.setState({}); ";
            }
            if (this.n.platform == 3) {
                e = "if ((this as any)._bind) { (this as any)._bind(); } ";
            }
            if (this.n.platform == 1) {
                e = "this._detector.markForCheck(); ";
            }
            let f = this.w(a, false);
            let g = "null";
            switch (b) {
                case 2:
                    f = "any[]";
                    break;
                case 1:
                    f = "any[]";
                    g = this.w(a, false);
                    break;
                case 0: break;
            }
            this.g.l("private " + this.ap(a) + ": " + f + " = " + g + ";");
            if (b == 2 || (c != null && c.isAsync)) {
                this.g.l("private " + this.ap(a) + "Fetching: boolean = false;");
            }
            this.g.l("public get " + CodeGeneratingComponentRenderer.d4(a) + "(): " + f + " {");
            this.g.h();
            if (c != null && c.isAsync) {
                this.g.l("if (this." + this.ap(a) + " == null && !this." + this.ap(a) + "Fetching)");
                this.g.l("{");
                this.g.h();
                this.g.l("this." + this.ap(a) + "Fetching = true;");
                this.g.l("( async () => { this." + this.ap(a) + " = await (await " + this.w(a, false) + ".fetch()); " + e + " })();");
                this.g.f();
                this.g.l("}");
            }
            else {
                if (b == 2) {
                    this.g.l("if (this." + this.ap(a) + " == null && !this." + this.ap(a) + "Fetching)");
                    this.g.l("{");
                    this.g.h();
                    this.g.l(this.ap(a) + "Fetching = true;");
                    this.g.l("( async () => { this." + this.ap(a) + " = await (await fetch('" + d + "')).json(); " + e + " })();");
                    this.g.f();
                    this.g.l("}");
                }
                if (b == 0) {
                    this.g.l("if (this." + this.ap(a) + " == null)");
                    this.g.l("{");
                    this.g.h();
                    this.g.l("this." + this.ap(a) + " = new " + this.w(a, false) + "();");
                    this.g.f();
                    this.g.l("}");
                }
            }
            this.g.l("return this." + this.ap(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        ah(a, b) {
            super.ah(a, b);
            this.g.l("private " + this.ap(a) + ": " + this.am(b.e) + " = " + this.an(b) + ";");
            this.g.l("public get " + CodeGeneratingComponentRenderer.d4(a) + "(): " + this.am(b.e) + " {");
            this.g.h();
            this.g.l("return this." + this.ap(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        ai(a, b, c, d) {
            super.ai(a, b, c, d);
            this.g.l("private " + this.ap(a) + ": " + this.w(a, false) + "Item = null;");
            this.g.l("public get " + CodeGeneratingComponentRenderer.d4(a) + "(): " + this.w(a, false) + "Item {");
            this.g.h();
            this.g.l("if (this." + this.ap(a) + " == null)");
            this.g.l("{");
            this.g.h();
            this.g.l("this." + this.ap(a) + " = ");
            this.aq(this.w(a, true), b, c, d, this.g);
            this.g.l("}");
            this.g.l("return this." + this.ap(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        y(a, b) {
            b.addPackageImport("./" + a, a);
        }
        z(a, b, c) {
            c.addDefaultImport("./" + b, a);
        }
        aa(a, b) {
            this.l = CodeGeneratingImportManager.merge(this.l, b);
        }
        ac(a, b, c) {
            let d = 0;
            if (c == this.k) {
                c.addPackageImport("./" + a.c, a.d);
            }
            else {
                c.addPackageImport("./" + a.c, a.d);
            }
            b.l("export class " + a.d + " {");
            b.h();
            b.l("public constructor(init: Partial<" + a.d + ">) {");
            b.h();
            b.l("Object.assign(this, init);");
            b.f();
            b.l("}");
            b.l("");
            for (let e = 0; e < a.a.length; e++) {
                let f = a.a[e];
                let g = a.b[e];
                b.l("public " + this.v(f) + ": " + this.ao(a, f, g) + ";");
            }
            b.f();
            b.l("");
            b.l("}");
        }
        am(a) {
            switch (a) {
                case 1: return "boolean";
                case 0: return "number";
                case 2: return "string";
                case 3: return "any";
            }
            return "any";
        }
        ao(a, b, c) {
            if (c.d) {
                return a.d + "_" + CodeGeneratingComponentRenderer.d5(this.v(b)) + "Item[]";
            }
            else if (c.f) {
                return a.d + "_" + this.v(b);
            }
            else {
                switch (c.c) {
                    case 1: return "boolean";
                    case 0: return "number";
                    case 2: return "string";
                    case 3: return "any";
                }
            }
            return "any";
        }
        aj(a, b, c, d) {
            super.aj(a, b, c, d);
            d.f();
            d.l("}");
            d.f();
            d.l("}");
        }
        x(a, b, c, d, e, f) {
            super.x(a, b, c, d, e, f);
            f.j("this.push(");
            this.aq(this.w(a, true), b, c, d, f);
            f.l(");");
        }
        aq(a, b, c, d, e) {
            if (d == null) {
                e.j("null");
                return;
            }
            if (typeCast(JsonDictionaryArray.$, d) !== null) {
                let f = d;
                e.l("[");
                e.h();
                if (f.items != null && f.items.length > 0) {
                    for (let g = 0; g < f.items.length; g++) {
                        if (g > 0) {
                            e.l(",");
                        }
                        this.aq(a, b, c, f.items[g], e);
                    }
                }
                e.f();
                e.l("]");
            }
            else if (typeCast(JsonDictionaryObject.$, d) !== null) {
                let h = d;
                e.l("new " + a + "(");
                e.l("{");
                e.h();
                for (let i = 0; i < b.length; i++) {
                    let j = b[i];
                    let k = c[i];
                    if (i > 0) {
                        e.l(",");
                    }
                    e.j(this.v(j) + ": ");
                    if (k.d) {
                        let l = a + "_" + CodeGeneratingComponentRenderer.d5(this.v(j)) + "Item";
                        this.aq(l, k.a, k.b, h.item(j), e);
                    }
                    else if (k.f) {
                        let m = a + "_" + CodeGeneratingComponentRenderer.d5(this.v(j));
                        this.aq(m, k.a, k.b, h.item(j), e);
                    }
                    else {
                        let n = h.item(j);
                        e.j(this.t(k, n));
                    }
                }
                e.l("");
                e.f();
                e.j("})");
            }
        }
        t(a, b) {
            if (b.value == null) {
                return ("null");
            }
            switch (a.c) {
                case 1: return (b.value ? "true" : "false");
                case 0: return (b.value.toString());
                case 2: return ("`" + b.value + "`");
                case 3: return ("null");
            }
            return "";
        }
        an(a) {
            switch (a.e) {
                case 1: return (a.value ? "true" : "false");
                case 0: return (a.value.toString());
                case 2: return ("`" + a.value + "`");
                case 3: return "null";
            }
            return "";
        }
    }
    TSCodeGeneratingComponentRendererDataEmitter.$t = /*@__PURE__*/ markType(TSCodeGeneratingComponentRendererDataEmitter, 'TSCodeGeneratingComponentRendererDataEmitter', CodeGeneratingComponentRendererDataEmitter.$);
    return TSCodeGeneratingComponentRendererDataEmitter;
})();
/**
 * @hidden
 */
export let AngularTemplateCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class AngularTemplateCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        ao(a, b, c, d) {
            d = super.ao(a, b, c, d);
            let e = b.q();
            this.f.l("private " + " _" + CodeGeneratingComponentRenderer.d4(d) + ": " + e + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.d4(d) + "(): " + e + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.d4(d) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k(c);
            this.f.l("this._" + CodeGeneratingComponentRenderer.d4(d) + " = " + d + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.d4(d) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.d5(d);
        }
        ap(a, b, c, d, e) {
            e = super.ap(a, b, c, d, e);
            let f = null;
            if (c.count > 0) {
                f = c._inner[0].q() + "[]";
            }
            else {
                let g = a.b.specificExternalType;
                f = CodeGenerationItemBuilder.n(this.k.platform, g, b, this.ai, this.aj) + "[]";
            }
            this.f.l("private" + " _" + CodeGeneratingComponentRenderer.d4(e) + ": " + f + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.d4(e) + "(): " + f + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.d4(e) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k("let " + CodeGeneratingComponentRenderer.d4(e) + ": " + f + " = [];");
            for (let h = 0; h < c.count; h++) {
                this.f.k(d._inner[h]);
                this.f.k(e + ".push(" + c._inner[h].t + ")");
            }
            this.f.l("this._" + CodeGeneratingComponentRenderer.d4(e) + " = " + e + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.d4(e) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.d5(e);
        }
        i(a, b) {
            return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
            this.bq = 0;
            this.bo = new HashSet$1(String_$type, 0);
            this.bp = new HashSet$1(String_$type, 0);
        }
        aq(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        get_w() {
            return true;
        }
        get_a4() {
            return "Igx";
        }
        a8(a, b, c) {
            super.a8(a, b, c);
            this.h.l("<" + this.br(a) + "");
            if (!this.bo.contains(this.am)) {
                this.bo.add_1(this.am);
                if (!this.bp.contains(this.an)) {
                    this.bp.add_1(this.an);
                    this.a7(this.an);
                }
                this.f.l("@ViewChild(\"" + CodeGeneratingComponentRenderer.d4(this.am) + "\", { static: true } )");
                this.f.l("private " + CodeGeneratingComponentRenderer.d4(this.am) + ": " + this.an);
            }
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            WebImportsHalper.a(this.k, this.a3, this.a2, this.a4, a, b, c);
        }
        br(a) {
            if (stringEndsWith(a, "Component")) {
                a = a.substr(0, a.length - ("Component").length);
            }
            return CodeGeneratingComponentRenderer.d7(a);
        }
        bm() {
            this.h.l(">");
            this.h.h();
        }
        bh(a) {
            this.h.f();
            this.h.l("</" + this.br(a) + ">");
            super.bh(a);
        }
        be(a, b, c, d) {
            let e = CodeGeneratingComponentRenderer.d4(b);
            if (this.q(d, c)) {
                this.bn(c);
                if (this.t(c)) {
                    e = "(" + e + ")";
                }
                else {
                    e = "[" + e + "]";
                }
            }
            if (a > 0) {
                this.h.l("");
            }
            this.h.j(e);
            this.h.j("=");
            this.h.j("\"");
            this.h.j(this.a0(d, c, true));
            this.h.j("\"");
            if (e == "name") {
                this.h.l("");
                this.h.j("#" + CodeGeneratingComponentRenderer.d4(this.a0(d, c, true)));
            }
        }
        au(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.bn(a);
                return "this." + CodeGeneratingComponentRenderer.d4(a.a.e) + "($event)";
            }
            return this.ax();
        }
        ad(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        ax() {
            return "null";
        }
        as(a) {
            return CodeGeneratingComponentRenderer.d4(a.a.e);
        }
        ay(a) {
            return CodeGeneratingComponentRenderer.d4(a.a.e);
        }
        bf(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.u(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.u(this, false);
                }
            }
        }
    }
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(AngularTemplateCodeGeneratingComponentRendererCodeEmitter, 'AngularTemplateCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return AngularTemplateCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter extends WebCodeBehindCodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        i(a, b) {
            return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
        }
        get_a4() {
            return "Igc";
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            WebImportsHalper.a(this.k, this.a3, this.a2, this.a4, a, b, c);
        }
    }
    AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let DotNetCodeGeneratingComponentRendererDataEmitter = /*@__PURE__*/ (() => {
    class DotNetCodeGeneratingComponentRendererDataEmitter extends CodeGeneratingComponentRendererDataEmitter {
        constructor(a, b) {
            super(a, b);
            this.al = false;
        }
        aa(a, b) {
            super.aa(a, b);
            this.f.l("<Compile Include=\"" + a + ".cs\">");
            this.f.h();
            this.f.l("<SubType>Code</SubType>");
            this.f.f();
            this.f.l("</Compile>");
        }
        ae(a) {
            a.l("using System;");
            a.l("using System.Collections.Generic;");
            if (!this.al) {
                this.al = true;
            }
            super.ae(a);
        }
        ad(a) {
            super.ad(a);
        }
        w(a, b = false) {
            a = CodeGeneratingComponentRenderer.d5(a);
            a = stringReplace(a, "-", "_");
            if (b) {
                a += "Item";
            }
            return a;
        }
        v(a) {
            a = stringReplace(a, "-", "_");
            return CodeGeneratingComponentRenderer.d5(a);
        }
        ap(a) {
            a = stringReplace(a, "-", "_");
            return "_" + CodeGeneratingComponentRenderer.d4(a);
        }
        ak(a, b, c, d, e) {
            super.ak(a, b, c, d, e);
            let f = this.r(this.w(a, true), this.w(a, false), b, c);
            for (let g of fromEnum(f)) {
                this.ac(g, d, e);
            }
            d.l("");
            d.l("public class " + this.w(a, false));
            d.h();
            d.l(": List<" + this.w(a, true) + ">");
            d.f();
            d.l("{");
            d.h();
            d.l("public " + this.w(a, false) + "()");
            d.l("{");
            d.h();
        }
        ag(a, b, c) {
            super.ag(a, b, c);
            let d = this.ap(a) + " = " + "new " + this.w(a, false) + "(); ";
            let e = "";
            if (this.n.platform == 5) {
                e = " StateHasChanged();";
            }
            else {
                e = " OnPropertyChanged(\"" + this.w(a, false) + "\");";
            }
            if (c != null && c.isAsync) {
                d = this.w(a, false) + ".Fetch().ContinueWith((t) => {" + this.ap(a) + " = t.Result; " + e + " });";
            }
            this.g.l("private " + this.w(a, false) + " " + this.ap(a) + " = null;");
            this.g.l("public " + this.w(a, false) + " " + this.v(a));
            this.g.l("{");
            this.g.h();
            this.g.l("get");
            this.g.l("{");
            this.g.h();
            this.g.l("if (" + this.ap(a) + " == null)");
            this.g.l("{");
            this.g.h();
            this.g.l(d);
            this.g.f();
            this.g.l("}");
            this.g.l("return " + this.ap(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        ah(a, b) {
            super.ah(a, b);
            this.g.l("private " + this.am(b.e) + " " + this.ap(a) + " = " + this.an(b) + ";");
            this.g.l("public " + this.am(b.e) + " " + this.v(a));
            this.g.l("{");
            this.g.h();
            this.g.l("get");
            this.g.l("{");
            this.g.h();
            this.g.l("return " + this.ap(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        ai(a, b, c, d) {
            super.ai(a, b, c, d);
            this.g.l("private " + this.w(a, false) + "Item " + this.ap(a) + " = null;");
            this.g.l("public " + this.w(a, false) + "Item " + this.v(a));
            this.g.l("{");
            this.g.h();
            this.g.l("get");
            this.g.l("{");
            this.g.h();
            this.g.l("if (" + this.ap(a) + " == null)");
            this.g.l("{");
            this.g.h();
            this.g.l("_" + a + " = ");
            this.aq(this.w(a, true), b, c, d, this.g);
            this.g.f();
            this.g.l("}");
            this.g.l("return " + this.ap(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        get_s() {
            return ".cs";
        }
        ac(a, b, c) {
            b.l("public class " + a.d);
            b.l("{");
            b.h();
            for (let d = 0; d < a.a.length; d++) {
                let e = a.a[d];
                let f = a.b[d];
                b.l("public " + this.ao(a, e, f) + " " + this.v(e) + " { get; set; }");
            }
            b.f();
            b.l("}");
        }
        am(a) {
            switch (a) {
                case 1: return "bool";
                case 0: return "double";
                case 2: return "string";
                case 3: return "object";
            }
            return "object";
        }
        ao(a, b, c) {
            if (c.d) {
                return "List<" + a.d + "_" + this.v(b) + "Item>";
            }
            else if (c.f) {
                return a.d + "_" + this.v(b);
            }
            else {
                switch (c.c) {
                    case 1: return "bool";
                    case 0: return "double";
                    case 2: return "string";
                    case 3: return "object";
                }
            }
            return "object";
        }
        aj(a, b, c, d) {
            super.aj(a, b, c, d);
            d.f();
            d.l("}");
            d.f();
            d.l("}");
        }
        x(a, b, c, d, e, f) {
            super.x(a, b, c, d, e, f);
            f.j("this.Add(");
            this.aq(this.w(a, true), b, c, d, f);
            f.l(");");
        }
        aq(a, b, c, d, e) {
            if (d == null) {
                e.j("null");
                return;
            }
            if (typeCast(JsonDictionaryArray.$, d) !== null) {
                let f = d;
                e.l("new List<" + a + ">()");
                e.l("{");
                e.h();
                if (f.items != null && f.items.length > 0) {
                    for (let g = 0; g < f.items.length; g++) {
                        if (g > 0) {
                            e.l(",");
                        }
                        this.aq(a, b, c, f.items[g], e);
                    }
                }
                e.f();
                e.l("}");
            }
            else if (typeCast(JsonDictionaryObject.$, d) !== null) {
                let h = d;
                e.l("new " + a + "()");
                e.l("{");
                e.h();
                for (let i = 0; i < b.length; i++) {
                    let j = b[i];
                    let k = c[i];
                    if (!h.g(j)) {
                        continue;
                    }
                    if (i > 0) {
                        e.l(",");
                    }
                    e.j(j + " = ");
                    if (k.d) {
                        let l = a + "_" + this.v(j) + "Item";
                        this.aq(l, k.a, k.b, h.item(j), e);
                    }
                    else if (k.f) {
                        let m = a + "_" + this.v(j);
                        this.aq(m, k.a, k.b, h.item(j), e);
                    }
                    else {
                        let n = h.item(j);
                        e.j(this.t(k, n));
                    }
                }
                e.l("");
                e.f();
                e.j("}");
            }
        }
        t(a, b) {
            switch (a.c) {
                case 1: return (b.value ? "true" : "false");
                case 0:
                    if (b == null || b.value == null) {
                        return "double.NaN";
                    }
                    return (b.value.toString());
                case 2: if (b.value == null) {
                    return "null";
                }
                else {
                    return ("@\"" + stringReplace(b.value, "\"", "\"\"") + "\"");
                }
                case 3: return ("null");
            }
            return "";
        }
        an(a) {
            switch (a.e) {
                case 1: return (a.value ? "true" : "false");
                case 0: return (a.value.toString());
                case 2: return ("@\"" + stringReplace(a.value, "\"", "\"\"") + "\"");
                case 3: return "null";
            }
            return "";
        }
    }
    DotNetCodeGeneratingComponentRendererDataEmitter.$t = /*@__PURE__*/ markType(DotNetCodeGeneratingComponentRendererDataEmitter, 'DotNetCodeGeneratingComponentRendererDataEmitter', CodeGeneratingComponentRendererDataEmitter.$);
    return DotNetCodeGeneratingComponentRendererDataEmitter;
})();
/**
 * @hidden
 */
export let WPFXamlCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WPFXamlCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        i(a, b) {
            return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        get_x() {
            return false;
        }
        constructor(a, b) {
            super(a, b);
        }
        aq(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        a8(a, b, c) {
            super.a8(a, b, c);
            this.h.l("<ig:" + a + "");
        }
        bm() {
            this.h.l(">");
            this.h.h();
        }
        bh(a) {
            this.h.f();
            this.h.l("</ig:" + a + ">");
            super.bh(a);
        }
        be(a, b, c, d) {
            if (b == "Label" && c.b != null && stringContains(c.b.owningType, "Axis")) {
                if (c.c != null && stringContains(c.c, "{")) {
                    if (a > 0) {
                        this.h.l("");
                    }
                    this.h.j(b);
                    this.h.j("=");
                    this.h.j("\"");
                    this.h.j("{}" + this.a0(d, c, true));
                    this.h.j("\"");
                    return;
                }
            }
            if (a > 0) {
                this.h.l("");
            }
            this.h.j(b);
            this.h.j("=");
            this.h.j("\"");
            this.h.j(this.a0(d, c, true));
            this.h.j("\"");
        }
        ad(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        ax() {
            return "{x:Null}";
        }
        as(a) {
            return "{Binding ElementName=" + a.a.e + "}";
        }
        bo(a) {
            a = stringReplace(a, "-", "_");
            return CodeGeneratingComponentRenderer.d5(a);
        }
        ay(a) {
            this.bn(a);
            return "{Binding " + this.bo(a.a.e) + "}";
        }
        bf(a, b, c, d) {
            this.h.j("<ig:");
            {
                this.h.j(this.an);
                this.h.j(".");
            }
            this.h.j(b);
            this.h.l(">");
            this.h.h();
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.u(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.u(this, false);
                }
            }
            this.h.f();
            this.h.j("</ig:");
            {
                this.h.j(this.an);
                this.h.j(".");
            }
            this.h.j(b);
            this.h.l(">");
        }
        get_a4() {
            return "Xam";
        }
    }
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WPFXamlCodeGeneratingComponentRendererCodeEmitter, 'WPFXamlCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WPFXamlCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        get r() {
            return this.get_r();
        }
        i(a, b) {
            return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        get_x() {
            return false;
        }
        constructor(a, b) {
            super(a, b);
        }
    }
    WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let CodeGeneratingComponentRenderer = /*@__PURE__*/ (() => {
    class CodeGeneratingComponentRenderer extends ComponentRenderer {
        loadCodeJson(a) {
            this.loadJson(a, (b) => ((() => {
                let $ret = new CodeGenerationContainer();
                $ret.id = b;
                return $ret;
            })()));
        }
        cw(a) {
            this.dw.b = a;
        }
        n() {
            return false;
        }
        static d4(a) {
            if (a == null) {
                return null;
            }
            if (a.length == 1) {
                return a.toLowerCase();
            }
            return a.substr(0, 1).toLowerCase() + a.substr(1);
        }
        static d6(a) {
            if (a == null) {
                return null;
            }
            if (a.length == 1) {
                return a.toLowerCase();
            }
            if (Base.equalsStatic(a.toUpperCase(), a)) {
                return a;
            }
            return a.substr(0, 1).toLowerCase() + a.substr(1);
        }
        static d5(a) {
            if (a == null) {
                return null;
            }
            if (a.length == 1) {
                return a.toUpperCase();
            }
            return a.substr(0, 1).toUpperCase() + a.substr(1);
        }
        static d7(a) {
            if (a == null) {
                return null;
            }
            let b = new List$1(String_$type, 0);
            let c = 0;
            for (let d = 0; d < a.length; d++) {
                let e = a.charAt(d);
                let f = e.toUpperCase();
                let g = e.toLowerCase();
                let h = (c == 0 && b.count > 0);
                let i = (c > 1);
                if (f == e) {
                    if (h) {
                        b.add('-');
                    }
                    c++;
                }
                else if (g == e) {
                    if (i) {
                        b.insert(b.count - 1, '-');
                    }
                    c = 0;
                }
                else {
                    c = 0;
                }
                b.add(g);
            }
            let j = new StringBuilder(0);
            for (let k = 0; k < b.count; k++) {
                j.h(b._inner[k]);
            }
            return j.toString();
        }
        aq() {
            return true;
        }
        ad(a) {
            super.ad(a);
            this.d0 = a;
            return true;
        }
        ak(a) {
            super.ak(a);
            this.d2 = a;
            return true;
        }
        ae(a) {
            super.ae(a);
            this.dy = a;
            return true;
        }
        af(a) {
            super.af(a);
            this.dz = a;
            return true;
        }
        ag(a) {
            super.ag(a);
            this.d1 = a;
            return true;
        }
        dq(a, b, c) {
            switch (this.dx) {
                case 1: if (b.forceCodeBehind) {
                    return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new AngularTemplateCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                case 2: if (b.forceCodeBehind) {
                    return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new ReactMarkupCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                case 3: if (b.forceCodeBehind) {
                    return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                case 4: break;
                case 0: if (b.forceCodeBehind) {
                    return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new WPFXamlCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                case 5: if (b.forceCodeBehind) {
                    return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new BlazorRazorCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
            }
            return null;
        }
        dr(a, b) {
            switch (this.dx) {
                case 1: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
                case 2: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
                case 3: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
                case 4: break;
                case 0:
                case 5: return new DotNetCodeGeneratingComponentRendererDataEmitter(a, b);
            }
            return null;
        }
        ds(a, b) {
            switch (this.dx) {
                case 1: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.dx);
                case 2: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.dx);
                case 3: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.dx);
                case 4: break;
                case 0:
                case 5: return new DotNetCodeGeneratingComponentRendererHandlerEmitter(a, b);
            }
            return null;
        }
        du(a, b) {
            switch (this.dx) {
                case 1: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.dx);
                case 2: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.dx);
                case 3: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.dx);
                case 4: break;
                case 0:
                case 5: return new DotNetCodeGeneratingComponentRendererTemplateEmitter(a, b);
            }
            return null;
        }
        dt(a, b) {
            switch (this.dx) {
                case 1: return new TSCodeGeneratingComponentRendererModuleEmitter(this.dx, a, b);
                case 2: return new TSCodeGeneratingComponentRendererModuleEmitter(this.dx, a, b);
                case 3: return new TSCodeGeneratingComponentRendererModuleEmitter(this.dx, a, b);
                case 4: break;
                case 0: return new DotNetCodeGeneratingComponentRendererModuleEmitter(this.dx, a, b);
                case 5: return new BlazorCodeGeneratingComponentRendererModuleEmitter(this.dx, a, b);
            }
            return null;
        }
        emitCode(a) {
            let b = new CodeGeneratingRendererResult();
            b.platform = this.dx;
            b.generationOptions = this.dw;
            for (let c of fromEnum(this.a6.keys)) {
                let d = c;
                let e = d.id;
                for (let f of fromEnum(d.a)) {
                    if (f.i("Name") != null) {
                        b.trackDescriptionField(e, f.i("Name"));
                    }
                    if (f.i("name") != null) {
                        b.trackDescriptionField(e, f.i("name"));
                    }
                }
            }
            for (let g of fromEnum(this.a6.keys)) {
                let h = g;
                h.c(this.context, b, this, this.dw);
            }
            if (this.d0 != null) {
                let i = this.dt(this.dw, b);
                i.n(this.d0, this.context);
            }
            if (this.d2 != null) {
                if (typeCast(JsonDictionaryObject.$, this.d2) !== null) {
                    let j = this.d2;
                    let m = j.e();
                    for (let l = 0; l < m.length; l++) {
                        let k = m[l];
                        b.defineString(k, j.item(k).value);
                    }
                }
            }
            {
                let n = this.dr(this.dw, b);
                n.ab(this.d1);
            }
            let o = this.ds(this.dw, b);
            let p = new List$1(String_$type, 0);
            let q = new List$1(String_$type, 0);
            if (this.dy != null) {
                if (typeCast(JsonDictionaryArray.$, this.dy) !== null) {
                    let r = this.dy;
                    if (r.items != null) {
                        let u = r.items;
                        for (let t = 0; t < u.length; t++) {
                            let s = u[t];
                            p.add(s.value.toString());
                        }
                    }
                }
                else if (typeCast(JsonDictionaryValue.$, this.dy) !== null) {
                    let v = this.dy;
                    p.add(v.value.toString());
                }
            }
            if (this.dz != null) {
                if (typeCast(JsonDictionaryArray.$, this.dz) !== null) {
                    let w = this.dz;
                    if (w.items != null) {
                        let z = w.items;
                        for (let y = 0; y < z.length; y++) {
                            let x = z[y];
                            p.add(x.value.toString());
                        }
                    }
                }
                else if (typeCast(JsonDictionaryValue.$, this.dz) !== null) {
                    let aa = this.dz;
                    p.add(aa.value.toString());
                }
            }
            o.s(p.toArray(), q.toArray());
            let ab = this.du(this.dw, b);
            ab.p();
            this.eb(b, a);
            return b;
        }
        eb(a, b) {
            b.execute(a);
        }
        bb() {
            return runOn(this, this.d9);
        }
        d8(a) {
            a = stringReplace(a, "-", "_");
            if (!this.dw.e) {
                switch (this.dx) {
                    case 1:
                    case 2:
                    case 3: return CodeGeneratingComponentRenderer.d4(a);
                    case 4:
                    case 0:
                    case 5: return CodeGeneratingComponentRenderer.d5(a);
                }
            }
            return a;
        }
        d9(a) {
            let b = a;
            if (stringStartsWith(a, "{")) {
                if (!stringStartsWith(a, "{[")) {
                    a = a.substr(1, a.length - 2);
                    a = "{" + this.d8(a) + "}";
                    b = a;
                }
            }
            else {
                b = this.d8(a);
            }
            return b;
        }
        constructor(a, b) {
            super();
            this.dw = null;
            this.dx = 0;
            this.d0 = null;
            this.d1 = null;
            this.d2 = null;
            this.dz = null;
            this.dy = null;
            this.dw = b;
            this.dx = a;
            this.preserveKeyOrder = true;
            ComponentRenderer.platform = CodeGeneratingComponentRenderer.d3(a);
            this.adapter = new CodeGenerationRendererAdapter(a, this.adapter);
            this.context = new TypeDescriptionContext(this.adapter, ComponentRenderer.platform);
            this.i = new CodeGenerationTransformer(a);
        }
        static d3(a) {
            switch (a) {
                case 1: return 3;
                case 2: return 2;
                case 3: return 4;
                case 4: return 8;
                case 0: return 0;
                case 5: return 12;
            }
            return 0;
        }
    }
    CodeGeneratingComponentRenderer.$t = /*@__PURE__*/ markType(CodeGeneratingComponentRenderer, 'CodeGeneratingComponentRenderer', ComponentRenderer.$);
    return CodeGeneratingComponentRenderer;
})();
/**
 * @hidden
 */
export let CodeGenerationContainer = /*@__PURE__*/ (() => {
    class CodeGenerationContainer extends Base {
        get id() {
            return this._id;
        }
        set id(a) {
            this._id = a;
        }
        constructor() {
            super();
            this._id = null;
            this.a = null;
            this.a = new List$1(CodeGenerationItemBuilder.$, 0);
        }
        c(a, b, c, d) {
            let e = new StringBuilder(0);
            let f = new StringBuilder(0);
            let g = new StringBuilder(0);
            let h = new CodeGeneratingImportManager();
            for (let i of fromEnum(this.a)) {
                let j = c.dq(i, d, b);
                i.u(j, true);
                let k = j.toString();
                b.addContainerResult(this.id, k);
                j.ba(e);
                j.bd(f);
                j.a9(g);
                j.bb(a, h);
            }
            b.addContainerResult("bindingFields", e.toString());
            b.addContainerResult("bindingInit", f.toString());
            b.addContainerResult("bindingCode", g.toString());
            b.addImportResult("bindingImports", h);
        }
    }
    CodeGenerationContainer.$t = /*@__PURE__*/ markType(CodeGenerationContainer, 'CodeGenerationContainer');
    return CodeGenerationContainer;
})();
/**
 * @hidden
 */
export let CodeGenerationItemBuilder = /*@__PURE__*/ (() => {
    class CodeGenerationItemBuilder extends Base {
        o() {
            let a = this;
            let b = null;
            if (a.i("Name") != null) {
                b = a.i("Name");
            }
            if (a.i("name") != null) {
                b = a.i("name");
            }
            if (b != null) {
                return b;
            }
            return null;
        }
        constructor(a, b, c, d) {
            super();
            this.g = new List$1(String_$type, 0);
            this.f = new Dictionary$2(String_$type, CodeGenerationItemBuilderPropertyInfo.$, 0);
            this.c = null;
            this.b = 0;
            this.j = null;
            this.h = null;
            this.t = null;
            this.b = a;
            this.j = b;
            this.c = c;
            this.h = d;
        }
        a(a) {
            let b = this.c.getMetadata(this.j, a);
            return new CodeGenerationItemBuilder(this.b, this.r(b), this.c, this.h);
        }
        r(a) {
            return a.specificExternalType != null ? a.specificExternalType : a.specificType;
        }
        x(a, b, c, d) {
            if (!this.f.containsKey(a)) {
                this.g.add(a);
            }
            ;
            this.f.item(a, ((() => {
                let $ret = new CodeGenerationItemBuilderPropertyInfo();
                $ret.b = b;
                $ret.c = c;
                $ret.a = d;
                return $ret;
            })()));
        }
        v(a, b, c) {
            let d = null;
            if (!this.f.containsKey(a)) {
                this.g.add(a);
            }
            ;
            if (this.f.containsKey(a)) {
                let e = this.f.item(a);
                d = e.c;
            }
            this.f.item(a, ((() => {
                let $ret = new CodeGenerationItemBuilderPropertyInfo();
                $ret.b = b;
                $ret.c = d;
                $ret.a = c;
                return $ret;
            })()));
        }
        w(a, b) {
            if (this.f.containsKey(a)) {
                this.f.removeItem(a);
            }
        }
        i(a) {
            if (this.f.containsKey(a)) {
                return this.f.item(a).c;
            }
            return null;
        }
        u(a, b) {
            let c = this.m();
            if (b) {
                let d = this.p();
                a.a2 = d;
            }
            let e = false;
            let f = false;
            if (this.f.containsKey("Name")) {
                e = true;
                a.bl(this.f.item("Name").c);
            }
            else if (this.f.containsKey("name")) {
                e = true;
                a.bl(this.f.item("name").c);
            }
            else {
                if (b && this.t != null) {
                    e = true;
                    a.bl(this.t);
                }
                else if (this.e(this.c, a)) {
                    f = true;
                    a.bk(c);
                    if (a.x) {
                        this.g.add("name");
                        this.f.addItem("name", ((() => {
                            let $ret = new CodeGenerationItemBuilderPropertyInfo();
                            $ret.c = a.am;
                            return $ret;
                        })()));
                    }
                    else {
                        this.g.add("Name");
                        this.f.addItem("Name", ((() => {
                            let $ret = new CodeGenerationItemBuilderPropertyInfo();
                            $ret.c = a.am;
                            return $ret;
                        })()));
                    }
                }
            }
            a.a8(this.q(), this.j, this.c);
            let g = 0;
            for (let h of fromEnum(this.g)) {
                let i = this.f.item(h);
                if (i.b == null && h.toLowerCase() == "type") {
                    continue;
                }
                if (this.d(a, h, i)) {
                    a.be(g, h, i, this.c);
                    g++;
                }
            }
            a.bm();
            let j = 0;
            for (let k of fromEnum(this.g)) {
                let l = this.f.item(k);
                if (l.b == null && k.toLowerCase() == "type") {
                    continue;
                }
                if (!this.d(a, k, l)) {
                    a.bf(j, k, l, this.c);
                    j++;
                }
            }
            a.bg(j);
            a.bh(c);
            if (e) {
                a.bj();
            }
            else {
                if (f) {
                    a.bi();
                }
            }
        }
        e(a, b) {
            if (b.n) {
                return true;
            }
            for (let c of fromEnum(this.g)) {
                let d = this.f.item(c);
                if (b.q(a, d)) {
                    return true;
                }
            }
            return false;
        }
        d(a, b, c) {
            if (c.b == null) {
                if (b.toLowerCase() == "name") {
                    return true;
                }
            }
            switch (c.b.knownType) {
                case 10:
                case 7: return true;
                case 4: return c.c == null || !(typeCast(LinearGradientBrush.$, c.c) !== null);
                case 6:
                    return true;
                    break;
                case 23: return true;
                case 9:
                    if (a.v(this.c, c)) {
                        return true;
                    }
                    return false;
                case 5: return true;
                case 16: return true;
                case 20: return true;
                case 22: return true;
                case 15: return false;
                case 3: return true;
                case 19: return true;
                case 8:
                    if (a.v(this.c, c)) {
                        return true;
                    }
                    if (c.b.specificExternalType == "string") {
                        return true;
                    }
                    if (c.a != null) {
                        return true;
                    }
                    return false;
                case 13: return false;
                case 18: return true;
                case 1: return true;
                case 11: return true;
                case 14: return true;
                case 12: return true;
                case 2: return true;
                case 21: return true;
                case 17: return true;
                case 0: return false;
            }
            return false;
        }
        q() {
            return this.m();
        }
        get s() {
            return this.j;
        }
        static n(a, b, c, d, e) {
            if (a == 0 || a == 4) {
                let f = c.getMetadata(b, "_qualifiedName");
                if (f != null) {
                    let g = f.specificExternalType;
                    let h = g.lastIndexOf('.');
                    if (h >= 0) {
                        return g.substr(h + 1);
                    }
                    else {
                        return g;
                    }
                }
                return b;
            }
            else if ((a == 5 || a == 2 || a == 1 || a == 3) && c.getMetadata(b, "__qualifiedNameTS") != null) {
                let i = c.getMetadata(b, "__qualifiedNameTS");
                if (c.getMetadata(b, "__skipSuffix") != null) {
                    return d + i.specificExternalType;
                }
                return d + i.specificExternalType + e;
            }
            else {
                return d + b + e;
            }
        }
        m() {
            return CodeGenerationItemBuilder.n(this.b, this.j, this.c, this.k, this.l);
        }
        p() {
            let a = this.c.getMetadata(this.j, "_qualifiedName");
            if (a != null) {
                let b = a.specificExternalType;
                let c = b.split('.');
                if (c.length > 2) {
                    return c[c.length - 2].toLowerCase();
                }
            }
            if (this.j == "Legend" || this.j == "ItemLegend") {
                return "charts";
            }
            return "core";
        }
        get_k() {
            switch (this.b) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
                case 4: return "Ultra";
                case 0: return "Xam";
                case 5: return "Igb";
            }
            return "";
        }
        get k() {
            return this.get_k();
        }
        get_l() {
            switch (this.b) {
                case 1: return "Component";
                case 2: return "";
                case 3: return "Component";
                case 4: return "";
                case 0: return "";
            }
            return "";
        }
        get l() {
            return this.get_l();
        }
    }
    CodeGenerationItemBuilder.$t = /*@__PURE__*/ markType(CodeGenerationItemBuilder, 'CodeGenerationItemBuilder');
    return CodeGenerationItemBuilder;
})();
/**
 * @hidden
 */
export let CodeGenerationRendererAdapter = /*@__PURE__*/ (() => {
    class CodeGenerationRendererAdapter extends Base {
        constructor(a, b) {
            super();
            this.a = 0;
            this.b = null;
            this.b = b;
            this.a = a;
        }
        createColorCollection(a) {
            let b = new List$1(Base.$, 0);
            for (let c = 0; c < a.length; c++) {
                b.add1(a[c]);
            }
            return b;
        }
        createBrushCollection(a) {
            let b = new List$1(Base.$, 0);
            for (let c = 0; c < a.length; c++) {
                b.add1(a[c]);
            }
            return b;
        }
        createDoubleCollection(a) {
            let b = new List$1(Base.$, 0);
            for (let c = 0; c < a.length; c++) {
                b.add1(a[c]);
            }
            return b.toArray();
        }
        createObject(a, b, c) {
            let d = new CodeGenerationItemBuilder(this.a, a, c, b);
            return d;
        }
        coerceToEnum(a, b, c) {
            return c;
        }
        onUIThread(a, b) {
            b();
        }
        setOrUpdateCollectionOnTarget(a, b, c, d, e, f) {
            let g = this.getPropertyValue(e, b);
            if (g == null) {
                let h = new List$1(Base.$, 0);
                let i = f;
                for (let j = 0; j < i.count; j++) {
                    h.add1(i.item(j));
                }
                this.setPropertyValue(e, b, c, h, null, null);
            }
            else {
                let k = f;
                let l = g;
                l.clear();
                for (let m = 0; m < k.count; m++) {
                    l.add(k.item(m));
                }
            }
        }
        c(a) {
            if (a == null) {
                return false;
            }
            return false;
        }
        onPendingRef(a, b, c, d) {
            let e = a;
            e.v(b, c, d);
        }
        setPropertyValue(a, b, c, d, e, f) {
            if (!(typeCast(CodeGenerationItemBuilder.$, a) !== null)) {
                this.b.setPropertyValue(a, b, c, d, e, f);
                return;
            }
            while (stringContains(b, ".")) {
                let g = b.split('.');
                let h = this.getPropertyValue(a, g[0]);
                if (h != null && !this.c(h)) {
                    b = "";
                    for (let i = 1; i < g.length; i++) {
                        if (i > 1) {
                            b += ".";
                        }
                        b += g[i];
                    }
                    a = h;
                }
                else {
                    let j = a;
                    let k = j.a(g[0]);
                    h = k;
                    this.setPropertyValue(a, g[0], c, h, null, f);
                    b = "";
                    for (let l = 1; l < g.length; l++) {
                        if (l > 1) {
                            b += ".";
                        }
                        b += g[l];
                    }
                    a = h;
                }
            }
            let m = a;
            m.x(b, c, d, f);
        }
        getPropertyValue(a, b) {
            while (stringContains(b, ".")) {
                let c = b.split('.');
                let d = this.getPropertyValue(a, c[0]);
                if (d != null) {
                    b = "";
                    for (let e = 1; e < c.length; e++) {
                        if (e > 1) {
                            b += ".";
                        }
                        b += c[e];
                    }
                    a = d;
                }
                else {
                    let f = a;
                    let g = f.a(c[0]);
                    d = g;
                    for (let h = 1; h < c.length; h++) {
                        if (h > 1) {
                            b += ".";
                        }
                        b += c[h];
                    }
                    a = d;
                }
            }
            let i = a;
            return i.i(b);
        }
        getRootObject(a) {
            let b = a;
            let c = b.a.count > 0 ? b.a._inner[0] : null;
            return c;
        }
        clearContainer(a, b, c) {
            let d = a;
            let e = d.a.count > 0 ? d.a._inner[0] : null;
            d.a.clear();
            c(false);
        }
        forPropertyValueItem(a, b, c) {
            let d = this.getPropertyValue(a, b);
            if (typeCast(IList_$type, d) !== null) {
                let e = d;
                for (let f = 0; f < e.count; f++) {
                    c(e.item(f));
                }
            }
            else if (typeCast(Array_$type, d) !== null) {
                let g = d;
                for (let h = 0; h < g.length; h++) {
                    c(g[h]);
                }
            }
        }
        clearCollection(a, b, c) {
            let d = this.getPropertyValue(a, b);
            if (typeCast(IList_$type, d) !== null) {
                d.clear();
            }
            else if (typeCast(Array_$type, d) !== null) {
                this.setPropertyValue(a, b, c, null, d, null);
            }
        }
        addItemToCollection(a, b, c, d, e) {
            let f = this.getPropertyValue(c, a);
            if (typeCast(IList_$type, f) !== null) {
                f.insert(d, e);
            }
            else if (typeCast(Array_$type, f) !== null) {
                let g = f;
                let h = new List$1(Base.$, 0);
                for (let i = 0; i < g.count; i++) {
                    h.add1(g.item(i));
                }
                h.insert1(d, e);
                let j = [g.count + 1];
                for (let k = 0; k < j.length; k++) {
                    j[k] = h._inner[k];
                }
                this.setPropertyValue(c, a, b, j, f, null);
            }
        }
        resetPropertyOnTarget(a, b, c, d) {
            while (stringContains(b, ".")) {
                let e = b.split('.');
                let f = this.getPropertyValue(d, e[0]);
                if (f != null && !this.c(f)) {
                    b = "";
                    for (let g = 1; g < e.length; g++) {
                        if (g > 1) {
                            b += ".";
                        }
                        b += e[g];
                    }
                    d = f;
                }
                else {
                    let h = d;
                    let i = h.a(e[0]);
                    f = i;
                    this.setPropertyValue(d, e[0], c, f, null, null);
                    b = "";
                    for (let j = 1; j < e.length; j++) {
                        if (j > 1) {
                            b += ".";
                        }
                        b += e[j];
                    }
                    d = f;
                }
            }
            let k = d;
            k.w(b, c);
        }
        replaceItemInCollection(a, b, c, d, e) {
            let f = this.getPropertyValue(c, a);
            if (typeCast(IList_$type, f) !== null) {
                f.item(d, e);
            }
            else if (typeCast(Array_$type, f) !== null) {
                let g = f;
                let h = new List$1(Base.$, 0);
                for (let i = 0; i < g.count; i++) {
                    h.add1(g.item(i));
                }
                h._inner[d] = e;
                let j = [g.count + 1];
                for (let k = 0; k < j.length; k++) {
                    j[k] = h._inner[k];
                }
                this.setPropertyValue(c, a, b, j, f, null);
            }
        }
        replaceRootItem(a, b, c, d) {
            let e = this.createObject(b, a, c);
            let f = a;
            let g = f.a.count > 0 ? f.a._inner[0] : null;
            f.a.clear();
            {
                f.a.add(e);
            }
            d(false);
        }
        removeItemFromCollection(a, b, c, d) {
            let e = this.getPropertyValue(c, a);
            if (typeCast(IList_$type, e) !== null) {
                e.removeAt(d);
            }
            else if (typeCast(Array_$type, e) !== null) {
                let f = e;
                let g = new List$1(Base.$, 0);
                for (let h = 0; h < f.count; h++) {
                    g.add1(f.item(h));
                }
                g.removeAt(d);
                let i = [f.count - 1];
                for (let j = 0; j < i.length; j++) {
                    i[j] = g._inner[j];
                }
                this.setPropertyValue(c, a, b, i, e, null);
            }
        }
        removeRootItem(a, b, c) {
            let d = a;
            let e = d.a.count > 0 ? d.a._inner[0] : null;
            d.a.clear();
            c(false);
        }
        flushChanges(a) {
        }
    }
    CodeGenerationRendererAdapter.$t = /*@__PURE__*/ markType(CodeGenerationRendererAdapter, 'CodeGenerationRendererAdapter', Base.$, [IComponentRendererAdapter_$type]);
    return CodeGenerationRendererAdapter;
})();
