// igniteui-react-core/esm2015/lib/ColumnSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeriesDescription } from "./HorizontalAnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnSeriesDescription = /*@__PURE__*/ (() => {
    class ColumnSeriesDescription extends HorizontalAnchoredCategorySeriesDescription {
        get_type() {
            return "ColumnSeries";
        }
        constructor() {
            super();
            this.e1 = 0;
            this.e2 = 0;
            this.e5 = null;
        }
        get radiusX() {
            return this.e1;
        }
        set radiusX(a) {
            this.e1 = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.e2;
        }
        set radiusY(a) {
            this.e2 = a;
            this.g("RadiusY");
        }
        get consolidatedColumnVerticalPosition() {
            return this.e5;
        }
        set consolidatedColumnVerticalPosition(a) {
            this.e5 = a;
            this.g("ConsolidatedColumnVerticalPosition");
        }
    }
    ColumnSeriesDescription.$t = /*@__PURE__*/ markType(ColumnSeriesDescription, 'ColumnSeriesDescription', HorizontalAnchoredCategorySeriesDescription.$);
    return ColumnSeriesDescription;
})();
