// igniteui-react-core/esm2015/lib/ColumnSummaryDescriptionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnSummaryDescriptionDescription = /*@__PURE__*/ (() => {
    class ColumnSummaryDescriptionDescription extends Description {
        get_type() {
            return "ColumnSummaryDescription";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.s = null;
            this.u = null;
            this.p = null;
            this.r = null;
            this.w = null;
            this.i = null;
            this.q = null;
            this.h = null;
            this.m = 0;
            this.t = null;
            this.v = null;
        }
        get field() {
            return this.s;
        }
        set field(a) {
            this.s = a;
            this.g("Field");
        }
        get operand() {
            return this.u;
        }
        set operand(a) {
            this.u = a;
            this.g("Operand");
        }
        get calculatorDisplayName() {
            return this.p;
        }
        set calculatorDisplayName(a) {
            this.p = a;
            this.g("CalculatorDisplayName");
        }
        get displayName() {
            return this.r;
        }
        set displayName(a) {
            this.r = a;
            this.g("DisplayName");
        }
        get valueFormat() {
            return this.w;
        }
        set valueFormat(a) {
            this.w = a;
            this.g("ValueFormat");
        }
        get valueFormatSpecifiers() {
            return this.i;
        }
        set valueFormatSpecifiers(a) {
            this.i = a;
            this.g("ValueFormatSpecifiers");
        }
        get displayFormat() {
            return this.q;
        }
        set displayFormat(a) {
            this.q = a;
            this.g("DisplayFormat");
        }
        get displayFormatSpecifiers() {
            return this.h;
        }
        set displayFormatSpecifiers(a) {
            this.h = a;
            this.g("DisplayFormatSpecifiers");
        }
        get maxFractionDigits() {
            return this.m;
        }
        set maxFractionDigits(a) {
            this.m = a;
            this.g("MaxFractionDigits");
        }
        get formatTextRef() {
            return this.t;
        }
        set formatTextRef(a) {
            this.t = a;
            this.g("FormatTextRef");
        }
        get provideCalculatorRef() {
            return this.v;
        }
        set provideCalculatorRef(a) {
            this.v = a;
            this.g("ProvideCalculatorRef");
        }
    }
    ColumnSummaryDescriptionDescription.$t = /*@__PURE__*/ markType(ColumnSummaryDescriptionDescription, 'ColumnSummaryDescriptionDescription', Description.$);
    ColumnSummaryDescriptionDescription.__marshalByValue = true;
    ColumnSummaryDescriptionDescription.__marshalByValueAlias = "ColumnSummaryDescription";
    return ColumnSummaryDescriptionDescription;
})();
