// igniteui-react-core/esm2015/lib/DataGridFilterDialogDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridFilterDialogDescription = /*@__PURE__*/ (() => {
    class DataGridFilterDialogDescription extends Description {
        get_type() {
            return "DataGridFilterDialog";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.k = 0;
            this.j = 0;
            this.s = null;
            this.l = 0;
            this.p = null;
            this.r = null;
            this.q = null;
            this.t = null;
        }
        get isAnimationEnabled() {
            return this.h;
        }
        set isAnimationEnabled(a) {
            this.h = a;
            this.g("IsAnimationEnabled");
        }
        get pixelScalingRatio() {
            return this.k;
        }
        set pixelScalingRatio(a) {
            this.k = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.j;
        }
        set actualPixelScalingRatio(a) {
            this.j = a;
            this.g("ActualPixelScalingRatio");
        }
        get iconColor() {
            return this.s;
        }
        set iconColor(a) {
            this.s = a;
            this.g("IconColor");
        }
        get viewSize() {
            return this.l;
        }
        set viewSize(a) {
            this.l = a;
            this.g("ViewSize");
        }
        get dialogOpeningRef() {
            return this.p;
        }
        set dialogOpeningRef(a) {
            this.p = a;
            this.g("DialogOpeningRef");
        }
        get filterChangingRef() {
            return this.r;
        }
        set filterChangingRef(a) {
            this.r = a;
            this.g("FilterChangingRef");
        }
        get filterChangedRef() {
            return this.q;
        }
        set filterChangedRef(a) {
            this.q = a;
            this.g("FilterChangedRef");
        }
        get renderCompletedRef() {
            return this.t;
        }
        set renderCompletedRef(a) {
            this.t = a;
            this.g("RenderCompletedRef");
        }
    }
    DataGridFilterDialogDescription.$t = /*@__PURE__*/ markType(DataGridFilterDialogDescription, 'DataGridFilterDialogDescription', Description.$);
    return DataGridFilterDialogDescription;
})();
