// igniteui-react-core/esm2015/lib/DataChartMouseButtonEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataChartMouseButtonEventArgsDescription = /*@__PURE__*/ (() => {
    class DataChartMouseButtonEventArgsDescription extends Description {
        get_type() {
            return "DataChartMouseButtonEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = false;
            this.n = null;
            this.h = null;
            this.j = null;
        }
        get handled() {
            return this.l;
        }
        set handled(a) {
            this.l = a;
            this.g("Handled");
        }
        get itemRef() {
            return this.n;
        }
        set itemRef(a) {
            this.n = a;
            this.g("ItemRef");
        }
        get series() {
            return this.h;
        }
        set series(a) {
            this.h = a;
            this.g("Series");
        }
        get chart() {
            return this.j;
        }
        set chart(a) {
            this.j = a;
            this.g("Chart");
        }
    }
    DataChartMouseButtonEventArgsDescription.$t = /*@__PURE__*/ markType(DataChartMouseButtonEventArgsDescription, 'DataChartMouseButtonEventArgsDescription', Description.$);
    return DataChartMouseButtonEventArgsDescription;
})();
