// igniteui-react-core/esm2015/lib/DataLegendStylingColumnEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataLegendStylingColumnEventArgsDescription = /*@__PURE__*/ (() => {
    class DataLegendStylingColumnEventArgsDescription extends Description {
        get_type() {
            return "DataLegendStylingColumnEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.k = 0;
            this.p = null;
            this.h = 0;
            this.u = null;
            this.t = null;
            this.s = null;
            this.n = null;
            this.q = null;
            this.v = null;
            this.o = null;
            this.r = null;
            this.w = null;
        }
        get columnIndex() {
            return this.j;
        }
        set columnIndex(a) {
            this.j = a;
            this.g("ColumnIndex");
        }
        get seriesIndex() {
            return this.k;
        }
        set seriesIndex(a) {
            this.k = a;
            this.g("SeriesIndex");
        }
        get seriesTitle() {
            return this.p;
        }
        set seriesTitle(a) {
            this.p = a;
            this.g("SeriesTitle");
        }
        get valueOriginal() {
            return this.h;
        }
        set valueOriginal(a) {
            this.h = a;
            this.g("ValueOriginal");
        }
        get valueMemberPath() {
            return this.u;
        }
        set valueMemberPath(a) {
            this.u = a;
            this.g("ValueMemberPath");
        }
        get valueMemberLabel() {
            return this.t;
        }
        set valueMemberLabel(a) {
            this.t = a;
            this.g("ValueMemberLabel");
        }
        get valueAbbreviation() {
            return this.s;
        }
        set valueAbbreviation(a) {
            this.s = a;
            this.g("ValueAbbreviation");
        }
        get labelText() {
            return this.n;
        }
        set labelText(a) {
            this.n = a;
            this.g("LabelText");
        }
        get unitsText() {
            return this.q;
        }
        set unitsText(a) {
            this.q = a;
            this.g("UnitsText");
        }
        get valueText() {
            return this.v;
        }
        set valueText(a) {
            this.v = a;
            this.g("ValueText");
        }
        get labelTextColor() {
            return this.o;
        }
        set labelTextColor(a) {
            this.o = a;
            this.g("LabelTextColor");
        }
        get unitsTextColor() {
            return this.r;
        }
        set unitsTextColor(a) {
            this.r = a;
            this.g("UnitsTextColor");
        }
        get valueTextColor() {
            return this.w;
        }
        set valueTextColor(a) {
            this.w = a;
            this.g("ValueTextColor");
        }
    }
    DataLegendStylingColumnEventArgsDescription.$t = /*@__PURE__*/ markType(DataLegendStylingColumnEventArgsDescription, 'DataLegendStylingColumnEventArgsDescription', Description.$);
    return DataLegendStylingColumnEventArgsDescription;
})();
