// igniteui-react-core/esm2015/lib/DatePickerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DatePickerDescription = /*@__PURE__*/ (() => {
    class DatePickerDescription extends Description {
        get_type() {
            return "DatePicker";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.w = new Date();
            this.v = new Date();
            this.ap = null;
            this.at = null;
            this.aq = null;
            this.ab = 0;
            this.ar = null;
            this.as = null;
            this.av = null;
            this.u = new Date();
            this.t = new Date();
            this.ai = null;
            this.af = null;
            this.ay = null;
            this.ac = 0;
            this.az = null;
            this.a0 = null;
            this.ax = null;
            this.an = null;
            this.k = false;
            this.l = false;
            this.h = false;
            this.j = false;
            this.aj = null;
            this.ak = null;
            this.m = false;
            this.ah = null;
            this.al = null;
            this.i = false;
            this.aw = null;
            this.am = null;
            this.au = null;
            this.ao = null;
            this.ag = null;
            this.a1 = null;
        }
        get value() {
            return this.w;
        }
        set value(a) {
            this.w = a;
            this.g("Value");
        }
        get today() {
            return this.v;
        }
        set today(a) {
            this.v = a;
            this.g("Today");
        }
        get label() {
            return this.ap;
        }
        set label(a) {
            this.ap = a;
            this.g("Label");
        }
        get labelTextColor() {
            return this.at;
        }
        set labelTextColor(a) {
            this.at = a;
            this.g("LabelTextColor");
        }
        get labelFontFamily() {
            return this.aq;
        }
        set labelFontFamily(a) {
            this.aq = a;
            this.g("LabelFontFamily");
        }
        get labelFontSize() {
            return this.ab;
        }
        set labelFontSize(a) {
            this.ab = a;
            this.g("LabelFontSize");
        }
        get labelFontStyle() {
            return this.ar;
        }
        set labelFontStyle(a) {
            this.ar = a;
            this.g("LabelFontStyle");
        }
        get labelFontWeight() {
            return this.as;
        }
        set labelFontWeight(a) {
            this.as = a;
            this.g("LabelFontWeight");
        }
        get placeholder() {
            return this.av;
        }
        set placeholder(a) {
            this.av = a;
            this.g("Placeholder");
        }
        get minDate() {
            return this.u;
        }
        set minDate(a) {
            this.u = a;
            this.g("MinDate");
        }
        get maxDate() {
            return this.t;
        }
        set maxDate(a) {
            this.t = a;
            this.g("MaxDate");
        }
        get density() {
            return this.ai;
        }
        set density(a) {
            this.ai = a;
            this.g("Density");
        }
        get baseTheme() {
            return this.af;
        }
        set baseTheme(a) {
            this.af = a;
            this.g("BaseTheme");
        }
        get textFontFamily() {
            return this.ay;
        }
        set textFontFamily(a) {
            this.ay = a;
            this.g("TextFontFamily");
        }
        get textFontSize() {
            return this.ac;
        }
        set textFontSize(a) {
            this.ac = a;
            this.g("TextFontSize");
        }
        get textFontStyle() {
            return this.az;
        }
        set textFontStyle(a) {
            this.az = a;
            this.g("TextFontStyle");
        }
        get textFontWeight() {
            return this.a0;
        }
        set textFontWeight(a) {
            this.a0 = a;
            this.g("TextFontWeight");
        }
        get textColor() {
            return this.ax;
        }
        set textColor(a) {
            this.ax = a;
            this.g("TextColor");
        }
        get iconColor() {
            return this.an;
        }
        set iconColor(a) {
            this.an = a;
            this.g("IconColor");
        }
        get showClearButton() {
            return this.k;
        }
        set showClearButton(a) {
            this.k = a;
            this.g("ShowClearButton");
        }
        get showTodayButton() {
            return this.l;
        }
        set showTodayButton(a) {
            this.l = a;
            this.g("ShowTodayButton");
        }
        get allowTextInput() {
            return this.h;
        }
        set allowTextInput(a) {
            this.h = a;
            this.g("AllowTextInput");
        }
        get openOnFocus() {
            return this.j;
        }
        set openOnFocus(a) {
            this.j = a;
            this.g("OpenOnFocus");
        }
        get firstDayOfWeek() {
            return this.aj;
        }
        set firstDayOfWeek(a) {
            this.aj = a;
            this.g("FirstDayOfWeek");
        }
        get firstWeekOfYear() {
            return this.ak;
        }
        set firstWeekOfYear(a) {
            this.ak = a;
            this.g("FirstWeekOfYear");
        }
        get showWeekNumbers() {
            return this.m;
        }
        set showWeekNumbers(a) {
            this.m = a;
            this.g("ShowWeekNumbers");
        }
        get dateFormat() {
            return this.ah;
        }
        set dateFormat(a) {
            this.ah = a;
            this.g("DateFormat");
        }
        get formatString() {
            return this.al;
        }
        set formatString(a) {
            this.al = a;
            this.g("FormatString");
        }
        get isDisabled() {
            return this.i;
        }
        set isDisabled(a) {
            this.i = a;
            this.g("IsDisabled");
        }
        get selectedValueChangedRef() {
            return this.aw;
        }
        set selectedValueChangedRef(a) {
            this.aw = a;
            this.g("SelectedValueChangedRef");
        }
        get gotFocusRef() {
            return this.am;
        }
        set gotFocusRef(a) {
            this.am = a;
            this.g("GotFocusRef");
        }
        get lostFocusRef() {
            return this.au;
        }
        set lostFocusRef(a) {
            this.au = a;
            this.g("LostFocusRef");
        }
        get keyDownRef() {
            return this.ao;
        }
        set keyDownRef(a) {
            this.ao = a;
            this.g("KeyDownRef");
        }
        get changingRef() {
            return this.ag;
        }
        set changingRef(a) {
            this.ag = a;
            this.g("ChangingRef");
        }
        get valueChangeRef() {
            return this.a1;
        }
        set valueChangeRef(a) {
            this.a1 = a;
            this.g("ValueChangeRef");
        }
    }
    DatePickerDescription.$t = /*@__PURE__*/ markType(DatePickerDescription, 'DatePickerDescription', Description.$);
    return DatePickerDescription;
})();
