// igniteui-react-core/esm2015/lib/DefinitionBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DefinitionBaseDescription = /*@__PURE__*/ (() => {
    class DefinitionBaseDescription extends Description {
        get_type() {
            return "DefinitionBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.ax = null;
            this.ay = null;
            this.an = null;
            this.a2 = null;
            this.aq = null;
            this.u = 0;
            this.w = 0;
            this.v = 0;
            this.t = 0;
            this.q = 0;
            this.s = 0;
            this.r = 0;
            this.p = 0;
            this.y = 0;
            this.aa = 0;
            this.z = 0;
            this.x = 0;
            this.ap = null;
            this.ao = null;
            this.ar = null;
            this.ba = null;
            this.av = null;
            this.a9 = null;
            this.au = null;
            this.a7 = null;
            this.as = null;
            this.i = 0;
            this.a6 = null;
            this.bc = null;
            this.a8 = null;
            this.at = null;
            this.bb = null;
            this.aw = null;
            this.k = 0;
            this.h = 0;
            this.a3 = null;
            this.j = 0;
            this.a4 = null;
            this.a5 = null;
            this.a0 = null;
            this.a1 = null;
            this.az = null;
        }
        get background() {
            return this.ax;
        }
        set background(a) {
            this.ax = a;
            this.g("Background");
        }
        get border() {
            return this.ay;
        }
        set border(a) {
            this.ay = a;
            this.g("Border");
        }
        get activationBorder() {
            return this.an;
        }
        set activationBorder(a) {
            this.an = a;
            this.g("ActivationBorder");
        }
        get errorBorder() {
            return this.a2;
        }
        set errorBorder(a) {
            this.a2 = a;
            this.g("ErrorBorder");
        }
        get actualBorder() {
            return this.aq;
        }
        set actualBorder(a) {
            this.aq = a;
            this.g("ActualBorder");
        }
        get borderLeftWidth() {
            return this.u;
        }
        set borderLeftWidth(a) {
            this.u = a;
            this.g("BorderLeftWidth");
        }
        get borderTopWidth() {
            return this.w;
        }
        set borderTopWidth(a) {
            this.w = a;
            this.g("BorderTopWidth");
        }
        get borderRightWidth() {
            return this.v;
        }
        set borderRightWidth(a) {
            this.v = a;
            this.g("BorderRightWidth");
        }
        get borderBottomWidth() {
            return this.t;
        }
        set borderBottomWidth(a) {
            this.t = a;
            this.g("BorderBottomWidth");
        }
        get activationBorderLeftWidth() {
            return this.q;
        }
        set activationBorderLeftWidth(a) {
            this.q = a;
            this.g("ActivationBorderLeftWidth");
        }
        get activationBorderTopWidth() {
            return this.s;
        }
        set activationBorderTopWidth(a) {
            this.s = a;
            this.g("ActivationBorderTopWidth");
        }
        get activationBorderRightWidth() {
            return this.r;
        }
        set activationBorderRightWidth(a) {
            this.r = a;
            this.g("ActivationBorderRightWidth");
        }
        get activationBorderBottomWidth() {
            return this.p;
        }
        set activationBorderBottomWidth(a) {
            this.p = a;
            this.g("ActivationBorderBottomWidth");
        }
        get errorBorderLeftWidth() {
            return this.y;
        }
        set errorBorderLeftWidth(a) {
            this.y = a;
            this.g("ErrorBorderLeftWidth");
        }
        get errorBorderTopWidth() {
            return this.aa;
        }
        set errorBorderTopWidth(a) {
            this.aa = a;
            this.g("ErrorBorderTopWidth");
        }
        get errorBorderRightWidth() {
            return this.z;
        }
        set errorBorderRightWidth(a) {
            this.z = a;
            this.g("ErrorBorderRightWidth");
        }
        get errorBorderBottomWidth() {
            return this.x;
        }
        set errorBorderBottomWidth(a) {
            this.x = a;
            this.g("ErrorBorderBottomWidth");
        }
        get actualBackground() {
            return this.ap;
        }
        set actualBackground(a) {
            this.ap = a;
            this.g("ActualBackground");
        }
        get actualActivationBorder() {
            return this.ao;
        }
        set actualActivationBorder(a) {
            this.ao = a;
            this.g("ActualActivationBorder");
        }
        get actualErrorBorder() {
            return this.ar;
        }
        set actualErrorBorder(a) {
            this.ar = a;
            this.g("ActualErrorBorder");
        }
        get stickyRowBackground() {
            return this.ba;
        }
        set stickyRowBackground(a) {
            this.ba = a;
            this.g("StickyRowBackground");
        }
        get actualStickyRowBackground() {
            return this.av;
        }
        set actualStickyRowBackground(a) {
            this.av = a;
            this.g("ActualStickyRowBackground");
        }
        get pinnedRowBackground() {
            return this.a9;
        }
        set pinnedRowBackground(a) {
            this.a9 = a;
            this.g("PinnedRowBackground");
        }
        get actualPinnedRowBackground() {
            return this.au;
        }
        set actualPinnedRowBackground(a) {
            this.au = a;
            this.g("ActualPinnedRowBackground");
        }
        get lastStickyRowBackground() {
            return this.a7;
        }
        set lastStickyRowBackground(a) {
            this.a7 = a;
            this.g("LastStickyRowBackground");
        }
        get actualLastStickyRowBackground() {
            return this.as;
        }
        set actualLastStickyRowBackground(a) {
            this.as = a;
            this.g("ActualLastStickyRowBackground");
        }
        get contentOpacity() {
            return this.i;
        }
        set contentOpacity(a) {
            this.i = a;
            this.g("ContentOpacity");
        }
        get horizontalAlignment() {
            return this.a6;
        }
        set horizontalAlignment(a) {
            this.a6 = a;
            this.g("HorizontalAlignment");
        }
        get verticalAlignment() {
            return this.bc;
        }
        set verticalAlignment(a) {
            this.bc = a;
            this.g("VerticalAlignment");
        }
        get lineBreakMode() {
            return this.a8;
        }
        set lineBreakMode(a) {
            this.a8 = a;
            this.g("LineBreakMode");
        }
        get actualLineBreakMode() {
            return this.at;
        }
        set actualLineBreakMode(a) {
            this.at = a;
            this.g("ActualLineBreakMode");
        }
        get textColor() {
            return this.bb;
        }
        set textColor(a) {
            this.bb = a;
            this.g("TextColor");
        }
        get actualTextColor() {
            return this.aw;
        }
        set actualTextColor(a) {
            this.aw = a;
            this.g("ActualTextColor");
        }
        get pinnedRowOpacity() {
            return this.k;
        }
        set pinnedRowOpacity(a) {
            this.k = a;
            this.g("PinnedRowOpacity");
        }
        get actualPinnedRowOpacity() {
            return this.h;
        }
        set actualPinnedRowOpacity(a) {
            this.h = a;
            this.g("ActualPinnedRowOpacity");
        }
        get fontFamily() {
            return this.a3;
        }
        set fontFamily(a) {
            this.a3 = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.j;
        }
        set fontSize(a) {
            this.j = a;
            this.g("FontSize");
        }
        get fontStyle() {
            return this.a4;
        }
        set fontStyle(a) {
            this.a4 = a;
            this.g("FontStyle");
        }
        get fontWeight() {
            return this.a5;
        }
        set fontWeight(a) {
            this.a5 = a;
            this.g("FontWeight");
        }
        get dataBindingRef() {
            return this.a0;
        }
        set dataBindingRef(a) {
            this.a0 = a;
            this.g("DataBindingRef");
        }
        get dataBoundRef() {
            return this.a1;
        }
        set dataBoundRef(a) {
            this.a1 = a;
            this.g("DataBoundRef");
        }
        get cellStyleKeyRequestedRef() {
            return this.az;
        }
        set cellStyleKeyRequestedRef(a) {
            this.az = a;
            this.g("CellStyleKeyRequestedRef");
        }
    }
    DefinitionBaseDescription.$t = /*@__PURE__*/ markType(DefinitionBaseDescription, 'DefinitionBaseDescription', Description.$);
    return DefinitionBaseDescription;
})();
