// igniteui-react-core/esm2015/lib/DomainChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DomainChartDescription = /*@__PURE__*/ (() => {
    class DomainChartDescription extends Description {
        get_type() {
            return "DomainChart";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.ei = 0;
            this.ev = 0;
            this.ew = 0;
            this.ex = 0;
            this.eu = 0;
            this.eq = 0;
            this.es = 0;
            this.er = 0;
            this.ep = 0;
            this.lf = null;
            this.li = null;
            this.ef = 0;
            this.ey = 0;
            this.eo = 0;
            this.co = 0;
            this.h6 = 0;
            this.ll = null;
            this.lg = null;
            this.lj = null;
            this.i2 = null;
            this.da = 0;
            this.db = 0;
            this.cu = 0;
            this.cv = 0;
            this.i0 = null;
            this.i1 = null;
            this.bn = false;
            this.p = null;
            this.m = null;
            this.o = null;
            this.l = null;
            this.kb = null;
            this.h2 = 0;
            this.h3 = 0;
            this.kd = null;
            this.kc = null;
            this.bo = false;
            this.ke = null;
            this.q = null;
            this.dz = 0;
            this.d0 = 0;
            this.d1 = 0;
            this.d2 = 0;
            this.bp = false;
            this.kj = null;
            this.bq = false;
            this.d4 = 0;
            this.d5 = 0;
            this.d6 = 0;
            this.d7 = 0;
            this.kf = null;
            this.kg = null;
            this.d3 = 0;
            this.kh = null;
            this.ki = null;
            this.i5 = null;
            this.n = null;
            this.i3 = null;
            this.i4 = null;
            this.i6 = null;
            this.i7 = null;
            this.i8 = null;
            this.c1 = 0;
            this.c2 = 0;
            this.c3 = 0;
            this.c4 = 0;
            this.cw = 0;
            this.cx = 0;
            this.cy = 0;
            this.cz = 0;
            this.bm = false;
            this.i9 = null;
            this.c0 = 0;
            this.ja = null;
            this.jb = null;
            this.jn = null;
            this.jo = null;
            this.dh = 0;
            this.jp = null;
            this.jq = null;
            this.jr = null;
            this.jm = null;
            this.di = 0;
            this.dj = 0;
            this.dk = 0;
            this.dl = 0;
            this.dd = 0;
            this.de = 0;
            this.df = 0;
            this.dg = 0;
            this.jx = null;
            this.jy = null;
            this.dn = 0;
            this.jz = null;
            this.j0 = null;
            this.jh = null;
            this.ji = null;
            this.jj = null;
            this.dc = 0;
            this.jk = null;
            this.jl = null;
            this.js = null;
            this.jt = null;
            this.ju = null;
            this.dm = 0;
            this.jv = null;
            this.jw = null;
            this.cq = 0;
            this.cr = 0;
            this.cs = 0;
            this.ct = 0;
            this.iz = null;
            this.j5 = null;
            this.j6 = null;
            this.j7 = null;
            this.dv = 0;
            this.dw = 0;
            this.dx = 0;
            this.dy = 0;
            this.j8 = null;
            this.du = 0;
            this.j9 = null;
            this.ka = null;
            this.dq = 0;
            this.dr = 0;
            this.ds = 0;
            this.dt = 0;
            this.j1 = null;
            this.j2 = null;
            this.dp = 0;
            this.j3 = null;
            this.j4 = null;
            this.jc = null;
            this.jd = null;
            this.c6 = 0;
            this.c7 = 0;
            this.c8 = 0;
            this.c9 = 0;
            this.je = null;
            this.c5 = 0;
            this.jf = null;
            this.jg = null;
            this.ae = null;
            this.u = null;
            this.t = null;
            this.s = null;
            this.af = null;
            this.y = null;
            this.x = null;
            this.w = null;
            this.z = null;
            this.ku = null;
            this.km = null;
            this.kt = null;
            this.kv = null;
            this.iy = null;
            this.bb = null;
            this.v = null;
            this.r = null;
            this.j = null;
            this.ad = null;
            this.h = null;
            this.i = null;
            this.kz = null;
            this.bt = false;
            this.bv = false;
            this.bx = false;
            this.bz = false;
            this.bu = false;
            this.bd = null;
            this.ip = null;
            this.ld = null;
            this.lh = null;
            this.le = null;
            this.ln = null;
            this.en = 0;
            this.et = 0;
            this.k3 = null;
            this.k2 = null;
            this.k1 = null;
            this.eg = 0;
            this.eh = 0;
            this.ac = null;
            this.k0 = null;
            this.aa = null;
            this.ab = null;
            this.h5 = 0;
            this.cn = 0;
            this.bg = false;
            this.ag = null;
            this.lm = null;
            this.ez = 0;
            this.bf = false;
            this.ek = 0;
            this.em = 0;
            this.el = 0;
            this.ej = 0;
            this.iq = null;
            this.k5 = null;
            this.k6 = null;
            this.ko = null;
            this.kn = null;
            this.kw = null;
            this.ky = null;
            this.kx = null;
            this.h7 = 0;
            this.lk = null;
            this.iv = null;
            this.bl = false;
            this.ix = null;
            this.cp = 0;
            this.iw = null;
            this.bk = false;
            this.ir = null;
            this.it = null;
            this.is = null;
            this.iu = null;
            this.h0 = 0;
            this.h1 = 0;
            this.bw = false;
            this.by = false;
            this.br = false;
            this.kk = null;
            this.kl = null;
            this.h4 = 0;
            this.bh = false;
            this.bj = false;
            this.bi = false;
            this.ij = null;
            this.k = null;
            this.im = null;
            this.io = null;
            this.ik = null;
            this.ii = null;
            this.ah = null;
            this.ai = null;
            this.e0 = 0;
            this.kq = null;
            this.lp = null;
            this.ks = null;
            this.lr = null;
            this.kp = null;
            this.kr = null;
            this.ec = 0;
            this.eb = 0;
            this.d8 = 0;
            this.d9 = 0;
            this.ea = 0;
            this.ee = 0;
            this.ed = 0;
            this.bs = false;
            this.lo = null;
            this.lq = null;
            this.e4 = 0;
            this.e3 = 0;
            this.e1 = 0;
            this.e7 = 0;
            this.e2 = 0;
            this.e6 = 0;
            this.e5 = 0;
            this.b0 = false;
            this.e9 = 0;
            this.e8 = 0;
            this.fb = 0;
            this.fa = 0;
            this.k4 = null;
            this.lc = null;
            this.k8 = null;
            this.k9 = null;
            this.la = null;
            this.k7 = null;
            this.lb = null;
            this.il = null;
            this.ih = null;
        }
        get pixelScalingRatio() {
            return this.ei;
        }
        set pixelScalingRatio(a) {
            this.ei = a;
            this.g("PixelScalingRatio");
        }
        get titleLeftMargin() {
            return this.ev;
        }
        set titleLeftMargin(a) {
            this.ev = a;
            this.g("TitleLeftMargin");
        }
        get titleRightMargin() {
            return this.ew;
        }
        set titleRightMargin(a) {
            this.ew = a;
            this.g("TitleRightMargin");
        }
        get titleTopMargin() {
            return this.ex;
        }
        set titleTopMargin(a) {
            this.ex = a;
            this.g("TitleTopMargin");
        }
        get titleBottomMargin() {
            return this.eu;
        }
        set titleBottomMargin(a) {
            this.eu = a;
            this.g("TitleBottomMargin");
        }
        get subtitleLeftMargin() {
            return this.eq;
        }
        set subtitleLeftMargin(a) {
            this.eq = a;
            this.g("SubtitleLeftMargin");
        }
        get subtitleTopMargin() {
            return this.es;
        }
        set subtitleTopMargin(a) {
            this.es = a;
            this.g("SubtitleTopMargin");
        }
        get subtitleRightMargin() {
            return this.er;
        }
        set subtitleRightMargin(a) {
            this.er = a;
            this.g("SubtitleRightMargin");
        }
        get subtitleBottomMargin() {
            return this.ep;
        }
        set subtitleBottomMargin(a) {
            this.ep = a;
            this.g("SubtitleBottomMargin");
        }
        get subtitleTextColor() {
            return this.lf;
        }
        set subtitleTextColor(a) {
            this.lf = a;
            this.g("SubtitleTextColor");
        }
        get titleTextColor() {
            return this.li;
        }
        set titleTextColor(a) {
            this.li = a;
            this.g("TitleTextColor");
        }
        get leftMargin() {
            return this.ef;
        }
        set leftMargin(a) {
            this.ef = a;
            this.g("LeftMargin");
        }
        get topMargin() {
            return this.ey;
        }
        set topMargin(a) {
            this.ey = a;
            this.g("TopMargin");
        }
        get rightMargin() {
            return this.eo;
        }
        set rightMargin(a) {
            this.eo = a;
            this.g("RightMargin");
        }
        get bottomMargin() {
            return this.co;
        }
        set bottomMargin(a) {
            this.co = a;
            this.g("BottomMargin");
        }
        get transitionDuration() {
            return this.h6;
        }
        set transitionDuration(a) {
            this.h6 = a;
            this.g("TransitionDuration");
        }
        get transitionEasingFunctionRef() {
            return this.ll;
        }
        set transitionEasingFunctionRef(a) {
            this.ll = a;
            this.g("TransitionEasingFunctionRef");
        }
        get subtitleTextStyle() {
            return this.lg;
        }
        set subtitleTextStyle(a) {
            this.lg = a;
            this.g("SubtitleTextStyle");
        }
        get titleTextStyle() {
            return this.lj;
        }
        set titleTextStyle(a) {
            this.lj = a;
            this.g("TitleTextStyle");
        }
        get dataToolTipGroupingMode() {
            return this.i2;
        }
        set dataToolTipGroupingMode(a) {
            this.i2 = a;
            this.g("DataToolTipGroupingMode");
        }
        get dataToolTipPositionOffsetX() {
            return this.da;
        }
        set dataToolTipPositionOffsetX(a) {
            this.da = a;
            this.g("DataToolTipPositionOffsetX");
        }
        get dataToolTipPositionOffsetY() {
            return this.db;
        }
        set dataToolTipPositionOffsetY(a) {
            this.db = a;
            this.g("DataToolTipPositionOffsetY");
        }
        get dataToolTipDefaultPositionOffsetX() {
            return this.cu;
        }
        set dataToolTipDefaultPositionOffsetX(a) {
            this.cu = a;
            this.g("DataToolTipDefaultPositionOffsetX");
        }
        get dataToolTipDefaultPositionOffsetY() {
            return this.cv;
        }
        set dataToolTipDefaultPositionOffsetY(a) {
            this.cv = a;
            this.g("DataToolTipDefaultPositionOffsetY");
        }
        get dataToolTipGroupedPositionModeX() {
            return this.i0;
        }
        set dataToolTipGroupedPositionModeX(a) {
            this.i0 = a;
            this.g("DataToolTipGroupedPositionModeX");
        }
        get dataToolTipGroupedPositionModeY() {
            return this.i1;
        }
        set dataToolTipGroupedPositionModeY(a) {
            this.i1 = a;
            this.g("DataToolTipGroupedPositionModeY");
        }
        get dataToolTipShouldUpdateWhenSeriesDataChanges() {
            return this.bn;
        }
        set dataToolTipShouldUpdateWhenSeriesDataChanges(a) {
            this.bn = a;
            this.g("DataToolTipShouldUpdateWhenSeriesDataChanges");
        }
        get dataToolTipIncludedSeries() {
            return this.p;
        }
        set dataToolTipIncludedSeries(a) {
            this.p = a;
            this.g("DataToolTipIncludedSeries");
        }
        get dataToolTipExcludedSeries() {
            return this.m;
        }
        set dataToolTipExcludedSeries(a) {
            this.m = a;
            this.g("DataToolTipExcludedSeries");
        }
        get dataToolTipIncludedColumns() {
            return this.o;
        }
        set dataToolTipIncludedColumns(a) {
            this.o = a;
            this.g("DataToolTipIncludedColumns");
        }
        get dataToolTipExcludedColumns() {
            return this.l;
        }
        set dataToolTipExcludedColumns(a) {
            this.l = a;
            this.g("DataToolTipExcludedColumns");
        }
        get dataToolTipValueFormatAbbreviation() {
            return this.kb;
        }
        set dataToolTipValueFormatAbbreviation(a) {
            this.kb = a;
            this.g("DataToolTipValueFormatAbbreviation");
        }
        get dataToolTipValueFormatMaxFractions() {
            return this.h2;
        }
        set dataToolTipValueFormatMaxFractions(a) {
            this.h2 = a;
            this.g("DataToolTipValueFormatMaxFractions");
        }
        get dataToolTipValueFormatMinFractions() {
            return this.h3;
        }
        set dataToolTipValueFormatMinFractions(a) {
            this.h3 = a;
            this.g("DataToolTipValueFormatMinFractions");
        }
        get dataToolTipValueFormatMode() {
            return this.kd;
        }
        set dataToolTipValueFormatMode(a) {
            this.kd = a;
            this.g("DataToolTipValueFormatMode");
        }
        get dataToolTipValueFormatCulture() {
            return this.kc;
        }
        set dataToolTipValueFormatCulture(a) {
            this.kc = a;
            this.g("DataToolTipValueFormatCulture");
        }
        get dataToolTipValueFormatUseGrouping() {
            return this.bo;
        }
        set dataToolTipValueFormatUseGrouping(a) {
            this.bo = a;
            this.g("DataToolTipValueFormatUseGrouping");
        }
        get dataToolTipValueFormatString() {
            return this.ke;
        }
        set dataToolTipValueFormatString(a) {
            this.ke = a;
            this.g("DataToolTipValueFormatString");
        }
        get dataToolTipValueFormatSpecifiers() {
            return this.q;
        }
        set dataToolTipValueFormatSpecifiers(a) {
            this.q = a;
            this.g("DataToolTipValueFormatSpecifiers");
        }
        get dataToolTipValueRowMarginBottom() {
            return this.dz;
        }
        set dataToolTipValueRowMarginBottom(a) {
            this.dz = a;
            this.g("DataToolTipValueRowMarginBottom");
        }
        get dataToolTipValueRowMarginLeft() {
            return this.d0;
        }
        set dataToolTipValueRowMarginLeft(a) {
            this.d0 = a;
            this.g("DataToolTipValueRowMarginLeft");
        }
        get dataToolTipValueRowMarginRight() {
            return this.d1;
        }
        set dataToolTipValueRowMarginRight(a) {
            this.d1 = a;
            this.g("DataToolTipValueRowMarginRight");
        }
        get dataToolTipValueRowMarginTop() {
            return this.d2;
        }
        set dataToolTipValueRowMarginTop(a) {
            this.d2 = a;
            this.g("DataToolTipValueRowMarginTop");
        }
        get dataToolTipValueRowVisible() {
            return this.bp;
        }
        set dataToolTipValueRowVisible(a) {
            this.bp = a;
            this.g("DataToolTipValueRowVisible");
        }
        get dataToolTipValueTextWhenMissingData() {
            return this.kj;
        }
        set dataToolTipValueTextWhenMissingData(a) {
            this.kj = a;
            this.g("DataToolTipValueTextWhenMissingData");
        }
        get dataToolTipValueTextUseSeriesColors() {
            return this.bq;
        }
        set dataToolTipValueTextUseSeriesColors(a) {
            this.bq = a;
            this.g("DataToolTipValueTextUseSeriesColors");
        }
        get dataToolTipValueTextMarginBottom() {
            return this.d4;
        }
        set dataToolTipValueTextMarginBottom(a) {
            this.d4 = a;
            this.g("DataToolTipValueTextMarginBottom");
        }
        get dataToolTipValueTextMarginLeft() {
            return this.d5;
        }
        set dataToolTipValueTextMarginLeft(a) {
            this.d5 = a;
            this.g("DataToolTipValueTextMarginLeft");
        }
        get dataToolTipValueTextMarginRight() {
            return this.d6;
        }
        set dataToolTipValueTextMarginRight(a) {
            this.d6 = a;
            this.g("DataToolTipValueTextMarginRight");
        }
        get dataToolTipValueTextMarginTop() {
            return this.d7;
        }
        set dataToolTipValueTextMarginTop(a) {
            this.d7 = a;
            this.g("DataToolTipValueTextMarginTop");
        }
        get dataToolTipValueTextColor() {
            return this.kf;
        }
        set dataToolTipValueTextColor(a) {
            this.kf = a;
            this.g("DataToolTipValueTextColor");
        }
        get dataToolTipValueTextFontFamily() {
            return this.kg;
        }
        set dataToolTipValueTextFontFamily(a) {
            this.kg = a;
            this.g("DataToolTipValueTextFontFamily");
        }
        get dataToolTipValueTextFontSize() {
            return this.d3;
        }
        set dataToolTipValueTextFontSize(a) {
            this.d3 = a;
            this.g("DataToolTipValueTextFontSize");
        }
        get dataToolTipValueTextFontStyle() {
            return this.kh;
        }
        set dataToolTipValueTextFontStyle(a) {
            this.kh = a;
            this.g("DataToolTipValueTextFontStyle");
        }
        get dataToolTipValueTextFontWeight() {
            return this.ki;
        }
        set dataToolTipValueTextFontWeight(a) {
            this.ki = a;
            this.g("DataToolTipValueTextFontWeight");
        }
        get dataToolTipHeaderFormatString() {
            return this.i5;
        }
        set dataToolTipHeaderFormatString(a) {
            this.i5 = a;
            this.g("DataToolTipHeaderFormatString");
        }
        get dataToolTipHeaderFormatSpecifiers() {
            return this.n;
        }
        set dataToolTipHeaderFormatSpecifiers(a) {
            this.n = a;
            this.g("DataToolTipHeaderFormatSpecifiers");
        }
        get dataToolTipHeaderFormatCulture() {
            return this.i3;
        }
        set dataToolTipHeaderFormatCulture(a) {
            this.i3 = a;
            this.g("DataToolTipHeaderFormatCulture");
        }
        get dataToolTipHeaderFormatDate() {
            return this.i4;
        }
        set dataToolTipHeaderFormatDate(a) {
            this.i4 = a;
            this.g("DataToolTipHeaderFormatDate");
        }
        get dataToolTipHeaderFormatTime() {
            return this.i6;
        }
        set dataToolTipHeaderFormatTime(a) {
            this.i6 = a;
            this.g("DataToolTipHeaderFormatTime");
        }
        get dataToolTipHeaderText() {
            return this.i7;
        }
        set dataToolTipHeaderText(a) {
            this.i7 = a;
            this.g("DataToolTipHeaderText");
        }
        get dataToolTipHeaderTextColor() {
            return this.i8;
        }
        set dataToolTipHeaderTextColor(a) {
            this.i8 = a;
            this.g("DataToolTipHeaderTextColor");
        }
        get dataToolTipHeaderTextMarginBottom() {
            return this.c1;
        }
        set dataToolTipHeaderTextMarginBottom(a) {
            this.c1 = a;
            this.g("DataToolTipHeaderTextMarginBottom");
        }
        get dataToolTipHeaderTextMarginLeft() {
            return this.c2;
        }
        set dataToolTipHeaderTextMarginLeft(a) {
            this.c2 = a;
            this.g("DataToolTipHeaderTextMarginLeft");
        }
        get dataToolTipHeaderTextMarginRight() {
            return this.c3;
        }
        set dataToolTipHeaderTextMarginRight(a) {
            this.c3 = a;
            this.g("DataToolTipHeaderTextMarginRight");
        }
        get dataToolTipHeaderTextMarginTop() {
            return this.c4;
        }
        set dataToolTipHeaderTextMarginTop(a) {
            this.c4 = a;
            this.g("DataToolTipHeaderTextMarginTop");
        }
        get dataToolTipHeaderRowMarginBottom() {
            return this.cw;
        }
        set dataToolTipHeaderRowMarginBottom(a) {
            this.cw = a;
            this.g("DataToolTipHeaderRowMarginBottom");
        }
        get dataToolTipHeaderRowMarginLeft() {
            return this.cx;
        }
        set dataToolTipHeaderRowMarginLeft(a) {
            this.cx = a;
            this.g("DataToolTipHeaderRowMarginLeft");
        }
        get dataToolTipHeaderRowMarginRight() {
            return this.cy;
        }
        set dataToolTipHeaderRowMarginRight(a) {
            this.cy = a;
            this.g("DataToolTipHeaderRowMarginRight");
        }
        get dataToolTipHeaderRowMarginTop() {
            return this.cz;
        }
        set dataToolTipHeaderRowMarginTop(a) {
            this.cz = a;
            this.g("DataToolTipHeaderRowMarginTop");
        }
        get dataToolTipHeaderRowVisible() {
            return this.bm;
        }
        set dataToolTipHeaderRowVisible(a) {
            this.bm = a;
            this.g("DataToolTipHeaderRowVisible");
        }
        get dataToolTipHeaderTextFontFamily() {
            return this.i9;
        }
        set dataToolTipHeaderTextFontFamily(a) {
            this.i9 = a;
            this.g("DataToolTipHeaderTextFontFamily");
        }
        get dataToolTipHeaderTextFontSize() {
            return this.c0;
        }
        set dataToolTipHeaderTextFontSize(a) {
            this.c0 = a;
            this.g("DataToolTipHeaderTextFontSize");
        }
        get dataToolTipHeaderTextFontStyle() {
            return this.ja;
        }
        set dataToolTipHeaderTextFontStyle(a) {
            this.ja = a;
            this.g("DataToolTipHeaderTextFontStyle");
        }
        get dataToolTipHeaderTextFontWeight() {
            return this.jb;
        }
        set dataToolTipHeaderTextFontWeight(a) {
            this.jb = a;
            this.g("DataToolTipHeaderTextFontWeight");
        }
        get dataToolTipSummaryTitleTextColor() {
            return this.jn;
        }
        set dataToolTipSummaryTitleTextColor(a) {
            this.jn = a;
            this.g("DataToolTipSummaryTitleTextColor");
        }
        get dataToolTipSummaryTitleTextFontFamily() {
            return this.jo;
        }
        set dataToolTipSummaryTitleTextFontFamily(a) {
            this.jo = a;
            this.g("DataToolTipSummaryTitleTextFontFamily");
        }
        get dataToolTipSummaryTitleTextFontSize() {
            return this.dh;
        }
        set dataToolTipSummaryTitleTextFontSize(a) {
            this.dh = a;
            this.g("DataToolTipSummaryTitleTextFontSize");
        }
        get dataToolTipSummaryTitleTextFontStyle() {
            return this.jp;
        }
        set dataToolTipSummaryTitleTextFontStyle(a) {
            this.jp = a;
            this.g("DataToolTipSummaryTitleTextFontStyle");
        }
        get dataToolTipSummaryTitleTextFontWeight() {
            return this.jq;
        }
        set dataToolTipSummaryTitleTextFontWeight(a) {
            this.jq = a;
            this.g("DataToolTipSummaryTitleTextFontWeight");
        }
        get dataToolTipSummaryType() {
            return this.jr;
        }
        set dataToolTipSummaryType(a) {
            this.jr = a;
            this.g("DataToolTipSummaryType");
        }
        get dataToolTipSummaryTitleText() {
            return this.jm;
        }
        set dataToolTipSummaryTitleText(a) {
            this.jm = a;
            this.g("DataToolTipSummaryTitleText");
        }
        get dataToolTipSummaryTitleTextMarginBottom() {
            return this.di;
        }
        set dataToolTipSummaryTitleTextMarginBottom(a) {
            this.di = a;
            this.g("DataToolTipSummaryTitleTextMarginBottom");
        }
        get dataToolTipSummaryTitleTextMarginLeft() {
            return this.dj;
        }
        set dataToolTipSummaryTitleTextMarginLeft(a) {
            this.dj = a;
            this.g("DataToolTipSummaryTitleTextMarginLeft");
        }
        get dataToolTipSummaryTitleTextMarginRight() {
            return this.dk;
        }
        set dataToolTipSummaryTitleTextMarginRight(a) {
            this.dk = a;
            this.g("DataToolTipSummaryTitleTextMarginRight");
        }
        get dataToolTipSummaryTitleTextMarginTop() {
            return this.dl;
        }
        set dataToolTipSummaryTitleTextMarginTop(a) {
            this.dl = a;
            this.g("DataToolTipSummaryTitleTextMarginTop");
        }
        get dataToolTipSummaryRowMarginBottom() {
            return this.dd;
        }
        set dataToolTipSummaryRowMarginBottom(a) {
            this.dd = a;
            this.g("DataToolTipSummaryRowMarginBottom");
        }
        get dataToolTipSummaryRowMarginLeft() {
            return this.de;
        }
        set dataToolTipSummaryRowMarginLeft(a) {
            this.de = a;
            this.g("DataToolTipSummaryRowMarginLeft");
        }
        get dataToolTipSummaryRowMarginRight() {
            return this.df;
        }
        set dataToolTipSummaryRowMarginRight(a) {
            this.df = a;
            this.g("DataToolTipSummaryRowMarginRight");
        }
        get dataToolTipSummaryRowMarginTop() {
            return this.dg;
        }
        set dataToolTipSummaryRowMarginTop(a) {
            this.dg = a;
            this.g("DataToolTipSummaryRowMarginTop");
        }
        get dataToolTipSummaryValueTextColor() {
            return this.jx;
        }
        set dataToolTipSummaryValueTextColor(a) {
            this.jx = a;
            this.g("DataToolTipSummaryValueTextColor");
        }
        get dataToolTipSummaryValueTextFontFamily() {
            return this.jy;
        }
        set dataToolTipSummaryValueTextFontFamily(a) {
            this.jy = a;
            this.g("DataToolTipSummaryValueTextFontFamily");
        }
        get dataToolTipSummaryValueTextFontSize() {
            return this.dn;
        }
        set dataToolTipSummaryValueTextFontSize(a) {
            this.dn = a;
            this.g("DataToolTipSummaryValueTextFontSize");
        }
        get dataToolTipSummaryValueTextFontStyle() {
            return this.jz;
        }
        set dataToolTipSummaryValueTextFontStyle(a) {
            this.jz = a;
            this.g("DataToolTipSummaryValueTextFontStyle");
        }
        get dataToolTipSummaryValueTextFontWeight() {
            return this.j0;
        }
        set dataToolTipSummaryValueTextFontWeight(a) {
            this.j0 = a;
            this.g("DataToolTipSummaryValueTextFontWeight");
        }
        get dataToolTipSummaryLabelText() {
            return this.jh;
        }
        set dataToolTipSummaryLabelText(a) {
            this.jh = a;
            this.g("DataToolTipSummaryLabelText");
        }
        get dataToolTipSummaryLabelTextColor() {
            return this.ji;
        }
        set dataToolTipSummaryLabelTextColor(a) {
            this.ji = a;
            this.g("DataToolTipSummaryLabelTextColor");
        }
        get dataToolTipSummaryLabelTextFontFamily() {
            return this.jj;
        }
        set dataToolTipSummaryLabelTextFontFamily(a) {
            this.jj = a;
            this.g("DataToolTipSummaryLabelTextFontFamily");
        }
        get dataToolTipSummaryLabelTextFontSize() {
            return this.dc;
        }
        set dataToolTipSummaryLabelTextFontSize(a) {
            this.dc = a;
            this.g("DataToolTipSummaryLabelTextFontSize");
        }
        get dataToolTipSummaryLabelTextFontStyle() {
            return this.jk;
        }
        set dataToolTipSummaryLabelTextFontStyle(a) {
            this.jk = a;
            this.g("DataToolTipSummaryLabelTextFontStyle");
        }
        get dataToolTipSummaryLabelTextFontWeight() {
            return this.jl;
        }
        set dataToolTipSummaryLabelTextFontWeight(a) {
            this.jl = a;
            this.g("DataToolTipSummaryLabelTextFontWeight");
        }
        get dataToolTipSummaryUnitsText() {
            return this.js;
        }
        set dataToolTipSummaryUnitsText(a) {
            this.js = a;
            this.g("DataToolTipSummaryUnitsText");
        }
        get dataToolTipSummaryUnitsTextColor() {
            return this.jt;
        }
        set dataToolTipSummaryUnitsTextColor(a) {
            this.jt = a;
            this.g("DataToolTipSummaryUnitsTextColor");
        }
        get dataToolTipSummaryUnitsTextFontFamily() {
            return this.ju;
        }
        set dataToolTipSummaryUnitsTextFontFamily(a) {
            this.ju = a;
            this.g("DataToolTipSummaryUnitsTextFontFamily");
        }
        get dataToolTipSummaryUnitsTextFontSize() {
            return this.dm;
        }
        set dataToolTipSummaryUnitsTextFontSize(a) {
            this.dm = a;
            this.g("DataToolTipSummaryUnitsTextFontSize");
        }
        get dataToolTipSummaryUnitsTextFontStyle() {
            return this.jv;
        }
        set dataToolTipSummaryUnitsTextFontStyle(a) {
            this.jv = a;
            this.g("DataToolTipSummaryUnitsTextFontStyle");
        }
        get dataToolTipSummaryUnitsTextFontWeight() {
            return this.jw;
        }
        set dataToolTipSummaryUnitsTextFontWeight(a) {
            this.jw = a;
            this.g("DataToolTipSummaryUnitsTextFontWeight");
        }
        get dataToolTipBadgeMarginBottom() {
            return this.cq;
        }
        set dataToolTipBadgeMarginBottom(a) {
            this.cq = a;
            this.g("DataToolTipBadgeMarginBottom");
        }
        get dataToolTipBadgeMarginLeft() {
            return this.cr;
        }
        set dataToolTipBadgeMarginLeft(a) {
            this.cr = a;
            this.g("DataToolTipBadgeMarginLeft");
        }
        get dataToolTipBadgeMarginRight() {
            return this.cs;
        }
        set dataToolTipBadgeMarginRight(a) {
            this.cs = a;
            this.g("DataToolTipBadgeMarginRight");
        }
        get dataToolTipBadgeMarginTop() {
            return this.ct;
        }
        set dataToolTipBadgeMarginTop(a) {
            this.ct = a;
            this.g("DataToolTipBadgeMarginTop");
        }
        get dataToolTipBadgeShape() {
            return this.iz;
        }
        set dataToolTipBadgeShape(a) {
            this.iz = a;
            this.g("DataToolTipBadgeShape");
        }
        get dataToolTipUnitsDisplayMode() {
            return this.j5;
        }
        set dataToolTipUnitsDisplayMode(a) {
            this.j5 = a;
            this.g("DataToolTipUnitsDisplayMode");
        }
        get dataToolTipUnitsText() {
            return this.j6;
        }
        set dataToolTipUnitsText(a) {
            this.j6 = a;
            this.g("DataToolTipUnitsText");
        }
        get dataToolTipUnitsTextColor() {
            return this.j7;
        }
        set dataToolTipUnitsTextColor(a) {
            this.j7 = a;
            this.g("DataToolTipUnitsTextColor");
        }
        get dataToolTipUnitsTextMarginBottom() {
            return this.dv;
        }
        set dataToolTipUnitsTextMarginBottom(a) {
            this.dv = a;
            this.g("DataToolTipUnitsTextMarginBottom");
        }
        get dataToolTipUnitsTextMarginLeft() {
            return this.dw;
        }
        set dataToolTipUnitsTextMarginLeft(a) {
            this.dw = a;
            this.g("DataToolTipUnitsTextMarginLeft");
        }
        get dataToolTipUnitsTextMarginRight() {
            return this.dx;
        }
        set dataToolTipUnitsTextMarginRight(a) {
            this.dx = a;
            this.g("DataToolTipUnitsTextMarginRight");
        }
        get dataToolTipUnitsTextMarginTop() {
            return this.dy;
        }
        set dataToolTipUnitsTextMarginTop(a) {
            this.dy = a;
            this.g("DataToolTipUnitsTextMarginTop");
        }
        get dataToolTipUnitsTextFontFamily() {
            return this.j8;
        }
        set dataToolTipUnitsTextFontFamily(a) {
            this.j8 = a;
            this.g("DataToolTipUnitsTextFontFamily");
        }
        get dataToolTipUnitsTextFontSize() {
            return this.du;
        }
        set dataToolTipUnitsTextFontSize(a) {
            this.du = a;
            this.g("DataToolTipUnitsTextFontSize");
        }
        get dataToolTipUnitsTextFontStyle() {
            return this.j9;
        }
        set dataToolTipUnitsTextFontStyle(a) {
            this.j9 = a;
            this.g("DataToolTipUnitsTextFontStyle");
        }
        get dataToolTipUnitsTextFontWeight() {
            return this.ka;
        }
        set dataToolTipUnitsTextFontWeight(a) {
            this.ka = a;
            this.g("DataToolTipUnitsTextFontWeight");
        }
        get dataToolTipTitleTextMarginBottom() {
            return this.dq;
        }
        set dataToolTipTitleTextMarginBottom(a) {
            this.dq = a;
            this.g("DataToolTipTitleTextMarginBottom");
        }
        get dataToolTipTitleTextMarginLeft() {
            return this.dr;
        }
        set dataToolTipTitleTextMarginLeft(a) {
            this.dr = a;
            this.g("DataToolTipTitleTextMarginLeft");
        }
        get dataToolTipTitleTextMarginRight() {
            return this.ds;
        }
        set dataToolTipTitleTextMarginRight(a) {
            this.ds = a;
            this.g("DataToolTipTitleTextMarginRight");
        }
        get dataToolTipTitleTextMarginTop() {
            return this.dt;
        }
        set dataToolTipTitleTextMarginTop(a) {
            this.dt = a;
            this.g("DataToolTipTitleTextMarginTop");
        }
        get dataToolTipTitleTextColor() {
            return this.j1;
        }
        set dataToolTipTitleTextColor(a) {
            this.j1 = a;
            this.g("DataToolTipTitleTextColor");
        }
        get dataToolTipTitleTextFontFamily() {
            return this.j2;
        }
        set dataToolTipTitleTextFontFamily(a) {
            this.j2 = a;
            this.g("DataToolTipTitleTextFontFamily");
        }
        get dataToolTipTitleTextFontSize() {
            return this.dp;
        }
        set dataToolTipTitleTextFontSize(a) {
            this.dp = a;
            this.g("DataToolTipTitleTextFontSize");
        }
        get dataToolTipTitleTextFontStyle() {
            return this.j3;
        }
        set dataToolTipTitleTextFontStyle(a) {
            this.j3 = a;
            this.g("DataToolTipTitleTextFontStyle");
        }
        get dataToolTipTitleTextFontWeight() {
            return this.j4;
        }
        set dataToolTipTitleTextFontWeight(a) {
            this.j4 = a;
            this.g("DataToolTipTitleTextFontWeight");
        }
        get dataToolTipLabelDisplayMode() {
            return this.jc;
        }
        set dataToolTipLabelDisplayMode(a) {
            this.jc = a;
            this.g("DataToolTipLabelDisplayMode");
        }
        get dataToolTipLabelTextColor() {
            return this.jd;
        }
        set dataToolTipLabelTextColor(a) {
            this.jd = a;
            this.g("DataToolTipLabelTextColor");
        }
        get dataToolTipLabelTextMarginBottom() {
            return this.c6;
        }
        set dataToolTipLabelTextMarginBottom(a) {
            this.c6 = a;
            this.g("DataToolTipLabelTextMarginBottom");
        }
        get dataToolTipLabelTextMarginLeft() {
            return this.c7;
        }
        set dataToolTipLabelTextMarginLeft(a) {
            this.c7 = a;
            this.g("DataToolTipLabelTextMarginLeft");
        }
        get dataToolTipLabelTextMarginRight() {
            return this.c8;
        }
        set dataToolTipLabelTextMarginRight(a) {
            this.c8 = a;
            this.g("DataToolTipLabelTextMarginRight");
        }
        get dataToolTipLabelTextMarginTop() {
            return this.c9;
        }
        set dataToolTipLabelTextMarginTop(a) {
            this.c9 = a;
            this.g("DataToolTipLabelTextMarginTop");
        }
        get dataToolTipLabelTextFontFamily() {
            return this.je;
        }
        set dataToolTipLabelTextFontFamily(a) {
            this.je = a;
            this.g("DataToolTipLabelTextFontFamily");
        }
        get dataToolTipLabelTextFontSize() {
            return this.c5;
        }
        set dataToolTipLabelTextFontSize(a) {
            this.c5 = a;
            this.g("DataToolTipLabelTextFontSize");
        }
        get dataToolTipLabelTextFontStyle() {
            return this.jf;
        }
        set dataToolTipLabelTextFontStyle(a) {
            this.jf = a;
            this.g("DataToolTipLabelTextFontStyle");
        }
        get dataToolTipLabelTextFontWeight() {
            return this.jg;
        }
        set dataToolTipLabelTextFontWeight(a) {
            this.jg = a;
            this.g("DataToolTipLabelTextFontWeight");
        }
        get sortDescriptions() {
            return this.ae;
        }
        set sortDescriptions(a) {
            this.ae = a;
            this.g("SortDescriptions");
        }
        get groupSortDescriptions() {
            return this.u;
        }
        set groupSortDescriptions(a) {
            this.u = a;
            this.g("GroupSortDescriptions");
        }
        get groupDescriptions() {
            return this.t;
        }
        set groupDescriptions(a) {
            this.t = a;
            this.g("GroupDescriptions");
        }
        get filterExpressions() {
            return this.s;
        }
        set filterExpressions(a) {
            this.s = a;
            this.g("FilterExpressions");
        }
        get summaryDescriptions() {
            return this.af;
        }
        set summaryDescriptions(a) {
            this.af = a;
            this.g("SummaryDescriptions");
        }
        get initialSortDescriptions() {
            return this.y;
        }
        set initialSortDescriptions(a) {
            this.y = a;
            this.g("InitialSortDescriptions");
        }
        get initialGroupSortDescriptions() {
            return this.x;
        }
        set initialGroupSortDescriptions(a) {
            this.x = a;
            this.g("InitialGroupSortDescriptions");
        }
        get initialGroupDescriptions() {
            return this.w;
        }
        set initialGroupDescriptions(a) {
            this.w = a;
            this.g("InitialGroupDescriptions");
        }
        get initialSummaryDescriptions() {
            return this.z;
        }
        set initialSummaryDescriptions(a) {
            this.z = a;
            this.g("InitialSummaryDescriptions");
        }
        get initialSorts() {
            return this.ku;
        }
        set initialSorts(a) {
            this.ku = a;
            this.g("InitialSorts");
        }
        get groupSorts() {
            return this.km;
        }
        set groupSorts(a) {
            this.km = a;
            this.g("GroupSorts");
        }
        get initialGroups() {
            return this.kt;
        }
        set initialGroups(a) {
            this.kt = a;
            this.g("InitialGroups");
        }
        get initialSummaries() {
            return this.kv;
        }
        set initialSummaries(a) {
            this.kv = a;
            this.g("InitialSummaries");
        }
        get dataSourceRef() {
            return this.iy;
        }
        set dataSourceRef(a) {
            this.iy = a;
            this.g("DataSourceRef");
        }
        get dataSource() {
            return this.bb;
        }
        set dataSource(a) {
            this.bb = a;
            this.g("DataSource");
        }
        get includedProperties() {
            return this.v;
        }
        set includedProperties(a) {
            this.v = a;
            this.g("IncludedProperties");
        }
        get excludedProperties() {
            return this.r;
        }
        set excludedProperties(a) {
            this.r = a;
            this.g("ExcludedProperties");
        }
        get brushes() {
            return this.j;
        }
        set brushes(a) {
            this.j = a;
            this.g("Brushes");
        }
        get outlines() {
            return this.ad;
        }
        set outlines(a) {
            this.ad = a;
            this.g("Outlines");
        }
        get actualBrushes() {
            return this.h;
        }
        set actualBrushes(a) {
            this.h = a;
            this.g("ActualBrushes");
        }
        get actualOutlines() {
            return this.i;
        }
        set actualOutlines(a) {
            this.i = a;
            this.g("ActualOutlines");
        }
        get legendRef() {
            return this.kz;
        }
        set legendRef(a) {
            this.kz = a;
            this.g("LegendRef");
        }
        get isHorizontalZoomEnabled() {
            return this.bt;
        }
        set isHorizontalZoomEnabled(a) {
            this.bt = a;
            this.g("IsHorizontalZoomEnabled");
        }
        get isVerticalZoomEnabled() {
            return this.bv;
        }
        set isVerticalZoomEnabled(a) {
            this.bv = a;
            this.g("IsVerticalZoomEnabled");
        }
        get shouldDisplayMockData() {
            return this.bx;
        }
        set shouldDisplayMockData(a) {
            this.bx = a;
            this.g("ShouldDisplayMockData");
        }
        get shouldUseSkeletonStyleForMockData() {
            return this.bz;
        }
        set shouldUseSkeletonStyleForMockData(a) {
            this.bz = a;
            this.g("ShouldUseSkeletonStyleForMockData");
        }
        get isSeriesHighlightingEnabled() {
            return this.bu;
        }
        set isSeriesHighlightingEnabled(a) {
            this.bu = a;
            this.g("IsSeriesHighlightingEnabled");
        }
        get windowRect() {
            return this.bd;
        }
        set windowRect(a) {
            this.bd = a;
            this.g("WindowRect");
        }
        get chartTitle() {
            return this.ip;
        }
        set chartTitle(a) {
            this.ip = a;
            this.g("ChartTitle");
        }
        get subtitle() {
            return this.ld;
        }
        set subtitle(a) {
            this.ld = a;
            this.g("Subtitle");
        }
        get titleAlignment() {
            return this.lh;
        }
        set titleAlignment(a) {
            this.lh = a;
            this.g("TitleAlignment");
        }
        get subtitleAlignment() {
            return this.le;
        }
        set subtitleAlignment(a) {
            this.le = a;
            this.g("SubtitleAlignment");
        }
        get unknownValuePlotting() {
            return this.ln;
        }
        set unknownValuePlotting(a) {
            this.ln = a;
            this.g("UnknownValuePlotting");
        }
        get resolution() {
            return this.en;
        }
        set resolution(a) {
            this.en = a;
            this.g("Resolution");
        }
        get thickness() {
            return this.et;
        }
        set thickness(a) {
            this.et = a;
            this.g("Thickness");
        }
        get outlineMode() {
            return this.k3;
        }
        set outlineMode(a) {
            this.k3 = a;
            this.g("OutlineMode");
        }
        get markerOutlineMode() {
            return this.k2;
        }
        set markerOutlineMode(a) {
            this.k2 = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.k1;
        }
        set markerFillMode(a) {
            this.k1 = a;
            this.g("MarkerFillMode");
        }
        get markerFillOpacity() {
            return this.eg;
        }
        set markerFillOpacity(a) {
            this.eg = a;
            this.g("MarkerFillOpacity");
        }
        get markerThickness() {
            return this.eh;
        }
        set markerThickness(a) {
            this.eh = a;
            this.g("MarkerThickness");
        }
        get markerTypes() {
            return this.ac;
        }
        set markerTypes(a) {
            this.ac = a;
            this.g("MarkerTypes");
        }
        get markerAutomaticBehavior() {
            return this.k0;
        }
        set markerAutomaticBehavior(a) {
            this.k0 = a;
            this.g("MarkerAutomaticBehavior");
        }
        get markerBrushes() {
            return this.aa;
        }
        set markerBrushes(a) {
            this.aa = a;
            this.g("MarkerBrushes");
        }
        get markerOutlines() {
            return this.ab;
        }
        set markerOutlines(a) {
            this.ab = a;
            this.g("MarkerOutlines");
        }
        get markerMaxCount() {
            return this.h5;
        }
        set markerMaxCount(a) {
            this.h5 = a;
            this.g("MarkerMaxCount");
        }
        get areaFillOpacity() {
            return this.cn;
        }
        set areaFillOpacity(a) {
            this.cn = a;
            this.g("AreaFillOpacity");
        }
        get animateSeriesWhenAxisRangeChanges() {
            return this.bg;
        }
        set animateSeriesWhenAxisRangeChanges(a) {
            this.bg = a;
            this.g("AnimateSeriesWhenAxisRangeChanges");
        }
        get trendLineBrushes() {
            return this.ag;
        }
        set trendLineBrushes(a) {
            this.ag = a;
            this.g("TrendLineBrushes");
        }
        get trendLineType() {
            return this.lm;
        }
        set trendLineType(a) {
            this.lm = a;
            this.g("TrendLineType");
        }
        get trendLineThickness() {
            return this.ez;
        }
        set trendLineThickness(a) {
            this.ez = a;
            this.g("TrendLineThickness");
        }
        get alignsGridLinesToPixels() {
            return this.bf;
        }
        set alignsGridLinesToPixels(a) {
            this.bf = a;
            this.g("AlignsGridLinesToPixels");
        }
        get plotAreaMarginLeft() {
            return this.ek;
        }
        set plotAreaMarginLeft(a) {
            this.ek = a;
            this.g("PlotAreaMarginLeft");
        }
        get plotAreaMarginTop() {
            return this.em;
        }
        set plotAreaMarginTop(a) {
            this.em = a;
            this.g("PlotAreaMarginTop");
        }
        get plotAreaMarginRight() {
            return this.el;
        }
        set plotAreaMarginRight(a) {
            this.el = a;
            this.g("PlotAreaMarginRight");
        }
        get plotAreaMarginBottom() {
            return this.ej;
        }
        set plotAreaMarginBottom(a) {
            this.ej = a;
            this.g("PlotAreaMarginBottom");
        }
        get computedPlotAreaMarginMode() {
            return this.iq;
        }
        set computedPlotAreaMarginMode(a) {
            this.iq = a;
            this.g("ComputedPlotAreaMarginMode");
        }
        get seriesPlotAreaMarginHorizontalMode() {
            return this.k5;
        }
        set seriesPlotAreaMarginHorizontalMode(a) {
            this.k5 = a;
            this.g("SeriesPlotAreaMarginHorizontalMode");
        }
        get seriesPlotAreaMarginVerticalMode() {
            return this.k6;
        }
        set seriesPlotAreaMarginVerticalMode(a) {
            this.k6 = a;
            this.g("SeriesPlotAreaMarginVerticalMode");
        }
        get highlightingMode() {
            return this.ko;
        }
        set highlightingMode(a) {
            this.ko = a;
            this.g("HighlightingMode");
        }
        get highlightingBehavior() {
            return this.kn;
        }
        set highlightingBehavior(a) {
            this.kn = a;
            this.g("HighlightingBehavior");
        }
        get legendHighlightingMode() {
            return this.kw;
        }
        set legendHighlightingMode(a) {
            this.kw = a;
            this.g("LegendHighlightingMode");
        }
        get legendItemBadgeShape() {
            return this.ky;
        }
        set legendItemBadgeShape(a) {
            this.ky = a;
            this.g("LegendItemBadgeShape");
        }
        get legendItemBadgeMode() {
            return this.kx;
        }
        set legendItemBadgeMode(a) {
            this.kx = a;
            this.g("LegendItemBadgeMode");
        }
        get trendLinePeriod() {
            return this.h7;
        }
        set trendLinePeriod(a) {
            this.h7 = a;
            this.g("TrendLinePeriod");
        }
        get toolTipType() {
            return this.lk;
        }
        set toolTipType(a) {
            this.lk = a;
            this.g("ToolTipType");
        }
        get crosshairsDisplayMode() {
            return this.iv;
        }
        set crosshairsDisplayMode(a) {
            this.iv = a;
            this.g("CrosshairsDisplayMode");
        }
        get crosshairsSnapToData() {
            return this.bl;
        }
        set crosshairsSnapToData(a) {
            this.bl = a;
            this.g("CrosshairsSnapToData");
        }
        get crosshairsLineVerticalStroke() {
            return this.ix;
        }
        set crosshairsLineVerticalStroke(a) {
            this.ix = a;
            this.g("CrosshairsLineVerticalStroke");
        }
        get crosshairsLineThickness() {
            return this.cp;
        }
        set crosshairsLineThickness(a) {
            this.cp = a;
            this.g("CrosshairsLineThickness");
        }
        get crosshairsLineHorizontalStroke() {
            return this.iw;
        }
        set crosshairsLineHorizontalStroke(a) {
            this.iw = a;
            this.g("CrosshairsLineHorizontalStroke");
        }
        get crosshairsAnnotationEnabled() {
            return this.bk;
        }
        set crosshairsAnnotationEnabled(a) {
            this.bk = a;
            this.g("CrosshairsAnnotationEnabled");
        }
        get crosshairsAnnotationXAxisBackground() {
            return this.ir;
        }
        set crosshairsAnnotationXAxisBackground(a) {
            this.ir = a;
            this.g("CrosshairsAnnotationXAxisBackground");
        }
        get crosshairsAnnotationYAxisBackground() {
            return this.it;
        }
        set crosshairsAnnotationYAxisBackground(a) {
            this.it = a;
            this.g("CrosshairsAnnotationYAxisBackground");
        }
        get crosshairsAnnotationXAxisTextColor() {
            return this.is;
        }
        set crosshairsAnnotationXAxisTextColor(a) {
            this.is = a;
            this.g("CrosshairsAnnotationXAxisTextColor");
        }
        get crosshairsAnnotationYAxisTextColor() {
            return this.iu;
        }
        set crosshairsAnnotationYAxisTextColor(a) {
            this.iu = a;
            this.g("CrosshairsAnnotationYAxisTextColor");
        }
        get crosshairsAnnotationXAxisPrecision() {
            return this.h0;
        }
        set crosshairsAnnotationXAxisPrecision(a) {
            this.h0 = a;
            this.g("CrosshairsAnnotationXAxisPrecision");
        }
        get crosshairsAnnotationYAxisPrecision() {
            return this.h1;
        }
        set crosshairsAnnotationYAxisPrecision(a) {
            this.h1 = a;
            this.g("CrosshairsAnnotationYAxisPrecision");
        }
        get shouldAvoidAxisAnnotationCollisions() {
            return this.bw;
        }
        set shouldAvoidAxisAnnotationCollisions(a) {
            this.bw = a;
            this.g("ShouldAvoidAxisAnnotationCollisions");
        }
        get shouldPanOnMaximumZoom() {
            return this.by;
        }
        set shouldPanOnMaximumZoom(a) {
            this.by = a;
            this.g("ShouldPanOnMaximumZoom");
        }
        get finalValueAnnotationsVisible() {
            return this.br;
        }
        set finalValueAnnotationsVisible(a) {
            this.br = a;
            this.g("FinalValueAnnotationsVisible");
        }
        get finalValueAnnotationsBackground() {
            return this.kk;
        }
        set finalValueAnnotationsBackground(a) {
            this.kk = a;
            this.g("FinalValueAnnotationsBackground");
        }
        get finalValueAnnotationsTextColor() {
            return this.kl;
        }
        set finalValueAnnotationsTextColor(a) {
            this.kl = a;
            this.g("FinalValueAnnotationsTextColor");
        }
        get finalValueAnnotationsPrecision() {
            return this.h4;
        }
        set finalValueAnnotationsPrecision(a) {
            this.h4 = a;
            this.g("FinalValueAnnotationsPrecision");
        }
        get autoCalloutsVisible() {
            return this.bh;
        }
        set autoCalloutsVisible(a) {
            this.bh = a;
            this.g("AutoCalloutsVisible");
        }
        get calloutsVisible() {
            return this.bj;
        }
        set calloutsVisible(a) {
            this.bj = a;
            this.g("CalloutsVisible");
        }
        get calloutStyleUpdatingEventEnabled() {
            return this.bi;
        }
        set calloutStyleUpdatingEventEnabled(a) {
            this.bi = a;
            this.g("CalloutStyleUpdatingEventEnabled");
        }
        get calloutsDataSourceRef() {
            return this.ij;
        }
        set calloutsDataSourceRef(a) {
            this.ij = a;
            this.g("CalloutsDataSourceRef");
        }
        get calloutsAllowedPositions() {
            return this.k;
        }
        set calloutsAllowedPositions(a) {
            this.k = a;
            this.g("CalloutsAllowedPositions");
        }
        get calloutsXMemberPath() {
            return this.im;
        }
        set calloutsXMemberPath(a) {
            this.im = a;
            this.g("CalloutsXMemberPath");
        }
        get calloutsYMemberPath() {
            return this.io;
        }
        set calloutsYMemberPath(a) {
            this.io = a;
            this.g("CalloutsYMemberPath");
        }
        get calloutsLabelMemberPath() {
            return this.ik;
        }
        set calloutsLabelMemberPath(a) {
            this.ik = a;
            this.g("CalloutsLabelMemberPath");
        }
        get calloutsContentMemberPath() {
            return this.ii;
        }
        set calloutsContentMemberPath(a) {
            this.ii = a;
            this.g("CalloutsContentMemberPath");
        }
        get valueLines() {
            return this.ah;
        }
        set valueLines(a) {
            this.ah = a;
            this.g("ValueLines");
        }
        get valueLinesBrushes() {
            return this.ai;
        }
        set valueLinesBrushes(a) {
            this.ai = a;
            this.g("ValueLinesBrushes");
        }
        get valueLinesThickness() {
            return this.e0;
        }
        set valueLinesThickness(a) {
            this.e0 = a;
            this.g("ValueLinesThickness");
        }
        get horizontalViewScrollbarMode() {
            return this.kq;
        }
        set horizontalViewScrollbarMode(a) {
            this.kq = a;
            this.g("HorizontalViewScrollbarMode");
        }
        get verticalViewScrollbarMode() {
            return this.lp;
        }
        set verticalViewScrollbarMode(a) {
            this.lp = a;
            this.g("VerticalViewScrollbarMode");
        }
        get horizontalViewScrollbarPosition() {
            return this.ks;
        }
        set horizontalViewScrollbarPosition(a) {
            this.ks = a;
            this.g("HorizontalViewScrollbarPosition");
        }
        get verticalViewScrollbarPosition() {
            return this.lr;
        }
        set verticalViewScrollbarPosition(a) {
            this.lr = a;
            this.g("VerticalViewScrollbarPosition");
        }
        get horizontalViewScrollbarFill() {
            return this.kp;
        }
        set horizontalViewScrollbarFill(a) {
            this.kp = a;
            this.g("HorizontalViewScrollbarFill");
        }
        get horizontalViewScrollbarOutline() {
            return this.kr;
        }
        set horizontalViewScrollbarOutline(a) {
            this.kr = a;
            this.g("HorizontalViewScrollbarOutline");
        }
        get horizontalViewScrollbarStrokeThickness() {
            return this.ec;
        }
        set horizontalViewScrollbarStrokeThickness(a) {
            this.ec = a;
            this.g("HorizontalViewScrollbarStrokeThickness");
        }
        get horizontalViewScrollbarMaxOpacity() {
            return this.eb;
        }
        set horizontalViewScrollbarMaxOpacity(a) {
            this.eb = a;
            this.g("HorizontalViewScrollbarMaxOpacity");
        }
        get horizontalViewScrollbarCornerRadius() {
            return this.d8;
        }
        set horizontalViewScrollbarCornerRadius(a) {
            this.d8 = a;
            this.g("HorizontalViewScrollbarCornerRadius");
        }
        get horizontalViewScrollbarHeight() {
            return this.d9;
        }
        set horizontalViewScrollbarHeight(a) {
            this.d9 = a;
            this.g("HorizontalViewScrollbarHeight");
        }
        get horizontalViewScrollbarInset() {
            return this.ea;
        }
        set horizontalViewScrollbarInset(a) {
            this.ea = a;
            this.g("HorizontalViewScrollbarInset");
        }
        get horizontalViewScrollbarTrackStartInset() {
            return this.ee;
        }
        set horizontalViewScrollbarTrackStartInset(a) {
            this.ee = a;
            this.g("HorizontalViewScrollbarTrackStartInset");
        }
        get horizontalViewScrollbarTrackEndInset() {
            return this.ed;
        }
        set horizontalViewScrollbarTrackEndInset(a) {
            this.ed = a;
            this.g("HorizontalViewScrollbarTrackEndInset");
        }
        get horizontalViewScrollbarShouldAddAutoTrackInsets() {
            return this.bs;
        }
        set horizontalViewScrollbarShouldAddAutoTrackInsets(a) {
            this.bs = a;
            this.g("HorizontalViewScrollbarShouldAddAutoTrackInsets");
        }
        get verticalViewScrollbarFill() {
            return this.lo;
        }
        set verticalViewScrollbarFill(a) {
            this.lo = a;
            this.g("VerticalViewScrollbarFill");
        }
        get verticalViewScrollbarOutline() {
            return this.lq;
        }
        set verticalViewScrollbarOutline(a) {
            this.lq = a;
            this.g("VerticalViewScrollbarOutline");
        }
        get verticalViewScrollbarStrokeThickness() {
            return this.e4;
        }
        set verticalViewScrollbarStrokeThickness(a) {
            this.e4 = a;
            this.g("VerticalViewScrollbarStrokeThickness");
        }
        get verticalViewScrollbarMaxOpacity() {
            return this.e3;
        }
        set verticalViewScrollbarMaxOpacity(a) {
            this.e3 = a;
            this.g("VerticalViewScrollbarMaxOpacity");
        }
        get verticalViewScrollbarCornerRadius() {
            return this.e1;
        }
        set verticalViewScrollbarCornerRadius(a) {
            this.e1 = a;
            this.g("VerticalViewScrollbarCornerRadius");
        }
        get verticalViewScrollbarWidth() {
            return this.e7;
        }
        set verticalViewScrollbarWidth(a) {
            this.e7 = a;
            this.g("VerticalViewScrollbarWidth");
        }
        get verticalViewScrollbarInset() {
            return this.e2;
        }
        set verticalViewScrollbarInset(a) {
            this.e2 = a;
            this.g("VerticalViewScrollbarInset");
        }
        get verticalViewScrollbarTrackStartInset() {
            return this.e6;
        }
        set verticalViewScrollbarTrackStartInset(a) {
            this.e6 = a;
            this.g("VerticalViewScrollbarTrackStartInset");
        }
        get verticalViewScrollbarTrackEndInset() {
            return this.e5;
        }
        set verticalViewScrollbarTrackEndInset(a) {
            this.e5 = a;
            this.g("VerticalViewScrollbarTrackEndInset");
        }
        get verticalViewScrollbarShouldAddAutoTrackInsets() {
            return this.b0;
        }
        set verticalViewScrollbarShouldAddAutoTrackInsets(a) {
            this.b0 = a;
            this.g("VerticalViewScrollbarShouldAddAutoTrackInsets");
        }
        get windowRectMinWidth() {
            return this.e9;
        }
        set windowRectMinWidth(a) {
            this.e9 = a;
            this.g("WindowRectMinWidth");
        }
        get windowRectMinHeight() {
            return this.e8;
        }
        set windowRectMinHeight(a) {
            this.e8 = a;
            this.g("WindowRectMinHeight");
        }
        get windowSizeMinWidth() {
            return this.fb;
        }
        set windowSizeMinWidth(a) {
            this.fb = a;
            this.g("WindowSizeMinWidth");
        }
        get windowSizeMinHeight() {
            return this.fa;
        }
        set windowSizeMinHeight(a) {
            this.fa = a;
            this.g("WindowSizeMinHeight");
        }
        get seriesAddedRef() {
            return this.k4;
        }
        set seriesAddedRef(a) {
            this.k4 = a;
            this.g("SeriesAddedRef");
        }
        get seriesRemovedRef() {
            return this.lc;
        }
        set seriesRemovedRef(a) {
            this.lc = a;
            this.g("SeriesRemovedRef");
        }
        get seriesPointerEnterRef() {
            return this.k8;
        }
        set seriesPointerEnterRef(a) {
            this.k8 = a;
            this.g("SeriesPointerEnterRef");
        }
        get seriesPointerLeaveRef() {
            return this.k9;
        }
        set seriesPointerLeaveRef(a) {
            this.k9 = a;
            this.g("SeriesPointerLeaveRef");
        }
        get seriesPointerMoveRef() {
            return this.la;
        }
        set seriesPointerMoveRef(a) {
            this.la = a;
            this.g("SeriesPointerMoveRef");
        }
        get seriesPointerDownRef() {
            return this.k7;
        }
        set seriesPointerDownRef(a) {
            this.k7 = a;
            this.g("SeriesPointerDownRef");
        }
        get seriesPointerUpRef() {
            return this.lb;
        }
        set seriesPointerUpRef(a) {
            this.lb = a;
            this.g("SeriesPointerUpRef");
        }
        get calloutStyleUpdatingRef() {
            return this.il;
        }
        set calloutStyleUpdatingRef(a) {
            this.il = a;
            this.g("CalloutStyleUpdatingRef");
        }
        get calloutRenderStyleUpdatingRef() {
            return this.ih;
        }
        set calloutRenderStyleUpdatingRef(a) {
            this.ih = a;
            this.g("CalloutRenderStyleUpdatingRef");
        }
    }
    DomainChartDescription.$t = /*@__PURE__*/ markType(DomainChartDescription, 'DomainChartDescription', Description.$);
    return DomainChartDescription;
})();
