// igniteui-react-core/esm2015/lib/FinancialSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FinancialSeriesDescription = /*@__PURE__*/ (() => {
    class FinancialSeriesDescription extends SeriesDescription {
        get_type() {
            return "FinancialSeries";
        }
        constructor() {
            super();
            this.dk = null;
            this.dr = null;
            this.ds = null;
            this.dl = null;
            this.di = null;
            this.dj = null;
            this.dh = null;
            this.dq = null;
            this.dc = false;
            this.dm = null;
            this.dd = false;
            this.dg = null;
            this.dn = null;
            this.dp = null;
        }
        get negativeBrush() {
            return this.dk;
        }
        set negativeBrush(a) {
            this.dk = a;
            this.g("NegativeBrush");
        }
        get xAxisRef() {
            return this.dr;
        }
        set xAxisRef(a) {
            this.dr = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.ds;
        }
        set yAxisRef(a) {
            this.ds = a;
            this.g("YAxisRef");
        }
        get openMemberPath() {
            return this.dl;
        }
        set openMemberPath(a) {
            this.dl = a;
            this.g("OpenMemberPath");
        }
        get highMemberPath() {
            return this.di;
        }
        set highMemberPath(a) {
            this.di = a;
            this.g("HighMemberPath");
        }
        get lowMemberPath() {
            return this.dj;
        }
        set lowMemberPath(a) {
            this.dj = a;
            this.g("LowMemberPath");
        }
        get closeMemberPath() {
            return this.dh;
        }
        set closeMemberPath(a) {
            this.dh = a;
            this.g("CloseMemberPath");
        }
        get volumeMemberPath() {
            return this.dq;
        }
        set volumeMemberPath(a) {
            this.dq = a;
            this.g("VolumeMemberPath");
        }
        get isCustomCategoryStyleAllowed() {
            return this.dc;
        }
        set isCustomCategoryStyleAllowed(a) {
            this.dc = a;
            this.g("IsCustomCategoryStyleAllowed");
        }
        get transitionInMode() {
            return this.dm;
        }
        set transitionInMode(a) {
            this.dm = a;
            this.g("TransitionInMode");
        }
        get isTransitionInEnabled() {
            return this.dd;
        }
        set isTransitionInEnabled(a) {
            this.dd = a;
            this.g("IsTransitionInEnabled");
        }
        get assigningCategoryStyleRef() {
            return this.dg;
        }
        set assigningCategoryStyleRef(a) {
            this.dg = a;
            this.g("AssigningCategoryStyleRef");
        }
        get typicalRef() {
            return this.dn;
        }
        set typicalRef(a) {
            this.dn = a;
            this.g("TypicalRef");
        }
        get typicalBasedOnRef() {
            return this.dp;
        }
        set typicalBasedOnRef(a) {
            this.dp = a;
            this.g("TypicalBasedOnRef");
        }
    }
    FinancialSeriesDescription.$t = /*@__PURE__*/ markType(FinancialSeriesDescription, 'FinancialSeriesDescription', SeriesDescription.$);
    return FinancialSeriesDescription;
})();
