// igniteui-react-core/esm2015/lib/GeographicPolylineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicPolylineSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicPolylineSeriesDescription extends GeographicShapeSeriesBaseDescription {
        get_type() {
            return "GeographicPolylineSeries";
        }
        constructor() {
            super();
            this.d7 = null;
            this.d5 = null;
            this.d6 = null;
            this.d2 = 0;
            this.d1 = 0;
            this.d8 = null;
        }
        get shapeStyleSelectorRef() {
            return this.d7;
        }
        set shapeStyleSelectorRef(a) {
            this.d7 = a;
            this.g("ShapeStyleSelectorRef");
        }
        get shapeFill() {
            return this.d5;
        }
        set shapeFill(a) {
            this.d5 = a;
            this.g("ShapeFill");
        }
        get shapeStroke() {
            return this.d6;
        }
        set shapeStroke(a) {
            this.d6 = a;
            this.g("ShapeStroke");
        }
        get shapeStrokeThickness() {
            return this.d2;
        }
        set shapeStrokeThickness(a) {
            this.d2 = a;
            this.g("ShapeStrokeThickness");
        }
        get shapeOpacity() {
            return this.d1;
        }
        set shapeOpacity(a) {
            this.d1 = a;
            this.g("ShapeOpacity");
        }
        get styleShapeRef() {
            return this.d8;
        }
        set styleShapeRef(a) {
            this.d8 = a;
            this.g("StyleShapeRef");
        }
    }
    GeographicPolylineSeriesDescription.$t = /*@__PURE__*/ markType(GeographicPolylineSeriesDescription, 'GeographicPolylineSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicPolylineSeriesDescription;
})();
