// igniteui-react-core/esm2015/lib/GeographicMarkerSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicMarkerSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicMarkerSeriesDescription extends GeographicMapSeriesHostDescription {
        get_type() {
            return "GeographicMarkerSeries";
        }
        constructor() {
            super();
            this.dt = null;
            this.dg = 0;
            this.dr = null;
            this.dp = null;
            this.ds = null;
            this.dn = null;
            this.dq = null;
            this.di = 0;
            this.dm = null;
            this.dk = null;
            this.dl = null;
        }
        get markerType() {
            return this.dt;
        }
        set markerType(a) {
            this.dt = a;
            this.g("MarkerType");
        }
        get markerThickness() {
            return this.dg;
        }
        set markerThickness(a) {
            this.dg = a;
            this.g("MarkerThickness");
        }
        get markerOutlineMode() {
            return this.dr;
        }
        set markerOutlineMode(a) {
            this.dr = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.dp;
        }
        set markerFillMode(a) {
            this.dp = a;
            this.g("MarkerFillMode");
        }
        get markerTemplateRef() {
            return this.ds;
        }
        set markerTemplateRef(a) {
            this.ds = a;
            this.g("MarkerTemplateRef");
        }
        get markerBrush() {
            return this.dn;
        }
        set markerBrush(a) {
            this.dn = a;
            this.g("MarkerBrush");
        }
        get markerOutline() {
            return this.dq;
        }
        set markerOutline(a) {
            this.dq = a;
            this.g("MarkerOutline");
        }
        get maximumMarkers() {
            return this.di;
        }
        set maximumMarkers(a) {
            this.di = a;
            this.g("MaximumMarkers");
        }
        get actualMarkerTemplateRef() {
            return this.dm;
        }
        set actualMarkerTemplateRef(a) {
            this.dm = a;
            this.g("ActualMarkerTemplateRef");
        }
        get actualMarkerBrush() {
            return this.dk;
        }
        set actualMarkerBrush(a) {
            this.dk = a;
            this.g("ActualMarkerBrush");
        }
        get actualMarkerOutline() {
            return this.dl;
        }
        set actualMarkerOutline(a) {
            this.dl = a;
            this.g("ActualMarkerOutline");
        }
    }
    GeographicMarkerSeriesDescription.$t = /*@__PURE__*/ markType(GeographicMarkerSeriesDescription, 'GeographicMarkerSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicMarkerSeriesDescription;
})();
