// igniteui-react-core/esm2015/lib/GeographicSymbolSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMarkerSeriesDescription } from "./GeographicMarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicSymbolSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicSymbolSeriesDescription extends GeographicMarkerSeriesDescription {
        get_type() {
            return "GeographicSymbolSeries";
        }
        constructor() {
            super();
            this.ec = null;
            this.ed = null;
            this.ee = null;
            this.d4 = false;
            this.d3 = false;
            this.eb = null;
            this.d7 = 0;
            this.ea = null;
            this.d9 = null;
        }
        get latitudeMemberPath() {
            return this.ec;
        }
        set latitudeMemberPath(a) {
            this.ec = a;
            this.g("LatitudeMemberPath");
        }
        get longitudeMemberPath() {
            return this.ed;
        }
        set longitudeMemberPath(a) {
            this.ed = a;
            this.g("LongitudeMemberPath");
        }
        get markerCollisionAvoidance() {
            return this.ee;
        }
        set markerCollisionAvoidance(a) {
            this.ee = a;
            this.g("MarkerCollisionAvoidance");
        }
        get isCustomScatterStyleAllowed() {
            return this.d4;
        }
        set isCustomScatterStyleAllowed(a) {
            this.d4 = a;
            this.g("IsCustomScatterStyleAllowed");
        }
        get isCustomScatterMarkerStyleAllowed() {
            return this.d3;
        }
        set isCustomScatterMarkerStyleAllowed(a) {
            this.d3 = a;
            this.g("IsCustomScatterMarkerStyleAllowed");
        }
        get itemSearchMode() {
            return this.eb;
        }
        set itemSearchMode(a) {
            this.eb = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.d7;
        }
        set itemSearchThreshold(a) {
            this.d7 = a;
            this.g("ItemSearchThreshold");
        }
        get assigningScatterStyleRef() {
            return this.ea;
        }
        set assigningScatterStyleRef(a) {
            this.ea = a;
            this.g("AssigningScatterStyleRef");
        }
        get assigningScatterMarkerStyleRef() {
            return this.d9;
        }
        set assigningScatterMarkerStyleRef(a) {
            this.d9 = a;
            this.g("AssigningScatterMarkerStyleRef");
        }
    }
    GeographicSymbolSeriesDescription.$t = /*@__PURE__*/ markType(GeographicSymbolSeriesDescription, 'GeographicSymbolSeriesDescription', GeographicMarkerSeriesDescription.$);
    return GeographicSymbolSeriesDescription;
})();
