// igniteui-react-core/esm2015/lib/GeographicMapImageryDescription_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
import { SeriesViewerDescription } from "./SeriesViewerDescription";
/**
 * @hidden
 */
export let GeographicMapImageryDescription = /*@__PURE__*/ (() => {
    class GeographicMapImageryDescription extends Description {
        get_type() {
            return "GeographicMapImagery";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
            this.t = null;
            this.l = false;
            this.u = null;
            this.n = 0;
            this.s = null;
            this.r = null;
            this.p = null;
            this.q = null;
        }
        get windowRect() {
            return this.j;
        }
        set windowRect(a) {
            this.j = a;
            this.g("WindowRect");
        }
        get geographicMap() {
            return this.h;
        }
        set geographicMap(a) {
            this.h = a;
            this.g("GeographicMap");
        }
        get referer() {
            return this.t;
        }
        set referer(a) {
            this.t = a;
            this.g("Referer");
        }
        get isHorizontalWrappingEnabled() {
            return this.l;
        }
        set isHorizontalWrappingEnabled(a) {
            this.l = a;
            this.g("IsHorizontalWrappingEnabled");
        }
        get userAgent() {
            return this.u;
        }
        set userAgent(a) {
            this.u = a;
            this.g("UserAgent");
        }
        get opacity() {
            return this.n;
        }
        set opacity(a) {
            this.n = a;
            this.g("Opacity");
        }
        get imageTilesReadyRef() {
            return this.s;
        }
        set imageTilesReadyRef(a) {
            this.s = a;
            this.g("ImageTilesReadyRef");
        }
        get imagesChangedRef() {
            return this.r;
        }
        set imagesChangedRef(a) {
            this.r = a;
            this.g("ImagesChangedRef");
        }
        get cancellingImageRef() {
            return this.p;
        }
        set cancellingImageRef(a) {
            this.p = a;
            this.g("CancellingImageRef");
        }
        get downloadingImageRef() {
            return this.q;
        }
        set downloadingImageRef(a) {
            this.q = a;
            this.g("DownloadingImageRef");
        }
    }
    GeographicMapImageryDescription.$t = /*@__PURE__*/ markType(GeographicMapImageryDescription, 'GeographicMapImageryDescription', Description.$);
    return GeographicMapImageryDescription;
})();
/**
 * @hidden
 */
export let GeographicMapDescription = /*@__PURE__*/ (() => {
    class GeographicMapDescription extends SeriesViewerDescription {
        get_type() {
            return "GeographicMap";
        }
        constructor() {
            super();
            this.he = false;
            this.hq = null;
            this.hb = false;
            this.hn = null;
            this.hd = false;
            this.hc = false;
            this.g8 = null;
            this.g7 = null;
            this.g5 = null;
            this.hk = 0;
            this.hj = 0;
            this.hp = null;
            this.hr = null;
            this.ho = null;
        }
        get zoomable() {
            return this.he;
        }
        set zoomable(a) {
            this.he = a;
            this.g("Zoomable");
        }
        get resizeBehavior() {
            return this.hq;
        }
        set resizeBehavior(a) {
            this.hq = a;
            this.g("ResizeBehavior");
        }
        get isHorizontalWrappingEnabled() {
            return this.hb;
        }
        set isHorizontalWrappingEnabled(a) {
            this.hb = a;
            this.g("IsHorizontalWrappingEnabled");
        }
        get backgroundTilingMode() {
            return this.hn;
        }
        set backgroundTilingMode(a) {
            this.hn = a;
            this.g("BackgroundTilingMode");
        }
        get useWorldRectForZoomBounds() {
            return this.hd;
        }
        set useWorldRectForZoomBounds(a) {
            this.hd = a;
            this.g("UseWorldRectForZoomBounds");
        }
        get suppressZoomResetOnWorldRectChange() {
            return this.hc;
        }
        set suppressZoomResetOnWorldRectChange(a) {
            this.hc = a;
            this.g("SuppressZoomResetOnWorldRectChange");
        }
        get worldRect() {
            return this.g8;
        }
        set worldRect(a) {
            this.g8 = a;
            this.g("WorldRect");
        }
        get actualWorldRect() {
            return this.g7;
        }
        set actualWorldRect(a) {
            this.g7 = a;
            this.g("ActualWorldRect");
        }
        get backgroundContent() {
            return this.g5;
        }
        set backgroundContent(a) {
            this.g5 = a;
            this.g("BackgroundContent");
        }
        get windowScale() {
            return this.hk;
        }
        set windowScale(a) {
            this.hk = a;
            this.g("WindowScale");
        }
        get actualWindowScale() {
            return this.hj;
        }
        set actualWindowScale(a) {
            this.hj = a;
            this.g("ActualWindowScale");
        }
        get imageTilesReadyRef() {
            return this.hp;
        }
        set imageTilesReadyRef(a) {
            this.hp = a;
            this.g("ImageTilesReadyRef");
        }
        get width() {
            return this.hr;
        }
        set width(a) {
            this.hr = a;
            this.g("Width");
        }
        get height() {
            return this.ho;
        }
        set height(a) {
            this.ho = a;
            this.g("Height");
        }
    }
    GeographicMapDescription.$t = /*@__PURE__*/ markType(GeographicMapDescription, 'GeographicMapDescription', SeriesViewerDescription.$);
    return GeographicMapDescription;
})();
