// igniteui-react-core/esm2015/lib/GeographicContourLineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicXYTriangulatingSeriesDescription } from "./GeographicXYTriangulatingSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicContourLineSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicContourLineSeriesDescription extends GeographicXYTriangulatingSeriesDescription {
        get_type() {
            return "GeographicContourLineSeries";
        }
        constructor() {
            super();
            this.d0 = null;
            this.dw = null;
            this.dv = null;
            this.dt = null;
            this.dz = null;
        }
        get valueMemberPath() {
            return this.d0;
        }
        set valueMemberPath(a) {
            this.d0 = a;
            this.g("ValueMemberPath");
        }
        get fillScale() {
            return this.dw;
        }
        set fillScale(a) {
            this.dw = a;
            this.g("FillScale");
        }
        get actualFillScale() {
            return this.dv;
        }
        set actualFillScale(a) {
            this.dv = a;
            this.g("ActualFillScale");
        }
        get valueResolver() {
            return this.dt;
        }
        set valueResolver(a) {
            this.dt = a;
            this.g("ValueResolver");
        }
        get triangulationStatusChangedRef() {
            return this.dz;
        }
        set triangulationStatusChangedRef(a) {
            this.dz = a;
            this.g("TriangulationStatusChangedRef");
        }
    }
    GeographicContourLineSeriesDescription.$t = /*@__PURE__*/ markType(GeographicContourLineSeriesDescription, 'GeographicContourLineSeriesDescription', GeographicXYTriangulatingSeriesDescription.$);
    return GeographicContourLineSeriesDescription;
})();
