// igniteui-react-core/esm2015/lib/GeographicXYTriangulatingSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicXYTriangulatingSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicXYTriangulatingSeriesDescription extends GeographicMapSeriesHostDescription {
        get_type() {
            return "GeographicXYTriangulatingSeries";
        }
        constructor() {
            super();
            this.dh = null;
            this.dg = null;
            this.di = null;
            this.dj = null;
            this.dk = null;
            this.dl = null;
        }
        get longitudeMemberPath() {
            return this.dh;
        }
        set longitudeMemberPath(a) {
            this.dh = a;
            this.g("LongitudeMemberPath");
        }
        get latitudeMemberPath() {
            return this.dg;
        }
        set latitudeMemberPath(a) {
            this.dg = a;
            this.g("LatitudeMemberPath");
        }
        get trianglesSourceRef() {
            return this.di;
        }
        set trianglesSourceRef(a) {
            this.di = a;
            this.g("TrianglesSourceRef");
        }
        get triangleVertexMemberPath1() {
            return this.dj;
        }
        set triangleVertexMemberPath1(a) {
            this.dj = a;
            this.g("TriangleVertexMemberPath1");
        }
        get triangleVertexMemberPath2() {
            return this.dk;
        }
        set triangleVertexMemberPath2(a) {
            this.dk = a;
            this.g("TriangleVertexMemberPath2");
        }
        get triangleVertexMemberPath3() {
            return this.dl;
        }
        set triangleVertexMemberPath3(a) {
            this.dl = a;
            this.g("TriangleVertexMemberPath3");
        }
    }
    GeographicXYTriangulatingSeriesDescription.$t = /*@__PURE__*/ markType(GeographicXYTriangulatingSeriesDescription, 'GeographicXYTriangulatingSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicXYTriangulatingSeriesDescription;
})();
