// igniteui-react-core/esm2015/lib/HeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let HeaderDescription = /*@__PURE__*/ (() => {
    class HeaderDescription extends DefinitionBaseDescription {
        get_type() {
            return "Header";
        }
        constructor() {
            super();
            this.cg = null;
            this.ce = null;
            this.ch = null;
            this.cf = null;
            this.b6 = 0;
            this.b8 = 0;
            this.b7 = 0;
            this.b5 = 0;
        }
        get sortIndicatorColor() {
            return this.cg;
        }
        set sortIndicatorColor(a) {
            this.cg = a;
            this.g("SortIndicatorColor");
        }
        get actualSortIndicatorColor() {
            return this.ce;
        }
        set actualSortIndicatorColor(a) {
            this.ce = a;
            this.g("ActualSortIndicatorColor");
        }
        get sortIndicatorStyle() {
            return this.ch;
        }
        set sortIndicatorStyle(a) {
            this.ch = a;
            this.g("SortIndicatorStyle");
        }
        get actualSortIndicatorStyle() {
            return this.cf;
        }
        set actualSortIndicatorStyle(a) {
            this.cf = a;
            this.g("ActualSortIndicatorStyle");
        }
        get paddingLeft() {
            return this.b6;
        }
        set paddingLeft(a) {
            this.b6 = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.b8;
        }
        set paddingTop(a) {
            this.b8 = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.b7;
        }
        set paddingRight(a) {
            this.b7 = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.b5;
        }
        set paddingBottom(a) {
            this.b5 = a;
            this.g("PaddingBottom");
        }
    }
    HeaderDescription.$t = /*@__PURE__*/ markType(HeaderDescription, 'HeaderDescription', DefinitionBaseDescription.$);
    HeaderDescription.__canMarshalByValue = true;
    HeaderDescription.__marshalByValueAlias = "Header";
    return HeaderDescription;
})();
