// igniteui-react-core/esm2015/lib/HighlightingInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let HighlightingInfoDescription = /*@__PURE__*/ (() => {
    class HighlightingInfoDescription extends Description {
        get_type() {
            return "HighlightingInfo";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.s = 0;
            this.r = 0;
            this.v = null;
            this.p = 0;
            this.l = false;
            this.k = false;
            this.j = false;
            this.h = null;
        }
        get startIndex() {
            return this.s;
        }
        set startIndex(a) {
            this.s = a;
            this.g("StartIndex");
        }
        get endIndex() {
            return this.r;
        }
        set endIndex(a) {
            this.r = a;
            this.g("EndIndex");
        }
        get state() {
            return this.v;
        }
        set state(a) {
            this.v = a;
            this.g("State");
        }
        get progress() {
            return this.p;
        }
        set progress(a) {
            this.p = a;
            this.g("Progress");
        }
        get isMarker() {
            return this.l;
        }
        set isMarker(a) {
            this.l = a;
            this.g("IsMarker");
        }
        get isFullRange() {
            return this.k;
        }
        set isFullRange(a) {
            this.k = a;
            this.g("IsFullRange");
        }
        get isExclusive() {
            return this.j;
        }
        set isExclusive(a) {
            this.j = a;
            this.g("IsExclusive");
        }
        get series() {
            return this.h;
        }
        set series(a) {
            this.h = a;
            this.g("Series");
        }
    }
    HighlightingInfoDescription.$t = /*@__PURE__*/ markType(HighlightingInfoDescription, 'HighlightingInfoDescription', Description.$);
    return HighlightingInfoDescription;
})();
