// igniteui-react-core/esm2015/lib/HorizontalRangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RangeCategorySeriesDescription } from "./RangeCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let HorizontalRangeCategorySeriesDescription = /*@__PURE__*/ (() => {
    class HorizontalRangeCategorySeriesDescription extends RangeCategorySeriesDescription {
        get_type() {
            return "HorizontalRangeCategorySeries";
        }
        constructor() {
            super();
            this.ej = null;
            this.ek = null;
            this.eh = null;
            this.ef = null;
            this.ei = null;
            this.eg = null;
        }
        get xAxisRef() {
            return this.ej;
        }
        set xAxisRef(a) {
            this.ej = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.ek;
        }
        set yAxisRef(a) {
            this.ek = a;
            this.g("YAxisRef");
        }
        get lowMemberAsLegendLabel() {
            return this.eh;
        }
        set lowMemberAsLegendLabel(a) {
            this.eh = a;
            this.g("LowMemberAsLegendLabel");
        }
        get highMemberAsLegendLabel() {
            return this.ef;
        }
        set highMemberAsLegendLabel(a) {
            this.ef = a;
            this.g("HighMemberAsLegendLabel");
        }
        get lowMemberAsLegendUnit() {
            return this.ei;
        }
        set lowMemberAsLegendUnit(a) {
            this.ei = a;
            this.g("LowMemberAsLegendUnit");
        }
        get highMemberAsLegendUnit() {
            return this.eg;
        }
        set highMemberAsLegendUnit(a) {
            this.eg = a;
            this.g("HighMemberAsLegendUnit");
        }
    }
    HorizontalRangeCategorySeriesDescription.$t = /*@__PURE__*/ markType(HorizontalRangeCategorySeriesDescription, 'HorizontalRangeCategorySeriesDescription', RangeCategorySeriesDescription.$);
    return HorizontalRangeCategorySeriesDescription;
})();
