// igniteui-react-core/esm2015/lib/ItemToolTipLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ItemToolTipLayerDescription = /*@__PURE__*/ (() => {
    class ItemToolTipLayerDescription extends AnnotationLayerDescription {
        get_type() {
            return "ItemToolTipLayer";
        }
        constructor() {
            super();
            this.en = null;
            this.eo = null;
            this.ek = false;
            this.ej = false;
        }
        get targetSeriesName() {
            return this.en;
        }
        set targetSeriesName(a) {
            this.en = a;
            this.g("TargetSeriesName");
        }
        get targetSeriesRef() {
            return this.eo;
        }
        set targetSeriesRef(a) {
            this.eo = a;
            this.g("TargetSeriesRef");
        }
        get useInterpolation() {
            return this.ek;
        }
        set useInterpolation(a) {
            this.ek = a;
            this.g("UseInterpolation");
        }
        get skipUnknownValues() {
            return this.ej;
        }
        set skipUnknownValues(a) {
            this.ej = a;
            this.g("SkipUnknownValues");
        }
    }
    ItemToolTipLayerDescription.$t = /*@__PURE__*/ markType(ItemToolTipLayerDescription, 'ItemToolTipLayerDescription', AnnotationLayerDescription.$);
    return ItemToolTipLayerDescription;
})();
