// igniteui-react-core/esm2015/lib/LinearGaugeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let LinearGaugeDescription = /*@__PURE__*/ (() => {
    class LinearGaugeDescription extends Description {
        get fontFamily() {
            return this.cv;
        }
        set fontFamily(a) {
            this.cv = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.ab;
        }
        set fontSize(a) {
            this.ab = a;
            this.g("FontSize");
        }
        get fontWeight() {
            return this.cx;
        }
        set fontWeight(a) {
            this.cx = a;
            this.g("FontWeight");
        }
        get fontStyle() {
            return this.cw;
        }
        set fontStyle(a) {
            this.cw = a;
            this.g("FontStyle");
        }
        get width() {
            return this.db;
        }
        set width(a) {
            this.db = a;
            this.g("Width");
        }
        get height() {
            return this.cz;
        }
        set height(a) {
            this.cz = a;
            this.g("Height");
        }
        get background() {
            return this.cq;
        }
        set background(a) {
            this.cq = a;
            this.g("Background");
        }
        get_type() {
            return "LinearGauge";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.cv = null;
            this.ab = 0;
            this.cx = null;
            this.cw = null;
            this.db = null;
            this.cz = null;
            this.cq = null;
            this.ba = 0;
            this.c6 = null;
            this.k = null;
            this.i = null;
            this.j = null;
            this.ai = 0;
            this.w = 0;
            this.ah = 0;
            this.v = 0;
            this.bb = 0;
            this.c5 = null;
            this.c3 = null;
            this.ay = 0;
            this.a1 = 0;
            this.az = 0;
            this.a2 = 0;
            this.ap = 0;
            this.at = 0;
            this.ao = 0;
            this.as = 0;
            this.ar = 0;
            this.av = 0;
            this.aq = 0;
            this.au = 0;
            this.ac = 0;
            this.a6 = 0;
            this.a7 = 0;
            this.ae = 0;
            this.ad = 0;
            this.af = 0;
            this.ag = 0;
            this.aj = 0;
            this.a8 = 0;
            this.a5 = 0;
            this.a9 = 0;
            this.c9 = null;
            this.cu = null;
            this.an = 0;
            this.c2 = null;
            this.c4 = null;
            this.aw = 0;
            this.al = 0;
            this.ak = 0;
            this.am = 0;
            this.c1 = null;
            this.q = false;
            this.cr = null;
            this.cs = null;
            this.aa = 0;
            this.y = 0;
            this.z = 0;
            this.a3 = 0;
            this.a0 = 0;
            this.c7 = null;
            this.c8 = null;
            this.a4 = 0;
            this.p = false;
            this.c0 = null;
            this.h = null;
            this.ck = 0;
            this.cj = 0;
            this.r = false;
            this.ct = null;
            this.ax = 0;
            this.x = 0;
            this.cy = null;
            this.cp = null;
            this.da = null;
            this.co = null;
            this.cn = null;
        }
        get transitionProgress() {
            return this.ba;
        }
        set transitionProgress(a) {
            this.ba = a;
            this.g("TransitionProgress");
        }
        get orientation() {
            return this.c6;
        }
        set orientation(a) {
            this.c6 = a;
            this.g("Orientation");
        }
        get ranges() {
            return this.k;
        }
        set ranges(a) {
            this.k = a;
            this.g("Ranges");
        }
        get rangeBrushes() {
            return this.i;
        }
        set rangeBrushes(a) {
            this.i = a;
            this.g("RangeBrushes");
        }
        get rangeOutlines() {
            return this.j;
        }
        set rangeOutlines(a) {
            this.j = a;
            this.g("RangeOutlines");
        }
        get minimumValue() {
            return this.ai;
        }
        set minimumValue(a) {
            this.ai = a;
            this.g("MinimumValue");
        }
        get actualMinimumValue() {
            return this.w;
        }
        set actualMinimumValue(a) {
            this.w = a;
            this.g("ActualMinimumValue");
        }
        get maximumValue() {
            return this.ah;
        }
        set maximumValue(a) {
            this.ah = a;
            this.g("MaximumValue");
        }
        get actualMaximumValue() {
            return this.v;
        }
        set actualMaximumValue(a) {
            this.v = a;
            this.g("ActualMaximumValue");
        }
        get value() {
            return this.bb;
        }
        set value(a) {
            this.bb = a;
            this.g("Value");
        }
        get needleShape() {
            return this.c5;
        }
        set needleShape(a) {
            this.c5 = a;
            this.g("NeedleShape");
        }
        get needleName() {
            return this.c3;
        }
        set needleName(a) {
            this.c3 = a;
            this.g("NeedleName");
        }
        get rangeInnerExtent() {
            return this.ay;
        }
        set rangeInnerExtent(a) {
            this.ay = a;
            this.g("RangeInnerExtent");
        }
        get scaleInnerExtent() {
            return this.a1;
        }
        set scaleInnerExtent(a) {
            this.a1 = a;
            this.g("ScaleInnerExtent");
        }
        get rangeOuterExtent() {
            return this.az;
        }
        set rangeOuterExtent(a) {
            this.az = a;
            this.g("RangeOuterExtent");
        }
        get scaleOuterExtent() {
            return this.a2;
        }
        set scaleOuterExtent(a) {
            this.a2 = a;
            this.g("ScaleOuterExtent");
        }
        get needleInnerExtent() {
            return this.ap;
        }
        set needleInnerExtent(a) {
            this.ap = a;
            this.g("NeedleInnerExtent");
        }
        get needleOuterExtent() {
            return this.at;
        }
        set needleOuterExtent(a) {
            this.at = a;
            this.g("NeedleOuterExtent");
        }
        get needleInnerBaseWidth() {
            return this.ao;
        }
        set needleInnerBaseWidth(a) {
            this.ao = a;
            this.g("NeedleInnerBaseWidth");
        }
        get needleOuterBaseWidth() {
            return this.as;
        }
        set needleOuterBaseWidth(a) {
            this.as = a;
            this.g("NeedleOuterBaseWidth");
        }
        get needleInnerPointWidth() {
            return this.ar;
        }
        set needleInnerPointWidth(a) {
            this.ar = a;
            this.g("NeedleInnerPointWidth");
        }
        get needleOuterPointWidth() {
            return this.av;
        }
        set needleOuterPointWidth(a) {
            this.av = a;
            this.g("NeedleOuterPointWidth");
        }
        get needleInnerPointExtent() {
            return this.aq;
        }
        set needleInnerPointExtent(a) {
            this.aq = a;
            this.g("NeedleInnerPointExtent");
        }
        get needleOuterPointExtent() {
            return this.au;
        }
        set needleOuterPointExtent(a) {
            this.au = a;
            this.g("NeedleOuterPointExtent");
        }
        get interval() {
            return this.ac;
        }
        set interval(a) {
            this.ac = a;
            this.g("Interval");
        }
        get ticksPostInitial() {
            return this.a6;
        }
        set ticksPostInitial(a) {
            this.a6 = a;
            this.g("TicksPostInitial");
        }
        get ticksPreTerminal() {
            return this.a7;
        }
        set ticksPreTerminal(a) {
            this.a7 = a;
            this.g("TicksPreTerminal");
        }
        get labelInterval() {
            return this.ae;
        }
        set labelInterval(a) {
            this.ae = a;
            this.g("LabelInterval");
        }
        get labelExtent() {
            return this.ad;
        }
        set labelExtent(a) {
            this.ad = a;
            this.g("LabelExtent");
        }
        get labelsPostInitial() {
            return this.af;
        }
        set labelsPostInitial(a) {
            this.af = a;
            this.g("LabelsPostInitial");
        }
        get labelsPreTerminal() {
            return this.ag;
        }
        set labelsPreTerminal(a) {
            this.ag = a;
            this.g("LabelsPreTerminal");
        }
        get minorTickCount() {
            return this.aj;
        }
        set minorTickCount(a) {
            this.aj = a;
            this.g("MinorTickCount");
        }
        get tickStartExtent() {
            return this.a8;
        }
        set tickStartExtent(a) {
            this.a8 = a;
            this.g("TickStartExtent");
        }
        get tickEndExtent() {
            return this.a5;
        }
        set tickEndExtent(a) {
            this.a5 = a;
            this.g("TickEndExtent");
        }
        get tickStrokeThickness() {
            return this.a9;
        }
        set tickStrokeThickness(a) {
            this.a9 = a;
            this.g("TickStrokeThickness");
        }
        get tickBrush() {
            return this.c9;
        }
        set tickBrush(a) {
            this.c9 = a;
            this.g("TickBrush");
        }
        get fontBrush() {
            return this.cu;
        }
        set fontBrush(a) {
            this.cu = a;
            this.g("FontBrush");
        }
        get needleBreadth() {
            return this.an;
        }
        set needleBreadth(a) {
            this.an = a;
            this.g("NeedleBreadth");
        }
        get needleBrush() {
            return this.c2;
        }
        set needleBrush(a) {
            this.c2 = a;
            this.g("NeedleBrush");
        }
        get needleOutline() {
            return this.c4;
        }
        set needleOutline(a) {
            this.c4 = a;
            this.g("NeedleOutline");
        }
        get needleStrokeThickness() {
            return this.aw;
        }
        set needleStrokeThickness(a) {
            this.aw = a;
            this.g("NeedleStrokeThickness");
        }
        get minorTickStartExtent() {
            return this.al;
        }
        set minorTickStartExtent(a) {
            this.al = a;
            this.g("MinorTickStartExtent");
        }
        get minorTickEndExtent() {
            return this.ak;
        }
        set minorTickEndExtent(a) {
            this.ak = a;
            this.g("MinorTickEndExtent");
        }
        get minorTickStrokeThickness() {
            return this.am;
        }
        set minorTickStrokeThickness(a) {
            this.am = a;
            this.g("MinorTickStrokeThickness");
        }
        get minorTickBrush() {
            return this.c1;
        }
        set minorTickBrush(a) {
            this.c1 = a;
            this.g("MinorTickBrush");
        }
        get isScaleInverted() {
            return this.q;
        }
        set isScaleInverted(a) {
            this.q = a;
            this.g("IsScaleInverted");
        }
        get backingBrush() {
            return this.cr;
        }
        set backingBrush(a) {
            this.cr = a;
            this.g("BackingBrush");
        }
        get backingOutline() {
            return this.cs;
        }
        set backingOutline(a) {
            this.cs = a;
            this.g("BackingOutline");
        }
        get backingStrokeThickness() {
            return this.aa;
        }
        set backingStrokeThickness(a) {
            this.aa = a;
            this.g("BackingStrokeThickness");
        }
        get backingInnerExtent() {
            return this.y;
        }
        set backingInnerExtent(a) {
            this.y = a;
            this.g("BackingInnerExtent");
        }
        get backingOuterExtent() {
            return this.z;
        }
        set backingOuterExtent(a) {
            this.z = a;
            this.g("BackingOuterExtent");
        }
        get scaleStartExtent() {
            return this.a3;
        }
        set scaleStartExtent(a) {
            this.a3 = a;
            this.g("ScaleStartExtent");
        }
        get scaleEndExtent() {
            return this.a0;
        }
        set scaleEndExtent(a) {
            this.a0 = a;
            this.g("ScaleEndExtent");
        }
        get scaleBrush() {
            return this.c7;
        }
        set scaleBrush(a) {
            this.c7 = a;
            this.g("ScaleBrush");
        }
        get scaleOutline() {
            return this.c8;
        }
        set scaleOutline(a) {
            this.c8 = a;
            this.g("ScaleOutline");
        }
        get scaleStrokeThickness() {
            return this.a4;
        }
        set scaleStrokeThickness(a) {
            this.a4 = a;
            this.g("ScaleStrokeThickness");
        }
        get isNeedleDraggingEnabled() {
            return this.p;
        }
        set isNeedleDraggingEnabled(a) {
            this.p = a;
            this.g("IsNeedleDraggingEnabled");
        }
        get labelFormat() {
            return this.c0;
        }
        set labelFormat(a) {
            this.c0 = a;
            this.g("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.h;
        }
        set labelFormatSpecifiers(a) {
            this.h = a;
            this.g("LabelFormatSpecifiers");
        }
        get transitionDuration() {
            return this.ck;
        }
        set transitionDuration(a) {
            this.ck = a;
            this.g("TransitionDuration");
        }
        get showToolTipTimeout() {
            return this.cj;
        }
        set showToolTipTimeout(a) {
            this.cj = a;
            this.g("ShowToolTipTimeout");
        }
        get showToolTip() {
            return this.r;
        }
        set showToolTip(a) {
            this.r = a;
            this.g("ShowToolTip");
        }
        get font() {
            return this.ct;
        }
        set font(a) {
            this.ct = a;
            this.g("Font");
        }
        get pixelScalingRatio() {
            return this.ax;
        }
        set pixelScalingRatio(a) {
            this.ax = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.x;
        }
        set actualPixelScalingRatio(a) {
            this.x = a;
            this.g("ActualPixelScalingRatio");
        }
        get formatLabelRef() {
            return this.cy;
        }
        set formatLabelRef(a) {
            this.cy = a;
            this.g("FormatLabelRef");
        }
        get alignLabelRef() {
            return this.cp;
        }
        set alignLabelRef(a) {
            this.cp = a;
            this.g("AlignLabelRef");
        }
        get valueChangedRef() {
            return this.da;
        }
        set valueChangedRef(a) {
            this.da = a;
            this.g("ValueChangedRef");
        }
        get actualMinimumValueChangeRef() {
            return this.co;
        }
        set actualMinimumValueChangeRef(a) {
            this.co = a;
            this.g("ActualMinimumValueChangeRef");
        }
        get actualMaximumValueChangeRef() {
            return this.cn;
        }
        set actualMaximumValueChangeRef(a) {
            this.cn = a;
            this.g("ActualMaximumValueChangeRef");
        }
    }
    LinearGaugeDescription.$t = /*@__PURE__*/ markType(LinearGaugeDescription, 'LinearGaugeDescription', Description.$);
    return LinearGaugeDescription;
})();
