// igniteui-react-core/esm2015/lib/ListPanelItemModelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ListPanelItemModelDescription = /*@__PURE__*/ (() => {
    class ListPanelItemModelDescription extends Description {
        get_type() {
            return "ListPanelItemModel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = 0;
            this.j = 0;
            this.n = null;
            this.h = false;
        }
        get dataRow() {
            return this.l;
        }
        set dataRow(a) {
            this.l = a;
            this.g("DataRow");
        }
        get rowHeight() {
            return this.j;
        }
        set rowHeight(a) {
            this.j = a;
            this.g("RowHeight");
        }
        get rowObjectRef() {
            return this.n;
        }
        set rowObjectRef(a) {
            this.n = a;
            this.g("RowObjectRef");
        }
        get isModelDirty() {
            return this.h;
        }
        set isModelDirty(a) {
            this.h = a;
            this.g("IsModelDirty");
        }
    }
    ListPanelItemModelDescription.$t = /*@__PURE__*/ markType(ListPanelItemModelDescription, 'ListPanelItemModelDescription', Description.$);
    return ListPanelItemModelDescription;
})();
