// igniteui-react-core/esm2015/lib/MarkerSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let MarkerSeriesDescription = /*@__PURE__*/ (() => {
    class MarkerSeriesDescription extends SeriesDescription {
        get_type() {
            return "MarkerSeries";
        }
        constructor() {
            super();
            this.dm = null;
            this.dk = null;
            this.di = null;
            this.dl = null;
            this.dc = 0;
            this.dg = null;
            this.dh = null;
            this.de = null;
            this.dj = null;
            this.df = null;
        }
        get markerType() {
            return this.dm;
        }
        set markerType(a) {
            this.dm = a;
            this.g("MarkerType");
        }
        get markerOutlineMode() {
            return this.dk;
        }
        set markerOutlineMode(a) {
            this.dk = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.di;
        }
        set markerFillMode(a) {
            this.di = a;
            this.g("MarkerFillMode");
        }
        get markerTemplateRef() {
            return this.dl;
        }
        set markerTemplateRef(a) {
            this.dl = a;
            this.g("MarkerTemplateRef");
        }
        get markerThickness() {
            return this.dc;
        }
        set markerThickness(a) {
            this.dc = a;
            this.g("MarkerThickness");
        }
        get actualMarkerTemplateRef() {
            return this.dg;
        }
        set actualMarkerTemplateRef(a) {
            this.dg = a;
            this.g("ActualMarkerTemplateRef");
        }
        get markerBrush() {
            return this.dh;
        }
        set markerBrush(a) {
            this.dh = a;
            this.g("MarkerBrush");
        }
        get actualMarkerBrush() {
            return this.de;
        }
        set actualMarkerBrush(a) {
            this.de = a;
            this.g("ActualMarkerBrush");
        }
        get markerOutline() {
            return this.dj;
        }
        set markerOutline(a) {
            this.dj = a;
            this.g("MarkerOutline");
        }
        get actualMarkerOutline() {
            return this.df;
        }
        set actualMarkerOutline(a) {
            this.df = a;
            this.g("ActualMarkerOutline");
        }
    }
    MarkerSeriesDescription.$t = /*@__PURE__*/ markType(MarkerSeriesDescription, 'MarkerSeriesDescription', SeriesDescription.$);
    return MarkerSeriesDescription;
})();
