// igniteui-react-core/esm2015/lib/MultiColumnComboBoxDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let MultiColumnComboBoxDescription = /*@__PURE__*/ (() => {
    class MultiColumnComboBoxDescription extends Description {
        get_type() {
            return "MultiColumnComboBox";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.u = false;
            this.j = null;
            this.cz = null;
            this.cl = null;
            this.c0 = null;
            this.cm = null;
            this.c1 = null;
            this.cn = null;
            this.at = 0;
            this.ab = 0;
            this.r = false;
            this.au = 0;
            this.av = 0;
            this.aw = 0;
            this.ax = 0;
            this.ac = 0;
            this.ad = 0;
            this.ae = 0;
            this.af = 0;
            this.az = 0;
            this.ay = 0;
            this.a0 = 0;
            this.a1 = 0;
            this.ah = 0;
            this.ag = 0;
            this.ai = 0;
            this.aj = 0;
            this.i = null;
            this.c5 = null;
            this.co = null;
            this.s = false;
            this.c6 = null;
            this.cp = null;
            this.a3 = 0;
            this.ak = 0;
            this.c7 = null;
            this.cq = null;
            this.a4 = 0;
            this.al = 0;
            this.a5 = 0;
            this.am = 0;
            this.dc = null;
            this.cr = null;
            this.a7 = 0;
            this.an = 0;
            this.a8 = 0;
            this.ao = 0;
            this.c4 = null;
            this.de = null;
            this.di = null;
            this.cs = null;
            this.df = null;
            this.a9 = 0;
            this.dg = null;
            this.dh = null;
            this.p = false;
            this.dk = null;
            this.ct = null;
            this.dl = null;
            this.cu = null;
            this.dq = null;
            this.cv = null;
            this.dm = null;
            this.ba = 0;
            this.dn = null;
            this.dp = null;
            this.dr = null;
            this.t = false;
            this.dw = null;
            this.cw = null;
            this.c8 = null;
            this.a6 = 0;
            this.c9 = null;
            this.da = null;
            this.dz = null;
            this.cx = null;
            this.bb = 0;
            this.ap = 0;
            this.d0 = null;
            this.cy = null;
            this.bc = 0;
            this.aq = 0;
            this.bd = 0;
            this.ar = 0;
            this.be = 0;
            this.as = 0;
            this.q = false;
            this.dt = null;
            this.dx = null;
            this.du = null;
            this.k = null;
            this.h = null;
            this.cj = null;
            this.db = null;
            this.dj = null;
            this.dy = null;
            this.ds = null;
            this.dd = null;
            this.c2 = null;
            this.c3 = null;
            this.dv = null;
            this.d1 = null;
            this.a2 = 0;
        }
        get showClearButton() {
            return this.u;
        }
        set showClearButton(a) {
            this.u = a;
            this.g("ShowClearButton");
        }
        get fields() {
            return this.j;
        }
        set fields(a) {
            this.j = a;
            this.g("Fields");
        }
        get backgroundColor() {
            return this.cz;
        }
        set backgroundColor(a) {
            this.cz = a;
            this.g("BackgroundColor");
        }
        get actualBackgroundColor() {
            return this.cl;
        }
        set actualBackgroundColor(a) {
            this.cl = a;
            this.g("ActualBackgroundColor");
        }
        get baseTheme() {
            return this.c0;
        }
        set baseTheme(a) {
            this.c0 = a;
            this.g("BaseTheme");
        }
        get actualBaseTheme() {
            return this.cm;
        }
        set actualBaseTheme(a) {
            this.cm = a;
            this.g("ActualBaseTheme");
        }
        get borderColor() {
            return this.c1;
        }
        set borderColor(a) {
            this.c1 = a;
            this.g("BorderColor");
        }
        get actualBorderColor() {
            return this.cn;
        }
        set actualBorderColor(a) {
            this.cn = a;
            this.g("ActualBorderColor");
        }
        get borderWidth() {
            return this.at;
        }
        set borderWidth(a) {
            this.at = a;
            this.g("BorderWidth");
        }
        get actualBorderWidth() {
            return this.ab;
        }
        set actualBorderWidth(a) {
            this.ab = a;
            this.g("ActualBorderWidth");
        }
        get columnHeadersVisible() {
            return this.r;
        }
        set columnHeadersVisible(a) {
            this.r = a;
            this.g("ColumnHeadersVisible");
        }
        get contentPaddingBottom() {
            return this.au;
        }
        set contentPaddingBottom(a) {
            this.au = a;
            this.g("ContentPaddingBottom");
        }
        get contentPaddingLeft() {
            return this.av;
        }
        set contentPaddingLeft(a) {
            this.av = a;
            this.g("ContentPaddingLeft");
        }
        get contentPaddingRight() {
            return this.aw;
        }
        set contentPaddingRight(a) {
            this.aw = a;
            this.g("ContentPaddingRight");
        }
        get contentPaddingTop() {
            return this.ax;
        }
        set contentPaddingTop(a) {
            this.ax = a;
            this.g("ContentPaddingTop");
        }
        get actualContentPaddingBottom() {
            return this.ac;
        }
        set actualContentPaddingBottom(a) {
            this.ac = a;
            this.g("ActualContentPaddingBottom");
        }
        get actualContentPaddingLeft() {
            return this.ad;
        }
        set actualContentPaddingLeft(a) {
            this.ad = a;
            this.g("ActualContentPaddingLeft");
        }
        get actualContentPaddingRight() {
            return this.ae;
        }
        set actualContentPaddingRight(a) {
            this.ae = a;
            this.g("ActualContentPaddingRight");
        }
        get actualContentPaddingTop() {
            return this.af;
        }
        set actualContentPaddingTop(a) {
            this.af = a;
            this.g("ActualContentPaddingTop");
        }
        get cornerRadiusBottomRight() {
            return this.az;
        }
        set cornerRadiusBottomRight(a) {
            this.az = a;
            this.g("CornerRadiusBottomRight");
        }
        get cornerRadiusBottomLeft() {
            return this.ay;
        }
        set cornerRadiusBottomLeft(a) {
            this.ay = a;
            this.g("CornerRadiusBottomLeft");
        }
        get cornerRadiusTopLeft() {
            return this.a0;
        }
        set cornerRadiusTopLeft(a) {
            this.a0 = a;
            this.g("CornerRadiusTopLeft");
        }
        get cornerRadiusTopRight() {
            return this.a1;
        }
        set cornerRadiusTopRight(a) {
            this.a1 = a;
            this.g("CornerRadiusTopRight");
        }
        get actualCornerRadiusBottomRight() {
            return this.ah;
        }
        set actualCornerRadiusBottomRight(a) {
            this.ah = a;
            this.g("ActualCornerRadiusBottomRight");
        }
        get actualCornerRadiusBottomLeft() {
            return this.ag;
        }
        set actualCornerRadiusBottomLeft(a) {
            this.ag = a;
            this.g("ActualCornerRadiusBottomLeft");
        }
        get actualCornerRadiusTopLeft() {
            return this.ai;
        }
        set actualCornerRadiusTopLeft(a) {
            this.ai = a;
            this.g("ActualCornerRadiusTopLeft");
        }
        get actualCornerRadiusTopRight() {
            return this.aj;
        }
        set actualCornerRadiusTopRight(a) {
            this.aj = a;
            this.g("ActualCornerRadiusTopRight");
        }
        get dataSourceDesiredProperties() {
            return this.i;
        }
        set dataSourceDesiredProperties(a) {
            this.i = a;
            this.g("DataSourceDesiredProperties");
        }
        get density() {
            return this.c5;
        }
        set density(a) {
            this.c5 = a;
            this.g("Density");
        }
        get actualDensity() {
            return this.co;
        }
        set actualDensity(a) {
            this.co = a;
            this.g("ActualDensity");
        }
        get dropDownButtonVisible() {
            return this.s;
        }
        set dropDownButtonVisible(a) {
            this.s = a;
            this.g("DropDownButtonVisible");
        }
        get focusBorderColor() {
            return this.c6;
        }
        set focusBorderColor(a) {
            this.c6 = a;
            this.g("FocusBorderColor");
        }
        get actualFocusBorderColor() {
            return this.cp;
        }
        set actualFocusBorderColor(a) {
            this.cp = a;
            this.g("ActualFocusBorderColor");
        }
        get focusBorderWidth() {
            return this.a3;
        }
        set focusBorderWidth(a) {
            this.a3 = a;
            this.g("FocusBorderWidth");
        }
        get actualFocusBorderWidth() {
            return this.ak;
        }
        set actualFocusBorderWidth(a) {
            this.ak = a;
            this.g("ActualFocusBorderWidth");
        }
        get focusUnderlineColor() {
            return this.c7;
        }
        set focusUnderlineColor(a) {
            this.c7 = a;
            this.g("FocusUnderlineColor");
        }
        get actualFocusUnderlineColor() {
            return this.cq;
        }
        set actualFocusUnderlineColor(a) {
            this.cq = a;
            this.g("ActualFocusUnderlineColor");
        }
        get focusUnderlineOpacity() {
            return this.a4;
        }
        set focusUnderlineOpacity(a) {
            this.a4 = a;
            this.g("FocusUnderlineOpacity");
        }
        get actualFocusUnderlineOpacity() {
            return this.al;
        }
        set actualFocusUnderlineOpacity(a) {
            this.al = a;
            this.g("ActualFocusUnderlineOpacity");
        }
        get focusUnderlineRippleOpacity() {
            return this.a5;
        }
        set focusUnderlineRippleOpacity(a) {
            this.a5 = a;
            this.g("FocusUnderlineRippleOpacity");
        }
        get actualFocusUnderlineRippleOpacity() {
            return this.am;
        }
        set actualFocusUnderlineRippleOpacity(a) {
            this.am = a;
            this.g("ActualFocusUnderlineRippleOpacity");
        }
        get hoverUnderlineColor() {
            return this.dc;
        }
        set hoverUnderlineColor(a) {
            this.dc = a;
            this.g("HoverUnderlineColor");
        }
        get actualHoverUnderlineColor() {
            return this.cr;
        }
        set actualHoverUnderlineColor(a) {
            this.cr = a;
            this.g("ActualHoverUnderlineColor");
        }
        get hoverUnderlineOpacity() {
            return this.a7;
        }
        set hoverUnderlineOpacity(a) {
            this.a7 = a;
            this.g("HoverUnderlineOpacity");
        }
        get actualHoverUnderlineOpacity() {
            return this.an;
        }
        set actualHoverUnderlineOpacity(a) {
            this.an = a;
            this.g("ActualHoverUnderlineOpacity");
        }
        get hoverUnderlineWidth() {
            return this.a8;
        }
        set hoverUnderlineWidth(a) {
            this.a8 = a;
            this.g("HoverUnderlineWidth");
        }
        get actualHoverUnderlineWidth() {
            return this.ao;
        }
        set actualHoverUnderlineWidth(a) {
            this.ao = a;
            this.g("ActualHoverUnderlineWidth");
        }
        get dataSourceRef() {
            return this.c4;
        }
        set dataSourceRef(a) {
            this.c4 = a;
            this.g("DataSourceRef");
        }
        get label() {
            return this.de;
        }
        set label(a) {
            this.de = a;
            this.g("Label");
        }
        get labelTextColor() {
            return this.di;
        }
        set labelTextColor(a) {
            this.di = a;
            this.g("LabelTextColor");
        }
        get actualLabelTextColor() {
            return this.cs;
        }
        set actualLabelTextColor(a) {
            this.cs = a;
            this.g("ActualLabelTextColor");
        }
        get labelFontFamily() {
            return this.df;
        }
        set labelFontFamily(a) {
            this.df = a;
            this.g("LabelFontFamily");
        }
        get labelFontSize() {
            return this.a9;
        }
        set labelFontSize(a) {
            this.a9 = a;
            this.g("LabelFontSize");
        }
        get labelFontStyle() {
            return this.dg;
        }
        set labelFontStyle(a) {
            this.dg = a;
            this.g("LabelFontStyle");
        }
        get labelFontWeight() {
            return this.dh;
        }
        set labelFontWeight(a) {
            this.dh = a;
            this.g("LabelFontWeight");
        }
        get actualLabelVisible() {
            return this.p;
        }
        set actualLabelVisible(a) {
            this.p = a;
            this.g("ActualLabelVisible");
        }
        get noMatchesFoundLabel() {
            return this.dk;
        }
        set noMatchesFoundLabel(a) {
            this.dk = a;
            this.g("NoMatchesFoundLabel");
        }
        get actualNoMatchesFoundLabel() {
            return this.ct;
        }
        set actualNoMatchesFoundLabel(a) {
            this.ct = a;
            this.g("ActualNoMatchesFoundLabel");
        }
        get noMatchesFoundLabelBackgroundColor() {
            return this.dl;
        }
        set noMatchesFoundLabelBackgroundColor(a) {
            this.dl = a;
            this.g("NoMatchesFoundLabelBackgroundColor");
        }
        get actualNoMatchesFoundLabelBackgroundColor() {
            return this.cu;
        }
        set actualNoMatchesFoundLabelBackgroundColor(a) {
            this.cu = a;
            this.g("ActualNoMatchesFoundLabelBackgroundColor");
        }
        get noMatchesFoundLabelTextColor() {
            return this.dq;
        }
        set noMatchesFoundLabelTextColor(a) {
            this.dq = a;
            this.g("NoMatchesFoundLabelTextColor");
        }
        get actualNoMatchesFoundLabelTextColor() {
            return this.cv;
        }
        set actualNoMatchesFoundLabelTextColor(a) {
            this.cv = a;
            this.g("ActualNoMatchesFoundLabelTextColor");
        }
        get noMatchesFoundLabelFontFamily() {
            return this.dm;
        }
        set noMatchesFoundLabelFontFamily(a) {
            this.dm = a;
            this.g("NoMatchesFoundLabelFontFamily");
        }
        get noMatchesFoundLabelFontSize() {
            return this.ba;
        }
        set noMatchesFoundLabelFontSize(a) {
            this.ba = a;
            this.g("NoMatchesFoundLabelFontSize");
        }
        get noMatchesFoundLabelFontStyle() {
            return this.dn;
        }
        set noMatchesFoundLabelFontStyle(a) {
            this.dn = a;
            this.g("NoMatchesFoundLabelFontStyle");
        }
        get noMatchesFoundLabelFontWeight() {
            return this.dp;
        }
        set noMatchesFoundLabelFontWeight(a) {
            this.dp = a;
            this.g("NoMatchesFoundLabelFontWeight");
        }
        get placeholder() {
            return this.dr;
        }
        set placeholder(a) {
            this.dr = a;
            this.g("Placeholder");
        }
        get rowSeparatorsVisible() {
            return this.t;
        }
        set rowSeparatorsVisible(a) {
            this.t = a;
            this.g("RowSeparatorsVisible");
        }
        get textColor() {
            return this.dw;
        }
        set textColor(a) {
            this.dw = a;
            this.g("TextColor");
        }
        get actualTextColor() {
            return this.cw;
        }
        set actualTextColor(a) {
            this.cw = a;
            this.g("ActualTextColor");
        }
        get fontFamily() {
            return this.c8;
        }
        set fontFamily(a) {
            this.c8 = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.a6;
        }
        set fontSize(a) {
            this.a6 = a;
            this.g("FontSize");
        }
        get fontStyle() {
            return this.c9;
        }
        set fontStyle(a) {
            this.c9 = a;
            this.g("FontStyle");
        }
        get fontWeight() {
            return this.da;
        }
        set fontWeight(a) {
            this.da = a;
            this.g("FontWeight");
        }
        get underlineColor() {
            return this.dz;
        }
        set underlineColor(a) {
            this.dz = a;
            this.g("UnderlineColor");
        }
        get actualUnderlineColor() {
            return this.cx;
        }
        set actualUnderlineColor(a) {
            this.cx = a;
            this.g("ActualUnderlineColor");
        }
        get underlineOpacity() {
            return this.bb;
        }
        set underlineOpacity(a) {
            this.bb = a;
            this.g("UnderlineOpacity");
        }
        get actualUnderlineOpacity() {
            return this.ap;
        }
        set actualUnderlineOpacity(a) {
            this.ap = a;
            this.g("ActualUnderlineOpacity");
        }
        get underlineRippleColor() {
            return this.d0;
        }
        set underlineRippleColor(a) {
            this.d0 = a;
            this.g("UnderlineRippleColor");
        }
        get actualUnderlineRippleColor() {
            return this.cy;
        }
        set actualUnderlineRippleColor(a) {
            this.cy = a;
            this.g("ActualUnderlineRippleColor");
        }
        get underlineRippleOpacity() {
            return this.bc;
        }
        set underlineRippleOpacity(a) {
            this.bc = a;
            this.g("UnderlineRippleOpacity");
        }
        get actualUnderlineRippleOpacity() {
            return this.aq;
        }
        set actualUnderlineRippleOpacity(a) {
            this.aq = a;
            this.g("ActualUnderlineRippleOpacity");
        }
        get underlineRippleWidth() {
            return this.bd;
        }
        set underlineRippleWidth(a) {
            this.bd = a;
            this.g("UnderlineRippleWidth");
        }
        get actualUnderlineRippleWidth() {
            return this.ar;
        }
        set actualUnderlineRippleWidth(a) {
            this.ar = a;
            this.g("ActualUnderlineRippleWidth");
        }
        get underlineWidth() {
            return this.be;
        }
        set underlineWidth(a) {
            this.be = a;
            this.g("UnderlineWidth");
        }
        get actualUnderlineWidth() {
            return this.as;
        }
        set actualUnderlineWidth(a) {
            this.as = a;
            this.g("ActualUnderlineWidth");
        }
        get allowFilter() {
            return this.q;
        }
        set allowFilter(a) {
            this.q = a;
            this.g("AllowFilter");
        }
        get sortMode() {
            return this.dt;
        }
        set sortMode(a) {
            this.dt = a;
            this.g("SortMode");
        }
        get textField() {
            return this.dx;
        }
        set textField(a) {
            this.dx = a;
            this.g("TextField");
        }
        get text() {
            return this.du;
        }
        set text(a) {
            this.du = a;
            this.g("Text");
        }
        get valueField() {
            return this.k;
        }
        set valueField(a) {
            this.k = a;
            this.g("ValueField");
        }
        get actualValueField() {
            return this.h;
        }
        set actualValueField(a) {
            this.h = a;
            this.g("ActualValueField");
        }
        get value() {
            return this.cj;
        }
        set value(a) {
            this.cj = a;
            this.g("Value");
        }
        get gotFocusRef() {
            return this.db;
        }
        set gotFocusRef(a) {
            this.db = a;
            this.g("GotFocusRef");
        }
        get lostFocusRef() {
            return this.dj;
        }
        set lostFocusRef(a) {
            this.dj = a;
            this.g("LostFocusRef");
        }
        get textValueChangedRef() {
            return this.dy;
        }
        set textValueChangedRef(a) {
            this.dy = a;
            this.g("TextValueChangedRef");
        }
        get selectedValueChangedRef() {
            return this.ds;
        }
        set selectedValueChangedRef(a) {
            this.ds = a;
            this.g("SelectedValueChangedRef");
        }
        get keyDownRef() {
            return this.dd;
        }
        set keyDownRef(a) {
            this.dd = a;
            this.g("KeyDownRef");
        }
        get changeRef() {
            return this.c2;
        }
        set changeRef(a) {
            this.c2 = a;
            this.g("ChangeRef");
        }
        get changingRef() {
            return this.c3;
        }
        set changingRef(a) {
            this.c3 = a;
            this.g("ChangingRef");
        }
        get textChangeRef() {
            return this.dv;
        }
        set textChangeRef(a) {
            this.dv = a;
            this.g("TextChangeRef");
        }
        get valueChangeRef() {
            return this.d1;
        }
        set valueChangeRef(a) {
            this.d1 = a;
            this.g("ValueChangeRef");
        }
        get defaultColumnWidth() {
            return this.a2;
        }
        set defaultColumnWidth(a) {
            this.a2 = a;
            this.g("DefaultColumnWidth");
        }
    }
    MultiColumnComboBoxDescription.$t = /*@__PURE__*/ markType(MultiColumnComboBoxDescription, 'MultiColumnComboBoxDescription', Description.$);
    return MultiColumnComboBoxDescription;
})();
