// igniteui-react-core/esm2015/lib/NativeUICheckbox.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { markType } from "./type";
/**
 * @hidden
 */
export let NativeUICheckbox = /*@__PURE__*/ (() => {
    class NativeUICheckbox extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bg = null;
        }
        get_e() {
            return 6;
        }
        get bi() {
            let a = this.ae(4);
            return a;
        }
        set bi(a) {
            this.a0(4, a);
        }
        get bj() {
            let a = this.ae(7);
            return a;
        }
        set bj(a) {
            this.a0(7, a);
        }
        get bk() {
            let a = this.ae(17);
            return a;
        }
        set bk(a) {
            this.a0(17, a);
        }
        get bh() {
            return this.bg;
        }
        set bh(a) {
            let b = this.bg;
            this.bg = a;
            if (this.bg != b) {
                if (b != null) {
                    this.ay(2, b);
                }
                if (this.bg != null) {
                    this.an(2, this.bg);
                }
            }
        }
    }
    NativeUICheckbox.$t = /*@__PURE__*/ markType(NativeUICheckbox, 'NativeUICheckbox', NativeUIComponent.$);
    return NativeUICheckbox;
})();
