// igniteui-react-core/esm2015/lib/NativeUISlider.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { markType } from "./type";
/**
 * @hidden
 */
export let NativeUISlider = /*@__PURE__*/ (() => {
    class NativeUISlider extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bg = null;
            this.bh = null;
        }
        get_e() {
            return 5;
        }
        get bn() {
            let a = this.ae(2);
            return a;
        }
        set bn(a) {
            this.a0(2, a);
        }
        get bl() {
            let a = this.ae(12);
            return a;
        }
        set bl(a) {
            this.a0(12, a);
        }
        get bk() {
            let a = this.ae(13);
            return a;
        }
        set bk(a) {
            this.a0(13, a);
        }
        get bm() {
            let a = this.ae(14);
            return a;
        }
        set bm(a) {
            this.a0(14, a);
        }
        get bi() {
            return this.bg;
        }
        set bi(a) {
            let b = this.bg;
            this.bg = a;
            if (this.bg != b) {
                if (b != null) {
                    this.ay(4, b);
                }
                if (this.bg != null) {
                    this.an(4, this.bg);
                }
            }
        }
        get bj() {
            return this.bh;
        }
        set bj(a) {
            let b = this.bh;
            this.bh = a;
            if (this.bh != b) {
                if (b != null) {
                    this.ay(0, b);
                }
                if (this.bh != null) {
                    this.an(0, this.bh);
                }
            }
        }
    }
    NativeUISlider.$t = /*@__PURE__*/ markType(NativeUISlider, 'NativeUISlider', NativeUIComponent.$);
    return NativeUISlider;
})();
