// igniteui-react-core/esm2015/lib/NativeUIPopup.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIContainer } from "./NativeUIContainer";
import { NativeUIPopupAnimationType_$type } from "./NativeUIPopupAnimationType";
import { NativeUIPopupPointerPosition_$type } from "./NativeUIPopupPointerPosition";
import { typeGetValue, enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIPopup = /*@__PURE__*/ (() => {
    class NativeUIPopup extends NativeUIContainer {
        constructor() {
            super(...arguments);
            this.br = null;
            this.bp = null;
        }
        get bn() {
            return this.d;
        }
        get by() {
            return this.ae(22);
        }
        get bu() {
            return this.ae(25);
        }
        set bu(a) {
            this.a0(25, a);
        }
        get b0() {
            return typeGetValue(this.ae(26));
        }
        set b0(a) {
            this.a0(26, a);
        }
        get bv() {
            return this.ae(39);
        }
        set bv(a) {
            this.a0(39, a);
        }
        get bo() {
            return EnumUtil.getEnumValue(NativeUIPopupAnimationType_$type, this.ae(35));
        }
        set bo(a) {
            this.a0(35, enumGetBox(NativeUIPopupAnimationType_$type, a));
        }
        get b6() {
            return this.ae(23);
        }
        set b6(a) {
            this.a0(23, a);
        }
        get bz() {
            return this.ae(17);
        }
        set bz(a) {
            this.a0(17, a);
        }
        get_u() {
            return this.ae(36);
        }
        set_u(a) {
            this.a0(36, a);
        }
        get bw() {
            return this.ae(37);
        }
        set bw(a) {
            this.a0(37, a);
        }
        get b1() {
            return typeGetValue(this.ae(40));
        }
        set b1(a) {
            this.a0(40, a);
        }
        get bx() {
            return this.ae(41);
        }
        set bx(a) {
            this.a0(41, a);
        }
        get bt() {
            return EnumUtil.getEnumValue(NativeUIPopupPointerPosition_$type, this.ae(43));
        }
        set bt(a) {
            this.a0(43, enumGetBox(NativeUIPopupPointerPosition_$type, a));
        }
        get b7() {
            return this.ae(42);
        }
        set b7(a) {
            this.a0(42, a);
        }
        get bs() {
            return this.br;
        }
        set bs(a) {
            let b = this.br;
            this.br = a;
            if (this.br != b) {
                if (b != null) {
                    this.ay(5, b);
                }
                if (this.br != null) {
                    this.an(5, this.br);
                }
            }
        }
        get bq() {
            return this.bp;
        }
        set bq(a) {
            let b = this.bp;
            this.bp = a;
            if (this.bp != b) {
                if (b != null) {
                    this.ay(6, b);
                }
                if (this.bp != null) {
                    this.an(6, this.bp);
                }
            }
        }
        b5(a, b, c) {
            this.bn.showRelativeToExclusionRect(this.ak, a, b, c);
        }
        b2() {
            this.bn.close(this.ak);
        }
        b3() {
            this.bn.notifyContentSizeChanged(this.ak);
        }
        b4(a) {
            this.bn.notifyExclusionRectChanged(this.ak, a);
        }
    }
    NativeUIPopup.$t = /*@__PURE__*/ markType(NativeUIPopup, 'NativeUIPopup', NativeUIContainer.$);
    return NativeUIPopup;
})();
