// igniteui-react-core/esm2015/lib/NativeUI.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { UIElementPropertyAccessorFactory } from "./UIElementPropertyAccessorFactory";
import { MatrixTransform } from "./MatrixTransform";
import { TransformMatrixHelpler } from "./TransformMatrixHelpler";
/**
 * @hidden
 */
export let NativeUI = /*@__PURE__*/ (() => {
    class NativeUI extends Base {
        static ae(a, b, c) {
            let d = UIElementPropertyAccessorFactory.c.a(a, 0);
            d.c(b, c);
        }
        static k(a, b) {
            let c = UIElementPropertyAccessorFactory.c.a(a, 0);
            return c.b(b);
        }
        static w(a, b) {
            NativeUI.ae(1, a, b);
        }
        static f(a) {
            return NativeUI.k(1, a);
        }
        static l(a, b) {
            NativeUI.ae(2, a, b);
        }
        static ah(a) {
            return NativeUI.k(2, a);
        }
        static m(a, b) {
            NativeUI.ae(3, a, b);
        }
        static ai(a) {
            return NativeUI.k(3, a);
        }
        static o(a, b) {
            NativeUI.ae(4, a, b);
        }
        static am(a) {
            return NativeUI.k(4, a);
        }
        static n(a, b) {
            NativeUI.ae(5, a, b);
        }
        static x(a, b) {
            NativeUI.ae(6, a, b);
        }
        static ao(a) {
            return NativeUI.k(6, a);
        }
        static ag(a) {
            return NativeUI.k(5, a);
        }
        static p(a, b) {
            NativeUI.ae(7, a, b);
        }
        static aj(a) {
            return NativeUI.k(7, a);
        }
        static r(a, b) {
            NativeUI.ae(8, a, b);
        }
        static c(a) {
            return NativeUI.k(8, a);
        }
        static af(a, b) {
            NativeUI.ae(9, a, b);
        }
        static j(a) {
            return NativeUI.k(9, a);
        }
        static s(a, b) {
            NativeUI.ae(10, a, b);
        }
        static d(a) {
            return NativeUI.k(10, a);
        }
        static u(a, b) {
            NativeUI.ae(11, a, b);
        }
        static ac(a, b) {
            NativeUI.ae(12, a, b);
        }
        static e(a) {
            return NativeUI.k(11, a);
        }
        static i(a) {
            return NativeUI.k(12, a);
        }
        static v(a, b) {
            NativeUI.ae(13, a, b);
        }
        static an(a) {
            return NativeUI.k(13, a);
        }
        static ad(a, b) {
            let c = TransformMatrixHelpler.f(b);
            let d = new MatrixTransform();
            d.j = c;
            NativeUI.ae(14, a, d);
        }
        static al(a) {
            return NativeUI.k(14, a);
        }
        static y(a, b) {
            NativeUI.ae(15, a, b);
        }
        static q(a, b) {
            NativeUI.ae(16, a, b);
        }
        static ak(a) {
            return NativeUI.k(15, a);
        }
        static ab(a, b) {
            NativeUI.ae(17, a, b);
        }
        static h(a) {
            return NativeUI.k(17, a);
        }
        static z(a, b) {
            NativeUI.ae(18, a, b);
        }
        static a(a) {
            return NativeUI.k(18, a);
        }
        static aa(a, b) {
            NativeUI.ae(19, a, b);
        }
        static g(a) {
            return NativeUI.k(19, a);
        }
        static t(a, b) {
            NativeUI.ae(21, a, b);
        }
        static b(a) {
            return NativeUI.k(21, a);
        }
    }
    NativeUI.$t = /*@__PURE__*/ markType(NativeUI, 'NativeUI');
    return NativeUI;
})();
