// igniteui-react-core/esm2015/lib/NativeUIAbsolutePanel.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIContainer } from "./NativeUIContainer";
import { markType } from "./type";
/**
 * @hidden
 */
export let NativeUIAbsolutePanel = /*@__PURE__*/ (() => {
    class NativeUIAbsolutePanel extends NativeUIContainer {
        get_e() {
            return 9;
        }
        get bn() {
            return this.d;
        }
        bq(a, b) {
            this.bn.setX(this.ak, a, b);
        }
        br(a, b) {
            this.bn.setY(this.ak, a, b);
        }
        bo(a) {
            return this.bn.getX(this.ak, a);
        }
        bp(a) {
            return this.bn.getY(this.ak, a);
        }
    }
    NativeUIAbsolutePanel.$t = /*@__PURE__*/ markType(NativeUIAbsolutePanel, 'NativeUIAbsolutePanel', NativeUIContainer.$);
    return NativeUIAbsolutePanel;
})();
