// igniteui-react-core/esm2015/lib/NativeUIDropDown.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIDropDown = /*@__PURE__*/ (() => {
    class NativeUIDropDown extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bh = null;
        }
        get_e() {
            return 4;
        }
        get bg() {
            let a = this.ae(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bg(a) {
            this.a0(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get bi() {
            return this.bh;
        }
        set bi(a) {
            let b = this.bh;
            this.bh = a;
            if (this.bh != b) {
                if (b != null) {
                    this.ay(3, b);
                }
                if (this.bh != null) {
                    this.an(3, this.bh);
                }
            }
        }
        get bk() {
            let a = this.ae(9);
            return a;
        }
        set bk(a) {
            this.a0(9, a);
        }
        get bj() {
            let a = this.ae(8);
            return a;
        }
        set bj(a) {
            this.a0(8, a);
        }
    }
    NativeUIDropDown.$t = /*@__PURE__*/ markType(NativeUIDropDown, 'NativeUIDropDown', NativeUIComponent.$);
    return NativeUIDropDown;
})();
