// igniteui-react-core/esm2015/lib/NativeUIInput.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIInputDisplayType_$type } from "./NativeUIInputDisplayType";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { NativeUIInputType_$type } from "./NativeUIInputType";
import { NativeUIComponentHorizontalAlignment_$type } from "./NativeUIComponentHorizontalAlignment";
/**
 * @hidden
 */
export let NativeUIInput = /*@__PURE__*/ (() => {
    class NativeUIInput extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bk = null;
            this.bg = null;
        }
        get bh() {
            return this.d;
        }
        get bm() {
            let a = this.ae(0);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIInputDisplayType_$type, a);
        }
        set bm(a) {
            this.a0(0, enumGetBox(NativeUIInputDisplayType_$type, a));
        }
        get bj() {
            let a = this.ae(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bj(a) {
            this.a0(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get bq() {
            let a = this.ae(2);
            return a;
        }
        set bq(a) {
            this.a0(2, a);
        }
        get bl() {
            return this.bk;
        }
        set bl(a) {
            let b = this.bk;
            this.bk = a;
            if (this.bk != b) {
                if (b != null) {
                    this.ay(0, b);
                }
                if (this.bk != null) {
                    this.an(0, this.bk);
                }
            }
        }
        get bn() {
            let a = this.ae(19);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIInputType_$type, a);
        }
        set bn(a) {
            this.a0(19, enumGetBox(NativeUIInputType_$type, a));
        }
        get bi() {
            let a = this.ae(20);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIComponentHorizontalAlignment_$type, a);
        }
        set bi(a) {
            this.a0(20, enumGetBox(NativeUIComponentHorizontalAlignment_$type, a));
        }
        get bt() {
            let a = this.ae(33);
            return a == null ? null : a;
        }
        set bt(a) {
            this.a0(33, a);
        }
        get bp() {
            let a = this.ae(21);
            return a == null ? false : a;
        }
        set bp(a) {
            this.a0(21, a);
        }
        get bo() {
            let a = this.ae(32);
            return a == null ? false : a;
        }
        set bo(a) {
            this.a0(32, a);
        }
        br() {
            this.bh.select(this.ak);
        }
        bs(a, b) {
            this.bh.setSelectionRange(this.ak, a, b);
        }
    }
    NativeUIInput.$t = /*@__PURE__*/ markType(NativeUIInput, 'NativeUIInput', NativeUIComponent.$);
    return NativeUIInput;
})();
