// igniteui-react-core/esm2015/lib/NumericAxisBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AxisDescription } from "./AxisDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let NumericAxisBaseDescription = /*@__PURE__*/ (() => {
    class NumericAxisBaseDescription extends AxisDescription {
        get_type() {
            return "NumericAxisBase";
        }
        constructor() {
            super();
            this.es = null;
            this.d7 = 0;
            this.d1 = 0;
            this.d4 = 0;
            this.d6 = 0;
            this.d0 = 0;
            this.d3 = 0;
            this.d5 = 0;
            this.dz = 0;
            this.d8 = 0;
            this.d2 = 0;
            this.d9 = 0;
            this.dt = false;
            this.dq = false;
            this.dr = false;
            this.el = 0;
            this.et = null;
            this.en = null;
            this.dm = null;
            this.ds = false;
            this.dp = false;
            this.eq = null;
            this.ep = null;
            this.eo = null;
            this.er = null;
        }
        get autoRangeBufferMode() {
            return this.es;
        }
        set autoRangeBufferMode(a) {
            this.es = a;
            this.g("AutoRangeBufferMode");
        }
        get minimumValue() {
            return this.d7;
        }
        set minimumValue(a) {
            this.d7 = a;
            this.g("MinimumValue");
        }
        get actualMinimumValue() {
            return this.d1;
        }
        set actualMinimumValue(a) {
            this.d1 = a;
            this.g("ActualMinimumValue");
        }
        get actualVisibleMinimumValue() {
            return this.d4;
        }
        set actualVisibleMinimumValue(a) {
            this.d4 = a;
            this.g("ActualVisibleMinimumValue");
        }
        get maximumValue() {
            return this.d6;
        }
        set maximumValue(a) {
            this.d6 = a;
            this.g("MaximumValue");
        }
        get actualMaximumValue() {
            return this.d0;
        }
        set actualMaximumValue(a) {
            this.d0 = a;
            this.g("ActualMaximumValue");
        }
        get actualVisibleMaximumValue() {
            return this.d3;
        }
        set actualVisibleMaximumValue(a) {
            this.d3 = a;
            this.g("ActualVisibleMaximumValue");
        }
        get interval() {
            return this.d5;
        }
        set interval(a) {
            this.d5 = a;
            this.g("Interval");
        }
        get actualInterval() {
            return this.dz;
        }
        set actualInterval(a) {
            this.dz = a;
            this.g("ActualInterval");
        }
        get minorInterval() {
            return this.d8;
        }
        set minorInterval(a) {
            this.d8 = a;
            this.g("MinorInterval");
        }
        get actualMinorInterval() {
            return this.d2;
        }
        set actualMinorInterval(a) {
            this.d2 = a;
            this.g("ActualMinorInterval");
        }
        get referenceValue() {
            return this.d9;
        }
        set referenceValue(a) {
            this.d9 = a;
            this.g("ReferenceValue");
        }
        get isLogarithmic() {
            return this.dt;
        }
        set isLogarithmic(a) {
            this.dt = a;
            this.g("IsLogarithmic");
        }
        get actualIsLogarithmic() {
            return this.dq;
        }
        set actualIsLogarithmic(a) {
            this.dq = a;
            this.g("ActualIsLogarithmic");
        }
        get favorLabellingScaleEnd() {
            return this.dr;
        }
        set favorLabellingScaleEnd(a) {
            this.dr = a;
            this.g("FavorLabellingScaleEnd");
        }
        get logarithmBase() {
            return this.el;
        }
        set logarithmBase(a) {
            this.el = a;
            this.g("LogarithmBase");
        }
        get formatAbbreviatedLabelRef() {
            return this.et;
        }
        set formatAbbreviatedLabelRef(a) {
            this.et = a;
            this.g("FormatAbbreviatedLabelRef");
        }
        get abbreviatedLabelFormat() {
            return this.en;
        }
        set abbreviatedLabelFormat(a) {
            this.en = a;
            this.g("AbbreviatedLabelFormat");
        }
        get abbreviatedLabelFormatSpecifiers() {
            return this.dm;
        }
        set abbreviatedLabelFormatSpecifiers(a) {
            this.dm = a;
            this.g("AbbreviatedLabelFormatSpecifiers");
        }
        get isFormattingAbbreviatedLargeNumber() {
            return this.ds;
        }
        set isFormattingAbbreviatedLargeNumber(a) {
            this.ds = a;
            this.g("IsFormattingAbbreviatedLargeNumber");
        }
        get abbreviateLargeNumbers() {
            return this.dp;
        }
        set abbreviateLargeNumbers(a) {
            this.dp = a;
            this.g("AbbreviateLargeNumbers");
        }
        get actualMinimumValueChangeRef() {
            return this.eq;
        }
        set actualMinimumValueChangeRef(a) {
            this.eq = a;
            this.g("ActualMinimumValueChangeRef");
        }
        get actualMaximumValueChangeRef() {
            return this.ep;
        }
        set actualMaximumValueChangeRef(a) {
            this.ep = a;
            this.g("ActualMaximumValueChangeRef");
        }
        get actualIntervalChangeRef() {
            return this.eo;
        }
        set actualIntervalChangeRef(a) {
            this.eo = a;
            this.g("ActualIntervalChangeRef");
        }
        get actualMinorIntervalChangeRef() {
            return this.er;
        }
        set actualMinorIntervalChangeRef(a) {
            this.er = a;
            this.g("ActualMinorIntervalChangeRef");
        }
    }
    NumericAxisBaseDescription.$t = /*@__PURE__*/ markType(NumericAxisBaseDescription, 'NumericAxisBaseDescription', AxisDescription.$);
    return NumericAxisBaseDescription;
})();
