// igniteui-react-core/esm2015/lib/NumericColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let NumericColumnDescription = /*@__PURE__*/ (() => {
    class NumericColumnDescription extends DataGridColumnDescription {
        get_type() {
            return "NumericColumn";
        }
        constructor() {
            super();
            this.f1 = null;
            this.f3 = null;
            this.f2 = null;
            this.f4 = null;
            this.fo = 0;
            this.fp = 0;
            this.fq = 0;
            this.fm = false;
            this.f0 = null;
            this.fh = null;
            this.fz = null;
            this.fx = null;
            this.fv = null;
            this.fj = null;
            this.fw = null;
            this.fy = null;
        }
        get negativePrefix() {
            return this.f1;
        }
        set negativePrefix(a) {
            this.f1 = a;
            this.g("NegativePrefix");
        }
        get positivePrefix() {
            return this.f3;
        }
        set positivePrefix(a) {
            this.f3 = a;
            this.g("PositivePrefix");
        }
        get negativeSuffix() {
            return this.f2;
        }
        set negativeSuffix(a) {
            this.f2 = a;
            this.g("NegativeSuffix");
        }
        get positiveSuffix() {
            return this.f4;
        }
        set positiveSuffix(a) {
            this.f4 = a;
            this.g("PositiveSuffix");
        }
        get maxFractionDigits() {
            return this.fo;
        }
        set maxFractionDigits(a) {
            this.fo = a;
            this.g("MaxFractionDigits");
        }
        get minFractionDigits() {
            return this.fp;
        }
        set minFractionDigits(a) {
            this.fp = a;
            this.g("MinFractionDigits");
        }
        get minIntegerDigits() {
            return this.fq;
        }
        set minIntegerDigits(a) {
            this.fq = a;
            this.g("MinIntegerDigits");
        }
        get showGroupingSeparator() {
            return this.fm;
        }
        set showGroupingSeparator(a) {
            this.fm = a;
            this.g("ShowGroupingSeparator");
        }
        get formatString() {
            return this.f0;
        }
        set formatString(a) {
            this.f0 = a;
            this.g("FormatString");
        }
        get formatSpecifiers() {
            return this.fh;
        }
        set formatSpecifiers(a) {
            this.fh = a;
            this.g("FormatSpecifiers");
        }
        get formatOverrideRef() {
            return this.fz;
        }
        set formatOverrideRef(a) {
            this.fz = a;
            this.g("FormatOverrideRef");
        }
        get editorType() {
            return this.fx;
        }
        set editorType(a) {
            this.fx = a;
            this.g("EditorType");
        }
        get editorDataSourceRef() {
            return this.fv;
        }
        set editorDataSourceRef(a) {
            this.fv = a;
            this.g("EditorDataSourceRef");
        }
        get editorDataSource() {
            return this.fj;
        }
        set editorDataSource(a) {
            this.fj = a;
            this.g("EditorDataSource");
        }
        get editorTextField() {
            return this.fw;
        }
        set editorTextField(a) {
            this.fw = a;
            this.g("EditorTextField");
        }
        get editorValueField() {
            return this.fy;
        }
        set editorValueField(a) {
            this.fy = a;
            this.g("EditorValueField");
        }
    }
    NumericColumnDescription.$t = /*@__PURE__*/ markType(NumericColumnDescription, 'NumericColumnDescription', DataGridColumnDescription.$);
    NumericColumnDescription.__canMarshalByValue1 = true;
    NumericColumnDescription.__marshalByValueAlias1 = "NumericColumn";
    return NumericColumnDescription;
})();
