// igniteui-react-core/esm2015/lib/RadialPieSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnchoredRadialSeriesDescription } from "./AnchoredRadialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RadialPieSeriesDescription = /*@__PURE__*/ (() => {
    class RadialPieSeriesDescription extends AnchoredRadialSeriesDescription {
        get_type() {
            return "RadialPieSeries";
        }
        constructor() {
            super();
            this.ev = 0;
            this.ew = 0;
        }
        get radiusX() {
            return this.ev;
        }
        set radiusX(a) {
            this.ev = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.ew;
        }
        set radiusY(a) {
            this.ew = a;
            this.g("RadiusY");
        }
    }
    RadialPieSeriesDescription.$t = /*@__PURE__*/ markType(RadialPieSeriesDescription, 'RadialPieSeriesDescription', AnchoredRadialSeriesDescription.$);
    return RadialPieSeriesDescription;
})();
