// igniteui-react-core/esm2015/lib/RadialGaugeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let RadialGaugeDescription = /*@__PURE__*/ (() => {
    class RadialGaugeDescription extends Description {
        get fontFamily() {
            return this.cv;
        }
        set fontFamily(a) {
            this.cv = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.ad;
        }
        set fontSize(a) {
            this.ad = a;
            this.g("FontSize");
        }
        get fontWeight() {
            return this.cx;
        }
        set fontWeight(a) {
            this.cx = a;
            this.g("FontWeight");
        }
        get fontStyle() {
            return this.cw;
        }
        set fontStyle(a) {
            this.cw = a;
            this.g("FontStyle");
        }
        get width() {
            return this.de;
        }
        set width(a) {
            this.de = a;
            this.g("Width");
        }
        get height() {
            return this.cz;
        }
        set height(a) {
            this.cz = a;
            this.g("Height");
        }
        get background() {
            return this.co;
        }
        set background(a) {
            this.co = a;
            this.g("Background");
        }
        get_type() {
            return "RadialGauge";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.cv = null;
            this.ad = 0;
            this.cx = null;
            this.cw = null;
            this.de = null;
            this.cz = null;
            this.co = null;
            this.i = null;
            this.j = null;
            this.ai = 0;
            this.u = 0;
            this.ah = 0;
            this.t = 0;
            this.ae = 0;
            this.ab = 0;
            this.ac = 0;
            this.ba = 0;
            this.a4 = 0;
            this.a1 = 0;
            this.da = null;
            this.cj = 0;
            this.dc = null;
            this.c2 = null;
            this.c3 = null;
            this.aw = 0;
            this.ap = 0;
            this.c7 = null;
            this.ax = 0;
            this.aq = 0;
            this.ao = 0;
            this.an = 0;
            this.av = 0;
            this.au = 0;
            this.at = 0;
            this.ar = 0;
            this.c6 = null;
            this.a5 = 0;
            this.c4 = null;
            this.c5 = null;
            this.ay = 0;
            this.as = 0;
            this.a2 = 0;
            this.af = 0;
            this.ag = 0;
            this.a7 = 0;
            this.a6 = 0;
            this.a8 = 0;
            this.db = null;
            this.cu = null;
            this.al = 0;
            this.ak = 0;
            this.am = 0;
            this.c1 = null;
            this.aj = 0;
            this.c8 = null;
            this.cp = null;
            this.cq = null;
            this.aa = 0;
            this.y = 0;
            this.z = 0;
            this.a3 = 0;
            this.c9 = null;
            this.w = 0;
            this.x = 0;
            this.cr = null;
            this.k = null;
            this.a0 = 0;
            this.cs = null;
            this.q = false;
            this.p = false;
            this.ct = null;
            this.c0 = null;
            this.h = null;
            this.a9 = 0;
            this.az = 0;
            this.v = 0;
            this.cy = null;
            this.cn = null;
            this.dd = null;
            this.cm = null;
            this.cl = null;
        }
        get rangeBrushes() {
            return this.i;
        }
        set rangeBrushes(a) {
            this.i = a;
            this.g("RangeBrushes");
        }
        get rangeOutlines() {
            return this.j;
        }
        set rangeOutlines(a) {
            this.j = a;
            this.g("RangeOutlines");
        }
        get minimumValue() {
            return this.ai;
        }
        set minimumValue(a) {
            this.ai = a;
            this.g("MinimumValue");
        }
        get actualMinimumValue() {
            return this.u;
        }
        set actualMinimumValue(a) {
            this.u = a;
            this.g("ActualMinimumValue");
        }
        get maximumValue() {
            return this.ah;
        }
        set maximumValue(a) {
            this.ah = a;
            this.g("MaximumValue");
        }
        get actualMaximumValue() {
            return this.t;
        }
        set actualMaximumValue(a) {
            this.t = a;
            this.g("ActualMaximumValue");
        }
        get interval() {
            return this.ae;
        }
        set interval(a) {
            this.ae = a;
            this.g("Interval");
        }
        get centerX() {
            return this.ab;
        }
        set centerX(a) {
            this.ab = a;
            this.g("CenterX");
        }
        get centerY() {
            return this.ac;
        }
        set centerY(a) {
            this.ac = a;
            this.g("CenterY");
        }
        get value() {
            return this.ba;
        }
        set value(a) {
            this.ba = a;
            this.g("Value");
        }
        get scaleStartAngle() {
            return this.a4;
        }
        set scaleStartAngle(a) {
            this.a4 = a;
            this.g("ScaleStartAngle");
        }
        get scaleEndAngle() {
            return this.a1;
        }
        set scaleEndAngle(a) {
            this.a1 = a;
            this.g("ScaleEndAngle");
        }
        get scaleSweepDirection() {
            return this.da;
        }
        set scaleSweepDirection(a) {
            this.da = a;
            this.g("ScaleSweepDirection");
        }
        get transitionDuration() {
            return this.cj;
        }
        set transitionDuration(a) {
            this.cj = a;
            this.g("TransitionDuration");
        }
        get transitionEasingFunctionRef() {
            return this.dc;
        }
        set transitionEasingFunctionRef(a) {
            this.dc = a;
            this.g("TransitionEasingFunctionRef");
        }
        get needleBrush() {
            return this.c2;
        }
        set needleBrush(a) {
            this.c2 = a;
            this.g("NeedleBrush");
        }
        get needleOutline() {
            return this.c3;
        }
        set needleOutline(a) {
            this.c3 = a;
            this.g("NeedleOutline");
        }
        get needleStartExtent() {
            return this.aw;
        }
        set needleStartExtent(a) {
            this.aw = a;
            this.g("NeedleStartExtent");
        }
        get needleEndExtent() {
            return this.ap;
        }
        set needleEndExtent(a) {
            this.ap = a;
            this.g("NeedleEndExtent");
        }
        get needleShape() {
            return this.c7;
        }
        set needleShape(a) {
            this.c7 = a;
            this.g("NeedleShape");
        }
        get needleStartWidthRatio() {
            return this.ax;
        }
        set needleStartWidthRatio(a) {
            this.ax = a;
            this.g("NeedleStartWidthRatio");
        }
        get needleEndWidthRatio() {
            return this.aq;
        }
        set needleEndWidthRatio(a) {
            this.aq = a;
            this.g("NeedleEndWidthRatio");
        }
        get needleBaseFeatureWidthRatio() {
            return this.ao;
        }
        set needleBaseFeatureWidthRatio(a) {
            this.ao = a;
            this.g("NeedleBaseFeatureWidthRatio");
        }
        get needleBaseFeatureExtent() {
            return this.an;
        }
        set needleBaseFeatureExtent(a) {
            this.an = a;
            this.g("NeedleBaseFeatureExtent");
        }
        get needlePointFeatureWidthRatio() {
            return this.av;
        }
        set needlePointFeatureWidthRatio(a) {
            this.av = a;
            this.g("NeedlePointFeatureWidthRatio");
        }
        get needlePointFeatureExtent() {
            return this.au;
        }
        set needlePointFeatureExtent(a) {
            this.au = a;
            this.g("NeedlePointFeatureExtent");
        }
        get needlePivotWidthRatio() {
            return this.at;
        }
        set needlePivotWidthRatio(a) {
            this.at = a;
            this.g("NeedlePivotWidthRatio");
        }
        get needlePivotInnerWidthRatio() {
            return this.ar;
        }
        set needlePivotInnerWidthRatio(a) {
            this.ar = a;
            this.g("NeedlePivotInnerWidthRatio");
        }
        get needlePivotShape() {
            return this.c6;
        }
        set needlePivotShape(a) {
            this.c6 = a;
            this.g("NeedlePivotShape");
        }
        get scaleStartExtent() {
            return this.a5;
        }
        set scaleStartExtent(a) {
            this.a5 = a;
            this.g("ScaleStartExtent");
        }
        get needlePivotBrush() {
            return this.c4;
        }
        set needlePivotBrush(a) {
            this.c4 = a;
            this.g("NeedlePivotBrush");
        }
        get needlePivotOutline() {
            return this.c5;
        }
        set needlePivotOutline(a) {
            this.c5 = a;
            this.g("NeedlePivotOutline");
        }
        get needleStrokeThickness() {
            return this.ay;
        }
        set needleStrokeThickness(a) {
            this.ay = a;
            this.g("NeedleStrokeThickness");
        }
        get needlePivotStrokeThickness() {
            return this.as;
        }
        set needlePivotStrokeThickness(a) {
            this.as = a;
            this.g("NeedlePivotStrokeThickness");
        }
        get scaleEndExtent() {
            return this.a2;
        }
        set scaleEndExtent(a) {
            this.a2 = a;
            this.g("ScaleEndExtent");
        }
        get labelExtent() {
            return this.af;
        }
        set labelExtent(a) {
            this.af = a;
            this.g("LabelExtent");
        }
        get labelInterval() {
            return this.ag;
        }
        set labelInterval(a) {
            this.ag = a;
            this.g("LabelInterval");
        }
        get tickStartExtent() {
            return this.a7;
        }
        set tickStartExtent(a) {
            this.a7 = a;
            this.g("TickStartExtent");
        }
        get tickEndExtent() {
            return this.a6;
        }
        set tickEndExtent(a) {
            this.a6 = a;
            this.g("TickEndExtent");
        }
        get tickStrokeThickness() {
            return this.a8;
        }
        set tickStrokeThickness(a) {
            this.a8 = a;
            this.g("TickStrokeThickness");
        }
        get tickBrush() {
            return this.db;
        }
        set tickBrush(a) {
            this.db = a;
            this.g("TickBrush");
        }
        get fontBrush() {
            return this.cu;
        }
        set fontBrush(a) {
            this.cu = a;
            this.g("FontBrush");
        }
        get minorTickStartExtent() {
            return this.al;
        }
        set minorTickStartExtent(a) {
            this.al = a;
            this.g("MinorTickStartExtent");
        }
        get minorTickEndExtent() {
            return this.ak;
        }
        set minorTickEndExtent(a) {
            this.ak = a;
            this.g("MinorTickEndExtent");
        }
        get minorTickStrokeThickness() {
            return this.am;
        }
        set minorTickStrokeThickness(a) {
            this.am = a;
            this.g("MinorTickStrokeThickness");
        }
        get minorTickBrush() {
            return this.c1;
        }
        set minorTickBrush(a) {
            this.c1 = a;
            this.g("MinorTickBrush");
        }
        get minorTickCount() {
            return this.aj;
        }
        set minorTickCount(a) {
            this.aj = a;
            this.g("MinorTickCount");
        }
        get scaleBrush() {
            return this.c8;
        }
        set scaleBrush(a) {
            this.c8 = a;
            this.g("ScaleBrush");
        }
        get backingBrush() {
            return this.cp;
        }
        set backingBrush(a) {
            this.cp = a;
            this.g("BackingBrush");
        }
        get backingOutline() {
            return this.cq;
        }
        set backingOutline(a) {
            this.cq = a;
            this.g("BackingOutline");
        }
        get backingStrokeThickness() {
            return this.aa;
        }
        set backingStrokeThickness(a) {
            this.aa = a;
            this.g("BackingStrokeThickness");
        }
        get backingOuterExtent() {
            return this.y;
        }
        set backingOuterExtent(a) {
            this.y = a;
            this.g("BackingOuterExtent");
        }
        get backingOversweep() {
            return this.z;
        }
        set backingOversweep(a) {
            this.z = a;
            this.g("BackingOversweep");
        }
        get scaleOversweep() {
            return this.a3;
        }
        set scaleOversweep(a) {
            this.a3 = a;
            this.g("ScaleOversweep");
        }
        get scaleOversweepShape() {
            return this.c9;
        }
        set scaleOversweepShape(a) {
            this.c9 = a;
            this.g("ScaleOversweepShape");
        }
        get backingCornerRadius() {
            return this.w;
        }
        set backingCornerRadius(a) {
            this.w = a;
            this.g("BackingCornerRadius");
        }
        get backingInnerExtent() {
            return this.x;
        }
        set backingInnerExtent(a) {
            this.x = a;
            this.g("BackingInnerExtent");
        }
        get backingShape() {
            return this.cr;
        }
        set backingShape(a) {
            this.cr = a;
            this.g("BackingShape");
        }
        get ranges() {
            return this.k;
        }
        set ranges(a) {
            this.k = a;
            this.g("Ranges");
        }
        get radiusMultiplier() {
            return this.a0;
        }
        set radiusMultiplier(a) {
            this.a0 = a;
            this.g("RadiusMultiplier");
        }
        get duplicateLabelOmissionStrategy() {
            return this.cs;
        }
        set duplicateLabelOmissionStrategy(a) {
            this.cs = a;
            this.g("DuplicateLabelOmissionStrategy");
        }
        get isNeedleDraggingEnabled() {
            return this.q;
        }
        set isNeedleDraggingEnabled(a) {
            this.q = a;
            this.g("IsNeedleDraggingEnabled");
        }
        get isNeedleDraggingConstrained() {
            return this.p;
        }
        set isNeedleDraggingConstrained(a) {
            this.p = a;
            this.g("IsNeedleDraggingConstrained");
        }
        get font() {
            return this.ct;
        }
        set font(a) {
            this.ct = a;
            this.g("Font");
        }
        get labelFormat() {
            return this.c0;
        }
        set labelFormat(a) {
            this.c0 = a;
            this.g("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.h;
        }
        set labelFormatSpecifiers(a) {
            this.h = a;
            this.g("LabelFormatSpecifiers");
        }
        get transitionProgress() {
            return this.a9;
        }
        set transitionProgress(a) {
            this.a9 = a;
            this.g("TransitionProgress");
        }
        get pixelScalingRatio() {
            return this.az;
        }
        set pixelScalingRatio(a) {
            this.az = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.v;
        }
        set actualPixelScalingRatio(a) {
            this.v = a;
            this.g("ActualPixelScalingRatio");
        }
        get formatLabelRef() {
            return this.cy;
        }
        set formatLabelRef(a) {
            this.cy = a;
            this.g("FormatLabelRef");
        }
        get alignLabelRef() {
            return this.cn;
        }
        set alignLabelRef(a) {
            this.cn = a;
            this.g("AlignLabelRef");
        }
        get valueChangedRef() {
            return this.dd;
        }
        set valueChangedRef(a) {
            this.dd = a;
            this.g("ValueChangedRef");
        }
        get actualMinimumValueChangeRef() {
            return this.cm;
        }
        set actualMinimumValueChangeRef(a) {
            this.cm = a;
            this.g("ActualMinimumValueChangeRef");
        }
        get actualMaximumValueChangeRef() {
            return this.cl;
        }
        set actualMaximumValueChangeRef(a) {
            this.cl = a;
            this.g("ActualMaximumValueChangeRef");
        }
    }
    RadialGaugeDescription.$t = /*@__PURE__*/ markType(RadialGaugeDescription, 'RadialGaugeDescription', Description.$);
    return RadialGaugeDescription;
})();
