// igniteui-react-core/esm2015/lib/RadialBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RadialBaseDescription = /*@__PURE__*/ (() => {
    class RadialBaseDescription extends MarkerSeriesDescription {
        get_type() {
            return "RadialBase";
        }
        constructor() {
            super();
            this.dz = false;
            this.dy = false;
            this.d3 = null;
            this.d6 = null;
            this.dx = false;
            this.d5 = null;
            this.d4 = null;
        }
        get isCustomRadialStyleAllowed() {
            return this.dz;
        }
        set isCustomRadialStyleAllowed(a) {
            this.dz = a;
            this.g("IsCustomRadialStyleAllowed");
        }
        get isCustomRadialMarkerStyleAllowed() {
            return this.dy;
        }
        set isCustomRadialMarkerStyleAllowed(a) {
            this.dy = a;
            this.g("IsCustomRadialMarkerStyleAllowed");
        }
        get angleAxisRef() {
            return this.d3;
        }
        set angleAxisRef(a) {
            this.d3 = a;
            this.g("AngleAxisRef");
        }
        get valueAxisRef() {
            return this.d6;
        }
        set valueAxisRef(a) {
            this.d6 = a;
            this.g("ValueAxisRef");
        }
        get clipSeriesToBounds() {
            return this.dx;
        }
        set clipSeriesToBounds(a) {
            this.dx = a;
            this.g("ClipSeriesToBounds");
        }
        get assigningRadialStyleRef() {
            return this.d5;
        }
        set assigningRadialStyleRef(a) {
            this.d5 = a;
            this.g("AssigningRadialStyleRef");
        }
        get assigningRadialMarkerStyleRef() {
            return this.d4;
        }
        set assigningRadialMarkerStyleRef(a) {
            this.d4 = a;
            this.g("AssigningRadialMarkerStyleRef");
        }
    }
    RadialBaseDescription.$t = /*@__PURE__*/ markType(RadialBaseDescription, 'RadialBaseDescription', MarkerSeriesDescription.$);
    return RadialBaseDescription;
})();
