// igniteui-react-core/esm2015/lib/RangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RangeCategorySeriesDescription = /*@__PURE__*/ (() => {
    class RangeCategorySeriesDescription extends CategorySeriesDescription {
        get_type() {
            return "RangeCategorySeries";
        }
        constructor() {
            super();
            this.ec = null;
            this.eb = null;
        }
        get lowMemberPath() {
            return this.ec;
        }
        set lowMemberPath(a) {
            this.ec = a;
            this.g("LowMemberPath");
        }
        get highMemberPath() {
            return this.eb;
        }
        set highMemberPath(a) {
            this.eb = a;
            this.g("HighMemberPath");
        }
    }
    RangeCategorySeriesDescription.$t = /*@__PURE__*/ markType(RangeCategorySeriesDescription, 'RangeCategorySeriesDescription', CategorySeriesDescription.$);
    return RangeCategorySeriesDescription;
})();
