// igniteui-react-core/esm2015/lib/ScatterTriangulationSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterTriangulationSeriesDescription = /*@__PURE__*/ (() => {
    class ScatterTriangulationSeriesDescription extends SeriesDescription {
        get_type() {
            return "ScatterTriangulationSeries";
        }
        constructor() {
            super();
            this.dk = null;
            this.dp = null;
            this.dh = null;
            this.dl = null;
            this.dc = null;
            this.dd = null;
            this.de = null;
            this.df = null;
            this.di = null;
            this.dm = null;
            this.dj = null;
            this.dn = null;
            this.dg = null;
        }
        get xMemberPath() {
            return this.dk;
        }
        set xMemberPath(a) {
            this.dk = a;
            this.g("XMemberPath");
        }
        get yMemberPath() {
            return this.dp;
        }
        set yMemberPath(a) {
            this.dp = a;
            this.g("YMemberPath");
        }
        get xAxisRef() {
            return this.dh;
        }
        set xAxisRef(a) {
            this.dh = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.dl;
        }
        set yAxisRef(a) {
            this.dl = a;
            this.g("YAxisRef");
        }
        get trianglesSourceRef() {
            return this.dc;
        }
        set trianglesSourceRef(a) {
            this.dc = a;
            this.g("TrianglesSourceRef");
        }
        get triangleVertexMemberPath1() {
            return this.dd;
        }
        set triangleVertexMemberPath1(a) {
            this.dd = a;
            this.g("TriangleVertexMemberPath1");
        }
        get triangleVertexMemberPath2() {
            return this.de;
        }
        set triangleVertexMemberPath2(a) {
            this.de = a;
            this.g("TriangleVertexMemberPath2");
        }
        get triangleVertexMemberPath3() {
            return this.df;
        }
        set triangleVertexMemberPath3(a) {
            this.df = a;
            this.g("TriangleVertexMemberPath3");
        }
        get xMemberAsLegendLabel() {
            return this.di;
        }
        set xMemberAsLegendLabel(a) {
            this.di = a;
            this.g("XMemberAsLegendLabel");
        }
        get yMemberAsLegendLabel() {
            return this.dm;
        }
        set yMemberAsLegendLabel(a) {
            this.dm = a;
            this.g("YMemberAsLegendLabel");
        }
        get xMemberAsLegendUnit() {
            return this.dj;
        }
        set xMemberAsLegendUnit(a) {
            this.dj = a;
            this.g("XMemberAsLegendUnit");
        }
        get yMemberAsLegendUnit() {
            return this.dn;
        }
        set yMemberAsLegendUnit(a) {
            this.dn = a;
            this.g("YMemberAsLegendUnit");
        }
        get triangulationStatusChangedRef() {
            return this.dg;
        }
        set triangulationStatusChangedRef(a) {
            this.dg = a;
            this.g("TriangulationStatusChangedRef");
        }
    }
    ScatterTriangulationSeriesDescription.$t = /*@__PURE__*/ markType(ScatterTriangulationSeriesDescription, 'ScatterTriangulationSeriesDescription', SeriesDescription.$);
    return ScatterTriangulationSeriesDescription;
})();
