// igniteui-react-core/esm2015/lib/ScatterBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterBaseDescription = /*@__PURE__*/ (() => {
    class ScatterBaseDescription extends MarkerSeriesDescription {
        get_type() {
            return "ScatterBase";
        }
        constructor() {
            super();
            this.el = null;
            this.ep = null;
            this.eo = null;
            this.es = null;
            this.em = null;
            this.eq = null;
            this.en = null;
            this.er = null;
            this.ek = null;
            this.ej = null;
            this.ee = null;
            this.d3 = 0;
            this.dx = null;
            this.d7 = 0;
            this.ei = null;
            this.d8 = 0;
            this.d6 = 0;
            this.eh = null;
            this.d5 = 0;
            this.ed = null;
            this.d0 = false;
            this.dz = false;
            this.eg = null;
            this.ef = null;
        }
        get xAxisRef() {
            return this.el;
        }
        set xAxisRef(a) {
            this.el = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.ep;
        }
        set yAxisRef(a) {
            this.ep = a;
            this.g("YAxisRef");
        }
        get xMemberPath() {
            return this.eo;
        }
        set xMemberPath(a) {
            this.eo = a;
            this.g("XMemberPath");
        }
        get yMemberPath() {
            return this.es;
        }
        set yMemberPath(a) {
            this.es = a;
            this.g("YMemberPath");
        }
        get xMemberAsLegendLabel() {
            return this.em;
        }
        set xMemberAsLegendLabel(a) {
            this.em = a;
            this.g("XMemberAsLegendLabel");
        }
        get yMemberAsLegendLabel() {
            return this.eq;
        }
        set yMemberAsLegendLabel(a) {
            this.eq = a;
            this.g("YMemberAsLegendLabel");
        }
        get xMemberAsLegendUnit() {
            return this.en;
        }
        set xMemberAsLegendUnit(a) {
            this.en = a;
            this.g("XMemberAsLegendUnit");
        }
        get yMemberAsLegendUnit() {
            return this.er;
        }
        set yMemberAsLegendUnit(a) {
            this.er = a;
            this.g("YMemberAsLegendUnit");
        }
        get trendLineType() {
            return this.ek;
        }
        set trendLineType(a) {
            this.ek = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.ej;
        }
        set trendLineBrush(a) {
            this.ej = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.ee;
        }
        set actualTrendLineBrush(a) {
            this.ee = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.d3;
        }
        set trendLineThickness(a) {
            this.d3 = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.dx;
        }
        set trendLineDashArray(a) {
            this.dx = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.d7;
        }
        set trendLinePeriod(a) {
            this.d7 = a;
            this.g("TrendLinePeriod");
        }
        get markerCollisionAvoidance() {
            return this.ei;
        }
        set markerCollisionAvoidance(a) {
            this.ei = a;
            this.g("MarkerCollisionAvoidance");
        }
        get trendLineZIndex() {
            return this.d8;
        }
        set trendLineZIndex(a) {
            this.d8 = a;
            this.g("TrendLineZIndex");
        }
        get maximumMarkers() {
            return this.d6;
        }
        set maximumMarkers(a) {
            this.d6 = a;
            this.g("MaximumMarkers");
        }
        get itemSearchMode() {
            return this.eh;
        }
        set itemSearchMode(a) {
            this.eh = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.d5;
        }
        set itemSearchThreshold(a) {
            this.d5 = a;
            this.g("ItemSearchThreshold");
        }
        get actualItemSearchMode() {
            return this.ed;
        }
        set actualItemSearchMode(a) {
            this.ed = a;
            this.g("ActualItemSearchMode");
        }
        get isCustomScatterStyleAllowed() {
            return this.d0;
        }
        set isCustomScatterStyleAllowed(a) {
            this.d0 = a;
            this.g("IsCustomScatterStyleAllowed");
        }
        get isCustomScatterMarkerStyleAllowed() {
            return this.dz;
        }
        set isCustomScatterMarkerStyleAllowed(a) {
            this.dz = a;
            this.g("IsCustomScatterMarkerStyleAllowed");
        }
        get assigningScatterStyleRef() {
            return this.eg;
        }
        set assigningScatterStyleRef(a) {
            this.eg = a;
            this.g("AssigningScatterStyleRef");
        }
        get assigningScatterMarkerStyleRef() {
            return this.ef;
        }
        set assigningScatterMarkerStyleRef(a) {
            this.ef = a;
            this.g("AssigningScatterMarkerStyleRef");
        }
    }
    ScatterBaseDescription.$t = /*@__PURE__*/ markType(ScatterBaseDescription, 'ScatterBaseDescription', MarkerSeriesDescription.$);
    return ScatterBaseDescription;
})();
