// igniteui-react-core/esm2015/lib/SectionHeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SectionHeaderDescription = /*@__PURE__*/ (() => {
    class SectionHeaderDescription extends DefinitionBaseDescription {
        get_type() {
            return "SectionHeader";
        }
        constructor() {
            super();
            this.cf = null;
            this.ce = null;
            this.b7 = 0;
            this.b9 = 0;
            this.b8 = 0;
            this.b6 = 0;
            this.b4 = false;
            this.cg = null;
        }
        get selectedBackground() {
            return this.cf;
        }
        set selectedBackground(a) {
            this.cf = a;
            this.g("SelectedBackground");
        }
        get actualSelectedBackground() {
            return this.ce;
        }
        set actualSelectedBackground(a) {
            this.ce = a;
            this.g("ActualSelectedBackground");
        }
        get paddingLeft() {
            return this.b7;
        }
        set paddingLeft(a) {
            this.b7 = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.b9;
        }
        set paddingTop(a) {
            this.b9 = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.b8;
        }
        set paddingRight(a) {
            this.b8 = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.b6;
        }
        set paddingBottom(a) {
            this.b6 = a;
            this.g("PaddingBottom");
        }
        get isCollapsable() {
            return this.b4;
        }
        set isCollapsable(a) {
            this.b4 = a;
            this.g("IsCollapsable");
        }
        get summaryDisplayMode() {
            return this.cg;
        }
        set summaryDisplayMode(a) {
            this.cg = a;
            this.g("SummaryDisplayMode");
        }
    }
    SectionHeaderDescription.$t = /*@__PURE__*/ markType(SectionHeaderDescription, 'SectionHeaderDescription', DefinitionBaseDescription.$);
    return SectionHeaderDescription;
})();
