// igniteui-react-core/esm2015/lib/SeriesViewerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SeriesViewerDescription = /*@__PURE__*/ (() => {
    class SeriesViewerDescription extends Description {
        get_type() {
            return "SeriesViewer";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.h = null;
            this.k = null;
            this.i = null;
            this.j = null;
            this.d6 = null;
            this.r = null;
            this.d5 = null;
            this.ab = false;
            this.ag = false;
            this.aj = false;
            this.dw = null;
            this.dz = null;
            this.e5 = null;
            this.d1 = null;
            this.e7 = null;
            this.dy = null;
            this.d0 = null;
            this.bc = 0;
            this.bb = 0;
            this.a8 = 0;
            this.a9 = 0;
            this.ba = 0;
            this.be = 0;
            this.bd = 0;
            this.ac = false;
            this.e4 = null;
            this.e6 = null;
            this.bx = 0;
            this.bw = 0;
            this.bu = 0;
            this.b0 = 0;
            this.bv = 0;
            this.bz = 0;
            this.by = 0;
            this.am = false;
            this.dv = null;
            this.d2 = null;
            this.d4 = null;
            this.d3 = null;
            this.u = null;
            this.w = null;
            this.al = false;
            this.ah = false;
            this.dh = 0;
            this.df = 0;
            this.dg = 0;
            this.ae = false;
            this.dq = null;
            this.dl = null;
            this.af = false;
            this.fb = null;
            this.b4 = 0;
            this.b3 = 0;
            this.a4 = 0;
            this.a3 = 0;
            this.b6 = 0;
            this.b5 = 0;
            this.ev = null;
            this.dr = null;
            this.dx = null;
            this.e3 = null;
            this.fc = null;
            this.d8 = null;
            this.ds = null;
            this.eg = null;
            this.dt = null;
            this.d7 = null;
            this.v = null;
            this.b1 = 0;
            this.b2 = 0;
            this.dp = null;
            this.e0 = null;
            this.es = null;
            this.e2 = null;
            this.eu = null;
            this.e1 = null;
            this.et = null;
            this.bs = 0;
            this.bq = 0;
            this.br = 0;
            this.bo = 0;
            this.bn = 0;
            this.bl = 0;
            this.bm = 0;
            this.bj = 0;
            this.eo = null;
            this.bt = 0;
            this.bg = 0;
            this.bi = 0;
            this.a7 = 0;
            this.a6 = 0;
            this.a5 = 0;
            this.ad = false;
            this.bh = 0;
            this.bf = 0;
            this.a0 = 0;
            this.az = 0;
            this.ai = false;
            this.t = null;
            this.a1 = 0;
            this.a2 = 0;
            this.ec = null;
            this.ed = null;
            this.d9 = null;
            this.ea = null;
            this.eb = null;
            this.ee = null;
            this.eh = null;
            this.ek = null;
            this.el = null;
            this.em = null;
            this.ei = null;
            this.ej = null;
            this.e9 = null;
            this.e8 = null;
            this.fa = null;
            this.en = null;
            this.dm = null;
            this.du = null;
            this.ef = null;
            this.dn = null;
            this.ex = null;
            this.bp = 0;
            this.ez = null;
            this.ey = null;
            this.ep = null;
            this.bk = 0;
            this.er = null;
            this.eq = null;
            this.ak = false;
            this.ew = null;
        }
        get series() {
            return this.l;
        }
        set series(a) {
            this.l = a;
            this.g("Series");
        }
        get brushes() {
            return this.h;
        }
        set brushes(a) {
            this.h = a;
            this.g("Brushes");
        }
        get outlines() {
            return this.k;
        }
        set outlines(a) {
            this.k = a;
            this.g("Outlines");
        }
        get markerBrushes() {
            return this.i;
        }
        set markerBrushes(a) {
            this.i = a;
            this.g("MarkerBrushes");
        }
        get markerOutlines() {
            return this.j;
        }
        set markerOutlines(a) {
            this.j = a;
            this.g("MarkerOutlines");
        }
        get markerAutomaticBehavior() {
            return this.d6;
        }
        set markerAutomaticBehavior(a) {
            this.d6 = a;
            this.g("MarkerAutomaticBehavior");
        }
        get crosshairPoint() {
            return this.r;
        }
        set crosshairPoint(a) {
            this.r = a;
            this.g("CrosshairPoint");
        }
        get legendRef() {
            return this.d5;
        }
        set legendRef(a) {
            this.d5 = a;
            this.g("LegendRef");
        }
        get animateSeriesWhenAxisRangeChanges() {
            return this.ab;
        }
        set animateSeriesWhenAxisRangeChanges(a) {
            this.ab = a;
            this.g("AnimateSeriesWhenAxisRangeChanges");
        }
        get isWindowSyncedToVisibleRange() {
            return this.ag;
        }
        set isWindowSyncedToVisibleRange(a) {
            this.ag = a;
            this.g("IsWindowSyncedToVisibleRange");
        }
        get shouldSimulateHoverMoveCrosshairPoint() {
            return this.aj;
        }
        set shouldSimulateHoverMoveCrosshairPoint(a) {
            this.aj = a;
            this.g("ShouldSimulateHoverMoveCrosshairPoint");
        }
        get highlightingMode() {
            return this.dw;
        }
        set highlightingMode(a) {
            this.dw = a;
            this.g("HighlightingMode");
        }
        get horizontalViewScrollbarMode() {
            return this.dz;
        }
        set horizontalViewScrollbarMode(a) {
            this.dz = a;
            this.g("HorizontalViewScrollbarMode");
        }
        get verticalViewScrollbarMode() {
            return this.e5;
        }
        set verticalViewScrollbarMode(a) {
            this.e5 = a;
            this.g("VerticalViewScrollbarMode");
        }
        get horizontalViewScrollbarPosition() {
            return this.d1;
        }
        set horizontalViewScrollbarPosition(a) {
            this.d1 = a;
            this.g("HorizontalViewScrollbarPosition");
        }
        get verticalViewScrollbarPosition() {
            return this.e7;
        }
        set verticalViewScrollbarPosition(a) {
            this.e7 = a;
            this.g("VerticalViewScrollbarPosition");
        }
        get horizontalViewScrollbarFill() {
            return this.dy;
        }
        set horizontalViewScrollbarFill(a) {
            this.dy = a;
            this.g("HorizontalViewScrollbarFill");
        }
        get horizontalViewScrollbarOutline() {
            return this.d0;
        }
        set horizontalViewScrollbarOutline(a) {
            this.d0 = a;
            this.g("HorizontalViewScrollbarOutline");
        }
        get horizontalViewScrollbarStrokeThickness() {
            return this.bc;
        }
        set horizontalViewScrollbarStrokeThickness(a) {
            this.bc = a;
            this.g("HorizontalViewScrollbarStrokeThickness");
        }
        get horizontalViewScrollbarMaxOpacity() {
            return this.bb;
        }
        set horizontalViewScrollbarMaxOpacity(a) {
            this.bb = a;
            this.g("HorizontalViewScrollbarMaxOpacity");
        }
        get horizontalViewScrollbarCornerRadius() {
            return this.a8;
        }
        set horizontalViewScrollbarCornerRadius(a) {
            this.a8 = a;
            this.g("HorizontalViewScrollbarCornerRadius");
        }
        get horizontalViewScrollbarHeight() {
            return this.a9;
        }
        set horizontalViewScrollbarHeight(a) {
            this.a9 = a;
            this.g("HorizontalViewScrollbarHeight");
        }
        get horizontalViewScrollbarInset() {
            return this.ba;
        }
        set horizontalViewScrollbarInset(a) {
            this.ba = a;
            this.g("HorizontalViewScrollbarInset");
        }
        get horizontalViewScrollbarTrackStartInset() {
            return this.be;
        }
        set horizontalViewScrollbarTrackStartInset(a) {
            this.be = a;
            this.g("HorizontalViewScrollbarTrackStartInset");
        }
        get horizontalViewScrollbarTrackEndInset() {
            return this.bd;
        }
        set horizontalViewScrollbarTrackEndInset(a) {
            this.bd = a;
            this.g("HorizontalViewScrollbarTrackEndInset");
        }
        get horizontalViewScrollbarShouldAddAutoTrackInsets() {
            return this.ac;
        }
        set horizontalViewScrollbarShouldAddAutoTrackInsets(a) {
            this.ac = a;
            this.g("HorizontalViewScrollbarShouldAddAutoTrackInsets");
        }
        get verticalViewScrollbarFill() {
            return this.e4;
        }
        set verticalViewScrollbarFill(a) {
            this.e4 = a;
            this.g("VerticalViewScrollbarFill");
        }
        get verticalViewScrollbarOutline() {
            return this.e6;
        }
        set verticalViewScrollbarOutline(a) {
            this.e6 = a;
            this.g("VerticalViewScrollbarOutline");
        }
        get verticalViewScrollbarStrokeThickness() {
            return this.bx;
        }
        set verticalViewScrollbarStrokeThickness(a) {
            this.bx = a;
            this.g("VerticalViewScrollbarStrokeThickness");
        }
        get verticalViewScrollbarMaxOpacity() {
            return this.bw;
        }
        set verticalViewScrollbarMaxOpacity(a) {
            this.bw = a;
            this.g("VerticalViewScrollbarMaxOpacity");
        }
        get verticalViewScrollbarCornerRadius() {
            return this.bu;
        }
        set verticalViewScrollbarCornerRadius(a) {
            this.bu = a;
            this.g("VerticalViewScrollbarCornerRadius");
        }
        get verticalViewScrollbarWidth() {
            return this.b0;
        }
        set verticalViewScrollbarWidth(a) {
            this.b0 = a;
            this.g("VerticalViewScrollbarWidth");
        }
        get verticalViewScrollbarInset() {
            return this.bv;
        }
        set verticalViewScrollbarInset(a) {
            this.bv = a;
            this.g("VerticalViewScrollbarInset");
        }
        get verticalViewScrollbarTrackStartInset() {
            return this.bz;
        }
        set verticalViewScrollbarTrackStartInset(a) {
            this.bz = a;
            this.g("VerticalViewScrollbarTrackStartInset");
        }
        get verticalViewScrollbarTrackEndInset() {
            return this.by;
        }
        set verticalViewScrollbarTrackEndInset(a) {
            this.by = a;
            this.g("VerticalViewScrollbarTrackEndInset");
        }
        get verticalViewScrollbarShouldAddAutoTrackInsets() {
            return this.am;
        }
        set verticalViewScrollbarShouldAddAutoTrackInsets(a) {
            this.am = a;
            this.g("VerticalViewScrollbarShouldAddAutoTrackInsets");
        }
        get highlightingBehavior() {
            return this.dv;
        }
        set highlightingBehavior(a) {
            this.dv = a;
            this.g("HighlightingBehavior");
        }
        get legendHighlightingMode() {
            return this.d2;
        }
        set legendHighlightingMode(a) {
            this.d2 = a;
            this.g("LegendHighlightingMode");
        }
        get legendItemBadgeShape() {
            return this.d4;
        }
        set legendItemBadgeShape(a) {
            this.d4 = a;
            this.g("LegendItemBadgeShape");
        }
        get legendItemBadgeMode() {
            return this.d3;
        }
        set legendItemBadgeMode(a) {
            this.d3 = a;
            this.g("LegendItemBadgeMode");
        }
        get effectiveViewport() {
            return this.u;
        }
        set effectiveViewport(a) {
            this.u = a;
            this.g("EffectiveViewport");
        }
        get windowRect() {
            return this.w;
        }
        set windowRect(a) {
            this.w = a;
            this.g("WindowRect");
        }
        get useTiledZooming() {
            return this.al;
        }
        set useTiledZooming(a) {
            this.al = a;
            this.g("UseTiledZooming");
        }
        get preferHigherResolutionTiles() {
            return this.ah;
        }
        set preferHigherResolutionTiles(a) {
            this.ah = a;
            this.g("PreferHigherResolutionTiles");
        }
        get zoomTileCacheSize() {
            return this.dh;
        }
        set zoomTileCacheSize(a) {
            this.dh = a;
            this.g("ZoomTileCacheSize");
        }
        get highlightingTransitionDuration() {
            return this.df;
        }
        set highlightingTransitionDuration(a) {
            this.df = a;
            this.g("HighlightingTransitionDuration");
        }
        get scrollbarsAnimationDuration() {
            return this.dg;
        }
        set scrollbarsAnimationDuration(a) {
            this.dg = a;
            this.g("ScrollbarsAnimationDuration");
        }
        get isPagePanningAllowed() {
            return this.ae;
        }
        set isPagePanningAllowed(a) {
            this.ae = a;
            this.g("IsPagePanningAllowed");
        }
        get contentHitTestMode() {
            return this.dq;
        }
        set contentHitTestMode(a) {
            this.dq = a;
            this.g("ContentHitTestMode");
        }
        get actualContentHitTestMode() {
            return this.dl;
        }
        set actualContentHitTestMode(a) {
            this.dl = a;
            this.g("ActualContentHitTestMode");
        }
        get isSurfaceInteractionDisabled() {
            return this.af;
        }
        set isSurfaceInteractionDisabled(a) {
            this.af = a;
            this.g("IsSurfaceInteractionDisabled");
        }
        get windowResponse() {
            return this.fb;
        }
        set windowResponse(a) {
            this.fb = a;
            this.g("WindowResponse");
        }
        get windowRectMinWidth() {
            return this.b4;
        }
        set windowRectMinWidth(a) {
            this.b4 = a;
            this.g("WindowRectMinWidth");
        }
        get windowRectMinHeight() {
            return this.b3;
        }
        set windowRectMinHeight(a) {
            this.b3 = a;
            this.g("WindowRectMinHeight");
        }
        get actualWindowRectMinWidth() {
            return this.a4;
        }
        set actualWindowRectMinWidth(a) {
            this.a4 = a;
            this.g("ActualWindowRectMinWidth");
        }
        get actualWindowRectMinHeight() {
            return this.a3;
        }
        set actualWindowRectMinHeight(a) {
            this.a3 = a;
            this.g("ActualWindowRectMinHeight");
        }
        get windowSizeMinWidth() {
            return this.b6;
        }
        set windowSizeMinWidth(a) {
            this.b6 = a;
            this.g("WindowSizeMinWidth");
        }
        get windowSizeMinHeight() {
            return this.b5;
        }
        set windowSizeMinHeight(a) {
            this.b5 = a;
            this.g("WindowSizeMinHeight");
        }
        get syncChannel() {
            return this.ev;
        }
        set syncChannel(a) {
            this.ev = a;
            this.g("SyncChannel");
        }
        get crosshairVisibility() {
            return this.dr;
        }
        set crosshairVisibility(a) {
            this.dr = a;
            this.g("CrosshairVisibility");
        }
        get horizontalCrosshairBrush() {
            return this.dx;
        }
        set horizontalCrosshairBrush(a) {
            this.dx = a;
            this.g("HorizontalCrosshairBrush");
        }
        get verticalCrosshairBrush() {
            return this.e3;
        }
        set verticalCrosshairBrush(a) {
            this.e3 = a;
            this.g("VerticalCrosshairBrush");
        }
        get zoomCoercionMode() {
            return this.fc;
        }
        set zoomCoercionMode(a) {
            this.fc = a;
            this.g("ZoomCoercionMode");
        }
        get plotAreaBackground() {
            return this.d8;
        }
        set plotAreaBackground(a) {
            this.d8 = a;
            this.g("PlotAreaBackground");
        }
        get defaultInteraction() {
            return this.ds;
        }
        set defaultInteraction(a) {
            this.ds = a;
            this.g("DefaultInteraction");
        }
        get rightButtonDefaultInteraction() {
            return this.eg;
        }
        set rightButtonDefaultInteraction(a) {
            this.eg = a;
            this.g("RightButtonDefaultInteraction");
        }
        get dragModifier() {
            return this.dt;
        }
        set dragModifier(a) {
            this.dt = a;
            this.g("DragModifier");
        }
        get panModifier() {
            return this.d7;
        }
        set panModifier(a) {
            this.d7 = a;
            this.g("PanModifier");
        }
        get previewRect() {
            return this.v;
        }
        set previewRect(a) {
            this.v = a;
            this.g("PreviewRect");
        }
        get windowPositionHorizontal() {
            return this.b1;
        }
        set windowPositionHorizontal(a) {
            this.b1 = a;
            this.g("WindowPositionHorizontal");
        }
        get windowPositionVertical() {
            return this.b2;
        }
        set windowPositionVertical(a) {
            this.b2 = a;
            this.g("WindowPositionVertical");
        }
        get chartTitle() {
            return this.dp;
        }
        set chartTitle(a) {
            this.dp = a;
            this.g("ChartTitle");
        }
        get titleHorizontalAlignment() {
            return this.e0;
        }
        set titleHorizontalAlignment(a) {
            this.e0 = a;
            this.g("TitleHorizontalAlignment");
        }
        get subtitleHorizontalAlignment() {
            return this.es;
        }
        set subtitleHorizontalAlignment(a) {
            this.es = a;
            this.g("SubtitleHorizontalAlignment");
        }
        get titleTextStyle() {
            return this.e2;
        }
        set titleTextStyle(a) {
            this.e2 = a;
            this.g("TitleTextStyle");
        }
        get subtitleTextStyle() {
            return this.eu;
        }
        set subtitleTextStyle(a) {
            this.eu = a;
            this.g("SubtitleTextStyle");
        }
        get titleTextColor() {
            return this.e1;
        }
        set titleTextColor(a) {
            this.e1 = a;
            this.g("TitleTextColor");
        }
        get subtitleTextColor() {
            return this.et;
        }
        set subtitleTextColor(a) {
            this.et = a;
            this.g("SubtitleTextColor");
        }
        get titleTopMargin() {
            return this.bs;
        }
        set titleTopMargin(a) {
            this.bs = a;
            this.g("TitleTopMargin");
        }
        get titleLeftMargin() {
            return this.bq;
        }
        set titleLeftMargin(a) {
            this.bq = a;
            this.g("TitleLeftMargin");
        }
        get titleRightMargin() {
            return this.br;
        }
        set titleRightMargin(a) {
            this.br = a;
            this.g("TitleRightMargin");
        }
        get titleBottomMargin() {
            return this.bo;
        }
        set titleBottomMargin(a) {
            this.bo = a;
            this.g("TitleBottomMargin");
        }
        get subtitleTopMargin() {
            return this.bn;
        }
        set subtitleTopMargin(a) {
            this.bn = a;
            this.g("SubtitleTopMargin");
        }
        get subtitleLeftMargin() {
            return this.bl;
        }
        set subtitleLeftMargin(a) {
            this.bl = a;
            this.g("SubtitleLeftMargin");
        }
        get subtitleRightMargin() {
            return this.bm;
        }
        set subtitleRightMargin(a) {
            this.bm = a;
            this.g("SubtitleRightMargin");
        }
        get subtitleBottomMargin() {
            return this.bj;
        }
        set subtitleBottomMargin(a) {
            this.bj = a;
            this.g("SubtitleBottomMargin");
        }
        get subtitle() {
            return this.eo;
        }
        set subtitle(a) {
            this.eo = a;
            this.g("Subtitle");
        }
        get topMargin() {
            return this.bt;
        }
        set topMargin(a) {
            this.bt = a;
            this.g("TopMargin");
        }
        get leftMargin() {
            return this.bg;
        }
        set leftMargin(a) {
            this.bg = a;
            this.g("LeftMargin");
        }
        get rightMargin() {
            return this.bi;
        }
        set rightMargin(a) {
            this.bi = a;
            this.g("RightMargin");
        }
        get bottomMargin() {
            return this.a7;
        }
        set bottomMargin(a) {
            this.a7 = a;
            this.g("BottomMargin");
        }
        get autoMarginWidth() {
            return this.a6;
        }
        set autoMarginWidth(a) {
            this.a6 = a;
            this.g("AutoMarginWidth");
        }
        get autoMarginHeight() {
            return this.a5;
        }
        set autoMarginHeight(a) {
            this.a5 = a;
            this.g("AutoMarginHeight");
        }
        get isAntiAliasingEnabledDuringInteraction() {
            return this.ad;
        }
        set isAntiAliasingEnabledDuringInteraction(a) {
            this.ad = a;
            this.g("IsAntiAliasingEnabledDuringInteraction");
        }
        get pixelScalingRatio() {
            return this.bh;
        }
        set pixelScalingRatio(a) {
            this.bh = a;
            this.g("PixelScalingRatio");
        }
        get interactionPixelScalingRatio() {
            return this.bf;
        }
        set interactionPixelScalingRatio(a) {
            this.bf = a;
            this.g("InteractionPixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.a0;
        }
        set actualPixelScalingRatio(a) {
            this.a0 = a;
            this.g("ActualPixelScalingRatio");
        }
        get actualInteractionPixelScalingRatio() {
            return this.az;
        }
        set actualInteractionPixelScalingRatio(a) {
            this.az = a;
            this.g("ActualInteractionPixelScalingRatio");
        }
        get shouldPanOnMaximumZoom() {
            return this.ai;
        }
        set shouldPanOnMaximumZoom(a) {
            this.ai = a;
            this.g("ShouldPanOnMaximumZoom");
        }
        get actualWindowRect() {
            return this.t;
        }
        set actualWindowRect(a) {
            this.t = a;
            this.g("ActualWindowRect");
        }
        get actualWindowPositionHorizontal() {
            return this.a1;
        }
        set actualWindowPositionHorizontal(a) {
            this.a1 = a;
            this.g("ActualWindowPositionHorizontal");
        }
        get actualWindowPositionVertical() {
            return this.a2;
        }
        set actualWindowPositionVertical(a) {
            this.a2 = a;
            this.g("ActualWindowPositionVertical");
        }
        get plotAreaMouseLeftButtonDownRef() {
            return this.ec;
        }
        set plotAreaMouseLeftButtonDownRef(a) {
            this.ec = a;
            this.g("PlotAreaMouseLeftButtonDownRef");
        }
        get plotAreaMouseLeftButtonUpRef() {
            return this.ed;
        }
        set plotAreaMouseLeftButtonUpRef(a) {
            this.ed = a;
            this.g("PlotAreaMouseLeftButtonUpRef");
        }
        get plotAreaClickedRef() {
            return this.d9;
        }
        set plotAreaClickedRef(a) {
            this.d9 = a;
            this.g("PlotAreaClickedRef");
        }
        get plotAreaMouseEnterRef() {
            return this.ea;
        }
        set plotAreaMouseEnterRef(a) {
            this.ea = a;
            this.g("PlotAreaMouseEnterRef");
        }
        get plotAreaMouseLeaveRef() {
            return this.eb;
        }
        set plotAreaMouseLeaveRef(a) {
            this.eb = a;
            this.g("PlotAreaMouseLeaveRef");
        }
        get plotAreaMouseOverRef() {
            return this.ee;
        }
        set plotAreaMouseOverRef(a) {
            this.ee = a;
            this.g("PlotAreaMouseOverRef");
        }
        get seriesCursorMouseMoveRef() {
            return this.eh;
        }
        set seriesCursorMouseMoveRef(a) {
            this.eh = a;
            this.g("SeriesCursorMouseMoveRef");
        }
        get seriesMouseLeftButtonDownRef() {
            return this.ek;
        }
        set seriesMouseLeftButtonDownRef(a) {
            this.ek = a;
            this.g("SeriesMouseLeftButtonDownRef");
        }
        get seriesMouseLeftButtonUpRef() {
            return this.el;
        }
        set seriesMouseLeftButtonUpRef(a) {
            this.el = a;
            this.g("SeriesMouseLeftButtonUpRef");
        }
        get seriesMouseMoveRef() {
            return this.em;
        }
        set seriesMouseMoveRef(a) {
            this.em = a;
            this.g("SeriesMouseMoveRef");
        }
        get seriesMouseEnterRef() {
            return this.ei;
        }
        set seriesMouseEnterRef(a) {
            this.ei = a;
            this.g("SeriesMouseEnterRef");
        }
        get seriesMouseLeaveRef() {
            return this.ej;
        }
        set seriesMouseLeaveRef(a) {
            this.ej = a;
            this.g("SeriesMouseLeaveRef");
        }
        get viewerManipulationStartingRef() {
            return this.e9;
        }
        set viewerManipulationStartingRef(a) {
            this.e9 = a;
            this.g("ViewerManipulationStartingRef");
        }
        get viewerManipulationEndingRef() {
            return this.e8;
        }
        set viewerManipulationEndingRef(a) {
            this.e8 = a;
            this.g("ViewerManipulationEndingRef");
        }
        get windowRectChangedRef() {
            return this.fa;
        }
        set windowRectChangedRef(a) {
            this.fa = a;
            this.g("WindowRectChangedRef");
        }
        get sizeChangedRef() {
            return this.en;
        }
        set sizeChangedRef(a) {
            this.en = a;
            this.g("SizeChangedRef");
        }
        get actualWindowRectChangedRef() {
            return this.dm;
        }
        set actualWindowRectChangedRef(a) {
            this.dm = a;
            this.g("ActualWindowRectChangedRef");
        }
        get gridAreaRectChangedRef() {
            return this.du;
        }
        set gridAreaRectChangedRef(a) {
            this.du = a;
            this.g("GridAreaRectChangedRef");
        }
        get refreshCompletedRef() {
            return this.ef;
        }
        set refreshCompletedRef(a) {
            this.ef = a;
            this.g("RefreshCompletedRef");
        }
        get background() {
            return this.dn;
        }
        set background(a) {
            this.dn = a;
            this.g("Background");
        }
        get titleFontFamily() {
            return this.ex;
        }
        set titleFontFamily(a) {
            this.ex = a;
            this.g("TitleFontFamily");
        }
        get titleFontSize() {
            return this.bp;
        }
        set titleFontSize(a) {
            this.bp = a;
            this.g("TitleFontSize");
        }
        get titleFontWeight() {
            return this.ez;
        }
        set titleFontWeight(a) {
            this.ez = a;
            this.g("TitleFontWeight");
        }
        get titleFontStyle() {
            return this.ey;
        }
        set titleFontStyle(a) {
            this.ey = a;
            this.g("TitleFontStyle");
        }
        get subtitleFontFamily() {
            return this.ep;
        }
        set subtitleFontFamily(a) {
            this.ep = a;
            this.g("SubtitleFontFamily");
        }
        get subtitleFontSize() {
            return this.bk;
        }
        set subtitleFontSize(a) {
            this.bk = a;
            this.g("SubtitleFontSize");
        }
        get subtitleFontWeight() {
            return this.er;
        }
        set subtitleFontWeight(a) {
            this.er = a;
            this.g("SubtitleFontWeight");
        }
        get subtitleFontStyle() {
            return this.eq;
        }
        set subtitleFontStyle(a) {
            this.eq = a;
            this.g("SubtitleFontStyle");
        }
        get synchronizeHorizontally() {
            return this.ak;
        }
        set synchronizeHorizontally(a) {
            this.ak = a;
            this.g("SynchronizeHorizontally");
        }
        get synchronizeVertically() {
            return this.ew;
        }
        set synchronizeVertically(a) {
            this.ew = a;
            this.g("SynchronizeVertically");
        }
    }
    SeriesViewerDescription.$t = /*@__PURE__*/ markType(SeriesViewerDescription, 'SeriesViewerDescription', Description.$);
    return SeriesViewerDescription;
})();
