// igniteui-react-core/esm2015/lib/SplineSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeriesDescription } from "./HorizontalAnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SplineSeriesBaseDescription = /*@__PURE__*/ (() => {
    class SplineSeriesBaseDescription extends HorizontalAnchoredCategorySeriesDescription {
        get_type() {
            return "SplineSeriesBase";
        }
        constructor() {
            super();
            this.e1 = false;
            this.e3 = null;
        }
        get isSplineShapePartOfRange() {
            return this.e1;
        }
        set isSplineShapePartOfRange(a) {
            this.e1 = a;
            this.g("IsSplineShapePartOfRange");
        }
        get splineType() {
            return this.e3;
        }
        set splineType(a) {
            this.e3 = a;
            this.g("SplineType");
        }
    }
    SplineSeriesBaseDescription.$t = /*@__PURE__*/ markType(SplineSeriesBaseDescription, 'SplineSeriesBaseDescription', HorizontalAnchoredCategorySeriesDescription.$);
    return SplineSeriesBaseDescription;
})();
