// igniteui-react-core/esm2015/lib/StackedSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let StackedSeriesBaseDescription = /*@__PURE__*/ (() => {
    class StackedSeriesBaseDescription extends CategorySeriesDescription {
        get_type() {
            return "StackedSeriesBase";
        }
        constructor() {
            super();
            this.eb = null;
            this.ed = false;
            this.ee = false;
            this.eh = null;
        }
        get series() {
            return this.eb;
        }
        set series(a) {
            this.eb = a;
            this.g("Series");
        }
        get autoGenerateSeries() {
            return this.ed;
        }
        set autoGenerateSeries(a) {
            this.ed = a;
            this.g("AutoGenerateSeries");
        }
        get reverseLegendOrder() {
            return this.ee;
        }
        set reverseLegendOrder(a) {
            this.ee = a;
            this.g("ReverseLegendOrder");
        }
        get seriesCreatedRef() {
            return this.eh;
        }
        set seriesCreatedRef(a) {
            this.eh = a;
            this.g("SeriesCreatedRef");
        }
    }
    StackedSeriesBaseDescription.$t = /*@__PURE__*/ markType(StackedSeriesBaseDescription, 'StackedSeriesBaseDescription', CategorySeriesDescription.$);
    return StackedSeriesBaseDescription;
})();
