// igniteui-react-core/esm2015/lib/StackedFragmentSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let StackedFragmentSeriesDescription = /*@__PURE__*/ (() => {
    class StackedFragmentSeriesDescription extends Description {
        get_type() {
            return "StackedFragmentSeries";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.cf = null;
            this.ce = null;
            this.br = null;
            this.h = null;
            this.n = false;
            this.j = false;
            this.ar = 0;
            this.ag = 0;
            this.cx = null;
            this.b6 = null;
            this.q = false;
            this.m = false;
            this.as = 0;
            this.ah = 0;
            this.at = 0;
            this.ai = 0;
            this.cl = null;
            this.bx = null;
            this.ci = null;
            this.bu = null;
            this.cj = null;
            this.bv = null;
            this.ck = null;
            this.bw = null;
            this.ch = null;
            this.cg = null;
            this.bt = null;
            this.bs = null;
            this.cm = null;
            this.by = null;
            this.co = null;
            this.b0 = null;
            this.cq = null;
            this.b2 = null;
            this.cr = null;
            this.b3 = null;
            this.cn = null;
            this.bz = null;
            this.cp = null;
            this.b1 = null;
            this.cs = null;
            this.ao = 0;
            this.ad = 0;
            this.ct = null;
            this.b4 = null;
            this.ak = 0;
            this.z = 0;
            this.am = 0;
            this.ab = 0;
            this.cu = null;
            this.b5 = null;
            this.al = 0;
            this.aa = 0;
            this.ap = 0;
            this.ae = 0;
            this.k = false;
            this.o = false;
            this.aq = 0;
            this.af = 0;
            this.au = 0;
            this.aj = 0;
            this.an = 0;
            this.ac = 0;
            this.bp = null;
            this.cy = null;
            this.bj = 0;
            this.bk = 0;
            this.cz = null;
            this.c0 = null;
            this.p = false;
            this.c1 = null;
            this.l = false;
            this.bi = 0;
            this.b9 = null;
            this.ca = null;
            this.b8 = null;
            this.b7 = null;
            this.bh = 0;
            this.c4 = null;
            this.c2 = null;
            this.c3 = null;
            this.cb = null;
            this.cc = null;
            this.c5 = null;
            this.cd = null;
            this.cv = null;
            this.cw = null;
        }
        get dataSourceRef() {
            return this.cf;
        }
        set dataSourceRef(a) {
            this.cf = a;
            this.g("DataSourceRef");
        }
        get brush() {
            return this.ce;
        }
        set brush(a) {
            this.ce = a;
            this.g("Brush");
        }
        get actualBrush() {
            return this.br;
        }
        set actualBrush(a) {
            this.br = a;
            this.g("ActualBrush");
        }
        get dashArray() {
            return this.h;
        }
        set dashArray(a) {
            this.h = a;
            this.g("DashArray");
        }
        get isDropShadowEnabled() {
            return this.n;
        }
        set isDropShadowEnabled(a) {
            this.n = a;
            this.g("IsDropShadowEnabled");
        }
        get actualIsDropShadowEnabled() {
            return this.j;
        }
        set actualIsDropShadowEnabled(a) {
            this.j = a;
            this.g("ActualIsDropShadowEnabled");
        }
        get shadowBlur() {
            return this.ar;
        }
        set shadowBlur(a) {
            this.ar = a;
            this.g("ShadowBlur");
        }
        get actualShadowBlur() {
            return this.ag;
        }
        set actualShadowBlur(a) {
            this.ag = a;
            this.g("ActualShadowBlur");
        }
        get shadowColor() {
            return this.cx;
        }
        set shadowColor(a) {
            this.cx = a;
            this.g("ShadowColor");
        }
        get actualShadowColor() {
            return this.b6;
        }
        set actualShadowColor(a) {
            this.b6 = a;
            this.g("ActualShadowColor");
        }
        get useSingleShadow() {
            return this.q;
        }
        set useSingleShadow(a) {
            this.q = a;
            this.g("UseSingleShadow");
        }
        get actualUseSingleShadow() {
            return this.m;
        }
        set actualUseSingleShadow(a) {
            this.m = a;
            this.g("ActualUseSingleShadow");
        }
        get shadowOffsetX() {
            return this.as;
        }
        set shadowOffsetX(a) {
            this.as = a;
            this.g("ShadowOffsetX");
        }
        get actualShadowOffsetX() {
            return this.ah;
        }
        set actualShadowOffsetX(a) {
            this.ah = a;
            this.g("ActualShadowOffsetX");
        }
        get shadowOffsetY() {
            return this.at;
        }
        set shadowOffsetY(a) {
            this.at = a;
            this.g("ShadowOffsetY");
        }
        get actualShadowOffsetY() {
            return this.ai;
        }
        set actualShadowOffsetY(a) {
            this.ai = a;
            this.g("ActualShadowOffsetY");
        }
        get lineCap() {
            return this.cl;
        }
        set lineCap(a) {
            this.cl = a;
            this.g("LineCap");
        }
        get actualLineCap() {
            return this.bx;
        }
        set actualLineCap(a) {
            this.bx = a;
            this.g("ActualLineCap");
        }
        get legendItemBadgeTemplateRef() {
            return this.ci;
        }
        set legendItemBadgeTemplateRef(a) {
            this.ci = a;
            this.g("LegendItemBadgeTemplateRef");
        }
        get actualLegendItemBadgeTemplateRef() {
            return this.bu;
        }
        set actualLegendItemBadgeTemplateRef(a) {
            this.bu = a;
            this.g("ActualLegendItemBadgeTemplateRef");
        }
        get legendItemTemplateRef() {
            return this.cj;
        }
        set legendItemTemplateRef(a) {
            this.cj = a;
            this.g("LegendItemTemplateRef");
        }
        get actualLegendItemTemplateRef() {
            return this.bv;
        }
        set actualLegendItemTemplateRef(a) {
            this.bv = a;
            this.g("ActualLegendItemTemplateRef");
        }
        get legendItemVisibility() {
            return this.ck;
        }
        set legendItemVisibility(a) {
            this.ck = a;
            this.g("LegendItemVisibility");
        }
        get actualLegendItemVisibility() {
            return this.bw;
        }
        set actualLegendItemVisibility(a) {
            this.bw = a;
            this.g("ActualLegendItemVisibility");
        }
        get legendItemBadgeShape() {
            return this.ch;
        }
        set legendItemBadgeShape(a) {
            this.ch = a;
            this.g("LegendItemBadgeShape");
        }
        get legendItemBadgeMode() {
            return this.cg;
        }
        set legendItemBadgeMode(a) {
            this.cg = a;
            this.g("LegendItemBadgeMode");
        }
        get actualLegendItemBadgeShape() {
            return this.bt;
        }
        set actualLegendItemBadgeShape(a) {
            this.bt = a;
            this.g("ActualLegendItemBadgeShape");
        }
        get actualLegendItemBadgeMode() {
            return this.bs;
        }
        set actualLegendItemBadgeMode(a) {
            this.bs = a;
            this.g("ActualLegendItemBadgeMode");
        }
        get markerBrush() {
            return this.cm;
        }
        set markerBrush(a) {
            this.cm = a;
            this.g("MarkerBrush");
        }
        get actualMarkerBrush() {
            return this.by;
        }
        set actualMarkerBrush(a) {
            this.by = a;
            this.g("ActualMarkerBrush");
        }
        get markerOutline() {
            return this.co;
        }
        set markerOutline(a) {
            this.co = a;
            this.g("MarkerOutline");
        }
        get actualMarkerOutline() {
            return this.b0;
        }
        set actualMarkerOutline(a) {
            this.b0 = a;
            this.g("ActualMarkerOutline");
        }
        get markerTemplateRef() {
            return this.cq;
        }
        set markerTemplateRef(a) {
            this.cq = a;
            this.g("MarkerTemplateRef");
        }
        get actualMarkerTemplateRef() {
            return this.b2;
        }
        set actualMarkerTemplateRef(a) {
            this.b2 = a;
            this.g("ActualMarkerTemplateRef");
        }
        get markerType() {
            return this.cr;
        }
        set markerType(a) {
            this.cr = a;
            this.g("MarkerType");
        }
        get actualMarkerType() {
            return this.b3;
        }
        set actualMarkerType(a) {
            this.b3 = a;
            this.g("ActualMarkerType");
        }
        get markerFillMode() {
            return this.cn;
        }
        set markerFillMode(a) {
            this.cn = a;
            this.g("MarkerFillMode");
        }
        get actualMarkerFillMode() {
            return this.bz;
        }
        set actualMarkerFillMode(a) {
            this.bz = a;
            this.g("ActualMarkerFillMode");
        }
        get markerOutlineMode() {
            return this.cp;
        }
        set markerOutlineMode(a) {
            this.cp = a;
            this.g("MarkerOutlineMode");
        }
        get actualMarkerOutlineMode() {
            return this.b1;
        }
        set actualMarkerOutlineMode(a) {
            this.b1 = a;
            this.g("ActualMarkerOutlineMode");
        }
        get name() {
            return this.cs;
        }
        set name(a) {
            this.cs = a;
            this.g("Name");
        }
        get opacity() {
            return this.ao;
        }
        set opacity(a) {
            this.ao = a;
            this.g("Opacity");
        }
        get actualOpacity() {
            return this.ad;
        }
        set actualOpacity(a) {
            this.ad = a;
            this.g("ActualOpacity");
        }
        get outline() {
            return this.ct;
        }
        set outline(a) {
            this.ct = a;
            this.g("Outline");
        }
        get actualOutline() {
            return this.b4;
        }
        set actualOutline(a) {
            this.b4 = a;
            this.g("ActualOutline");
        }
        get areaFillOpacity() {
            return this.ak;
        }
        set areaFillOpacity(a) {
            this.ak = a;
            this.g("AreaFillOpacity");
        }
        get actualAreaFillOpacity() {
            return this.z;
        }
        set actualAreaFillOpacity(a) {
            this.z = a;
            this.g("ActualAreaFillOpacity");
        }
        get markerFillOpacity() {
            return this.am;
        }
        set markerFillOpacity(a) {
            this.am = a;
            this.g("MarkerFillOpacity");
        }
        get actualMarkerFillOpacity() {
            return this.ab;
        }
        set actualMarkerFillOpacity(a) {
            this.ab = a;
            this.g("ActualMarkerFillOpacity");
        }
        get outlineMode() {
            return this.cu;
        }
        set outlineMode(a) {
            this.cu = a;
            this.g("OutlineMode");
        }
        get actualOutlineMode() {
            return this.b5;
        }
        set actualOutlineMode(a) {
            this.b5 = a;
            this.g("ActualOutlineMode");
        }
        get highlightingFadeOpacity() {
            return this.al;
        }
        set highlightingFadeOpacity(a) {
            this.al = a;
            this.g("HighlightingFadeOpacity");
        }
        get actualHighlightingFadeOpacity() {
            return this.aa;
        }
        set actualHighlightingFadeOpacity(a) {
            this.aa = a;
            this.g("ActualHighlightingFadeOpacity");
        }
        get radiusX() {
            return this.ap;
        }
        set radiusX(a) {
            this.ap = a;
            this.g("RadiusX");
        }
        get actualRadiusX() {
            return this.ae;
        }
        set actualRadiusX(a) {
            this.ae = a;
            this.g("ActualRadiusX");
        }
        get actualIsSplineShapePartOfRange() {
            return this.k;
        }
        set actualIsSplineShapePartOfRange(a) {
            this.k = a;
            this.g("ActualIsSplineShapePartOfRange");
        }
        get isSplineShapePartOfRange() {
            return this.o;
        }
        set isSplineShapePartOfRange(a) {
            this.o = a;
            this.g("IsSplineShapePartOfRange");
        }
        get radiusY() {
            return this.aq;
        }
        set radiusY(a) {
            this.aq = a;
            this.g("RadiusY");
        }
        get actualRadiusY() {
            return this.af;
        }
        set actualRadiusY(a) {
            this.af = a;
            this.g("ActualRadiusY");
        }
        get thickness() {
            return this.au;
        }
        set thickness(a) {
            this.au = a;
            this.g("Thickness");
        }
        get actualThickness() {
            return this.aj;
        }
        set actualThickness(a) {
            this.aj = a;
            this.g("ActualThickness");
        }
        get markerThickness() {
            return this.an;
        }
        set markerThickness(a) {
            this.an = a;
            this.g("MarkerThickness");
        }
        get actualMarkerThickness() {
            return this.ac;
        }
        set actualMarkerThickness(a) {
            this.ac = a;
            this.g("ActualMarkerThickness");
        }
        get title() {
            return this.bp;
        }
        set title(a) {
            this.bp = a;
            this.g("Title");
        }
        get transitionEasingFunctionRef() {
            return this.cy;
        }
        set transitionEasingFunctionRef(a) {
            this.cy = a;
            this.g("TransitionEasingFunctionRef");
        }
        get transitionDuration() {
            return this.bj;
        }
        set transitionDuration(a) {
            this.bj = a;
            this.g("TransitionDuration");
        }
        get transitionInDuration() {
            return this.bk;
        }
        set transitionInDuration(a) {
            this.bk = a;
            this.g("TransitionInDuration");
        }
        get transitionInEasingFunctionRef() {
            return this.cz;
        }
        set transitionInEasingFunctionRef(a) {
            this.cz = a;
            this.g("TransitionInEasingFunctionRef");
        }
        get transitionInMode() {
            return this.c0;
        }
        set transitionInMode(a) {
            this.c0 = a;
            this.g("TransitionInMode");
        }
        get isTransitionInEnabled() {
            return this.p;
        }
        set isTransitionInEnabled(a) {
            this.p = a;
            this.g("IsTransitionInEnabled");
        }
        get transitionInSpeedType() {
            return this.c1;
        }
        set transitionInSpeedType(a) {
            this.c1 = a;
            this.g("TransitionInSpeedType");
        }
        get actualIsTransitionInEnabled() {
            return this.l;
        }
        set actualIsTransitionInEnabled(a) {
            this.l = a;
            this.g("ActualIsTransitionInEnabled");
        }
        get actualTransitionInDuration() {
            return this.bi;
        }
        set actualTransitionInDuration(a) {
            this.bi = a;
            this.g("ActualTransitionInDuration");
        }
        get actualTransitionInMode() {
            return this.b9;
        }
        set actualTransitionInMode(a) {
            this.b9 = a;
            this.g("ActualTransitionInMode");
        }
        get actualTransitionInSpeedType() {
            return this.ca;
        }
        set actualTransitionInSpeedType(a) {
            this.ca = a;
            this.g("ActualTransitionInSpeedType");
        }
        get actualTransitionInEasingFunctionRef() {
            return this.b8;
        }
        set actualTransitionInEasingFunctionRef(a) {
            this.b8 = a;
            this.g("ActualTransitionInEasingFunctionRef");
        }
        get actualTransitionEasingFunctionRef() {
            return this.b7;
        }
        set actualTransitionEasingFunctionRef(a) {
            this.b7 = a;
            this.g("ActualTransitionEasingFunctionRef");
        }
        get actualTransitionDuration() {
            return this.bh;
        }
        set actualTransitionDuration(a) {
            this.bh = a;
            this.g("ActualTransitionDuration");
        }
        get valueMemberPath() {
            return this.c4;
        }
        set valueMemberPath(a) {
            this.c4 = a;
            this.g("ValueMemberPath");
        }
        get valueMemberAsLegendLabel() {
            return this.c2;
        }
        set valueMemberAsLegendLabel(a) {
            this.c2 = a;
            this.g("ValueMemberAsLegendLabel");
        }
        get valueMemberAsLegendUnit() {
            return this.c3;
        }
        set valueMemberAsLegendUnit(a) {
            this.c3 = a;
            this.g("ValueMemberAsLegendUnit");
        }
        get actualValueMemberAsLegendLabel() {
            return this.cb;
        }
        set actualValueMemberAsLegendLabel(a) {
            this.cb = a;
            this.g("ActualValueMemberAsLegendLabel");
        }
        get actualValueMemberAsLegendUnit() {
            return this.cc;
        }
        set actualValueMemberAsLegendUnit(a) {
            this.cc = a;
            this.g("ActualValueMemberAsLegendUnit");
        }
        get visibility() {
            return this.c5;
        }
        set visibility(a) {
            this.c5 = a;
            this.g("Visibility");
        }
        get actualVisibility() {
            return this.cd;
        }
        set actualVisibility(a) {
            this.cd = a;
            this.g("ActualVisibility");
        }
        get parentOrLocalBrush() {
            return this.cv;
        }
        set parentOrLocalBrush(a) {
            this.cv = a;
            this.g("ParentOrLocalBrush");
        }
        get propertyUpdatedRef() {
            return this.cw;
        }
        set propertyUpdatedRef(a) {
            this.cw = a;
            this.g("PropertyUpdatedRef");
        }
    }
    StackedFragmentSeriesDescription.$t = /*@__PURE__*/ markType(StackedFragmentSeriesDescription, 'StackedFragmentSeriesDescription', Description.$);
    return StackedFragmentSeriesDescription;
})();
