// igniteui-react-core/esm2015/lib/SummaryRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SummaryRowDescription = /*@__PURE__*/ (() => {
    class SummaryRowDescription extends DefinitionBaseDescription {
        get_type() {
            return "SummaryRow";
        }
        constructor() {
            super();
            this.cg = null;
            this.ce = null;
            this.ck = null;
            this.cf = null;
            this.ch = null;
            this.b4 = 0;
            this.ci = null;
            this.cj = null;
            this.b7 = 0;
            this.b9 = 0;
            this.b8 = 0;
            this.b6 = 0;
        }
        get selectedBackground() {
            return this.cg;
        }
        set selectedBackground(a) {
            this.cg = a;
            this.g("SelectedBackground");
        }
        get actualSelectedBackground() {
            return this.ce;
        }
        set actualSelectedBackground(a) {
            this.ce = a;
            this.g("ActualSelectedBackground");
        }
        get summaryLabelTextColor() {
            return this.ck;
        }
        set summaryLabelTextColor(a) {
            this.ck = a;
            this.g("SummaryLabelTextColor");
        }
        get actualSummaryLabelTextColor() {
            return this.cf;
        }
        set actualSummaryLabelTextColor(a) {
            this.cf = a;
            this.g("ActualSummaryLabelTextColor");
        }
        get summaryLabelFontFamily() {
            return this.ch;
        }
        set summaryLabelFontFamily(a) {
            this.ch = a;
            this.g("SummaryLabelFontFamily");
        }
        get summaryLabelFontSize() {
            return this.b4;
        }
        set summaryLabelFontSize(a) {
            this.b4 = a;
            this.g("SummaryLabelFontSize");
        }
        get summaryLabelFontStyle() {
            return this.ci;
        }
        set summaryLabelFontStyle(a) {
            this.ci = a;
            this.g("SummaryLabelFontStyle");
        }
        get summaryLabelFontWeight() {
            return this.cj;
        }
        set summaryLabelFontWeight(a) {
            this.cj = a;
            this.g("SummaryLabelFontWeight");
        }
        get paddingLeft() {
            return this.b7;
        }
        set paddingLeft(a) {
            this.b7 = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.b9;
        }
        set paddingTop(a) {
            this.b9 = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.b8;
        }
        set paddingRight(a) {
            this.b8 = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.b6;
        }
        set paddingBottom(a) {
            this.b6 = a;
            this.g("PaddingBottom");
        }
    }
    SummaryRowDescription.$t = /*@__PURE__*/ markType(SummaryRowDescription, 'SummaryRowDescription', DefinitionBaseDescription.$);
    return SummaryRowDescription;
})();
