// igniteui-react-core/esm2015/lib/ToolActionRadioDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionRadioDescription = /*@__PURE__*/ (() => {
    class ToolActionRadioDescription extends ToolActionDescription {
        get_type() {
            return "ToolActionRadio";
        }
        constructor() {
            super();
            this.ce = null;
            this.cc = false;
        }
        get channel() {
            return this.ce;
        }
        set channel(a) {
            this.ce = a;
            this.g("Channel");
        }
        get isChecked() {
            return this.cc;
        }
        set isChecked(a) {
            this.cc = a;
            this.g("IsChecked");
        }
    }
    ToolActionRadioDescription.$t = /*@__PURE__*/ markType(ToolActionRadioDescription, 'ToolActionRadioDescription', ToolActionDescription.$);
    return ToolActionRadioDescription;
})();
