// igniteui-react-core/esm2015/lib/ToolActionInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let ToolActionInfo = /*@__PURE__*/ (() => {
    class ToolActionInfo extends Base {
        constructor() {
            super();
            this._parentId = null;
            this._commandId = null;
            this._commandArgument = null;
            this._commandArgumentValue = null;
            this._paddingLeft = 0;
            this._paddingRight = 0;
            this._paddingBottom = 0;
            this._paddingTop = 0;
            this._name = null;
            this._title = null;
            this._subtitle = null;
            this._iconName = null;
            this._iconCollectionName = null;
            this._iconViewBoxLeft = 0;
            this._iconViewBoxTop = 0;
            this._iconViewBoxWidth = 0;
            this._iconViewBoxHeight = 0;
            this._isDisabled = false;
            this._textColor = null;
            this._disabledTextColor = null;
            this._actions = null;
            this.iconCollectionName = "default";
            this.iconViewBoxLeft = NaN;
            this.iconViewBoxTop = NaN;
            this.iconViewBoxWidth = NaN;
            this.iconViewBoxHeight = NaN;
            this.paddingLeft = NaN;
            this.paddingRight = NaN;
            this.paddingBottom = NaN;
            this.paddingTop = NaN;
            this.isDisabled = false;
        }
        get parentId() {
            return this._parentId;
        }
        set parentId(a) {
            this._parentId = a;
        }
        get commandId() {
            return this._commandId;
        }
        set commandId(a) {
            this._commandId = a;
        }
        get commandArgument() {
            return this._commandArgument;
        }
        set commandArgument(a) {
            this._commandArgument = a;
        }
        get commandArgumentValue() {
            return this._commandArgumentValue;
        }
        set commandArgumentValue(a) {
            this._commandArgumentValue = a;
        }
        get paddingLeft() {
            return this._paddingLeft;
        }
        set paddingLeft(a) {
            this._paddingLeft = a;
        }
        get paddingRight() {
            return this._paddingRight;
        }
        set paddingRight(a) {
            this._paddingRight = a;
        }
        get paddingBottom() {
            return this._paddingBottom;
        }
        set paddingBottom(a) {
            this._paddingBottom = a;
        }
        get paddingTop() {
            return this._paddingTop;
        }
        set paddingTop(a) {
            this._paddingTop = a;
        }
        get name() {
            return this._name;
        }
        set name(a) {
            this._name = a;
        }
        get title() {
            return this._title;
        }
        set title(a) {
            this._title = a;
        }
        get subtitle() {
            return this._subtitle;
        }
        set subtitle(a) {
            this._subtitle = a;
        }
        get iconName() {
            return this._iconName;
        }
        set iconName(a) {
            this._iconName = a;
        }
        get iconCollectionName() {
            return this._iconCollectionName;
        }
        set iconCollectionName(a) {
            this._iconCollectionName = a;
        }
        get iconViewBoxLeft() {
            return this._iconViewBoxLeft;
        }
        set iconViewBoxLeft(a) {
            this._iconViewBoxLeft = a;
        }
        get iconViewBoxTop() {
            return this._iconViewBoxTop;
        }
        set iconViewBoxTop(a) {
            this._iconViewBoxTop = a;
        }
        get iconViewBoxWidth() {
            return this._iconViewBoxWidth;
        }
        set iconViewBoxWidth(a) {
            this._iconViewBoxWidth = a;
        }
        get iconViewBoxHeight() {
            return this._iconViewBoxHeight;
        }
        set iconViewBoxHeight(a) {
            this._iconViewBoxHeight = a;
        }
        get isDisabled() {
            return this._isDisabled;
        }
        set isDisabled(a) {
            this._isDisabled = a;
        }
        get textColor() {
            return this._textColor;
        }
        set textColor(a) {
            this._textColor = a;
        }
        get disabledTextColor() {
            return this._disabledTextColor;
        }
        set disabledTextColor(a) {
            this._disabledTextColor = a;
        }
        get actions() {
            return this._actions;
        }
        set actions(a) {
            this._actions = a;
        }
        get_b() {
            return 0;
        }
        get b() {
            return this.get_b();
        }
    }
    ToolActionInfo.$t = /*@__PURE__*/ markType(ToolActionInfo, 'ToolActionInfo');
    return ToolActionInfo;
})();
