// igniteui-react-core/esm2015/lib/ToolActionSeparatorInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionInfoDescription } from "./ToolActionInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionSeparatorInfoDescription = /*@__PURE__*/ (() => {
    class ToolActionSeparatorInfoDescription extends ToolActionInfoDescription {
        get_type() {
            return "ToolActionSeparatorInfo";
        }
        constructor() {
            super();
            this.a3 = 0;
            this.a1 = false;
        }
        get size() {
            return this.a3;
        }
        set size(a) {
            this.a3 = a;
            this.g("Size");
        }
        get isGroupHeaderSeparator() {
            return this.a1;
        }
        set isGroupHeaderSeparator(a) {
            this.a1 = a;
            this.g("IsGroupHeaderSeparator");
        }
    }
    ToolActionSeparatorInfoDescription.$t = /*@__PURE__*/ markType(ToolActionSeparatorInfoDescription, 'ToolActionSeparatorInfoDescription', ToolActionInfoDescription.$);
    ToolActionSeparatorInfoDescription.__marshalByValue1 = true;
    ToolActionSeparatorInfoDescription.__marshalByValueAlias1 = "ToolActionSeparatorInfo";
    return ToolActionSeparatorInfoDescription;
})();
