// igniteui-react-core/esm2015/lib/TreemapNodeStyleMappingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { TreemapNodeStyleDescription } from "./TreemapNodeStyleDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let TreemapNodeStyleMappingDescription = /*@__PURE__*/ (() => {
    class TreemapNodeStyleMappingDescription extends TreemapNodeStyleDescription {
        get_type() {
            return "TreemapNodeStyleMapping";
        }
        constructor() {
            super();
            this.a3 = null;
            this.ax = 0;
            this.aw = 0;
            this.a0 = null;
            this.a2 = null;
        }
        get targetType() {
            return this.a3;
        }
        set targetType(a) {
            this.a3 = a;
            this.g("TargetType");
        }
        get minimumValue() {
            return this.ax;
        }
        set minimumValue(a) {
            this.ax = a;
            this.g("MinimumValue");
        }
        get maximumValue() {
            return this.aw;
        }
        set maximumValue(a) {
            this.aw = a;
            this.g("MaximumValue");
        }
        get value() {
            return this.a0;
        }
        set value(a) {
            this.a0 = a;
            this.g("Value");
        }
        get mappingMode() {
            return this.a2;
        }
        set mappingMode(a) {
            this.a2 = a;
            this.g("MappingMode");
        }
    }
    TreemapNodeStyleMappingDescription.$t = /*@__PURE__*/ markType(TreemapNodeStyleMappingDescription, 'TreemapNodeStyleMappingDescription', TreemapNodeStyleDescription.$);
    return TreemapNodeStyleMappingDescription;
})();
