// igniteui-react-core/esm2015/lib/TreemapNodeStylingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let TreemapNodeStylingEventArgsDescription = /*@__PURE__*/ (() => {
    class TreemapNodeStylingEventArgsDescription extends Description {
        get_type() {
            return "TreemapNodeStylingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.o = 0;
            this.n = 0;
            this.t = null;
            this.u = null;
            this.w = null;
            this.v = null;
            this.m = 0;
            this.x = null;
            this.l = 0;
            this.j = false;
            this.h = null;
        }
        get value() {
            return this.o;
        }
        set value(a) {
            this.o = a;
            this.g("Value");
        }
        get sum() {
            return this.n;
        }
        set sum(a) {
            this.n = a;
            this.g("Sum");
        }
        get customValueRef() {
            return this.t;
        }
        set customValueRef(a) {
            this.t = a;
            this.g("CustomValueRef");
        }
        get itemRef() {
            return this.u;
        }
        set itemRef(a) {
            this.u = a;
            this.g("ItemRef");
        }
        get parentItemRef() {
            return this.w;
        }
        set parentItemRef(a) {
            this.w = a;
            this.g("ParentItemRef");
        }
        get label() {
            return this.v;
        }
        set label(a) {
            this.v = a;
            this.g("Label");
        }
        get parentValue() {
            return this.m;
        }
        set parentValue(a) {
            this.m = a;
            this.g("ParentValue");
        }
        get parentLabel() {
            return this.x;
        }
        set parentLabel(a) {
            this.x = a;
            this.g("ParentLabel");
        }
        get parentSum() {
            return this.l;
        }
        set parentSum(a) {
            this.l = a;
            this.g("ParentSum");
        }
        get isParent() {
            return this.j;
        }
        set isParent(a) {
            this.j = a;
            this.g("IsParent");
        }
        get style() {
            return this.h;
        }
        set style(a) {
            this.h = a;
            this.g("Style");
        }
    }
    TreemapNodeStylingEventArgsDescription.$t = /*@__PURE__*/ markType(TreemapNodeStylingEventArgsDescription, 'TreemapNodeStylingEventArgsDescription', Description.$);
    return TreemapNodeStylingEventArgsDescription;
})();
