// igniteui-react-core/esm2015/lib/WaterfallSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeriesDescription } from "./HorizontalAnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WaterfallSeriesDescription = /*@__PURE__*/ (() => {
    class WaterfallSeriesDescription extends HorizontalAnchoredCategorySeriesDescription {
        get_type() {
            return "WaterfallSeries";
        }
        constructor() {
            super();
            this.e5 = null;
            this.e1 = 0;
            this.e2 = 0;
            this.e6 = null;
        }
        get negativeBrush() {
            return this.e5;
        }
        set negativeBrush(a) {
            this.e5 = a;
            this.g("NegativeBrush");
        }
        get radiusX() {
            return this.e1;
        }
        set radiusX(a) {
            this.e1 = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.e2;
        }
        set radiusY(a) {
            this.e2 = a;
            this.g("RadiusY");
        }
        get negativeOutline() {
            return this.e6;
        }
        set negativeOutline(a) {
            this.e6 = a;
            this.g("NegativeOutline");
        }
    }
    WaterfallSeriesDescription.$t = /*@__PURE__*/ markType(WaterfallSeriesDescription, 'WaterfallSeriesDescription', HorizontalAnchoredCategorySeriesDescription.$);
    return WaterfallSeriesDescription;
})();
