// igniteui-react-core/esm2015/lib/WebActionStripDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebDisplayDensityBaseElementDescription } from "./WebDisplayDensityBaseElementDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebActionStripDescription = /*@__PURE__*/ (() => {
    class WebActionStripDescription extends WebDisplayDensityBaseElementDescription {
        get_type() {
            return "WebActionStrip";
        }
        constructor() {
            super();
            this.s = null;
            this.m = null;
            this.q = false;
            this.o = null;
            this.u = null;
        }
        get context() {
            return this.s;
        }
        set context(a) {
            this.s = a;
            this.g("Context");
        }
        get actionButtons() {
            return this.m;
        }
        set actionButtons(a) {
            this.m = a;
            this.g("ActionButtons");
        }
        get hidden() {
            return this.q;
        }
        set hidden(a) {
            this.q = a;
            this.g("Hidden");
        }
        get resourceStrings() {
            return this.o;
        }
        set resourceStrings(a) {
            this.o = a;
            this.g("ResourceStrings");
        }
        get name() {
            return this.u;
        }
        set name(a) {
            this.u = a;
            this.g("Name");
        }
    }
    WebActionStripDescription.$t = /*@__PURE__*/ markType(WebActionStripDescription, 'WebActionStripDescription', WebDisplayDensityBaseElementDescription.$);
    return WebActionStripDescription;
})();
