// igniteui-react-core/esm2015/lib/WebComboChangeEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebComboChangeEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebComboChangeEventArgsDetailDescription extends Description {
        get_type() {
            return "WebComboChangeEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.k = null;
            this.n = null;
        }
        get newValue() {
            return this.h;
        }
        set newValue(a) {
            this.h = a;
            this.g("NewValue");
        }
        get items() {
            return this.k;
        }
        set items(a) {
            this.k = a;
            this.g("Items");
        }
        get changeType() {
            return this.n;
        }
        set changeType(a) {
            this.n = a;
            this.g("ChangeType");
        }
    }
    WebComboChangeEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebComboChangeEventArgsDetailDescription, 'WebComboChangeEventArgsDetailDescription', Description.$);
    WebComboChangeEventArgsDetailDescription.__marshalByValue = true;
    WebComboChangeEventArgsDetailDescription.__marshalByValueAlias = "ComboChangeEventArgsDetail";
    return WebComboChangeEventArgsDetailDescription;
})();
