// igniteui-react-core/esm2015/lib/WebColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnDescription = /*@__PURE__*/ (() => {
    class WebColumnDescription extends Description {
        get_type() {
            return "WebColumn";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.bm = null;
            this.bp = null;
            this.bz = null;
            this.ad = false;
            this.ab = false;
            this.u = false;
            this.r = false;
            this.s = false;
            this.z = false;
            this.o = false;
            this.v = false;
            this.w = false;
            this.ac = false;
            this.p = false;
            this.q = false;
            this.x = false;
            this.b0 = null;
            this.bu = null;
            this.bq = null;
            this.a8 = null;
            this.br = null;
            this.a7 = null;
            this.bi = null;
            this.bj = null;
            this.t = false;
            this.ae = false;
            this.aa = false;
            this.bk = null;
            this.a0 = 0;
            this.ay = 0;
            this.a1 = 0;
            this.az = 0;
            this.a6 = null;
            this.bv = null;
            this.y = false;
            this.ba = null;
            this.j = null;
            this.l = null;
            this.by = null;
            this.bh = null;
            this.bs = null;
            this.bt = null;
            this.bl = null;
            this.bn = null;
            this.af = false;
            this.h = null;
            this.a9 = null;
            this.bw = null;
            this.bo = null;
            this.bx = null;
        }
        get field() {
            return this.bm;
        }
        set field(a) {
            this.bm = a;
            this.g("Field");
        }
        get header() {
            return this.bp;
        }
        set header(a) {
            this.bp = a;
            this.g("Header");
        }
        get title() {
            return this.bz;
        }
        set title(a) {
            this.bz = a;
            this.g("Title");
        }
        get sortable() {
            return this.ad;
        }
        set sortable(a) {
            this.ad = a;
            this.g("Sortable");
        }
        get selectable() {
            return this.ab;
        }
        set selectable(a) {
            this.ab = a;
            this.g("Selectable");
        }
        get groupable() {
            return this.u;
        }
        set groupable(a) {
            this.u = a;
            this.g("Groupable");
        }
        get editable() {
            return this.r;
        }
        set editable(a) {
            this.r = a;
            this.g("Editable");
        }
        get filterable() {
            return this.s;
        }
        set filterable(a) {
            this.s = a;
            this.g("Filterable");
        }
        get resizable() {
            return this.z;
        }
        set resizable(a) {
            this.z = a;
            this.g("Resizable");
        }
        get autosizeHeader() {
            return this.o;
        }
        set autosizeHeader(a) {
            this.o = a;
            this.g("AutosizeHeader");
        }
        get hasSummary() {
            return this.v;
        }
        set hasSummary(a) {
            this.v = a;
            this.g("HasSummary");
        }
        get hidden() {
            return this.w;
        }
        set hidden(a) {
            this.w = a;
            this.g("Hidden");
        }
        get selected() {
            return this.ac;
        }
        set selected(a) {
            this.ac = a;
            this.g("Selected");
        }
        get disableHiding() {
            return this.p;
        }
        set disableHiding(a) {
            this.p = a;
            this.g("DisableHiding");
        }
        get disablePinning() {
            return this.q;
        }
        set disablePinning(a) {
            this.q = a;
            this.g("DisablePinning");
        }
        get movable() {
            return this.x;
        }
        set movable(a) {
            this.x = a;
            this.g("Movable");
        }
        get width() {
            return this.b0;
        }
        set width(a) {
            this.b0 = a;
            this.g("Width");
        }
        get maxWidth() {
            return this.bu;
        }
        set maxWidth(a) {
            this.bu = a;
            this.g("MaxWidth");
        }
        get headerClasses() {
            return this.bq;
        }
        set headerClasses(a) {
            this.bq = a;
            this.g("HeaderClasses");
        }
        get headerStyles() {
            return this.a8;
        }
        set headerStyles(a) {
            this.a8 = a;
            this.g("HeaderStyles");
        }
        get headerGroupClasses() {
            return this.br;
        }
        set headerGroupClasses(a) {
            this.br = a;
            this.g("HeaderGroupClasses");
        }
        get headerGroupStyles() {
            return this.a7;
        }
        set headerGroupStyles(a) {
            this.a7 = a;
            this.g("HeaderGroupStyles");
        }
        get cellClassesRef() {
            return this.bi;
        }
        set cellClassesRef(a) {
            this.bi = a;
            this.g("CellClassesRef");
        }
        get cellStylesRef() {
            return this.bj;
        }
        set cellStylesRef(a) {
            this.bj = a;
            this.g("CellStylesRef");
        }
        get filteringIgnoreCase() {
            return this.t;
        }
        set filteringIgnoreCase(a) {
            this.t = a;
            this.g("FilteringIgnoreCase");
        }
        get sortingIgnoreCase() {
            return this.ae;
        }
        set sortingIgnoreCase(a) {
            this.ae = a;
            this.g("SortingIgnoreCase");
        }
        get searchable() {
            return this.aa;
        }
        set searchable(a) {
            this.aa = a;
            this.g("Searchable");
        }
        get dataType() {
            return this.bk;
        }
        set dataType(a) {
            this.bk = a;
            this.g("DataType");
        }
        get rowEnd() {
            return this.a0;
        }
        set rowEnd(a) {
            this.a0 = a;
            this.g("RowEnd");
        }
        get colEnd() {
            return this.ay;
        }
        set colEnd(a) {
            this.ay = a;
            this.g("ColEnd");
        }
        get rowStart() {
            return this.a1;
        }
        set rowStart(a) {
            this.a1 = a;
            this.g("RowStart");
        }
        get colStart() {
            return this.az;
        }
        set colStart(a) {
            this.az = a;
            this.g("ColStart");
        }
        get additionalTemplateContext() {
            return this.a6;
        }
        set additionalTemplateContext(a) {
            this.a6 = a;
            this.g("AdditionalTemplateContext");
        }
        get minWidth() {
            return this.bv;
        }
        set minWidth(a) {
            this.bv = a;
            this.g("MinWidth");
        }
        get pinned() {
            return this.y;
        }
        set pinned(a) {
            this.y = a;
            this.g("Pinned");
        }
        get summaries() {
            return this.ba;
        }
        set summaries(a) {
            this.ba = a;
            this.g("Summaries");
        }
        get filters() {
            return this.j;
        }
        set filters(a) {
            this.j = a;
            this.g("Filters");
        }
        get sortStrategy() {
            return this.l;
        }
        set sortStrategy(a) {
            this.l = a;
            this.g("SortStrategy");
        }
        get summaryTemplateRef() {
            return this.by;
        }
        set summaryTemplateRef(a) {
            this.by = a;
            this.g("SummaryTemplateRef");
        }
        get bodyTemplateRef() {
            return this.bh;
        }
        set bodyTemplateRef(a) {
            this.bh = a;
            this.g("BodyTemplateRef");
        }
        get headerTemplateRef() {
            return this.bs;
        }
        set headerTemplateRef(a) {
            this.bs = a;
            this.g("HeaderTemplateRef");
        }
        get inlineEditorTemplateRef() {
            return this.bt;
        }
        set inlineEditorTemplateRef(a) {
            this.bt = a;
            this.g("InlineEditorTemplateRef");
        }
        get errorTemplateRef() {
            return this.bl;
        }
        set errorTemplateRef(a) {
            this.bl = a;
            this.g("ErrorTemplateRef");
        }
        get filterCellTemplateRef() {
            return this.bn;
        }
        set filterCellTemplateRef(a) {
            this.bn = a;
            this.g("FilterCellTemplateRef");
        }
        get visibleWhenCollapsed() {
            return this.af;
        }
        set visibleWhenCollapsed(a) {
            this.af = a;
            this.g("VisibleWhenCollapsed");
        }
        get pipeArgs() {
            return this.h;
        }
        set pipeArgs(a) {
            this.h = a;
            this.g("PipeArgs");
        }
        get parentColumn() {
            return this.a9;
        }
        set parentColumn(a) {
            this.a9 = a;
            this.g("ParentColumn");
        }
        get name() {
            return this.bw;
        }
        set name(a) {
            this.bw = a;
            this.g("Name");
        }
        get formatterRef() {
            return this.bo;
        }
        set formatterRef(a) {
            this.bo = a;
            this.g("FormatterRef");
        }
        get summaryFormatterRef() {
            return this.bx;
        }
        set summaryFormatterRef(a) {
            this.bx = a;
            this.g("SummaryFormatterRef");
        }
    }
    WebColumnDescription.$t = /*@__PURE__*/ markType(WebColumnDescription, 'WebColumnDescription', Description.$);
    WebColumnDescription.__canMarshalByValue = true;
    WebColumnDescription.__marshalByValueAlias = "Column";
    return WebColumnDescription;
})();
