// igniteui-react-core/esm2015/lib/WebCellDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCellDescription = /*@__PURE__*/ (() => {
    class WebCellDescription extends Description {
        get_type() {
            return "WebCell";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.k = null;
            this.i = null;
            this.h = null;
        }
        get primaryKey() {
            return this.j;
        }
        set primaryKey(a) {
            this.j = a;
            this.g("PrimaryKey");
        }
        get state() {
            return this.k;
        }
        set state(a) {
            this.k = a;
            this.g("State");
        }
        get pendingValue() {
            return this.i;
        }
        set pendingValue(a) {
            this.i = a;
            this.g("PendingValue");
        }
        get editValue() {
            return this.h;
        }
        set editValue(a) {
            this.h = a;
            this.g("EditValue");
        }
    }
    WebCellDescription.$t = /*@__PURE__*/ markType(WebCellDescription, 'WebCellDescription', Description.$);
    return WebCellDescription;
})();
