// igniteui-react-core/esm2015/lib/WebComboItemDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebComboItemDescription = /*@__PURE__*/ (() => {
    class WebComboItemDescription extends Description {
        get_type() {
            return "WebComboItem";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = 0;
            this.j = false;
            this.h = false;
            this.i = false;
        }
        get index() {
            return this.n;
        }
        set index(a) {
            this.n = a;
            this.g("Index");
        }
        get selected() {
            return this.j;
        }
        set selected(a) {
            this.j = a;
            this.g("Selected");
        }
        get active() {
            return this.h;
        }
        set active(a) {
            this.h = a;
            this.g("Active");
        }
        get hideCheckbox() {
            return this.i;
        }
        set hideCheckbox(a) {
            this.i = a;
            this.g("HideCheckbox");
        }
    }
    WebComboItemDescription.$t = /*@__PURE__*/ markType(WebComboItemDescription, 'WebComboItemDescription', Description.$);
    return WebComboItemDescription;
})();
