// igniteui-react-core/esm2015/lib/WebColumnSelectionEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnSelectionEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebColumnSelectionEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebColumnSelectionEventArgsDetail";
        }
        constructor() {
            super();
            this.o = null;
            this.n = null;
            this.m = null;
            this.p = null;
        }
        get oldSelection() {
            return this.o;
        }
        set oldSelection(a) {
            this.o = a;
            this.g("OldSelection");
        }
        get newSelection() {
            return this.n;
        }
        set newSelection(a) {
            this.n = a;
            this.g("NewSelection");
        }
        get added() {
            return this.m;
        }
        set added(a) {
            this.m = a;
            this.g("Added");
        }
        get removed() {
            return this.p;
        }
        set removed(a) {
            this.p = a;
            this.g("Removed");
        }
    }
    WebColumnSelectionEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebColumnSelectionEventArgsDetailDescription, 'WebColumnSelectionEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebColumnSelectionEventArgsDetailDescription.__marshalByValue1 = true;
    WebColumnSelectionEventArgsDetailDescription.__marshalByValueAlias1 = "ColumnSelectionEventArgsDetail";
    return WebColumnSelectionEventArgsDetailDescription;
})();
