// igniteui-react-core/esm2015/lib/WebDisplayDensityBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDisplayDensityBaseDescription = /*@__PURE__*/ (() => {
    class WebDisplayDensityBaseDescription extends Description {
        get_type() {
            return "WebDisplayDensityBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
        }
        get displayDensity() {
            return this.i;
        }
        set displayDensity(a) {
            this.i = a;
            this.g("DisplayDensity");
        }
        get densityChangedRef() {
            return this.h;
        }
        set densityChangedRef(a) {
            this.h = a;
            this.g("DensityChangedRef");
        }
    }
    WebDisplayDensityBaseDescription.$t = /*@__PURE__*/ markType(WebDisplayDensityBaseDescription, 'WebDisplayDensityBaseDescription', Description.$);
    return WebDisplayDensityBaseDescription;
})();
