// igniteui-react-core/esm2015/lib/WebDisplayDensityBaseElementDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDisplayDensityBaseElementDescription = /*@__PURE__*/ (() => {
    class WebDisplayDensityBaseElementDescription extends Description {
        get_type() {
            return "WebDisplayDensityBaseElement";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
        }
        get displayDensity() {
            return this.h;
        }
        set displayDensity(a) {
            this.h = a;
            this.g("DisplayDensity");
        }
        get onDensityChangedRef() {
            return this.i;
        }
        set onDensityChangedRef(a) {
            this.i = a;
            this.g("OnDensityChangedRef");
        }
    }
    WebDisplayDensityBaseElementDescription.$t = /*@__PURE__*/ markType(WebDisplayDensityBaseElementDescription, 'WebDisplayDensityBaseElementDescription', Description.$);
    return WebDisplayDensityBaseElementDescription;
})();
