// igniteui-react-core/esm2015/lib/WebDropdownItemDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDropdownItemDescription = /*@__PURE__*/ (() => {
    class WebDropdownItemDescription extends Description {
        get_type() {
            return "WebDropdownItem";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.j = false;
            this.h = false;
            this.i = false;
        }
        get value() {
            return this.n;
        }
        set value(a) {
            this.n = a;
            this.g("Value");
        }
        get selected() {
            return this.j;
        }
        set selected(a) {
            this.j = a;
            this.g("Selected");
        }
        get active() {
            return this.h;
        }
        set active(a) {
            this.h = a;
            this.g("Active");
        }
        get disabled() {
            return this.i;
        }
        set disabled(a) {
            this.i = a;
            this.g("Disabled");
        }
    }
    WebDropdownItemDescription.$t = /*@__PURE__*/ markType(WebDropdownItemDescription, 'WebDropdownItemDescription', Description.$);
    return WebDropdownItemDescription;
})();
