// igniteui-react-core/esm2015/lib/WebDropdownDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDropdownDescription = /*@__PURE__*/ (() => {
    class WebDropdownDescription extends Description {
        get_type() {
            return "WebDropdown";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = false;
            this.k = false;
            this.u = null;
            this.v = null;
            this.w = null;
            this.h = false;
            this.r = 0;
            this.i = false;
            this.l = false;
            this.x = null;
            this.t = null;
        }
        get keepOpenOnSelect() {
            return this.j;
        }
        set keepOpenOnSelect(a) {
            this.j = a;
            this.g("KeepOpenOnSelect");
        }
        get open() {
            return this.k;
        }
        set open(a) {
            this.k = a;
            this.g("Open");
        }
        get placement() {
            return this.u;
        }
        set placement(a) {
            this.u = a;
            this.g("Placement");
        }
        get positionStrategy() {
            return this.v;
        }
        set positionStrategy(a) {
            this.v = a;
            this.g("PositionStrategy");
        }
        get scrollStrategy() {
            return this.w;
        }
        set scrollStrategy(a) {
            this.w = a;
            this.g("ScrollStrategy");
        }
        get flip() {
            return this.h;
        }
        set flip(a) {
            this.h = a;
            this.g("Flip");
        }
        get distance() {
            return this.r;
        }
        set distance(a) {
            this.r = a;
            this.g("Distance");
        }
        get keepOpenOnOutsideClick() {
            return this.i;
        }
        set keepOpenOnOutsideClick(a) {
            this.i = a;
            this.g("KeepOpenOnOutsideClick");
        }
        get sameWidth() {
            return this.l;
        }
        set sameWidth(a) {
            this.l = a;
            this.g("SameWidth");
        }
        get size() {
            return this.x;
        }
        set size(a) {
            this.x = a;
            this.g("Size");
        }
        get changeRef() {
            return this.t;
        }
        set changeRef(a) {
            this.t = a;
            this.g("ChangeRef");
        }
    }
    WebDropdownDescription.$t = /*@__PURE__*/ markType(WebDropdownDescription, 'WebDropdownDescription', Description.$);
    return WebDropdownDescription;
})();
