// igniteui-react-core/esm2015/lib/WebFormDataDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebFormDataDescription = /*@__PURE__*/ (() => {
    class WebFormDataDescription extends Description {
        get_type() {
            return "WebFormData";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
        }
        get keys() {
            return this.h;
        }
        set keys(a) {
            this.h = a;
            this.g("Keys");
        }
        get values() {
            return this.i;
        }
        set values(a) {
            this.i = a;
            this.g("Values");
        }
    }
    WebFormDataDescription.$t = /*@__PURE__*/ markType(WebFormDataDescription, 'WebFormDataDescription', Description.$);
    WebFormDataDescription.__marshalByValue = true;
    WebFormDataDescription.__marshalByValueAlias = "FormData";
    return WebFormDataDescription;
})();
