// igniteui-react-core/esm2015/lib/WebGridKeydownEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridKeydownEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridKeydownEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebGridKeydownEventArgsDetail";
        }
        constructor() {
            super();
            this.s = null;
            this.p = null;
            this.n = false;
        }
        get targetType() {
            return this.s;
        }
        set targetType(a) {
            this.s = a;
            this.g("TargetType");
        }
        get target() {
            return this.p;
        }
        set target(a) {
            this.p = a;
            this.g("Target");
        }
        get cancel() {
            return this.n;
        }
        set cancel(a) {
            this.n = a;
            this.g("Cancel");
        }
    }
    WebGridKeydownEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridKeydownEventArgsDetailDescription, 'WebGridKeydownEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridKeydownEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridKeydownEventArgsDetailDescription.__marshalByValueAlias1 = "GridKeydownEventArgsDetail";
    return WebGridKeydownEventArgsDetailDescription;
})();
