// igniteui-react-core/esm2015/lib/WebGridSelectionRangeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridSelectionRangeDescription = /*@__PURE__*/ (() => {
    class WebGridSelectionRangeDescription extends Description {
        get_type() {
            return "WebGridSelectionRange";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.i = 0;
            this.o = null;
            this.n = null;
        }
        get rowStart() {
            return this.j;
        }
        set rowStart(a) {
            this.j = a;
            this.g("RowStart");
        }
        get rowEnd() {
            return this.i;
        }
        set rowEnd(a) {
            this.i = a;
            this.g("RowEnd");
        }
        get columnStart() {
            return this.o;
        }
        set columnStart(a) {
            this.o = a;
            this.g("ColumnStart");
        }
        get columnEnd() {
            return this.n;
        }
        set columnEnd(a) {
            this.n = a;
            this.g("ColumnEnd");
        }
    }
    WebGridSelectionRangeDescription.$t = /*@__PURE__*/ markType(WebGridSelectionRangeDescription, 'WebGridSelectionRangeDescription', Description.$);
    WebGridSelectionRangeDescription.__marshalByValue = true;
    WebGridSelectionRangeDescription.__marshalByValueAlias = "GridSelectionRange";
    return WebGridSelectionRangeDescription;
})();
