// igniteui-react-core/esm2015/lib/WebGridDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebGridBaseDirectiveDescription } from "./WebGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridDescription = /*@__PURE__*/ (() => {
    class WebGridDescription extends WebGridBaseDirectiveDescription {
        get_type() {
            return "WebGrid";
        }
        constructor() {
            super();
            this.gp = false;
            this.g1 = null;
            this.gz = null;
            this.g5 = null;
            this.gy = null;
            this.gv = 0;
            this.gk = null;
            this.gj = null;
            this.gq = false;
            this.gn = null;
            this.g0 = null;
            this.g2 = null;
            this.g4 = null;
            this.gr = false;
            this.gx = null;
            this.g3 = null;
        }
        get groupsExpanded() {
            return this.gp;
        }
        set groupsExpanded(a) {
            this.gp = a;
            this.g("GroupsExpanded");
        }
        get dropAreaTemplateRef() {
            return this.g1;
        }
        set dropAreaTemplateRef(a) {
            this.g1 = a;
            this.g("DropAreaTemplateRef");
        }
        get detailTemplateRef() {
            return this.gz;
        }
        set detailTemplateRef(a) {
            this.gz = a;
            this.g("DetailTemplateRef");
        }
        get id() {
            return this.g5;
        }
        set id(a) {
            this.g5 = a;
            this.g("Id");
        }
        get dataRef() {
            return this.gy;
        }
        set dataRef(a) {
            this.gy = a;
            this.g("DataRef");
        }
        get totalItemCount() {
            return this.gv;
        }
        set totalItemCount(a) {
            this.gv = a;
            this.g("TotalItemCount");
        }
        get groupingExpressions() {
            return this.gk;
        }
        set groupingExpressions(a) {
            this.gk = a;
            this.g("GroupingExpressions");
        }
        get groupingExpansionState() {
            return this.gj;
        }
        set groupingExpansionState(a) {
            this.gj = a;
            this.g("GroupingExpansionState");
        }
        get hideGroupedColumns() {
            return this.gq;
        }
        set hideGroupedColumns(a) {
            this.gq = a;
            this.g("HideGroupedColumns");
        }
        get groupStrategy() {
            return this.gn;
        }
        set groupStrategy(a) {
            this.gn = a;
            this.g("GroupStrategy");
        }
        get dropAreaMessage() {
            return this.g0;
        }
        set dropAreaMessage(a) {
            this.g0 = a;
            this.g("DropAreaMessage");
        }
        get groupByRowSelectorTemplateRef() {
            return this.g2;
        }
        set groupByRowSelectorTemplateRef(a) {
            this.g2 = a;
            this.g("GroupByRowSelectorTemplateRef");
        }
        get groupRowTemplateRef() {
            return this.g4;
        }
        set groupRowTemplateRef(a) {
            this.g4 = a;
            this.g("GroupRowTemplateRef");
        }
        get showGroupArea() {
            return this.gr;
        }
        set showGroupArea(a) {
            this.gr = a;
            this.g("ShowGroupArea");
        }
        get dataPreLoadRef() {
            return this.gx;
        }
        set dataPreLoadRef(a) {
            this.gx = a;
            this.g("DataPreLoadRef");
        }
        get groupingDoneRef() {
            return this.g3;
        }
        set groupingDoneRef(a) {
            this.g3 = a;
            this.g("GroupingDoneRef");
        }
    }
    WebGridDescription.$t = /*@__PURE__*/ markType(WebGridDescription, 'WebGridDescription', WebGridBaseDirectiveDescription.$);
    return WebGridDescription;
})();
