// igniteui-react-core/esm2015/lib/WebGridBaseDirectiveDescription_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebDisplayDensityBaseDescription } from "./WebDisplayDensityBaseDescription";
import { markType } from "./type";
import { WebDisplayDensityBaseElementDescription } from "./WebDisplayDensityBaseElementDescription";
/**
 * @hidden
 */
export let WebGridBaseDirectiveDescription = /*@__PURE__*/ (() => {
    class WebGridBaseDirectiveDescription extends WebDisplayDensityBaseDescription {
        get_type() {
            return "WebGridBaseDirective";
        }
        constructor() {
            super();
            this.bj = 0;
            this.aw = false;
            this.n = null;
            this.a0 = false;
            this.ck = null;
            this.br = null;
            this.cy = null;
            this.bk = 0;
            this.aa = null;
            this.ai = null;
            this.c7 = null;
            this.dq = null;
            this.c1 = null;
            this.y = null;
            this.cg = null;
            this.di = null;
            this.c6 = null;
            this.de = null;
            this.dj = null;
            this.c8 = null;
            this.cw = null;
            this.cv = null;
            this.cl = null;
            this.du = null;
            this.dv = null;
            this.dw = null;
            this.r = null;
            this.p = null;
            this.ak = null;
            this.cp = null;
            this.ad = null;
            this.ac = null;
            this.cz = null;
            this.c0 = null;
            this.ay = false;
            this.a1 = false;
            this.d4 = null;
            this.a2 = false;
            this.bi = 0;
            this.cb = null;
            this.cj = null;
            this.az = false;
            this.a4 = false;
            this.ci = null;
            this.aq = null;
            this.av = false;
            this.au = false;
            this.cq = null;
            this.d1 = null;
            this.d0 = null;
            this.a5 = false;
            this.ag = null;
            this.am = null;
            this.as = null;
            this.dt = null;
            this.cx = null;
            this.dp = null;
            this.ch = null;
            this.q = null;
            this.ax = false;
            this.by = null;
            this.dm = null;
            this.b7 = null;
            this.ao = null;
            this.o = null;
            this.bl = 0;
            this.a3 = false;
            this.m = null;
            this.co = null;
            this.bs = null;
            this.cu = null;
            this.bt = null;
            this.cr = null;
            this.d3 = null;
            this.ds = null;
            this.dn = null;
            this.b8 = null;
            this.b3 = null;
            this.b4 = null;
            this.bw = null;
            this.bx = null;
            this.bu = null;
            this.bv = null;
            this.dg = null;
            this.dd = null;
            this.df = null;
            this.dh = null;
            this.bz = null;
            this.b6 = null;
            this.dx = null;
            this.dy = null;
            this.cm = null;
            this.cn = null;
            this.c5 = null;
            this.da = null;
            this.c9 = null;
            this.c4 = null;
            this.b5 = null;
            this.cc = null;
            this.cf = null;
            this.ca = null;
            this.b9 = null;
            this.b2 = null;
            this.b0 = null;
            this.b1 = null;
            this.ct = null;
            this.dc = null;
            this.db = null;
            this.cs = null;
            this.dr = null;
            this.dl = null;
            this.dk = null;
            this.bq = null;
            this.dz = null;
            this.d2 = null;
            this.c2 = null;
            this.c3 = null;
            this.ce = null;
            this.cd = null;
        }
        get snackbarDisplayTime() {
            return this.bj;
        }
        set snackbarDisplayTime(a) {
            this.bj = a;
            this.g("SnackbarDisplayTime");
        }
        get autoGenerate() {
            return this.aw;
        }
        set autoGenerate(a) {
            this.aw = a;
            this.g("AutoGenerate");
        }
        get autoGenerateExclude() {
            return this.n;
        }
        set autoGenerateExclude(a) {
            this.n = a;
            this.g("AutoGenerateExclude");
        }
        get moving() {
            return this.a0;
        }
        set moving(a) {
            this.a0 = a;
            this.g("Moving");
        }
        get emptyGridTemplateRef() {
            return this.ck;
        }
        set emptyGridTemplateRef(a) {
            this.ck = a;
            this.g("EmptyGridTemplateRef");
        }
        get addRowEmptyTemplateRef() {
            return this.br;
        }
        set addRowEmptyTemplateRef(a) {
            this.br = a;
            this.g("AddRowEmptyTemplateRef");
        }
        get loadingGridTemplateRef() {
            return this.cy;
        }
        set loadingGridTemplateRef(a) {
            this.cy = a;
            this.g("LoadingGridTemplateRef");
        }
        get summaryRowHeight() {
            return this.bk;
        }
        set summaryRowHeight(a) {
            this.bk = a;
            this.g("SummaryRowHeight");
        }
        get dataCloneStrategy() {
            return this.aa;
        }
        set dataCloneStrategy(a) {
            this.aa = a;
            this.g("DataCloneStrategy");
        }
        get clipboardOptions() {
            return this.ai;
        }
        set clipboardOptions(a) {
            this.ai = a;
            this.g("ClipboardOptions");
        }
        get rowClassesRef() {
            return this.c7;
        }
        set rowClassesRef(a) {
            this.c7 = a;
            this.g("RowClassesRef");
        }
        get rowStylesRef() {
            return this.dq;
        }
        set rowStylesRef(a) {
            this.dq = a;
            this.g("RowStylesRef");
        }
        get primaryKey() {
            return this.c1;
        }
        set primaryKey(a) {
            this.c1 = a;
            this.g("PrimaryKey");
        }
        get actionStrip() {
            return this.y;
        }
        set actionStrip(a) {
            this.y = a;
            this.g("ActionStrip");
        }
        get dragGhostCustomTemplateRef() {
            return this.cg;
        }
        set dragGhostCustomTemplateRef(a) {
            this.cg = a;
            this.g("DragGhostCustomTemplateRef");
        }
        get rowEditTextTemplateRef() {
            return this.di;
        }
        set rowEditTextTemplateRef(a) {
            this.di = a;
            this.g("RowEditTextTemplateRef");
        }
        get rowAddTextTemplateRef() {
            return this.c6;
        }
        set rowAddTextTemplateRef(a) {
            this.c6 = a;
            this.g("RowAddTextTemplateRef");
        }
        get rowEditActionsTemplateRef() {
            return this.de;
        }
        set rowEditActionsTemplateRef(a) {
            this.de = a;
            this.g("RowEditActionsTemplateRef");
        }
        get rowExpandedIndicatorTemplateRef() {
            return this.dj;
        }
        set rowExpandedIndicatorTemplateRef(a) {
            this.dj = a;
            this.g("RowExpandedIndicatorTemplateRef");
        }
        get rowCollapsedIndicatorTemplateRef() {
            return this.c8;
        }
        set rowCollapsedIndicatorTemplateRef(a) {
            this.c8 = a;
            this.g("RowCollapsedIndicatorTemplateRef");
        }
        get headerExpandedIndicatorTemplateRef() {
            return this.cw;
        }
        set headerExpandedIndicatorTemplateRef(a) {
            this.cw = a;
            this.g("HeaderExpandedIndicatorTemplateRef");
        }
        get headerCollapsedIndicatorTemplateRef() {
            return this.cv;
        }
        set headerCollapsedIndicatorTemplateRef(a) {
            this.cv = a;
            this.g("HeaderCollapsedIndicatorTemplateRef");
        }
        get excelStyleHeaderIconTemplateRef() {
            return this.cl;
        }
        set excelStyleHeaderIconTemplateRef(a) {
            this.cl = a;
            this.g("ExcelStyleHeaderIconTemplateRef");
        }
        get sortAscendingHeaderIconTemplateRef() {
            return this.du;
        }
        set sortAscendingHeaderIconTemplateRef(a) {
            this.du = a;
            this.g("SortAscendingHeaderIconTemplateRef");
        }
        get sortDescendingHeaderIconTemplateRef() {
            return this.dv;
        }
        set sortDescendingHeaderIconTemplateRef(a) {
            this.dv = a;
            this.g("SortDescendingHeaderIconTemplateRef");
        }
        get sortHeaderIconTemplateRef() {
            return this.dw;
        }
        set sortHeaderIconTemplateRef(a) {
            this.dw = a;
            this.g("SortHeaderIconTemplateRef");
        }
        get toolbar() {
            return this.r;
        }
        set toolbar(a) {
            this.r = a;
            this.g("Toolbar");
        }
        get paginationComponents() {
            return this.p;
        }
        set paginationComponents(a) {
            this.p = a;
            this.g("PaginationComponents");
        }
        get resourceStrings() {
            return this.ak;
        }
        set resourceStrings(a) {
            this.ak = a;
            this.g("ResourceStrings");
        }
        get filteringLogic() {
            return this.cp;
        }
        set filteringLogic(a) {
            this.cp = a;
            this.g("FilteringLogic");
        }
        get filteringExpressionsTree() {
            return this.ad;
        }
        set filteringExpressionsTree(a) {
            this.ad = a;
            this.g("FilteringExpressionsTree");
        }
        get advancedFilteringExpressionsTree() {
            return this.ac;
        }
        set advancedFilteringExpressionsTree(a) {
            this.ac = a;
            this.g("AdvancedFilteringExpressionsTree");
        }
        get locale() {
            return this.cz;
        }
        set locale(a) {
            this.cz = a;
            this.g("Locale");
        }
        get pagingMode() {
            return this.c0;
        }
        set pagingMode(a) {
            this.c0 = a;
            this.g("PagingMode");
        }
        get hideRowSelectors() {
            return this.ay;
        }
        set hideRowSelectors(a) {
            this.ay = a;
            this.g("HideRowSelectors");
        }
        get rowDraggable() {
            return this.a1;
        }
        set rowDraggable(a) {
            this.a1 = a;
            this.g("RowDraggable");
        }
        get validationTrigger() {
            return this.d4;
        }
        set validationTrigger(a) {
            this.d4 = a;
            this.g("ValidationTrigger");
        }
        get rowEditable() {
            return this.a2;
        }
        set rowEditable(a) {
            this.a2 = a;
            this.g("RowEditable");
        }
        get rowHeight() {
            return this.bi;
        }
        set rowHeight(a) {
            this.bi = a;
            this.g("RowHeight");
        }
        get columnWidth() {
            return this.cb;
        }
        set columnWidth(a) {
            this.cb = a;
            this.g("ColumnWidth");
        }
        get emptyGridMessage() {
            return this.cj;
        }
        set emptyGridMessage(a) {
            this.cj = a;
            this.g("EmptyGridMessage");
        }
        get isLoading() {
            return this.az;
        }
        set isLoading(a) {
            this.az = a;
            this.g("IsLoading");
        }
        get shouldGenerate() {
            return this.a4;
        }
        set shouldGenerate(a) {
            this.a4 = a;
            this.g("ShouldGenerate");
        }
        get emptyFilteredGridMessage() {
            return this.ci;
        }
        set emptyFilteredGridMessage(a) {
            this.ci = a;
            this.g("EmptyFilteredGridMessage");
        }
        get pinning() {
            return this.aq;
        }
        set pinning(a) {
            this.aq = a;
            this.g("Pinning");
        }
        get allowFiltering() {
            return this.av;
        }
        set allowFiltering(a) {
            this.av = a;
            this.g("AllowFiltering");
        }
        get allowAdvancedFiltering() {
            return this.au;
        }
        set allowAdvancedFiltering(a) {
            this.au = a;
            this.g("AllowAdvancedFiltering");
        }
        get filterMode() {
            return this.cq;
        }
        set filterMode(a) {
            this.cq = a;
            this.g("FilterMode");
        }
        get summaryPosition() {
            return this.d1;
        }
        set summaryPosition(a) {
            this.d1 = a;
            this.g("SummaryPosition");
        }
        get summaryCalculationMode() {
            return this.d0;
        }
        set summaryCalculationMode(a) {
            this.d0 = a;
            this.g("SummaryCalculationMode");
        }
        get showSummaryOnCollapse() {
            return this.a5;
        }
        set showSummaryOnCollapse(a) {
            this.a5 = a;
            this.g("ShowSummaryOnCollapse");
        }
        get filterStrategy() {
            return this.ag;
        }
        set filterStrategy(a) {
            this.ag = a;
            this.g("FilterStrategy");
        }
        get sortStrategy() {
            return this.am;
        }
        set sortStrategy(a) {
            this.am = a;
            this.g("SortStrategy");
        }
        get sortingOptions() {
            return this.as;
        }
        set sortingOptions(a) {
            this.as = a;
            this.g("SortingOptions");
        }
        get selectedRowsRef() {
            return this.dt;
        }
        set selectedRowsRef(a) {
            this.dt = a;
            this.g("SelectedRowsRef");
        }
        get headSelectorTemplateRef() {
            return this.cx;
        }
        set headSelectorTemplateRef(a) {
            this.cx = a;
            this.g("HeadSelectorTemplateRef");
        }
        get rowSelectorTemplateRef() {
            return this.dp;
        }
        set rowSelectorTemplateRef(a) {
            this.dp = a;
            this.g("RowSelectorTemplateRef");
        }
        get dragIndicatorIconTemplateRef() {
            return this.ch;
        }
        set dragIndicatorIconTemplateRef(a) {
            this.ch = a;
            this.g("DragIndicatorIconTemplateRef");
        }
        get sortingExpressions() {
            return this.q;
        }
        set sortingExpressions(a) {
            this.q = a;
            this.g("SortingExpressions");
        }
        get batchEditing() {
            return this.ax;
        }
        set batchEditing(a) {
            this.ax = a;
            this.g("BatchEditing");
        }
        get cellSelection() {
            return this.by;
        }
        set cellSelection(a) {
            this.by = a;
            this.g("CellSelection");
        }
        get rowSelection() {
            return this.dm;
        }
        set rowSelection(a) {
            this.dm = a;
            this.g("RowSelection");
        }
        get columnSelection() {
            return this.b7;
        }
        set columnSelection(a) {
            this.b7 = a;
            this.g("ColumnSelection");
        }
        get outlet() {
            return this.ao;
        }
        set outlet(a) {
            this.ao = a;
            this.g("Outlet");
        }
        get columns() {
            return this.o;
        }
        set columns(a) {
            this.o = a;
            this.g("Columns");
        }
        get totalRecords() {
            return this.bl;
        }
        set totalRecords(a) {
            this.bl = a;
            this.g("TotalRecords");
        }
        get selectRowOnClick() {
            return this.a3;
        }
        set selectRowOnClick(a) {
            this.a3 = a;
            this.g("SelectRowOnClick");
        }
        get actualColumns() {
            return this.m;
        }
        set actualColumns(a) {
            this.m = a;
            this.g("ActualColumns");
        }
        get filteringExpressionsTreeChangeRef() {
            return this.co;
        }
        set filteringExpressionsTreeChangeRef(a) {
            this.co = a;
            this.g("FilteringExpressionsTreeChangeRef");
        }
        get advancedFilteringExpressionsTreeChangeRef() {
            return this.bs;
        }
        set advancedFilteringExpressionsTreeChangeRef(a) {
            this.bs = a;
            this.g("AdvancedFilteringExpressionsTreeChangeRef");
        }
        get gridScrollRef() {
            return this.cu;
        }
        set gridScrollRef(a) {
            this.cu = a;
            this.g("GridScrollRef");
        }
        get cellClickRef() {
            return this.bt;
        }
        set cellClickRef(a) {
            this.bt = a;
            this.g("CellClickRef");
        }
        get formGroupCreatedRef() {
            return this.cr;
        }
        set formGroupCreatedRef(a) {
            this.cr = a;
            this.g("FormGroupCreatedRef");
        }
        get validationStatusChangeRef() {
            return this.d3;
        }
        set validationStatusChangeRef(a) {
            this.d3 = a;
            this.g("ValidationStatusChangeRef");
        }
        get selectedRef() {
            return this.ds;
        }
        set selectedRef(a) {
            this.ds = a;
            this.g("SelectedRef");
        }
        get rowSelectionChangingRef() {
            return this.dn;
        }
        set rowSelectionChangingRef(a) {
            this.dn = a;
            this.g("RowSelectionChangingRef");
        }
        get columnSelectionChangingRef() {
            return this.b8;
        }
        set columnSelectionChangingRef(a) {
            this.b8 = a;
            this.g("ColumnSelectionChangingRef");
        }
        get columnPinRef() {
            return this.b3;
        }
        set columnPinRef(a) {
            this.b3 = a;
            this.g("ColumnPinRef");
        }
        get columnPinnedRef() {
            return this.b4;
        }
        set columnPinnedRef(a) {
            this.b4 = a;
            this.g("ColumnPinnedRef");
        }
        get cellEditEnterRef() {
            return this.bw;
        }
        set cellEditEnterRef(a) {
            this.bw = a;
            this.g("CellEditEnterRef");
        }
        get cellEditExitRef() {
            return this.bx;
        }
        set cellEditExitRef(a) {
            this.bx = a;
            this.g("CellEditExitRef");
        }
        get cellEditRef() {
            return this.bu;
        }
        set cellEditRef(a) {
            this.bu = a;
            this.g("CellEditRef");
        }
        get cellEditDoneRef() {
            return this.bv;
        }
        set cellEditDoneRef(a) {
            this.bv = a;
            this.g("CellEditDoneRef");
        }
        get rowEditEnterRef() {
            return this.dg;
        }
        set rowEditEnterRef(a) {
            this.dg = a;
            this.g("RowEditEnterRef");
        }
        get rowEditRef() {
            return this.dd;
        }
        set rowEditRef(a) {
            this.dd = a;
            this.g("RowEditRef");
        }
        get rowEditDoneRef() {
            return this.df;
        }
        set rowEditDoneRef(a) {
            this.df = a;
            this.g("RowEditDoneRef");
        }
        get rowEditExitRef() {
            return this.dh;
        }
        set rowEditExitRef(a) {
            this.dh = a;
            this.g("RowEditExitRef");
        }
        get columnInitRef() {
            return this.bz;
        }
        set columnInitRef(a) {
            this.bz = a;
            this.g("ColumnInitRef");
        }
        get columnsAutogeneratedRef() {
            return this.b6;
        }
        set columnsAutogeneratedRef(a) {
            this.b6 = a;
            this.g("ColumnsAutogeneratedRef");
        }
        get sortingRef() {
            return this.dx;
        }
        set sortingRef(a) {
            this.dx = a;
            this.g("SortingRef");
        }
        get sortingDoneRef() {
            return this.dy;
        }
        set sortingDoneRef(a) {
            this.dy = a;
            this.g("SortingDoneRef");
        }
        get filteringRef() {
            return this.cm;
        }
        set filteringRef(a) {
            this.cm = a;
            this.g("FilteringRef");
        }
        get filteringDoneRef() {
            return this.cn;
        }
        set filteringDoneRef(a) {
            this.cn = a;
            this.g("FilteringDoneRef");
        }
        get rowAddedRef() {
            return this.c5;
        }
        set rowAddedRef(a) {
            this.c5 = a;
            this.g("RowAddedRef");
        }
        get rowDeletedRef() {
            return this.da;
        }
        set rowDeletedRef(a) {
            this.da = a;
            this.g("RowDeletedRef");
        }
        get rowDeleteRef() {
            return this.c9;
        }
        set rowDeleteRef(a) {
            this.c9 = a;
            this.g("RowDeleteRef");
        }
        get rowAddRef() {
            return this.c4;
        }
        set rowAddRef(a) {
            this.c4 = a;
            this.g("RowAddRef");
        }
        get columnResizedRef() {
            return this.b5;
        }
        set columnResizedRef(a) {
            this.b5 = a;
            this.g("ColumnResizedRef");
        }
        get contextMenuRef() {
            return this.cc;
        }
        set contextMenuRef(a) {
            this.cc = a;
            this.g("ContextMenuRef");
        }
        get doubleClickRef() {
            return this.cf;
        }
        set doubleClickRef(a) {
            this.cf = a;
            this.g("DoubleClickRef");
        }
        get columnVisibilityChangingRef() {
            return this.ca;
        }
        set columnVisibilityChangingRef(a) {
            this.ca = a;
            this.g("ColumnVisibilityChangingRef");
        }
        get columnVisibilityChangedRef() {
            return this.b9;
        }
        set columnVisibilityChangedRef(a) {
            this.b9 = a;
            this.g("ColumnVisibilityChangedRef");
        }
        get columnMovingStartRef() {
            return this.b2;
        }
        set columnMovingStartRef(a) {
            this.b2 = a;
            this.g("ColumnMovingStartRef");
        }
        get columnMovingRef() {
            return this.b0;
        }
        set columnMovingRef(a) {
            this.b0 = a;
            this.g("ColumnMovingRef");
        }
        get columnMovingEndRef() {
            return this.b1;
        }
        set columnMovingEndRef(a) {
            this.b1 = a;
            this.g("ColumnMovingEndRef");
        }
        get gridKeydownRef() {
            return this.ct;
        }
        set gridKeydownRef(a) {
            this.ct = a;
            this.g("GridKeydownRef");
        }
        get rowDragStartRef() {
            return this.dc;
        }
        set rowDragStartRef(a) {
            this.dc = a;
            this.g("RowDragStartRef");
        }
        get rowDragEndRef() {
            return this.db;
        }
        set rowDragEndRef(a) {
            this.db = a;
            this.g("RowDragEndRef");
        }
        get gridCopyRef() {
            return this.cs;
        }
        set gridCopyRef(a) {
            this.cs = a;
            this.g("GridCopyRef");
        }
        get rowToggleRef() {
            return this.dr;
        }
        set rowToggleRef(a) {
            this.dr = a;
            this.g("RowToggleRef");
        }
        get rowPinningRef() {
            return this.dl;
        }
        set rowPinningRef(a) {
            this.dl = a;
            this.g("RowPinningRef");
        }
        get rowPinnedRef() {
            return this.dk;
        }
        set rowPinnedRef(a) {
            this.dk = a;
            this.g("RowPinnedRef");
        }
        get activeNodeChangeRef() {
            return this.bq;
        }
        set activeNodeChangeRef(a) {
            this.bq = a;
            this.g("ActiveNodeChangeRef");
        }
        get sortingExpressionsChangeRef() {
            return this.dz;
        }
        set sortingExpressionsChangeRef(a) {
            this.dz = a;
            this.g("SortingExpressionsChangeRef");
        }
        get toolbarExportingRef() {
            return this.d2;
        }
        set toolbarExportingRef(a) {
            this.d2 = a;
            this.g("ToolbarExportingRef");
        }
        get rangeSelectedRef() {
            return this.c2;
        }
        set rangeSelectedRef(a) {
            this.c2 = a;
            this.g("RangeSelectedRef");
        }
        get renderedRef() {
            return this.c3;
        }
        set renderedRef(a) {
            this.c3 = a;
            this.g("RenderedRef");
        }
        get dataChangingRef() {
            return this.ce;
        }
        set dataChangingRef(a) {
            this.ce = a;
            this.g("DataChangingRef");
        }
        get dataChangedRef() {
            return this.cd;
        }
        set dataChangedRef(a) {
            this.cd = a;
            this.g("DataChangedRef");
        }
    }
    WebGridBaseDirectiveDescription.$t = /*@__PURE__*/ markType(WebGridBaseDirectiveDescription, 'WebGridBaseDirectiveDescription', WebDisplayDensityBaseDescription.$);
    return WebGridBaseDirectiveDescription;
})();
/**
 * @hidden
 */
export let WebGridToolbarDescription = /*@__PURE__*/ (() => {
    class WebGridToolbarDescription extends WebDisplayDensityBaseElementDescription {
        get_type() {
            return "WebGridToolbar";
        }
        constructor() {
            super();
            this.q = false;
            this.o = null;
            this.s = null;
            this.m = null;
        }
        get showProgress() {
            return this.q;
        }
        set showProgress(a) {
            this.q = a;
            this.g("ShowProgress");
        }
        get grid() {
            return this.o;
        }
        set grid(a) {
            this.o = a;
            this.g("Grid");
        }
        get name() {
            return this.s;
        }
        set name(a) {
            this.s = a;
            this.g("Name");
        }
        get tools() {
            return this.m;
        }
        set tools(a) {
            this.m = a;
            this.g("Tools");
        }
    }
    WebGridToolbarDescription.$t = /*@__PURE__*/ markType(WebGridToolbarDescription, 'WebGridToolbarDescription', WebDisplayDensityBaseElementDescription.$);
    return WebGridToolbarDescription;
})();
