// igniteui-react-core/esm2015/lib/WebPinRowEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebCancelableEventArgsDetailDescription } from "./WebCancelableEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPinRowEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebPinRowEventArgsDetailDescription extends WebCancelableEventArgsDetailDescription {
        get_type() {
            return "WebPinRowEventArgsDetail";
        }
        constructor() {
            super();
            this.t = null;
            this.m = null;
            this.r = 0;
            this.p = false;
        }
        get rowID() {
            return this.t;
        }
        set rowID(a) {
            this.t = a;
            this.g("RowID");
        }
        get row() {
            return this.m;
        }
        set row(a) {
            this.m = a;
            this.g("Row");
        }
        get insertAtIndex() {
            return this.r;
        }
        set insertAtIndex(a) {
            this.r = a;
            this.g("InsertAtIndex");
        }
        get isPinned() {
            return this.p;
        }
        set isPinned(a) {
            this.p = a;
            this.g("IsPinned");
        }
    }
    WebPinRowEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebPinRowEventArgsDetailDescription, 'WebPinRowEventArgsDetailDescription', WebCancelableEventArgsDetailDescription.$);
    WebPinRowEventArgsDetailDescription.__marshalByValue1 = true;
    WebPinRowEventArgsDetailDescription.__marshalByValueAlias1 = "PinRowEventArgsDetail";
    return WebPinRowEventArgsDetailDescription;
})();
