// igniteui-react-core/esm2015/lib/WebPivotDimensionDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotDimensionDetailDescription = /*@__PURE__*/ (() => {
    class WebPivotDimensionDetailDescription extends Description {
        get_type() {
            return "WebPivotDimensionDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.t = null;
            this.m = false;
            this.h = null;
            this.u = null;
            this.r = null;
            this.v = null;
            this.o = 0;
            this.s = null;
        }
        get childLevel() {
            return this.j;
        }
        set childLevel(a) {
            this.j = a;
            this.g("ChildLevel");
        }
        get memberName() {
            return this.t;
        }
        set memberName(a) {
            this.t = a;
            this.g("MemberName");
        }
        get enabled() {
            return this.m;
        }
        set enabled(a) {
            this.m = a;
            this.g("Enabled");
        }
        get filter() {
            return this.h;
        }
        set filter(a) {
            this.h = a;
            this.g("Filter");
        }
        get sortDirection() {
            return this.u;
        }
        set sortDirection(a) {
            this.u = a;
            this.g("SortDirection");
        }
        get dataType() {
            return this.r;
        }
        set dataType(a) {
            this.r = a;
            this.g("DataType");
        }
        get width() {
            return this.v;
        }
        set width(a) {
            this.v = a;
            this.g("Width");
        }
        get level() {
            return this.o;
        }
        set level(a) {
            this.o = a;
            this.g("Level");
        }
        get memberFunctionRef() {
            return this.s;
        }
        set memberFunctionRef(a) {
            this.s = a;
            this.g("MemberFunctionRef");
        }
    }
    WebPivotDimensionDetailDescription.$t = /*@__PURE__*/ markType(WebPivotDimensionDetailDescription, 'WebPivotDimensionDetailDescription', Description.$);
    WebPivotDimensionDetailDescription.__marshalByValue = true;
    WebPivotDimensionDetailDescription.__marshalByValueAlias = "PivotDimensionDetail";
    return WebPivotDimensionDetailDescription;
})();
