// igniteui-react-core/esm2015/lib/WebPinningConfigDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPinningConfigDescription = /*@__PURE__*/ (() => {
    class WebPinningConfigDescription extends Description {
        get_type() {
            return "WebPinningConfig";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
        }
        get columns() {
            return this.h;
        }
        set columns(a) {
            this.h = a;
            this.g("Columns");
        }
        get rows() {
            return this.i;
        }
        set rows(a) {
            this.i = a;
            this.g("Rows");
        }
    }
    WebPinningConfigDescription.$t = /*@__PURE__*/ markType(WebPinningConfigDescription, 'WebPinningConfigDescription', Description.$);
    return WebPinningConfigDescription;
})();
