// igniteui-react-core/esm2015/lib/WebPathSegmentDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPathSegmentDescription = /*@__PURE__*/ (() => {
    class WebPathSegmentDescription extends Description {
        get_type() {
            return "WebPathSegment";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
        }
        get rowID() {
            return this.h;
        }
        set rowID(a) {
            this.h = a;
            this.g("RowID");
        }
        get rowIslandKey() {
            return this.j;
        }
        set rowIslandKey(a) {
            this.j = a;
            this.g("RowIslandKey");
        }
    }
    WebPathSegmentDescription.$t = /*@__PURE__*/ markType(WebPathSegmentDescription, 'WebPathSegmentDescription', Description.$);
    return WebPathSegmentDescription;
})();
