// igniteui-react-core/esm2015/lib/WebPositionSettingsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPositionSettingsDescription = /*@__PURE__*/ (() => {
    class WebPositionSettingsDescription extends Description {
        get_type() {
            return "WebPositionSettings";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.l = null;
            this.n = null;
            this.m = null;
            this.o = null;
            this.h = null;
        }
        get target() {
            return this.j;
        }
        set target(a) {
            this.j = a;
            this.g("Target");
        }
        get horizontalDirection() {
            return this.l;
        }
        set horizontalDirection(a) {
            this.l = a;
            this.g("HorizontalDirection");
        }
        get verticalDirection() {
            return this.n;
        }
        set verticalDirection(a) {
            this.n = a;
            this.g("VerticalDirection");
        }
        get horizontalStartPoint() {
            return this.m;
        }
        set horizontalStartPoint(a) {
            this.m = a;
            this.g("HorizontalStartPoint");
        }
        get verticalStartPoint() {
            return this.o;
        }
        set verticalStartPoint(a) {
            this.o = a;
            this.g("VerticalStartPoint");
        }
        get minSize() {
            return this.h;
        }
        set minSize(a) {
            this.h = a;
            this.g("MinSize");
        }
    }
    WebPositionSettingsDescription.$t = /*@__PURE__*/ markType(WebPositionSettingsDescription, 'WebPositionSettingsDescription', Description.$);
    return WebPositionSettingsDescription;
})();
