// igniteui-react-core/esm2015/lib/WebRowSelectionEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowSelectionEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebRowSelectionEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebRowSelectionEventArgsDetail";
        }
        constructor() {
            super();
            this.s = null;
            this.r = null;
            this.q = null;
            this.t = null;
            this.n = false;
        }
        get oldSelectionRef() {
            return this.s;
        }
        set oldSelectionRef(a) {
            this.s = a;
            this.g("OldSelectionRef");
        }
        get newSelectionRef() {
            return this.r;
        }
        set newSelectionRef(a) {
            this.r = a;
            this.g("NewSelectionRef");
        }
        get addedRef() {
            return this.q;
        }
        set addedRef(a) {
            this.q = a;
            this.g("AddedRef");
        }
        get removedRef() {
            return this.t;
        }
        set removedRef(a) {
            this.t = a;
            this.g("RemovedRef");
        }
        get allRowsSelected() {
            return this.n;
        }
        set allRowsSelected(a) {
            this.n = a;
            this.g("AllRowsSelected");
        }
    }
    WebRowSelectionEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebRowSelectionEventArgsDetailDescription, 'WebRowSelectionEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowSelectionEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowSelectionEventArgsDetailDescription.__marshalByValueAlias1 = "RowSelectionEventArgsDetail";
    return WebRowSelectionEventArgsDetailDescription;
})();
