// igniteui-react-core/esm2015/lib/WebRowDragStartEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowDragStartEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebRowDragStartEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebRowDragStartEventArgsDetail";
        }
        constructor() {
            super();
            this.p = null;
            this.m = null;
        }
        get dragDirective() {
            return this.p;
        }
        set dragDirective(a) {
            this.p = a;
            this.g("DragDirective");
        }
        get dragData() {
            return this.m;
        }
        set dragData(a) {
            this.m = a;
            this.g("DragData");
        }
    }
    WebRowDragStartEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebRowDragStartEventArgsDetailDescription, 'WebRowDragStartEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowDragStartEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowDragStartEventArgsDetailDescription.__marshalByValueAlias1 = "RowDragStartEventArgsDetail";
    return WebRowDragStartEventArgsDetailDescription;
})();
