// igniteui-react-core/esm2015/lib/WebRowDataEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowDataEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebRowDataEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebRowDataEventArgsDetail";
        }
        constructor() {
            super();
            this.n = null;
            this.o = null;
        }
        get data() {
            return this.n;
        }
        set data(a) {
            this.n = a;
            this.g("Data");
        }
        get primaryKey() {
            return this.o;
        }
        set primaryKey(a) {
            this.o = a;
            this.g("PrimaryKey");
        }
    }
    WebRowDataEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebRowDataEventArgsDetailDescription, 'WebRowDataEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowDataEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowDataEventArgsDetailDescription.__marshalByValueAlias1 = "RowDataEventArgsDetail";
    return WebRowDataEventArgsDetailDescription;
})();
