// igniteui-react-core/esm2015/lib/WebRatingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRatingDescription = /*@__PURE__*/ (() => {
    class WebRatingDescription extends Description {
        get_type() {
            return "WebRating";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.p = 0;
            this.q = 0;
            this.y = null;
            this.x = null;
            this.aa = null;
            this.r = 0;
            this.h = false;
            this.i = false;
            this.j = false;
            this.k = false;
            this.z = null;
            this.v = null;
            this.w = null;
        }
        get max() {
            return this.p;
        }
        set max(a) {
            this.p = a;
            this.g("Max");
        }
        get step() {
            return this.q;
        }
        set step(a) {
            this.q = a;
            this.g("Step");
        }
        get name() {
            return this.y;
        }
        set name(a) {
            this.y = a;
            this.g("Name");
        }
        get label() {
            return this.x;
        }
        set label(a) {
            this.x = a;
            this.g("Label");
        }
        get valueFormat() {
            return this.aa;
        }
        set valueFormat(a) {
            this.aa = a;
            this.g("ValueFormat");
        }
        get value() {
            return this.r;
        }
        set value(a) {
            this.r = a;
            this.g("Value");
        }
        get disabled() {
            return this.h;
        }
        set disabled(a) {
            this.h = a;
            this.g("Disabled");
        }
        get hoverPreview() {
            return this.i;
        }
        set hoverPreview(a) {
            this.i = a;
            this.g("HoverPreview");
        }
        get readonly() {
            return this.j;
        }
        set readonly(a) {
            this.j = a;
            this.g("Readonly");
        }
        get single() {
            return this.k;
        }
        set single(a) {
            this.k = a;
            this.g("Single");
        }
        get size() {
            return this.z;
        }
        set size(a) {
            this.z = a;
            this.g("Size");
        }
        get changeRef() {
            return this.v;
        }
        set changeRef(a) {
            this.v = a;
            this.g("ChangeRef");
        }
        get hoverRef() {
            return this.w;
        }
        set hoverRef(a) {
            this.w = a;
            this.g("HoverRef");
        }
    }
    WebRatingDescription.$t = /*@__PURE__*/ markType(WebRatingDescription, 'WebRatingDescription', Description.$);
    return WebRatingDescription;
})();
