// igniteui-react-core/esm2015/lib/WebSliderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebSliderBaseDescription } from "./WebSliderBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSliderDescription = /*@__PURE__*/ (() => {
    class WebSliderDescription extends WebSliderBaseDescription {
        get_type() {
            return "WebSlider";
        }
        constructor() {
            super();
            this.aq = 0;
            this.as = null;
            this.au = null;
            this.at = null;
        }
        get value() {
            return this.aq;
        }
        set value(a) {
            this.aq = a;
            this.g("Value");
        }
        get ariaLabel() {
            return this.as;
        }
        set ariaLabel(a) {
            this.as = a;
            this.g("AriaLabel");
        }
        get inputRef() {
            return this.au;
        }
        set inputRef(a) {
            this.au = a;
            this.g("InputRef");
        }
        get changeRef() {
            return this.at;
        }
        set changeRef(a) {
            this.at = a;
            this.g("ChangeRef");
        }
    }
    WebSliderDescription.$t = /*@__PURE__*/ markType(WebSliderDescription, 'WebSliderDescription', WebSliderBaseDescription.$);
    return WebSliderDescription;
})();
