// igniteui-react-core/esm2015/lib/WebSelectGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebDropdownGroupDescription } from "./WebDropdownGroupDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSelectGroupDescription = /*@__PURE__*/ (() => {
    class WebSelectGroupDescription extends WebDropdownGroupDescription {
        get_type() {
            return "WebSelectGroup";
        }
        constructor() {
            super();
            this.i = null;
            this.k = false;
        }
        get items() {
            return this.i;
        }
        set items(a) {
            this.i = a;
            this.g("Items");
        }
        get disabled() {
            return this.k;
        }
        set disabled(a) {
            this.k = a;
            this.g("Disabled");
        }
    }
    WebSelectGroupDescription.$t = /*@__PURE__*/ markType(WebSelectGroupDescription, 'WebSelectGroupDescription', WebDropdownGroupDescription.$);
    return WebSelectGroupDescription;
})();
