// igniteui-react-core/esm2015/lib/WebStepDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebStepDescription = /*@__PURE__*/ (() => {
    class WebStepDescription extends Description {
        get_type() {
            return "WebStep";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = false;
            this.h = false;
            this.n = false;
            this.k = false;
            this.i = false;
            this.o = false;
            this.ac = null;
            this.ad = null;
            this.ab = null;
            this.z = 0;
            this.j = false;
            this.m = false;
            this.p = false;
        }
        get invalid() {
            return this.l;
        }
        set invalid(a) {
            this.l = a;
            this.g("Invalid");
        }
        get active() {
            return this.h;
        }
        set active(a) {
            this.h = a;
            this.g("Active");
        }
        get optional() {
            return this.n;
        }
        set optional(a) {
            this.n = a;
            this.g("Optional");
        }
        get disabled() {
            return this.k;
        }
        set disabled(a) {
            this.k = a;
            this.g("Disabled");
        }
        get complete() {
            return this.i;
        }
        set complete(a) {
            this.i = a;
            this.g("Complete");
        }
        get previousComplete() {
            return this.o;
        }
        set previousComplete(a) {
            this.o = a;
            this.g("PreviousComplete");
        }
        get stepType() {
            return this.ac;
        }
        set stepType(a) {
            this.ac = a;
            this.g("StepType");
        }
        get titlePosition() {
            return this.ad;
        }
        set titlePosition(a) {
            this.ad = a;
            this.g("TitlePosition");
        }
        get orientation() {
            return this.ab;
        }
        set orientation(a) {
            this.ab = a;
            this.g("Orientation");
        }
        get index() {
            return this.z;
        }
        set index(a) {
            this.z = a;
            this.g("Index");
        }
        get contentTop() {
            return this.j;
        }
        set contentTop(a) {
            this.j = a;
            this.g("ContentTop");
        }
        get linearDisabled() {
            return this.m;
        }
        set linearDisabled(a) {
            this.m = a;
            this.g("LinearDisabled");
        }
        get visited() {
            return this.p;
        }
        set visited(a) {
            this.p = a;
            this.g("Visited");
        }
    }
    WebStepDescription.$t = /*@__PURE__*/ markType(WebStepDescription, 'WebStepDescription', Description.$);
    return WebStepDescription;
})();
