// igniteui-react-core/esm2015/lib/WebSortingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebCancelableEventArgsDetailDescription } from "./WebCancelableEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSortingEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebSortingEventArgsDetailDescription extends WebCancelableEventArgsDetailDescription {
        get_type() {
            return "WebSortingEventArgsDetail";
        }
        constructor() {
            super();
            this.n = null;
            this.m = null;
        }
        get sortingExpressions() {
            return this.n;
        }
        set sortingExpressions(a) {
            this.n = a;
            this.g("SortingExpressions");
        }
        get groupingExpressions() {
            return this.m;
        }
        set groupingExpressions(a) {
            this.m = a;
            this.g("GroupingExpressions");
        }
    }
    WebSortingEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebSortingEventArgsDetailDescription, 'WebSortingEventArgsDetailDescription', WebCancelableEventArgsDetailDescription.$);
    WebSortingEventArgsDetailDescription.__marshalByValue1 = true;
    WebSortingEventArgsDetailDescription.__marshalByValueAlias1 = "SortingEventArgsDetail";
    return WebSortingEventArgsDetailDescription;
})();
