// igniteui-react-core/esm2015/lib/WebSelectDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebDropdownDescription } from "./WebDropdownDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSelectDescription = /*@__PURE__*/ (() => {
    class WebSelectDescription extends WebDropdownDescription {
        get_type() {
            return "WebSelect";
        }
        constructor() {
            super();
            this.au = null;
            this.as = null;
            this.af = false;
            this.ai = false;
            this.ag = false;
            this.ah = false;
            this.ae = false;
            this.ar = null;
            this.at = null;
            this.ap = null;
            this.aq = null;
            this.ao = null;
        }
        get value() {
            return this.au;
        }
        set value(a) {
            this.au = a;
            this.g("Value");
        }
        get name() {
            return this.as;
        }
        set name(a) {
            this.as = a;
            this.g("Name");
        }
        get disabled() {
            return this.af;
        }
        set disabled(a) {
            this.af = a;
            this.g("Disabled");
        }
        get required() {
            return this.ai;
        }
        set required(a) {
            this.ai = a;
            this.g("Required");
        }
        get invalid() {
            return this.ag;
        }
        set invalid(a) {
            this.ag = a;
            this.g("Invalid");
        }
        get outlined() {
            return this.ah;
        }
        set outlined(a) {
            this.ah = a;
            this.g("Outlined");
        }
        get autofocus() {
            return this.ae;
        }
        set autofocus(a) {
            this.ae = a;
            this.g("Autofocus");
        }
        get label() {
            return this.ar;
        }
        set label(a) {
            this.ar = a;
            this.g("Label");
        }
        get placeholder() {
            return this.at;
        }
        set placeholder(a) {
            this.at = a;
            this.g("Placeholder");
        }
        get dir() {
            return this.ap;
        }
        set dir(a) {
            this.ap = a;
            this.g("Dir");
        }
        get focusRef() {
            return this.aq;
        }
        set focusRef(a) {
            this.aq = a;
            this.g("FocusRef");
        }
        get blurRef() {
            return this.ao;
        }
        set blurRef(a) {
            this.ao = a;
            this.g("BlurRef");
        }
    }
    WebSelectDescription.$t = /*@__PURE__*/ markType(WebSelectDescription, 'WebSelectDescription', WebDropdownDescription.$);
    return WebSelectDescription;
})();
