// igniteui-react-core/esm2015/lib/WebStepperDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebStepperDescription = /*@__PURE__*/ (() => {
    class WebStepperDescription extends Description {
        get_type() {
            return "WebStepper";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.q = null;
            this.r = null;
            this.k = false;
            this.j = false;
            this.p = null;
            this.s = null;
            this.o = null;
            this.n = null;
        }
        get steps() {
            return this.h;
        }
        set steps(a) {
            this.h = a;
            this.g("Steps");
        }
        get orientation() {
            return this.q;
        }
        set orientation(a) {
            this.q = a;
            this.g("Orientation");
        }
        get stepType() {
            return this.r;
        }
        set stepType(a) {
            this.r = a;
            this.g("StepType");
        }
        get linear() {
            return this.k;
        }
        set linear(a) {
            this.k = a;
            this.g("Linear");
        }
        get contentTop() {
            return this.j;
        }
        set contentTop(a) {
            this.j = a;
            this.g("ContentTop");
        }
        get dir() {
            return this.p;
        }
        set dir(a) {
            this.p = a;
            this.g("Dir");
        }
        get titlePosition() {
            return this.s;
        }
        set titlePosition(a) {
            this.s = a;
            this.g("TitlePosition");
        }
        get activeStepChangingRef() {
            return this.o;
        }
        set activeStepChangingRef(a) {
            this.o = a;
            this.g("ActiveStepChangingRef");
        }
        get activeStepChangedRef() {
            return this.n;
        }
        set activeStepChangedRef(a) {
            this.n = a;
            this.g("ActiveStepChangedRef");
        }
    }
    WebStepperDescription.$t = /*@__PURE__*/ markType(WebStepperDescription, 'WebStepperDescription', Description.$);
    return WebStepperDescription;
})();
