// igniteui-react-core/esm2015/lib/WebSnackbarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSnackbarDescription = /*@__PURE__*/ (() => {
    class WebSnackbarDescription extends Description {
        get_type() {
            return "WebSnackbar";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = false;
            this.l = 0;
            this.h = false;
            this.o = null;
            this.n = null;
        }
        get open() {
            return this.i;
        }
        set open(a) {
            this.i = a;
            this.g("Open");
        }
        get displayTime() {
            return this.l;
        }
        set displayTime(a) {
            this.l = a;
            this.g("DisplayTime");
        }
        get keepOpen() {
            return this.h;
        }
        set keepOpen(a) {
            this.h = a;
            this.g("KeepOpen");
        }
        get actionText() {
            return this.o;
        }
        set actionText(a) {
            this.o = a;
            this.g("ActionText");
        }
        get actionRef() {
            return this.n;
        }
        set actionRef(a) {
            this.n = a;
            this.g("ActionRef");
        }
    }
    WebSnackbarDescription.$t = /*@__PURE__*/ markType(WebSnackbarDescription, 'WebSnackbarDescription', Description.$);
    return WebSnackbarDescription;
})();
