// igniteui-react-core/esm2015/lib/WebToggleDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebToggleDirectiveDescription = /*@__PURE__*/ (() => {
    class WebToggleDirectiveDescription extends Description {
        get_type() {
            return "WebToggleDirective";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.l = null;
            this.m = null;
            this.i = null;
            this.j = null;
            this.h = null;
        }
        get id() {
            return this.k;
        }
        set id(a) {
            this.k = a;
            this.g("Id");
        }
        get openedRef() {
            return this.l;
        }
        set openedRef(a) {
            this.l = a;
            this.g("OpenedRef");
        }
        get openingRef() {
            return this.m;
        }
        set openingRef(a) {
            this.m = a;
            this.g("OpeningRef");
        }
        get closedRef() {
            return this.i;
        }
        set closedRef(a) {
            this.i = a;
            this.g("ClosedRef");
        }
        get closingRef() {
            return this.j;
        }
        set closingRef(a) {
            this.j = a;
            this.g("ClosingRef");
        }
        get appendedRef() {
            return this.h;
        }
        set appendedRef(a) {
            this.h = a;
            this.g("AppendedRef");
        }
    }
    WebToggleDirectiveDescription.$t = /*@__PURE__*/ markType(WebToggleDirectiveDescription, 'WebToggleDirectiveDescription', Description.$);
    return WebToggleDirectiveDescription;
})();
