// igniteui-react-core/esm2015/lib/XYChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DomainChartDescription } from "./DomainChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let XYChartDescription = /*@__PURE__*/ (() => {
    class XYChartDescription extends DomainChartDescription {
        get_type() {
            return "XYChart";
        }
        constructor() {
            super();
            this.q9 = null;
            this.rr = null;
            this.pe = 0;
            this.pg = 0;
            this.pf = 0;
            this.pd = 0;
            this.pv = 0;
            this.px = 0;
            this.pw = 0;
            this.pu = 0;
            this.re = null;
            this.rw = null;
            this.q7 = null;
            this.q8 = null;
            this.pp = 0;
            this.p6 = 0;
            this.po = 0;
            this.p5 = 0;
            this.pr = 0;
            this.p8 = 0;
            this.pq = 0;
            this.p7 = 0;
            this.pn = 0;
            this.p4 = 0;
            this.rp = null;
            this.r7 = null;
            this.rf = null;
            this.rx = null;
            this.rq = null;
            this.r8 = null;
            this.rd = null;
            this.rv = null;
            this.ri = null;
            this.r0 = null;
            this.ph = 0;
            this.py = 0;
            this.pi = 0;
            this.pz = 0;
            this.rk = null;
            this.r2 = null;
            this.rl = null;
            this.r3 = null;
            this.pj = 0;
            this.p0 = 0;
            this.pk = 0;
            this.p1 = 0;
            this.rm = null;
            this.r4 = null;
            this.pl = 0;
            this.p2 = 0;
            this.rn = null;
            this.r5 = null;
            this.rj = null;
            this.r1 = null;
            this.pc = 0;
            this.pt = 0;
            this.pb = 0;
            this.ps = 0;
            this.pm = 0;
            this.p3 = 0;
            this.o7 = false;
            this.o8 = false;
            this.ro = null;
            this.r6 = null;
            this.rb = null;
            this.rt = null;
            this.rg = null;
            this.ry = null;
            this.rh = null;
            this.rz = null;
            this.ru = null;
            this.rc = null;
            this.ra = null;
            this.o3 = null;
            this.rs = null;
            this.o4 = null;
        }
        get xAxisFormatLabelRef() {
            return this.q9;
        }
        set xAxisFormatLabelRef(a) {
            this.q9 = a;
            this.g("XAxisFormatLabelRef");
        }
        get yAxisFormatLabelRef() {
            return this.rr;
        }
        set yAxisFormatLabelRef(a) {
            this.rr = a;
            this.g("YAxisFormatLabelRef");
        }
        get xAxisLabelLeftMargin() {
            return this.pe;
        }
        set xAxisLabelLeftMargin(a) {
            this.pe = a;
            this.g("XAxisLabelLeftMargin");
        }
        get xAxisLabelTopMargin() {
            return this.pg;
        }
        set xAxisLabelTopMargin(a) {
            this.pg = a;
            this.g("XAxisLabelTopMargin");
        }
        get xAxisLabelRightMargin() {
            return this.pf;
        }
        set xAxisLabelRightMargin(a) {
            this.pf = a;
            this.g("XAxisLabelRightMargin");
        }
        get xAxisLabelBottomMargin() {
            return this.pd;
        }
        set xAxisLabelBottomMargin(a) {
            this.pd = a;
            this.g("XAxisLabelBottomMargin");
        }
        get yAxisLabelLeftMargin() {
            return this.pv;
        }
        set yAxisLabelLeftMargin(a) {
            this.pv = a;
            this.g("YAxisLabelLeftMargin");
        }
        get yAxisLabelTopMargin() {
            return this.px;
        }
        set yAxisLabelTopMargin(a) {
            this.px = a;
            this.g("YAxisLabelTopMargin");
        }
        get yAxisLabelRightMargin() {
            return this.pw;
        }
        set yAxisLabelRightMargin(a) {
            this.pw = a;
            this.g("YAxisLabelRightMargin");
        }
        get yAxisLabelBottomMargin() {
            return this.pu;
        }
        set yAxisLabelBottomMargin(a) {
            this.pu = a;
            this.g("YAxisLabelBottomMargin");
        }
        get xAxisLabelTextColor() {
            return this.re;
        }
        set xAxisLabelTextColor(a) {
            this.re = a;
            this.g("XAxisLabelTextColor");
        }
        get yAxisLabelTextColor() {
            return this.rw;
        }
        set yAxisLabelTextColor(a) {
            this.rw = a;
            this.g("YAxisLabelTextColor");
        }
        get actualXAxisLabelTextColor() {
            return this.q7;
        }
        set actualXAxisLabelTextColor(a) {
            this.q7 = a;
            this.g("ActualXAxisLabelTextColor");
        }
        get actualYAxisLabelTextColor() {
            return this.q8;
        }
        set actualYAxisLabelTextColor(a) {
            this.q8 = a;
            this.g("ActualYAxisLabelTextColor");
        }
        get xAxisTitleMargin() {
            return this.pp;
        }
        set xAxisTitleMargin(a) {
            this.pp = a;
            this.g("XAxisTitleMargin");
        }
        get yAxisTitleMargin() {
            return this.p6;
        }
        set yAxisTitleMargin(a) {
            this.p6 = a;
            this.g("YAxisTitleMargin");
        }
        get xAxisTitleLeftMargin() {
            return this.po;
        }
        set xAxisTitleLeftMargin(a) {
            this.po = a;
            this.g("XAxisTitleLeftMargin");
        }
        get yAxisTitleLeftMargin() {
            return this.p5;
        }
        set yAxisTitleLeftMargin(a) {
            this.p5 = a;
            this.g("YAxisTitleLeftMargin");
        }
        get xAxisTitleTopMargin() {
            return this.pr;
        }
        set xAxisTitleTopMargin(a) {
            this.pr = a;
            this.g("XAxisTitleTopMargin");
        }
        get yAxisTitleTopMargin() {
            return this.p8;
        }
        set yAxisTitleTopMargin(a) {
            this.p8 = a;
            this.g("YAxisTitleTopMargin");
        }
        get xAxisTitleRightMargin() {
            return this.pq;
        }
        set xAxisTitleRightMargin(a) {
            this.pq = a;
            this.g("XAxisTitleRightMargin");
        }
        get yAxisTitleRightMargin() {
            return this.p7;
        }
        set yAxisTitleRightMargin(a) {
            this.p7 = a;
            this.g("YAxisTitleRightMargin");
        }
        get xAxisTitleBottomMargin() {
            return this.pn;
        }
        set xAxisTitleBottomMargin(a) {
            this.pn = a;
            this.g("XAxisTitleBottomMargin");
        }
        get yAxisTitleBottomMargin() {
            return this.p4;
        }
        set yAxisTitleBottomMargin(a) {
            this.p4 = a;
            this.g("YAxisTitleBottomMargin");
        }
        get xAxisTitleTextColor() {
            return this.rp;
        }
        set xAxisTitleTextColor(a) {
            this.rp = a;
            this.g("XAxisTitleTextColor");
        }
        get yAxisTitleTextColor() {
            return this.r7;
        }
        set yAxisTitleTextColor(a) {
            this.r7 = a;
            this.g("YAxisTitleTextColor");
        }
        get xAxisLabelTextStyle() {
            return this.rf;
        }
        set xAxisLabelTextStyle(a) {
            this.rf = a;
            this.g("XAxisLabelTextStyle");
        }
        get yAxisLabelTextStyle() {
            return this.rx;
        }
        set yAxisLabelTextStyle(a) {
            this.rx = a;
            this.g("YAxisLabelTextStyle");
        }
        get xAxisTitleTextStyle() {
            return this.rq;
        }
        set xAxisTitleTextStyle(a) {
            this.rq = a;
            this.g("XAxisTitleTextStyle");
        }
        get yAxisTitleTextStyle() {
            return this.r8;
        }
        set yAxisTitleTextStyle(a) {
            this.r8 = a;
            this.g("YAxisTitleTextStyle");
        }
        get xAxisLabelRef() {
            return this.rd;
        }
        set xAxisLabelRef(a) {
            this.rd = a;
            this.g("XAxisLabelRef");
        }
        get yAxisLabelRef() {
            return this.rv;
        }
        set yAxisLabelRef(a) {
            this.rv = a;
            this.g("YAxisLabelRef");
        }
        get xAxisMajorStroke() {
            return this.ri;
        }
        set xAxisMajorStroke(a) {
            this.ri = a;
            this.g("XAxisMajorStroke");
        }
        get yAxisMajorStroke() {
            return this.r0;
        }
        set yAxisMajorStroke(a) {
            this.r0 = a;
            this.g("YAxisMajorStroke");
        }
        get xAxisMajorStrokeThickness() {
            return this.ph;
        }
        set xAxisMajorStrokeThickness(a) {
            this.ph = a;
            this.g("XAxisMajorStrokeThickness");
        }
        get yAxisMajorStrokeThickness() {
            return this.py;
        }
        set yAxisMajorStrokeThickness(a) {
            this.py = a;
            this.g("YAxisMajorStrokeThickness");
        }
        get xAxisMinorStrokeThickness() {
            return this.pi;
        }
        set xAxisMinorStrokeThickness(a) {
            this.pi = a;
            this.g("XAxisMinorStrokeThickness");
        }
        get yAxisMinorStrokeThickness() {
            return this.pz;
        }
        set yAxisMinorStrokeThickness(a) {
            this.pz = a;
            this.g("YAxisMinorStrokeThickness");
        }
        get xAxisStrip() {
            return this.rk;
        }
        set xAxisStrip(a) {
            this.rk = a;
            this.g("XAxisStrip");
        }
        get yAxisStrip() {
            return this.r2;
        }
        set yAxisStrip(a) {
            this.r2 = a;
            this.g("YAxisStrip");
        }
        get xAxisStroke() {
            return this.rl;
        }
        set xAxisStroke(a) {
            this.rl = a;
            this.g("XAxisStroke");
        }
        get yAxisStroke() {
            return this.r3;
        }
        set yAxisStroke(a) {
            this.r3 = a;
            this.g("YAxisStroke");
        }
        get xAxisStrokeThickness() {
            return this.pj;
        }
        set xAxisStrokeThickness(a) {
            this.pj = a;
            this.g("XAxisStrokeThickness");
        }
        get yAxisStrokeThickness() {
            return this.p0;
        }
        set yAxisStrokeThickness(a) {
            this.p0 = a;
            this.g("YAxisStrokeThickness");
        }
        get xAxisTickLength() {
            return this.pk;
        }
        set xAxisTickLength(a) {
            this.pk = a;
            this.g("XAxisTickLength");
        }
        get yAxisTickLength() {
            return this.p1;
        }
        set yAxisTickLength(a) {
            this.p1 = a;
            this.g("YAxisTickLength");
        }
        get xAxisTickStroke() {
            return this.rm;
        }
        set xAxisTickStroke(a) {
            this.rm = a;
            this.g("XAxisTickStroke");
        }
        get yAxisTickStroke() {
            return this.r4;
        }
        set yAxisTickStroke(a) {
            this.r4 = a;
            this.g("YAxisTickStroke");
        }
        get xAxisTickStrokeThickness() {
            return this.pl;
        }
        set xAxisTickStrokeThickness(a) {
            this.pl = a;
            this.g("XAxisTickStrokeThickness");
        }
        get yAxisTickStrokeThickness() {
            return this.p2;
        }
        set yAxisTickStrokeThickness(a) {
            this.p2 = a;
            this.g("YAxisTickStrokeThickness");
        }
        get xAxisTitle() {
            return this.rn;
        }
        set xAxisTitle(a) {
            this.rn = a;
            this.g("XAxisTitle");
        }
        get yAxisTitle() {
            return this.r5;
        }
        set yAxisTitle(a) {
            this.r5 = a;
            this.g("YAxisTitle");
        }
        get xAxisMinorStroke() {
            return this.rj;
        }
        set xAxisMinorStroke(a) {
            this.rj = a;
            this.g("XAxisMinorStroke");
        }
        get yAxisMinorStroke() {
            return this.r1;
        }
        set yAxisMinorStroke(a) {
            this.r1 = a;
            this.g("YAxisMinorStroke");
        }
        get xAxisLabelAngle() {
            return this.pc;
        }
        set xAxisLabelAngle(a) {
            this.pc = a;
            this.g("XAxisLabelAngle");
        }
        get yAxisLabelAngle() {
            return this.pt;
        }
        set yAxisLabelAngle(a) {
            this.pt = a;
            this.g("YAxisLabelAngle");
        }
        get xAxisExtent() {
            return this.pb;
        }
        set xAxisExtent(a) {
            this.pb = a;
            this.g("XAxisExtent");
        }
        get yAxisExtent() {
            return this.ps;
        }
        set yAxisExtent(a) {
            this.ps = a;
            this.g("YAxisExtent");
        }
        get xAxisTitleAngle() {
            return this.pm;
        }
        set xAxisTitleAngle(a) {
            this.pm = a;
            this.g("XAxisTitleAngle");
        }
        get yAxisTitleAngle() {
            return this.p3;
        }
        set yAxisTitleAngle(a) {
            this.p3 = a;
            this.g("YAxisTitleAngle");
        }
        get xAxisInverted() {
            return this.o7;
        }
        set xAxisInverted(a) {
            this.o7 = a;
            this.g("XAxisInverted");
        }
        get yAxisInverted() {
            return this.o8;
        }
        set yAxisInverted(a) {
            this.o8 = a;
            this.g("YAxisInverted");
        }
        get xAxisTitleAlignment() {
            return this.ro;
        }
        set xAxisTitleAlignment(a) {
            this.ro = a;
            this.g("XAxisTitleAlignment");
        }
        get yAxisTitleAlignment() {
            return this.r6;
        }
        set yAxisTitleAlignment(a) {
            this.r6 = a;
            this.g("YAxisTitleAlignment");
        }
        get xAxisLabelHorizontalAlignment() {
            return this.rb;
        }
        set xAxisLabelHorizontalAlignment(a) {
            this.rb = a;
            this.g("XAxisLabelHorizontalAlignment");
        }
        get yAxisLabelHorizontalAlignment() {
            return this.rt;
        }
        set yAxisLabelHorizontalAlignment(a) {
            this.rt = a;
            this.g("YAxisLabelHorizontalAlignment");
        }
        get xAxisLabelVerticalAlignment() {
            return this.rg;
        }
        set xAxisLabelVerticalAlignment(a) {
            this.rg = a;
            this.g("XAxisLabelVerticalAlignment");
        }
        get yAxisLabelVerticalAlignment() {
            return this.ry;
        }
        set yAxisLabelVerticalAlignment(a) {
            this.ry = a;
            this.g("YAxisLabelVerticalAlignment");
        }
        get xAxisLabelVisibility() {
            return this.rh;
        }
        set xAxisLabelVisibility(a) {
            this.rh = a;
            this.g("XAxisLabelVisibility");
        }
        get yAxisLabelVisibility() {
            return this.rz;
        }
        set yAxisLabelVisibility(a) {
            this.rz = a;
            this.g("YAxisLabelVisibility");
        }
        get yAxisLabelLocation() {
            return this.ru;
        }
        set yAxisLabelLocation(a) {
            this.ru = a;
            this.g("YAxisLabelLocation");
        }
        get xAxisLabelLocation() {
            return this.rc;
        }
        set xAxisLabelLocation(a) {
            this.rc = a;
            this.g("XAxisLabelLocation");
        }
        get xAxisLabelFormat() {
            return this.ra;
        }
        set xAxisLabelFormat(a) {
            this.ra = a;
            this.g("XAxisLabelFormat");
        }
        get xAxisLabelFormatSpecifiers() {
            return this.o3;
        }
        set xAxisLabelFormatSpecifiers(a) {
            this.o3 = a;
            this.g("XAxisLabelFormatSpecifiers");
        }
        get yAxisLabelFormat() {
            return this.rs;
        }
        set yAxisLabelFormat(a) {
            this.rs = a;
            this.g("YAxisLabelFormat");
        }
        get yAxisLabelFormatSpecifiers() {
            return this.o4;
        }
        set yAxisLabelFormatSpecifiers(a) {
            this.o4 = a;
            this.g("YAxisLabelFormatSpecifiers");
        }
    }
    XYChartDescription.$t = /*@__PURE__*/ markType(XYChartDescription, 'XYChartDescription', DomainChartDescription.$);
    return XYChartDescription;
})();
